---
title: "Medgenome India cross-species data"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---
<!-- comment out for pdf compiling -->
<!-- some css for formatting html -->
<!-- <style>
    h1.title {
        font-size: 40px;
        font-family: Serif;
        text-align: center;
        font-weight: normal;
        /* color: DarkRed; */
    }
    h1 {
        font-size: 24px;
        font-family: Serif;
        font-weight: bold;
        /* font-weight: bold; */
        /* text-align: center; */
        /* color: DarkRed; */
    }
    h2 {
        font-size: 22px;
        font-family: Serif;
        font-weight: bold;
        /* text-align: center; */
        /* color: DarkRed; */
    }
    h3 {
        font-size: 20px;
        font-family: Serif;
        font-weight: bold;
        /* text-align: center; */
        /* color: DarkRed; */
    }
    body .main-container {
        /* max-width: 1000px; */
        font-size: 18px;
        font-family: Serif;
    }
</style> -->

```{css echo=FALSE, warning=FALSE}
/* To make hoverable links. (does not have to be called hint) Usage: */
/* [Message to show on hover]{.hint} */
.hint {
  visibility: hidden;
}

.hint::before {
  visibility: visible;
  content: "Hint";
  color: blue;
}

.hint:hover {
  visibility: visible;
  font-weight: bold;
}

.hint:hover::before {
  display: none;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))
knitr::opts_chunk$set(cache=TRUE)
knitr::opts_chunk$set(echo=FALSE, message=FALSE, error=FALSE, warning=FALSE, results='hide', fig.show='hide')
# loadhistory(".Rhistory")
# load libraries
library(ggplot2)
library(kableExtra)
library(readxl)
library(knitr)
library(tidyverse)
library(viridis)
library(hrbrthemes)
library(ggthemes)
library(RColorBrewer)
library(scales)
library(dplyr)
library(gridExtra)
library(ggVennDiagram)
library(vegan)
library(ape)
library(ggforce)
library(VennDiagram)
library(circlize)
library(phyloseq)
library(ggdendro)
library(dendextend)
library(ComplexHeatmap)

# useful function(s)
make_theme <- function(theme_name=theme_classic() ,max_colors=0, palettefill="Pastel1", palettecolor="Dark2", modify_guide = T,
                        setFill=TRUE, setCol=TRUE,
                        guide_nrow=2, guide_nrow_byrow=TRUE, leg_pos="top", leg_size=12,
                        x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                        y_angle=0 ,y_vj=0, y_hj=0, y_size=12){
  n_11 = c("BrBG", "PiYG", "PRGn", "PuOr", "RdBu", "RdGy", "RdYlBu", "RdYlGn", "Spectral")
  n_12 = c("Paired", "Set3")
  n_8 = c("Accent", "Dark2", "Pastel2", "Set2")
  if (palettefill %in% n_12) {
    n_f = 12
  } else {
    if (palettefill %in% n_11) {
      n_f = 11
    } else {
      if (palettefill %in% n_8) {
        n_f  = 8
      } else {
        n_f = 9
      }
    }
  }
  if (palettecolor %in% n_12) {
    n_c = 12
  } else {
    if (palettecolor %in% n_11) {
      n_c = 11
    } else {
      if (palettecolor %in% n_8) {
        n_c  = 8
      } else {
        n_c = 9
      }
    }
  }
  getFill = colorRampPalette(brewer.pal(n_f, palettefill))
  getColor = colorRampPalette(brewer.pal(n_c, palettecolor))
  theme_params <- theme(axis.text.x = element_text(angle = x_angle,
    vjust = x_vj, hjust=x_hj,
    size = x_size),
    axis.text.y = element_text(angle = y_angle,
      vjust = y_vj, hjust=y_hj,
      size = y_size),
      # axis.title.x = element_text(margin=margin(t=5)),
      # axis.title.y = element_text(margin=margin(r=10)),
      legend.position=leg_pos,
      legend.text = element_text(size=leg_size)
    )
  if (modify_guide == T) {
    guide_params <- guides(fill = guide_legend(
                                    nrow=guide_nrow,
                                    byrow=guide_nrow_byrow
                                  ),
                          col = guide_legend(
                                    nrow=guide_nrow,
                                    byrow=guide_nrow_byrow
                                  )
                    )
  my_theme <- list(
                theme_name,
                theme_params,
                guide_params
              )
  } else {
    my_theme <- list(
                  theme_name,
                  theme_params
                )
  }
  if(setFill) {
    if (n_f < max_colors) {
      my_theme <- list(
                    my_theme,
                    scale_fill_manual(values = getFill(max_colors), na.value="grey")
                  )

    } else {
      my_theme <- list(
                    my_theme,
                    scale_fill_brewer(palette=palettefill, na.value="grey")
                  )
    }
  }
  if(setCol) {
    if (n_c < max_colors) {
      my_theme <- list(
                    my_theme,
                    scale_color_manual(values = getColor(max_colors), na.value="grey")
                  )

    } else {
      my_theme <- list(
                    my_theme,
                    scale_color_brewer(palette=palettecolor, na.value="grey")
                  )
    }
  }
  return(my_theme)
}

get_phylotype <- function(SDP){
  # only works if sdps are written in the right format
  # phylotype_x (eg. firm5_1)
  phy = strsplit(SDP, "_")[[1]][1]
  return(phy)
}
get_host_from_colony <- function(colony_name){
  # only works if sdps are written in the right format
  # phylotype_x (eg. Am_xx)
  host_name = strsplit(colony_name, "_")[[1]][1]
  if (host_name == "Am"){
    return("Apis mellifera")
  }
  if (host_name == "Ac"){
    return("Apis cerana")
  }
  if (host_name == "Ad"){
    return("Apis dorsata")
  }
  if (host_name == "Af"){
    return("Apis florea")
  }
  if (host_name == "Aa"){
    return("Apis andreniformis")
  }
  return(NA)
}

get_sample_name <- function(magname){
  if (startsWith(magname, "MAG_")){
    paste0(head(strsplit(strsplit(magname, "MAG_")[[1]][2], "_")[[1]], -1), collapse="_")
  } else {
    strsplit(magname, "_MAG")[[1]][1]
  }
}

get_host_name <- function(magname){
  if (startsWith(magname, "MAG_")){
    sample_name = strsplit(magname, "MAG_")[[1]][2]
    if (grepl("Dr|Gr", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Am", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Ac", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("Apis dorsata")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("Apis florea")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
  else {
    sample_name = strsplit(magname, "_MAG")[[1]][1]
    if (grepl("Dr|Gr", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Am", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Ac", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("Apis dorsata")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("Apis florea")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
}

get_host_from_sample_name <- function(sample_name){
  if (grepl("Am", sample_name)) {
    return("Apis mellifera")
  }
  if (grepl("Ac", sample_name)) {
    return("Apis cerana")
  }
  if (grepl("M1.|M2.|M3.|M4.|M5.|M6.|M7.|M8.|M9.|DrY|GrY|AmAi|AmIu", sample_name)) {
    return("Apis mellifera")
  }
  if (grepl("C1.|C2.|C3.|C4.|C5.|C6.|C7.|C8.|C9.|AcCh|AcKn", sample_name)) {
    return("Apis cerana")
  }
  if (grepl("D1.|D2.|D3.|D4.|D5.|D6.|D7.|D8.|D9.", sample_name)) {
    return("Apis dorsata")
  }
  if (grepl("F1.|F2.|F3.|F4.|F5.|F6.|F7.|F8.|F9.", sample_name)) {
    return("Apis florea")
  }
  if (grepl("A1.|A2.|A3.|A4.|A5.|A6.|A7.|A8.|A9.", sample_name)) {
    return("Apis andreniformis")
  }
  return(NA)
}

get_location_from_sample_name <- function(sample_name){
  if (grepl("Am|Ac", sample_name)) {
    return("Japan")
  }
  if (grepl("DrY|GrY", sample_name)) {
    return("Switzerland")
  }
  if (grepl("M1.", sample_name)) {
    return("India")
  }
  if (grepl("M2.|M3.|M4.|M5.|M6.|M7.|M8.|M9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("C1.|C2.|C3.", sample_name)) {
    return("India")
  }
  if (grepl("C4.|C5.|C6.|C7.|C8.|C9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("D1.|D2.|D3.", sample_name)) {
    return("India")
  }
  if (grepl("D4.|D5.|D6.|D7.|D8.|D9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("F1.|F2.|F3.", sample_name)) {
    return("India")
  }
  if (grepl("F4.|F5.|F6.|F7.|F8.|F9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("A1.|A2.|A3.|A4.|A5.|A6.|A7.|A8.|A9.", sample_name)) {
    return("Malaysia")
  }
  return(NA)
}

get_origin_name <- function(magname){
  if (startsWith(magname, "MAG_")){
    sample_name = strsplit(magname, "MAG_")[[1]][2]
    if (grepl("Dr|Gr", sample_name)) {
      return("Switzerland, Engel apiary")
    }
    if (grepl("Am", sample_name)) {
      return("Japan")
    }
    if (grepl("Ac", sample_name)) {
      return("Japan")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("India")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("India")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("India")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("India")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
  else {
    sample_name = strsplit(magname, "_MAG")[[1]][1]
    if (grepl("Dr|Gr", sample_name)) {
      return("Switzerland, Engel apiary")
    }
    if (grepl("Am", sample_name)) {
      return("Japan")
    }
    if (grepl("Ac", sample_name)) {
      return("Japan")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("India")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("India")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("India")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("India")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
}
get_only_legend <- function(plot) {
  # get tabular interpretation of plot
  plot_table <- ggplot_gtable(ggplot_build(plot))
  #  Mark only legend in plot
  legend_plot <- which(sapply(plot_table$grobs, function(x) x$name) == "guide-box")
  # extract legend
  legend <- plot_table$grobs[[legend_plot]]
  # return legend
  return(legend)
}
format_genome_name <- function(genome){
  MAG = strsplit(genome, ".fa")[[1]]
  return(MAG)
}
get_sdp <- function(cluster){
  if (is.na(clusters_sdp[cluster])) {
    return(cluster)
  } else{
    return(clusters_sdp[cluster])
  }
}
extend_colors_family <- function(names_vec){
  final_list <- list()
  for (a_name in names_vec) {
    if (a_name %in% names(familyColors)) {
      final_list[a_name] = familyColors[a_name]
    } else {
      final_list[a_name] = "grey"
    }
  }
  return(final_list)
}

extend_colors_genera <- function(names_vec){
  final_list <- list()
  for (a_name in names_vec) {
    if (a_name %in% names(genusColors)) {
      final_list[a_name] = genusColors[a_name]
    } else {
      final_list[a_name] = "grey"
    }
  }
  return(final_list)
}
paginate_save <- function(plot, variable, plot_name, pass_nrow = 5, pass_ncol = 5, position = "top", pass_scales = "fixed"){
  for(i in 1:n_pages(plot)){
  plot +
    facet_wrap_paginate(as.formula(paste("~", variable)), ncol = pass_ncol, nrow = pass_nrow, strip.position = position, scales = pass_scales, page = i)
    ggsave(filename = paste0(strsplit(plot_name, ".pdf")[[1]][1], "_", i, '.pdf'))
  }
}
```

```{r include_readme_text, child='README.md'}
# htmltools::includeMarkdown("README.md")
```

# Results


```{r predefined_vectors}
samples_IN <- c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5",
              "C1.1", "C1.2", "C1.3", "C1.4", "C1.5",
              "C2.1", "C2.2", "C2.3", "C2.4", "C2.5",
              "C3.1", "C3.2", "C3.3", "C3.4", "C3.5",
              "D1.1","D1.2","D1.3","D1.4","D1.5",
              "D2.1","D2.2","D2.3","D2.4","D2.5",
              "D3.1","D3.2","D3.3","D3.4","D3.5",
              "F1.1","F1.2","F1.3","F1.4","F1.5",
              "F2.1","F2.2","F2.3","F2.4","F2.5",
              "F3.1","F3.2","F3.3","F3.4","F3.5"
            )
samples_KE <- c("AmAi01", "AmAi02", "AmAi03", "AmAi04", "AmAi05",
              "AmAi06", "AmAi07", "AmAi08", "AmAi09", "AmAi10",
              "AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05",
              "AmIu06", "AmIu07", "AmIu08", "AmIu09", "AmIu10",
              "AcKn01", "AcKn02", "AcKn03", "AcKn04", "AcKn05",
              "AcKn06", "AcKn07", "AcKn08", "AcKn09", "AcKn10",
              "AcCh01", "AcCh02", "AcCh03", "AcCh04", "AcCh05",
              "AcCh06", "AcCh07", "AcCh08", "AcCh09", "AcCh10",
              "DrY1_F1", "DrY1_F2", "DrY1_F3", "DrY1_F4", "DrY1_F5", "DrY1_F6",
              "DrY1_N1", "DrY1_N2", "DrY1_N3", "DrY1_N4", "DrY1_N5", "DrY1_N6",
              "DrY1_W1", "DrY1_W2", "DrY1_W3", "DrY1_W4", "DrY1_W5", "DrY1_W6",
              "DrY2_F1", "DrY2_F2", "DrY2_F3", "DrY2_F4", "DrY2_F5", "DrY2_F6",
              "DrY2_N1", "DrY2_N2", "DrY2_N3", "DrY2_N4", "DrY2_N5", "DrY2_N6",
              "DrY2_W1", "DrY2_W2", "DrY2_W3", "DrY2_W4", "DrY2_W5", "DrY2_W6",
              "GrY2_F1", "GrY2_F2", "GrY2_F3", "GrY2_F4", "GrY2_F5", "GrY2_F6",
              "GrY2_N1", "GrY2_N2", "GrY2_N3", "GrY2_N4", "GrY2_N5", "GrY2_N6",
              "GrY2_W1", "GrY2_W2", "GrY2_W3", "GrY2_W4", "GrY2_W5", "GrY2_W6"
            )
samples_MY <- c("M2.1", "M2.2", "M2.3", "M2.4", "M2.5",
              "M3.1", "M3.2", "M3.3", "M3.4", "M3.5",
              "M4.1", "M4.2", "M4.3", "M4.4", "M4.5",
              "M5.1", "M5.2", "M5.3", "M5.4", "M5.5",
              "M6.1", "M6.2", "M6.3", "M6.4", "M6.5",
              "M7.1", "M7.2", "M7.3", "M7.4", "M7.5",
              "C4.1", "C4.2", "C4.3", "C4.4", "C4.5",
              "C5.1", "C5.2", "C5.3", "C5.4", "C5.5",
              "C6.1", "C6.2", "C6.3", "C6.4", "C6.5",
              "C7.1", "C7.2", "C7.3", "C7.4", "C7.5",
              "C8.1", "C8.2", "C8.3", "C8.4", "C8.5",
              "C9.1", "C9.2", "C9.3", "C9.4", "C9.5",
              "D4.1","D4.2","D4.3","D4.4","D4.5",
              "D5.1","D5.2","D5.3","D5.4","D5.5",
              "D6.1","D6.2","D6.3","D6.4","D6.5",
              "D7.1","D7.2","D7.3","D7.4","D7.5",
              "D8.1","D8.2","D8.3","D8.4","D8.5",
              "D9.1","D9.2","D9.3","D9.4","D9.5",
              "F4.1","F4.2","F4.3","F4.4","F4.5",
              "F5.1","F5.2","F5.3","F5.4","F5.5",
              "F6.1","F6.2","F6.3","F6.4","F6.5",
              "F7.1","F7.2","F7.3","F7.4","F7.5",
              "F8.1","F8.2","F8.3","F8.4","F8.5",
              "F9.1","F9.2","F9.3","F9.4","F9.5",
              "A1.1","A1.2","A1.3","A1.4","A1.5",
              "A2.1","A2.2","A2.3","A2.4","A2.5",
              "A3.1","A3.2","A3.3","A3.4","A3.5",
              "A4.1","A4.2","A4.3","A4.4","A4.5",
              "A5.1","A5.2","A5.3","A5.4","A5.5",
              "A6.1","A6.2","A6.3","A6.4","A6.5"
            )
samples <- c(samples_IN, samples_KE)
if (samples_KE[1] %in% samples) {
  samples <- c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5",
              "AmAi01", "AmAi02", "AmAi03", "AmAi04", "AmAi05",
              "AmAi06", "AmAi07", "AmAi08", "AmAi09", "AmAi10",
              "AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05",
              "AmIu06", "AmIu07", "AmIu08", "AmIu09", "AmIu10",
              "DrY1_F1", "DrY1_F2", "DrY1_F3", "DrY1_F4", "DrY1_F5", "DrY1_F6",
              "DrY1_N1", "DrY1_N2", "DrY1_N3", "DrY1_N4", "DrY1_N5", "DrY1_N6",
              "DrY1_W1", "DrY1_W2", "DrY1_W3", "DrY1_W4", "DrY1_W5", "DrY1_W6",
              "DrY2_F1", "DrY2_F2", "DrY2_F3", "DrY2_F4", "DrY2_F5", "DrY2_F6",
              "DrY2_N1", "DrY2_N2", "DrY2_N3", "DrY2_N4", "DrY2_N5", "DrY2_N6",
              "DrY2_W1", "DrY2_W2", "DrY2_W3", "DrY2_W4", "DrY2_W5", "DrY2_W6",
              "GrY2_F1", "GrY2_F2", "GrY2_F3", "GrY2_F4", "GrY2_F5", "GrY2_F6",
              "GrY2_N1", "GrY2_N2", "GrY2_N3", "GrY2_N4", "GrY2_N5", "GrY2_N6",
              "GrY2_W1", "GrY2_W2", "GrY2_W3", "GrY2_W4", "GrY2_W5", "GrY2_W6",
              "C1.1", "C1.2", "C1.3", "C1.4", "C1.5",
              "C2.1", "C2.2", "C2.3", "C2.4", "C2.5",
              "C3.1", "C3.2", "C3.3", "C3.4", "C3.5",
              "AcKn01", "AcKn02", "AcKn03", "AcKn04", "AcKn05",
              "AcKn06", "AcKn07", "AcKn08", "AcKn09", "AcKn10",
              "AcCh01", "AcCh02", "AcCh03", "AcCh04", "AcCh05",
              "AcCh06", "AcCh07", "AcCh08", "AcCh09", "AcCh10",
              "D1.1","D1.2","D1.3","D1.4","D1.5",
              "D2.1","D2.2","D2.3","D2.4","D2.5",
              "D3.1","D3.2","D3.3","D3.4","D3.5",
              "F1.1","F1.2","F1.3","F1.4","F1.5",
              "F2.1","F2.2","F2.3","F2.4","F2.5",
              "F3.1","F3.2","F3.3","F3.4","F3.5"
    )
}
colonies <- c("M_1", "M_1", "M_1", "M_1", "M_1",
             "M_DrY2_F","M_DrY2_F","M_Ai","M_Iu",
              "C_1", "C_1", "C_1", "C_1", "C_1",
              "C_2", "C_2", "C_2", "C_2", "C_2",
              "C_3", "C_3", "C_3", "C_3", "C_3",
              "C_Ch","C_Kn",
              "D_1","D_1","D_1","D_1","D_1",
              "D_2","D_2","D_2","D_2","D_2",
              "D_3","D_3","D_3","D_3","D_3",
              "F_1","F_1","F_1","F_1","F_1",
              "F_2","F_2","F_2","F_2","F_2",
              "F_3","F_3","F_3","F_3","F_3")
host_order <- c("Apis mellifera", "Apis cerana", "Apis dorsata", "Apis florea", "Apis andreniformis")
host_order_color <- c("Apis mellifera" = brewer.pal(9, "Pastel1")[2], "Apis cerana" = brewer.pal(9, "Pastel1")[1], "Apis dorsata" = brewer.pal(9, "Pastel1")[4], "Apis florea" = brewer.pal(9, "Pastel1")[3], "Apis andreniformis" = brewer.pal(9, "Pastel1")[5])
host_order_color_dark <- c("Apis mellifera" = brewer.pal(9, "Set1")[2], "Apis cerana" = brewer.pal(9, "Set1")[1], "Apis dorsata" = brewer.pal(9, "Set1")[4], "Apis florea" = brewer.pal(9, "Set1")[3], "Apis andreniformis" = brewer.pal(9, "Set1")[5])
colony_order <- c("M_1", "M_Iu", "M_Ai", "M_DrY2_F", "C_1", "C_2", "C_3", "C_Kn", "C_Ch", "D_1", "D_2", "D_3", "F_1", "F_2", "F_3")
location_order <- c("AIST_Am", "UT_Am", "Bee park, GKVK_Am","Les Droites_Am",
                    "NCBS campus_Ac", "Bee park, GKVK_Ac", "Chiba_Ac", "Kanagawa_Ac",
                    "Biological sciences building, IISc_Ad","House near NCBS_Ad","Naideli hostel_Ad",
                    "Bangalore outskirts_Af")
phylotypes <- c("firm4", "firm5", "api", "bifido", "bom", "com", "bapis", "fper", "lkun", "snod", "gilli")
phylotypes_heatmap_order <- c("snod", "gilli", "firm4", "firm5", "bifido", "bapis", "fper", "api", "lkun", "bom", "com")
sdps <- c('firm4_1', 'firm4_2',
          'firm5_1', 'firm5_2', 'firm5_3', 'firm5_4', 'firm5_7', 'firm5_bombus',
          # 'bifido_1', 'bifido_2', 'bifido_bombus',
          'bifido_1.1', 'bifido_1.2', 'bifido_1.3', 'bifido_1.4', 'bifido_1.5', 'bifido_2', 'bifido_1_cerana', 'bifido_bombus',
          'api_1', 'api_apis_dorsa', 'api_bombus',
          'bom_1', 'bom_apis_melli', 'bom_bombus',
          'com_1', 'com_drosophila', 'com_monarch',
          'bapis',
          'fper_1',
          'lkun',
          'snod_1', 'snod_2', 'snod_bombus',
          'gilli_1', 'gilli_2', 'gilli_3', 'gilli_4', 'gilli_5', 'gilli_6',
          'gilli_apis_andre', 'gilli_apis_dorsa', 'gilli_bombus')
# Data_dir <- "04_CoreCov_211018_Medgenome_india_samples"

#########################################################
# Vector of colors for Phylotypes and SDPs and families
#########################################################
# Family colors
# species <- c('s__Bombilactobacillus mellis', 's__Lactobacillus panisapium', 's__', 's__Gilliamella apicola_E', 's__Bombilactobacillus mellifer', 's__Lactobacillus apis', 's__Snodgrassella alvi', 's__Lactobacillus melliventris', 's__Lactobacillus helsingborgensis', 's__Frischella perrara', 's__Enterobacter hormaechei_A', 's__Apibacter sp002964915', 's__Snodgrassella alvi_E', 's__Frischella japonica', 's__Gilliamella apicola_F', 's__Spiroplasma melliferum', 's__Bartonella apis', 's__Apibacter adventoris', 's__Apilactobacillus kunkeei_A', 's__Hafnia paralvei', 's__Pantoea vagans', 's__Gilliamella apicola_K', 's__Gilliamella apicola', 's__Snodgrassella alvi_G', 's__Bifidobacterium indicum', 's__Gilliamella apicola_N', 's__Klebsiella variicola', 's__Commensalibacter sp003202795', 's__Gilliamella apicola_Q')
# speciesColors
genera <- c(
    "g__Lactobacillus",
    "g__Bifidobacterium",
    "g__Bombilactobacillus",
    "g__Snodgrassella",
    "g__Gilliamella",
    "g__Apibacter",
    "g__Bartonella",
    "g__Frischella",
    "g__Commensalibacter",
    "g__Apilactobacillus",
    "g__Bombella",
    "g__Dysgonomonas",
    "g__Enterobacter",
    "g__Pectinatus",
    "g__Spiroplasma",
    "g__Zymobacter",
    "g__Entomomonas",
    "g__Saezia",
    "g__Parolsenella",
    "g__WRHT01",
    "g__"
  )
phy_group_dict = c("firm4" = "g__Bombilactobacillus",
            "g__Bombilactobacillus_outgroup" = "g__Bombilactobacillus",
            "firm5" = "g__Lactobacillus",
            "lacto" = "g__Lactobacillus",
            "g__Lactobacillus_outgroup" = "g__Lactobacillus",
            "bifido" = "g__Bifidobacterium",
            "g__Bifidobacterium_outgroup" = "g__Bifidobacterium",
            "gilli" = "g__Gilliamella",
            "entero" = "g__Gilliamella",
            "g__Gilliamella_outgroup" = "g__Gilliamella",
            "fper" = "g__Frischella",
            "g__Frischella_outgroup" = "g__Frischella",
            "snod" = "g__Snodgrassella",
            "g__Snodgrassella_outgroup" = "g__Snodgrassella",
            "bapis" = "g__Bartonella",
            "g__Bartonella_outgroup" = "g__Bartonella",
            # "" = "g__Enterobacter",
            "g__Enterobacter_outgroup" = "g__Enterobacter",
            # "" = "g__",
            # "" = "g__Pectinatus",
            "g__Pectinatus_outgroup" = "g__Pectinatus",
            "api" = "g__Apibacter",
            "g__Apibacter_outgroup" = "g__Apibacter",
            # "" = "g__Dysgonomonas",
            "g__Dysgonomonas_outgroup" = "g__Dysgonomonas",
            # "" = "g__Spiroplasma",
            "g__Spiroplasma_outgroup" = "g__Spiroplasma",
            # "" = "g__Zymobacter",
            "g__Zymobacter_outgroup" = "g__Zymobacter",
            # "" = "g__Entomomonas",
            "g__Entomomonas_outgroup" = "g__Entomomonas",
            # "" = "g__Saezia",
            "g__Saezia_outgroup" = "g__Saezia",
            # "" = "g__Parolsenella",
            "g__Parolsenella_outgroup" = "g__Parolsenella",
            # "" = "g__WRHT01",
            "g__WRHT01_outgroup" = "g__WRHT01",
            "com" = "g__Commensalibacter",
            "g__Commensalibacter_outgroup" = "g__Commensalibacter",
            "lkun" = "g__Apilactobacillus",
            "g__Apilactobacillus_outgroup" = "g__Apilactobacillus",
            "bom" = "g__Bombella",
            "g__Bombella_outgroup" = "g__Bombella"
          )
genusColors <- list("g__Bombilactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[1],
                    "g__Lactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[4],
                    "g__Bifidobacterium" = brewer.pal(11, "Spectral")[3],
                    "g__Gilliamella" = brewer.pal(11, "Spectral")[11],
                    "g__Frischella" = brewer.pal(11, "Spectral")[8],
                    "g__Bartonella" = brewer.pal(11, "Spectral")[7],
                    "g__Snodgrassella" = brewer.pal(11, "Spectral")[10],
                    "g__Apibacter" = brewer.pal(11, "Spectral")[4],
                    "g__Commensalibacter" = brewer.pal(11, "Spectral")[6],
                    "g__Bombella" = brewer.pal(11, "Spectral")[5],
                    "g__Apilactobacillus" = brewer.pal(11, "Spectral")[9],
                    "g__Dysgonomonas" = brewer.pal(11, "Spectral")[2],
                    "g__Spiroplasma" = brewer.pal(8, "Set1")[8],
                    "g__WRHT01" = brewer.pal(8, "Dark2")[3],
                    "g__Pectinatus" = brewer.pal(8, "Dark2")[1],
                    "g__Enterobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[1],
                    "g__Zymobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[2],
                    "g__Entomomonas"= head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
                    "g__Saezia" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
                    "g__Parolsenella" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[8],
                    "g__" = "#000000"
)

families <- c("f__Lactobacillaceae", "f__Bifidobacteriaceae", "f__Enterobacteriaceae", "f__Neisseriaceae", "f__Rhizobiaceae_A", "f__Selenomonadaceae", "f__Weeksellaceae", "f__Dysgonomonadaceae", "f__Mycoplasmataceae", "f__Halomonadaceae", "f__Pseudomonadaceae", "f__Burkholderiaceae", "f__Atopobiaceae", "f__Desulfovibrionaceae", "f__Acetobacteraceae", "f__", "f__Streptococcaceae")
familyColors <- list(
  "f__Lactobacillaceae" = brewer.pal(11, "Spectral")[1],
  "f__Bifidobacteriaceae" = brewer.pal(11, "Spectral")[3],
  "f__Enterobacteriaceae" = brewer.pal(11, "Spectral")[11],
  "f__Neisseriaceae" = brewer.pal(11, "Spectral")[10],
  "f__Rhizobiaceae_A" = brewer.pal(11, "Spectral")[7],
  "f__Weeksellaceae" = brewer.pal(11, "Spectral")[4],
  "f__Acetobacteraceae" = brewer.pal(11, "Spectral")[6],
  "f__Dysgonomonadaceae" = brewer.pal(11, "Spectral")[2],
  "f__Mycoplasmataceae" = brewer.pal(8, "Set1")[8],
  "f__Desulfovibrionaceae" = brewer.pal(8, "Dark2")[3],
  "f__Selenomonadaceae" = brewer.pal(8, "Dark2")[1],
  "f__Halomonadaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[2],
  "f__Pseudomonadaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
  "f__Burkholderiaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
  "f__Atopobiaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[8],
  "f__Streptococcaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[9],
  "f__" = "#000000"
)
#
motus <- c("Lactobacillus", "Bifidobacterium", "Gilliamella", "Frischella", "Snodgrassella", "Bartonella", "Bombella", "Acetobacteraceae", "Dysgonomonadaceae", "Spiroplasma", "Flavobacteriaceae", "Fructobacillus", "unassigned")
motuColors <- list(
  "Lactobacillus" = brewer.pal(11, "Spectral")[1],
  "Bifidobacterium" = brewer.pal(11, "Spectral")[3],
  "Flavobacteriaceae" = brewer.pal(11, "Spectral")[4],
  "Bombella" = brewer.pal(11, "Spectral")[5],
  "Acetobacteraceae" = brewer.pal(11, "Spectral")[6],
  "Bartonella" = brewer.pal(11, "Spectral")[7],
  "Frischella" = brewer.pal(11, "Spectral")[8],
  "Gilliamella" = brewer.pal(11, "Spectral")[11],
  "Snodgrassella" = brewer.pal(11, "Spectral")[10],
  "Dysgonomonadaceae" = brewer.pal(11, "Spectral")[2],
  "Fructobacillus" = brewer.pal(11, "Spectral")[9],
  "Spiroplasma" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[2], "#FFFFFF"))(10), -1)[8],
  "unassigned" = "black"
)
#
PhylotypeColors <- brewer.pal(11,"Spectral")
names(PhylotypeColors) <- phylotypes
# each color from the Spectral palatte corresponds to a phylotype
# each SDP of the phylotype gets a color made by colorRampPalette
# it falls in the sange from the color of the phylotype to #FFFFFF (white)
SDPColors <- c()
# 'firm4'
# 'firm4_1''firm4_2':2
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(3), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(3), -1))
# 'firm5'
# 'firm5_1''firm5_2''firm5_3''firm5_4''firm5_7''firm5_bombus':6
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[2], "#FFFFFF"))(7), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[2], "#FFFFFF"))(7), -1))
# 'bifido'
# # 'bifido_1''bifido_2''bifido_bombus':3 - no
# 'bifido_1.1', 'bifido_1.2', 'bifido_1.3', 'bifido_1.4', 'bifido_1.5', 'bifido_2', 'bifido_1_cerana' 'bifido_bombus': 8
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[3], "#FFFFFF"))(9), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[3], "#FFFFFF"))(4), -1))
# 'api'
# 'api_1''api_apis_dorsa''api_bombus':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[4], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[4], "#FFFFFF"))(4), -1))
# 'bom''bom_apis_melli''bom_bombus':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[5], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[5], "#FFFFFF"))(4), -1))
# 'com'
# 'com_1''com_drosophila''com_monarch':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[6], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[6], "#FFFFFF"))(4), -1))
# 'bapis'
# 'bapis':1
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[7], "#FFFFFF"))(2), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[7], "#FFFFFF"))(2), -1))
# 'fper'
# 'fper_1':1
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[8], "#FFFFFF"))(2), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[8], "#FFFFFF"))(2), -1))
# 'lkun'
# 'lkun':1
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[9], "#FFFFFF"))(2), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[9], "#FFFFFF"))(2), -1))
# 'snod'
# 'snod_1''snod_2''snod_bombus':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[10], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[10], "#FFFFFF"))(4), -1))
# 'gilli'
# 'gilli_1''gilli_2''gilli_3''gilli_4''gilli_5''gilli_6''gilli_apis_andre''gilli_apis_dorsa''gilli_bombus':9
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[11], "#FFFFFF"))(10), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[11], "#FFFFFF"))(10), -1))
names(SDPColors) <- sdps
# coverage_df <- data.frame(cbind("ID" = c(),
#                                 "SDP" = c(),
#                                 "Phylotype" = c(),
#                                 "Coverage" = c()
#                               )
#                         )
#
#
# for (phy in phylotypes){
#   df_phy_coord <- read.csv(paste0(Data_dir,"/",phy,"_corecov_coord.txt"), sep = "\t", header = TRUE)
#   df_phy_coord$Sample <- as.character(lapply(df_phy_coord$Sample, function(x) strsplit(x, "_m")[[1]][1]))
#   colnames(df_phy_coord) <- c("SDP", "ID", "Coverage", "PTR")
#   df_phy_coord <- df_phy_coord %>%
#           select(ID, SDP, Coverage) %>%
#             mutate(Phylotype = phy)
#   coverage_df <- bind_rows(coverage_df, df_phy_coord)
# }
Groups <- c("g__Bombilactobacillus",
            "g__Lactobacillus",
            "g__Bifidobacterium",
            "g__Gilliamella",
            "g__Frischella",
            "g__Snodgrassella",
            "g__Bartonella",
            "g__Enterobacter",
            # "g__",
            "g__Pectinatus",
            "g__Apibacter",
            "g__Dysgonomonas",
            "g__Spiroplasma",
            # "g__Zymobacter",
            "g__Entomomonas",
            "g__Saezia",
            "g__Parolsenella",
            "g__WRHT01",
            "g__Commensalibacter",
            "g__Apilactobacillus",
            "g__Bombella")
# path to files of interest - currently path as in curnagl
# to be written and checked manually
qpcr_results <- c(
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate1/plate1-2022-11-17_115607_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate2/plate2-2022-11-17_142003_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate3/plate3-2022-11-17_161529_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate4/plate4-2022-11-17_181034_Results.xls"
  # "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate5/plate5-2023-01-xx_xxxxxx_Results.xls"
  # "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate6/plate6-2023-01-xx_xxxxxx_Results.xls"
  # "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate7/plate7-2023-01-xx_xxxxxx_Results.xls"
)
path_df_meta_complete <- "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/DataCollection/221108_metadata_compiled.xlsx"
df_meta_complete <- read_xlsx(path_df_meta_complete, sheet = "Compiled", range = "A368:AH567", col_names = F)
colnames(df_meta_complete) <- read_xlsx(path_df_meta_complete, sheet = "Compiled", range = "A1:AH1", col_names = F)
df_meta_complete <- df_meta_complete %>%
                      rename(ID_long = ID) %>%
                        rename(ID = Short_name)
workig_dir <- "/scratch/aprasad/221220_Medgenome_india_samples"
```

```{r read_mapping_data, results='hold'}
setwd(working_dir)
# make mapping summary from flagstat tsv
df_reads <- data.frame()
for (sample in samples) {
  if (file.exists(paste0("fastqc/raw/", sample, "_R1_fastqc.zip"))) {
    number_raw_R1 <- read.csv(unz(paste0("fastqc/raw/", sample, "_R1_fastqc.zip"), paste0(sample, "_R1_fastqc/fastqc_data.txt")), sep = "\t") %>%
                filter(.[[1]] == "Total Sequences") %>%
                  pull() %>%
                    as.integer()  
  } else {
    number_raw_R1 <- NA
  }
  if (file.exists(paste0("fastqc/raw/", sample, "_R2_fastqc.zip"))) {
    number_raw_R2 <- read.csv(unz(paste0("fastqc/raw/", sample, "_R2_fastqc.zip"), paste0(sample, "_R2_fastqc/fastqc_data.txt")), sep = "\t") %>%
                filter(.[[1]] == "Total Sequences") %>%
                  pull() %>%
                    as.integer()
  } else {
    number_raw_R2 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"))) {
    number_trimmed_R1 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"), paste0(sample, "_R1_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R1 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"))) {
    number_trimmed_R2 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"), paste0(sample, "_R2_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R2 <- NA
  }
  raw_reads <- number_raw_R1 + number_raw_R2
  trimmed <- number_trimmed_R1 + number_trimmed_R2
  if (file.exists(paste0("02_HostMapping/", sample, "_flagstat.tsv"))) {
    mapped_host_db <- read.csv(paste0("02_HostMapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                      filter(.[[3]] == "with itself and mate mapped") %>%
                        pull(1) %>%
                          as.integer()
  } else {
   mapped_host_db <- NA 
  }
  unmapped_host_db <- trimmed - mapped_host_db
  if (file.exists(paste0("04_MicrobiomeMappingDirect/", sample, "_flagstat.tsv"))) {
    mapped_mic_db <- read.csv(paste0("04_MicrobiomeMappingDirect/", sample, "_flagstat.tsv"), sep = "\t") %>%
                      filter(.[[3]] == "with itself and mate mapped") %>%
                        pull(1) %>%
                          as.integer()
  } else {
   mapped_mic_db <- NA 
  }
  unmapped_mic_db <- trimmed - mapped_mic_db
  if (file.exists(paste0("03_MicrobiomeMapping/", sample, "_flagstat.tsv"))) {
    mapped_mic_db_host_filtered <- read.csv(paste0("03_MicrobiomeMapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                      filter(.[[3]] == "with itself and mate mapped") %>%
                        pull(1) %>%
                          as.integer()
  } else {
   mapped_mic_db_host_filtered <- NA 
  }
  unmapped_sequential <- unmapped_host_db - mapped_mic_db_host_filtered
  values_to_bind <- c(sample, as.integer(c(raw_reads, trimmed, mapped_host_db, unmapped_host_db, mapped_mic_db, unmapped_mic_db, mapped_mic_db_host_filtered, unmapped_sequential)))
  df_reads <- rbind(df_reads, values_to_bind)
}
# parse this later
# read.csv(paste0("02_HostMapping/", sample, "_coverage.tsv"), sep = "\t")
df_colnames <- c("Sample", "Raw", "Trimmed", "Mapped_host", "Unmapped_host", "Mapped_microbiome", "Unmapped_microbiome", "Mapped_microbiome_filtered", "Unmapped_filtered")
colnames(df_reads) <- df_colnames
df_reads <- df_reads %>%
              mutate(across(!c("Sample"), as.integer))
df_meta <- read.csv("config/Metadata_211018_Medgenome_india_samples.csv", sep = ',')
colnames(df_meta)[which(colnames(df_meta) == "ID")] <- "Sample"
df_meta$SpeciesID <- recode(df_meta$SpeciesID, "Am" = "Apis mellifera", "Ac" = "Apis cerana", "Af" = "Apis florea", "Ad" = "Apis dorsata")
df_meta <- df_meta %>%
            filter(Sample %in% samples) %>%
              arrange(match(Sample, samples))
df_meta %>% group_by(SpeciesID) %>% tally()
df_meta %>% group_by(SpeciesID, Country) %>% tally()
df_meta %>% group_by(SpeciesID, Country, Colony) %>% tally()
df_plot_reads <- pivot_longer(df_reads, !Sample, values_to = "Number", names_to = "Type") %>%
                  merge(select(df_meta, Sample, SpeciesID), by="Sample")
df_plot_reads$Number <- as.integer(df_plot_reads$Number)
df_meta_complete
```

## Summary of DNA extraction results

```{r DNA_concentrations}
df_concentrations <- df_meta_complete %>%
  select(ID, Concentration) %>%
    left_join(df_meta, by = c("ID" = "Sample")) %>%
      mutate(Host = Vectorize(get_host_from_sample_name)(ID)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(ID))

ggplot(df_concentrations %>% filter(ID %in% c(samples_IN, samples_MY)), 
        aes(y = factor(ID, rev(c(samples_IN, samples_MY))),
            x = Concentration,
            # fill = factor(Host, host_order),
            shape = Location,
            color = factor(Host, host_order)
          )
        ) +
  geom_point() +
  facet_wrap(~Host, scales = "free") +
  geom_hline(yintercept = 6.5, linetype = "solid") +
   geom_hline(yintercept = 14.5, linetype = "solid") +
   geom_hline(yintercept = 22.5, linetype = "solid") +
   geom_hline(yintercept = 30.5, linetype = "solid") +
    labs(color = "Host species", y = "Sample name", x = "Concentration (ng/uL) - Aliquot of 10 uL") +
    make_theme(setFill = F, setCol = F, guide_nrow = 3) +
    geom_vline(xintercept = 0, linetype = "solid", alpha = 0.5) +
    geom_vline(xintercept = 1, linetype = "solid", color = "#1a9850", alpha = 1) +
    geom_vline(xintercept = 5, linetype = "solid", color = "#fee08b", alpha = 1) +
    geom_vline(xintercept = 10, linetype = "solid", color = "#d73027", alpha = 1) +
      scale_fill_manual(values=host_order_color) +
      scale_color_manual(values=host_order_color_dark)
      ggsave("Figures/00-DNA_Concentrations_compared.pdf")

# write.csv(df_concentrations %>% filter(ID %in% samples_MY) %>% select(ID, Concentration, Host), "Figures/Concentrations_for_plotting_edited_manually.csv", row.names = F)
df_concentrations_edited <- read.csv("Figures/Concentrations_for_plotting_edited_manually.csv")
ggplot(df_concentrations_edited %>% filter(ID %in% samples_MY) %>% mutate(Host_name = Vectorize(get_host_from_sample_name)(ID)), 
        aes(y = factor(ID, rev(samples_MY)),
            x = Concentration,
            # fill = factor(Host, host_order),
            color = factor(Host_name, host_order)
          )
        ) +
  geom_point() +
  facet_wrap(~Host, scales = "free") +
  geom_hline(yintercept = 6.5, linetype = "solid") +
   geom_hline(yintercept = 14.5, linetype = "solid") +
   geom_hline(yintercept = 22.5, linetype = "solid") +
   geom_hline(yintercept = 30.5, linetype = "solid") +
    labs(color = "Host species", y = "Sample name", x = "Concentration (ng/uL) - Aliquot of 10 uL") +
    make_theme(setFill = F, setCol = F, guide_nrow = 3) +
    geom_vline(xintercept = 0, linetype = "solid", alpha = 0.5) +
    geom_vline(xintercept = 1, linetype = "solid", color = "#d73027", alpha = 1) +
    geom_vline(xintercept = 5, linetype = "solid", color = "#fee08b", alpha = 1) +
    geom_vline(xintercept = 10, linetype = "solid", color = "#1a9850", alpha = 1) +
      scale_fill_manual(values=host_order_color) +
      scale_color_manual(values=host_order_color_dark)
      ggsave("Figures/00-DNA_Concentrations_with_subset.pdf")

ggplot(df_concentrations %>% filter(ID %in% samples_MY),
        aes(y = factor(ID, rev(samples_MY)),
            x = Concentration,
            # fill = factor(Host, host_order),
            color = factor(Host, host_order)
          )
        ) +
  geom_point() +
  facet_wrap(~Host, scales = "free") +
  geom_hline(yintercept = 6.5, linetype = "solid") +
   geom_hline(yintercept = 14.5, linetype = "solid") +
   geom_hline(yintercept = 22.5, linetype = "solid") +
   geom_hline(yintercept = 30.5, linetype = "solid") +
    labs(color = "Host species", y = "Sample name", x = "Concentration (ng/uL) - Aliquot of 10 uL") +
    make_theme(setFill = F, setCol = F, guide_nrow = 3) +
    geom_vline(xintercept = 0, linetype = "solid", alpha = 0.5) +
    geom_vline(xintercept = 1, linetype = "solid", color = "#d73027", alpha = 1) +
    geom_vline(xintercept = 5, linetype = "solid", color = "#fee08b", alpha = 1) +
    geom_vline(xintercept = 10, linetype = "solid", color = "#1a9850", alpha = 1) +
      scale_fill_manual(values=host_order_color) +
      scale_color_manual(values=host_order_color_dark)
      ggsave("Figures/00-DNA_Concentrations.pdf")
```

## Total number of reads before and after trimming

```{r plots_total_species}
Total_reads <- ggplot(filter(df_plot_reads, Type %in% c("Raw", "Trimmed")), aes(x=factor(Sample, levels = samples), y=Number, fill = Type)) +
                        geom_bar(stat="identity", position = "dodge") +
                          # ggtitle("Total reads per sample") +
                            xlab("Sample") +
                            ylab("Number of reads (paired end)") +
                            geom_vline(xintercept = 5.5, linetype="solid") +
                            geom_vline(xintercept = 20.5, linetype="solid") +
                            geom_vline(xintercept = 35.5, linetype="solid") +
                            geom_hline(yintercept = 30e+6, linetype="dotted") +
                            geom_hline(yintercept = 50e+6, linetype="dotted") +
                              make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 1,
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=7,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
                                scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6))
                                  ggsave("Figures/00a-Total_reads_trimming.pdf")
```

```{r show_plots_total_species, dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'Total reads by species'}
Total_reads
# knitr::include_graphics("Figures/Total_reads_species.pdf")
```

## Total number of reads per species after trimming

```{r plots_total_trimmed}
Total_reads_species <- ggplot(filter(df_plot_reads, Type %in% c("Trimmed")), aes(x=factor(Sample, levels = samples), y=Number, fill = SpeciesID)) +
                        geom_bar(stat="identity", position = "dodge") +
                          # ggtitle("Total reads per sample") +
                            ylab("Sample") +
                            xlab("Number of reads (paired end)") +
                            geom_vline(xintercept = 5.5, linetype="solid") +
                            geom_vline(xintercept = 20.5, linetype="solid") +
                            geom_vline(xintercept = 35.5, linetype="solid") +
                            geom_hline(yintercept = 30e+6, linetype="dotted") +
                            make_theme(theme_name=theme_few(), leg_pos="bottom",
                                       guide_nrow = 1,
                                       x_angle=45 ,x_vj=1.2, x_hj=1, x_size=7,
                                       y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
                                scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                scale_fill_manual(values=host_order_color)
                                  ggsave("Figures/00b-Total_reads_species_after_trimming.pdf")
```

```{r show_plots_total_trimmed, dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'Total reads by species'}
Total_reads_species
# knitr::include_graphics("Figures/Total_reads_species.pdf")
```

## Proportion of reads mapping to host and microbiome database

```{r plots_num_mapped_mic}
df_plot_reads_prop <- select(df_reads, Sample, Trimmed, Mapped_host, Mapped_microbiome_filtered, Unmapped_filtered) %>%
  summarise(Sample, "None" = Unmapped_filtered/Trimmed*100,
            "Host" = Mapped_host/Trimmed*100,
            "Microbiome" = Mapped_microbiome_filtered/Trimmed*100) %>%
              pivot_longer(!Sample, names_to = "Type", values_to = "Number")
Mapped_Unmapped_reads_prop <- ggplot(df_plot_reads_prop, aes(x=factor(Sample, levels = samples),
                                                            y=Number,
                                                            fill=factor(Type, levels = c("None", "Host", "Microbiome")))) +
                        geom_bar(stat="identity", position = "stack") +
                          # ggtitle("Total reads per sample") +
                            labs(y = "Sample",
                                 x = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                            geom_vline(xintercept = 5.5, linetype="solid") +
                            geom_vline(xintercept = 20.5, linetype="solid") +
                            geom_vline(xintercept = 35.5, linetype="solid") +
                            make_theme(theme_name=theme_few(), leg_pos="bottom",
                                       guide_nrow = 1,
                                       x_angle=45 ,x_vj=1.2, x_hj=1, x_size=7,
                                       y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                                      ggsave("Figures/00c-Mapped_Unmapped_reads_prop.pdf")
```

```{r show_plots_num_mapped_mic, dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'Proportion of reads mapped to the microbiome database and host database'}
Mapped_Unmapped_reads_prop
# knitr::include_graphics("Figures/Mapped_Unmapped_reads_prop.pdf")
```

## Number of reads mapping to host and microbiome database

```{r plots_num_mapped_mic}
df_plot_reads_num <- select(df_reads, Sample, Trimmed, Mapped_host, Mapped_microbiome_filtered, Unmapped_filtered) %>%
  summarise(Sample, "None" = Unmapped_filtered,
            "Host" = Mapped_host,
            "Microbiome" = Mapped_microbiome_filtered) %>%
              pivot_longer(!Sample, names_to = "Type", values_to = "Number")
Mapped_Unmapped_reads_num <- ggplot(df_plot_reads_num, aes(x=factor(Sample, levels = samples),
                                                            y=Number,
                                                            fill=factor(Type, levels = c("None", "Host", "Microbiome")))) +
                        geom_bar(stat="identity", position = "stack") +
                          # ggtitle("Total reads per sample") +
                            labs(x = "Sample",
                                 y = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                            geom_vline(xintercept = 5.5, linetype="solid") +
                            geom_vline(xintercept = 20.5, linetype="solid") +
                            geom_vline(xintercept = 35.5, linetype="solid") +
                            geom_hline(yintercept = 10e+6, linetype="dotted") +
                            geom_hline(yintercept = 1e+6, linetype="dotted") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                            make_theme(theme_name=theme_few(), leg_pos="bottom",
                                       guide_nrow = 1,
                                       x_angle=45 ,x_vj=1.2, x_hj=1, x_size=7,
                                       y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                                      ggsave("Figures/00e-Mapped_Unmapped_reads_num.pdf")
```

```{r show_plots_num_mapped_mic, dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'Proportion of reads mapped to the microbiome database and host database'}
Mapped_Unmapped_reads_num
# knitr::include_graphics("Figures/Mapped_Unmapped_reads_prop.pdf")
```

## Micorbiome reads lost by non_specific mapping to host

```{r plots_prop_mapped_mic}
df_plot_reads_non_specific <- df_reads %>%
  mutate(Unmapped_microbiome_filtered = Mapped_host + Unmapped_filtered) %>%
    select(Sample, Mapped_microbiome, Unmapped_microbiome, Mapped_microbiome_filtered, Unmapped_microbiome_filtered) %>%
              pivot_longer(!Sample, names_to = "Type", values_to = "Number") %>%
                mutate(Host_filtered = ifelse(Type %in% c("Mapped_microbiome", "Unmapped_microbiome"), "Direct mapping", "Host filtered")) %>%
                  mutate(Type = ifelse(Type %in% c("Mapped_microbiome", "Mapped_microbiome_filtered"), "Mapped", "Unmapped"))


Non_specific_reads <- ggplot(df_plot_reads_non_specific %>% filter(Type == "Mapped"),
                             aes(x=factor(Sample, samples),
                                 y=Number,
                                 fill=Host_filtered)) +
                        geom_bar(stat="identity", position = "dodge") +
                            labs(x = "Sample",
                                 y = "Number of reads (paired end)",
                                 fill = "Mapping strategy") +
                            geom_vline(xintercept = 5.5, linetype="solid") +
                            geom_vline(xintercept = 20.5, linetype="solid") +
                            geom_vline(xintercept = 35.5, linetype="solid") +
                            geom_hline(yintercept = 10e+6, linetype="dotted") +
                            geom_hline(yintercept = 1e+6, linetype="dotted") +
                              scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 1, setFill = F,
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=7,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
                                scale_fill_manual(values=brewer.pal(4, "Paired")[c(3,4)])
                                      ggsave("Figures/00d-Mapping_non_specific.pdf")

                                      # ggsave("Figures/00d-Mapping_non_specific.pdf")
```

```{r show_plots_prop_mapped_mic, dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'Proportion of reads mapped to the microbiome database and host database'}
Non_specific_reads
```

## Community profiling with mOTUs

```{r motus_summary}
df_motus_raw <- read.csv("02_motus_profile/samples_merged.motus", sep = "\t", skip = 2, stringsAsFactors=FALSE)
colnames(df_motus_raw)[1] <- "taxonomy"
df_motus_raw <- df_motus_raw %>%
              mutate(across(!c("taxonomy"), as.numeric)) %>%
              mutate(sum_ab = rowSums(across(where(is.numeric)))) %>%
                filter(sum_ab > 0) %>%
                  select(!sum_ab) %>%
                    column_to_rownames("taxonomy")

df_motus <- as.data.frame(t(df_motus_raw)) %>%
              rownames_to_column("Sample") %>%
              pivot_longer(!Sample, names_to = "motu", values_to = "rel_ab") %>%
                group_by(Sample, motu) %>%
                  mutate(Present = ifelse(rel_ab > 0, 1, 0)) %>%
                    group_by(motu) %>%
                     mutate(Prevalence_num = sum(Present)) %>%
                      mutate(Prevalence = mean(Present))

condense_motu <- function(motu){
  motu_condensed = strsplit(motu, " ")[[1]][1]
  return(motu_condensed)
}

df_motus_combined <- left_join(df_motus, df_meta)  %>%
                      group_by(SpeciesID, motu) %>%
                        mutate(Present_host = ifelse(rel_ab > 0, 1, 0)) %>%
                          group_by(SpeciesID, motu) %>%
                            mutate(Prevalence_num_host = sum(Present_host)) %>%
                            mutate(Prevalence_host = mean(Present_host)) %>%
                              mutate(mean_rel_ab_host = mean(rel_ab)) %>%
                              group_by(Sample) %>%
                              mutate(mean_rel_ab = mean(rel_ab)) %>%
                                mutate(motu_condensed = Vectorize(condense_motu)(motu))

motu_list <- df_motus_combined %>%
              pull(motu) %>% unique

high_motu_list <- df_motus_combined %>%
                        filter(Prevalence_host >= 0.5) %>%
                          pull(motu_condensed) %>% unique

df_assigned_plot <- df_motus_combined %>%
                      group_by(Sample) %>%
                        mutate(sum = sum(rel_ab)) %>%
                          mutate(unassigned = rel_ab[motu == "unassigned"]) %>%
                            mutate("assigned" = sum - unassigned) %>%
                              select(Sample, assigned, unassigned) %>%
                                pivot_longer(!Sample, values_to = "proportion", names_to = "Type")

assigned_plot <- ggplot(df_assigned_plot, aes(y = factor(Sample, samples), x = proportion, fill = factor(Type, levels = c("unassigned", "assigned")))) +
                  geom_bar(position = "stack", stat = "identity") +
                  labs(fill = "Type", x = "Proportion", y = "Sample") +
                    make_theme(leg_pos = "right", guide_nrow = 20, leg_size = 7,
                               y_size = 8, y_hj = 1, y_vj = 0.5
                    )
                    # ggsave("Figures/04-motus_unassigned.pdf")

extend_colors_motus <- function(names_vec){
  final_list <- list()
  for (a_name in names_vec) {
    if (a_name %in% names(motuColors)) {
      final_list[a_name] = motuColors[a_name]
    } else {
      final_list[a_name] = "grey"
    }
  }
  return(final_list)
}

plot_motus_high_prev <- ggplot(df_motus_combined, aes(x = factor(Sample, samples), y = rel_ab, fill = factor(motu_condensed, motus))) +
                geom_bar(position = "stack", stat = "identity") +
                labs(fill = "mOTU", x = "mOTUs2 Relative abundance", y = "Sample") +
                  make_theme(max_colors = length(unique(motus)),
                             setFill = F,
                             leg_pos = "bottom",
                             guide_nrow = 3,
                             leg_size = 8,
                             x_angle=45 ,x_vj=1.2, x_hj=1, x_size=7,
                             y_angle=0 ,y_vj=0, y_hj=0, y_size=12
                  ) +
                  scale_fill_manual(values=motuColors)
                    ggsave("Figures/04a-motus.pdf")

df_motus_filt <- filter(df_motus_combined, Prevalence_num >= 1 & rel_ab > 0.01)
plot_motus_filt <- ggplot(df_motus_filt, aes(y = factor(Sample, samples), x = rel_ab, fill = factor(motu_condensed))) +
                geom_bar(position = "stack", stat = "identity") +
                labs(fill = "mOTU", x = "mOTUs2 Relative abundance", y = "Sample") +
                  make_theme(max_colors = length(unique(df_motus_filt$motu_condensed)),
                             palettefill = "Set1",
                             leg_pos = "right",
                             guide_nrow = 20,
                             leg_size = 7,
                             y_size = 8, y_hj = 1,
                  )
                  ggsave("Figures/04b-motus_filt.pdf")

ggplot(df_motus_combined %>% filter(rel_ab > 0.01), aes(y = factor(Sample, samples), x = rel_ab, fill = factor(motu))) +
                geom_bar(position = "stack", stat = "identity") +
                labs(fill = "mOTU", x = "mOTUs2 Relative abundance", y = "Sample") +
                  make_theme(max_colors = length(unique(df_motus_combined$motu)),
                             leg_pos = "none",
                             leg_size = 7,
                             y_size = 8, y_hj = 1
                  )
ggsave("Figures/04c-motus_all.pdf")
grid.arrange(get_only_legend(ggplot(df_motus_combined  %>% filter(rel_ab > 0.01), aes(y = factor(Sample, samples), x = rel_ab, fill = factor(motu))) +
                geom_bar(position = "stack", stat = "identity") +
                labs(fill = "mOTU", x = "mOTUs2 Relative abundance", y = "Sample") +
                  make_theme(max_colors = length(unique(df_motus_combined$motu)),
                             leg_size = 7,
                             guide_nrow = 40,
                             y_size = 8, y_hj = 1
                  ) +
                  theme(legend.key=element_blank(), legend.key.size=unit(5,"point"))
                )
              )
ggsave("Figures/04c-motus_all_legend.pdf")
```

```{r motus_summary_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'mOTUs Summary'}
# g <- grid.arrange(arrangeGrob(assigned_plot, ncol = 2, widths = c(4, 2)), plot_motus_high_prev, nrow = 2, heights = c(1.2, 2))
#   ggsave("Figures/04a-motus.pdf", g)
plot_motus_filt
plot_motus_high_prev
```

# MAG analysis

```{r more_imports}
# setwd("'/Users/aiswarya/OneDrive - Université de Lausanne/Work/2-2-Projects/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/211018_Medgenome_india_samples'")
df_gtdbk_bac <- read.csv("06_MAG_binning/gtdbtk_out_dir/classify/gtdbtk.bac120.summary.tsv", sep = "\t")
drep_Bdb <- read.csv("06_MAG_binning/drep_results/data_tables/Bdb.csv", sep = ",")

drep_gInformation <- read.csv("06_MAG_binning/drep_results/data_tables/genomeInfo.csv", sep = ",")
df_assembly <- read.csv("05_Assembly/MapToAssembly/Assembly_mapping_summary.csv", sep = ',')
df_assembly <- merge(df_assembly, df_meta, by="Sample")
df_assembly$SpeciesID <- recode(df_assembly$SpeciesID, "Am" = "Apis mellifera", "Ac" = "Apis cerana", "Af" = "Apis florea", "Ad" = "Apis dorsata")
drep_Sdb <- read.csv("06_MAG_binning/drep_results/data_tables/Sdb.csv", sep = ",")
drep_Widb <- read.csv("06_MAG_binning/drep_results/data_tables/Widb.csv", sep = ",")
cluster_info <- drep_Widb %>%
                  select(genome, score, cluster, cluster_members, furthest_cluster_member) %>%
                    rename(score_representative = score) %>%
                    mutate(representative_genome = Vectorize(format_genome_name)(genome)) %>%
                      select(!genome)
MAGs_collated <- read.csv("06_MAG_binning/all_GenomeInfo_auto.tsv", sep = "\t")

# make sure that number per cluster is 1 for all
MAGs_collated_info <- left_join(MAGs_collated, mutate(drep_gInformation, genome = Vectorize(format_genome_name)(genome)), by = c("ID"="genome"))

# Number of MAGs coming from each host species
vis_magOTUs_df_all <- MAGs_collated_info %>%
                        group_by(Sample) %>%
                            mutate(all_quality = ifelse(completeness > 50 & contamination < 5 & N50 > 10000, "Pass", "Fail")) %>%
                            mutate(Completeness_quality = cut(completeness ,breaks=c(-1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 100, 110),
                                                            labels = c("<10", "10-20", "20-30", "30-40", "40-50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-100", "100"))
                                                          ) %>%
                              mutate(Contamination_quality = cut(contamination ,breaks=c(-1, 0, 5, 10, 100),
                                                              labels = c("0", "0-5", "5-10", ">10"))
                                                          ) %>%
                              mutate(N50_quality = cut(N50 ,breaks=c(0, 10000, 50000, 100000, 200000, 500000, 1000000, 2000000, Inf), labels = c("<10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", "1-2Mb", ">2Mb"))) %>%
                                mutate(sample = Vectorize(get_sample_name)(ID)) %>%
                                filter(grepl("MAG_", ID)) %>%
                                  mutate(Host = Vectorize(get_host_name)(ID)) # %>%
                                    # mutate(num_contigs = Vectorize(get_num_contigs_per_bin)(ID))

samples_am <- c(vis_magOTUs_df_all %>% filter(Host == "Apis mellifera") %>% pull(Sample) %>% unique %>% as.vector)
samples_ac <- c(vis_magOTUs_df_all %>% filter(Host == "Apis cerana") %>% pull(Sample) %>% unique %>% as.vector)
samples_ad <-c(vis_magOTUs_df_all %>% filter(Host == "Apis dorsata") %>% pull(Sample) %>% unique %>% as.vector)
samples_af <-c(vis_magOTUs_df_all %>% filter(Host == "Apis florea") %>% pull(Sample) %>% unique %>% as.vector)

contig_fates_df_am <- data.frame()
for (sample in samples_am) {
  contig_fates_df_am <- rbind(contig_fates_df_am, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
}
contig_fates_df_am <- cbind(contig_fates_df_am, host = rep("Apis mellifera", dim(contig_fates_df_am)[[1]]))
contig_fates_df_am_pf <- contig_fates_df_am %>%
                              group_by(sample, passed_filter) %>%
                                summarise(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
contig_fates_df_am_bin <- contig_fates_df_am %>%
                              group_by(sample, binned) %>%
                                summarise(binned_length = sum(length), binned_num = n(), .groups = "keep")


contig_fates_df_ac <- data.frame()
for (sample in samples_ac) {
  contig_fates_df_ac <- rbind(contig_fates_df_ac, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
}
contig_fates_df_ac <- cbind(contig_fates_df_ac, host = rep("Apis cerana", dim(contig_fates_df_ac)[[1]]))
contig_fates_df_ac_pf <- contig_fates_df_ac %>%
                              group_by(sample, passed_filter) %>%
                                summarise(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
contig_fates_df_ac_bin <- contig_fates_df_ac %>%
                              group_by(sample, binned) %>%
                                summarise(binned_length = sum(length), binned_num = n(), .groups = "keep")

contig_fates_df_ad <- data.frame()
for (sample in samples_ad) {
  contig_fates_df_ad <- rbind(contig_fates_df_ad, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
}
contig_fates_df_ad <- cbind(contig_fates_df_ad, host = rep("Apis dorsata", dim(contig_fates_df_ad)[[1]]))
contig_fates_df_ad_pf <- contig_fates_df_ad %>%
                              group_by(sample, passed_filter) %>%
                                summarise(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
contig_fates_df_ad_bin <- contig_fates_df_ad %>%
                              group_by(sample, binned) %>%
                                summarise(binned_length = sum(length), binned_num = n(), .groups = "keep")

contig_fates_df_af <- data.frame()
for (sample in samples_af) {
  contig_fates_df_af <- rbind(contig_fates_df_af, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
}
contig_fates_df_af <- cbind(contig_fates_df_af, host = rep("Apis florea", dim(contig_fates_df_af)[[1]]))
contig_fates_df_af_pf <- contig_fates_df_af %>%
                              group_by(sample, passed_filter) %>%
                                summarise(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
contig_fates_df_af_bin <- contig_fates_df_af %>%
                              group_by(sample, binned) %>%
                                summarise(binned_length = sum(length), binned_num = n(), .groups = "keep")

contigs_length_df <- rbind(
                        summarise(group_by(contig_fates_df_am, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop"),
                        summarise(group_by(contig_fates_df_ac, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop"),
                        summarise(group_by(contig_fates_df_ad, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop"),
                        summarise(group_by(contig_fates_df_af, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop")
                  )

length_bin_sum_df <- contigs_length_df %>%
                    mutate(length_bin = cut(length ,breaks=c(0, 500, 1000, 10000, 50000, 100000, 200000, 500000, 1000000, Inf), labels = c("500", "0.5-1kb", "1-10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", ">1Mb"))) %>%
                      group_by(sample, binned, length_bin) %>%
                        summarise(length_bin_sum = sum(length), num_contigs = n())
length_bin_sum_df_passed <- contigs_length_df %>%
                    filter(passed == "P") %>%
                    mutate(length_bin = cut(length ,breaks=c(0, 500, 1000, 10000, 50000, 100000, 200000, 500000, 1000000, Inf), labels = c("500", "0.5-1kb", "1-10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", ">1Mb"))) %>%
                      group_by(sample, binned, length_bin) %>%
                        summarise(length_bin_sum = sum(length), num_contigs = n())

all_depths <- data.frame()
for (sample in samples) {
  sample_contig_depths <- read.csv(paste0("06_MAG_binning/backmapping/",sample,"/",sample,"_mapped_to_",sample,".depth"), sep = "\t")
  depth_column <- paste0(sample,"_mapped_to_",sample,".bam")
  all_depths <- rbind(all_depths, rename(select(sample_contig_depths, contigName, all_of(depth_column)), depth = depth_column))
}
contigs_depths_df <- left_join(filter(contigs_length_df, passed == "P"), all_depths, by = c("contig" = "contigName"))
rm(all_depths)
vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(Cluster) %>%
                            mutate(Num_mags = n()) %>%
                              mutate(Prevalence_overall = Num_mags/length(samples))

samples_am <- c(vis_magOTUs_df_all %>% filter(Host == "Apis mellifera") %>% pull(Sample) %>% unique %>% as.vector)
samples_ac <- c(vis_magOTUs_df_all %>% filter(Host == "Apis cerana") %>% pull(Sample) %>% unique %>% as.vector)
samples_ad <-c(vis_magOTUs_df_all %>% filter(Host == "Apis dorsata") %>% pull(Sample) %>% unique %>% as.vector)
samples_af <-c(vis_magOTUs_df_all %>% filter(Host == "Apis florea") %>% pull(Sample) %>% unique %>% as.vector)

vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(Cluster, Host) %>%
                            mutate(Present = n()) %>%
                            mutate(Prevalence = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence)) %>%
                              left_join(summarise(group_by(contigs_depths_df, bin), mean_coverage = mean(depth), .groups = "drop"), by = c("ID" = "bin")) %>%
                                arrange(Genus)

vis_magOTUs_df_all <- left_join(vis_magOTUs_df_all, cluster_info, by = c("Cluster" = "cluster")) %>%
                        left_join(drep_Sdb %>%
                                    mutate(ID = Vectorize(format_genome_name)(genome)) %>%
                                    select(!genome)
                        )

vis_magOTUs_df <- vis_magOTUs_df_all %>%
                    filter(all_quality == "Pass")


vis_magOTUs_df$Host <- as.factor(recode(vis_magOTUs_df$Host, Am="Apis mellifera", Ac="Apis cerana", Ad="Apis dorsata", Af="Apis florea"))
vis_magOTUs_df$Cluster <- as.factor(vis_magOTUs_df$Cluster)
vis_magOTUs_df$sample <- as.factor(vis_magOTUs_df$sample)
vis_magOTUs_df$Family <- as.factor(vis_magOTUs_df$Family)
vis_magOTUs_df$Genus <- as.factor(vis_magOTUs_df$Genus)
vis_magOTUs_df <- vis_magOTUs_df[order(vis_magOTUs_df$Genus), ]
```

## Mapping to Assembly

The host-filtered reads were assembled. The proportion of reads that mapped back to the assembly and some information about the assemblies such as assembly size, and information about contigs are shown below.

```{r Assembly_mapping_summary}
assembly_sizes <- ggplot(df_assembly, aes(y = factor(Sample, levels=samples),
                                          x = Assembly.size,
                                          fill = factor(SpeciesID, host_order)
                                        )
                                      ) +
                    geom_bar(stat = "identity") +
                    geom_vline(xintercept = 2000000) +
                      labs(x = "Total Assembly Size", y = "Sample", fill = "Host Species") +
                        scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                          make_theme(setFill = F,
                                     leg_size = 6,
                                     x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                     y_angle=0 ,y_vj=0, y_hj=1, y_size=7,
                                     guide_nrow = 1) +
                            scale_fill_manual(values=host_order_color)
                          # ggsave("Figures/03-AssemblySizes.pdf")

df_assembly_plot <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    rename(Unmapped = Number.unmapped, Mapped = Number.mapped) %>%
     pivot_longer(cols = 3:4, names_to ="Type", values_to = "Number")

number_mapped_assembly <- ggplot(df_assembly, aes(y = factor(Sample, levels=samples), x = Number.mapped, fill = factor(SpeciesID, host_order))) +
                            geom_bar(stat = "identity") +
                            labs(x = "Number of reads mapped to assembly", y = "Sample", fill = "Type") +
                            scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(setFill = F,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=1, y_size=7,
                                         guide_nrow = 1) +
                                scale_fill_manual(values=host_order_color)
                              # ggsave("Figures/03-NumberAssembled.pdf")

df_assembly_plot <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    mutate(percent_mapped = Number.mapped/Number.of.reads*100) %>%
    mutate(percent_unmapped = Number.unmapped/Number.of.reads*100) %>%
    rename(Unmapped = percent_unmapped, Mapped = percent_mapped, Unmapped_number = Number.unmapped, Mapped_number = Number.mapped) %>%
    select(Sample, Mapped, Unmapped) %>%
     pivot_longer(cols = 2:3, names_to ="Type", values_to = "Percent")

df_assembly_plot_number <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    mutate(percent_mapped = Number.mapped/Number.of.reads*100) %>%
    mutate(percent_unmapped = Number.unmapped/Number.of.reads*100) %>%
    rename(Unmapped = Number.unmapped, Mapped = Number.mapped) %>%
    select(Sample, Mapped, Unmapped) %>%
     pivot_longer(cols = 2:3, names_to ="Type", values_to = "Number")

number_mapped_assembly <- ggplot(df_assembly_plot_number, aes(y = factor(Sample, levels=samples), x = Number, fill = factor(Type, levels = c("Unmapped", "Mapped")))) +
                            geom_bar(stat = "identity") +
                            labs(x = "Number of reads (paired end)", y = "Sample", fill = "Type") +
                            scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(x_angle = 60, x_hj = 1, x_vj = 1, max_colors = length(unique(df_assembly_plot$Type)), x_size = 7, guide_nrow = 1)
                              # ggsave("Figures/03-NumberMappedtoAssembly.pdf")

Number_contig_plot <- ggplot(df_assembly, aes(y = factor(Sample, levels=samples), x = Number.of.filtered.scaffolds, fill = factor(SpeciesID, levels = rev(host_order)))) +
                            geom_bar(stat = "identity") +
                            labs(y = "Sample", x = "Number of scaffolds", fill = "Host") +
                            scale_x_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            scale_fill_manual(values=host_order_color) +
                              make_theme(setFill = F,
                                         leg_size = 6,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=1, y_size=7,
                                         guide_nrow = 1)
                              ggsave("Figures/03a-NumberOfFilteredcontigs.pdf")

percent_mapped_assembly <- ggplot(df_assembly_plot, aes(y = factor(Sample, levels=samples), x = Percent, fill = factor(Type, levels = c("Unmapped", "Mapped")))) +
                            geom_bar(stat = "identity") +
                            labs(y = "Percentage of reads", x = "Sample", fill = "Type") +
                              make_theme(x_angle = 60, x_hj = 1, x_vj = 1, max_colors = length(unique(df_assembly_plot$Type)), x_size = 7, guide_nrow = 1)
                              # ggsave("Figures/03-ProportionMappedtoAssembly.pdf")
Total_data_species <- ggplot(mutate(filter(df_plot_reads, Type == "Raw" ), amount_data = Number*2*150), aes(y=factor(Sample, levels = samples),
                                      x=amount_data,
                                      fill = SpeciesID)) +
                        geom_bar(stat="identity") +
                            ylab("Sample") +
                            xlab("Number of reads (paired end)") +
                              make_theme(setFill = F,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=1, y_size=7,
                                         guide_nrow = 1, leg_pos = "none") +
                                scale_x_continuous(labels=unit_format(unit = "G", scale = 1e-9)) +
                                scale_fill_manual(values=host_order_color)
```

```{r Assembly_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
Total_reads_species_temp <- ggplot(filter(df_plot_reads, Type == "Raw"), aes(y=factor(Sample, levels = samples),
                                      x=Number,
                                      fill = SpeciesID)) +
                        geom_bar(stat="identity") +
                          # ggtitle("Total reads per sample") +
                            ylab("Sample") +
                            xlab("Number of reads (paired end)") +
                              make_theme(setFill = F,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=1, y_size=7,
                                         guide_nrow = 1) +
                                scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                scale_fill_manual(values=host_order_color)
g <- grid.arrange(
      arrangeGrob(
        assembly_sizes + make_theme(setFill = F, setCol = F, leg_size = 6, y_size = 5, y_hj = 1, y_vj = 1, x_size = 8, guide_nrow = 1, leg_pos = "none"),
        number_mapped_assembly + make_theme(setFill = F, setCol = F, leg_size = 6, y_size = 5, y_hj = 1, y_vj = 1, x_size = 8, guide_nrow = 1, leg_pos = "none"),
        nrow = 1
      ),
      get_only_legend(assembly_sizes),
      arrangeGrob(
        Total_data_species + make_theme(setFill = F, setCol = F, leg_size = 6, y_size = 5, y_hj = 1, y_vj = 1, x_size = 8, guide_nrow = 2, leg_pos = "none"),
        Total_reads_species_temp + make_theme(setFill = F, setCol = F, leg_size = 6, y_size = 5, y_hj = 1, y_vj = 1, x_size = 8, guide_nrow = 1, leg_pos = "none"),
        nrow = 1
      ),
      nrow = 3, heights = c(8,1,8)
 )
 g
 ggsave("Figures/03b-Assembly_summary.pdf", g)
Number_contig_plot
```

## contig length and fate (binning outcome)

```{r contig_binning_summary}
contig_fates_df_pf <- rbind(contig_fates_df_am_pf,
                            contig_fates_df_ac_pf,
                            contig_fates_df_ad_pf,
                            contig_fates_df_af_pf
                      )
amount_pass_fail <- ggplot(contig_fates_df_pf, aes(y = factor(sample, samples), x = pass_fail_length, fill = passed_filter)) +
  geom_bar(stat = "identity") +
  labs(y= "Sample", x = "Amount of data passed or failed") +
  # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  make_theme(palettefill = "Set1", leg_pos = "bottom", guide_nrow = 1)

contig_fates_df_bin <- rbind(contig_fates_df_am_bin,
                            contig_fates_df_ac_bin,
                            contig_fates_df_ad_bin,
                            contig_fates_df_af_bin
                      )
contigs_binned_length_plot <- ggplot(contig_fates_df_bin, aes(y = sample, x = binned_length, fill = binned)) +
  geom_bar(stat = "identity") +
    labs(x = "Sample", y = "Amount of data binned or unbinned") +
    scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
      make_theme(palettefill = "Set1", guide_nrow = 1)


contig_fates_df_am_mag <- contig_fates_df_am %>%
                            # filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
contig_fates_df_ac_mag <- contig_fates_df_ac %>%
                            # filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
contig_fates_df_ad_mag <- contig_fates_df_ad %>%
                            # filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
contig_fates_df_af_mag <- contig_fates_df_af %>%
                            # filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
contigs_binned_length_am_plot_genus <- ggplot(contig_fates_df_am_mag, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05e-contigs_binned_unbinned_by_genus_am.pdf")
contigs_binned_length_am_plot_genus <- ggplot(contig_fates_df_am_mag, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      ggsave("Figures/05-contigs_binned_unbinned_by_genus_am.pdf")

contigs_binned_length_ac_plot_genus <- ggplot(contig_fates_df_ac_mag, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_ac.pdf")

contigs_binned_length_ad_plot_genus <- ggplot(contig_fates_df_ad_mag, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_ad.pdf")

contigs_binned_length_af_plot_genus <- ggplot(contig_fates_df_af_mag, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_af.pdf")

contig_fates_df_am_mag_binned <- contig_fates_df_am %>%
                            filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
rm(contig_fates_df_am)
contig_fates_df_ac_mag_binned <- contig_fates_df_ac %>%
                            filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
rm(contig_fates_df_ac)
contig_fates_df_ad_mag_binned <- contig_fates_df_ad %>%
                            filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
rm(contig_fates_df_ad)
contig_fates_df_af_mag_binned <- contig_fates_df_af %>%
                            filter(binned == "Y") %>%
                            group_by(sample, bin_name) %>%
                              summarise(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
                                left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
rm(contig_fates_df_af)
contigs_binned_length_am_plot_genus_binned <- ggplot(contig_fates_df_am_mag_binned, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_am.pdf")

contigs_binned_length_ac_plot_genus_binned <- ggplot(contig_fates_df_ac_mag_binned, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_ac.pdf")

contigs_binned_length_ad_plot_genus_binned <- ggplot(contig_fates_df_ad_mag_binned, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_ad.pdf")

contigs_binned_length_af_plot_genus_binned <- ggplot(contig_fates_df_af_mag_binned, aes(x = sample, y = len_contigs, fill = factor(Genus, genera))) +
                                  geom_bar(stat = "identity") +
                                    labs(x = "Sample", y = "Sum of length of scaffolds in bin", fill = "Genus") +
                                    # scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)) +
                                    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                                      make_theme(setFill = F,
                                          leg_pos = "right", guide_nrow = 22,
                                          x_angle = 30, x_hj = 1, x_vj = 1
                                        ) +
                                      scale_fill_manual(values=genusColors)
                                      # ggsave("Figures/05-contigs_binned_unbinned_by_genus_af.pdf")

contig_length_host_plot_am <- ggplot(filter(length_bin_sum_df, sample %in% samples_am), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05c-contig_length_histogram_am.pdf")
contig_length_host_plot_am_passed <- ggplot(filter(length_bin_sum_df_passed, sample %in% samples_am), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05d-contig_length_histogram_am_passed.pdf")
contig_length_host_plot_ac <- ggplot(filter(length_bin_sum_df, sample %in% samples_ac), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05c-contig_length_histogram_ac.pdf")
contig_length_host_plot_ac_passed <- ggplot(filter(length_bin_sum_df_passed, sample %in% samples_ac), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05d-contig_length_histogram_ac_passed.pdf")
contig_length_host_plot_ad <- ggplot(filter(length_bin_sum_df, sample %in% samples_ad), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05c-contig_length_histogram_ad.pdf")
contig_length_host_plot_ad_passed <- ggplot(filter(length_bin_sum_df_passed, sample %in% samples_ad), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05d-contig_length_histogram_ad_passed.pdf")
contig_length_host_plot_af <- ggplot(filter(length_bin_sum_df, sample %in% samples_af), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05c-contig_length_histogram_af.pdf")
contig_length_host_plot_af_passed <- ggplot(filter(length_bin_sum_df_passed, sample %in% samples_af), aes(x = length_bin, y = length_bin_sum, fill = binned)) +
                      geom_bar(stat = "identity") +
                        geom_text(aes(label = num_contigs), angle = 0, size = 1, vjust = 1) +
                        # geom_text(aes(label = num_contigs), angle = 90, size = 2) +
                          labs(x = "length of contig", y = "Total bases from contigs in bin", fill = "binned") +
                            scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              facet_wrap(~ sample, scales = "free") +
                                make_theme(x_angle = 40, x_size = 7, x_hj = 1, x_vj = 1, leg_pos = "none")
                                  ggsave("Figures/05d-contig_length_histogram_af_passed.pdf")

contigs_depths_df_genus <- left_join(contigs_depths_df, select(vis_magOTUs_df_all, ID, Genus, Cluster), by = c("bin" = "ID")) %>%
  ungroup() %>%
  select(!Host) %>%
  left_join(rename(select(df_meta, Sample, SpeciesID), Host = SpeciesID), by = c("sample" = "Sample"))


temp <- ggplot(filter(contigs_depths_df_genus, Host == "Apis mellifera"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              guide_nrow = 7, leg_size = 12,
                              x_size = 5, x_angle = 30, x_hj = 1, x_vj = 1
                            ) +
                            scale_color_manual(values=genusColors)+
                              guides(size = "none", alpha = "none")
genus_legend <- get_only_legend(temp)
remove(temp)

contig_len_vs_depth_am <- ggplot(filter(contigs_depths_df_genus, Host == "Apis mellifera"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_am), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05g-length_vs_depth_contigs_all_am.pdf")
contig_len_vs_depth_am_binned <- ggplot(filter(contigs_depths_df_genus, Host == "Apis mellifera" & binned != "N"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_am), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05h-length_vs_depth_contigs_binned_am.pdf")
contig_len_vs_depth_ac <- ggplot(filter(contigs_depths_df_genus, Host == "Apis cerana"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_ac), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05g-length_vs_depth_contigs_all_ac.pdf")
contig_len_vs_depth_ac_binned <- ggplot(filter(contigs_depths_df_genus, Host == "Apis cerana" & binned != "N"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_ac), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05h-length_vs_depth_contigs_binned_ac.pdf")
contig_len_vs_depth_ad <- ggplot(filter(contigs_depths_df_genus, Host == "Apis dorsata"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_ad), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05g-length_vs_depth_contigs_all_ad.pdf")
contig_len_vs_depth_ad_binned <- ggplot(filter(contigs_depths_df_genus, Host == "Apis dorsata" & binned != "N"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_ad), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05h-length_vs_depth_contigs_binned_ad.pdf")
contig_len_vs_depth_af <- ggplot(filter(contigs_depths_df_genus, Host == "Apis florea"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_af), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05g-length_vs_depth_contigs_all_af.pdf")
contig_len_vs_depth_af_binned <- ggplot(filter(contigs_depths_df_genus, Host == "Apis florea" & binned != "N"), aes(x = bin, y = depth, color = Genus, size = length, alpha = 0.5)) +
                          geom_point() +
                            make_theme(setFill = F, setCol = F,
                              leg_pos = "bottom",
                              x_size = 5, x_angle = 50, x_hj = 1, x_vj = 1
                            ) +
                            scale_size_continuous(labels=unit_format(unit = "K", scale = 1e-4)) +
                            theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                            scale_color_manual(values=genusColors) +
                            scale_y_continuous(trans="log10")+
                            facet_wrap(~ factor(sample, samples_af), scales="free") +
                              guides(color = "none", alpha = "none")
                              ggsave("Figures/05h-length_vs_depth_contigs_binned_af.pdf")
```

```{r contig_binning_summary_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
amount_pass_fail
  ggsave("Figures/05a-data-contigs_passed_failed.pdf", amount_pass_fail)
contigs_binned_length_plot
  ggsave("Figures/05b-contigs_binned_unbinned.pdf", contigs_binned_length_plot)
binned_unbinned_contigs_length_genus <- grid.arrange(
  arrangeGrob(
    contigs_binned_length_am_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    contigs_binned_length_ac_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    nrow = 1
  ),
  arrangeGrob(
    contigs_binned_length_ad_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    contigs_binned_length_af_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    nrow = 1
  ),
  get_only_legend(contigs_binned_length_am_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "bottom", guide_nrow = 6, leg_size = 7)),
  nrow = 3, heights = c(2,2,1)
)
  ggsave("Figures/05e-contigs_binned_unbinned_by_genus.pdf", binned_unbinned_contigs_length_genus)

binned_unbinned_contigs_length_genus_scaled <- grid.arrange(
  arrangeGrob(
    contigs_binned_length_am_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)),
    contigs_binned_length_ac_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)),
    nrow = 1
  ),
  arrangeGrob(
    contigs_binned_length_ad_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)),
    contigs_binned_length_af_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 6e+8)),
    nrow = 1
  ),
  get_only_legend(contigs_binned_length_am_plot_genus + make_theme(setFill = F, setCol = F, leg_pos = "bottom", guide_nrow = 6, leg_size = 7)),
  nrow = 3, heights = c(2,2,1)
)
  ggsave("Figures/05e-contigs_binned_unbinned_by_genus_scaled.pdf", binned_unbinned_contigs_length_genus_scaled)
binned_unbinned_contigs_length_genus_binned <- grid.arrange(
  arrangeGrob(
    contigs_binned_length_am_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    contigs_binned_length_ac_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    nrow = 1
  ),
  arrangeGrob(
    contigs_binned_length_ad_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    contigs_binned_length_af_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)),
    nrow = 1
  ),
  get_only_legend(contigs_binned_length_am_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "bottom", guide_nrow = 6, leg_size = 7)),
  nrow = 3, heights = c(2,2,1)
)
  ggsave("Figures/05f-contigs_by_genus_binned.pdf", binned_unbinned_contigs_length_genus_binned)

binned_unbinned_contigs_length_genus_scaled_binned <- grid.arrange(
  arrangeGrob(
    contigs_binned_length_am_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 1.5e+8)),
    contigs_binned_length_ac_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 1.5e+8)),
    nrow = 1
  ),
  arrangeGrob(
    contigs_binned_length_ad_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 1.5e+8)),
    contigs_binned_length_af_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "none", x_size = 7, x_angle = 30, x_hj = 1, x_vj = 1) + scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), limits = c(0, 1.5e+8)),
    nrow = 1
  ),
  get_only_legend(contigs_binned_length_am_plot_genus_binned + make_theme(setFill = F, setCol = F, leg_pos = "bottom", guide_nrow = 6, leg_size = 7)),
  nrow = 3, heights = c(2,2,1)
)
  ggsave("Figures/05f-contigs_by_genus_binned_scaled.pdf", binned_unbinned_contigs_length_genus_scaled_binned)
contig_length_host_plot_am
contig_length_host_plot_am_passed
contig_length_host_plot_ac
contig_length_host_plot_ac_passed
contig_length_host_plot_ad
contig_length_host_plot_ad_passed
contig_length_host_plot_af
contig_length_host_plot_af_passed
contig_len_vs_depth_am
contig_len_vs_depth_am_binned
contig_len_vs_depth_ac
contig_len_vs_depth_ac_binned
contig_len_vs_depth_ad
contig_len_vs_depth_ad_binned
contig_len_vs_depth_af
contig_len_vs_depth_af_binned
```

## MAGs prevalence and abundance

Mean of mean contig coverage (calculated for metabat2) as a proxy for abundance of MAG in it's own sample.

Total MAGs recovered - distribution of quality / total + per sample

```{r visualize_coverage}
vis_magOTUs_df_all_means <- vis_magOTUs_df_all %>%
                        group_by(Sample) %>%
                          summarise(Sample, completeness, contamination, N50, Host) %>%
                            mutate(Completeness_mean = mean(completeness)) %>%
                              mutate(Contamination_mean = mean(contamination)) %>%
                              mutate(N50_mean = mean(N50))
Completeness_hist <- ggplot(vis_magOTUs_df_all, aes(x = completeness, fill = Host)) +
    geom_histogram(binwidth=2) +
    geom_vline(xintercept = 50) +
      make_theme(palettefill="Spectral")
N50_hist <- ggplot(vis_magOTUs_df_all, aes(x = N50, fill = Host)) +
    geom_histogram(bins = 150) +
      scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
      geom_vline(xintercept = 10000) +
      make_theme(palettefill="Spectral")
Contamination_hist <- ggplot(vis_magOTUs_df_all, aes(x = contamination, fill = Host)) +
    geom_histogram(binwidth=2) +
    geom_vline(xintercept = 5) +
      make_theme(palettefill="Spectral")
completeness_per_sample <- ggplot(vis_magOTUs_df_all, aes(y = factor(Sample, levels = samples), fill = Completeness_quality)) +
    geom_bar(position = "stack") +
    labs(fill = "Quality", y = "Sample") +
      make_theme(palettefill="RdYlGn", max_colors = length(levels(vis_magOTUs_df_all$Completeness_quality)))
N50_per_sample <- ggplot(vis_magOTUs_df_all, aes(y = factor(Sample, levels = samples), fill = N50_quality)) +
        geom_bar(position = "stack") +
        labs(fill = "Quality", y = "Sample") +
        make_theme(palettefill="RdYlBu", max_colors = length(levels(vis_magOTUs_df_all$N50_quality)))
contamination_per_sample <- ggplot(vis_magOTUs_df_all, aes(y = factor(Sample, levels = samples), fill = Contamination_quality)) +
    geom_bar(position = "stack") +
    labs(fill = "Quality", y = "Sample") +
      make_theme(palettefill="RdYlBu", max_colors = length(levels(vis_magOTUs_df_all$Contamination_quality)))
MAG_quality_per_sample <- ggplot(vis_magOTUs_df_all, aes(y = factor(Sample, levels = samples), fill = all_quality)) +
    geom_bar(position = "stack") +
    labs(fill = "Quality", y = "Sample") +
      make_theme(palettefill="Set1",)

prev_vs_abud_all <- ggplot(vis_magOTUs_df_all, aes(x = mean_coverage, y = Prevalence, size = completeness, color = Genus, alpha = 0.5)) +
                    geom_point(position = position_jitter(w = 0, h = 0.05)) +
                      make_theme(setFill = F, setCol = F,
                        leg_pos = "bottom",
                        guide_nrow = 8,
                        leg_size = 12
                      ) +
                      theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                      scale_color_manual(values=genusColors) +
                      scale_x_continuous(trans = "log10") +
                        scale_alpha(guide = "none")

prev_vs_abud <- ggplot(vis_magOTUs_df, aes(x = mean_coverage, y = Prevalence, size = completeness, color = Genus, alpha = 0.5)) +
  geom_point(position = position_jitter(w = 0, h = 0.05)) +
    make_theme(setFill = F, setCol = F,
      leg_pos = "bottom",
      guide_nrow = 8,
      leg_size = 12
    ) +
    theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
    scale_color_manual(values=genusColors) +
    scale_x_continuous(trans = "log10") +
    # facet_wrap(~ factor(Host, host_order)) +
      scale_alpha(guide = "none")

prev_overall_vs_abud_all <- ggplot(vis_magOTUs_df_all, aes(x = mean_coverage, y = Prevalence_overall, size = completeness, color = Genus, alpha = 0.5)) +
                    geom_point(position = position_jitter(w = 0, h = 0.05)) +
                      make_theme(setFill = F, setCol = F,
                        leg_pos = "bottom",
                        guide_nrow = 8,
                        leg_size = 12
                      ) +
                      theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
                      scale_color_manual(values=genusColors) +
                      scale_x_continuous(trans = "log10") +
                        scale_alpha(guide = "none")

prev_overall_vs_abud <- ggplot(vis_magOTUs_df, aes(x = mean_coverage, y = Prevalence_overall, size = completeness, color = Genus, alpha = 0.5)) +
  geom_point(position = position_jitter(w = 0, h = 0.05)) +
    make_theme(setFill = F, setCol = F,
      leg_pos = "bottom",
      guide_nrow = 8,
      leg_size = 12
    ) +
    theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
    scale_color_manual(values=genusColors) +
    scale_x_continuous(trans = "log10") +
    # facet_wrap(~ factor(Host, host_order)) +
      scale_alpha(guide = "none")

prev_vs_abud_all_host <- ggplot(vis_magOTUs_df_all, aes(x = mean_coverage, y = Prevalence, size = completeness, color = Genus, alpha = 0.5)) +
  geom_point(position = position_jitter(w = 0, h = 0.05)) +
    make_theme(setFill = F, setCol = F,
      leg_pos = "none",
      guide_nrow = 8,
      leg_size = 12
    ) +
    ggtitle("MAGs with > 70% completeness and < 5% contamination") +
    theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
    scale_color_manual(values=genusColors) +
    scale_x_continuous(trans = "log10") +
    facet_wrap(~ factor(Host, host_order)) +
      scale_alpha(guide = "none")
      ggsave("Figures/07a-prev_vs_coverage_all_MAGs_genus_by_host.pdf")

prev_vs_abud_host <- ggplot(vis_magOTUs_df, aes(x = mean_coverage, y = Prevalence, size = completeness, color = Genus, alpha = 0.5)) +
  geom_point(position = position_jitter(w = 0, h = 0.05)) +
    make_theme(setFill = F, setCol = F,
      leg_pos = "none",
      guide_nrow = 8,
      leg_size = 12
    ) +
    ggtitle("MAGs with > 70% completeness and < 5% contamination") +
    theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
    scale_color_manual(values=genusColors) +
    scale_x_continuous(trans = "log10") +
    facet_wrap(~ factor(Host, host_order)) +
      scale_alpha(guide = "none")
      ggsave("Figures/07a-prev_vs_coverage_filtered_MAGs_genus_by_host.pdf")

prev_vs_abud_all_host <- ggplot(vis_magOTUs_df_all, aes(x = mean_coverage, y = Prevalence, color = Genus)) +
  geom_point(position = position_jitter(w = 0, h = 0.05)) +
    make_theme(setFill = F, setCol = F,
      leg_pos = "none",
      guide_nrow = 8,
      leg_size = 12
    ) +
    ggtitle("MAGs with > 70% completeness and < 5% contamination") +
    theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
    scale_color_manual(values=genusColors) +
    scale_x_continuous(trans = "log10") +
    facet_wrap(~ factor(Host, host_order)) +
      scale_alpha(guide = "none")
      ggsave("Figures/07a-prev_vs_coverage_all_MAGs_genus_by_host_no_size.pdf")

prev_vs_abud_host <- ggplot(vis_magOTUs_df, aes(x = mean_coverage, y = Prevalence, color = Genus)) +
  geom_point(position = position_jitter(w = 0, h = 0.05)) +
    make_theme(setFill = F, setCol = F,
      leg_pos = "none",
      guide_nrow = 8,
      leg_size = 12
    ) +
    ggtitle("MAGs with > 70% completeness and < 5% contamination") +
    theme(legend.margin=margin(-1,-1,-1,-1), legend.box="vertical") +
    scale_color_manual(values=genusColors) +
    scale_x_continuous(trans = "log10") +
    facet_wrap(~ factor(Host, host_order)) +
      scale_alpha(guide = "none")
      ggsave("Figures/07a-prev_vs_coverage_filtered_MAGs_genus_by_host_no_size.pdf")
```

```{r visualize_coverage_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
legend_hist_host <- get_only_legend(Completeness_hist)
g  <- arrangeGrob(
  arrangeGrob(
      Completeness_hist + make_theme(setFill = F, setCol = F, leg_pos = "none"),
      Contamination_hist + make_theme(setFill = F, setCol = F, leg_pos = "none"),
      N50_hist + make_theme(setFill = F, setCol = F, leg_pos = "none"),
      nrow = 2,
      layout_matrix = rbind(c(1,2), c(3,3))
    ),
    legend_hist_host,
    heights = c(10, 1)
  )
  ggsave("Figures/06a-QC_MAG_histogram.pdf", g)
g <- grid.arrange(
    MAG_quality_per_sample + make_theme(setFill = F, setCol = F, leg_size = 7, y_size = 5, leg_pos = "right", guide_nrow = 2),
    N50_per_sample + make_theme(setFill = F, setCol = F, leg_size = 7, y_size = 5, leg_pos = "right", guide_nrow = 7),
    contamination_per_sample + make_theme(setFill = F, setCol = F, leg_size = 7, y_size = 5, leg_pos = "right", guide_nrow = 4),
    completeness_per_sample + make_theme(setFill = F, setCol = F, leg_size = 7, y_size = 5, leg_pos = "right", guide_nrow = 10)
  )
  ggsave("Figures/06b-QC_MAG_per_sample.pdf", g)
g <- grid.arrange(prev_vs_abud_all + make_theme(setFill = F, setCol = F, leg_pos = "none"),
             prev_vs_abud + make_theme(setFill = F, setCol = F, leg_pos = "none") + ggtitle("MAGs with > 70% completeness and < 5% contamination"),
             genus_legend,
             heights = c(3,3,2)
           )
    ggsave("Figures/07a-prev_vs_coverage_MAGs_genus.pdf", g)
g <- grid.arrange(prev_overall_vs_abud_all + make_theme(setFill = F, setCol = F, leg_pos = "none"),
             prev_overall_vs_abud + make_theme(setFill = F, setCol = F, leg_pos = "none") + ggtitle("MAGs with > 70% completeness and < 5% contamination"),
             genus_legend,
             heights = c(3,3,2)
           )
    ggsave("Figures/07a-prev_overall_vs_coverage_MAGs_genus.pdf", g)
vis_magOTUs_df %>%
  group_by(Host) %>%
  # filter(mean_coverage > 0) %>%
    mutate(N_MAGs = n_distinct(ID)) %>%
    mutate(N_magOTUs = n_distinct(Cluster)) %>%
    mutate(N_Genera = n_distinct(Genus)) %>%
      summarise(N_MAGs, N_magOTUs, N_Genera) %>%
        unique()
```

There are a total of `r dim(MAGs_collated)[[1]]` MAGs present. There were `r dim(filter(vis_magOTUs_df_all, completeness > 95 & contamination < 5))[[1]]` high quality MAGs (> 95% complete and < 5% redundant) and `r dim(filter(vis_magOTUs_df_all, completeness > 50 & contamination < 10))[[1]]` with > 50% completion and < 10% redundancy.

## MAG taxonomy affiliation per sample


```{r mag_genus_summary}
genus_MAG_quality_host_all <- ggplot(vis_magOTUs_df_all, aes(y = Genus, fill = Host)) +
        geom_bar(position = "stack") +
        labs(fill = "Host", y = "Genus", x = "Number of MAGs") +
        make_theme(palettefill = "Spectral")
      ggsave("Figures/07-QC_per_Genus_per_host_all.pdf")

genus_MAG_quality_host <- ggplot(filter(vis_magOTUs_df_all, all_quality=="Pass"), aes(y = Genus, fill = Host)) +
        geom_bar(position = "stack") +
        labs(fill = "Host", y = "Genus", x = "Number of MAGs") +
        make_theme(palettefill = "Spectral")
      ggsave("Figures/07-QC_per_Genus_per_host_passed.pdf")
```

```{r mags_genus_plot,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, fig.cap = 'MAG quality summarised'}
genus_MAG_quality_host_all
genus_MAG_quality_host
```

The common threshold (thumb-rule) of >70% completeness and contamination <5% along with N50 >10Kb (N50 threshold does not exclude any genomes) seems appropriate as it does not exclude too many MAGs. Completeness is currently the most powerful criterion.

There are a total of `r vis_magOTUs_df_all %>% pull(Cluster) %>% unique %>% length` clusters. They represent `r vis_magOTUs_df_all %>% pull(Family) %>% unique %>% length` families. Only `r vis_magOTUs_df_all %>% filter(all_quality == "Pass") %>% pull(Cluster) %>% unique %>% length` of them are represented by MAGs that cross the threshold comprising `r vis_magOTUs_df_all %>% filter(all_quality == "Pass") %>% pull(Family) %>% unique %>% length` families.


The passed MAGs include `r filter(vis_magOTUs_df_all, all_quality=="Pass") %>% pull(Genus) %>% unique %>% length` <!-- 26  --> out of `r vis_magOTUs_df_all %>% pull(Genus) %>% unique %>% length` <!-- 34 --> genera.
I exclude the following because they only contain 1 MAG: `r filter(vis_magOTUs_df_all, all_quality=="Pass") %>% group_by(Genus) %>% summarise(number = n()) %>% filter(number <2)`
<!-- g__Pantoea, g__JAATFO01, g__Hafnia, g__Floricoccus, g__Klebsiella -->
`r filter(vis_magOTUs_df_all, all_quality=="Pass") %>% pull(Species) %>% unique %>% length` <!-- 29 --> out of `r vis_magOTUs_df_all %>% pull(Species) %>% unique %>% length` <!-- 51 --> species.

```{r magOTU_summary}
line_list <- c()
for (num in 1:94){
  add_line <- geom_vline(xintercept=num+0.5, size=0.1, color="black")
  # add_line <- geom_vline(xintercept=num+0.5, size=0.1, alpha=0.5, color="grey")
  line_list <- c(line_list, add_line)
}

magOTUs_per_sample <- ggplot(vis_magOTUs_df_all, aes(y = factor(Cluster), x = factor(sample, samples), fill = Host)) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster")+
                              make_theme(setFill=F,
                              # make_theme(palettefill="Spectral", max_colors = length(unique(vis_magOTUs_df$Cluster)),
                              leg_pos="none", guide_nrow=6,
                              y_hj=1, y_size=7, leg_size=8, y_vj=0.5,
                              x_vj=0, x_hj=1, x_size=6, x_angle=90) +
                              scale_fill_manual(values=host_order_color) +
                              line_list
                                    ggsave("Figures/08a-magOTUs_per_sample.pdf")

magOTUs_per_sample_genus <- ggplot(vis_magOTUs_df_all, aes(y = factor(Cluster), x = factor(sample, samples), fill = factor(Genus, genera))) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster")+
                              make_theme(setFill=F,
                              # make_theme(palettefill="Spectral", max_colors = length(unique(vis_magOTUs_df$Cluster)),
                              leg_pos="none", guide_nrow=6,
                              y_hj=1, y_size=7, leg_size=8, y_vj=0.5,
                              x_vj=0, x_hj=1, x_size=6, x_angle=90) +
                              scale_fill_manual(values=genusColors, guide = F) +
                              line_list
                                    ggsave("Figures/08b-magOTU_per_sample_genus.pdf")

magOTUs_per_sample_by_host_genus <- ggplot(vis_magOTUs_df_all, aes(y = Cluster, x = sample, fill = factor(Genus, genera))) +
        geom_tile() +
        labs(y = "Cluster", x = "Prevalence", fill = "Genus") +
        make_theme(setFill = F, setCol = F,
                   y_size = 2, y_hj = 1.5, y_vj = 0.5,
                   x_size = 7, x_angle = 40, x_hj = 1, x_vj = 1,
                   leg_size = 5, leg_pos = "none") +
        scale_fill_manual(values=genusColors) +
          facet_wrap(~ factor(Host, host_order), scales = "free")
      ggsave("Figures/08c-magOTU_by_host_genus.pdf")
# contigs_depths_df %>%
#   filter(bin == "MAG_C2.4_8")
#
# vis_magOTUs_df_all %>%
#   filter(Cluster == "116_1") %>%
#     select(Class, Genus, ID, completeness, length)

magOTUs_per_sample_by_host_coverage <- ggplot(vis_magOTUs_df_all, aes(y = Cluster, x = sample, fill = mean_coverage)) +
        geom_tile() +
        labs(y = "Cluster", x = "Prevalence", fill = "Log of mean of contig mean coverage") +
        make_theme(setFill = F, setCol = F,
                   y_size = 7, y_hj = 1, y_vj = 0.5,
                   x_size = 7, x_angle = 40, x_hj = 1, x_vj = 1,
                   guide_nrow = 1,
                   leg_pos = "bottom"
                 ) +
          guides(fill = guide_colorbar(barhwight = 1, barwidth = 10)) +
          scale_fill_gradientn(colors=brewer.pal(5, "RdYlGn"), na.value = "transparent",
                              trans = "log10") +
          facet_wrap(~ factor(Host, host_order), scales = "free")
          ggsave("Figures/08c-magOTU_by_host_MAGs_coverage.pdf")

magOTUs_per_sample_by_host_prevalence <- ggplot(vis_magOTUs_df_all, aes(y = Cluster, x = sample, fill = Prevalence)) +
        geom_tile() +
        labs(y = "Cluster", x = "Prevalence", fill = "Prevalence within host") +
        make_theme(setFill = F, setCol = F,
                   y_size = 7, y_hj = 1, y_vj = 0.5,
                   x_size = 7, x_angle = 40, x_hj = 1, x_vj = 1,
                   guide_nrow = 1,
                   leg_pos = "bottom"
                 ) +
          guides(fill = guide_colorbar(barhwight = 1, barwidth = 10)) +
          scale_fill_gradientn(colors=brewer.pal(5, "RdYlGn"), na.value = "transparent") +
          facet_wrap(~ factor(Host, host_order), scales = "free")
          ggsave("Figures/08c-magOTU_by_host_MAGs_prevalence.pdf")

vis_magOTUs_df_all_shared_cluster <- vis_magOTUs_df_all %>%
                              group_by(Cluster) %>%
                                mutate(Num_hosts = n_distinct(Host)) %>%
                                  filter(Num_hosts > 1)

magOTUs_shared_per_sample_genus <- ggplot(vis_magOTUs_df_all_shared_cluster, aes(y = factor(Cluster), x = factor(sample, samples), fill = factor(Genus, genera))) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster")+
                              make_theme(setFill=F,
                              leg_pos="none", guide_nrow=6,
                              y_hj=1, y_size=7, leg_size=8, y_vj=0.5,
                              x_vj=0, x_hj=1, x_size=6, x_angle=90) +
                              scale_fill_manual(values=genusColors) +
                              line_list
                                    ggsave("Figures/08d-magOTU_shared_per_sample_genus.pdf")

magOTUs_shared_per_sample_prev_abund <- ggplot(vis_magOTUs_df_all_shared_cluster, aes(y = factor(Cluster), x = factor(sample, samples), fill = Prevalence)) +
                            geom_tile() +
                            geom_text(aes(label = round(mean_coverage, 2)), size = 1) +
                              labs(x = "Sample", y = "Cluster", fill = "Prevalence within host")+
                              make_theme(setFill=F,
                              y_hj=1, y_size=7, leg_size=8, y_vj=0.5,
                              x_vj=0, x_hj=1, x_size=6, x_angle=90) +
                              guides(fill = guide_colorbar(barhwight = 1, barwidth = 10)) +
                              scale_fill_gradientn(colors=brewer.pal(5, "RdYlGn"), na.value = "transparent", trans = "log10") +
                              line_list
                                    ggsave("Figures/08d-magOTUs_shared_per_sample_prev_abund.pdf")

magOTUs_per_sample_by_host_completeness_genus <- ggplot(vis_magOTUs_df_all, aes(y = Cluster, x = Num_mags, size = factor(Completeness_quality), color = Genus, alpha = 0.5)) +
        geom_point() +
        labs(y = "Cluster", x = "Number of MAGs", size = "Completeness") +
        make_theme(setFill = F, setCol = F,
                   y_size = 3, y_hj = 1, y_vj = 0.5,
                   leg_size = 5, leg_pos = "right") +
        scale_color_manual(values=genusColors) +
          facet_wrap(~ factor(Host, host_order)) +
            guides(color = "none", alpha = "none")
      ggsave("Figures/08e-magOTUs_by_host_completeness.pdf")

prev_abd_by_genus_completeness <- ggplot(vis_magOTUs_df_all, aes(y = Prevalence, x = mean_coverage, color = factor(Completeness_quality), alpha = 0.5)) +
        geom_point(position = position_jitter(w = 0, h = 0.05)) +
        labs(y = "Prevalence within host", x = "Mean of mean contig coverage", color = "Completeness") +
        make_theme(setFill = F, setCol = T,
                   palettecolor = "RdYlGn",
                   # y_size = 3, y_hj = 1, y_vj = 0.5,
                   x_angle = 30, x_hj = 1, x_vj = 1,
                   leg_size = 8, leg_pos = "right",
                   guide_nrow = 11
                 ) +
        scale_x_continuous(trans="log10") +
          facet_wrap(~ Genus) +
            guides(alpha = "none", color = "none")
            ggsave("Figures/08f-mags_prev_vs_abd_by_genus_completeness.pdf")

prev_abd_by_genus_host <- ggplot(vis_magOTUs_df_all, aes(y = Prevalence, x = mean_coverage, color = Host, alpha = 0.5)) +
        geom_point(position = position_jitter(w = 0, h = 0.05)) +
        labs(y = "Prevalence within host", x = "Mean of mean contig coverage", color = "Host species") +
        make_theme(setFill = F, setCol = F,
                   # y_size = 3, y_hj = 1, y_vj = 0.5,
                   x_angle = 30, x_hj = 1, x_vj = 1,
                   leg_size = 5, leg_pos = "right") +
        scale_color_manual(values=host_order_color) +
        scale_x_continuous(trans="log10") +
          facet_wrap(~ Genus) +
            guides(color = "none", alpha = "none")
            ggsave("Figures/08f-mags_prev_vs_abd_by_genus_host.pdf")
```

```{r magOTU_summary_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
magOTUs_per_sample
magOTUs_per_sample_genus
magOTUs_per_sample_by_host_genus
magOTUs_per_sample_by_host_coverage
magOTUs_per_sample_by_host_prevalence
magOTUs_shared_per_sample_genus
magOTUs_shared_per_sample_prev_abund
magOTUs_per_sample_by_host_completeness_genus
prev_abd_by_genus_completeness
prev_abd_by_genus_host
```


```{r magOTU_numbers_compare}
# vis_magOTUs_df_numClusters_all <- vis_magOTUs_df_all %>%
#                         group_by(sample) %>%
#                           summarise(sample, Host, number_of_clusters = n_distinct(Cluster), .groups="keep") %>%
#                             unique()

#                             test_all <- pairwise.wilcox.test(vis_magOTUs_df_numClusters_all$number_of_clusters, vis_magOTUs_df_numClusters_all$Host, p.adjust = "fdr")
#                             glm_all <- summary(glm(data = vis_magOTUs_df_numClusters_all, number_of_clusters ~ Host, family = "poisson"))



# vis_magOTUs_df_numClusters_all_plot <- ggplot(vis_magOTUs_df_numClusters_all, aes(x = factor(Host, levels = c("Apis florea", "Apis cerana", "Apis mellifera", "Apis dorsata")), y = number_of_clusters, fill = Host)) +
#                                         geom_boxplot(outlier.shape = NA) +
#                                         geom_jitter() +
#                                         labs(y = "Number of magOTUs per individual", x = "Host species") +
#                                           make_theme(leg_pos = "none", x_angle = 30, setFill = F, x_vj = 1, x_hj = 1, ) +
#                                           scale_fill_manual(values=host_order_color)
#                                           ggsave("Figures/09a-magOTU_number_per_sample_all.pdf")

# vis_magOTUs_df_numClusters <- vis_magOTUs_df %>%
#                         group_by(sample) %>%
#                           summarise(sample, Host, number_of_clusters = n_distinct(Cluster)) %>%
#                             unique

#                         test_passed <- pairwise.wilcox.test(vis_magOTUs_df_numClusters$number_of_clusters, vis_magOTUs_df_numClusters$Host, p.adjust = "fdr")
#                         glm_passed <- summary(glm(data = vis_magOTUs_df_numClusters, number_of_clusters ~ Host, family = "poisson"))


# vis_magOTUs_df_numClusters_passed_plot <- ggplot(vis_magOTUs_df_numClusters, aes(x = factor(Host, levels = c("Apis florea", "Apis cerana", "Apis mellifera", "Apis dorsata")), y = number_of_clusters, fill = Host)) +
#                                         geom_boxplot(outlier.shape = NA) +
#                                         geom_jitter() +
#                                         labs(y = "Number of magOTUs per individual", x = "Host species") +
#                                           make_theme(leg_pos = "none", x_angle = 30, setFill = F, x_hj = 1, x_vj = 1) +
#                                           scale_fill_manual(values=host_order_color)
#                                           ggsave("Figures/09a-magOTU_number_per_sample_passed.pdf")
# ```

# ```{r magOTU_numbers_compare_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
# grid.arrange(vis_magOTUs_df_numClusters_all_plot, vis_magOTUs_df_numClusters_passed_plot, nrow = 1)
# glm_all
# test_all
# glm_passed
# test_passed
# ```

# ```{r magOTU_clustering,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
# observations_host <- list(
#    `Apis mellifera` = c(pivot_wider(summarise(group_by(vis_magOTUs_df, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis mellifera`) %>% unlist),
#    `Apis cerana` = c(pivot_wider(summarise(group_by(vis_magOTUs_df, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis cerana`) %>% unlist),
#    `Apis dorsata` = c(pivot_wider(summarise(group_by(vis_magOTUs_df, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis dorsata`) %>% unlist),
#    `Apis florea` = c(pivot_wider(summarise(group_by(vis_magOTUs_df, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis florea`) %>% unlist)
# )

# # magOTU_passed_venn <- ggVennDiagram(observations_host) +
# #                         scale_color_manual(values=host_order_color) +
# #                             scale_fill_gradient(low = brewer.pal(8, "Blues")[1], high = brewer.pal(8, "Blues")[6]) +
# #                               scale_color_manual(values=host_order_color) +
# #                                 make_theme(theme_name = theme_void(), setFill = F, setCol = F, guide_nrow = 1) +
# #                                 theme(axis.text.x=element_blank(), axis.text.y=element_blank())
# #                                 ggsave("Figures/09b-magOTUs_venn_passed_MAGs.pdf")
# venn.diagram(
#   x = observations_host,
#   category.names = names(observations_host),
#   output = T,
#   filename = "Figures/09b-magOTUs_venn_passed_MAGs.pdf",
#   fill = host_order_color[!which(host_order_color == c("Apis andreniformis"))]
# )

# observations_host_all <- list(
#    `Apis mellifera` = c(pivot_wider(summarise(group_by(vis_magOTUs_df_all, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis mellifera`) %>% unlist),
#    `Apis cerana` = c(pivot_wider(summarise(group_by(vis_magOTUs_df_all, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis cerana`) %>% unlist),
#    `Apis dorsata` = c(pivot_wider(summarise(group_by(vis_magOTUs_df_all, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis dorsata`) %>% unlist),
#    `Apis florea` = c(pivot_wider(summarise(group_by(vis_magOTUs_df_all, Host), Cluster, .groups = "keep"), names_from = Host, values_from = Cluster, values_fn = list) %>% pull(`Apis florea`) %>% unlist)
# )
# # magOTU_all_venn <- ggVennDiagram(observations_host_all) +
# #                     scale_fill_gradient(low = brewer.pal(8, "Blues")[1], high = brewer.pal(8, "Blues")[6]) +
# #                       scale_color_manual(values=host_order_color) +
# #                         make_theme(theme_name = theme_void(), setFill = F, setCol = F, guide_nrow = 1) +
# #                           theme(axis.text.x=element_blank(), axis.text.y=element_blank())
# #                           ggsave("Figures/09b-magOTUs_venn_all_MAGs.pdf")

# venn.diagram(
#   x = observations_host_all,
#   category.names = names(observations_host_all),
#   output = T,
#   filename = "Figures/09b-magOTUs_venn_all_MAGs.pdf",
#   fill = host_order_color[!which(host_order_color == c("Apis andreniformis"))]
# )

# observations <- vis_magOTUs_df_all %>%
#                   group_by(sample) %>%
#                     summarise(Cluster, .groups="keep") %>%
#                       pivot_wider(names_from = sample, values_from = Cluster, values_fn = list)

# df_magOTUs_vegan <- data.frame(matrix(nrow = length(vis_magOTUs_df_all %>% pull(Sample) %>% unique), ncol = length(unique(vis_magOTUs_df_all$Cluster))))
# rownames(df_magOTUs_vegan) <- vis_magOTUs_df_all %>% pull(Sample) %>% unique
# colnames(df_magOTUs_vegan) <- unique(vis_magOTUs_df_all$Cluster)

# for (sample in rownames(df_magOTUs_vegan)) {
#   for (cluster in colnames(df_magOTUs_vegan)) {
#     if (cluster %in% observations[sample][[1]][[1]]) {
#       df_magOTUs_vegan[sample, cluster] = 1
#     } else {
#       df_magOTUs_vegan[sample, cluster] = 0
#     }
#   }
# }


# make_cum_curve <- function(samples_vector, pa_df, iterations, name = NA) {
#   num_clusters_matrix <- matrix(nrow = length(samples_vector), ncol = iterations)
#   for (iter in 1:iterations) {
#     clusters_found_cumulative <- c()
#     for (num_samples in 1:length(samples_vector)) {
#       num_new_clusters = 0
#       selected_samples <- sample(samples_vector, num_samples)
#       clusters_found <- colnames(pa_df[selected_samples, which(colSums(pa_df[selected_samples, ]) > 1)])
#       for (cluster in clusters_found) {
#         if (cluster %in% clusters_found_cumulative) {
#           invisible()
#         } else {
#           num_new_clusters <- num_new_clusters + 1
#           clusters_found_cumulative <- c(clusters_found_cumulative, cluster)
#         }
#       }
#       num_clusters_matrix[num_samples, iter] = length(clusters_found_cumulative)
#     }
#   }
#   num_clusters_df <- as.data.frame(num_clusters_matrix)
#   colnames(num_clusters_df) <- do.call(function(x) paste0("curve_", x), list(c(1:iterations)))
#   num_clusters_df <- cbind(sample_size = c(1:length(samples_vector)), num_clusters_df)
#   plot_cum_curve_df <- pivot_longer(num_clusters_df, !sample_size, values_to = "number_of_clusters", names_to = "curve")
#   plot_cum_curve_df <- cbind("name" = name, plot_cum_curve_df)
#   return(plot_cum_curve_df)
# }

# df_plot_cum_curve <- rbind(
#                         make_cum_curve(samples_am, df_magOTUs_vegan, 50, "Apis mellifera"),
#                         make_cum_curve(samples_ac, df_magOTUs_vegan, 50, "Apis cerana"),
#                         make_cum_curve(samples_ad, df_magOTUs_vegan, 50, "Apis dorsata"),
#                         make_cum_curve(samples_af, df_magOTUs_vegan, 50, "Apis florea")
#                       )

# magotu_accumulation_curve <- ggplot(data = df_plot_cum_curve, aes(x = sample_size, y = number_of_clusters, color = factor(name, host_order))) +
#                       geom_jitter(position = position_dodge(width=0.7)) +
#                         geom_smooth(se = FALSE) +
#                           labs(color = "Host species", x = "# Bees", y = "Number of magOTUs") +
#                           scale_color_manual(values=host_order_color) +
#                             make_theme(leg_pos = "bottom", setCol = F, guide_nrow = 1)
#                             ggsave("Figures/09c-magOTUs_accumulation_curve.pdf")

# pcoa_plot_by_host <- function(df_pcoa) {
#           matrix <- as.matrix(df_pcoa)
#           dist <- as.dist(matrix)
#           res_pcoa <- pcoa(dist)
#           ev1 <- res_pcoa$vectors[,1]
#           ev2 <- res_pcoa$vectors[,2]
#           df_pcoa_new <- data.frame(cbind(ev1,ev2))
#           df_pcoa_new$Sample <- rownames(df_pcoa_new)
#           rownames(df_pcoa_new) <- NULL
#           df_pcoa_new <- left_join(df_pcoa_new, select(df_meta, Sample, SpeciesID), by = "Sample")
#           perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
#           axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
#           axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
#           p <- ggplot(df_pcoa_new, aes(x = ev1,
#                                        y = ev2,
#                                        colour = factor(SpeciesID, levels = host_order)))+
#                 geom_point(stat="identity", size=2, shape=19) +
#                   labs(x=axis_x_title, y = axis_y_title, color = "Host") +
#                     make_theme(setCol = F, guide_nrow = 1) +
#                       scale_color_manual(values=host_order_color)
#           return(p)
# }

# pcoa_plot <- function(df_pcoa, variable=SpeciesID) {
#           matrix <- as.matrix(df_pcoa)
#           dist <- as.dist(matrix)
#           res_pcoa <- pcoa(dist)
#           ev1 <- res_pcoa$vectors[,1]
#           ev2 <- res_pcoa$vectors[,2]
#           df_pcoa_new <- data.frame(cbind(ev1,ev2))
#           df_pcoa_new$Sample <- rownames(df_pcoa_new)
#           rownames(df_pcoa_new) <- NULL
#           df_pcoa_new <- left_join(df_pcoa_new, select(df_meta, Sample, SpeciesID, matches(variable)), by = "Sample")
#           perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
#           axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
#           axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
#           p <- ggplot(df_pcoa_new, aes(x = ev1,
#                                        y = ev2,
#                                        colour = get(variable)))+
#                 geom_point(stat="identity", size=2, shape=19) +
#                   labs(x=axis_x_title, y = axis_y_title, color = variable) +
#                     make_theme( max_colors = length(unique(df_pcoa_new[, variable])), guide_nrow = 4 )
#           return(p)
# }

# dist_matrix <- as.matrix(vegdist(df_magOTUs_vegan, "jaccard"))

# pcoa_magotus <- pcoa_plot_by_host(dist_matrix)
#           ggsave("Figures/09d-magOTUs_pcoa.pdf")
# ```

# ```{r magOTU_clustering_plots,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
# knitr::include_graphics("Figures/09b-magOTUs_venn_all_MAGs.pdf")
# knitr::include_graphics("Figures/09b-magOTUs_venn_passed_MAGs.pdf")
# magotu_accumulation_curve
# pcoa_plot(dist_matrix, "Location_name")
#   ggsave("Figures/09d-magOTUs_pcoa_location.pdf")
# pcoa_plot(dist_matrix, "Colony")
#   ggsave("Figures/09d-magOTUs_pcoa_colony.pdf")
# pcoa_plot(dist_matrix, "Country")
#   ggsave("Figures/09d-magOTUs_pcoa_country.pdf")
# pcoa_plot(dist_matrix, "Run_ID")
#   ggsave("Figures/09d-magOTUs_pcoa_run_id.pdf")
# pcoa_magotus
# anosim(df_magOTUs_vegan, df_meta$SpeciesID, distance = "jaccard", permutations = 9999)
# adonis2(dist_matrix ~ SpeciesID, data = df_meta, permutations = 9999, method="jaccard")
# plot(betadisper(vegdist(df_magOTUs_vegan, "jaccard"), group=df_meta$SpeciesID),hull=FALSE, ellipse=TRUE)
# # ggsave("Figures/09d-magOTUs_pcoa_betadisp.pdf")
```

```{r orthofinder_results_for_phylo_tree}
Number_genomes_df <- data.frame()
for (group in Groups) {
  OG_numbers_df <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", group, "/OrthoFinder/Results_", group, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  genomes <- head(colnames(OG_numbers_df)[-1], -1)
  MAGs <- genomes[which(grepl("MAG", genomes, fixed=TRUE))]
  Isolates <- genomes[which(!grepl("MAG", genomes, fixed=TRUE))]
  Number_genomes_df <- rbind(Number_genomes_df,
                             cbind(
                                MAGs = length(MAGs),
                                Isolates = length(Isolates),
                                genomes = length(genomes),
                                Group = group
                              )
                            )
}
Number_genomes_df <- pivot_longer(Number_genomes_df, !Group, names_to = "Type", values_to = "Number") %>%
                      mutate(Number = as.numeric(Number))

Orthogroups_summary_df <- data.frame()
for (group in Groups) {
  OG_summary <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", group, "_Orthogroups_summary.csv"), sep = ",")
  Orthogroups_summary_df <- rbind(Orthogroups_summary_df,
                                  OG_summary
                            ) %>%
                              filter(present_in != "None")
}

orthogroups_present_in <- Orthogroups_summary_df %>%
  group_by(group, present_in) %>%
    summarise(Number = n(), .groups="drop")

num_genomes <- ggplot(Number_genomes_df %>% filter(Type != "genomes"),
                      aes(x = Number, y = factor(Group, genera), fill = Type)) +
                geom_bar(stat = "identity") +
                labs(x = "Number of genomes", y = "Genus group") +
                make_theme(leg_pos="right",
                           x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                           y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                ggsave("Figures/10a-Number_of_genomes_per_group.pdf")

# write.csv(vis_magOTUs_df %>% ungroup() %>% select(ID, Cluster, Genus), "Figures/magOTU_categories.csv")

vis_magOTUs_df_w_categories <- vis_magOTUs_df %>% 
                                left_join(read.csv("Figures/magOTU_categories.csv")) %>%
                                  mutate(Category = as.factor(Category))
# only filtered mags
num_genomes_by_magOTU <- ggplot(vis_magOTUs_df_w_categories,
                      aes(y = factor(Genus, genera), fill = factor(Category, c(1:15)))) +
                geom_bar(stat = "count") +
                labs(x = "Number of genomes", y = "Genus group") +
                make_theme(leg_pos="none", palettefill = "Paired",
                           max_colors = 15,
                           x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                           y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                ggsave("Figures/10d-Number_of_genomes_by_group_per_magOTU.pdf")

num_genomes_by_magOTU_panels <- ggplot(vis_magOTUs_df_w_categories,
                      aes(x = Cluster, fill = factor(Category, c(1:15)))) +
                geom_bar(stat = "count") +
                labs(y = "Number of genomes", x = "Genus group") +
                facet_wrap(~Genus, scales = "free_x") +
                geom_hline(yintercept = 1) +
                make_theme(leg_pos="none", palettefill = "Paired",
                           max_colors = 15,
                           x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                           y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                ggsave("Figures/10e-Number_of_genomes_per_group_per_magOTU.pdf")


plot_present_in <- ggplot(orthogroups_present_in,
                          aes(x = Number, y = factor(group, genera), fill = present_in)) +
                    geom_bar(stat = "identity") +
                    labs(x = "Number of orthogroups", y = "Genus group", fill = "Present in") +
                    make_theme(leg_pos="right",
                               x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                               y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                               ggsave("Figures/10b-Number_of_orthogroups_per_group.pdf")

ScpCore <- c("Isolates:ScpCore ; MAGs:ScpCore"
           )
ScpCoreMAGs <- c("Isolates:-Core ; MAGs:ScpCore",
                 "Isolates:Scp- ; MAGs:ScpCore",
                 "Isolates:-- ; MAGs:ScpCore"
           )
ScpCoreIsolates <- c("Isolates:ScpCore ; MAGs:-Core",
                     "Isolates:ScpCore ; MAGs:Scp-",
                     "Isolates:ScpCore ; MAGs:--",
                     "Isolates:ScpCore ; MAGs:-Core0.5"
           )
ScpCorehalfMAGs <- c("Isolates:ScpCore ; MAGs:ScpCore0.5"
           )

orthogroups_status <- Orthogroups_summary_df %>%
                            group_by(group, status) %>%
                            mutate(status = ifelse(status %in% c(ScpCore, ScpCorehalfMAGs, ScpCoreMAGs, ScpCoreIsolates), status, "Others")) %>%
                            mutate(status = ifelse(status %in% ScpCore, "Single-copy core orthogroups", status)) %>%
                            mutate(status = ifelse(status %in% ScpCorehalfMAGs, "Single-copy core in half of the MAGs and single-copy core in isolates", status)) %>%
                            mutate(status = ifelse(status %in% ScpCoreMAGs, "Single-copy core only in MAGs", status)) %>%
                            mutate(status = ifelse(status %in% ScpCoreIsolates, "Single-copy core only in Isolates", status)) %>%
                              summarise(Number = n(), .groups="drop")
status <- ggplot(orthogroups_status,
                  aes(x = Number, y = factor(group, genera), fill = status)) +
                geom_bar(stat = "identity", position = "stack") +
                labs(x = "Number of orthogroups", y = "Genus group", fill = "Core in") +
                scale_x_continuous(limits = c(0, 1500)) +
                make_theme(leg_pos="bottom", leg_size = 8, guide_nrow = 5,
                           max_colors = length(unique(orthogroups_status$status)),
                           palettefill = "Spectral",
                           x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                           y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
                           ggsave("Figures/10c-Orthogroups_status_per_group.pdf")
```
```{r show_orthofinder_results,  dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20}
num_genomes
plot_present_in
status
```

```{r winning_magOTUs}
score_hist <- ggplot(vis_magOTUs_df_all, aes(x = score, fill = factor(Host, host_order))) +
              geom_histogram(binwidth = 1) +
              labs(x = "Score", y = "Number of MAGs", fill = "Host") +
              geom_vline(xintercept = 15) +
                make_theme(setF = F,
                           leg_pos="right", leg_size = 7,
                           guide_nrow=4
                ) +
                  scale_fill_manual(values=host_order_color)
                  ggsave("Figures/11a-mag_drep_scores_histogram.pdf")
cluster_size_histogram <- ggplot(vis_magOTUs_df_all, aes(x = cluster_members, fill = factor(Genus, genera))) +
  geom_histogram(binwidth = 0.5) +
  labs(x = "Number of members", y = "Number of MAGs", fill = "Genus") +
    make_theme(setF = F,
               leg_pos="right", leg_size = 7,
               guide_nrow= 21
    ) +
      scale_fill_manual(values=genusColors)
      ggsave("Figures/11b-cluster_size_histogram.pdf")

get_num_clusters_and_MAGs_CMT <- function(df_info, cut_off_value){
  df_info %>%
    ungroup() %>%
      mutate(CMT = completeness - 5*contamination) %>%
        filter(CMT > cut_off_value) %>%
          summarise(MAGs = n_distinct(ID), Clusters = n_distinct(Cluster)) %>%
            as.list()
}
get_num_clusters_and_MAGs_CMT(vis_magOTUs_df_all, 0)[["MAGs"]]
get_num_clusters_and_MAGs_CMT(vis_magOTUs_df_all, 0)[["Clusters"]]

ggplot(vis_magOTUs_df_all, aes(y = Cluster, fill = factor(Genus, genera))) +
  geom_bar() +
  labs(x = "Number of MAGs", y = "Cluster", fill = "Genus") +
    make_theme(setF = F,
               leg_pos="right", leg_size = 7,
               guide_nrow= 21, y_size = 3
    ) +
    geom_vline(xintercept = 1) +
    facet_grid(~Host) +
      scale_fill_manual(values=genusColors)
cluster_completeness_number <- ggplot(vis_magOTUs_df_all, aes(y = Completeness_quality, x=cluster_members, fill = factor(Genus, genera))) +
  geom_bar(stat = "identity") +
  labs(x = "Completeness", y = "Cluster", fill = "Genus") +
    make_theme(setF = F,
               leg_pos="right", leg_size = 7,
               guide_nrow= 21, y_size = 3
    ) +
    geom_vline(xintercept = 1) +
    facet_wrap_paginate(~Cluster, ncol = 5, nrow = 7,
                      scales = "fixed", strip.position = "top", page = 5) +
      scale_fill_manual(values=genusColors)

paginate_save(cluster_completeness_number, "Cluster", "Figures/11x-clusters_completeness_numbers.pdf",
              pass_nrow = 7)
```
```{r show_winning_magOTUs, dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%', fig.align = 'center', fig.height = 20, }
score_hist
```

```{r figures_for_the_poster}
cluster_tax_info <- vis_magOTUs_df %>%
                      ungroup() %>%
                      select(Genus, Species, Cluster, Num_mags) %>%
                      unique() %>%
                        mutate(cluster_name = paste0(Genus, "-", Cluster))
corecov_data_dir <- "09_MagDatabaseProfiling/CoverageEstimation/Merged"
coords_df <- data.frame()
for (file in list.files(corecov_data_dir)) {
  if (endsWith(file, "_coord.txt")) {
    df_temp <- read.csv(paste0(corecov_data_dir, "/", file), sep = "\t", header = TRUE)
    coords_df <- rbind(coords_df, df_temp)
  }
}
samples_am <- samples[which(Vectorize(get_host_from_sample_name)(samples) == "Apis mellifera")]
samples_ac <- samples[which(Vectorize(get_host_from_sample_name)(samples) == "Apis cerana")]
samples_ad <- samples[which(Vectorize(get_host_from_sample_name)(samples) == "Apis dorsata")]
samples_af <- samples[which(Vectorize(get_host_from_sample_name)(samples) == "Apis florea")]
coverage_df <- coords_df %>%
  rename(Cluster = magOTU) %>%
    left_join(cluster_tax_info) %>%
      mutate(Host = Vectorize(get_host_name)(Sample)) %>%
        mutate(Detected = ifelse(Cov > 0, 1, 0)) %>%
          mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
          mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
          group_by(Host, Cluster) %>%
            mutate(Present = sum(Detected)) %>%
            mutate(Mean = mean(Cov)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence_host)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence_host)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence_host))
shorten_list <- function(covs_list_full, start = 1, end = 2) {
  covs_list_short = strsplit(covs_list_full, ",")[[1]][start:end]
  return(paste0(covs_list_short, collapse = ","))
}
trim_side_mean <- function(x, trim, type="both") {
    if (type == "both") {
        mean(x,trim)}
    else if (type == "right") {
        x <- sort(x)
        mean(x[1:floor(length(x)*(1-trim))])}
    else if (type == "left"){
        x <- sort(x)
        mean(x[max(1,floor(length(x)*trim)):length(x)])}
}
coverage_df_genus <- coords_df %>%
                rename(Cluster = magOTU) %>%
                  left_join(cluster_tax_info) %>%
                      mutate(Detected = ifelse(Cov > 1, 1, 0)) %>%
                        mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
                        mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
                        ungroup() %>%
                          group_by(Genus, Sample) %>%
                              summarise(Genus_Present = ifelse(any(Detected == 1), 1, 0), Num_mags = sum(Num_mags), Genus_cov = mean(Cov)) %>%
                               mutate(Host = Vectorize(get_host_name)(Sample)) %>%
                                group_by(Host, Genus) %>%
                                  mutate(Mean = round(trim_side_mean(Genus_cov, trim = 0.01)), 2) %>%
                                  mutate(Median = round(mean(Genus_cov)), 2) %>%
                                  mutate(covs_list = paste0(round(Genus_cov, 0), collapse = ",")) %>%
                                  mutate(covs_list = Vectorize(shorten_list)(covs_list, 1, 3)) %>%
                                  mutate(Present = sum(Genus_Present)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis cerana",Present/length(samples_ac), Prevalence_host)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis dorsata",Present/length(samples_ad), Prevalence_host)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis florea",Present/length(samples_af), Prevalence_host))

ggplot(coverage_df, aes(y = factor(cluster_name), x = factor(Sample, samples), fill = Cov)) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster")+
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar", trans = "log10") +
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar", trans = "log10") +
                                make_theme(setFill = F, modify_guide = F, y_size = 8,
                                  x_size = 0, x_angle = 30, x_hj = 1 , x_vj = 1,
                                  y_hj =1, leg_pos = "right"
                                )
                                ggsave(paste0("Figures/", "12_Host_magOTU_Coverage.pdf"))

ggplot(coverage_df, aes(y = factor(cluster_name), x = factor(Host, host_order), fill = Prevalence_host)) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster", fill = "Prevalence\n(Present = \nCov > 1)\n")+
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar") +
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                                coord_fixed(0.1) +
                                make_theme(setFill = F, modify_guide = F, y_size = 8,
                                  y_hj =1, leg_pos = "right"
                                )
                                ggsave(paste0("Figures/", "12_Host_magOTU_Prevalence.pdf"))

ggplot(coverage_df_genus %>% filter(Genus != "g__"), aes(y = factor(Genus, genera), x = factor(Host, host_order), fill = Prevalence_host)) +
                            geom_tile() +
                            geom_text(aes(label = ifelse(Mean > 0, Mean, ""))) +
                              labs(x = "Sample", y = "Cluster", fill = "Prevalence\n(Present = \nCov > 1)\nTruncated mean\n(0.01)")+
                              # scale_fill_gradient(na.value = "transparent", low = "#fff5f0" , high = "#99000d", guide = "colourbar") +
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar") +
                                guides(fill = guide_colourbar(nbin = 8)) +
                                make_theme(setFill = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "right", leg_size = 10
                                )
                                ggsave(paste0("Figures/", "12_Host_Genus_Prevalence.pdf"))
ggplot(coverage_df %>% filter(Genus != "g__"), aes(y = factor(Genus, genera), x = Cov, color =  factor(Genus, genera))) +
                            geom_point() +
                            # geom_text(aes(label = Mean)) +
                              labs(x = "Sample", y = "Cluster")+
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                              scale_color_manual(values = genusColors) +
                                # guides(fill = guide_colourbar(nbin = 8)) +
                                facet_wrap(~ Host) +
                                scale_x_continuous(trans = "log") +
                                make_theme(setCol = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "right", leg_size = 10
                                )

ggplot(coverage_df %>% filter(Genus != "g__"), aes(y = factor(cluster_name), x = Cov, color =  factor(Genus, genera))) +
                            geom_point() +
                            # geom_text(aes(label = Mean)) +
                              labs(x = "Sample", y = "Cluster")+
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                              scale_color_manual(values = genusColors) +
                                # guides(fill = guide_colourbar(nbin = 8)) +
                                facet_wrap(~ Host) +
                                scale_x_continuous(trans = "log") +
                                make_theme(setCol = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "none", leg_size = 0
                                )

abundance_df <- coords_df %>%
  rename(Cluster = magOTU) %>%
    left_join(cluster_tax_info %>% filter(!is.na(cluster_name))) %>%
      mutate(Host = Vectorize(get_host_name)(Sample)) %>%
        mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
          mutate(MedianCov = ifelse(is.na(MedianCov), 0, MedianCov)) %>%
            select(Sample, Cov, cluster_name, Host) %>%
              left_join(select(df_reads, Trimmed, Sample), by = "Sample") %>%
                  rename(NumReads = Trimmed) %>%
                    mutate(CovNorm = Cov/NumReads)

abundance_df_matrix <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          mutate(PresAb = ifelse(CovNorm > 0, 1, 0)) %>%
                          select(Sample, cluster_name, PresAb) %>%
                            pivot_wider(names_from=cluster_name, values_from=PresAb, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

abundance_df_matrix_pa <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          mutate(PresAb = ifelse(Cov > 1, 1, 0)) %>%
                          select(Sample, cluster_name, PresAb) %>%
                            pivot_wider(names_from=cluster_name, values_from=PresAb, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

abundance_df_matrix_Cov_rel <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample") %>%
                              mutate(across(-1)*100/rowSums(across(-1)))

abundance_df_matrix_Cov <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                          filter(Sample %in% samples_IN) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color)
                          )
row_ha = rowAnnotation(Genus = colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(cluster_tax_info %>% ungroup() %>% select(Genus, cluster_name), by = c(. = "cluster_name")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors))
                    )
pdf("Figures/16-presence_absence_heatmap.pdf")
Heatmap(t(abundance_df_matrix_pa), 
        col = c("#ffffff", "#1a88c9"),
        # col = brewer.pal(2, "Pastel1"),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
          cluster_rows = F)
dev.off()
pdf("Figures/16-relative_abundance_heatmap.pdf")
Heatmap(t(abundance_df_matrix_Cov_rel),
        col = colorRamp2(c(0, 10, 100), colors = c("#ffffff", "#1a88c9", "#1a3869")),
        # col = colorRamp2(c(0, 10, 100), colors = c(brewer.pal(9, "BrBG")[9], brewer.pal(9, "BrBG")[5], brewer.pal(9, "BrBG")[1])),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
          cluster_rows = F)
dev.off()
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_Cov) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color)
                          )
row_ha = rowAnnotation(Genus = colnames(abundance_df_matrix_Cov) %>% as.data.frame() %>% left_join(cluster_tax_info %>% ungroup() %>% select(Genus, cluster_name), by = c(. = "cluster_name")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors))
                    )
Heatmap(t(abundance_df_matrix_Cov),
        col = colorRamp2(c(0, 100, 1000), colors = c("#ffffff", "#1a88c9", "#08306b")),
        # col = colorRamp2(c(0, 10, 100), colors = c(brewer.pal(9, "BrBG")[9], brewer.pal(9, "BrBG")[5], brewer.pal(9, "BrBG")[1])),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
          cluster_rows = F)
data_otu %>% rowSums
data_otu <- abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                          mutate(Cov = round(Cov, 0)) %>%
                          group_by(Sample) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample") 
                            # column_to_rownames(var="Sample") %>%
                            #   mutate(across(-1)*100/rowSums(across(-1)))
data_grp <- df_meta %>% column_to_rownames("Sample") %>% select(SpeciesID, Location_name)
data_taxo <- vis_magOTUs_df %>% 
                    ungroup() %>% 
                      select(ID, Cluster) %>%
                        left_join(cluster_tax_info %>% select(Cluster, cluster_name)) %>%
                          left_join(df_gtdbk_bac %>% select(user_genome, classification), by = c("ID" = "user_genome")) %>%
                            ungroup() %>% 
                            select(!c(ID, Cluster)) %>%
                              unique %>%
                                column_to_rownames("cluster_name") %>%
                                  separate(classification, c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), sep = ";")

OTU = otu_table(as.matrix(data_otu), taxa_are_rows = FALSE) 
SAM = sample_data(data_grp)
TAX = tax_table(as.matrix(tax_table_data))
data_phylo <- phyloseq(OTU, TAX, SAM) 
S <- specnumber(data_otu)
raremax <- min(rowSums(data_otu))
Srare <- rarefy(data_otu, raremax)
colors_list = c(rep("#e41a1c", 20), rep("#377eb8", 20), rep("#e41a1c", 15), rep("#984ea3", 15), rep("#377eb8", 36), rep("#4daf4a", 15), rep("#377eb8", 23))
plot(S, Srare, col = colors_list,  xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)
# rarecurve(data_otu, step = 20, sample = raremax, cex = 0.6)
# https://stackoverflow.com/questions/22714775/how-do-i-colour-lines-separately-in-rarecurve-vegan-package
rarec <- function (x, step = 1, sample, xlab = "Sample Size", ylab = "Species", 
          label = TRUE, cols = c(rep('red', nrow(x) / 2), rep('blue', nrow(x) / 2)), ...) {
  tot <- rowSums(x)
  S <- specnumber(x)
  nr <- nrow(x)
  out <- lapply(seq_len(nr), function(i) {
    n <- seq(1, tot[i], by = step)
    if (n[length(n)] != tot[i]) 
      n <- c(n, tot[i])
    drop(rarefy(x[i, ], n))
  })
  Nmax <- sapply(out, function(x) max(attr(x, "Subsample")))
  Smax <- sapply(out, max)
  plot(c(1, max(Nmax)), c(1, max(Smax)), xlab = xlab, ylab = ylab, 
       type = "n", ...)
  if (!missing(sample)) {
    abline(v = sample)
    rare <- sapply(out, function(z) approx(x = attr(z, "Subsample"), 
                                           y = z, xout = sample, rule = 1)$y)
    abline(h = rare, lwd = 0.5)
  }
  for (ln in seq_len(length(out))) {
    N <- attr(out[[ln]], "Subsample")
    lines(N, out[[ln]], col = cols[ln], ...)
  }
  if (label) {
    ordilabel(cbind(tot, S), labels = rownames(x), ...)
  }
  invisible(out)
}
pdf("Figures/16-rarecurve_all.pdf")
rarec(data_otu, step = 50, sample = raremax, cols = colors_list, cex = 0.3)
dev.off()
pdf("Figures/16-rarecurve_IN.pdf")
colors_list_IN = c(rep("#e41a1c", 15), rep("#984ea3", 15), rep("#4daf4a", 15), rep("#377eb8", 5))
rarec(data_otu %>% filter(row.names(data_otu) %in% samples_IN), step = 50, sample = raremax, cols = colors_list_IN, cex = 0.5)
dev.off()

ggplot(df_reads_MAGsDB %>%
        select(Sample, MAGs_DB) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
        filter(Location %in% c("Japan", "India")) %>%
          left_join(as.data.frame(rowSums(data_otu)) %>% setNames("cov_sums") %>% rownames_to_column("Sample"), by = "Sample"),
       aes(x = MAGs_DB, y = cov_sums, 
           color = factor(Host, host_order), shape = Location)
    ) +
    labs(x = "Numer of MAGs DB mapped reads", y = "Sum of all coverages", color = "Host") +
    geom_point(size = 3) +
      make_theme(setCol = F) +
        scale_color_manual(values=host_order_color_dark) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/16-Coveragesum_vs_reads_mapped.pdf")

magOTU_clustering_Cov <- hclust(dist(abundance_df_matrix_Cov_rel), "ward.D")
magOTU_clustering_pa <- hclust(dist(abundance_df_matrix_pa))

Cov_heatmap <- ggplot(abundance_df, 
                      aes(y = factor(cluster_name),
                          x = factor(Sample,
                          levels=abundance_df_matrix_Cov_rel[order.dendrogram(as.dendrogram(magOTU_clustering_Cov)), ] %>% rownames()),
                          fill = Cov)) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster")+
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar", trans = "log10") +
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar", trans = "log10") +
                                make_theme(setFill = F, modify_guide = F, y_size = 8,
                                  x_size = 10, x_angle = 40, x_hj = 1 , x_vj = 1,
                                  y_hj =1, leg_pos = "right"
                                )

Pa_heatmap <- ggplot(abundance_df %>% mutate(PresAb = ifelse(Cov > 1, "Present", "Absent")),
                     aes(y = factor(cluster_name),
                         x = factor(Sample, 
                         levels=abundance_df_matrix_pa[order.dendrogram(as.dendrogram(magOTU_clustering_pa)), ] %>% rownames()),
                         fill = factor(PresAb))) +
                          geom_tile() +
                            labs(x = "Sample", y = "Cluster")+
                            scale_fill_manual(values=list("Present"= "#1a88c9", "Absent" = "#ffffff")) +
                            # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar", trans = "log10") +
                              make_theme(setFill = F, modify_guide = F, y_size = 8,
                                  x_size = 10, x_angle = 40, x_hj = 1 , x_vj = 1,
                                  y_hj =1, leg_pos = "right"
                              )



host_colors_dendro_cov <- label(as.dendrogram(magOTU_clustering_Cov))
host_colors_dendro_cov$cols <- recode(factor(as.vector(Vectorize(get_host_from_sample_name)(magOTU_clustering_Cov$labels))),
                                "Apis mellifera" = "#B3CDE3",
                                "Apis cerana" = "#FBB4AE",
                                "Apis dorsata" = "#DECBE4",
                                "Apis florea" =  "#CCEBC5",
                                "Apis andreniformis" = "#FED9A6"
                              )
p_cov <- grid.arrange(
    arrangeGrob(
      grid.rect(gp=gpar(col="white")),
      ggdendrogram(magOTU_clustering_Cov) +
        theme(axis.text.y = element_text(size = 0), axis.text.x = element_text(size = 5)),
      grid.rect(gp=gpar(col="white")),
      ncol = 3,
      widths = c(2, 13, 1)
    ),
    arrangeGrob(
      Cov_heatmap
    ),
    nrow = 2,
    heights = c(1,4)
)
p_cov
ggsave("Figures/12f_heatmap_coverage_clustering.pdf", p_cov)

p_pa <- grid.arrange(
    arrangeGrob(
      grid.rect(gp=gpar(col="white")),
      ggdendrogram(magOTU_clustering_pa) + theme(axis.text.y = element_text(size = 0), axis.text.x = element_text(size = 5)),
      grid.rect(gp=gpar(col="white")),
      ncol = 3,
      widths = c(1.5, 10, 1)
    ),
    arrangeGrob(
      Pa_heatmap
    ),
    nrow = 2,
    heights = c(1,4)
)
p_pa
ggsave("Figures/12f_heatmap_presence_absence_clustering.pdf", p_pa)

df_num_magOTUs <- abundance_df %>%
                    group_by(Sample) %>%
                      mutate(Detected = ifelse(CovNorm > 0, 1, 0)) %>%
                      summarise(Sample, Host = Vectorize(get_host_name)(Sample), number = sum(Detected), .groups="keep") %>%
                        unique()

                            test_all <- pairwise.wilcox.test(df_num_magOTUs$number, df_num_magOTUs$Host, p.adjust = "fdr")
                            glm_all <- summary(glm(data = df_num_magOTUs, number ~ Host, family = "poisson"))
test_all
glm_all


df_num_magOTUs_plot <- ggplot(df_num_magOTUs, aes(x = factor(Host, levels = c("Apis florea", "Apis cerana", "Apis mellifera", "Apis dorsata")), y = number, fill = Host)) +
                                        geom_boxplot(outlier.shape = NA) +
                                        geom_jitter(width = .07) +
                                        labs(y = "Number of magOTUs per individual", x = "Host species") +
                                          make_theme(leg_pos = "none", x_angle = 30, setFill = F, x_vj = 1, x_hj = 1, ) +
                                          scale_fill_manual(values=host_order_color)
                                          ggsave(paste0("Figures/", "12_BoxPlot_across_species.pdf"))
df_num_magOTUs_plot
observations_host_df <- abundance_df %>%
                    group_by(Sample) %>%
                      mutate(Detected = ifelse(CovNorm > 0, 1, 0)) %>%
                      filter(Detected == 1) %>%
                      filter(!is.na(cluster_name)) %>%
                      group_by(Host) %>%
                        summarise(Host, cluster_name) %>%
                        unique() %>%
                          pivot_wider(names_from = Host, values_from = cluster_name, values_fn = list)

observations_host <- list(
   `Apis mellifera` = observations_host_df %>% pull(`Apis mellifera`) %>% unlist,
   `Apis cerana` = observations_host_df %>% pull (`Apis cerana`) %>% unlist ,
   `Apis dorsata` = observations_host_df %>% pull (`Apis dorsata`) %>% unlist ,
   `Apis florea` = observations_host_df %>% pull (`Apis florea`) %>% unlist
)

venn.diagram(
  x = observations_host,
  category.names = names(observations_host),
  output = T,
  filename = paste0("Figures/", "12_venn_diagram.pdf"),
  fill = host_order_color
)

observations <- abundance_df %>%
                  group_by(Sample) %>%
                    summarise(cluster_name, .groups="keep") %>%
                      pivot_wider(names_from = Sample, values_from = cluster_name, values_fn = list)

df_magOTUs_vegan <- abundance_df_matrix_pa

make_cum_curve <- function(samples_vector, pa_df, iterations, name = NA) {
  num_clusters_matrix <- matrix(nrow = length(samples_vector), ncol = iterations)
  for (iter in 1:iterations) {
    clusters_found_cumulative <- c()
    for (num_samples in 1:length(samples_vector)) {
      num_new_clusters = 0
      selected_samples <- sample(samples_vector, num_samples)
      clusters_found <- colnames(pa_df[selected_samples, which(colSums(pa_df[selected_samples, ]) > 1)])
      for (cluster in clusters_found) {
        if (cluster %in% clusters_found_cumulative) {
          invisible()
        } else {
          num_new_clusters <- num_new_clusters + 1
          clusters_found_cumulative <- c(clusters_found_cumulative, cluster)
        }
      }
      num_clusters_matrix[num_samples, iter] = length(clusters_found_cumulative)
    }
  }
  num_clusters_df <- as.data.frame(num_clusters_matrix)
  colnames(num_clusters_df) <- do.call(function(x) paste0("curve_", x), list(c(1:iterations)))
  num_clusters_df <- cbind(sample_size = c(1:length(samples_vector)), num_clusters_df)
  plot_cum_curve_df <- pivot_longer(num_clusters_df, !sample_size, values_to = "number_of_clusters", names_to = "curve")
  plot_cum_curve_df <- cbind("name" = name, plot_cum_curve_df)
  return(plot_cum_curve_df)
}

df_plot_cum_curve <- rbind(
                        make_cum_curve(samples_am, df_magOTUs_vegan, 50, "Apis mellifera"),
                        make_cum_curve(samples_ac, df_magOTUs_vegan, 50, "Apis cerana"),
                        make_cum_curve(samples_ad, df_magOTUs_vegan, 50, "Apis dorsata"),
                        make_cum_curve(samples_af, df_magOTUs_vegan, 50, "Apis florea")
                      )

magotu_accumulation_curve <- ggplot(data = df_plot_cum_curve, aes(x = sample_size, y = number_of_clusters, color = factor(name, host_order))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(se = FALSE) +
                          labs(color = "Host species", x = "# Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color) +
                            make_theme(leg_pos = "bottom", setCol = F, guide_nrow = 1)
                            # ggsave("Figures/09c-magOTUs_accumulation_curve.pdf")
magotu_accumulation_curve
samples <- c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5",
              
              
              ,
              ,
              ,
              ,
              
    )
df_plot_cum_curve_colony <- rbind(
                        make_cum_curve(c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5"), df_magOTUs_vegan, 100, "India_Am"),
                        make_cum_curve(c("AmAi01", "AmAi02", "AmAi03", "AmAi04", "AmAi05",
                                         "AmAi06", "AmAi07", "AmAi08", "AmAi09", "AmAi10"), df_magOTUs_vegan, 100, "Japan_Ai_Am"),
                        make_cum_curve(c("AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05",
                                         "AmIu06", "AmIu07", "AmIu08", "AmIu09", "AmIu10"), df_magOTUs_vegan, 100, "Japan_Iu_Am"),
                        make_cum_curve(c("AcKn01", "AcKn02", "AcKn03", "AcKn04", "AcKn05",
                                         "AcKn06", "AcKn07", "AcKn08", "AcKn09", "AcKn10"), df_magOTUs_vegan, 100, "Japan_Kn_Ac"),
                        make_cum_curve(c("AcCh01", "AcCh02", "AcCh03", "AcCh04", "AcCh05",
                                         "AcCh06", "AcCh07", "AcCh08", "AcCh09", "AcCh10"), df_magOTUs_vegan, 100, "Japan_Ch_Ac"),
                        make_cum_curve(c("C1.1", "C1.2", "C1.3", "C1.4", "C1.5",
                                         "C2.1", "C2.2", "C2.3", "C2.4", "C2.5",
                                         "C3.1", "C3.2", "C3.3", "C3.4", "C3.5"), df_magOTUs_vegan, 100, "India_Ac"),
                        make_cum_curve(c("DrY1_F1", "DrY1_F2", "DrY1_F3", "DrY1_F4", "DrY1_F5", "DrY1_F6"), df_magOTUs_vegan, 100, "Switzerland_DrY1_F_Am"),
                        make_cum_curve(c("DrY1_N1", "DrY1_N2", "DrY1_N3", "DrY1_N4", "DrY1_N5", "DrY1_N6"), df_magOTUs_vegan, 100, "Switzerland_DrY1_N_Am"),
                        make_cum_curve(c("DrY1_W1", "DrY1_W2", "DrY1_W3", "DrY1_W4", "DrY1_W5", "DrY1_W6"), df_magOTUs_vegan, 100, "Switzerland_DrY1_W_Am"),
                        make_cum_curve(c("DrY2_F1", "DrY2_F2", "DrY2_F3", "DrY2_F4", "DrY2_F5", "DrY2_F6"), df_magOTUs_vegan, 100, "Switzerland_DrY2_F_Am"),
                        make_cum_curve(c("DrY2_N1", "DrY2_N2", "DrY2_N3", "DrY2_N4", "DrY2_N5", "DrY2_N6"), df_magOTUs_vegan, 100, "Switzerland_DrY2_N_Am"),
                        make_cum_curve(c("DrY2_W1", "DrY2_W2", "DrY2_W3", "DrY2_W4", "DrY2_W5", "DrY2_W6"), df_magOTUs_vegan, 100, "Switzerland_DrY2_W_Am"),
                        make_cum_curve(c("GrY2_F1", "GrY2_F2", "GrY2_F3", "GrY2_F4", "GrY2_F5", "GrY2_F6"), df_magOTUs_vegan, 100, "Switzerland_GrY2_F_Am"),
                        make_cum_curve(c("GrY2_N1", "GrY2_N2", "GrY2_N3", "GrY2_N4", "GrY2_N5", "GrY2_N6"), df_magOTUs_vegan, 100, "Switzerland_GrY2_N_Am"),
                        make_cum_curve(c("GrY2_W1", "GrY2_W2", "GrY2_W3", "GrY2_W4", "GrY2_W5", "GrY2_W6"), df_magOTUs_vegan, 100, "Switzerland_GrY2_W_Am"),
                        make_cum_curve(c("AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05"), df_magOTUs_vegan, 100, "Japan_AmIu"),
                        make_cum_curve(c("AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05"), df_magOTUs_vegan, 100, "Japan_AmIu"),
                        make_cum_curve(c("AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05"), df_magOTUs_vegan, 100, "Japan_AmIu"),
                        make_cum_curve(c("AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05"), df_magOTUs_vegan, 100, "Japan_AmIu"),
                        make_cum_curve(c("AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05"), df_magOTUs_vegan, 100, "Japan_AmIu"),
                        make_cum_curve(c("D1.1","D1.2","D1.3","D1.4","D1.5",
                                         "D2.1","D2.2","D2.3","D2.4","D2.5",
                                         "D3.1","D3.2","D3.3","D3.4","D3.5"), df_magOTUs_vegan, 100, "India_Ad"),
                        make_cum_curve(c("F1.1","F1.2","F1.3","F1.4","F1.5",
                                         "F2.1","F2.2","F2.3","F2.4","F2.5",
                                         "F3.1","F3.2","F3.3","F3.4","F3.5"), df_magOTUs_vegan, 100, "India_Af")
                      )

magotu_accumulation_curve_colony <- ggplot(data = df_plot_cum_curve_colony, aes(x = sample_size, y = number_of_clusters, color = name)) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method="gam", formula = y ~ s(x, bs = "cs", k=4)) +
                          labs(color = "Colony", x = "# Bees", y = "Number of magOTUs") +
                            make_theme(leg_pos = "bottom", setCol = T, guide_nrow = 6, max_colors = length(unique(df_plot_cum_curve_colony$name)))
                            # ggsave("Figures/09c-magOTUs_accumulation_curve.pdf")
magotu_accumulation_curve_colony
pcoa_plot_by_host <- function(df_pcoa) {
          matrix <- as.matrix(df_pcoa)
          dist <- as.dist(matrix)
          res_pcoa <- pcoa(dist)
          ev1 <- res_pcoa$vectors[,1]
          ev2 <- res_pcoa$vectors[,2]
          df_pcoa_new <- data.frame(cbind(ev1,ev2))
          df_pcoa_new$Sample <- rownames(df_pcoa_new)
          rownames(df_pcoa_new) <- NULL
          df_pcoa_new <- left_join(df_pcoa_new, select(df_meta, Sample, SpeciesID), by = "Sample")
          perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
          axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
          axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
          p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       colour = factor(SpeciesID, levels = host_order)))+
                geom_point(stat="identity", size=2, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = "Host") +
                    make_theme(setCol = F, guide_nrow = 1) +
                      scale_color_manual(values=host_order_color)
          return(p)
}

pcoa_plot <- function(df_pcoa, variable=SpeciesID) {
          matrix <- as.matrix(df_pcoa)
          dist <- as.dist(matrix)
          res_pcoa <- pcoa(dist)
          ev1 <- res_pcoa$vectors[,1]
          ev2 <- res_pcoa$vectors[,2]
          df_pcoa_new <- data.frame(cbind(ev1,ev2))
          df_pcoa_new$Sample <- rownames(df_pcoa_new)
          rownames(df_pcoa_new) <- NULL
          df_pcoa_new <- left_join(df_pcoa_new, select(df_meta, Sample, SpeciesID, matches(variable)), by = "Sample")
          perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
          axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
          axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
          p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       colour = get(variable)))+
                geom_point(stat="identity", aes(size=15, shape=14)) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme( max_colors = length(unique(df_pcoa_new[, variable])), guide_nrow = 4 )
          return(p)
}

dist_matrix <- as.matrix(vegdist(df_magOTUs_vegan, "jaccard"))

pcoa_magotus <- pcoa_plot_by_host(dist_matrix)
          # ggsave("Figures/09d-magOTUs_pcoa.pdf")
variable="SpeciesID"
df_pcoa <- dist_matrix
matrix <- as.matrix(df_pcoa)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, select(df_meta, Sample, SpeciesID, matches(variable)), by = "Sample")
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new, aes(x = ev1,
                        y = ev2,
                        color = SpeciesID))+
                geom_point(stat="identity", size = 3, shape = 16) +
                  labs(x=axis_x_title, y = axis_y_title, color = "Host") +
                  scale_color_manual(values = host_order_color) +
                    make_theme(setCol = F, guide_nrow = 4)
                    ggsave(paste0("Figures/", "12_PCoA_host_colors.pdf"))
```

```{r ani_plots}
drep_N <- read.csv("06_MAG_binning/drep_results/data_tables/Ndb.csv") %>%
            mutate(reference = Vectorize(format_genome_name)(reference)) %>%
              mutate(querry = Vectorize(format_genome_name)(querry))
# maybe do this genus-wise
matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% vis_magOTUs_df$ID) %>%
                    left_join(vis_magOTUs_df %>% ungroup() %>% select(ID, Cluster, Genus), by = c("reference" = "ID")) %>%
                      # filter(Genus %in% c("g__Bifidobacterium")) %>%
                      mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix
drep_dist_ordered <- matrix %>%
                      as.data.frame() %>%
                        rownames_to_column(var = "reference") %>%
                          pivot_longer(!reference, values_to = "ani", names_to = "querry") %>%
                            mutate(ani = as.numeric(ani))
ggplot(drep_dist_ordered, aes(x = reference, y = querry)) +
  geom_tile(aes(fill = ani)) %>%
    make_theme(setFill = F, modify_guide = F, leg_pos = "right",
               x_size = 0, y_size = 0
  ) +
    scale_fill_gradient2(low = muted("red"), mid = "white", high = muted("blue"), midpoint = 0.95, guide = "colourbar")
    # scale_fill_gradientn(colors=brewer.pal(5, "Spectral"), guide = "colourbar")
```
```{r mapping_to_MAGs}
df_reads_MAGsDB <- data.frame()
for (sample in samples) {
  if (file.exists(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"))) {
    number_trimmed_R1 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"), paste0(sample, "_R1_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R1 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"))) {
    number_trimmed_R2 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"), paste0(sample, "_R2_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R2 <- NA
  }
  raw_reads <- number_raw_R1 + number_raw_R2
  trimmed <- number_trimmed_R1 + number_trimmed_R2
  if (file.exists(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_flagstat.tsv"))) {
    mapped_full_db <- read.csv(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
    mapped_full_db <- NA
  }
  if (file.exists(paste0("09_MagDatabaseProfiling/SNVProfiling/Mapping/", sample, "_flagstat.tsv"))) {
    mapped_rep_db <- read.csv(paste0("09_MagDatabaseProfiling/SNVProfiling/Mapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
    mapped_rep_db <- NA
  }
  if (file.exists(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host_mapping_flagstat.tsv"))) {
    mapped_host_db <- read.csv(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host_mapping_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
    mapped_host_db <- NA
  }
  host_filtered <- trimmed - mapped_host_db
  unmapped_all <- trimmed - mapped_host_db - mapped_full_db
  if (file.exists(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host-filtered_flagstat.tsv"))) {
    mapped_full_db_host_filtered <- read.csv(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host-filtered_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
    mapped_full_db_host_filtered <- NA
  }
  values_to_bind <- c(sample, as.integer(c(raw_reads, trimmed, mapped_full_db, mapped_rep_db, host_filtered, mapped_full_db_host_filtered, mapped_host_db, unmapped_all)))
  df_reads_MAGsDB <- rbind(df_reads_MAGsDB, values_to_bind)
}
# setwd("~/NAS_recherche/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/211018_Medgenome_india_samples")
# df_reads_MAGsDB
# trimmed, mapped_full_db, mapped_rep_db, host_filtered, mapped_full_db_host_filtered
df_colnames <- c("Sample", "Raw", "Trimmed", "MAGs_DB", "Mapped_to_rep_MAGs", "Host_filtered_reads", "MAGs_DB_after_host_filtering", "Host", "Unmapped")
colnames(df_reads_MAGsDB) <- df_colnames
df_reads_MAGsDB <- df_reads_MAGsDB %>%
              mutate(across(!c("Sample"), as.integer))
df_reads_MAGsDB_plot <- df_reads_MAGsDB %>%
                          mutate(sum = Host_filtered_reads + MAGs_DB) %>%
                          mutate(perc_mags = MAGs_DB/Trimmed*100) %>%
                          mutate(perc_mags_no_host = MAGs_DB/(Trimmed-Host)*100) %>%
                          mutate(perc_unmapped_no_host = 100 - perc_mags_no_host) %>%
                          mutate(perc_host = Host/Trimmed*100) %>%
                          mutate(perc_unmapped_all = 100-perc_mags-perc_host) %>%
                          mutate(perc_mapped = MAGs_DB/Trimmed*100) %>%
                          mutate(perc_unmapped = 100-perc_mapped) %>%
                          mutate(perc_mapped_hf = MAGs_DB_after_host_filtering/host_filtered*100) %>%
                          mutate(perc_unmapped_hf = 100-perc_mapped_hf) %>%
                            pivot_longer(!Sample, names_to = "Type", values_to = "Number")
ggplot(df_reads_MAGsDB_plot %>% filter(Type %in% c("Trimmed", "Host_filtered_reads")),
       aes(y=factor(Sample, levels = samples),
                    x=Number,
                    fill=factor(Type, levels = c("Trimmed", "Host_filtered_reads")))) +
                        geom_bar(stat="identity", position = "dodge") +
                          # ggtitle("Total reads per sample") +
                            labs(y = "Sample",
                                 x = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                              scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="right",
                                         guide_nrow = 6,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=7,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
ggplot(df_reads_MAGsDB_plot %>% filter(Type %in% c("Trimmed", "Host_filtered_reads", "MAGs_DB")),
       aes(y=factor(Sample, levels = samples),
                    x=Number,
                    fill=factor(Type, levels = c("Trimmed", "Host_filtered_reads", "MAGs_DB")))) +
                        geom_bar(stat="identity", position = "dodge") +
                          # ggtitle("Total reads per sample") +
                            labs(y = "Sample",
                                 x = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                              scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=7)
ggplot(df_reads_MAGsDB_plot %>%
        filter(Type %in% c("Trimmed", "MAGs_DB")) %>%
          mutate(Host = Vectorize(get_host_from_sample_name)(Sample)),
       aes(x=factor(Sample, levels = samples),
                    y=Number,
                    fill=factor(Type, levels = c("Trimmed", "MAGs_DB")))) +
                        geom_bar(stat="identity", position = "dodge") +
                          # ggtitle("Total reads per sample") +
                            labs(x = "Sample",
                                 y = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                              facet_wrap(~Host, scales = "free_x") +
                              scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 1,
                                         x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
                              ggsave("Figures/13b-mapping_to_MAGs.pdf")
ggplot(df_reads_MAGsDB_plot %>%
        filter(Type %in% c("MAGs_DB", "Host", "Unmapped")) %>%
          mutate(Host = Vectorize(get_host_from_sample_name)(Sample)),
        aes(x=factor(Sample, levels = samples),
                    y=Number,
                    fill=factor(Type, levels = c("Unmapped", "Host", "MAGs_DB")))) +
                        geom_bar(stat="identity", position = "stack") +
                          # ggtitle("Total reads per sample") +
                            labs(x = "Sample",
                                 y = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                              facet_wrap(~Host, scales = "free_x") +
                              scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 1,
                                         x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
                              ggsave("Figures/13b-mapping_to_MAGs.pdf")
ggplot(df_reads_MAGsDB_plot %>%
        filter(Type %in% c("perc_mags", "perc_host", "perc_unmapped_all")),
        aes(y=factor(Sample, levels = samples),
                    x=Number,
                    fill=factor(Type, levels = c("perc_unmapped_all", "perc_host", "perc_mags")))) +
                        geom_bar(stat="identity", position = "stack") +
                          # ggtitle("Total reads per sample") +
                            labs(y = "Sample",
                                 x = "Percentage of reads (paired end)",
                                 fill = "Mapped or unmapped to MAGs DB") +
                              # scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="right",
                                         guide_nrow = 6,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=7)
                               ggsave("Figures/13a-mapping_to_MAGs_perc.pdf")
ggplot(df_reads_MAGsDB_plot %>%
        filter(Type %in% c("perc_mags_no_host", "perc_unmapped_no_host")),
       aes(y=factor(Sample, levels = samples),
                    x=Number,
                    fill=factor(Type, levels = c("perc_unmapped_no_host", "perc_mags_no_host")))) +
                        geom_bar(stat="identity", position = "stack") +
                          # ggtitle("Total reads per sample") +
                            labs(y = "Sample",
                                 x = "Percentage of reads (paired end)",
                                 fill = "Mapped or unmapped to MAGs DB") +
                              # scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="right",
                                         guide_nrow = 6,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=7)
                              ggsave("Figures/13-mapping_to_MAGs_no_host_perc.pdf")

ggplot(df_reads_MAGsDB_plot %>% filter(Type %in% c("MAGs_DB", "Mapped_to_rep_MAGs")),
       aes(y=factor(Sample, levels = samples),
                    x=Number,
                    fill=factor(Type, levels = c("MAGs_DB", "Mapped_to_rep_MAGs")))) +
                        geom_bar(stat="identity", position = "dodge") +
                          # ggtitle("Total reads per sample") +
                            labs(y = "Sample",
                                 x = "Percentage of reads (paired end)",
                                 fill = "Mapped to") +
                              scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
                              make_theme(theme_name=theme_few(), leg_pos="right",
                                         guide_nrow = 6,
                                         x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=7)

```

```{r parse_qPCR_results}
qpcr_df <- data.frame()
plate_num <- 1
for(x in qpcr_results){
  # read just the rows and columns with the values and leave out metadata cells
  # Limits the cells from rows=(36 to 132) and columns=(2 to 11)
  t <- read_excel(x, sheet = 1, range = cell_limits(c(36,2), c(132,11)))
  t <- t %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H1", "H2", "H3"), paste0("Control_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H4", "H5", "H6"), paste0("Control_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H7", "H8", "H9"), paste0("Control_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H10", "H11", "H12"), paste0("Control_", plate_num), `Sample Name`))
  # get only columns "Well Position", "Sample Name", "Target Name" and "CT" from the xls report
  t <- t[,c(1,3,4,8,9,10)]
  names(t)[1:4] <- c("Well", "Sample.Name","Target.Name","Ct")
  t$Ct <- as.numeric(t$Ct) # NAs will be introduced at the place of 'undetermined' values
  t$Target.Name <- factor(t$Target.Name)
  t$Sample.Name <- factor(t$Sample.Name)
  # t$sd <- factor(t$sd)
  t <- as.data.frame(t)
  # add in the plate number
  t <- cbind(t, "Plate" = as.factor(rep(plate_num, dim(t)[1])))
  # bind the previous plate values with the next before moving to the next file
  qpcr_df <- rbind(qpcr_df, t)
  plate_num = plate_num + 1
}
qpcr_df_info <- qpcr_df %>%
          mutate(Target.Name = ifelse(Target.Name == "UV_0356/0772", "Bacteria", "Host")) %>%
                filter(!is.na(Sample.Name)) %>%
                filter(!is.na(`Ct Mean`)) %>%
                  mutate(Host = ifelse(grepl("Control", Sample.Name), Sample.Name, NA)) %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample.Name)) %>%
                  mutate(Host = as.factor(Host)) %>%
                  filter(Target.Name == "Bacteria") %>%
                  group_by(Sample.Name) %>%
                    summarise(`Ct Mean`, Host, `Ct SD`, Plate) %>%
                      group_by(Host) %>%
                      unique() %>%
                        summarise(Sample.Name, `Ct Mean`, `Ct SD`, Plate) %>%
                          rename(Ct_mean = `Ct Mean`) %>%
                          rename(Ct_SD = `Ct SD`)
# qpcr_df_info
```


```{r parse_qPCR_results dev = 'pdf', results='hold', fig.show = 'hold', out.width = '100%'}
ggplot(qpcr_df_info %>% filter(!is.na(Host)) %>% unique(),
       aes(x = factor(Host, levels = c(host_order, "Control_1", "Control_2", "Control_3")),
           y = Ct_mean,
           fill = factor(Host, levels = c(host_order, "Control_1", "Control_2", "Control_3")),
         )
     ) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width=0.1, size = 2, aes(shape = Plate)) +
  labs(y = "Ct value with UV primers (mean)", x = "Host species") +
    make_theme(leg_pos = "none", x_angle = 30, setFill = F, x_vj = 1, x_hj = 1, ) +
      scale_fill_manual(values=host_order_color)
pairwise.wilcox.test(qpcr_df_info$Ct_mean, qpcr_df_info$Host, p.adjust = "fdr")

summary(glm(data = qpcr_df_info, Ct_mean ~ Host, family = "gaussian"))
```

```{r parse_instrain}
# instrain_df <- read.csv("10_instrain/rep_mags.IS.COMPARE/output/rep_mags.IS.COMPARE_genomeWide_compare.tsv", sep = "\t")
remove_extension <- function(name, extension) {
  return(strsplit(name, extension)[[1]][[1]])
}

# instrain_df %>%
#   mutate(name1 = Vectorize(remove_extension)(name1, ".bam")) %>%
#     mutate(name2 = Vectorize(remove_extension)(name2, ".bam")) %>%
#       mutate(genome = Vectorize(remove_extension)(genome, ".fna"))
sample_snps_df <- data.frame()
for (sample in samples) {
  df_read <- read.csv(paste0("10_instrain/",sample,"_profile.IS/output/",sample,"_profile.IS_genome_info.tsv"), sep = "\t") %>%
    mutate(genome = Vectorize(remove_extension)(genome, ".fna")) %>%
      left_join(vis_magOTUs_df_all %>% ungroup() %>% select(ID, Cluster, Genus, Ref_status, all_quality), by = c("genome" = "ID")) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
          mutate(breadth_difference = breadth - breadth_expected)
  sample_snps_df <- bind_rows(sample_snps_df, cbind(df_read, "sample"=sample))
}

snp_breadth_depth <- sample_snps_df %>%
                        mutate(label = paste0(Genus, "_", Cluster, "_", genome)) %>%
                            group_by(label) %>%
                            summarise(coverage, breadth, sample, Genus) %>%
                              mutate(coverage_factor = cut(coverage ,breaks=c(-1, 1e-04, 1e-03, 1e-02, 1e-01, 1, 10, 100, 1000, Inf),labels = c("<0.0001", "0.001", "0.01", "0.1", "1", "10", "100", "1000", ">1000")))

plot_breadth_by_sample <- ggplot(sample_snps_df,
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Genus,
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~sample, ncol = 3, nrow = 3, page = 1) +
    # scale_y_continuous(trans="log10") +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_vline(xintercept = 5.5) +
        geom_vline(xintercept = 20.5) +
        geom_vline(xintercept = 30.5) +
        theme(strip.text = element_text(size = 7)) +
        scale_color_manual(values=genusColors)

paginate_save(plot_breadth_by_sample, "sample", "Figures/13a-plot_breadth_by_sample.pdf", pass_nrow = 3, pass_ncol = 3)

plot_breadth_by_cluster <- ggplot(sample_snps_df,
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Host,
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~Cluster, ncol = 3, nrow = 3, page = 5) +
    # scale_y_continuous(trans="log10") +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_vline(xintercept = 5.5) +
        geom_vline(xintercept = 20.5) +
        geom_vline(xintercept = 30.5) +
        theme(strip.text = element_text(size = 7)) +
        scale_color_manual(values=host_order_color)

paginate_save(plot_breadth_by_cluster, "Cluster", "Figures/13b-plot_breadth_by_cluster.pdf", pass_nrow = 3, pass_ncol = 3)

plot_iRep <- ggplot(sample_snps_df,
                             aes(x = sample,
                                 y = iRep,
                                 color = Genus,
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~Cluster, ncol = 3, nrow = 3, page = 5) +
    # scale_y_continuous(trans="log10") +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_vline(xintercept = 5.5) +
        geom_vline(xintercept = 20.5) +
        geom_vline(xintercept = 30.5) +
        theme(strip.text = element_text(size = 7)) +
        scale_color_manual(values=genusColors)

paginate_save(plot_iRep, "Cluster", "Figures/13c-plot_iRep_by_sample.pdf", pass_nrow = 3, pass_ncol = 3)

plot_breadth <- ggplot(snp_breadth_depth,
                             aes(x = factor(sample, samples),
                                 y = breadth,
                                 # size = breadth,
                                 color = coverage_factor
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~label, ncol = 3, nrow = 3, page = 2) +
    # scale_y_continuous(trans="log10") +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_vline(xintercept = 5.5) +
        geom_vline(xintercept = 20.5) +
        geom_vline(xintercept = 30.5) +
        theme(strip.text = element_text(size = 7)) +
        scale_color_manual(values=brewer.pal(9, "RdYlGn"))

paginate_save(plot_breadth, "label", "Figures/14x-rep_genomes_breadth.pdf", pass_nrow = 3, pass_ncol = 3)

plot_breadth_depth <- ggplot(snp_breadth_depth %>% filter(breadth >= .5),
                             aes(x = factor(sample, samples),
                                 y = coverage,
                                 # size = breadth,
                                 color = Genus
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~label, ncol = 3, nrow = 3, page = 2) +
    scale_y_continuous(trans="log10") +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_vline(xintercept = 5.5) +
        geom_vline(xintercept = 20.5) +
        geom_vline(xintercept = 30.5) +
        theme(strip.text = element_text(size = 7)) +
        scale_color_manual(values=genusColors)

paginate_save(plot_breadth_depth, "label", "Figures/14x-rep_genomes_coverage.pdf", pass_nrow = 3, pass_ncol = 3)

shared_genera <- vis_magOTUs_df %>%
  ungroup() %>%
  group_by(Genus) %>%
    mutate(num_genus = n_distinct(Host)) %>%
      summarise(num_genus) %>%
      filter(num_genus > 1) %>%
        pull(Genus) %>%
          unique

snv_perc_plot_all <- ggplot(sample_snps_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = SNV_count/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y", "N")),
                             aes(x = Cluster,
                                 y = perc_var,
                                 color = factor(Host, host_order),
                                 shape = breadth_pass
                              )
                            ) +
      geom_point() +
      labs(x = "magOTU", y = "% variant sites", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)
            ggsave("Figures/15a-percentage_variable_sites_all.pdf")

snv_perc_plot <- ggplot(sample_snps_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = SNV_count/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = Cluster,
                                 y = perc_var,
                                 color = factor(Host, host_order),
                                 shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      labs(x = "magOTU", y = "% variant sites", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)
            ggsave("Figures/15b-percentage_variable_sites.pdf")

# snv_perc_plot_rare <- ggplot(sample_snps_df %>%
ggplot(sample_snps_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = consensus_divergent_sites/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = Cluster,
                                 y = reads_mean_PID,
                                 color = factor(Host, host_order)
                                 # shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      scale_y_continuous(labels=unit_format(unit = "%", scale = 1e+2)) +
      labs(x = "magOTU", y = "mean ANI (reads and reference)", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)

ggplot(sample_snps_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = consensus_divergent_sites/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = Cluster,
                                 y = reads_unfiltered_reads,
                                 color = factor(Host, host_order)
                                 # shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), trans = "log") +
      labs(x = "magOTU", y = "Number of reads mapped", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
        # geom_hline(yintercept = 10, color = "grey") +
        # geom_hline(yintercept = 100, color = "grey") +
        # geom_hline(yintercept = 1000, color = "grey") +
        geom_hline(yintercept = 10000, color = "grey") +
        # geom_hline(yintercept = 100000, color = "grey") +
        geom_hline(yintercept = 1000000, color = "grey") +
        # geom_hline(yintercept = 10000000, color = "grey") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)

```

```{r visualise_core_gene_summaries}
vis_magOTUs_df_all %>%
  filter(Host == "Apis mellifera") %>%
    filter(Genus == "g__Bifidobacterium") %>%
      pull(Cluster, completeness)
colnames(vis_magOTUs_df)

MAG_orthogroups_summary_df <- data.frame()
for (group in vis_magOTUs_df %>% filter(Genus != "g__") %>% pull(Genus) %>% unique) {
  if (file.exists(paste0("database/MAGs_database_Orthofinder/", group, "_Orthogroups_summary.csv"))) {
    OG_summary <- read.csv(paste0("database/MAGs_database_Orthofinder/", group, "_Orthogroups_summary.csv"), sep = ",")
  }
  MAG_orthogroups_summary_df <- rbind(MAG_orthogroups_summary_df,
                                  OG_summary
                            ) %>%
                              filter(present_in != "None")
    }

MAG_orthogroups_filt_df <- data.frame()
for (group in vis_magOTUs_df %>% filter(Genus != "g__") %>% pull(Genus) %>% unique) {
  if (file.exists(paste0("database/MAGs_database_Orthofinder/", group, "_Orthogroups_filtered_summary.csv"))) {
    OG_summary <- read.csv(paste0("database/MAGs_database_Orthofinder/", group, "_Orthogroups_filtered_summary.csv"), sep = ",")
  }
  MAG_orthogroups_filt_df <- rbind(MAG_orthogroups_filt_df,
                                  OG_summary
                            ) %>%
                              filter(present_in != "None")
    }

MAG_orthogroups_summary_df
MAG_orthogroups_filt_df
ggplot(MAG_orthogroups_filt_df,
       aes(
         y = group,
         fill = passed_filter
      )
    ) +
  geom_bar() +
   make_theme()

present_and_missing_df <- MAG_orthogroups_filt_df %>%
                            group_by(group) %>%
                              left_join(vis_magOTUs_df %>% ungroup() %>% group_by(Genus) %>% summarise(number_total = n()), by = c("group" = "Genus")) %>%
                                mutate(prevalence = ((number_total-missing_in)/number_total)*100) %>%
                                  mutate(prevalence_type = cut(prevalence ,breaks=c(-1, 50, 60, 70, 80, 90, 95, 99, 100),
                                                                  labels = c("<50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-99", "100"))
                                                                )

ggplot(present_and_missing_df,
       aes(
         y = group,
         fill = prevalence_type
      )
    ) +
  geom_bar() +
   make_theme(guide_nrow = 4)

```

```{r plot_SDP_validation}
density_plot <- function(infile, sdp) {
  data <- read.table(infile, h=T) %>%
           mutate(Host = Vectorize(get_host_from_scaffold)(ORF_id)) %>%
             mutate(Status = ifelse(Closest_SDP == sdp, "same", "different"))
  length_sdp <- length(data %>% filter(Status == "same") %>% pull)
  length_other<- length(data %>% filter(Status == "different") %>% pull)
  tot_length = length_sdp + length_other
  plot_title <- paste(sdp, ". Tot nb. orfs: ",tot_length, sep="")
  colors_vec <- list("same" = brewer.pal(9, "Set1")[2], "different" = brewer.pal(9, "Set1")[9])
  p <- ggplot() +
        geom_density(data=data, aes(x=Perc_id, group=Status, fill=Status, alpha=0.75)) + xlab(paste0("% identity to ", sdp)) +
          xlab(paste0("% identity to ", sdp)) +
          # geom_density(data=melt.df, aes(x=all_perc, group=all_names, fill=all_names, alpha=all_names))+ xlim(60,100) +xlab("% identity") +
            ggtitle(plot_title) +
              scale_fill_manual(values=colors_vec) +
                make_theme(setFill=F, setCol=F, leg_pos="none") + geom_vline(xintercept=95, linetype="dashed")
	 return(p)
}

get_host_from_scaffold <- function(orf_id){
  sample_name = strsplit(orf_id, "_")[[1]][[1]]
  get_host_from_sample_name(sample_name)
}
# infile = "12_species_validation/g__Frischella_137_1_perc_id.txt"
# sdp = "137_1"
density_plot_by_host <- function(infile, sdp) {
	 data <- read.table(infile, h=T) %>%
            mutate(Host = Vectorize(get_host_from_scaffold)(ORF_id)) %>%
              mutate(Status = ifelse(Closest_SDP == sdp, "same", "different"))
	 length_sdp <- length(data %>% filter(Status == "same") %>% pull)
	 length_other<- length(data %>% filter(Status == "different") %>% pull)
	 tot_length = length_sdp + length_other
	 plot_title <- paste(sdp, ". Tot nb. orfs: ",tot_length, sep="")
   colors_vec <- list("same" = brewer.pal(9, "Pastel1")[2], "different" = brewer.pal(9, "Pastel1")[1])
	 p <- ggplot() +
         geom_density(data=data, aes(x=Perc_id, group=Status, fill=Status, alpha=0.9)) + xlab(paste0("% identity to ", sdp)) +
           ggtitle(plot_title) +
             scale_fill_manual(values=colors_vec) +
             facet_wrap(~Host, scales = "free_y", nrow = 2) +
                make_theme(setFill=F, setCol=F, leg_pos="none") + geom_vline(xintercept=95, linetype="dashed")
	 return(p)
}

dir.create(file.path("Figures", "20_SDP_density_plots"))

##data acquisition
grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Apibacter/g__Apibacter_107_1_perc_id.txt", "107_1"),
    density_plot("12_species_validation/g__Apibacter/g__Apibacter_106_1_perc_id.txt", "106_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Apibacter.pdf")
density_plot_by_host("12_species_validation/g__Apibacter/g__Apibacter_107_1_perc_id.txt", "107_1")
  ggsave("Figures/20_SDP_density_plots/g__Apibacter_107_1.pdf")
density_plot_by_host("12_species_validation/g__Apibacter/g__Apibacter_106_1_perc_id.txt", "106_1")
  ggsave("Figures/20_SDP_density_plots/g__Apibacter_106_1.pdf")


grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Bartonella/g__Bartonella_92_1_perc_id.txt", "92_1"),
    density_plot("12_species_validation/g__Bartonella/g__Bartonella_93_1_perc_id.txt", "93_1"),
    density_plot("12_species_validation/g__Bartonella/g__Bartonella_94_1_perc_id.txt", "94_1"),
    density_plot("12_species_validation/g__Bartonella/g__Bartonella_94_2_perc_id.txt", "94_2"),
    density_plot("12_species_validation/g__Bartonella/g__Bartonella_96_0_perc_id.txt", "96_0"),
    density_plot("12_species_validation/g__Bartonella/g__Bartonella_116_1_perc_id.txt", "116_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Bartonella.pdf")
density_plot_by_host("12_species_validation/g__Bartonella/g__Bartonella_92_1_perc_id.txt", "92_1")
  ggsave("Figures/20_SDP_density_plots/g__Bartonella_92_1.pdf")
density_plot_by_host("12_species_validation/g__Bartonella/g__Bartonella_93_1_perc_id.txt", "93_1")
  ggsave("Figures/20_SDP_density_plots/g__Bartonella_93_1.pdf")
density_plot_by_host("12_species_validation/g__Bartonella/g__Bartonella_94_1_perc_id.txt", "94_1")
  ggsave("Figures/20_SDP_density_plots/g__Bartonella_94_1.pdf")
density_plot_by_host("12_species_validation/g__Bartonella/g__Bartonella_94_2_perc_id.txt", "94_2")
  ggsave("Figures/20_SDP_density_plots/g__Bartonella_94_2.pdf")
density_plot_by_host("12_species_validation/g__Bartonella/g__Bartonella_96_0_perc_id.txt", "96_0")
  ggsave("Figures/20_SDP_density_plots/g__Bartonella_96_0.pdf")
density_plot_by_host("12_species_validation/g__Bartonella/g__Bartonella_116_1_perc_id.txt", "116_1")
  ggsave("Figures/20_SDP_density_plots/g__Bartonella_116_1.pdf")

grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_14_1_perc_id.txt", "14_1"),
    # density_plot("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_14_3_perc_id.txt", "14_3"),
    density_plot("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_14_4_perc_id.txt", "14_4"),
    density_plot("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_17_1_perc_id.txt", "17_1"),
    density_plot("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_22_1_perc_id.txt", "22_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Bifidobacterium.pdf")
density_plot_by_host("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_14_1_perc_id.txt", "14_1")
  ggsave("Figures/20_SDP_density_plots/g__Bifidobacterium_14_1.pdf")
density_plot_by_host("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_14_3_perc_id.txt", "14_3")
  ggsave("Figures/20_SDP_density_plots/g__Bifidobacterium_14_3.pdf")
density_plot_by_host("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_14_4_perc_id.txt", "14_4")
  ggsave("Figures/20_SDP_density_plots/g__Bifidobacterium_14_4.pdf")
density_plot_by_host("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_17_1_perc_id.txt", "17_1")
  ggsave("Figures/20_SDP_density_plots/g__Bifidobacterium_17_1.pdf")
density_plot_by_host("12_species_validation/g__Bifidobacterium/g__Bifidobacterium_22_1_perc_id.txt", "22_1")
  ggsave("Figures/20_SDP_density_plots/g__Bifidobacterium_22_1.pdf")


grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Bombella/g__Bombella_87_1_perc_id.txt", "87_1"),
    density_plot("12_species_validation/g__Bombella/g__Bombella_88_0_perc_id.txt", "88_0")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Bombella.pdf")
density_plot_by_host("12_species_validation/g__Bombella/g__Bombella_87_1_perc_id.txt", "87_1")
  ggsave("Figures/20_SDP_density_plots/g__Bombella_87_1.pdf")
density_plot_by_host("12_species_validation/g__Bombella/g__Bombella_88_0_perc_id.txt", "88_0")
  ggsave("Figures/20_SDP_density_plots/g__Bombella_88_0.pdf")

grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_33_1_perc_id.txt", "33_1"),
    density_plot("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_69_1_perc_id.txt", "69_1"),
    density_plot("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_69_4_perc_id.txt", "69_4"),
    density_plot("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_70_1_perc_id.txt", "70_1"),
    density_plot("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_71_1_perc_id.txt", "71_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Bombilactobacillus.pdf")
density_plot_by_host("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_33_1_perc_id.txt", "33_1")
  ggsave("Figures/20_SDP_density_plots/g__Bombilactobacillus_33_1.pdf")
density_plot_by_host("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_69_1_perc_id.txt", "69_1")
  ggsave("Figures/20_SDP_density_plots/g__Bombilactobacillus_69_1.pdf")
density_plot_by_host("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_69_4_perc_id.txt", "69_4")
  ggsave("Figures/20_SDP_density_plots/g__Bombilactobacillus_69_4.pdf")
density_plot_by_host("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_70_1_perc_id.txt", "70_1")
  ggsave("Figures/20_SDP_density_plots/g__Bombilactobacillus_70_1.pdf")
density_plot_by_host("12_species_validation/g__Bombilactobacillus/g__Bombilactobacillus_71_1_perc_id.txt", "71_1")
  ggsave("Figures/20_SDP_density_plots/g__Bombilactobacillus_71_1.pdf")



grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Commensalibacter/g__Commensalibacter_85_1_perc_id.txt", "85_1"),
    density_plot("12_species_validation/g__Commensalibacter/g__Commensalibacter_84_1_perc_id.txt", "84_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Commensalibacter.pdf")
density_plot_by_host("12_species_validation/g__Commensalibacter/g__Commensalibacter_85_1_perc_id.txt", "85_1")
  ggsave("Figures/20_SDP_density_plots/g__Commensalibacter_85_1.pdf")
density_plot_by_host("12_species_validation/g__Commensalibacter/g__Commensalibacter_84_1_perc_id.txt", "84_1")
  ggsave("Figures/20_SDP_density_plots/g__Commensalibacter_84_1.pdf")

grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_77_1_perc_id.txt", "77_1"),
    density_plot("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_79_1_perc_id.txt", "79_1"),
    density_plot("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_80_1_perc_id.txt", "80_1"),
    density_plot("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_81_1_perc_id.txt", "81_1"),
    density_plot("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_82_1_perc_id.txt", "82_1"),
    density_plot("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_83_1_perc_id.txt", "83_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas.pdf")
density_plot_by_host("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_77_1_perc_id.txt", "77_1")
  ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas_77_1.pdf")
density_plot_by_host("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_79_1_perc_id.txt", "79_1")
  ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas_79_1.pdf")
density_plot_by_host("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_80_1_perc_id.txt", "80_1")
  ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas_80_1.pdf")
density_plot_by_host("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_81_1_perc_id.txt", "81_1")
  ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas_81_1.pdf")
density_plot_by_host("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_82_1_perc_id.txt", "82_1")
  ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas_82_1.pdf")
density_plot_by_host("12_species_validation/g__Dysgonomonas/g__Dysgonomonas_83_1_perc_id.txt", "83_1")
  ggsave("Figures/20_SDP_density_plots/g__Dysgonomonas_83_1.pdf")

grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Frischella/g__Frischella_134_1_perc_id.txt", "134_1"),
    density_plot("12_species_validation/g__Frischella/g__Frischella_137_1_perc_id.txt", "137_1"),
    density_plot("12_species_validation/g__Frischella/g__Frischella_139_1_perc_id.txt", "139_1"),
    density_plot("12_species_validation/g__Frischella/g__Frischella_140_1_perc_id.txt", "140_1"),
    density_plot("12_species_validation/g__Frischella/g__Frischella_141_1_perc_id.txt", "141_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Frischella_.pdf")
density_plot_by_host("12_species_validation/g__Frischella/g__Frischella_134_1_perc_id.txt", "134_1")
  ggsave("Figures/20_SDP_density_plots/g__Frischella_134_1.pdf")
density_plot_by_host("12_species_validation/g__Frischella/g__Frischella_137_1_perc_id.txt", "137_1")
  ggsave("Figures/20_SDP_density_plots/g__Frischella_137_1.pdf")
density_plot_by_host("12_species_validation/g__Frischella/g__Frischella_139_1_perc_id.txt", "139_1")
  ggsave("Figures/20_SDP_density_plots/g__Frischella_139_1.pdf")
density_plot_by_host("12_species_validation/g__Frischella/g__Frischella_140_1_perc_id.txt", "140_1")
  ggsave("Figures/20_SDP_density_plots/g__Frischella_140_1.pdf")
density_plot_by_host("12_species_validation/g__Frischella/g__Frischella_141_1_perc_id.txt", "141_1")
  ggsave("Figures/20_SDP_density_plots/g__Frischella_141_1.pdf")



grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Gilliamella/g__Gilliamella_167_1_perc_id.txt", "167_1")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Gilliamella.pdf")
density_plot_by_host("12_species_validation/g__Gilliamella/g__Gilliamella_167_1_perc_id.txt", "167_1")
  ggsave("Figures/20_SDP_density_plots/g__Gilliamella_167_1.pdf")

grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_40_2_perc_id.txt", "40_2"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_41_1_perc_id.txt", "41_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_48_1_perc_id.txt", "48_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_49_1_perc_id.txt", "49_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_52_1_perc_id.txt", "52_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_57_1_perc_id.txt", "57_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_60_1_perc_id.txt", "60_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_63_1_perc_id.txt", "63_1"),
    density_plot("12_species_validation/g__Lactobacillus/g__Lactobacillus_63_2_perc_id.txt", "63_2")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Lactobacillus.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_40_2_perc_id.txt", "40_2")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_40_2.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_41_1_perc_id.txt", "41_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_41_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_48_1_perc_id.txt", "48_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_48_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_49_1_perc_id.txt", "49_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_49_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_52_1_perc_id.txt", "52_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_52_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_57_1_perc_id.txt", "57_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_57_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_60_1_perc_id.txt", "60_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_60_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_63_1_perc_id.txt", "63_1")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_63_1.pdf")
density_plot_by_host("12_species_validation/g__Lactobacillus/g__Lactobacillus_63_2_perc_id.txt", "63_2")
  ggsave("Figures/20_SDP_density_plots/g__Lactobacillus_63_2.pdf")


grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Snodgrassella/g__Snodgrassella_117_1_perc_id.txt", "117_1"),
    density_plot("12_species_validation/g__Snodgrassella/g__Snodgrassella_120_1_perc_id.txt", "120_1"),
    density_plot("12_species_validation/g__Snodgrassella/g__Snodgrassella_121_2_perc_id.txt", "121_2"),
    density_plot("12_species_validation/g__Snodgrassella/g__Snodgrassella_122_1_perc_id.txt", "122_1"),
    density_plot("12_species_validation/g__Snodgrassella/g__Snodgrassella_123_1_perc_id.txt", "123_1"),
    density_plot("12_species_validation/g__Snodgrassella/g__Snodgrassella_124_0_perc_id.txt", "124_0")
  )
)
ggsave("Figures/20_SDP_density_plots/g__Snodgrassella.pdf")
density_plot_by_host("12_species_validation/g__Snodgrassella/g__Snodgrassella_117_1_perc_id.txt", "117_1")
  ggsave("Figures/20_SDP_density_plots/g__Snodgrassella_117_1.pdf")
density_plot_by_host("12_species_validation/g__Snodgrassella/g__Snodgrassella_120_1_perc_id.txt", "120_1")
  ggsave("Figures/20_SDP_density_plots/g__Snodgrassella_120_1.pdf")
density_plot_by_host("12_species_validation/g__Snodgrassella/g__Snodgrassella_121_2_perc_id.txt", "121_2")
  ggsave("Figures/20_SDP_density_plots/g__Snodgrassella_121_2.pdf")
density_plot_by_host("12_species_validation/g__Snodgrassella/g__Snodgrassella_122_1_perc_id.txt", "122_1")
  ggsave("Figures/20_SDP_density_plots/g__Snodgrassella_122_1.pdf")
density_plot_by_host("12_species_validation/g__Snodgrassella/g__Snodgrassella_123_1_perc_id.txt", "123_1")
  ggsave("Figures/20_SDP_density_plots/g__Snodgrassella_123_1.pdf")
density_plot_by_host("12_species_validation/g__Snodgrassella/g__Snodgrassella_124_0_perc_id.txt", "124_0")
  ggsave("Figures/20_SDP_density_plots/g__Snodgrassella_124_0.pdf")


grid.arrange(
  arrangeGrob(
    density_plot("12_species_validation/g__Zymobacter/g__Zymobacter_100_0_perc_id.txt", "100_0"),
    density_plot("12_species_validation/g__Zymobacter/g__Zymobacter_101_0_perc_id.txt", "101_0"),
    density_plot("12_species_validation/g__Zymobacter/g__Zymobacter_104_0_perc_id.txt", "104_0"),
    nrow = 2
  )
)
ggsave("Figures/20_SDP_density_plots/g__Zymobacter.pdf")
density_plot_by_host("12_species_validation/g__Zymobacter/g__Zymobacter_100_0_perc_id.txt", "100_0")
  ggsave("Figures/20_SDP_density_plots/g__Zymobacter_100_0.pdf")
density_plot_by_host("12_species_validation/g__Zymobacter/g__Zymobacter_101_0_perc_id.txt", "101_0")
  ggsave("Figures/20_SDP_density_plots/g__Zymobacter_101_0.pdf")
density_plot_by_host("12_species_validation/g__Zymobacter/g__Zymobacter_104_0_perc_id.txt", "104_0")
  ggsave("Figures/20_SDP_density_plots/g__Zymobacter_104_0.pdf")
```

```{r make_metadata_summary_MAGs}
# colnames(df_MAGs_metadata) <- c("Numbe_of_scaffolds", )
```
