---
title: "Honeybee cross-species analysis - 01"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Initialization

To intialize complete all of or relevant sections of `scripts/visualization/Load_data.R`.

## Plot number and Proportion mapped - host-filtering first

Number of reads that mapped to host and of those that did not, number that mapped to MAGs_db and total unmapped

```{r plot_MY_samples_1}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_hf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_hf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_hf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_hf" = "Unmapped",
                             "Host_mapped_hf" = "Mapped to Host",
                             "MAGs_DB_hf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +
  # new_scale_fill() +
  # geom_bar(data = df_reads_plot %>% filter(Type == "Raw"),
  #          aes(x=factor(Sample, levels = samples_chosen),
  #              y=Number
  #             ),
  #           stat="identity", position = "stack", 
  #           fill = NA,
  #           color = "black") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot %>% select(Host, Sample), qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_MY.pdf")
```

```{r plot_MY_samples_1_prop}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                group_by(Sample) %>%
                mutate(prop = Number/sum(Number)*100) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=prop,
               fill = factor(Type, levels = c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Proportion of reads (paired end)",
       fill = "Type") +
  scale_fill_manual(values=c("Unmapped_hf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_hf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_hf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_hf" = "Unmapped",
                             "Host_mapped_hf" = "Mapped to Host",
                              "MAGs_DB_hf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -5,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10,
                ymax = -5,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_MY_prop.pdf")
```

## Plot number and Proportion mapped - microbe-filtering first

```{r plot_MY_samples_1_mf}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_mf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_mf" = "Unmapped",
                             "Host_mapped_mf" = "Mapped to Host",
                             "MAGs_DB_mf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +
  # new_scale_fill() +
  # geom_bar(data = df_reads_plot %>% filter(Type == "Total_clean"),
  #          aes(x=factor(Sample, levels = samples_chosen),
  #              y=Number
  #             ),
  #           stat="identity", position = "stack", 
  #           fill = NA,
  #           color = "black") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_MY_mf.pdf")
```

```{r plot_MY_samples_1_prop_mf}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                group_by(Sample) %>%
                mutate(prop = Number/sum(Number)*100) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=prop,
               fill = factor(Type, levels = c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Proportion of reads (paired end)",
       fill = "Type") +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_mf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_mf" = "Unmapped",
                             "Host_mapped_mf" = "Mapped to Host",
                             "MAGs_DB_mf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -5,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10,
                ymax = -5,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_MY_prop_mf.pdf")
```

```{r plot_IN_MY_samples_1}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_hf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_hf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_hf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_hf" = "Unmapped",
                             "Host_mapped_hf" = "Mapped to Host",
                             "MAGs_DB_hf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free") +
  # new_scale_fill() +
  # geom_bar(data = df_reads_plot %>% filter(Type == "Raw"),
  #          aes(x=factor(Sample, levels = samples_chosen),
  #              y=Number
  #             ),
  #           stat="identity", position = "stack", 
  #           fill = NA,
  #           color = "black") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_IN_MY.pdf")
```

```{r plot_IN_MY_samples_1_prop}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                group_by(Sample) %>%
                mutate(prop = Number/sum(Number)*100) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=prop,
               fill = factor(Type, levels = c("Host_mapped_hf", "Unmapped_hf", "MAGs_DB_hf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Proportion of reads (paired end)",
       fill = "Type") +
  scale_fill_manual(values=c("Unmapped_hf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_hf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_hf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_hf" = "Unmapped",
                             "Host_mapped_hf" = "Mapped to Host",
                              "MAGs_DB_hf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -5,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10,
                ymax = -5,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_IN_MY_prop.pdf")
```

## Plot number and Proportion mapped - microbe-filtering first

```{r plot_IN_MY_samples_1_mf}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_mf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_mf" = "Unmapped",
                             "Host_mapped_mf" = "Mapped to Host",
                             "MAGs_DB_mf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free") +
  # new_scale_fill() +
  # geom_bar(data = df_reads_plot %>% filter(Type == "Total_clean"),
  #          aes(x=factor(Sample, levels = samples_chosen),
  #              y=Number
  #             ),
  #           stat="identity", position = "stack", 
  #           fill = NA,
  #           color = "black") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_IN_MY_mf.pdf")
```

```{r plot_IN_MY_samples_1_prop_mf}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf")) %>%
                mutate(Type = ifelse(Type == "Unmapped_host", "Unmapped", Type)) %>%
                group_by(Sample) %>%
                mutate(prop = Number/sum(Number)*100) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=prop,
               fill = factor(Type, levels = c("Host_mapped_mf", "Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Proportion of reads (paired end)",
       fill = "Type") +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "Host_mapped_mf" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_mf" = "Unmapped",
                             "Host_mapped_mf" = "Mapped to Host",
                             "MAGs_DB_mf" = "Mapped to MAGs-\n(after host filtering)"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -5,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10,
                ymax = -5,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_qpcr_numbers_IN_MY_prop_mf.pdf")
```

```{r plot_MY_samples_1_raw_cleaned}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Raw", "Total_clean")) %>%
                # mutate(Type = ifelse(Type == "Raw", "Total_clean", Type)) %>%
                group_by(Sample) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Raw", "Total_clean"))
              ),
           stat="identity", position = "dodge") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Raw" = brewer.pal(9, "Pastel1")[1],
                             "Total_clean" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Raw" = "Raw",
                             "Total_clean" = "Trimmed and clean"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x", ncol = 2) +
  new_scale_fill() +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_raw_trimmed.pdf")
```

# summarise sequencing depth

```{r}
df_reads_plot %>% filter(Sample %in% samples_MY & Type %in% c("Raw")) %>% group_by(Host) %>% summarise("Raw_mean" = format(mean(Number)/2, big.mark=","))
df_reads_plot %>% filter(Sample %in% samples_MY & Type %in% c("Total_clean")) %>% group_by(Host) %>% summarise("Clean_mean" = format(mean(Number)/2, big.mark=","))
df_reads_plot %>% filter(Sample %in% samples_MY & Type %in% c("MAGs_DB_mf")) %>% group_by(Host) %>% summarise("MAGs_mf" = format(mean(Number)/2, big.mark=","))
df_reads_plot %>% filter(Sample %in% samples_MY & Type %in% c("MAGs_DB_hf")) %>% group_by(Host) %>% summarise("MAGs_hf" = format(mean(Number)/2, big.mark=","))

df_reads_plot %>% filter(Sample %in% samples_IN & Type %in% c("Raw")) %>% group_by(Host) %>% summarise("Raw_mean" = format(mean(Number)/2, big.mark=","))
df_reads_plot %>% filter(Sample %in% samples_IN & Type %in% c("Total_clean")) %>% group_by(Host) %>% summarise("Clean_mean" = format(mean(Number)/2, big.mark=","))
df_reads_plot %>% filter(Sample %in% samples_IN & Type %in% c("MAGs_DB_mf")) %>% group_by(Host) %>% summarise("MAGs_mf" = format(mean(Number)/2, big.mark=","))
df_reads_plot %>% filter(Sample %in% samples_IN & Type %in% c("MAGs_DB_hf")) %>% group_by(Host) %>% summarise("MAGs_hf" = format(mean(Number)/2, big.mark=","))
```

```{r plot_MY_samples_1_raw_cleaned_mapped}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Raw", "Total_clean", "MAGs_DB_mf")) %>%
                # mutate(Type = ifelse(Type == "Raw", "Total_clean", "MAGs_DB_mf", Type)) %>%
                group_by(Sample) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Raw", "Total_clean", "MAGs_DB_mf"))
              ),
           stat="identity", position = "dodge") +
  labs(x = "Sample",
       y = "Proportion of reads (paired end)",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Raw" = brewer.pal(9, "Pastel1")[1],
                             "Total_clean" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Raw" = "Raw",
                             "Total_clean" = "Trimmed and clean",
                             "MAGs_DB_mf" = "Mapped to MAGs"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x", ncol = 2) +
  new_scale_fill() +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_raw_trimmed_mappedMAGs.pdf")
```

```{r plot_MY_samples_1_raw_cleaned_mapped}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("Raw", "Total_clean", "MAGs_DB_mf")) %>%
                # mutate(Type = ifelse(Type == "Raw", "Total_clean", "MAGs_DB_mf", Type)) %>%
                group_by(Sample) %>%
                ungroup() %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                mutate(Colony = Vectorize(get_colonyid_from_sample_name)(Sample)) 
ggplot() +
  geom_bar(data = df_to_plot,
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Raw", "Total_clean", "MAGs_DB_mf"))
              ),
           stat="identity", position = "dodge") +
  labs(x = "Sample",
       y = "Proportion of reads (paired end)",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Raw" = brewer.pal(9, "Pastel1")[1],
                             "Total_clean" = brewer.pal(9, "Pastel1")[2],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Raw" = "Raw",
                             "Total_clean" = "Trimmed and clean",
                             "MAGs_DB_mf" = "Mapped to MAGs"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x", ncol = 1) +
  new_scale_fill() +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Colony) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = Colony)
          ) +
  labs(fill = "Colony") +
  scale_fill_manual(values=brewer.pal(9, "Spectral")) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot, qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(setFill = F, leg_pos = "right",
             x_size = 0, modify_guide = F
            )
  ggsave("results/figures/01-Mapping_raw_trimmed_mappedMAGs_IN_MY.pdf")
```



```{r plot_GTF_sequencing_depths_1}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_mf", "Unmapped_mf")) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
ggplot() +
  geom_bar(data = filter(df_reads_plot, Sample %in% samples_chosen & Type %in% c("MAGs_DB_mf", "Unmapped_mf")),
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped_mf" = "Unmapped",
                             "MAGs_DB_mf" = "Mapped to MAGs"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +  
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, MAGs_DB_range) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10e+06,
                ymax = -20e+06,
                fill = factor(MAGs_DB_range, levels = c("Low", "Sufficient", "High")))
          ) +
  labs(fill = "#Mapped to MAGs DB") +
  scale_fill_manual(values=c("Low" = brewer.pal(9, "Paired")[6],
                             "Sufficient" = brewer.pal(9, "Paired")[3],
                             "High" = brewer.pal(9, "Paired")[4]
                            ),
                    labels=c("Low (< 15M)",
                             "Sufficient (15M - 30M)",
                             "High (> 30M)"
                            ),
                   ) +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Location) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = factor(Location, levels = location_order))
          ) +
  labs(fill = "Sampling location") +
  scale_fill_manual(values=location_order_color) +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Total_depth) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -30e+06,
                fill = factor(Total_depth, levels = c("Target_reached", "Insufficient")))
          ) +
  labs(fill = "#Mapped to MAGs DB") +
  scale_fill_manual(values=c("Target_reached" = brewer.pal(9, "Paired")[4],
                             "Insufficient" = brewer.pal(9, "Paired")[6]
                            ),
                    labels=c("Target reached (>50M)",
                             "Insufficient"
                            ),
                   ) +
  make_theme(theme_name=theme_few(), leg_pos="right",
             setFill = F, setCol = F,
             guide_nrow = 1,
             modify_guide = F,
             x_angle=45 ,x_vj=1.2, x_hj=1, x_size=0,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
  ggsave("results/figures/01-Sequencing_depth_summary_all.pdf")
```

```{r plot_GTF_sequencing_depths_2}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_mf", "Unmapped_mf")) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
ggplot() +
  geom_bar(data = filter(df_reads_plot, Sample %in% samples_chosen & Type %in% c("MAGs_DB_mf", "Unmapped_mf")),
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped",
                             "Mapped to MAGs"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +  
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, MAGs_DB_range) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10e+06,
                ymax = -20e+06,
                fill = factor(MAGs_DB_range, levels = c("Low", "Sufficient", "High")))
          ) +
  labs(fill = "#Mapped to MAGs DB") +
  scale_fill_manual(values=c("Low" = brewer.pal(9, "Paired")[6],
                             "Sufficient" = brewer.pal(9, "Paired")[3],
                             "High" = brewer.pal(9, "Paired")[4]
                            ),
                    labels=c("Low (< 15M)",
                             "Sufficient (15M - 30M)",
                             "High (> 30M)"
                            ),
                   ) +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Location) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = factor(Location, levels = location_order))
          ) +
  labs(fill = "Sampling location") +
  scale_fill_manual(values=location_order_color) +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Total_depth) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -30e+06,
                fill = factor(Total_depth, levels = c("Target_reached", "Insufficient")))
          ) +
  labs(fill = "#Mapped to MAGs DB") +
  scale_fill_manual(values=c("Target_reached" = brewer.pal(9, "Paired")[4],
                             "Insufficient" = brewer.pal(9, "Paired")[6]
                            ),
                    labels=c("Target reached (>30M)",
                             "Insufficient"
                            ),
                   ) +
  make_theme(theme_name=theme_few(), leg_pos="right",
             setFill = F, setCol = F,
             guide_nrow = 1,
             modify_guide = F,
             x_angle=45 ,x_vj=1.2, x_hj=1, x_size=5,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
  ggsave("results/figures/01-Sequencing_depth_summary_IN_MY.pdf")
```
```{r plot_GTF_sequencing_depths_3}
samples_chosen <- samples_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_mf", "Unmapped_mf")) %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
ggplot() +
  geom_bar(data = filter(df_reads_plot, Sample %in% samples_chosen & Type %in% c("MAGs_DB_mf", "Unmapped_mf")),
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("Unmapped_mf", "MAGs_DB_mf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("Unmapped_mf" = brewer.pal(9, "Pastel1")[1],
                             "MAGs_DB_mf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped",
                             "Mapped to MAGs"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free_x") +  
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, MAGs_DB_range) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10e+06,
                ymax = -20e+06,
                fill = factor(MAGs_DB_range, levels = c("Low", "Sufficient", "High")))
          ) +
  labs(fill = "#Mapped to MAGs DB") +
  scale_fill_manual(values=c("Low" = brewer.pal(9, "Paired")[6],
                             "Sufficient" = brewer.pal(9, "Paired")[3],
                             "High" = brewer.pal(9, "Paired")[4]
                            ),
                    labels=c("Low (< 15M)",
                             "Sufficient (15M - 30M)",
                             "High (> 30M)"
                            ),
                   ) +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Location) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0,
                ymax = -10e+06,
                fill = factor(Location, levels = location_order))
          ) +
  labs(fill = "Sampling location") +
  scale_fill_manual(values=location_order_color) +
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, Total_depth) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -20e+06,
                ymax = -30e+06,
                fill = factor(Total_depth, levels = c("Target_reached", "Insufficient")))
          ) +
  labs(fill = "#Mapped to MAGs DB") +
  scale_fill_manual(values=c("Target_reached" = brewer.pal(9, "Paired")[4],
                             "Insufficient" = brewer.pal(9, "Paired")[6]
                            ),
                    labels=c("Target reached (>30M)",
                             "Insufficient"
                            ),
                   ) +
  make_theme(theme_name=theme_few(), leg_pos="right",
             setFill = F, setCol = F,
             guide_nrow = 1,
             modify_guide = F,
             x_angle=45 ,x_vj=1.2, x_hj=1, x_size=5,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
  ggsave("results/figures/01-Sequencing_depth_summary_MY.pdf")
```

```{r plot_GTF_sequencing_depths_4}
samples_chosen <- samples_IN_MY
df_to_plot <- df_reads_plot %>%
                filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_hf", "Unmapped_hf", "Host_mapped_hf")) %>%
                pivot_wider(names_from = Type, values_from = Number) %>%
                mutate("non_MAG" = Unmapped_hf + Host_mapped_hf) %>%
                select(!Unmapped_hf) %>%
                select(!Host_mapped_hf) %>%
                pivot_longer(cols = c("MAGs_DB_hf", "non_MAG"), names_to = "Type", values_to = "Number") %>%
                mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
ggplot() +
  geom_bar(data = filter(df_to_plot, Sample %in% samples_chosen & Type %in% c("MAGs_DB_hf", "non_MAG")),
           aes(x=factor(Sample, levels = samples_chosen),
               y=Number,
               fill = factor(Type, levels = c("non_MAG", "MAGs_DB_hf"))
              ),
           stat="identity", position = "stack") +
  labs(x = "Sample",
       y = "Number of reads",
       fill = "Type") +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_fill_manual(values=c("non_MAG" = brewer.pal(9, "Pastel1")[1],
                             "MAGs_DB_hf" = brewer.pal(9, "Pastel1")[3]
                            ),
                    labels=c("Unmapped to MAGs",
                             "Mapped to MAGs"
                            )
                   ) +
  facet_wrap(~factor(Host, host_order), scale = "free") +  
  new_scale_fill() +
  geom_rect(data = left_join(df_to_plot, df_reads_updated) %>%
                    select(Sample, Host, MAGs_DB_range) %>%
                    unique() %>%
                    group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = -10e+06,
                ymax = -20e+06,
                fill = factor(MAGs_DB_range, levels = c("Low", "Sufficient", "High")))
          ) +
  geom_hline(yintercept = 10e+06, linetype = "dotted", color = "black") +
  labs(fill = "Reads Mapped to MAGs DB\n(after host filtering)") +
  scale_fill_manual(values=c("Low" = brewer.pal(9, "Paired")[6],
                             "Sufficient" = brewer.pal(9, "Paired")[3],
                             "High" = brewer.pal(9, "Paired")[4]
                            ),
                    labels=c("Low (< 15M)",
                             "Sufficient (15M - 30M)",
                             "High (> 30M)"
                            ),
                   ) +
  new_scale_fill() +
  labs(fill = "16S region copy numbers") +
  scale_fill_gradient(low = "#ffffd9", high = "#08519c",
                      trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
  geom_rect(data = left_join(df_to_plot %>% select(Host, Sample), qpcr_plot_df, by = c("Sample" = "Sample.Name", "Host")) %>%
                      select(Sample, Host, copy_num) %>%
                      unique() %>%
                      group_by(Host) %>%
                      mutate(SampleID = row_number()), 
            aes(xmin=SampleID - 0.5,
                xmax=SampleID + 0.5,
                ymin = 0e+06,
                ymax = -10e+06,
                fill = copy_num)
                ) +
  make_theme(theme_name=theme_few(), leg_pos="right",
             setFill = F, setCol = F,
             guide_nrow = 1,
             modify_guide = F,
             x_angle=45 ,x_vj=1.2, x_hj=1, x_size=0,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=12)
  ggsave("results/figures/01-Sequencing_depth_summary_IN_MY.pdf")
```

```{r}
df_reads_plot %>%
  filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_hf", "Unmapped_hf", "Host_mapped_hf")) %>%
  # filter(Type == "MAGs_DB_hf") %>% pull(Number, Sample) %>% view %>% format(big.mark=",")
  filter(Type == "MAGs_DB_hf") %>% pull(Number) %>% mean %>% format(big.mark=",")
df_reads_plot %>%
  filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_hf", "Unmapped_hf", "Host_mapped_hf")) %>%
  filter(Type == "MAGs_DB_hf") %>% pull(Number) %>% max %>% format(big.mark=",")
# mean 
# 19,202,788
# 53,253,634
# 2,252,750
# 4 samples less than this
df_reads_plot %>%
  # filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_hf", "Unmapped_hf", "Host_mapped_hf")) %>%
  filter(!is.na(Number)) %>%
  mutate(Number = as.numeric(Number)) %>%
  filter(Type == "Total_clean") %>% pull(Number) %>% mean() %>% format(big.mark=",")
df_reads_plot %>%
  # filter(Sample %in% samples_chosen & Type %in% c("MAGs_DB_hf", "Unmapped_hf", "Host_mapped_hf")) %>%
  filter(!is.na(Number)) %>%
  mutate(Number = as.numeric(Number)) %>%
  filter(Type == "Total_clean") %>% pull(Number) %>% min() %>% format(big.mark=",")
# 265,205,198
# 18,777,034
```