---
title: "Honeybee cross-species analysis - 06"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Initialization

To intialize complete all of or relevant sections of `scripts/visualization/Load_data.R`.

```{r}

```




# Go over all the below code for instrain plots again...
# instrain_mapping <- data.frame()
# for (sample in samples_IN) {
#   sample_mapping_scaffold_instrain <- read.csv(paste0("/scratch/aprasad/211018_Medgenome_india_samples/10_instrain/",sample,"_profile.IS/output/",sample,"_profile.IS_scaffold_info.tsv"), sep = "\t") %>%
#     mutate(ID = Vectorize(get_ID_from_scaffold)(scaffold)) %>%
#     select(ID, length, breadth) %>%
#       mutate(sample = sample)
#   instrain_mapping <- rbind(instrain_mapping, sample_mapping_instrain)
# }
ggplot() +
  geom_bar(data = plot_mapping_df_total %>% filter(Type %in% c("instrain_mapped", "instrain_summed")),
           aes(x = Reads,
               y = sample,
               fill = Type
           ),
           stat = "identity", position = "dodge"
  ) +
  # scale_fill_manual(values = genusColors) +
  make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 10,
                                         setFill = T, max_colors = 3,
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
    scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot() +
  geom_bar(data = plot_mapping_df_total %>% filter(Type %in% c("instrain_mapped", "Trimmed", "Mapped_microbiome")),
           aes(x = Reads,
               y = sample,
               fill = Type
           ),
           stat = "identity", position = "dodge"
  ) +
  # scale_fill_manual(values = genusColors) +
  make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 10,
                                         setFill = T, max_colors = 3,
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
    scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot() +
  geom_bar(data = plot_mapping_df,
           aes(x = Mapped,
               y = sample,
               fill = Genus
           ),
           stat = "identity", position = "stack"
  ) +
  # scale_fill_manual(values = genusColors) +
  make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 10,
                                         setFill = T, max_colors = length(unique(plot_mapping_df$Genus)),
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
    scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot() +
  geom_bar(data = plot_mapping_df,
           aes(x = Mapped,
               y = sample,
               fill = Genus
           ),
           stat = "identity", position = "stack"
  ) +
  geom_bar(data = plot_mapping_df_total  %>% filter(Type %in% c("Mapped_microbiome")),
           aes(x = Reads,
               y = sample
           ),
           stat = "identity", position ="stack", width = 0.5, fill = "black", alpha = 0.1
  ) +
  scale_fill_manual(values = extend_colors(unique(plot_mapping_df$Genus), genusColors)) +
  make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 10,
                                         setFill = F, max_colors = length(unique(plot_mapping_df$Genus)),
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
    scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
    
# ggplot() +
#   geom_bar(data = plot_mapping_df,
#            aes(x = Mapped,
#                y = sample,
#                fill = Genus
#            ),
#            stat = "identity", position = "stack"
#   ) +
#   # geom_bar(data = plot_mapping_df_total  %>% filter(Type %in% c("Mapped_microbiome")),
#   #          aes(x = Reads,
#   #              y = sample
#   #          ),
#   #          stat = "identity", position ="stack", width = 0.5, fill = "black", alpha = 0.25
#   # ) +
#   scale_fill_manual(values = extend_colors(unique(plot_mapping_df$Genus), genusColors)) +
#   make_theme(theme_name=theme_few(), leg_pos="bottom",
#                                          guide_nrow = 10,
#                                          setFill = F, max_colors = length(unique(plot_mapping_df$Genus)),
#                                          x_angle=45 ,x_vj=1.2, x_hj=1, x_size=12,
#                                          y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
#     scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot() +
  geom_bar(data = plot_mapping_df,
           aes(x = percent_mapped,
               y = sample,
               fill = Genus
           ),
           stat = "identity", position = "stack"
  ) +
  scale_fill_manual(values = extend_colors(unique(plot_mapping_df$Genus), genusColors)) +
  make_theme(theme_name=theme_few(), leg_pos="bottom",
                                         guide_nrow = 10,
                                         setFill = F, max_colors = length(unique(plot_mapping_df$Genus)),
                                         x_angle=45 ,x_vj=1.2, x_hj=1, x_size=12,
                                         y_angle=0 ,y_vj=0, y_hj=0, y_size=12) +
    scale_x_continuous(labels=unit_format(unit = "%"))


ggplot(coverage_df, aes(y = factor(cluster_name), x = factor(Sample, samples_IN_MY), fill = Cov)) +
                            geom_tile() +
                              labs(x = "Sample", y = "Cluster")+
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar", trans = "log10") +
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar", trans = "log10") +
                                make_theme(setFill = F, modify_guide = F, y_size = 8,
                                  x_size = 0, x_angle = 30, x_hj = 1 , x_vj = 1,
                                  y_hj =1, leg_pos = "right"
                                )
ggplot(coverage_df, aes(y = factor(cluster_name), x = factor(Host, host_order), fill = Prevalence_host)) +
                            geom_tile() +
                              labs(x = "Sample", y = "magOTU", fill = "Prevalence\n(Present = \nCov > 1)\n")+
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar") +
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                                coord_fixed(0.1) +
                                make_theme(setFill = F, modify_guide = F, y_size = 8,
                                  y_hj =1, leg_pos = "right"
                                )
                                ggsave("Figures/09-Prevalence_magOTU_by_host.pdf")

ggplot(coverage_df_genus %>% filter(Genus != "g__"), aes(y = factor(Genus, genera), x = factor(Host, host_order), fill = Prevalence_host)) +
                            geom_tile() +
                            geom_text(aes(label = ifelse(Mean > 0, Mean, ""))) +
                              labs(x = "Sample", y = "Genus", fill = "Prevalence\n(Present = \nCov > 1)\nTruncated mean\n(0.01)")+
                              # scale_fill_gradient(na.value = "transparent", low = "#fff5f0" , high = "#99000d", guide = "colourbar") +
                              scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar") +
                                guides(fill = guide_colourbar(nbin = 8)) +
                                make_theme(setFill = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "right", leg_size = 10
                                )

ggplot(coverage_df %>% filter(Genus != "g__") %>% mutate(Cov = round(Cov, 4)),
        aes(y = factor(Genus, genera), x = Cov,, color =  factor(Genus, genera))) +
                            geom_jitter() +
                            # geom_text(aes(label = Mean)) +
                              labs(x = "Coverage", y = "Cluster")+
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                              scale_color_manual(values = genusColors) +
                                # guides(fill = guide_colourbar(nbin = 8)) +
                                facet_wrap(~ factor(Host, host_order)) +
                                # scale_x_continuous(trans = "log", labels = scales::number_format(accuracy = 0.01)) +
                                make_theme(setCol = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "right", leg_size = 10
                                )

ggplot(coverage_df %>% filter(Genus != "g__") %>% mutate(Cov = round(Cov, 4)),
        aes(y = factor(Genus, genera), x = Cov,, color =  factor(Genus, genera))) +
                            geom_point() +
                            # geom_text(aes(label = Mean)) +
                              labs(x = "Coverage", y = "Cluster")+
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                              scale_color_manual(values = genusColors) +
                                # guides(fill = guide_colourbar(nbin = 8)) +
                                facet_wrap(~ factor(Host, host_order)) +
                                scale_x_continuous(trans = "log", labels = scales::number_format(accuracy = 0.01)) +
                                make_theme(setCol = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "right", leg_size = 10
                                )
ggplot(coverage_df %>% filter(Genus != "g__") %>% filter(Sample %in% samples_IN_MY) %>% mutate(Cov = round(Cov, 4)) %>% filter(Cov != 0),
        aes(y = factor(Genus, genera), x = Cov, shape = Location, color =  Location)) +
        # aes(y = factor(Genus, genera), x = Cov, shape = Location, color =  factor(Genus, genera))) +
                            geom_jitter() +
                            # geom_text(aes(label = Mean)) +
                              labs(x = "Coverage", y = "Cluster")+
                              # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
                            #   scale_color_manual(values = genusColors) +
                                # guides(fill = guide_colourbar(nbin = 8)) +
                                facet_wrap(~ factor(Host, host_order)) +
                                scale_x_continuous(trans = "log", labels = scales::number_format(accuracy = 0.01)) +
                                make_theme(setCol = F, modify_guide = F, y_size = 10,
                                  x_hj= 0.5, x_size = 10,
                                  y_hj =1, leg_pos = "right", leg_size = 10
                                )
                                ggsave(paste0("Figures/", "09-Coverage_scatter_location_compared.pdf"))

abundance_df <- coords_df %>%
  rename(Cluster = magOTU) %>%
    left_join(cluster_tax_info %>% filter(!is.na(cluster_name))) %>%
      mutate(Host = Vectorize(get_host_name)(Sample)) %>%
        mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
          mutate(MedianCov = ifelse(is.na(MedianCov), 0, MedianCov)) %>%
            select(Sample, Cov, cluster_name, Host) %>%
              left_join(select(df_reads, Trimmed, Sample), by = "Sample") %>%
                  rename(NumReads = Trimmed) %>%
                    mutate(CovNorm = Cov/NumReads)

abundance_df_matrix <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          mutate(PresAb = ifelse(CovNorm > 0, 1, 0)) %>%
                          select(Sample, cluster_name, PresAb) %>%
                            pivot_wider(names_from=cluster_name, values_from=PresAb, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

abundance_df_matrix_pa <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          mutate(PresAb = ifelse(Cov > 1, 1, 0)) %>%
                          select(Sample, cluster_name, PresAb) %>%
                            pivot_wider(names_from=cluster_name, values_from=PresAb, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

abundance_df_matrix_Cov_rel <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample") %>%
                              mutate(across(-1)*100/rowSums(across(-1)))

abundance_df_matrix_Cov <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                          filter(Sample %in% samples_IN) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

df_plot_cum_curve_m <- make_cum_curve(samples_am, df_magOTUs_vegan, 100, "Apis mellifera")
df_plot_cum_curve_c <- make_cum_curve(samples_ac, df_magOTUs_vegan, 100, "Apis cerana")
df_plot_cum_curve_d <- make_cum_curve(samples_ad, df_magOTUs_vegan, 100, "Apis dorsata")
df_plot_cum_curve_f <- make_cum_curve(samples_af, df_magOTUs_vegan, 100, "Apis florea")
df_plot_cum_curve_a <- make_cum_curve(samples_aa, df_magOTUs_vegan, 100, "Apis andreniformis")
df_plot_cum_curve <- rbind(df_plot_cum_curve_m, df_plot_cum_curve_c, df_plot_cum_curve_d, df_plot_cum_curve_f, df_plot_cum_curve_a)
ggplot(data = df_plot_cum_curve, aes(x = sample_size, y = number_of_clusters, color = factor(name, host_order))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(se = FALSE) +
                          labs(color = "Host species", x = "# Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color) +
                            make_theme(leg_pos = "bottom", setCol = F, guide_nrow = 1)


column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa[samples_IN,]) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color)
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa[samples_IN,]) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              col = list(Location = location_country_colors)
                          )
row_ha = rowAnnotation(Genus = colnames(abundance_df_matrix_pa[samples_IN,]) %>% as.data.frame() %>% left_join(cluster_tax_info %>% ungroup() %>% select(Genus, cluster_name), by = c(. = "cluster_name")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors))
                    )
pdf("Figures/09-presence_absence_heatmap_IN.pdf")

Heatmap(t(abundance_df_matrix_pa[samples_IN,]), 
        col = c("#ffffff", "#1a88c9"),
        # col = brewer.pal(2, "Pastel1"),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
        bottom_annotation = column_ba,
        column_names_gp = grid::gpar(fontsize = 0),
        heatmap_legend_param = list(title = "Present / Absent", color_bar = "discrete"),
          cluster_rows = F)
dev.off()
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa[samples_IN_MY,]) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color)
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa[samples_IN_MY,]) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              col = list(Location = location_country_colors)
                          )
row_ha = rowAnnotation(Genus = colnames(abundance_df_matrix_pa[samples_IN_MY,]) %>% as.data.frame() %>% left_join(cluster_tax_info %>% ungroup() %>% select(Genus, cluster_name), by = c(. = "cluster_name")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors))
                    )
pdf("Figures/09-presence_absence_heatmap_IN.pdf")

Heatmap(t(abundance_df_matrix_pa[samples_IN_MY,]), 
        col = c("#ffffff", "#1a88c9"),
        # col = brewer.pal(2, "Pastel1"),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
        bottom_annotation = column_ba,
        column_names_gp = grid::gpar(fontsize = 0),
        heatmap_legend_param = list(title = "Present / Absent", color_bar = "discrete"),
          cluster_rows = F)
dev.off()

column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color)
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              col = list(Location = location_country_colors)
                          )
row_ha = rowAnnotation(Genus = colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(cluster_tax_info %>% ungroup() %>% select(Genus, cluster_name), by = c(. = "cluster_name")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors))
                    )
pdf("Figures/09-presence_absence_heatmap.pdf")

Heatmap(t(abundance_df_matrix_pa), 
        col = c("#ffffff", "#1a88c9"),
        # col = brewer.pal(2, "Pastel1"),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
        bottom_annotation = column_ba,
        column_names_gp = grid::gpar(fontsize = 0),
        heatmap_legend_param = list(title = "Present / Absent", color_bar = "discrete"),
          cluster_rows = F)
dev.off()
# pdf("Figures/09-relative_abundance_heatmap.pdf")
Heatmap(t(abundance_df_matrix_Cov_rel),
        col = colorRamp2(c(0, 10, 100), colors = c("#ffffff", "#1a88c9", "#1a3869")),
        # col = colorRamp2(c(0, 10, 100), colors = c(brewer.pal(9, "BrBG")[9], brewer.pal(9, "BrBG")[5], brewer.pal(9, "BrBG")[1])),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
          cluster_rows = F)
# dev.off()
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_Cov) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color)
                          )
row_ha = rowAnnotation(Genus = colnames(abundance_df_matrix_Cov) %>% as.data.frame() %>% left_join(cluster_tax_info %>% ungroup() %>% select(Genus, cluster_name), by = c(. = "cluster_name")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors))
                    )
Heatmap(t(abundance_df_matrix_Cov),
        col = colorRamp2(c(0, 100, 1000), colors = c("#ffffff", "#1a88c9", "#08306b")),
        # col = colorRamp2(c(0, 10, 100), colors = c(brewer.pal(9, "BrBG")[9], brewer.pal(9, "BrBG")[5], brewer.pal(9, "BrBG")[1])),
        clustering_method_columns = 'ward.D2',
        top_annotation = column_ha, right_annotation = row_ha,
          cluster_rows = F)

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
        filter(Location %in% c("Malaysia", "India")) %>%
          left_join(as.data.frame(rowSums(data_otu)) %>% setNames("cov_sums") %>% rownames_to_column("Sample"), by = "Sample"),
       aes(x = MAGs_DB, y = cov_sums, 
           color = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads", y = "Sum of all coverages", color = "Host") +
    geom_point(size = 3) +
      make_theme(setCol = F) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot(df_reads %>% filter((Sample %in% samples)) %>% 
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
        ) +
    geom_point(aes(x = factor(Sample, samples),
                   y = MAGs_DB,
                   color = Location,
                   shape = factor(Host, host_order))) +
    geom_hline(yintercept = 1e+06) +
    # scale_color_manual(values=host_order_color_dark) +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(setCol = F)

ggplot(df_reads %>% filter((Sample %in% samples)) %>% 
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
        ) +
    geom_point(aes(x = factor(Sample, samples),
                   y = MAGs_DB,
                   shape = Location,
                   color = factor(Host, host_order))) +
    geom_hline(yintercept = 1e+06) +
    scale_color_manual(values=host_order_color_dark) +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(setCol = F)


ggplot(df_reads %>%
        select(Sample, MAGs_DB) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% ungroup() %>% select(Sample, Cov, Genus, cluster_name) %>% filter(Cov > 0) %>% filter(!is.na(cluster_name)) %>% group_by(Sample, Genus) %>%  mutate(Sample, Genus, genus_Cov = sum(Cov))),
       aes(x = MAGs_DB, y = genus_Cov, 
        #    color = Location)
        #    color = factor(Genus, genera), shape = Location)
           color = factor(Host, host_order), shape = Location)
    ) +
    # facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads", y = "Sum of all coverages", color = "Host") +
    geom_point(size = 3) +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(setCol = F, guide_nrow = 5,
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Lactobacillus")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Lactobacillus coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Lactobacillus.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Bombilactobacillus")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Bombilactobacillus coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Bombilactobacillus.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Bifidobacterium")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Bifidobacterium coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Bifidobacterium.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Gilliamella")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Gilliamella coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Gilliamella.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Snodgrassella")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Snodgrassella coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Snodgrassella.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Apibacter")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Apibacter coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Apibacter.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Frischella")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Frischella coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Frischella.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Dysgonomonas")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Dysgonomonas coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          scale_size(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Dysgonomonas.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Commensalibacter")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Commensalibacter coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          scale_size(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Commensalibacter.pdf")


ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Pectinatus")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Pectinatus coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          scale_size(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Pectinatus.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Bartonella")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Bartonella coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          scale_size(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Bartonella.pdf")
# coverage_df %>% filter(grepl("kun", Species)) %>% pull(Species)

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__Bombella")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__Bombella coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          scale_size(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__Bombella.pdf")

ggplot(df_reads %>%
        select(Sample, MAGs_DB, Trimmed) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
          left_join(coverage_df %>% 
                    ungroup() %>%
                    select(Sample, Cov, Genus, cluster_name) %>%
                    filter(Cov > 0)%>%
                    filter(!is.na(cluster_name)) %>%
                    group_by(Sample, Genus) %>% 
                    mutate(Sample, Genus, genus_Cov = sum(Cov)) %>%
                    filter(Genus == "g__WRHT01")
                    ),
       aes(x = MAGs_DB, y = genus_Cov, size = Trimmed,
           color = Location)
        #    color = factor(Genus, genera), shape = Location)
        #    color = factor(Host, host_order), shape = Location)
    ) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x = "Numer of MAGs DB mapped reads",
         y = "Sum of all coverages",
         color = "Host",
         size = "Number of trimmed reads",
         title = "g__WRHT01 coverage summed") +
    geom_point() +
    geom_hline(yintercept = 0) +
    geom_hline(yintercept = 1) +
    geom_hline(yintercept = 5, linetype = "dashed") +
    geom_hline(yintercept = 10, linetype = "dotted") +
      make_theme(guide_nrow = 3, palettecolor = "Set1",
                 x_angle = 0, x_size = 7) +
        # scale_color_manual(values=genusColors) +
        # scale_color_manual(values=host_order_color_dark) +
          scale_y_continuous(trans = "log10") +
          scale_size(labels=unit_format(unit = "M", scale = 1e-6)) +
          scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))
          scale_size(labels=unit_format(unit = "M", scale = 1e-6))
          ggsave("Figures/09-Coverage_vs_mapped_reads_g__WRHT01.pdf")

# vis_magOTUs_df %>% 
#   select(ID, Host, Genus, Cluster, length, score, furthest_cluster_member, score_representative, Num_mags) %>%
#     filter(Genus != "g__") %>%
#     ungroup() %>%
#       group_by(Cluster) %>%
#         mutate(Num_host = n_distinct(Host)) %>%
#           filter(Num_host >= 2) %>%
#             filter(Num_mags >= 3) %>%
#               summarise(Cluster, Genus, Num_mags, Num_host) %>% unique()
# vis_magOTUs_df %>% 
#   select(ID, Host, Genus, Cluster, length, score, furthest_cluster_member, score_representative, Num_mags) %>%
#     filter(Cluster == "33_1")

# code for betapart functions from
# https://github.com/cran/betapart

# Latest commit a222c6d on Feb 24, 2012
betapart.core <- function(x){
	
	# test for a binary numeric matrix or data.frame
	if(! is.matrix(x)){
		x<-as.matrix(x)
  	}
	
	if(! is.numeric(x))
    	stop("The data in x is not numeric.",call.=TRUE)
	
	# simple test for binary data
	xvals <-  unique(as.vector(x))
	if (any(!is.element(xvals, c(0, 1)))) 
        stop("The table contains values other than 0 and 1: data should be presence/absence.", call. = TRUE)

	shared <- x %*% t(x)
    not.shared <-  abs(sweep(shared, 2, diag(shared)))
		
	sumSi <- sum(diag(shared)) # species by site richness
    St <- sum(colSums(x) > 0)  # regional species richness
    a <- sumSi - St            # multi site shared species term

    sum.not.shared <- not.shared + t(not.shared)
    max.not.shared <- pmax(not.shared, t(not.shared))
    min.not.shared <- pmin(not.shared, t(not.shared))

	computations<-list(data=x, sumSi=sumSi, St=St, a=a, shared=shared, not.shared=not.shared, 
	                   sum.not.shared=sum.not.shared, max.not.shared=max.not.shared, 
	                   min.not.shared=min.not.shared)
    class(computations)<-"betapart"

	return(computations)
} 
# Latest commit 4b59d9c on Feb 24, 2012
beta.pair <- function(x, index.family="sorensen"){
	
	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
			beta.sim <- x$min.not.shared / (x$min.not.shared + x$shared)
			beta.sne <- ((x$max.not.shared - x$min.not.shared) / ((2 * x$shared) + x$sum.not.shared)) * (x$shared / (x$min.not.shared + x$shared))
            beta.sor <- x$sum.not.shared / (2 * x$shared + x$sum.not.shared)
                
            pairwise <- list(beta.sim=as.dist(beta.sim), beta.sne=as.dist(beta.sne),beta.sor=as.dist(beta.sor))},
		jaccard = {
			beta.jtu <- (2*x$min.not.shared) / ((2*x$min.not.shared) + x$shared)
	        beta.jne <- ((x$max.not.shared - x$min.not.shared) / (x$shared + x$sum.not.shared)) * (x$shared / ((2*x$min.not.shared) + x$shared))
	        beta.jac <- x$sum.not.shared / (x$shared + x$sum.not.shared)

	        pairwise <- list(beta.jtu=as.dist(beta.jtu), beta.jne=as.dist(beta.jne),beta.jac=as.dist(beta.jac))})

	return(pairwise)
}
beta.multi <- function(x, index.family="sorensen"){

	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}

	maxbibj <- sum(x$max.not.shared[lower.tri(x$max.not.shared)])
    minbibj <- sum(x$min.not.shared[lower.tri(x$min.not.shared)])
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
            beta.sim <- minbibj / (minbibj + x$a)
            beta.sne <- (x$a / (minbibj + x$a)) * ((maxbibj - minbibj) / ((2 * x$a) + maxbibj + minbibj))
            beta.sor <- (minbibj + maxbibj) / (minbibj + maxbibj + (2 * x$a))

           	multi <- list(beta.SIM=beta.sim, beta.SNE=beta.sne,beta.SOR=beta.sor)},
		jaccard = {
            beta.jtu <- (2*minbibj) / ((2*minbibj) + x$a)
            beta.jne <- (x$a / ((2*minbibj) + x$a)) * ((maxbibj - minbibj) / ((x$a) + maxbibj + minbibj))
            beta.jac <- (minbibj + maxbibj) / (minbibj + maxbibj + x$a)

           	multi <- list(beta.JTU=beta.jtu, beta.JNE=beta.jne, beta.JAC=beta.jac)})

	return(multi)

}
deep_samples <- df_reads %>% filter(Trimmed > 10e+6) %>% pull(Sample)
length(deep_samples)
df_magOTUs_vegan <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0) %>%
                        filter(row.names(.) %in% deep_samples)
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan, "sorensen")
str(dist_matrix_betapart)
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(dist_matrix_betapart$beta.sne, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(dist_matrix_betapart$beta.sim, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

dist_matrix_betapart <- beta.pair(df_magOTUs_vegan, "jaccard")
pcoa_plot(dist_matrix_betapart$beta.jac, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(dist_matrix_betapart$beta.jne, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(dist_matrix_betapart$beta.jtu, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

dist_matrix_betapart_multi <- beta.multi(df_magOTUs_vegan, "sorensen")
str(dist_matrix_betapart_multi)

# beta.sim dist object, dissimilarity matrix accounting for spatial turnover (replacement),
# measured as Simpson pair-wise dissimilarity
# beta.sne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Sorensen pair-wise dissimilarity
# beta.sor dist object, dissimilarity matrix accounting for total dissimilarity, measured as
# Sorensen pair-wise dissimilarity (a monotonic transformation of beta diversity)
# For index.family="jaccard" the three matrices are:
# beta.jtu dist dissimilarity matrix accounting for spatial turnover, measured as the turnoverfraction of Jaccard pair-wise dissimilarity
# beta.jne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Jaccard pair-wise dissimilarity
# beta.jac dist object, dissimilarity matrix accounting for beta diversity, measured as Jaccard pair-wise dissimilarity (a monotonic transformation of beta diversity)

grouping_vec <- df_meta_complete %>% filter(ID %in% rownames(df_magOTUs_vegan)) %>% pull(Species)
anosim(dist_matrix_betapart$beta.jtu, grouping_vec, distance = "jaccard", permutations = 9999)

df_to_test <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
                        # filter(rownames(.) %in% c(samples_IN, samples_MY))
                          # filter(rownames(.) %in% c(samples_am, samples_ac, samples_af, samples_ad))
adonis2(df_to_test ~ Species + Country, data = df_meta_complete %>% filter(ID %in% rownames(df_to_test)))

# distances from tree in this paper. Measured by eye for now as the number of nodes separating them.

get_host_divergence <- function(sample_1, sample_2) {
  host_1 <- get_host_from_sample_name(sample_1)
  host_1 <- strsplit(host_1, " ")[[1]][[2]]
  host_2 <- get_host_from_sample_name(sample_2)
  host_2 <- strsplit(host_2, " ")[[1]][[2]]
  mellifera <- c(0, 1, 2, 4, 4)
  cerana <- c(1, 0, 2, 4, 4)
  dorsata <- c(2, 2, 0, 3, 3)
  florea <- c(4, 4, 3, 0, 1)
  andreniformis <- c(4, 4, 3, 1, 0)
  host_divergence_mat <- rbind(mellifera, cerana, dorsata, florea, andreniformis)
  colnames(host_divergence_mat) <- c("mellifera", "cerana", "dorsata", "florea", "andreniformis")
  index_list <- list("mellifera" = 1, "cerana" = 2, "dorsata" = 3, "florea" = 4, "andreniformis" = 5)
  host_divergence <- host_divergence_mat[as.numeric(index_list[host_1]), as.numeric(index_list[host_2])]
  return(host_divergence)
}
df_magOTUs_vegan_IN_MY <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0) %>%
                      filter()
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_IN_MY, "jaccard")
sample_bac_div <- dist_matrix_betapart$beta.jac
sample_host_div <- data.frame()
for (sample in samples) {
  sample_host_div <- rbind(sample_host_div, rep(0, length((samples))))
}
colnames(sample_host_div) <- samples
rownames(sample_host_div) <- samples
for (sample_1 in samples) {
  for (sample_2 in samples) {
    ind_1 <- as.numeric(which(samples == sample_1))
    ind_2 <- as.numeric(which(samples == sample_2))
    sample_host_div[ind_1, ind_2] <- get_host_divergence(sample_1, sample_2)
  }
}

# samples_subset_am <- df_meta_complete %>% filter(Species == "Apis mellifera") %>% pull(ID)
# samples_subset_ac <- df_meta_complete %>% filter(Species == "Apis cerana") %>% pull(ID)
# samples_subset_ad <- df_meta_complete %>% filter(Species == "Apis dorsata") %>% pull(ID)
# samples_subset_af <- df_meta_complete %>% filter(Species == "Apis florea") %>% pull(ID)
# samples_subset_aa <- df_meta_complete %>% filter(Species == "Apis andreniformis") %>% pull(ID)


sample_host_div <- sample_host_div %>% 
  filter(row.names(.) %in% as.vector(as.matrix(sample_bac_div) %>% row.names))


sample_host_div_df <- sample_host_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                        pivot_longer(!Sample, names_to = "Compared", values_to = "dist_h")
sample_bac_div_df <- sample_bac_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                          pivot_longer(!Sample, names_to = "Compared", values_to = "dist_b") %>%
                          mutate(Host_s = Vectorize(get_host_from_sample_name)(Sample)) %>%
                          mutate(Host_c = Vectorize(get_host_from_sample_name)(Compared)) %>%
                          mutate(Type_b = ifelse(Host_s == Host_c, "Within", "Between")) %>%
                          mutate(Hosts_compared = paste0(Host_s, "_", Host_c))
dissimilarities_df_median <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, median_b = median(dist_b)) %>%
                        unique

# dissimilarities_df <- left_join(sample_bac_div_df, sample_host_div_df) %>%
#                         filter(Sample %in% c("M1.2", "C1.2", "D1.2", "F1.2", "A1.2"))

# mantel(sample_host_div, sample_bac_div, method="pearson", permutations=1000)

ggplot() +
  geom_point(data=dissimilarities_df,
             aes(x = factor(dist_h),
                 y = dist_b
                #  color = Host_s,
                #  shape = Type_b
             )
            ) +
  geom_point(data=dissimilarities_df_median,
             aes(x = factor(dist_h),
                 y = median_b
             ), size = 5, color = "red"
            ) +
    labs(x = "Host divergence (# nodes between)", y = "Microbiome dissimilarity (jaccard)") +
    # scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 5)

cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

split_by_pattern <- function(my_name, pattern, index) {
  return(strsplit(my_name, pattern)[[1]][[index]])
}
bac_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(dist_h, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = median_b) %>%
                    column_to_rownames("Compared1")
host_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(median_b, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = dist_h) %>%
                    column_to_rownames("Compared1")

bac_median_mat[1,]
host_median_mat[1,]
mantel(as.dist(bac_median_mat), as.dist(host_median_mat), method = "pearson", permutations = 9999)
mantel.test(bac_median_mat, host_median_mat, method = "pearson", permutations = 9999)
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")$estimate

pcoa_plot(as.dist(sample_host_div), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
