---
title: "Honeybee cross-species analysis - 09"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Initialization

To intialize complete all of or relevant sections of `scripts/visualization/Load_data.R`.

Due to large differences in total reads, plotting number of filtered pairs directly does not work.
Plotting percentage after normalizing across magOTUs using the factor `length*breadth` is more meaningful

```{r}
system("mkdir -p results/figures/10-instrain_plots/")
```

```{r instrain_plots_breadth_by_sample}
plot_breadth_by_sample <- ggplot(instrain_genomes_info_df %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100),
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Genus,
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~sample, ncol = 3, nrow = 3, page = 6) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_abline(slope=1, linetype = "dotted") +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=genusColors)

paginate_save(plot_breadth_by_sample, "sample", "results/figures/10-instrain_plots/1-breadth_vs_expected_by_sample.pdf", pass_nrow = 3, pass_ncol = 3)
```

```{r instrain_plots_breadth_by_magOTU}
plot_breadth_by_magOTU <- ggplot(instrain_genomes_info_df %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100),
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Host,
                              )
                            ) +
  geom_point() +
    facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 5) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_abline(slope=1, linetype = "dotted") +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=host_order_color_dark)

paginate_save(plot_breadth_by_magOTU, "magOTU", "results/figures/10-instrain_plots/1-breadth_vs_expected_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```
```{r instrain_plots_breadth_vs_num_reads}
plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth
                              ), alpha = 0.4, color = "black"
                            ) +
  geom_point(data = instrain_breadth_depth %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = Host
                              )
                            ) +
  scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 5,
                 setFill = F, setCol = F,
                 x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4)) +
        scale_color_manual(values=host_order_color_dark)

paginate_save(plot_breadth_num_reads, "magOTU", "results/figures/10-instrain_plots/2-mapped_reads_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```

```{r instrain_plots_breadth_vs_num_reads_coverage}
plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = coverage_factor
                              )
                            ) +
  geom_point(data = instrain_breadth_depth %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = coverage_factor,
                              )
                            ) +
  geom_point(data = instrain_breadth_depth %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                              ), shape = 1, color = "black", stroke = 0.25
                            ) +
  scale_color_manual(values = c("<0.0001" = "#a50f15",
                                "0.0001 - 0.001" = "#d73027",
                                "0.001 - 0.01" = "#f46d43",
                                "0.01 - 0.1" = "#fdae61",
                                "0.1 - 1" = "#fee08b",
                                "1 - 10" = "#a6d96a", 
                                "10 - 100" = "#74c476", 
                                "100 - 1000" = "#41ab5d", 
                                ">1000" = "#238b45"
                              )
                          ) +
  scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 9,
                 setFill = F, setCol = F, palettecolor = "RdYlGn",
                 x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        scale_size(range = c(1,4))

paginate_save(plot_breadth_num_reads, "magOTU", "results/figures/10-instrain_plots/2-mapped_reads_by_magOTU_w_coverage.pdf", pass_nrow = 3, pass_ncol = 3)
```

# plot prevalence of magOTUs and genera

```{r}
ggplot(instrain_prev_by_genus, aes(y = factor(Genus, genera), 
                          x = factor(Host, host_order),
                          fill = factor(presence_in_host))) +
  geom_tile(color="black") +
    labs(x = "Host", y = "Genus", fill = "Prevalence (%)") +
      make_theme(setFill = T, modify_guide = F, y_size = 12,
                 x_size = 14, x_angle = 0, x_hj = 0.5 , x_vj = 1,
                 y_hj = 1, leg_pos = "right", leg_size = 11)
      ggsave("results/figures/3-presence_of_genus.pdf")

ggplot(instrain_prev, aes(y = magOTU, 
                          x = factor(Host, host_order),
                          fill = prevalence_in_host)) +
  geom_tile() +
    labs(x = "Host", y = "Genus", fill = "Prevalence (%)") +
    scale_fill_gradient2(midpoint = 0.4, low = "#f7fbff",  mid = "#6baed6", high = "#08306b", guide = "colourbar", na.value = "#f7fbff") +
      make_theme(setFill = F, modify_guide = F, y_size = 12,
                 x_size = 14, x_angle = 30, x_hj = 1 , x_vj = 1,
                 y_hj = 1, leg_pos = "right", leg_size = 11)
      ggsave("results/figures/3-prevalence_in_host_of_magOTU.pdf")

prev_vs_cov <- ggplot(instrain_prev) +
  geom_point(aes(y = coverage, x = prevalence_in_host, color = Host), size = 0.2) +
    scale_color_manual(values=host_order_color_dark) +
    scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap(~Genus) +
    make_theme(setFill = F, setCol = F, 
               x_angle = 30, x_hj = 1, x_vj = 1)
  ggsave("results/figures/3-prevalence_vs_coverage.pdf")

prev_vs_cov_magOTU <- ggplot(instrain_prev) +
  geom_point(aes(y = coverage, x = prevalence_in_host, color = Host), size = 0.2) +
    scale_color_manual(values=host_order_color_dark) +
    scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~magOTU, ncol = 4, nrow = 4, page = 1) +
    make_theme(setFill = F, setCol = F, 
               x_angle = 30, x_hj = 1, x_vj = 1)
  # ggsave("results/figures/3-prevalence_vs_coverage_magOTU.pdf")
paginate_save(prev_vs_cov_magOTU, "magOTU", "results/figures/3-prevalence_vs_coverage_magOTU.pdf", pass_nrow = 4, pass_ncol = 4)
```

# Estimate host range
```{r}
#  update this with better estimates and summaries!
estimate_shannon <- function(pa_vector) {
  return(diversity(pa_vector, index = "shannon"))
}
get_host_range_magOTU <- function(magotu) {
  # get a shannon estimate for each host
  # and take their mean to estimate host range
  # shannon is high if Shannon diversity index 
  # will be higher when the vector has a 
  # more equal distribution of 1s and 0s
  # this way instead of a 1 or 0 read out 
  # of presence of magotu in the host, we
  # get a more nuanced estimate of host range
  # which is more sensitive to the number of
  # samples in a magotu is present in
  # so a low prevalence one that is present across
  # all hosts but in very few sample
  # will be considered to have a lower host range
  # than one that is present in all hosts
  # and in many samples
  shannon <- sapply(host_order, function(host) {
    pa_vec <- instrain_prev[instrain_prev$magOTU == magotu & instrain_prev$Host == host, "detected"]
    estimate_shannon(pa_vec)
  })
  host_range = mean(shannon)
  return(host_range)
}

# host <- "Apis mellifera"
# magotu <- "g__Lactobacillus_17_1-M1-2_1"
# instrain_prev[instrain_prev$magOTU == magotu & instrain_prev$Host == host, "detected"]
# sum(instrain_prev[instrain_prev$magOTU == magotu & instrain_prev$Host == host, "detected"])
# estimate_shannon(instrain_prev[instrain_prev$magOTU == magotu & instrain_prev$Host == host, "detected"])

get_host_range_magOTU_naive <- function(magotu) {
  detected <- sapply(host_order, function(host) {
    pa_vec <- instrain_prev[instrain_prev$magOTU == magotu & instrain_prev$Host == host, "detected"]
    sum(pa_vec) > 0
  })
  host_range = sum(as.numeric(detected))
  return(host_range)
}

instrain_host_range <- instrain_prev %>%
                          select(magOTU, Genus) %>%
                          unique() %>%
                          summarise(Genus, host_range = Vectorize(get_host_range_magOTU)(magOTU)) %>%
                          mutate(host_range_naive = Vectorize(get_host_range_magOTU_naive)(magOTU))

ggplot(instrain_host_range) +
  geom_bar(aes(x = host_range, y = magOTU, fill = Genus), stat = "identity") +
  # geom_point(aes(x = host_range_naive, y = magOTU, color = Genus), shape=3, size = 3) +
  geom_point(aes(x = host_range_naive, y = magOTU, color = Genus), size = 3) +
    labs(x = "Host range", y = "Number of magOTUs") +
    scale_color_manual(values=genusColors) +
    scale_fill_manual(values=genusColors_char) +
    make_theme(setFill = F, setCol = F)
  ggsave("results/figures/3-host_range.pdf")

library(ggalluvial)


# Duplicate rows based on sample count for each host


allu_df <- instrain_prev %>%
            group_by(magOTU, Host) %>%
            summarise(magOTU, Host, Genus, propSamples = sum(detected)/n(), numSamples = sum(detected)) %>%
            arrange(magOTU, Host) %>%
            filter(!is.na(Genus)) %>%
            unique()
View(allu_df)
ggplot(allu_df %>%
                      filter(propSamples >= 0.5),
      # aes(axis1 = magOTU, axis2 = Genus, axis3 = factor(Host, host_order), y = propSamples)) +
      aes(axis1 = factor(magOTU, colnames(abundance_df_matrix_pa)), axis2 = factor(Host, host_order), y = propSamples)) +
  geom_alluvium(aes(fill = factor(Genus, genera)), color = "black", size = .5) +
  ggtitle("Host range of magOTUs (>= 0.2 prevalence within host)") +
  scale_fill_manual(values=genusColors_char) +
  labs(y = "Number of samples", fill = "Genus") +
  ggnewscale::new_scale_fill() +
  geom_stratum(aes(fill = Host)) +
  scale_fill_manual(values=host_order_color) +
  # facet_wrap(~factor(Genus, genera), scales = "free") +
  geom_text(stat = "stratum", aes(label = ifelse(after_stat(x) == 2, as.character(after_stat(stratum)), NA))) +
  make_theme(setFill = F, leg_pos = "right", guide_nrow = 10) +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())
      ggsave("results/figures/3-host_range_alluvial_filt_50.pdf")

ggplot(allu_df %>%
                      filter(propSamples >= 0.2),
      # aes(axis1 = magOTU, axis2 = Genus, axis3 = factor(Host, host_order), y = propSamples)) +
      aes(axis1 = factor(magOTU, colnames(abundance_df_matrix_pa)), axis2 = factor(Host, host_order), y = propSamples)) +
  geom_alluvium(aes(fill = factor(Genus, genera)), color = "black", size = .5) +
  ggtitle("Host range of magOTUs (>= 0.2 prevalence within host)") +
  scale_fill_manual(values=genusColors_char) +
  labs(y = "Number of samples", fill = "Genus") +
  ggnewscale::new_scale_fill() +
  geom_stratum(aes(fill = Host)) +
  scale_fill_manual(values=host_order_color) +
  # facet_wrap(~factor(Genus, genera), scales = "free") +
  geom_text(stat = "stratum", aes(label = ifelse(after_stat(x) == 2, as.character(after_stat(stratum)), NA))) +
  make_theme(setFill = F, leg_pos = "right", guide_nrow = 10) +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())
      ggsave("results/figures/3-host_range_alluvial_filt_20.pdf")
system("mkdir -p results/figures/host_range_alluvials")

for (genus in genera) {
  ggplot(allu_df %>% filter(propSamples >= 0.2) %>%
                      filter(Genus == genus),
      # aes(axis1 = magOTU, axis2 = Genus, axis3 = factor(Host, host_order), y = propSamples)) +
      aes(axis1 = factor(magOTU, colnames(abundance_df_matrix_pa)), axis2 = factor(Host, host_order), y = propSamples)) +
  geom_alluvium(aes(fill = factor(Genus, genera)), color = "black", size = .5) +
  ggtitle("Host range of magOTUs (>= 0.2 prevalence within host)") +
  scale_fill_manual(values=genusColors_char) +
  labs(y = "Number of samples", fill = "Genus") +
  ggnewscale::new_scale_fill() +
  geom_stratum(aes(fill = Host)) +
  scale_fill_manual(values=host_order_color) +
  # facet_wrap(~factor(Genus, genera), scales = "free") +
  geom_text(stat = "stratum", aes(label = ifelse(after_stat(x) != 20, as.character(after_stat(stratum)), NA))) +
  make_theme(setFill = F, leg_pos = "right", guide_nrow = 10) +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())
      ggsave(paste0("results/figures/host_range_alluvials/3-host_range_alluvial_filt_20_",genus,".pdf"))
}

for (genus in genera) {
  ggplot(allu_df %>% filter(propSamples >= 0.2) %>%
                      filter(Genus == genus),
      # aes(axis1 = magOTU, axis2 = Genus, axis3 = factor(Host, host_order), y = propSamples)) +
      aes(axis1 = factor(magOTU, colnames(abundance_df_matrix_pa)), axis2 = factor(Host, host_order), y = propSamples)) +
  geom_alluvium(aes(fill = factor(Genus, genera)), color = "black", size = .5) +
  ggtitle("Host range of magOTUs (>= 0.2 prevalence within host)") +
  scale_fill_manual(values=genusColors_char) +
  labs(y = "Number of samples", fill = "Genus") +
  ggnewscale::new_scale_fill() +
  geom_stratum(aes(fill = Host)) +
  scale_fill_manual(values=host_order_color) +
  # facet_wrap(~factor(Genus, genera), scales = "free") +
  geom_text(stat = "stratum", aes(label = ifelse(after_stat(x) == 2, as.character(after_stat(stratum)), NA))) +
  make_theme(setFill = F, leg_pos = "right", guide_nrow = 10) +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())
      ggsave(paste0("results/figures/host_range_alluvials/3-host_range_alluvial_filt_20_",genus,"_wo_labels.pdf"))
}

ggplot(allu_df %>%
                      filter(propSamples >= 0.5),
      # aes(axis1 = magOTU, axis2 = Genus, axis3 = factor(Host, host_order), y = propSamples)) +
      aes(axis1 = factor(magOTU, colnames(abundance_df_matrix_pa)), axis2 = factor(Host, host_order), y = propSamples)) +
  geom_alluvium(aes(fill = factor(Genus, genera)), color = "black", size = .5) +
  ggtitle("Host range of magOTUs (>= 0.5 prevalence within host)") +
  scale_fill_manual(values=genusColors_char) +
  labs(y = "Number of samples", fill = "Genus") +
  ggnewscale::new_scale_fill() +
  geom_stratum(aes(fill = Host)) +
  scale_fill_manual(values=host_order_color) +
  # facet_wrap(~factor(Genus, genera), scales = "free") +
  geom_text(stat = "stratum", aes(label = ifelse(after_stat(x) == 2, as.character(after_stat(stratum)), NA))) +
  make_theme(setFill = F, leg_pos = "right", guide_nrow = 10) +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())
      ggsave("results/figures/3-host_range_alluvial_core_filt.pdf")

ggplot(allu_df, aes(axis1 = magOTU, axis2 = factor(Host, host_order), y = numSamples)) +
  geom_alluvium(aes(fill = factor(Genus, genera)), color = "black", size = .5) +
  scale_fill_manual(values=genusColors_char) +
  labs(x = "magOTU", y = "Number of samples") +
  ggnewscale::new_scale_fill() +
  geom_stratum(aes(fill = Host)) +
  scale_fill_manual(values=host_order_color) +
  # facet_wrap(~factor(Genus, genera), scales = "free") +
  geom_text(stat = "stratum", aes(label = ifelse(after_stat(x) == 2, as.character(after_stat(stratum)), NA))) +
  make_theme(setFill = F, leg_pos = "right", guide_nrow = 10) +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())
      ggsave("results/figures/3-host_range_alluvial_all.pdf")
```

```

# plot presence of magOTUs and genera

# Compare core_cov estimation and mapping estimation
```{r genus_heatmap}
# ggplot(instrain_breadth_depth_by_genus_host, aes(y = factor(Genus, genera), 
#                                                  x = factor(Host, host_order),
#                                                  fill = rel_abundance)) +
#                             geom_tile() +
#                               labs(x = "Host", y = "Genus", fill = "Mean relative\nabundance (%)")+
#                               scale_fill_gradient2(midpoint = 15, low = "#ffffff",  mid = "#6baed6", high = "#08306b", guide = "colourbar", na.value = "#ffffd9") +
#                                 make_theme(setFill = F, modify_guide = F, y_size = 12,
#                                   x_size = 14, x_angle = 0, x_hj = 0.5 , x_vj = 1,
#                                   y_hj =1, leg_pos = "right", leg_size = 11,
#                                   axis_x_title = 16, axis_y_title = 16,
#                                 )
#                                 ggsave("results/figures/09-mean_rel_abundance_from_mapping.pdf")

# ggplot(instrain_pa_by_genus_by_host, aes(y = factor(Genus, genera), 
#                                                  x = factor(Host, host_order),
#                                                  fill = Prevalence)) +
#                             geom_tile() +
#                             # geom_text(aes(label=Genus_cov)) +
#                               labs(x = "Host", y = "Genus", fill = "Prevalence\nin host (%)\n")+
#                               scale_fill_gradient2(midpoint = 0.5, low = "#ffffe5",  mid = "#41b6c4", high = "#081d58", guide = "colourbar") +
#                               # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar", trans = "log10") +
#                                 make_theme(setFill = F, modify_guide = F, y_size = 12,
#                                   x_size = 12, x_angle = 0, x_hj = 0.5 , x_vj = 1,
#                                   y_hj =1, leg_pos = "right", leg_size = 10,
#                                   axis_x_title = 16, axis_y_title = 16,
#                                 )
#                                 ggsave("results/figures/09-prevalence_from_core_cov.pdf")
```

```{r plots}
# ggplot(coverage_df, aes(y = factor(cluster_name), x = factor(Sample, samples_IN_MY), fill = Cov)) +
#                             geom_tile() +
#                               labs(x = "Sample", y = "Cluster")+
#                               scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar", trans = "log10") +
#                               # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar", trans = "log10") +
#                                 make_theme(setFill = F, modify_guide = F, y_size = 8,
#                                   x_size = 0, x_angle = 30, x_hj = 1 , x_vj = 1,
#                                   y_hj =1, leg_pos = "right",
#                                   leg_size = 11,
#                                   axis_x_title = 16, axis_y_title = 16,
#                                 )
# ggplot(coverage_df, aes(y = factor(cluster_name), x = factor(Host, host_order), fill = Prevalence_host)) +
#                             geom_tile() +
#                               labs(x = "Sample", y = "magOTU", fill = "Prevalence\n(Present = \nCov > 1)\n")+
#                               scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar") +
#                               # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
#                                 coord_fixed(0.1) +
#                                 make_theme(setFill = F, modify_guide = F, y_size = 8,
#                                   y_hj =1, leg_pos = "right"
#                                 )
#                                 ggsave("results/figures/09-Prevalence_magOTU_by_host.pdf")

# ggplot(coverage_df_genus %>% filter(Genus != "g__"), aes(y = factor(Genus, genera), x = factor(Host, host_order), fill = Prevalence_host)) +
#                             geom_tile() +
#                             geom_text(aes(label = ifelse(Mean > 0, Mean, ""))) +
#                               labs(x = "Sample", y = "Genus", fill = "Prevalence\n(Present = \nCov > 1)\nTruncated mean\n(0.01)")+
#                               # scale_fill_gradient(na.value = "transparent", low = "#fff5f0" , high = "#99000d", guide = "colourbar") +
#                               scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#1a88c9", guide = "colourbar") +
#                                 guides(fill = guide_colourbar(nbin = 8)) +
#                                 make_theme(setFill = F, modify_guide = F, y_size = 10,
#                                   x_hj= 0.5, x_size = 10,
#                                   y_hj =1, leg_pos = "right", leg_size = 10
#                                 )

# ggplot(coverage_df %>% filter(Genus != "g__") %>% mutate(Cov = round(Cov, 4)),
#         aes(y = factor(Genus, genera), x = Cov,, color =  factor(Genus, genera))) +
#                             geom_jitter() +
#                             # geom_text(aes(label = Mean)) +
#                               labs(x = "Coverage", y = "Cluster")+
#                               # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
#                               scale_color_manual(values = genusColors) +
#                                 # guides(fill = guide_colourbar(nbin = 8)) +
#                                 facet_wrap(~ factor(Host, host_order)) +
#                                 # scale_x_continuous(trans = "log", labels = scales::number_format(accuracy = 0.01)) +
#                                 make_theme(setCol = F, modify_guide = F, y_size = 10,
#                                   x_hj= 0.5, x_size = 10,
#                                   y_hj =1, leg_pos = "right", leg_size = 10
#                                 )

# ggplot(coverage_df %>% filter(Genus != "g__") %>% mutate(Cov = round(Cov, 4)),
#         aes(y = factor(Genus, genera), x = Cov,, color =  factor(Genus, genera))) +
#                             geom_point() +
#                             # geom_text(aes(label = Mean)) +
#                               labs(x = "Coverage", y = "Cluster")+
#                               # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
#                               scale_color_manual(values = genusColors) +
#                                 # guides(fill = guide_colourbar(nbin = 8)) +
#                                 facet_wrap(~ factor(Host, host_order)) +
#                                 scale_x_continuous(trans = "log", labels = scales::number_format(accuracy = 0.01)) +
#                                 make_theme(setCol = F, modify_guide = F, y_size = 10,
#                                   x_hj= 0.5, x_size = 10,
#                                   y_hj =1, leg_pos = "right", leg_size = 10
#                                 )
#                                 ggsave(paste0("results/figures/", "09-Coverage_scatter_genus_across_host.pdf"))
# ggplot(coverage_df %>% filter(Genus != "g__") %>% filter(Sample %in% samples_IN_MY) %>% mutate(Cov = round(Cov, 4)) %>% filter(Cov != 0),
#         aes(y = factor(Genus, genera), x = Cov, shape = Location, color =  Location)) +
#         # aes(y = factor(Genus, genera), x = Cov, shape = Location, color =  factor(Genus, genera))) +
#                             geom_jitter() +
#                             # geom_text(aes(label = Mean)) +
#                               labs(x = "Coverage", y = "Cluster")+
#                               # scale_fill_gradient(na.value = "transparent", low = "#ffffff" , high = "#ffc20e", guide = "colourbar") +
#                             #   scale_color_manual(values = genusColors) +
#                                 # guides(fill = guide_colourbar(nbin = 8)) +
#                                 facet_wrap(~ factor(Host, host_order)) +
#                                 scale_x_continuous(trans = "log", labels = scales::number_format(accuracy = 0.01)) +
#                                 make_theme(setCol = F, modify_guide = F, y_size = 10,
#                                   x_hj= 0.5, x_size = 10,
#                                   y_hj =1, leg_pos = "right", leg_size = 10
#                                 )
#                                 ggsave(paste0("results/figures/", "09-Coverage_scatter_location_compared.pdf"))

df_magOTUs_vegan <- abundance_df_matrix_pa

df_plot_cum_curve_m <- make_cum_curve(samples_AM, df_magOTUs_vegan, 500, "Apis mellifera")
df_plot_cum_curve_m_my <- make_cum_curve(samples_AM %>% as.data.frame %>% filter(. %in% samples_MY) %>% pull(.), df_magOTUs_vegan, 500, "Apis mellifera (Malaysia)")
df_plot_cum_curve_m_in <- make_cum_curve(samples_AM %>% as.data.frame %>% filter(. %in% samples_IN) %>% pull(.), df_magOTUs_vegan, 500, "Apis mellifera (India)")
df_plot_cum_curve_m_jp <- make_cum_curve(samples_AM %>% as.data.frame %>% filter(. %in% samples_JP) %>% pull(.), df_magOTUs_vegan, 500, "Apis mellifera (Japan)")
df_plot_cum_curve_m_sw <- make_cum_curve(samples_AM %>% as.data.frame %>% filter(. %in% samples_SW) %>% pull(.), df_magOTUs_vegan, 500, "Apis mellifera (Switzerland)")
df_plot_cum_curve_c <- make_cum_curve(samples_AC, df_magOTUs_vegan, 500, "Apis cerana")
df_plot_cum_curve_c_my <- make_cum_curve(samples_AC %>% as.data.frame %>% filter(. %in% samples_MY) %>% pull(.), df_magOTUs_vegan, 500, "Apis cerana (Malaysia)")
df_plot_cum_curve_c_in <- make_cum_curve(samples_AC %>% as.data.frame %>% filter(. %in% samples_IN) %>% pull(.), df_magOTUs_vegan, 500, "Apis cerana (India)")
df_plot_cum_curve_c_jp <- make_cum_curve(samples_AC %>% as.data.frame %>% filter(. %in% samples_JP) %>% pull(.), df_magOTUs_vegan, 500, "Apis cerana (Japan)")
df_plot_cum_curve_d <- make_cum_curve(samples_AD, df_magOTUs_vegan, 500, "Apis dorsata")
df_plot_cum_curve_d_my <- make_cum_curve(samples_AD %>% as.data.frame %>% filter(. %in% samples_MY) %>% pull(.), df_magOTUs_vegan, 500, "Apis dorsata (Malaysia)")
df_plot_cum_curve_d_in <- make_cum_curve(samples_AD %>% as.data.frame %>% filter(. %in% samples_IN) %>% pull(.), df_magOTUs_vegan, 500, "Apis dorsata (India)")
df_plot_cum_curve_f <- make_cum_curve(samples_AF, df_magOTUs_vegan, 500, "Apis florea")
df_plot_cum_curve_f_my <- make_cum_curve(samples_AF %>% as.data.frame %>% filter(. %in% samples_MY) %>% pull(.), df_magOTUs_vegan, 500, "Apis florea (Malaysia)")
df_plot_cum_curve_f_in <- make_cum_curve(samples_AF %>% as.data.frame %>% filter(. %in% samples_IN) %>% pull(.), df_magOTUs_vegan, 500, "Apis florea (India)")
df_plot_cum_curve_a <- make_cum_curve(samples_AA, df_magOTUs_vegan, 500, "Apis andreniformis")
df_plot_cum_curve_a_my <- make_cum_curve(samples_AA %>% as.data.frame %>% filter(. %in% samples_MY) %>% pull(.), df_magOTUs_vegan, 500, "Apis andreniformis (Malaysia)")
df_plot_cum_curve <- rbind(df_plot_cum_curve_m, df_plot_cum_curve_c, df_plot_cum_curve_d, df_plot_cum_curve_f, df_plot_cum_curve_a)
df_plot_cum_curve_by_loc <- rbind(df_plot_cum_curve_m_sw, df_plot_cum_curve_m_jp, df_plot_cum_curve_m_in, df_plot_cum_curve_m_my, 
                                  df_plot_cum_curve_c_jp, df_plot_cum_curve_c_in, df_plot_cum_curve_c_my, 
                                  df_plot_cum_curve_d_my, df_plot_cum_curve_d_in,
                                  df_plot_cum_curve_f_my, df_plot_cum_curve_f_in,
                                  df_plot_cum_curve_a_my)
ggplot(data = df_plot_cum_curve, aes(x = sample_size, y = number_of_clusters, color = factor(name, host_order))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method = "loess", se = FALSE) +
                          labs(color = "Host species", x = "Number of Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color) +
                            make_theme(leg_pos = "bottom", setCol = F,
                                       axis_x_title = 16, axis_y_title = 16,
                                       guide_nrow = 1)
                            ggsave("results/figures/09-Cumulative_curve.pdf")
host_order_color_ext <- c("Apis mellifera (India)" = "#4292c6",
                          "Apis mellifera (Malaysia)" = "#08519c",
                          "Apis mellifera (Japan)" = "#a6cee3",
                          "Apis mellifera (Switzerland)" = "#41b6c4",
                          "Apis cerana (Japan)" = "#fbb4ae", 
                          "Apis cerana (India)" = "#f46d43", 
                          "Apis cerana (Malaysia)" = "#e31a1c", 
                          "Apis dorsata (Malaysia)" = "#984ea3",
                          "Apis dorsata (India)" = "#cab2d6",
                          "Apis florea (Malaysia)" = "#4daf4a",
                          "Apis florea (India)" = "#b2df8a",
                          "Apis andreniformis (Malaysia)" = "#ff7f00")
ggplot(data = df_plot_cum_curve_by_loc, aes(x = sample_size, y = number_of_clusters, color = factor(name))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method = "loess", se = FALSE) +
                          labs(color = "Host species", x = "Number of Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color_ext) +
                            make_theme(leg_pos = "bottom", leg_size = 11,
                                       axis_x_title = 16, axis_y_title = 16,
                                       setCol = F, guide_nrow = 4) +
                              theme(legend.spacing.x = unit(1.0, 'cm'))
                            ggsave("results/figures/09-Cumulative_curve_by_location.pdf")

df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_pa %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_pa), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-presence_absence_heatmap.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_pa),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

# filtered for pretty
genera_filt <- c(
    "g__Lactobacillus",
    "g__Bifidobacterium",
    "g__Bombilactobacillus",
    "g__CALYQJ01",
    "g__Snodgrassella",
    "g__Gilliamella",
    "g__CALYQQ01",
    "g__Apibacter",
    "g__Bartonella_A",
    "g__Frischella",
    "g__Commensalibacter",
    "g__Apilactobacillus",
    "g__Dysgonomonas",
    "g__Enterobacter",
    "g__Pectinatus",
    "g__Entomomonas",
    "g__Saezia",
    "g__Parolsenella",
    "g__"
  )         
genusColors_filt <- list("g__Bombilactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[1],
                    "g__Lactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[4],
                    "g__CALYQJ01" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[6],
                    "g__Bifidobacterium" = brewer.pal(11, "Spectral")[3],
                    "g__Gilliamella" = brewer.pal(11, "Spectral")[11],
                    "g__CALYQQ01" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[11], "#FFFFFF"))(10), -1)[3],
                    "g__Frischella" = brewer.pal(11, "Spectral")[8],
                    "g__Bartonella_A" = brewer.pal(11, "Spectral")[7],
                    "g__Snodgrassella" = brewer.pal(11, "Spectral")[10],
                    "g__Apibacter" = brewer.pal(11, "Spectral")[4],
                    "g__Commensalibacter" = brewer.pal(8, "Dark2")[6],
                    "g__Apilactobacillus" = brewer.pal(11, "Spectral")[9],
                    "g__Dysgonomonas" = brewer.pal(11, "Spectral")[2],
                    "g__Pectinatus" = brewer.pal(8, "Dark2")[1],
                    "g__Enterobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[1],
                    "g__Entomomonas"= head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
                    "g__Saezia" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
                    "g__Parolsenella" = brewer.pal(11, "Spectral")[6],
                    "g__" = "#000000"
)
magotus <- unique(colnames(abundance_df_matrix_pa))
magotus_filt <- magotus %>% as.data.frame() %>% left_join(abundance_df %>% ungroup() %>% select(magOTU, Genus) %>% unique, by = c("." = "magOTU")) %>% filter(Genus %in% genera_filt) %>% select(!Genus) %>% pull(.) %>% as.character %>% unique
abundance_df_matrix_pa_filt <- abundance_df_matrix_pa[, magotus_filt] 
df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_pa_filt %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_pa_filt), 
                                gp = gpar(fill = "#081d58", border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(Location = location_order_color),
                              border = T
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors_filt)), border = T
                    )
pdf("results/figures/09-presence_absence_heatmap_filt.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_pa_filt),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()


df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color)
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-coverage_heatmap.pdf")

column_split <- rownames(abundance_df_matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix),
        col = colorRamp2(c(0, 10, 100, 1000, 5000, 10000), colors = c(brewer.pal(9, "YlGnBu")[1], brewer.pal(9, "YlGnBu")[2], brewer.pal(9, "YlGnBu")[4], brewer.pal(9, "YlGnBu")[6], brewer.pal(9, "YlGnBu")[8], brewer.pal(9, "YlGnBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Coverage"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()


# ggplot() +
#   geom_bar(data = abundance_df,
#            aes(x = coverage,
#                y = factor(sample, samples_MY),
#                fill = Genus
#            ), 
#           #  color = "white",
#            stat = "identity", position = "stack"
#   ) +
#   labs(x = "coverage", y = "sample", fill = "Genus") +
#   make_theme(theme_name=theme_few(), leg_pos="bottom",
#                      guide_nrow = 5,
#                      setFill = F, setCol = F,
#                      x_angle=45 ,x_vj=1, x_hj=1, x_size=4,
#                      y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#             scale_fill_manual(values=genusColors)
#   ggsave("results/figures/09-Abundance_by_sample_qPCR_adj.pdf")

# ggplot() +
#   geom_bar(data = df_Cov_copy_num_adj,
#            aes(x = Cov_adj,
#                y = factor(Sample, samples_MY),
#                fill = Genus
#            ), 
#           #  color = "white",
#            stat = "identity", position = "stack"
#   ) +
#   labs(x = "Cov Adjusted (# Microbial Reads)", y = "Sample", fill = "Genus") +
#   make_theme(theme_name=theme_few(), leg_pos="bottom",
#                      guide_nrow = 5,
#                      setFill = F, setCol = F,
#                      x_angle=45 ,x_vj=1, x_hj=1, x_size=4,
#                      y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#             scale_fill_manual(values=genusColors)
# ggplot() +
#   geom_bar(data = df_Cov_copy_num_adj,
#            aes(x = Cov_rel,
#                y = factor(Sample, samples_MY),
#                fill = Genus
#            ), 
#            color = "white",
#            stat = "identity", position = "stack"
#   ) +
#   labs(x = "Cov Relative", y = "Sample", fill = "Genus") +
#   make_theme(theme_name=theme_few(), leg_pos="bottom",
#                      guide_nrow = 5,
#                      setFill = F, setCol = F,
#                      x_angle=45 ,x_vj=1, x_hj=1, x_size=4,
#                      y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#             scale_fill_manual(values=genusColors)
# ggplot() +
#   geom_bar(data = df_Cov_copy_num_adj,
#            aes(x = Cov_adj_copy,
#                y = factor(Sample, samples_MY),
#                fill = Genus
#            ), 
#           #  color = "black",
#            stat = "identity", position = "stack"
#   ) +
#   labs(x = "Cov Relative * copy_num", y = "Sample", fill = "Genus") +
#   facet_wrap(~ Host, scales = "free") +
#   # scale_x_continuous(trans = "log") +
#   make_theme(theme_name=theme_few(), leg_pos="bottom",
#                      guide_nrow = 5,
#                      setFill = F, setCol = F,
#                      x_angle=45 ,x_vj=1, x_hj=1, x_size=4,
#                      y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#             scale_fill_manual(values=genusColors)
# ggplot(df_reads %>%
#         select(Sample, MAGs_DB, Trimmed) %>%
#         mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#         mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
#         filter(Location %in% c("Malaysia", "India")) %>%
#           left_join(as.data.frame(rowSums(data_otu)) %>% setNames("cov_sums") %>% rownames_to_column("Sample"), by = "Sample"),
#        aes(x = MAGs_DB, y = cov_sums, 
#            color = Location)
#         #    color = factor(Host, host_order), shape = Location)
#     ) +
#     facet_wrap(~factor(Host, host_order)) +
#     labs(x = "Numer of MAGs DB mapped reads", y = "Sum of all coverages", color = "Host") +
#     geom_point(size = 3) +
#       make_theme(setCol = F) +
#         # scale_color_manual(values=host_order_color_dark) +
#           scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6))

ggplot(df_reads %>% filter((Sample %in% samples)) %>% 
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
        ) +
    geom_point(aes(x = factor(Sample, samples),
                   y = MAGs_DB,
                   color = Location,
                   shape = factor(Host, host_order))) +
    geom_hline(yintercept = 1e+06) +
    # scale_color_manual(values=host_order_color_dark) +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(setCol = F)

ggplot(df_reads %>% filter((Sample %in% samples)) %>% 
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
        ) +
    geom_point(aes(x = factor(Sample, samples),
                   y = MAGs_DB,
                   shape = Location,
                   color = factor(Host, host_order))) +
    geom_hline(yintercept = 1e+06) +
    scale_color_manual(values=host_order_color_dark) +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(setCol = F)

# code for betapart functions from
# https://github.com/cran/betapart

# Latest commit a222c6d on Feb 24, 2012
betapart.core <- function(x){
	
	# test for a binary numeric matrix or data.frame
	if(! is.matrix(x)){
		x<-as.matrix(x)
  	}
	
	if(! is.numeric(x))
    	stop("The data in x is not numeric.",call.=TRUE)
	
	# simple test for binary data
	xvals <-  unique(as.vector(x))
	if (any(!is.element(xvals, c(0, 1)))) 
        stop("The table contains values other than 0 and 1: data should be presence/absence.", call. = TRUE)

	shared <- x %*% t(x)
    not.shared <-  abs(sweep(shared, 2, diag(shared)))
		
	sumSi <- sum(diag(shared)) # species by site richness
    St <- sum(colSums(x) > 0)  # regional species richness
    a <- sumSi - St            # multi site shared species term

    sum.not.shared <- not.shared + t(not.shared)
    max.not.shared <- pmax(not.shared, t(not.shared))
    min.not.shared <- pmin(not.shared, t(not.shared))

	computations<-list(data=x, sumSi=sumSi, St=St, a=a, shared=shared, not.shared=not.shared, 
	                   sum.not.shared=sum.not.shared, max.not.shared=max.not.shared, 
	                   min.not.shared=min.not.shared)
    class(computations)<-"betapart"

	return(computations)
} 
# Latest commit 4b59d9c on Feb 24, 2012
beta.pair <- function(x, index.family="sorensen"){
	
	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
			beta.sim <- x$min.not.shared / (x$min.not.shared + x$shared)
			beta.sne <- ((x$max.not.shared - x$min.not.shared) / ((2 * x$shared) + x$sum.not.shared)) * (x$shared / (x$min.not.shared + x$shared))
            beta.sor <- x$sum.not.shared / (2 * x$shared + x$sum.not.shared)
                
            pairwise <- list(beta.sim=as.dist(beta.sim), beta.sne=as.dist(beta.sne),beta.sor=as.dist(beta.sor))},
		jaccard = {
			beta.jtu <- (2*x$min.not.shared) / ((2*x$min.not.shared) + x$shared)
	        beta.jne <- ((x$max.not.shared - x$min.not.shared) / (x$shared + x$sum.not.shared)) * (x$shared / ((2*x$min.not.shared) + x$shared))
	        beta.jac <- x$sum.not.shared / (x$shared + x$sum.not.shared)

	        pairwise <- list(beta.jtu=as.dist(beta.jtu), beta.jne=as.dist(beta.jne),beta.jac=as.dist(beta.jac))})

	return(pairwise)
}
beta.multi <- function(x, index.family="sorensen"){

	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}

	maxbibj <- sum(x$max.not.shared[lower.tri(x$max.not.shared)])
    minbibj <- sum(x$min.not.shared[lower.tri(x$min.not.shared)])
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
            beta.sim <- minbibj / (minbibj + x$a)
            beta.sne <- (x$a / (minbibj + x$a)) * ((maxbibj - minbibj) / ((2 * x$a) + maxbibj + minbibj))
            beta.sor <- (minbibj + maxbibj) / (minbibj + maxbibj + (2 * x$a))

           	multi <- list(beta.SIM=beta.sim, beta.SNE=beta.sne,beta.SOR=beta.sor)},
		jaccard = {
            beta.jtu <- (2*minbibj) / ((2*minbibj) + x$a)
            beta.jne <- (x$a / ((2*minbibj) + x$a)) * ((maxbibj - minbibj) / ((x$a) + maxbibj + minbibj))
            beta.jac <- (minbibj + maxbibj) / (minbibj + maxbibj + x$a)

           	multi <- list(beta.JTU=beta.jtu, beta.JNE=beta.jne, beta.JAC=beta.jac)})

	return(multi)

}

df_magOTUs_vegan_pa <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0) %>%
                        filter(sample %in% samples_IN)
df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
pdf("results/figures/09-PCoA_sorensen_microbiome.pdf")
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()

# pcoa_plot(dist_matrix_betapart$beta.sne, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
# pcoa_plot(dist_matrix_betapart$beta.sim, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

# dist_matrix_betapart <- beta.pair(df_magOTUs_vegan, "jaccard")
# pdf("results/figures/09-PCoA_jaccard_microbiome.pdf")
# pcoa_plot(dist_matrix_betapart$beta.jac, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
# dev.off()
# pcoa_plot(dist_matrix_betapart$beta.jne, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
# pcoa_plot(dist_matrix_betapart$beta.jtu, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

dist_matrix_betapart_multi <- beta.multi(df_magOTUs_vegan, "sorensen")
str(dist_matrix_betapart_multi)

# beta.sim dist object, dissimilarity matrix accounting for spatial turnover (replacement),
# measured as Simpson pair-wise dissimilarity
# beta.sne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Sorensen pair-wise dissimilarity
# beta.sor dist object, dissimilarity matrix accounting for total dissimilarity, measured as
# Sorensen pair-wise dissimilarity (a monotonic transformation of beta diversity)
# For index.family="jaccard" the three matrices are:
# beta.jtu dist dissimilarity matrix accounting for spatial turnover, measured as the turnoverfraction of Jaccard pair-wise dissimilarity
# beta.jne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Jaccard pair-wise dissimilarity
# beta.jac dist object, dissimilarity matrix accounting for beta diversity, measured as Jaccard pair-wise dissimilarity (a monotonic transformation of beta diversity)

grouping_vec <- df_meta_complete %>% filter(ID %in% rownames(df_magOTUs_vegan)) %>% pull(Species)
anosim(dist_matrix_betapart$beta.sor, grouping_vec, distance = "sorenson", permutations = 9999)

df_to_test <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
                        # filter(rownames(.) %in% c(samples_IN, samples_MY))
                          # filter(rownames(.) %in% c(samples_am, samples_ac, samples_af, samples_ad))
adonis2(df_to_test ~ Species + Country, data = df_meta_complete %>% filter(ID %in% rownames(df_to_test)))
# Distances from
# https://link.springer.com/article/10.1007/s11756-022-01123-6/tables/4
# 		7	5	3	1	2
# mellifera	7		0.112	0.128	0.132	0.146
# cerana	5	0.112		0.133	0.13	0.119
# dorsata	3	0.128	0.133		0.095	0.108
# florea	1	0.132	0.13	0.095		0.047
# andreniformis	2	0.146	0.119	0.108	0.047	

get_host_divergence <- function(sample_1, sample_2) {
  host_1 <- get_host_from_sample_name(sample_1)
  host_1 <- strsplit(host_1, " ")[[1]][[2]]
  host_2 <- get_host_from_sample_name(sample_2)
  host_2 <- strsplit(host_2, " ")[[1]][[2]]
  mellifera <- c(0, 0.112, 0.128, 0.132, 0.146)
  cerana <- c(0.112, 0, 0.133, 0.13, 0.119)
  dorsata <- c(0.128, 0.133, 0, 0.095, 0.108)
  florea <- c(0.132, 0.13, 0.095, 0, 0.047)
  andreniformis <- c(0.146, 0.119, 0.108, 0.047, 0)
  # mellifera <- c(0, 1, 2, 4, 4)
  # cerana <- c(1, 0, 2, 4, 4)
  # dorsata <- c(2, 2, 0, 3, 3)
  # florea <- c(4, 4, 3, 0, 1)
  # andreniformis <- c(4, 4, 3, 1, 0)
  host_divergence_mat <- rbind(mellifera, cerana, dorsata, florea, andreniformis)
  colnames(host_divergence_mat) <- c("mellifera", "cerana", "dorsata", "florea", "andreniformis")
  index_list <- list("mellifera" = 1, "cerana" = 2, "dorsata" = 3, "florea" = 4, "andreniformis" = 5)
  host_divergence <- host_divergence_mat[as.numeric(index_list[host_1]), as.numeric(index_list[host_2])]
  return(host_divergence)
}
df_magOTUs_vegan_IN_MY <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_IN_MY, "sorensen")
sample_bac_div <- dist_matrix_betapart$beta.sor
sample_host_div <- data.frame()
for (sample in samples) {
  sample_host_div <- rbind(sample_host_div, rep(0, length((samples))))
}
colnames(sample_host_div) <- samples
rownames(sample_host_div) <- samples
for (sample_1 in samples) {
  for (sample_2 in samples) {
    ind_1 <- as.numeric(which(samples == sample_1))
    ind_2 <- as.numeric(which(samples == sample_2))
    sample_host_div[ind_1, ind_2] <- get_host_divergence(sample_1, sample_2)
  }
}

# samples_subset_am <- df_meta_complete %>% filter(Species == "Apis mellifera") %>% pull(ID)
# samples_subset_ac <- df_meta_complete %>% filter(Species == "Apis cerana") %>% pull(ID)
# samples_subset_ad <- df_meta_complete %>% filter(Species == "Apis dorsata") %>% pull(ID)
# samples_subset_af <- df_meta_complete %>% filter(Species == "Apis florea") %>% pull(ID)
# samples_subset_aa <- df_meta_complete %>% filter(Species == "Apis andreniformis") %>% pull(ID)

heatmap(sample_bac_div %>% as.matrix)

sample_host_div <- sample_host_div %>% 
  filter(row.names(.) %in% as.vector(as.matrix(sample_bac_div) %>% row.names))

sample_host_div_df <- sample_host_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                        pivot_longer(!Sample, names_to = "Compared", values_to = "dist_h")
sample_bac_div_df <- sample_bac_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                          pivot_longer(!Sample, names_to = "Compared", values_to = "dist_b") %>%
                          mutate(Host_s = Vectorize(get_host_from_sample_name)(Sample)) %>%
                          mutate(Host_c = Vectorize(get_host_from_sample_name)(Compared)) %>%
                          mutate(Type_b = ifelse(Host_s == Host_c, "Within", "Between")) %>%
                          mutate(Hosts_compared = paste0(Host_s, "_", Host_c))
dissimilarities_df_median <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, median_b = median(dist_b)) %>%
                        unique
dissimilarities_df <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared)
                        # filter(Sample %in% c("M1.2", "C1.2", "D1.2", "F1.2", "A1.2"))

mantel(sample_host_div, sample_bac_div, method="pearson", permutations=1000)

ggplot() +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_c,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_s,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
    labs(x = "Host divergence (# nodes between)", y = "Microbiome dissimilarity (jaccard)") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 1)
    ggplot("results/figures/09-Microbe_host_dissimilarities_violin.pdf")

ggplot() +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                #  fill = Host_c,
                 color = Host_c,
                #  shape = Type_b
             ), width=0, size = 4
            ) +
  geom_jitter(data=dissimilarities_df_median,
             aes(x = dist_h,
                 y = median_b
             ), size = 3, width = 0
            ) +
    labs(x = "Host divergence (based on 16S and 12S)", y = "Microbiome dissimilarity (jaccard)", color = "Host species") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 2,
               leg_size = 12,
               axis_x_title = 15, axis_y_title = 15
    )
  ggplot("results/figures/09-Microbe_host_dissimilarities.pdf")
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

split_by_pattern <- function(my_name, pattern, index) {
  return(strsplit(my_name, pattern)[[1]][[index]])
}
bac_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(dist_h, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = median_b) %>%
                    column_to_rownames("Compared1")
host_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(median_b, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = dist_h) %>%
                    column_to_rownames("Compared1")

bac_median_mat[1,]
host_median_mat[1,]

mantel(as.dist(bac_median_mat), as.dist(host_median_mat), method = "pearson", permutations = 9999)
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

# # edit tree to look prettier later
# plot_tree(ps, nodelabf=nodeplotblank, color="Host", label.tips = "Name", ladderize="left", plot.margin=0.3) + 
#   scale_color_manual(values = host_order_color)
#   ggsave("results/figures/09-Plot_bac120_tree_with_host.pdf")


dist_mat_unifrac <- UniFrac(prune_samples(sample_sums(ps) > 0, ps), weighted=F, normalized=T, parallel=FALSE, fast=TRUE)
pcoa_plot(dist_mat_unifrac, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

dist_mat_unifrac_wt <- UniFrac(prune_samples(sample_sums(ps) > 0, ps), weighted=T, normalized=F, parallel=FALSE, fast=TRUE)
pcoa_plot(dist_mat_unifrac_wt, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

reformat_host_name <- function(x) {
  paste0(strsplit(x, " ")[[1]], collapse = "_")
}
rename_mag_labels <- function(x) {
  return(paste0(tax_table_mat %>% as.data.frame %>% filter(rownames(.) == x) %>% pull(Name), "---", x))
}

host_tree_rooted <- root(host_tree, "Bombus_terrestris", resolve.root = TRUE)

mags_tree_Lacto_path <- "/scratch/aprasad/20230313_apis_species_comparison/results/11_phylogenies/02_orthofinder_results/g__Lactobacillus/Results_g__Lactobacillus_iqtree/Species_Tree/SpeciesTree_rooted.txt"
mag_tree_Lacto <- read.tree(mags_tree_Lacto_path)
  
associations_df <- abundance_df %>% filter(!is.na(magOTU)) %>%
                    mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0))  %>%
                    select(sample, Host, detected, magOTU, Genus) %>%
                    mutate(representative_genome = Vectorize(get_rep_genome_magotu)(magOTU, Genus)) %>%
                    select(!magOTU) %>%
                    filter(!is.na(representative_genome)) %>%
                    filter(!is.na(Genus)) %>%
                    left_join(vis_magOTUs_df_all %>% ungroup() %>% select(magOTU, ID), by = c("representative_genome" = "ID")) %>%
                      mutate(Host = Vectorize(reformat_host_name)(Host)) %>%
                      filter(representative_genome %in% mag_tree_Lacto$tip.label) %>%
                      filter(detected == 1) %>%
                      select(representative_genome, Host) %>%
                      unique()
mag_tree_Lacto$tip.label <- lapply(mag_tree_Lacto$tip.label, rename_mag_labels)
# cols <- unlist(lapply(as.character(associations_df$representative_genome), function(x) genusColors_char[strsplit(x, "---")[[1]]][[1]]))
cophyloplot(mag_tree_Lacto, collapse.singles(host_tree), assoc = associations_df, space = 100, gap = 15, lwd = 4, show.tip.label = F)
# nodelabels.cophylo(..., which=c("left","right"))
# edgelabels.cophylo(..., which=c("left","right"))
# tiplabels.cophylo(..., which=c("left","right"))
```



```{r plot_instrain_iRep}
# plot_iRep <- ggplot(instrain_genomes_info_df,
#                              aes(x = sample,
#                                  y = iRep,
#                                  color = Genus,
#                               )
#                             ) +
#   geom_point() +
#     facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 5) +
#     # scale_y_continuous(trans="log10") +
#       make_theme(theme_name=theme_few(), leg_pos="none",
#                  # guide_nrow = 6,
#                  setFill = F, setCol = F,
#                  x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
#                  y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#         geom_vline(xintercept = 5.5) +
#         geom_vline(xintercept = 20.5) +
#         geom_vline(xintercept = 30.5) +
#         theme(strip.text = element_text(size = 7)) +
#         scale_color_manual(values=genusColors)

# paginate_save(plot_iRep, "magOTU", "results/figures/10-instrain_plots/3-iRep_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```
```{r other_plots}
snv_perc_plot <- ggplot(instrain_genomes_info_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = SNV_count/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = magOTU,
                                 y = perc_var,
                                 color = factor(Host, host_order),
                                 shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      labs(x = "magOTU", y = "% variant sites", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)
            # ggsave("results/figures/15b-percentage_variable_sites.pdf")

# snv_perc_plot_rare <- ggplot(instrain_genomes_info_df %>%
ggplot(instrain_genomes_info_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = consensus_divergent_sites/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = magOTU,
                                 y = reads_mean_PID,
                                 color = factor(Host, host_order)
                                 # shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      scale_y_continuous(labels=unit_format(unit = "%", scale = 1e+2)) +
      labs(x = "magOTU", y = "mean ANI (reads and reference)", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)

ggplot(instrain_genomes_info_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = consensus_divergent_sites/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = magOTU,
                                 y = reads_unfiltered_reads,
                                 color = factor(Host, host_order)
                                 # shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6), trans = "log") +
      labs(x = "magOTU", y = "Number of reads mapped", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
        # geom_hline(yintercept = 10, color = "grey") +
        # geom_hline(yintercept = 100, color = "grey") +
        # geom_hline(yintercept = 1000, color = "grey") +
        geom_hline(yintercept = 10000, color = "grey") +
        # geom_hline(yintercept = 100000, color = "grey") +
        geom_hline(yintercept = 1000000, color = "grey") +
        # geom_hline(yintercept = 10000000, color = "grey") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)
```