---
title: "Honeybee cross-species analysis - 09"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Initialization

To intialize complete all of or relevant sections of `scripts/visualization/Load_data.R`.

Due to large differences in total reads, plotting number of filtered pairs directly does not work.
Plotting percentage after normalizing across magOTUs using the factor `length*breadth` is more meaningful


```{r}
system("mkdir -p results/figures/10-instrain_plots/")
```

```{r instrain_scaff_detected}
# plot number of true scaffolds vs detected scaffolds for each sample each magOTU
# showing breadth and coverage...

plot_scaff_detected_list <- list()
for (magotu in sort(unique(instrain_genomes_info_df$magOTU))) {
  plot_scaff_detected <- ggplot(instrain_genomes_info_df_all %>% 
                                mutate(perc_detected = detected_scaffolds/true_scaffolds*100) %>%
                                mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, "y", "n")) %>%
                                mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100) %>%
                                mutate(Genus = as.character(Genus)) %>%
                                mutate(Genus = ifelse(detected == "y", Genus, "g__")) %>%
                                mutate(Host = ifelse(detected == "y", Host, "g__")) %>%
                                filter(magOTU == magotu),
                             aes(y = perc_detected,
                                 x = breadth,
                                 color = Host,
                                 size = coverage_median
                              )
                            ) +
  geom_point() +
  scale_size_continuous(trans = scales::pseudo_log_trans(base = 10)) +
  labs(y = "Percentage of scaffolds detected (at least a read mapped)") +
  ggtitle(paste0(magotu)) +
  # scale_size_continuous(trans = "log10") +
    # facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 6) +
    geom_vline(xintercept = 50, linetype = "dotted") +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 6, leg_size = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=host_order_color_dark) +
        guides(color="none")
  plot_scaff_detected_list[[magotu]] <- plot_scaff_detected
}
pdf("results/figures/10-instrain_plots/1-scaffolds_detected_by_magOTU.pdf", width = 20, height = 25)
marrangeGrob(plot_scaff_detected_list, nrow = 3, ncol = 3)
dev.off()
```

```{r instrain_plots_breadth_by_sample}
plot_breadth_by_sample_list <- list()
for (a_sample in unique(instrain_genomes_info_df$sample)) {
  plot_breadth_by_sample <- ggplot(instrain_genomes_info_df_all %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100) %>%
                                filter(sample == a_sample),
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Genus,
                              )
                            ) +
  geom_point() +
    ggtitle(paste0(a_sample)) +
      make_theme(theme_name=theme_few(), leg_pos="none",
                 # guide_nrow = 6,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_abline(slope=1, linetype = "dotted") +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=genusColors)
  plot_breadth_by_sample_list[[a_sample]] <- plot_breadth_by_sample
}
pdf("results/figures/10-instrain_plots/1-breadth_vs_expected_by_sample.pdf", width = 20, height = 25)
marrangeGrob(plot_breadth_by_sample_list, nrow = 3, ncol = 3)
dev.off()
```


```{r instrain_plots_breadth_by_sample_with_coveraeg}
# redo using marrangeGrob !
plot_breadth_by_sample_w_cov_list <- list()
for (a_sample in unique(instrain_genomes_info_df$sample)) {
  plot_breadth_by_sample_w_cov <- ggplot(instrain_genomes_info_df_all %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100) %>%
                                filter(sample == a_sample),
                             aes(y = breadth,
                                 x = breadth_expected,
                                 color = Genus,
                                 size = coverage
                              ), alpha = 0.75
                            ) +
  geom_point() +
    ggtitle(paste0(a_sample)) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 12,
                 setFill = F, setCol = F,
                 x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        geom_abline(slope=1, linetype = "dotted") +
        theme(strip.text = element_text(size = 6)) +
        scale_color_manual(values=genusColors) +
        guides(color = guide_legend(show = FALSE))
  plot_breadth_by_sample_w_cov_list[[a_sample]] <- plot_breadth_by_sample_w_cov
}
pdf("results/figures/10-instrain_plots/1-breadth_vs_expected_w_cov_by_sample.pdf", width = 20, height = 25)
marrangeGrob(plot_breadth_by_sample_w_cov_list, nrow = 3, ncol = 3)
dev.off()
```

```{r instrain_plots_breadth_by_magOTU}
plot_breadth_by_magOTU_list <- list()
for (magotu in unique(instrain_genomes_info_df_all)) {
  plot_breadth_by_magOTU <- ggplot(instrain_genomes_info_df_all %>% mutate(breadth = breadth*100, breadth_expected =  breadth_expected*100),
                             aes(x = breadth,
                                 y = breadth_expected,
                                 color = Host,
                              )
                            ) +
                              geom_point() +
                                ggtitle(paste0(magotu)) +
                                  make_theme(theme_name=theme_few(), leg_pos="none",
                                            # guide_nrow = 6,
                                            setFill = F, setCol = F,
                                            x_angle=0 ,x_vj=0, x_hj=0, x_size=10,
                                            y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
                                    geom_abline(slope=1, linetype = "dotted") +
                                    theme(strip.text = element_text(size = 6)) +
                                    scale_color_manual(values=host_order_color_dark)
  plot_breadth_by_magOTU_list[[magotu]] <- plot_breadth_by_magOTU
}
pdf("results/figures/10-instrain_plots/1-breadth_vs_expected_by_magOTU.pdf", width = 20, height = 25)
marrangeGrob(plot_breadth_by_magOTU_list, nrow = 3, ncol = 3)
dev.off()
```

```{r instrain_plots_breadth_vs_num_reads}
# redo using marrangeGrob !

plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth
                              ), alpha = 0.4, color = "black"
                            ) +
  geom_point(data = instrain_breadth_depth_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = Host
                              )
                            ) +
  scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 5,
                 setFill = F, setCol = F,
                 x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4)) +
        scale_color_manual(values=host_order_color_dark)

paginate_save(plot_breadth_num_reads, "magOTU", "results/figures/10-instrain_plots/2-mapped_reads_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```
```{r instrain_plots_breadth_vs_coverage}
# redo using marrangeGrob

plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = coverage_median,
                                 size = breadth
                              ), alpha = 0.4, color = "black"
                            ) +
  geom_point(data = instrain_breadth_depth_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = coverage_median,
                                 size = breadth,
                                 color = Host
                              )
                            ) +
  scale_y_continuous(trans = "log10", labels = comma, limits = c(-0.1, NA)) +
    facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 5,
                 setFill = F, setCol = F,
                 x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4)) +
        scale_color_manual(values=host_order_color_dark)

paginate_save(plot_breadth_num_reads, "magOTU", "results/figures/10-instrain_plots/2-coverage_by_magOTU.pdf", pass_nrow = 3, pass_ncol = 3)
```

```{r instrain_plots_breadth_vs_num_reads_coverage}
# redo using marrangeGrob

plot_breadth_num_reads <- ggplot() +
  geom_point(data = instrain_breadth_depth_unfiltered_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = coverage_factor
                              )
                            ) +
  geom_point(data = instrain_breadth_depth_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                                 color = coverage_factor,
                              )
                            ) +
  geom_point(data = instrain_breadth_depth_all %>% mutate(breadth = breadth*100),
                             aes(x = factor(sample, samples),
                                 y = filtered_read_pair_count,
                                 size = breadth,
                              ), shape = 1, color = "black", stroke = 0.25
                            ) +
  scale_color_manual(values = c("<0.0001" = "#a50f15",
                                "0.0001 - 0.001" = "#d73027",
                                "0.001 - 0.01" = "#f46d43",
                                "0.01 - 0.1" = "#fdae61",
                                "0.1 - 1" = "#fee08b",
                                "1 - 10" = "#a6d96a", 
                                "10 - 100" = "#74c476", 
                                "100 - 1000" = "#41ab5d", 
                                ">1000" = "#238b45"
                              )
                          ) +
  scale_y_continuous(trans = "log10", labels = comma) +
    facet_wrap_paginate(~magOTU, ncol = 3, nrow = 3, page = 4) +
      make_theme(theme_name=theme_few(), leg_pos="right",
                 guide_nrow = 9,
                 setFill = F, setCol = F, palettecolor = "RdYlGn",
                 x_angle=45 ,x_vj=1, x_hj=1, x_size=0,
                 y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
        theme(strip.text = element_text(size = 10)) +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_hline(yintercept = 10, linetype = "dashed") +
        geom_hline(yintercept = 100, linetype = "dotted") +
        geom_vline(xintercept = 97, linetype = "dotted") +
        geom_vline(xintercept = 162, linetype = "dotted") +
        geom_vline(xintercept = 207, linetype = "dotted") +
        geom_vline(xintercept = 252, linetype = "dotted") +
        scale_size(range = c(1,4))

paginate_save(plot_breadth_num_reads, "magOTU", "results/figures/10-instrain_plots/2-mapped_reads_by_magOTU_w_coverage.pdf", pass_nrow = 3, pass_ncol = 3)
```

```{r coverage_nor_adj}
# plot_cov_df <- abundance_df %>%
#                 select(sample, Host, magOTU, coverage_median, coverage, coverage_adj, breadth) %>%
#                   pivot_longer(cols = c(coverage, coverage_adj, coverage_median), names_to = "cov_type", values_to = "coverage")
# plot_list <- list()
# for (magotu in unique(plot_cov_df$magOTU)) {
#   plot <- ggplot() +
#           geom_point(data = plot_cov_df %>% filter(magOTU == magotu),
#                              aes(x = factor(sample, samples),
#                                  y = coverage,
#                                 #  size = breadth,
#                                  shape = cov_type,
#                                  color = Host
#                               ), size = 3, alpha = 0.75
#                             ) +
#             scale_y_continuous(trans = "log10", labels = comma, limits = c(-0.1, NA)) +
#             make_theme(theme_name=theme_few(), leg_pos="right",
#                       guide_nrow = 5,
#                       setFill = F, setCol = F,
#                       x_angle=90 ,x_vj=1, x_hj=1, x_size=0,
#                       y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
#               scale_shape_manual(values = c("coverage" = 18, "coverage_adj" = 13, "coverage_median" = 15)) +
#               theme(strip.text = element_text(size = 10)) +
#               geom_hline(yintercept = 0, linetype = "solid") +
#               geom_hline(yintercept = 10, linetype = "dashed") +
#               geom_hline(yintercept = 100, linetype = "dotted") +
#               geom_vline(xintercept = 97, linetype = "dotted") +
#               geom_vline(xintercept = 162, linetype = "dotted") +
#               geom_vline(xintercept = 207, linetype = "dotted") +
#               geom_vline(xintercept = 252, linetype = "dotted") +
#               scale_size(range = c(1,4)) +
#               # scale_color_manual(values=c("#bebada","#fb8072", "#80b1d3"))
#               scale_color_manual(values=host_order_color_dark)
#   plot_list[[magotu]] <- plot
# }

# pdf("results/figures/10-instrain_plots/2-coverage_by_magOTU_w_coverage.pdf", width = 12, height = 15)
# marrangeGrob(plot_list, nrow = 5, ncol = 1)
# dev.off()
```


# Reads mapping to database

```{r}
colnames(df_reads_updated)

ggplot(df_reads_plot %>% 
          filter(!is.na(Number)) %>% 
          filter(Sample %in% samples_MY) %>% 
          filter(Type %in% c("Total_clean", "MAGs_DB_mf", "Host_mapped_hf")),
          aes(y = factor(Sample, samples_MY), x = Number, fill = Type)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Host, scale = "free_y") +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(leg_pos = "right", guide_nrow = 3)
ggsave("results/figures/10-instrain_plots/3-reads_mapped_to_each_db.pdf", width = 12, height = 15)

# of the non-host reads how much maps to mags db
df_reads_mod <- df_reads %>%
                  filter(Sample %in% samples_IN_MY) %>%
                  mutate(nonhost_hf = Total_clean - Host_mapped_hf) %>%
                  mutate(perc_nonhost_mag_hf = MAGs_DB_hf/nonhost_hf*100) %>%
                  mutate(perc_unmapped_nonhost_hf = 100 - perc_nonhost_mag_hf) %>%
                  mutate(MAG_DB_mapped = perc_nonhost_mag_hf) %>%
                  mutate(unmapped = perc_unmapped_nonhost_hf) %>%
                  select(MAG_DB_mapped, unmapped, Sample, nonhost_hf) %>%
                  pivot_longer(cols = c(MAG_DB_mapped, unmapped, nonhost_hf), names_to = "Type", values_to = "Number") %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample))

ggplot() +
  geom_bar(data = df_reads_mod %>% filter(Type != "nonhost_hf"),
            aes(x = Number, y = factor(Sample, samples_IN_MY), fill = factor(Type, c("unmapped", "MAG_DB_mapped"))),
            stat = "identity", position = "stack") +
  facet_wrap(~Host, scale = "free_y") +
  labs(fill = "Non-host reads", x = "Percentage", y = "Sample") +
  scale_fill_manual(values = c("MAG_DB_mapped" = "#ccebc5", "unmapped" = "#fbb4ae"), labels = c("MAG_DB_mapped" = "Mapped to MAGs", "unmapped" = "Unmapped to all")) +
  scale_x_continuous(labels=unit_format(unit = "%", scale = 1)) +
  new_scale_fill() +
  geom_tile(data = df_reads_mod %>% filter(Type == "nonhost_hf"), 
            aes(x = -8, width = 10, y = factor(Sample, samples_IN_MY), fill = Number), color = "#000000") +
    scale_fill_gradient2(low = "#f7fbff", mid = "#4292c6", high = "#08306b", midpoint = 4e+07,
                         label = scientific_10) +
    labs(fill = "Number of non-host reads") +
    geom_vline(xintercept = 80, linetype = "dotted") +
    make_theme(setCol = F, setFill = F, leg_pos = "right",
                guide_nrow = 3, modify_guide = F,
                x_angle = 45, x_vj = 1, x_hj = 1)
ggsave("results/figures/10-instrain_plots/3-perc_mapped_to_mag_vs_unmapped.pdf", width = 12, height = 15)

# plot qpcr Ct values against number of mag_db_mf and mag_db_hf

df_plot_ct <- df_reads_plot %>% 
                filter(!is.na(Number)) %>% 
                filter(Sample %in% samples_MY) %>% 
                filter(Type %in% c("Total_clean", "MAGs_DB_mf", "Host_mapped_hf")) %>%
                pivot_wider(names_from = Type, values_from = Number) %>%
                left_join(qpcr_plot_df %>% select(Ct_mean, Sample.Name), by = c("Sample" = "Sample.Name"))

ggplot(df_plot_ct) +
  geom_point(aes(y = MAGs_DB_mf, x = Ct_mean, color = Host), size = 3) +
  scale_x_reverse() +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_color_manual(values=host_order_color_dark) +
    make_theme(leg_pos = "right", guide_nrow = 5, setFill = F, setCol = F)
ggsave("results/figures/10-instrain_plots/3-reads_mapped_to_mag_vs_ct.pdf", width = 12, height = 15)

ggplot(df_plot_ct) +
  geom_point(aes(y = Host_mapped_hf, x = Ct_mean, color = Host), size = 3) +
  scale_x_reverse() +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_color_manual(values=host_order_color_dark) +
    make_theme(leg_pos = "right", guide_nrow = 5, setFill = F, setCol = F)
ggsave("results/figures/10-instrain_plots/3-reads_mapped_to_host_vs_ct.pdf", width = 12, height = 15)

ggplot(df_plot_ct) +
  geom_point(aes(y = Total_clean, x = Ct_mean, color = Host), size = 3) +
  scale_x_reverse() +
  scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  scale_color_manual(values=host_order_color_dark) +
    make_theme(leg_pos = "right", guide_nrow = 5, setFill = F, setCol = F)
ggsave("results/figures/10-instrain_plots/3-reads_sequenced_vs_ct.pdf", width = 12, height = 15)

```

# Estimate host range
```{r}
host_comp_order <- c("mellifera_mellifera",
                    "cerana_cerana",
                    "dorsata_dorsata",
                    "florea_florea",
                    "andreniformis_andreniformis",
                    "andreniformis_florea",
                    "cerana_mellifera",
                    "dorsata_mellifera",
                    "dorsata_florea",
                    "cerana_dorsata",
                    "andreniformis_dorsata",
                    "andreniformis_cerana",
                    "cerana_florea",
                    "florea_mellifera",
                    "andreniformis_mellifera")

# df_shared <- read.csv("results/figures/magotu_shared_same_diff.csv", stringsAsFactors = F) %>%
#               pivot_longer(cols = !c("X"), names_to = "host_comb", values_to = "prop_shared") %>%
#               left_join(magOTU_tax_info, by = c("X" = "magOTU"))

# p_list <- list()
# for (genus in unique(df_shared$Genus)) {
#   p <- ggplot(df_shared %>% filter(Genus == genus),
#   aes(x = factor(host_comb, host_comp_order), y = prop_shared, group = X)
#   ) +
#   geom_point(color = "#000000", size = 3) +
#   geom_point(aes(color = Genus), size = 2) +
#   scale_color_manual(values=genusColors) +
#   new_scale_color() +
#   geom_line(aes(color = X), size = 1) +
#   scale_color_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_shared %>% filter(Genus == genus) %>% pull(X)))), -2)) +
#     labs(x = "Host species pair", y = "") +
#     ggtitle(paste0(genus, " - Proportion of host species pairs that share the magOTU")) +
#     geom_vline(xintercept = 5.5, linetype = "solid", color = "#000000") +
#     make_theme(setFill = F, setCol = F, leg_pos = "none",
#                y_hj = 1, y_vj = 1, y_size = 7,
#                x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
#     ) +
#     ylim(0,1) +
#       theme(plot.title = element_text(size = 10, face = "bold"),
#             panel.background = element_rect(fill = "white", colour = "black"),
#             panel.grid.major = element_line(colour = "grey", linetype = "solid"),
#       )  
#   p_list[[genus]] <- p
# }
# pdf("results/figures/3-magotu_shared.pdf", width = 12, height = 15)
# marrangeGrob(p_list, nrow = 5, ncol = 2)
# dev.off()


# df_shared_ab <- read.csv("results/figures/magotu_shared_same_diff_ab.csv", stringsAsFactors = F) %>%
#               pivot_longer(cols = !c("X"), names_to = "host_comb", values_to = "prop_shared") %>%
#               left_join(magOTU_tax_info, by = c("X" = "magOTU"))

# p_list <- list()
# for (genus in unique(df_shared_ab$Genus)) {
#   p <- ggplot(df_shared_ab %>% filter(Genus == genus),
#   aes(x = factor(host_comb, host_comp_order), y = prop_shared, group = X)
#   ) +
#   geom_point(color = "#000000", size = 3) +
#   geom_point(aes(color = Genus), size = 2) +
#   scale_color_manual(values=genusColors) +
#   new_scale_color() +
#   geom_line(aes(color = X), size = 1) +
#   scale_color_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_shared_ab %>% filter(Genus == genus) %>% pull(X)))), -2)) +
#     labs(x = "Host species pair", y = "") +
#     ggtitle(paste0(genus, " - Proportion of host species pairs that share the magOTU")) +
#     geom_vline(xintercept = 5.5, linetype = "solid", color = "#000000") +
#     make_theme(setFill = F, setCol = F, leg_pos = "none",
#                y_hj = 1, y_vj = 1, y_size = 7,
#                x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
#     ) +
#     ylim(0,1) +
#       theme(plot.title = element_text(size = 10, face = "bold"),
#             panel.background = element_rect(fill = "white", colour = "black"),
#             panel.grid.major = element_line(colour = "grey", linetype = "solid"),
#       )  
#   p_list[[genus]] <- p
# }
# pdf("results/figures/3-magotu_shared_ab.pdf", width = 12, height = 15)
# marrangeGrob(p_list, nrow = 5, ncol = 2)
# dev.off()



# rhodes specificity index (with modification for std.)
# https://hal.sorbonne-universite.fr/hal-03246067/document
# implemented this in python and made "results/figures/magotu_rohdes_index.csv"

prevs_df <- read.csv("results/figures/magotu_prevs.csv") %>%
              pivot_longer(cols = !c("X"), names_to = "host", values_to = "prevalence_in_host") %>%
              mutate(Host = paste0("Apis ", host))


df_plot <- abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0)) %>%
                            select(sample, Host, magOTU, detected, Genus, coverage, copies_cov, rel_cov) %>%
                            left_join(prevs_df, by = c("magOTU" = "X", "Host" = "Host")) %>%
                            select(magOTU, coverage, copies_cov, rel_cov, prevalence_in_host, Genus, Host, sample) %>%
                            left_join(handmade_species_names %>% select(MAG_species_name_final, cluster_name_instrain), by = c("magOTU" = "cluster_name_instrain"))


df_rohdes <- read.csv("results/figures/magotu_rohdes_index.csv", stringsAsFactors = F) %>%
              left_join(magOTU_tax_info, by = c("X" = "magOTU")) %>%
              pivot_longer(cols = !c("X", "Genus"), names_to = "type", values_to = "rohdes_index") %>%
              mutate(specific = ifelse(rohdes_index == 1, "specific", "general")) %>%
              left_join(df_plot %>% select(MAG_species_name_final, magOTU) %>% unique, by = c("X" = "magOTU")) %>%
              mutate(X_old = X) %>%
              mutate(X = MAG_species_name_final)

p_list <- list()
for (genus in unique(df_rohdes$Genus)) {
  p <- ggplot(df_rohdes %>% filter(Genus == genus)
  ) +
  geom_bar(aes(x = X, 
               y = rohdes_index,
               fill = type, 
               color = factor(specific, c("specific", "general"))), size = 0.25, stat = "identity", position = position_dodge2(padding = 0.2)) +
    labs(x = "magOTU", y = "Rohde's index of host specificity", fill = "Input Type", color = "specificity") +
    ggtitle(paste0(genus, " - Rhodes specificity index")) +
    scale_color_manual(values=c("general" = "#ffffff","specific" = "#000000")) +
    make_theme(setFill = T, setCol = F, leg_pos = "right",
               y_hj = 1, y_vj = 1, y_size = 7, leg_size = 6,
               x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
    ) +
      theme(plot.title = element_text(size = 10, face = "bold"),
            panel.background = element_rect(fill = "white", colour = "black"),
            panel.grid.major = element_line(colour = "grey", linetype = "solid"),
      )
  p_list[[genus]] <- p
}
pdf("results/figures/3-magotu_rohdes_types.pdf", width = 12, height = 15)
marrangeGrob(p_list, nrow = 5, ncol = 2)
dev.off()

# Use prevalence, it seems to be the most robust... compare more quantitatively ???...

magotu_order <- df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% arrange(Genus, rohdes_index) %>% pull(X)
ggplot(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__")
  ) +
  geom_bar(aes(y = factor(X, magotu_order), 
               x = rohdes_index,
              #  color = factor(specific, c("specific", "general")),
               fill = Genus
              ), 
              color = "#000000",
              size = 0.25, stat = "identity", position = position_dodge2(padding = 0.2)) +
    labs(y = "magOTU", x = "Rohde's index of host specificity", fill = "Input Type", color = "specificity") +
    scale_color_manual(values=c("general" = "#ffffff","specific" = "#000000")) +
    scale_fill_manual(values=genusColors) +
    facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
    make_theme(setFill = F, setCol = F, leg_pos = "none",
               y_hj = 1, y_vj = 1, y_size = 7, leg_size = 6, guide_nrow = 22,
               x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
    ) +
      theme(plot.title = element_text(size = 10, face = "bold"),
            strip.text.y = element_text(angle = 0),
            panel.background = element_rect(fill = "white", colour = "black"),
            panel.grid.major = element_line(colour = "grey", linetype = "solid")
      )
 ggsave("results/figures/3-magotu_rohdes.pdf", width = 12, height = 15)


magotu_order <- df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% arrange(Genus, rohdes_index) %>% pull(X)
p_rohdes <- ggplot(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__")) +
            geom_bar(aes(y = factor(X, magotu_order), 
                         x = rohdes_index
                         ),
                     fill = "#f6e8c3",
                     color = "#000000",
                     size = 0.25, stat = "identity", position = position_dodge2(padding = 0.2)) +
    labs(y = "magOTU", x = "Rohde's index of host specificity", color = "specificity") +
    # scale_fill_manual(values=genusColors) +
    facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
    make_theme(setFill = F, setCol = F, leg_pos = "none",
               y_hj = 1, y_vj = 1, y_size = 7, leg_size = 6, guide_nrow = 22,
               x_angle = 30, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 7
    ) +
      theme(plot.title = element_text(size = 10, face = "bold"),
            strip.text.y = element_text(angle = 0),
            panel.background = element_rect(fill = "white", colour = "black"),
            panel.grid.major = element_line(colour = "grey", linetype = "solid")
      )
p_hp_m <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>% 
                                      unique %>% filter(Host == "Apis mellifera"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis mellifera")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis mellifera"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_c <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>%
                                      unique %>% filter(Host == "Apis cerana"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis cerana")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis cerana"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_d <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>% 
                                      unique %>% filter(Host == "Apis dorsata"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis dorsata")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis dorsata"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_f <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>% 
                                      unique %>% filter(Host == "Apis florea"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis florea")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis florea"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))
p_hp_a <- ggplot(data.frame(cbind(magOTU = magotu_order)) %>% 
                            left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X, X_old), by = c(magOTU = "X")) %>%
                            left_join(df_plot %>% 
                                      select(prevalence_in_host, Host, magOTU) %>%
                                      unique %>% filter(Host == "Apis andreniformis"), by = c("X_old" = "magOTU")) %>%
                            mutate(prevalence_in_host = ifelse(is.na(Host), 0, prevalence_in_host)) %>%
                            mutate(Host = "Apis andreniformis")) +
        geom_tile(aes(x = Host, y = factor(magOTU, magotu_order), fill = prevalence_in_host, width = 0.1), color = "#000000", size = 0.1) +
        scale_fill_gradient(low = "#FFFFFF", high = host_order_color_dark["Apis andreniformis"]) +
        theme_void() +
        facet_grid(factor(Genus, genera) ~ ., scales = "free", space = "free") +
        # make it square
        # coord_fixed() +
        # color by host
        # scale_x_discrete(position = "top") +
        theme(axis.title.x = element_text(),
              axis.text.x = element_text(),
              legend.position = "none", strip.text.y = element_text(size = 0))

pdf("results/figures/3-magotu_rohdes_with_host.pdf", width = 12, height = 15)
grid.arrange(p_rohdes, p_hp_m, p_hp_c, p_hp_d, p_hp_f, p_hp_a, nrow = 1, widths = c(7,1,1,1,1,1))
dev.off()


abundance_df %>% select(sample, coverage) %>% filter(coverage > 0) %>% pull(sample) %>% unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis mellifera") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis cerana") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis dorsata") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis florea") %>%
  pull(sample) %>% 
  unique %>% length
abundance_df %>% select(sample, coverage, Host) %>%
  filter(coverage > 0) %>%
  filter(Host == "Apis andreniformis") %>%
  pull(sample) %>% 
  unique %>% length

# abundance_df %>% select(sample, coverage, Host, magOTU) %>%
#   filter(coverage > 0) %>%
#   filter(Host == "Apis florea") %>% 
#   filter(grepl("g__Dysgono", magOTU)) %>% 
#   view

# data.frame(cbind(magOTU = magotu_order)) %>% 
#                             left_join(df_rohdes %>% filter(type == "prev") %>% filter(Genus != "g__") %>% select(Genus, X), by = c(magOTU = "X")) %>%
#                             left_join(df_plot %>% 
#                             select(prevalence_in_host, Host, magOTU)) %>%
#                             view


# plot rohdes (prev) vs. abundance per magOTU 
# (each magOTU is at the same prevalence regardless of what sample it comes from)
# the ones at high prevalence will also have more points

p_list_cov_prev <- list()
for (magotu in unique(df_plot$magOTU)) {
  p <- ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")) %>% filter(magOTU == magotu),
  aes(y = prevalence_in_host, x = copies_cov, color = Host)
  ) +
  geom_point(size = 3) +
  labs(y = "Prevalence in host", x = "Coverage (qPCR normalized)") +
  scale_color_manual(values=host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
  p_list_cov_prev[[magotu]] <- p
}

pdf("results/figures/3-magotu_prev_vs_abundance_copies.pdf", width = 12, height = 15)
marrangeGrob(p_list_cov_prev, nrow = 5, ncol = 2)
dev.off()

p_list_cov_prev <- list()
for (magotu in unique(df_plot$magOTU)) {
  p <- ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")) %>% filter(magOTU == magotu),
  aes(y = prevalence_in_host, x = rel_cov, color = Host)
  ) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.01, linetype = "dashed") +
  geom_vline(xintercept = 0.1, linetype = "dashed") +
  labs(y = "Prevalence in host", x = "Relative coverage") +
  scale_color_manual(values=host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
  p_list_cov_prev[[magotu]] <- p
}

pdf("results/figures/3-magotu_prev_vs_abundance_rel_cov.pdf", width = 12, height = 15)
marrangeGrob(p_list_cov_prev, nrow = 5, ncol = 2)
dev.off()


ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")),
  aes(y = prevalence_in_host, x = rel_cov, color = Genus, group = sample)
  ) +
  facet_wrap(~Host, ncol = 2) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.01, linetype = "dashed") +
  geom_vline(xintercept = 0.1, linetype = "dashed") +
  labs(y = "Prevalence in host", x = "Relative coverage") +
  scale_color_manual(values=genusColors) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
ggsave("results/figures/3-magotu_prev_vs_rel_cov.pdf", width = 12, height = 15)

# Could you look at the abundance of the highly specific ones vs the non-specific ones. Do the non-specific ones tend to have a ‘preferred’ host, in which they are most abundant?
ggplot(df_plot %>% left_join(df_rohdes %>% 
  select(!Genus) %>%
  filter(type == "prev"), by = c("magOTU" = "X")),
  aes(y = rohdes_index, x = rel_cov, color = Genus, group = sample)
  ) +
  facet_wrap(~Host, ncol = 2) +
  geom_point(size = 3) +
  geom_vline(xintercept = 0.01, linetype = "dashed") +
  geom_vline(xintercept = 0.1, linetype = "dashed") +
  labs(y = "Rohdes index", x = "Relative coverage") +
  scale_color_manual(values=genusColors) +
  make_theme(setFill = F, setCol = F, leg_pos = "none",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             axis_y_title = 10
  ) +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
ggsave("results/figures/3-magotu_rohdes_vs_rel_cov.pdf", width = 12, height = 15)

genera_to_plot <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 1) %>% 
        arrange(n) %>% pull(Genus) %>% unique

df_rohdes_prev <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 1) %>% select(!n) %>%
        left_join(df_rohdes %>% 
            select(!Genus) %>%
            filter(type == "prev"), by = c("magOTU" = "X_old")
          )

ggplot(df_rohdes_prev,
        aes(x = factor(Host, host_order), y = prevalence_in_host,
            color = rohdes_index, group = sample)
        ) +
  geom_point(size = 4, color = "black") +
  geom_point(size = 3) +
  scale_color_gradient2(low = "#d73027", mid = "#ffffbf", high = "#1a9850", midpoint = 0.5) +
  facet_wrap(~factor(Genus, genera_to_plot), ncol = 2) +
  geom_line(aes(group = magOTU), linewidth = 1.5, alpha = 1) +
  labs(x = "Host" , color = "Rohdes index", y = "Prevalence in host") +
  make_theme(setFill = F, setCol = F, leg_pos = "right",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             modify_guide = F,
             axis_y_title = 10
  ) +
  ylim(0,1) +
    theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )
ggsave("results/figures/3-magotu_rohdes_with_prev.pdf", width = 20, height = 25)

# using paginate_save over MAG_species_name_final

df_rohdes_prev <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 0) %>% select(!n) %>%
        left_join(df_rohdes %>% 
            select(!Genus) %>%
            filter(type == "prev"), by = c("magOTU" = "X_old", "MAG_species_name_final" = "MAG_species_name_final")
          )

p <- ggplot(df_rohdes_prev,
            aes(x = factor(Host, host_order), y = prevalence_in_host, color = rohdes_index, group = sample)) +
      geom_point(size = 4, color = "black") +
      geom_point(size = 3) +
      scale_color_gradient2(low = "#d73027", mid = "#ffffbf", high = "#1a9850", midpoint = 0.5) +
      facet_wrap_paginate(~MAG_species_name_final, ncol = 2, nrow = 7, page = 1) +
      geom_line(aes(group = magOTU), linewidth = 2, alpha = 1) +
      labs(x = "Host" , color = "Rohdes index", y = "Prevalence in host") +
      make_theme(setFill = F, setCol = F, leg_pos = "right",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             modify_guide = F,
             axis_y_title = 10
  ) +
  ylim(0,1) +
  theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )

paginate_save(p, "MAG_species_name_final", "results/figures/3-magotu_rohdes/3-magotu_rohdes_with_prev.pdf", pass_nrow = 6, pass_ncol = 2)


df_rohdes_prev <- df_plot %>% 
        group_by(magOTU) %>% mutate(n = n_distinct(Host)) %>% filter(n > 1) %>% select(!n) %>%
        left_join(df_rohdes %>% 
            select(!Genus) %>%
            filter(type == "prev"), by = c("magOTU" = "X_old", "MAG_species_name_final" = "MAG_species_name_final")
          )

p <- ggplot(df_rohdes_prev,
            aes(x = factor(Host, host_order), y = prevalence_in_host, color = rohdes_index, group = sample)) +
      geom_point(size = 4, color = "black") +
      geom_point(size = 3) +
      scale_color_gradient2(low = "#d73027", mid = "#ffffbf", high = "#1a9850", midpoint = 0.5) +
      facet_wrap_paginate(~MAG_species_name_final, ncol = 2, nrow = 7, page = 1) +
      geom_line(aes(group = magOTU), linewidth = 2, alpha = 1) +
      labs(x = "Host" , color = "Rohdes index", y = "Prevalence in host") +
      make_theme(setFill = F, setCol = F, leg_pos = "right",
             y_hj = 1, y_vj = 1, y_size = 10, leg_size = 6,
             x_angle = 40, x_hj = 1, x_vj = 1, x_size = 7, axis_x_title = 10,
             modify_guide = F,
             axis_y_title = 10
  ) +
  ylim(0,1) +
  theme(plot.title = element_text(size = 10, face = "bold"),
          panel.background = element_rect(fill = "white", colour = "black"),
          panel.grid.major = element_line(colour = "grey", linetype = "solid"),
    )

paginate_save(p, "MAG_species_name_final", "results/figures/3-magotu_rohdes/3-magotu_rohdes_with_prev_2_shared.pdf", pass_nrow = 6, pass_ncol = 2)



# How are the specialist and generaliist species related.  Do we see host specificity at the sub-species level for the generalist (did you try the SNP calling/instrrain approach?)

# also, we really need to do the phylogenies and formally test for co-diversification. 
# this will go hand in hand with specificity index. 
# the non-specific species may still cluster by host at the sub-species level.

```

# Test for correlation between reads and coverage

```{r}
df_rowsums = data.frame(rowSums(abundance_df_matrix_pa_all)) %>%
              rownames_to_column("Sample")
colnames(df_rowsums) = c("Sample", "rowsum")
df_rowsums_cov = data.frame(rowSums(abundance_df_matrix_all)) %>%
              rownames_to_column("Sample")
colnames(df_rowsums_cov) = c("Sample", "rowsum_cov")
plot_corr_species_num <- df_reads %>%
                          select(Sample, MAGs_DB_mf) %>%
                            left_join(df_rowsums) %>%
                            left_join(df_rowsums_cov) %>%
                            mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
                            mutate(Location = Vectorize(get_location_from_sample_name)(Sample))

ggplot(plot_corr_species_num, aes(x = MAGs_DB_mf, y = rowsum)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE) +
  stat_cor(digits = 3, size = 5) +
  scale_color_manual(values=host_order_color) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  labs(x = "Number of reads mapping to MAGs", y = "Number of magOTUs detected") +
  make_theme(leg_pos = "bottom", setCol = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 1)
  ggsave("results/figures/09-Correlation_species_num_reads.pdf")


ggplot(plot_corr_species_num, aes(x = MAGs_DB_mf, y = rowsum, color = Host)) +
  stat_cor(digits = 3, size = 5) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_point(aes(shape = Location), size = 3) +
  scale_color_manual(values=host_order_color_dark) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  labs(x = "Number of reads mapping to MAGs", y = "Number of magOTUs detected") +
  make_theme(leg_pos = "bottom", setCol = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 3)
  ggsave("results/figures/09-Correlation_species_num_reads_by_host.pdf")


ggplot(plot_corr_species_num, aes(x = MAGs_DB_mf, y = rowsum_cov, color = Host)) +
  stat_cor(digits = 3, size = 5) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_point(aes(shape = Location), size = 3) +
  scale_color_manual(values=host_order_color_dark) +
  scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
  labs(x = "Number of reads mapping to MAGs", y = "Sum of coverage of all magOTUs") +
  make_theme(leg_pos = "bottom", setCol = F,
             axis_x_title = 16, axis_y_title = 16,
             guide_nrow = 3)
  

# g <- ggplot(plot_corr_species_num, aes(x = MAGs_DB, y = rowsum, color = Host)) +
#   stat_cor(digits = 3, size = 5) +
#   geom_smooth(method = "lm", se = FALSE) +
#   geom_point(aes(shape = Location, name = Sample), size = 3) +
#   scale_color_manual(values=host_order_color_dark) +
#   scale_x_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
#   labs(x = "Number of reads mapping to MAGs", y = "Number of magOTUs detected") +
#   make_theme(leg_pos = "bottom", setCol = F,
#              axis_x_title = 16, axis_y_title = 16,
#              guide_nrow = 3)
# ggplotly(g)
```

# plot presence of magOTUs and genera


# Overall heatmaps

```{r plots}
df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_pa %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)


row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_pa), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color),
                              border = T
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-presence_absence_heatmap.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix_pa) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_pa),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

pdf("results/figures/09-presence_absence_heatmap_by_location.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix_pa) %>% 
                  as.data.frame() %>% 
                    mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% 
                    # mutate(Identifier = paste0(Host, "_", Location)) %>% 
                    # arrange(Identifier) %>%
                    pull(Host) %>% as.character %>%
                    as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_pa) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_pa),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 6),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
        cluster_columns = F,
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

# filtered for pretty
# Make sure to cross-check later which genera are to be included!
genera_filt <- c(
    "g__Lactobacillus",
    # "g__Bifidobacterium",
    "g__Bombilactobacillus",
    "g__CALYQJ01",
    # "g__Snodgrassella",
    # "g__Gilliamella",
    # "g__CALYQQ01",
    # "g__Apibacter",
    # "g__Bartonella_A",
    # "g__Frischella",
    # "g__Commensalibacter",
    "g__Apilactobacillus"
    # "g__Dysgonomonas",
    # "g__Enterobacter",
    # "g__Pectinatus",
    # "g__Entomomonas",
    # "g__Saezia",
    # "g__Parolsenella",
    # "g__"
  )         
genusColors_filt <- list("g__Bombilactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[1],
                    "g__Lactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[4],
                    "g__CALYQJ01" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[6],
                    # "g__Bifidobacterium" = brewer.pal(11, "Spectral")[3],
                    # "g__Gilliamella" = brewer.pal(11, "Spectral")[11],
                    # "g__CALYQQ01" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[11], "#FFFFFF"))(10), -1)[3],
                    # "g__Frischella" = brewer.pal(11, "Spectral")[8],
                    # "g__Bartonella_A" = brewer.pal(11, "Spectral")[7],
                    # "g__Snodgrassella" = brewer.pal(11, "Spectral")[10],
                    # "g__Apibacter" = brewer.pal(11, "Spectral")[4],
                    "g__Apilactobacillus" = brewer.pal(11, "Spectral")[9],
                    # "g__Commensalibacter" = brewer.pal(8, "Dark2")[6],
                    # "g__Apilactobacillus" = brewer.pal(11, "Spectral")[9],
                    # "g__Dysgonomonas" = brewer.pal(11, "Spectral")[2],
                    # "g__Pectinatus" = brewer.pal(8, "Dark2")[1],
                    # "g__Enterobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[1],
                    # "g__Entomomonas"= head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
                    # "g__Saezia" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
                    # "g__Parolsenella" = brewer.pal(11, "Spectral")[6],
                    "g__" = "#000000"
)
magotus <- unique(colnames(abundance_df_matrix_pa))
magotus_filt <- magotus %>% as.data.frame() %>% left_join(abundance_df %>% ungroup() %>% select(magOTU, Genus) %>% unique, by = c("." = "magOTU")) %>% filter(Genus %in% genera_filt) %>% select(!Genus) %>% pull(.) %>% as.character %>% unique
abundance_df_matrix_pa_filt <- abundance_df_matrix_pa[, magotus_filt] 
df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_pa_filt %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_pa_filt), 
                                gp = gpar(fill = "#081d58", border =NA, lty = "blank"),
                                baseline = 0, bar_width = 0.9),
                              col = list(Location = location_order_color),
                              border = T
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors_filt)), border = T
                    )
# pdf("results/figures/09-presence_absence_heatmap_filt.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_pa_filt) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_pa_filt),
        col = c("#e1ebf4", "#081d58"),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Presence"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

# dev.off()


df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
# replace NAs with 0
df_host_nums[is.na(df_host_nums)] <- 0
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color)
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-coverage_heatmap.pdf", width = 10, height = 15)

column_split <- rownames(abundance_df_matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix),
        col = colorRamp2(c(0.1, 1, 5, 10, 100, 1000), colors = c(brewer.pal(9, "Blues")[1], brewer.pal(9, "Blues")[2], brewer.pal(9, "Blues")[3], brewer.pal(9, "Blues")[4], brewer.pal(9, "Blues")[7], brewer.pal(9, "Blues")[8])),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Coverage"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()


df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_copies %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
df_host_nums[is.na(df_host_nums)] <- 0
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_copies) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
# make it 1 and 0 from numbers before counting
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_copies) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              `Number of magOTUs` = anno_barplot(rowSums(mutate_all(abundance_df_matrix_copies, function(x) ifelse(x > 0, 1, 0))), 
                                gp = gpar(fill = "#081d58", color = "#081d58"),
                                baseline = 0),
                              col = list(Location = location_order_color)
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_copies) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-coverage_heatmap_copies.pdf")

column_split <- rownames(abundance_df_matrix_copies) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_copies) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_copies),
        col = colorRamp2(c(0, 1e2, 1e4, 1e5, 1e6, 1e8), colors = c(brewer.pal(9, "Blues")[1], brewer.pal(9, "Blues")[2], brewer.pal(9, "Blues")[3], brewer.pal(9, "Blues")[4], brewer.pal(9, "Blues")[7], brewer.pal(9, "Blues")[8])),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Coverage"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

df_host_nums <- abundance_df %>% filter(!is.na(magOTU)) %>%
                  mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0)) %>%
                    select(Host, magOTU, detected) %>%
                      filter(magOTU %in% c(abundance_df_matrix_rel %>% colnames() %>% as.character())) %>%
                      group_by(magOTU, Host) %>%
                        summarise(number = sum(detected)) %>%
                          pivot_wider(magOTU, names_from = Host, values_from = number)
df_host_nums[is.na(df_host_nums)] <- 0
row_ra = rowAnnotation(`Host distribution` = anno_barplot(as.matrix(df_host_nums[,-1]), 
                                gp = gpar(fill = host_order_color[sort(names(host_order_color))], border =NA, lty = "blank"),
                                baseline = 0),
                              col = list(`Host distribution` = host_order_color
                              ), border = T
                          )
column_ha = HeatmapAnnotation(Host = rownames(abundance_df_matrix_rel) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
column_ba = HeatmapAnnotation(Location = rownames(abundance_df_matrix_rel) %>% as.data.frame() %>% mutate(Location = Vectorize(get_location_from_sample_name)(.)) %>% pull(Location),
                              # `Number of magOTUs` = anno_barplot(rowSums(abundance_df_matrix_rel), 
                              #   gp = gpar(fill = "#081d58", color = "#081d58"),
                              #   baseline = 0),
                              col = list(Location = location_order_color)
                              )
row_la = rowAnnotation(Genus = colnames(abundance_df_matrix_rel) %>% as.data.frame() %>% left_join(magOTU_tax_info %>% ungroup(), by = c(. = "magOTU")) %>% pull(Genus),
                       col = list(Genus = unlist(genusColors)), border = T
                    )
pdf("results/figures/09-coverage_heatmap_relative.pdf")

column_split <- rownames(abundance_df_matrix_rel) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.numeric
row_split <- colnames(abundance_df_matrix_rel) %>% as.data.frame() %>% left_join(magOTU_tax_info, by = c(. = "magOTU")) %>% pull(Genus) %>% as.factor %>% as.numeric
heatmap_obj = Heatmap(t(abundance_df_matrix_rel),
        col = colorRamp2(c(0, 1e-4, 1e-3, 1e-2, 1e-1, 1), colors = c(brewer.pal(9, "Blues")[1], brewer.pal(9, "Blues")[2], brewer.pal(9, "Blues")[3], brewer.pal(9, "Blues")[4], brewer.pal(9, "Blues")[7], brewer.pal(9, "Blues")[8])),
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Coverage"),
        top_annotation = column_ba, 
        right_annotation = row_ra,
        left_annotation = row_la,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

dev.off()

g_list <- list()
for (a_sample in unique(df_plot$sample)) {
  g <- ggplot(df_plot %>% filter(sample == a_sample), aes(x = prevalence_in_host, 
                    y = coverage)
                ) +
                geom_point(aes(color = Genus), size = 5) +
                geom_smooth(method = "lm", se = FALSE, color = "#000000") +
                ggtitle(a_sample) +
                stat_cor(digits = 3, size = 5) +
                scale_y_continuous(trans = "log10") +
                scale_color_manual(values=genusColors) +
                labs(x = "Prevalence in host", y = "Coverage") +
                make_theme(leg_pos = "none", setCol = F,
                           axis_x_title = 10, axis_y_title = 10,
                           guide_nrow = 1)
  g_list[[a_sample]] <- g
}
pdf("results/figures/09-abundance_vs_prevalence_in_host.pdf", width = 10, height = 15)
marrangeGrob(g_list, nrow = 5, ncol = 2)
dev.off()

```


# Community measures

Do not do this on the front-end of Curnagl. Run elsewhere or submit as a job!

```{r}
df_plot_cum_curve_by_loc <- read.csv("results/figures/magotu_cum_curves_loc_all.csv")

df_plot_cum_curve <- read.csv("results/figures/magotu_cum_curves.csv") %>%
                      mutate(host = paste0("Apis ", host))

ggplot(data = df_plot_cum_curve, aes(x = sample_size, y = num_magotus, color = factor(host, host_order))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method = "loess", se = F) +
                          labs(color = "Host species", x = "Number of Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color) +
                            make_theme(leg_pos = "bottom", setCol = F,
                                       axis_x_title = 16, axis_y_title = 16,
                                       guide_nrow = 1)
                            ggsave("results/figures/09-Cumulative_curve.pdf")

host_order_color_ext <- c("Apis mellifera (India)" = "#4292c6",
                          "Apis mellifera (Malaysia)" = "#08519c",
                          "Apis mellifera (Japan)" = "#a6cee3",
                          "Apis mellifera (Switzerland)" = "#41b6c4",
                          "Apis cerana (Japan)" = "#fbb4ae", 
                          "Apis cerana (India)" = "#f46d43", 
                          "Apis cerana (Malaysia)" = "#e31a1c", 
                          "Apis dorsata (Malaysia)" = "#984ea3",
                          "Apis dorsata (India)" = "#cab2d6",
                          "Apis florea (Malaysia)" = "#4daf4a",
                          "Apis florea (India)" = "#b2df8a",
                          "Apis andreniformis (Malaysia)" = "#ff7f00")
ggplot(data = df_plot_cum_curve_by_loc, aes(x = sample_size, y = num_magotus, color = factor(host))) +
                      geom_jitter(position = position_dodge(width=0.7)) +
                        geom_smooth(method = "loess", se = T) +
                          labs(color = "Host species", x = "Number of Bees", y = "Number of magOTUs") +
                          scale_color_manual(values=host_order_color_ext) +
                            make_theme(leg_pos = "bottom", leg_size = 11,
                                       axis_x_title = 16, axis_y_title = 16,
                                       setCol = F, guide_nrow = 4) +
                              theme(legend.spacing.x = unit(1.0, 'cm'))
                            ggsave("results/figures/09-Cumulative_curve_by_location.pdf")

# code for betapart functions from
# https://github.com/cran/betapart

# Latest commit a222c6d on Feb 24, 2012
betapart.core <- function(x){
	
	# test for a binary numeric matrix or data.frame
	if(! is.matrix(x)){
		x<-as.matrix(x)
  	}
	
	if(! is.numeric(x))
    	stop("The data in x is not numeric.",call.=TRUE)
	
	# simple test for binary data
	xvals <-  unique(as.vector(x))
	if (any(!is.element(xvals, c(0, 1)))) 
        stop("The table contains values other than 0 and 1: data should be presence/absence.", call. = TRUE)

	shared <- x %*% t(x)
    not.shared <-  abs(sweep(shared, 2, diag(shared)))
		
	sumSi <- sum(diag(shared)) # species by site richness
    St <- sum(colSums(x) > 0)  # regional species richness
    a <- sumSi - St            # multi site shared species term

    sum.not.shared <- not.shared + t(not.shared)
    max.not.shared <- pmax(not.shared, t(not.shared))
    min.not.shared <- pmin(not.shared, t(not.shared))

	computations<-list(data=x, sumSi=sumSi, St=St, a=a, shared=shared, not.shared=not.shared, 
	                   sum.not.shared=sum.not.shared, max.not.shared=max.not.shared, 
	                   min.not.shared=min.not.shared)
    class(computations)<-"betapart"

	return(computations)
} 
# Latest commit 4b59d9c on Feb 24, 2012
beta.pair <- function(x, index.family="sorensen"){
	
	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
			beta.sim <- x$min.not.shared / (x$min.not.shared + x$shared)
			beta.sne <- ((x$max.not.shared - x$min.not.shared) / ((2 * x$shared) + x$sum.not.shared)) * (x$shared / (x$min.not.shared + x$shared))
            beta.sor <- x$sum.not.shared / (2 * x$shared + x$sum.not.shared)
                
            pairwise <- list(beta.sim=as.dist(beta.sim), beta.sne=as.dist(beta.sne),beta.sor=as.dist(beta.sor))},
		jaccard = {
			beta.jtu <- (2*x$min.not.shared) / ((2*x$min.not.shared) + x$shared)
	        beta.jne <- ((x$max.not.shared - x$min.not.shared) / (x$shared + x$sum.not.shared)) * (x$shared / ((2*x$min.not.shared) + x$shared))
	        beta.jac <- x$sum.not.shared / (x$shared + x$sum.not.shared)

	        pairwise <- list(beta.jtu=as.dist(beta.jtu), beta.jne=as.dist(beta.jne),beta.jac=as.dist(beta.jac))})

	return(pairwise)
}
beta.multi <- function(x, index.family="sorensen"){

	# test for a valid index
	index.family <- match.arg(index.family, c('jaccard','sorensen'))
	
	# test for pre-existing betapart objects
	if (! inherits(x, "betapart")){	
		x <- betapart.core(x)
	}

	maxbibj <- sum(x$max.not.shared[lower.tri(x$max.not.shared)])
    minbibj <- sum(x$min.not.shared[lower.tri(x$min.not.shared)])
	
	# run the analysis given the index
	switch(index.family,
		sorensen = {
            beta.sim <- minbibj / (minbibj + x$a)
            beta.sne <- (x$a / (minbibj + x$a)) * ((maxbibj - minbibj) / ((2 * x$a) + maxbibj + minbibj))
            beta.sor <- (minbibj + maxbibj) / (minbibj + maxbibj + (2 * x$a))

           	multi <- list(beta.SIM=beta.sim, beta.SNE=beta.sne,beta.SOR=beta.sor)},
		jaccard = {
            beta.jtu <- (2*minbibj) / ((2*minbibj) + x$a)
            beta.jne <- (x$a / ((2*minbibj) + x$a)) * ((maxbibj - minbibj) / ((x$a) + maxbibj + minbibj))
            beta.jac <- (minbibj + maxbibj) / (minbibj + maxbibj + x$a)

           	multi <- list(beta.JTU=beta.jtu, beta.JNE=beta.jne, beta.JAC=beta.jac)})

	return(multi)

}

df_magOTUs_vegan_pa <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
# df_pcoa <- dist_matrix_betapart$beta.sor
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_pa, "sorensen")
pdf("results/figures/09-PCoA_sorensen_microbiome.pdf")
pcoa_plot(dist_matrix_betapart$beta.sor, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()

dist_matrix_atch <- vegdist(abundance_df_matrix_copies, method = "robust.aitchison")
pdf("results/figures/09-PCoA_atchinson_microbiome.pdf")
pcoa_plot(dist_matrix_atch, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()

dist_matrix_atch_rel <- vegdist(abundance_df_matrix_rel, method = "robust.aitchison")
pdf("results/figures/09-PCoA_atchinson_relative_ab_microbiome.pdf")
pcoa_plot(dist_matrix_atch_rel, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()


# beta.SIM value of the turnover component, measured as Simpson dissimilarity
# beta.SNE value of the nestedness component, measured as nestedness-resultant fraction of
# Sorensen dissimilarity
# beta.SOR value of the overall beta diversity, measured as Sorensen dissimilarity

pdf("results/figures/09-PCoA_sorensen_microbiome_turnover.pdf")
pcoa_plot(dist_matrix_betapart$beta.sim, df_meta_complete %>% select(ID, Species), variable = "Species", color_add=T, color_list=host_order_color_dark)
dev.off()



# pcoa_plot(dist_matrix_betapart$beta.sne, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
# pcoa_plot(dist_matrix_betapart$beta.sim, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

# dist_matrix_betapart <- beta.pair(df_magOTUs_vegan, "jaccard")
# pdf("results/figures/09-PCoA_jaccard_microbiome.pdf")
# pcoa_plot(dist_matrix_betapart$beta.jac, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
# dev.off()
# pcoa_plot(dist_matrix_betapart$beta.jne, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
# pcoa_plot(dist_matrix_betapart$beta.jtu, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

dist_matrix_betapart_multi <- beta.multi(df_magOTUs_vegan_pa, "sorensen")
str(dist_matrix_betapart_multi)

# beta.sim dist object, dissimilarity matrix accounting for spatial turnover (replacement),
# measured as Simpson pair-wise dissimilarity
# beta.sne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Sorensen pair-wise dissimilarity
# beta.sor dist object, dissimilarity matrix accounting for total dissimilarity, measured as
# Sorensen pair-wise dissimilarity (a monotonic transformation of beta diversity)
# For index.family="jaccard" the three matrices are:
# beta.jtu dist dissimilarity matrix accounting for spatial turnover, measured as the turnoverfraction of Jaccard pair-wise dissimilarity
# beta.jne dist object, dissimilarity matrix accounting for nestedness-resultant dissimilarity, measured as the nestedness-fraction of Jaccard pair-wise dissimilarity
# beta.jac dist object, dissimilarity matrix accounting for beta diversity, measured as Jaccard pair-wise dissimilarity (a monotonic transformation of beta diversity)

grouping_vec <- df_meta_complete %>% filter(ID %in% rownames(df_magOTUs_vegan_pa)) %>% pull(Species)
anosim(dist_matrix_betapart$beta.sor, grouping_vec, distance = "sorenson", permutations = 9999)

df_to_test <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
                        # filter(rownames(.) %in% c(samples_IN, samples_MY))
                          # filter(rownames(.) %in% c(samples_am, samples_ac, samples_af, samples_ad))
adonis2(df_to_test ~ Species + Country, data = df_meta_complete %>% filter(ID %in% rownames(df_to_test)))
# Distances from
# https://link.springer.com/article/10.1007/s11756-022-01123-6/tables/4
# 		7	5	3	1	2
# mellifera	7		0.112	0.128	0.132	0.146
# cerana	5	0.112		0.133	0.13	0.119
# dorsata	3	0.128	0.133		0.095	0.108
# florea	1	0.132	0.13	0.095		0.047
# andreniformis	2	0.146	0.119	0.108	0.047	

get_host_divergence <- function(sample_1, sample_2) {
  host_1 <- get_host_from_sample_name(sample_1)
  host_1 <- strsplit(host_1, " ")[[1]][[2]]
  host_2 <- get_host_from_sample_name(sample_2)
  host_2 <- strsplit(host_2, " ")[[1]][[2]]
  mellifera <- c(0, 0.112, 0.128, 0.132, 0.146)
  cerana <- c(0.112, 0, 0.133, 0.13, 0.119)
  dorsata <- c(0.128, 0.133, 0, 0.095, 0.108)
  florea <- c(0.132, 0.13, 0.095, 0, 0.047)
  andreniformis <- c(0.146, 0.119, 0.108, 0.047, 0)
  # mellifera <- c(0, 1, 2, 4, 4)
  # cerana <- c(1, 0, 2, 4, 4)
  # dorsata <- c(2, 2, 0, 3, 3)
  # florea <- c(4, 4, 3, 0, 1)
  # andreniformis <- c(4, 4, 3, 1, 0)
  host_divergence_mat <- rbind(mellifera, cerana, dorsata, florea, andreniformis)
  colnames(host_divergence_mat) <- c("mellifera", "cerana", "dorsata", "florea", "andreniformis")
  index_list <- list("mellifera" = 1, "cerana" = 2, "dorsata" = 3, "florea" = 4, "andreniformis" = 5)
  host_divergence <- host_divergence_mat[as.numeric(index_list[host_1]), as.numeric(index_list[host_2])]
  return(host_divergence)
}
df_magOTUs_vegan_IN_MY <- abundance_df_matrix_pa %>%
                      filter(rowSums(.) > 0)
dist_matrix_betapart <- beta.pair(df_magOTUs_vegan_IN_MY, "sorensen")
sample_bac_div <- dist_matrix_betapart$beta.sor
sample_host_div <- data.frame()
for (sample in samples) {
  sample_host_div <- rbind(sample_host_div, rep(0, length((samples))))
}
colnames(sample_host_div) <- samples
rownames(sample_host_div) <- samples
for (sample_1 in samples) {
  for (sample_2 in samples) {
    ind_1 <- as.numeric(which(samples == sample_1))
    ind_2 <- as.numeric(which(samples == sample_2))
    sample_host_div[ind_1, ind_2] <- get_host_divergence(sample_1, sample_2)
  }
}

# samples_subset_am <- df_meta_complete %>% filter(Species == "Apis mellifera") %>% pull(ID)
# samples_subset_ac <- df_meta_complete %>% filter(Species == "Apis cerana") %>% pull(ID)
# samples_subset_ad <- df_meta_complete %>% filter(Species == "Apis dorsata") %>% pull(ID)
# samples_subset_af <- df_meta_complete %>% filter(Species == "Apis florea") %>% pull(ID)
# samples_subset_aa <- df_meta_complete %>% filter(Species == "Apis andreniformis") %>% pull(ID)

heatmap(sample_bac_div %>% as.matrix)

sample_host_div <- sample_host_div %>% 
  filter(row.names(.) %in% as.vector(as.matrix(sample_bac_div) %>% row.names))

sample_host_div_df <- sample_host_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                        pivot_longer(!Sample, names_to = "Compared", values_to = "dist_h")
sample_bac_div_df <- sample_bac_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                          pivot_longer(!Sample, names_to = "Compared", values_to = "dist_b") %>%
                          mutate(Host_s = Vectorize(get_host_from_sample_name)(Sample)) %>%
                          mutate(Host_c = Vectorize(get_host_from_sample_name)(Compared)) %>%
                          mutate(Type_b = ifelse(Host_s == Host_c, "Within", "Between")) %>%
                          mutate(Hosts_compared = paste0(Host_s, "_", Host_c))
dissimilarities_df_median <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, median_b = median(dist_b)) %>%
                        unique
dissimilarities_df_mean <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, mean_b = median(dist_b)) %>%
                        unique
dissimilarities_df <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared)
                        # filter(Sample %in% c("M1.2", "C1.2", "D1.2", "F1.2", "A1.2"))

mantel(sample_host_div, sample_bac_div, method="pearson", permutations=1000)

ggplot() +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_c,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_s,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
    labs(x = "Host divergence (# nodes between)", y = "Microbiome dissimilarity (jaccard)") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 1)
    ggsave("results/figures/09-Microbe_host_dissimilarities_violin_sorenson.pdf")

ggplot() +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                #  fill = Host_c,
                 color = Host_c,
                #  shape = Type_b
             ), width=0.001, size = 4, alpha = 0.5
            ) +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                #  fill = Host_c,
                 color = Host_s,
                #  shape = Type_b
             ), width=0.001, size = 4, alpha = 0.5
            ) +
  # geom_jitter(data=dissimilarities_df_mean,
  #            aes(x = dist_h,
  #                y = mean_b
  #            ), size = 3, width = 0
  #           ) +
  geom_jitter(data=dissimilarities_df_median,
             aes(x = dist_h,
                 y = median_b
             ), size = 3, width = 0
            ) +
    labs(x = "Host divergence (based on 16S and 12S)", y = "Microbiome dissimilarity (jaccard)", color = "Host species") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 2,
               leg_size = 12,
               axis_x_title = 15, axis_y_title = 15
    )
  ggsave("results/figures/09-Microbe_host_dissimilarities_sorenson.pdf")
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

split_by_pattern <- function(my_name, pattern, index) {
  return(strsplit(my_name, pattern)[[1]][[index]])
}
bac_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(dist_h, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = median_b) %>%
                    column_to_rownames("Compared1")
host_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(median_b, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = dist_h) %>%
                    column_to_rownames("Compared1")

bac_median_mat[1,]
host_median_mat[1,]

mantel(as.dist(bac_median_mat), as.dist(host_median_mat), method = "pearson", permutations = 9999)
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

# sample_bac_div <- dist_matrix_atch
sample_bac_div <- dist_matrix_atch_rel
sample_host_div <- data.frame()
for (sample in samples) {
  sample_host_div <- rbind(sample_host_div, rep(0, length((samples))))
}
colnames(sample_host_div) <- samples
rownames(sample_host_div) <- samples
for (sample_1 in samples) {
  for (sample_2 in samples) {
    ind_1 <- as.numeric(which(samples == sample_1))
    ind_2 <- as.numeric(which(samples == sample_2))
    sample_host_div[ind_1, ind_2] <- get_host_divergence(sample_1, sample_2)
  }
}

# samples_subset_am <- df_meta_complete %>% filter(Species == "Apis mellifera") %>% pull(ID)
# samples_subset_ac <- df_meta_complete %>% filter(Species == "Apis cerana") %>% pull(ID)
# samples_subset_ad <- df_meta_complete %>% filter(Species == "Apis dorsata") %>% pull(ID)
# samples_subset_af <- df_meta_complete %>% filter(Species == "Apis florea") %>% pull(ID)
# samples_subset_aa <- df_meta_complete %>% filter(Species == "Apis andreniformis") %>% pull(ID)

heatmap(sample_bac_div %>% as.matrix)

sample_host_div <- sample_host_div %>% 
  filter(row.names(.) %in% as.vector(as.matrix(sample_bac_div) %>% row.names))

sample_host_div_df <- sample_host_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                        pivot_longer(!Sample, names_to = "Compared", values_to = "dist_h")
sample_bac_div_df <- sample_bac_div %>%
                        as.matrix %>%
                        as.data.frame %>%
                        rownames_to_column("Sample") %>%
                        filter(Sample %in% samples_IN_MY) %>%
                          pivot_longer(!Sample, names_to = "Compared", values_to = "dist_b") %>%
                          mutate(Host_s = Vectorize(get_host_from_sample_name)(Sample)) %>%
                          mutate(Host_c = Vectorize(get_host_from_sample_name)(Compared)) %>%
                          mutate(Type_b = ifelse(Host_s == Host_c, "Within", "Between")) %>%
                          mutate(Hosts_compared = paste0(Host_s, "_", Host_c))
dissimilarities_df_median <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, median_b = median(dist_b)) %>%
                        unique
dissimilarities_df_mean <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared) %>%
                        summarise(dist_h, mean_b = median(dist_b)) %>%
                        unique
dissimilarities_df <- left_join(sample_bac_div_df, sample_host_div_df) %>%
                        group_by(Hosts_compared)
                        # filter(Sample %in% c("M1.2", "C1.2", "D1.2", "F1.2", "A1.2"))

mantel(sample_host_div, sample_bac_div, method="pearson", permutations=1000)

ggplot() +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_c,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
  geom_violin(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                 fill = Host_s,
                #  color = Host_s,
                #  shape = Type_b
             ), width=0.015, 
             position = "identity",
             alpha = 0.5
            ) +
    labs(x = "Host divergence (# nodes between)", y = "Microbiome dissimilarity (jaccard)") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 1)
    ggsave("results/figures/09-Microbe_host_dissimilarities_violin_atchinson.pdf")

ggplot() +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                #  fill = Host_c,
                 color = Host_c,
                #  shape = Type_b
             ), width=0.001, size = 4, alpha = 0.5
            ) +
  geom_jitter(data=dissimilarities_df,
             aes(x = dist_h,
                 y = dist_b,
                 group = Hosts_compared,
                #  fill = Host_c,
                 color = Host_s,
                #  shape = Type_b
             ), width=0.001, size = 4, alpha = 0.5
            ) +
  # geom_jitter(data=dissimilarities_df_mean,
  #            aes(x = dist_h,
  #                y = mean_b
  #            ), size = 3, width = 0
  #           ) +
  geom_jitter(data=dissimilarities_df_median,
             aes(x = dist_h,
                 y = median_b
             ), size = 3, width = 0
            ) +
    labs(x = "Host divergence (based on 16S and 12S)", y = "Microbiome dissimilarity (jaccard)", color = "Host species") +
    scale_fill_manual(values = host_order_color) +
    scale_color_manual(values = host_order_color_dark) +
    make_theme(setCol = F, setFill = F, guide_nrow = 2,
               leg_size = 12,
               axis_x_title = 15, axis_y_title = 15
    )
  ggsave("results/figures/09-Microbe_host_dissimilarities_atchinson.pdf")
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

split_by_pattern <- function(my_name, pattern, index) {
  return(strsplit(my_name, pattern)[[1]][[index]])
}
bac_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(dist_h, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = median_b) %>%
                    column_to_rownames("Compared1")
host_median_mat <- dissimilarities_df_median %>%
                    ungroup() %>%
                    mutate(Compared1 = Vectorize(split_by_pattern)(Hosts_compared, "_", 1)) %>%
                    mutate(Compared2 = Vectorize(split_by_pattern)(Hosts_compared, "_", 2)) %>%
                    select(!c(median_b, Hosts_compared)) %>%
                    pivot_wider(Compared1, names_from = Compared2, values_from = dist_h) %>%
                    column_to_rownames("Compared1")

bac_median_mat[1,]
host_median_mat[1,]

mantel(as.dist(bac_median_mat), as.dist(host_median_mat), method = "pearson", permutations = 9999)
cor.test(dissimilarities_df_median$dist_h, dissimilarities_df_median$median_b, method = "pearson")

# edit tree to look prettier later
# plot_tree(ps, nodelabf=nodeplotblank, color="Host", label.tips = "Name", ladderize="left", plot.margin=0.3) + 
#   scale_color_manual(values = host_order_color)
#   ggsave("results/figures/09-Plot_bac120_tree_with_host.pdf")


# load tree(s) and then do this
dist_mat_unifrac <- UniFrac(prune_samples(sample_sums(ps) > 0, ps), weighted=F, normalized=T, parallel=FALSE, fast=TRUE)
pcoa_plot(dist_mat_unifrac, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

dist_mat_unifrac_wt <- UniFrac(prune_samples(sample_sums(ps) > 0, ps), weighted=T, normalized=F, parallel=FALSE, fast=TRUE)
pcoa_plot(dist_mat_unifrac_wt, df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)

reformat_host_name <- function(x) {
  paste0(strsplit(x, " ")[[1]], collapse = "_")
}
rename_mag_labels <- function(x) {
  return(paste0(tax_table_mat %>% as.data.frame %>% filter(rownames(.) == x) %>% pull(Name), "---", x))
}

host_tree_rooted <- root(host_tree, "Bombus_terrestris", resolve.root = TRUE)

mags_tree_Lacto_path <- "/scratch/aprasad/20230313_apis_species_comparison/results/11_phylogenies/02_orthofinder_results/g__Lactobacillus/Results_g__Lactobacillus_iqtree/Species_Tree/SpeciesTree_rooted.txt"
mag_tree_Lacto <- read.tree(mags_tree_Lacto_path)
  
associations_df <- abundance_df %>% filter(!is.na(magOTU)) %>%
                    mutate(detected = ifelse(coverage > 1 & breadth > 0.8, 1, 0))  %>%
                    select(sample, Host, detected, magOTU, Genus) %>%
                    mutate(representative_genome = Vectorize(get_rep_genome_magotu)(magOTU, Genus)) %>%
                    select(!magOTU) %>%
                    filter(!is.na(representative_genome)) %>%
                    filter(!is.na(Genus)) %>%
                    left_join(vis_magOTUs_df_all %>% ungroup() %>% select(magOTU, ID), by = c("representative_genome" = "ID")) %>%
                      mutate(Host = Vectorize(reformat_host_name)(Host)) %>%
                      filter(representative_genome %in% mag_tree_Lacto$tip.label) %>%
                      filter(detected == 1) %>%
                      select(representative_genome, Host) %>%
                      unique()
mag_tree_Lacto$tip.label <- lapply(mag_tree_Lacto$tip.label, rename_mag_labels)
# cols <- unlist(lapply(as.character(associations_df$representative_genome), function(x) genusColors_char[strsplit(x, "---")[[1]]][[1]]))
cophyloplot(mag_tree_Lacto, collapse.singles(host_tree), assoc = associations_df, space = 100, gap = 15, lwd = 4, show.tip.label = F)
# nodelabels.cophylo(..., which=c("left","right"))
# edgelabels.cophylo(..., which=c("left","right"))
# tiplabels.cophylo(..., which=c("left","right"))
```

# number of reads used for profiling and rarefaction curves

```{r}
# # save.image("/scratch/aprasad/20230313_apis_species_comparison/results/figures/workspace_generaldata_chunks_20230611.RData")
# load("/scratch/aprasad/20230313_apis_species_comparison/results/figures/workspace_generaldata_chunks_20230611.RData")
# abundance_df_matrix_copies_int <- abundance_df_matrix_copies %>% as.matrix %>% as.data.frame %>% mutate_all(as.integer)
# S <- specnumber(abundance_df_matrix_copies_int)
# raremax <- min(rowSums(abundance_df_matrix_copies_int))
# Srare <- rarefy(abundance_df_matrix_copies_int, raremax)
# pdf("results/figures/09-rarefaction_curve_test.pdf")
# rarecurve(abundance_df_matrix_copies_int, step = 20, 
#           sample = raremax, 
#           col = "blue", 
#           cex = 0.6,
#           main = "rarecurve()")
# dev.off()

# reads used by instrain:
df_mapping_info <- data.frame()
for (sample in samples_IN_MY) {
  mapping_info_path <- paste0("results/10_instrain/02_instrain_profile/", sample, "/output/", sample, "_mapping_info.tsv")
  if (file.exists(mapping_info_path)) {
    print(paste0("Reading ", sample))
    df_mapping <- read.csv(mapping_info_path, sep = "\t", skip = 1, nrows=1)
    df_mapping_info <- bind_rows(df_mapping_info, cbind(df_mapping, "sample"=sample))
  }
}

ggplot(df_mapping_info %>% mutate(Host = Vectorize(get_host_from_sample_name)(sample)), 
      aes(x = filtered_pairs,
          y = sample,
          fill = Host)) +
  geom_bar(stat = "identity") +
  labs(x = "Number of reads", y = "Sample") +
  scale_fill_manual(values = host_order_color) +
  scale_x_continuous(labels = scales::comma) +
  make_theme(setCol = F, setFill = F, guide_nrow = 1, y_size = 7)
  ggsave("results/figures/09-number_of_reads_filtered_instrain.pdf")

# correlation between number of reads and number of species detected?

spec_nums_df <- rowSums(abundance_df_matrix_pa[, magotu_order]) %>% as.data.frame %>% rownames_to_column("sample")
colnames(spec_nums_df) <- c("sample", "species_detected")

ggplot(df_mapping_info %>% 
        mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(sample)) %>%
        left_join(spec_nums_df) %>%
        mutate(species_detected = ifelse(is.na(species_detected), 0, species_detected)),
      aes(x = filtered_pairs,
          y = species_detected,
          color = Host)) +
  geom_smooth(method = "lm", se = FALSE) +
  stat_cor(digits = 3, size = 5) +
  geom_point(size = 3) +
  labs(x = "Number of reads", y = "Species") +
  scale_color_manual(values = host_order_color_dark) +
  make_theme(setCol = F, setFill = F, guide_nrow = 1)


```

# Bar plots by genus

```{r}
# plot abundance by genus

colnames(vis_magOTUs_df)
handmade_species_names <- read.csv("config/Species_MAG_Cluster-names.txt", header = T, sep = "\t") %>% 
colnames(abundance_df)
df_plot <- abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0)) %>%
                            select(sample, Host, magOTU, detected, Genus, coverage, copies_cov, rel_cov) %>%
                            left_join(prevs_df, by = c("magOTU" = "X", "Host" = "Host")) %>%
                            select(magOTU, coverage, copies_cov, rel_cov, prevalence_in_host, Genus, Host, sample) %>%
                            left_join(handmade_species_names %>% select(MAG_species_name_final, cluster_name_instrain), by = c("magOTU" = "cluster_name_instrain"))
ggplot(df_plot) +
  geom_bar(aes(x = sample,
               y = rel_cov,
               fill = Genus), stat = "identity") +
  # facet_grid(Host ~ ., scale = "free_x") +
  facet_grid(~Host, scale = "free_x") +
  labs(y = "Sample", x = "Relative abundance") +
  scale_fill_manual(values=genusColors) +
  make_theme(theme_name=theme_few(), leg_pos="none",
             # guide_nrow = 6,
             setFill = F, setCol = F,
             x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
ggsave("results/figures/09-relative_abundance_sep.pdf")

p_list <- list()
for (genus in unique(df_plot$Genus)) {
  p <- ggplot(df_plot %>% filter(Genus == genus) %>% group_by(sample) %>% mutate(rel_cov = rel_cov/sum(rel_cov))) +
    geom_bar(aes(x = sample,
                 y = rel_cov,
                 fill = MAG_species_name_final), color = "black", linewidth = 0.1, stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance", fill = "Species") +
    scale_fill_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_plot %>% filter(Genus == genus) %>% pull(MAG_species_name_final)))), -2)) +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="right",
               guide_nrow = 10,
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  p_list[[genus]] <- p
}

pdf("results/figures/09-relative_abundance_by_genus.pdf", width = 15, height = 10)
marrangeGrob(p_list, nrow = 2, ncol = 1)
dev.off()


# genus <- "g__Lactobacillus"


p_list <- list()
for (genus in unique(df_plot$Genus)) {
  p <- ggplot(df_plot %>% filter(Genus == genus) %>% group_by(sample) %>% mutate(rel_cov = rel_cov/sum(rel_cov))) +
    geom_bar(aes(x = sample,
                 y = rel_cov,
                 fill = MAG_species_name_final), color = "black", linewidth = 0.1, stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance", fill = "Species") +
    scale_fill_manual(values=make_col_list(df_plot$MAG_species_name_final)) +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="right",
               guide_nrow = 10,
              #  guide_nrow = length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU))),
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  p_list[[genus]] <- p
}

pdf("results/figures/09-relative_abundance_by_genus_dist_colors.pdf", width = 15, height = 10)
marrangeGrob(p_list, nrow = 2, ncol = 1)
dev.off()

p_list <- list()
for (genus in unique(df_plot$Genus)) {
  p <- ggplot(df_plot %>% filter(Genus == genus) %>% group_by(sample) %>% mutate(rel_cov = rel_cov/sum(rel_cov))) +
    geom_bar(aes(x = sample,
                 y = coverage,
                 fill = MAG_species_name_final), color = "black", linewidth = 0.1, stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance", fill = "Species") +
    scale_fill_manual(values=make_col_list(df_plot$MAG_species_name_final)) +
    scale_y_continuous(trans = "log") +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="right",
               guide_nrow = 10,
              #  guide_nrow = length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU))),
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  p_list[[genus]] <- p
}

pdf("results/figures/09-abundance_coverage_by_genus_dist_colors.pdf", width = 15, height = 10)
marrangeGrob(p_list, nrow = 2, ncol = 1)
dev.off()


p_list <- list()
for (genus in unique(df_plot$Genus)) {
  p <- ggplot(df_plot %>% filter(Genus == genus)) +
    geom_bar(aes(x = sample,
                 y = rel_cov,
                 fill = magOTU), color = "white", stat = "identity") +
    # facet_grid(factor(Host, host_order) ~ ., scale = "free_x") +
    facet_grid(~factor(Host, host_order), scale = "free_x") +
    ggtitle(genus) +
    labs(y = "Sample", x = "Relative abundance") +
    scale_fill_manual(values=head(colorRampPalette(c(genusColors[[genus]], "#FFFFFF"))(2+length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU)))), -2)) +
    # scale_fill_manual(values=genusColors) +
    make_theme(theme_name=theme_few(), leg_pos="none",
               guide_nrow = 10,
              #  guide_nrow = length(unique(df_plot %>% filter(Genus == genus) %>% pull(magOTU))),
               setFill = F, setCol = F,
               x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
               y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
  p_list[[genus]] <- p
}

pdf("results/figures/09-relative_abundance_sep_by_genus.pdf", width = 15, height = 10)
marrangeGrob(p_list, nrow = 2, ncol = 1)
dev.off()


ggplot(df_plot %>% select(magOTU, prevalence_in_host, sample, Host, Genus, rel_cov) %>%
          group_by(magOTU, Host) %>%
          mutate(rel_cov_mean = mean(rel_cov)) %>%
          mutate(rel_cov_sd = sd(rel_cov)) %>%
          ungroup() %>%
          select(magOTU, prevalence_in_host, rel_cov_mean, rel_cov_sd, Genus, Host) %>% unique
) +
  geom_point(aes(x = prevalence_in_host,
               y = rel_cov_mean), size = 5, color = "black") +
  geom_point(aes(x = prevalence_in_host,
               y = rel_cov_mean,
               color = Genus), size = 4) +
  facet_wrap(~Host, scales = "free", ncol = 2) +
  geom_errorbar(aes(x = prevalence_in_host,
                    ymin = rel_cov_mean - rel_cov_sd,
                    ymax = rel_cov_mean + rel_cov_sd,
                    color = Genus)) +
  geom_smooth(aes(x = prevalence_in_host,
                  y = rel_cov_mean), method = "glm") +
  stat_cor(aes(x = prevalence_in_host,
                  y = rel_cov_mean), method = "pearson") +
  labs(x = "Prevalence in host", y = "Relative abundance (mean in Host)") +
  scale_color_manual(values=genusColors) +
  make_theme(theme_name=theme_minimal(), leg_pos="none",
             # guide_nrow = 6,
             setFill = F, setCol = F,
             x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
             y_angle=0 ,y_vj=0, y_hj=0, y_size=10)
ggsave("results/figures/09-prevalence_vs_rel_abundance.pdf")
```



```{r other_plots}
ggplot(instrain_genomes_info_df %>%
        group_by(genome) %>%
          mutate(breadth_pass = ifelse(breadth >= 0.7, "Y", "N")) %>%
          mutate(perc_var = SNV_count/length*breadth*100) %>%
          filter(breadth_pass %in% c("Y")),
                             aes(x = magOTU,
                                 y = perc_var,
                                 color = factor(Host, host_order),
                                 shape = breadth_pass
                              )
                            ) +
      geom_jitter() +
      labs(x = "magOTU", y = "% variant sites", color = "Host species") +
        facet_wrap(~factor(Genus, shared_genera), scales = "free_x") +
          make_theme(theme_name=theme_few(), leg_pos="bottom",
                     guide_nrow = 3,
                     setFill = F, setCol = F,
                     x_angle=45 ,x_vj=1, x_hj=1, x_size=5,
                     y_angle=0 ,y_vj=0, y_hj=0, y_size=10) +
            scale_color_manual(values=host_order_color)
            ggsave("results/figures/15b-percentage_variable_sites.pdf")

```