---
title: "Honeybee cross-species analysis - 11"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Gene counts and information

Gene counts and information are collected by the `scripts/visualization/make_gene_info_summaries.py` and written to `results/figures/08-summarize_functions/genes_detected/{sample}_df_detected_genes_info.csv`. It lists all the genes considered detected (> 5 reads and coverage fraction > 0.9) Then, `scripts/visualization/summarize_gene_functions.py` collects the counts calculated from the bowtie mapping done of the cleaned reads to the assembly and extractes the appropriate gene information from the `results/figures/08-summarize_functions/genes_detected/{sample}_df_detected_genes_info.csv` files. The gene information is then written to `results/figures/08-summarize_functions/gene_info_tables/A1-1_df_detected_genes_info.csv` and the count information including start and end to `results/figures/08-summarize_functions/gene_info_tables/A1-1_df_detected_genes_cov.csv`. These two files can then be combined in R to collect all the necessary information about the genes for plotting.

The `scripts/visualization/summarize_gene_functions.py` further makes the files in `results/figures/08-summarize_functions/minpath_KO_collections/by_*` which contain per sample files either collected overall, per genus or per species with the list of KOs which can then be used as the input for minpath. It also run minpath and writes the output to `results/figures/08-summarize_functions/minpath_outputs/by_*`. The minpath output files are then parsed and combined with the information about gene / KO counts, `results/figures/08-summarize_functions/minpath_outputs/by_sample/{sample}_kos.path.rpkm` files are written with the rpkm corresponding to each pathway. The script also makes a directory of cazyme counts and KO counts per sample count aggregated per mag within the sample. This can then be aggregated at various levels within R.

```{r}
# df_ko_counts_sample <- data.frame()
# df_ko_counts_species <- data.frame()
# df_ko_counts_genus <- data.frame()
# for (sample in samples_IN_MY) {
#   df_ko <- read.csv(paste0("results/figures/08-summarize_functions/ko_counts/", sample, "_ko_counts.csv"))
#   if (nrow(df_ko) == 0) {
#     next
#   }
#   df_ko$sample <- sample
#   df_ko_sample <- df_ko %>%
#     group_by(ko, sample) %>%
#     summarize(count = sum(rpkm), .groups = "keep") %>%
#     ungroup()
#   df_ko_species <- df_ko %>%
#     group_by(ko, species, sample) %>%
#     summarize(count = sum(rpkm), .groups = "keep") %>%
#     ungroup()
#   df_ko_genus <- df_ko %>%
#     group_by(ko, genus, sample) %>%
#     summarize(count = sum(rpkm), .groups = "keep") %>%
#     ungroup()
#   df_ko_counts_sample <- rbind(df_ko_counts_sample, df_ko_sample)
#   df_ko_counts_species <- rbind(df_ko_counts_species, df_ko_species)
#   df_ko_counts_genus <- rbind(df_ko_counts_genus, df_ko_genus)
# }
# # save the data
# write.csv(df_ko_counts_sample, "results/figures/08-summarize_functions/ko_counts_sample.csv", row.names = FALSE)
# write.csv(df_ko_counts_species, "results/figures/08-summarize_functions/ko_counts_species.csv", row.names = FALSE)
# write.csv(df_ko_counts_genus, "results/figures/08-summarize_functions/ko_counts_genus.csv", row.names = FALSE)
df_ko_counts_sample <- read.csv("results/figures/08-summarize_functions/ko_counts_sample.csv")
ko_matrix <- df_ko_counts_sample %>% 
  pivot_wider(names_from = ko, values_from = count, values_fill = list(count = 0)) %>%
  column_to_rownames(var = "sample")
```

# Plots - KO

## KO per sample boxplot

Number of KOs per sample for each host

```{r}
kos_per_sample <- df_ko_counts_sample %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample) %>%
                    summarize(n_kos = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(kos_per_sample) +
  geom_boxplot(aes(y = n_kos, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_kos, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of KOs", fill = "Host") +
  scale_y_continuous(labels = label_number(suffix = " K", scale = 1e-3)) +
  # add stats
  geom_signif(data = kos_per_sample,
              aes(x = host, y = n_kos),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_kos_per_sample.pdf")

kos_per_sample_genus <- df_ko_counts_genus %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample, genus) %>%
                    summarize(n_kos = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(kos_per_sample_genus) +
  geom_boxplot(aes(y = n_kos, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_kos, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  facet_wrap(~genus, scales = "free") +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of KOs", fill = "Host") +
  scale_y_continuous(labels = label_number(suffix = " K", scale = 1e-3)) +
  # add stats
  geom_signif(data = kos_per_sample_genus %>%
                      group_by(genus) %>%
                      mutate(n_samples = n_distinct(samples)) %>%
                      filter(n_samples > 1) %>%
                      ungroup(),
              aes(x = host, y = n_kos),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_kos_per_sample_by_genus.pdf", width = 15, height = 20)

kos_per_sample_species <- df_ko_counts_species %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample, species) %>%
                    summarize(n_kos = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(kos_per_sample_species) +
  geom_boxplot(aes(y = n_kos, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_kos, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  facet_wrap(~species, scales = "free") +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of KOs", fill = "Host") +
  # scale_y_continuous(labels = label_number(suffix = " K", scale = 1e-3)) +
  # add stats
  geom_signif(data = kos_per_sample_species %>%
                      group_by(species) %>%
                      mutate(n_samples = n_distinct(samples)) %>%
                      filter(n_samples > 1) %>%
                      ungroup(),
              aes(x = host, y = n_kos),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_kos_per_sample_by_species.pdf", width = 30, height = 20)

```

## KO cumulative curve

```{r}
df_curves <- read.csv("results/figures/08-summarize_functions/cumulative_curves/ko_curves.tsv", sep = "\t")
ggplot(df_curves) +
  geom_smooth(aes(x = size, y = kos, color = host), method = "loess", se = T) +
  geom_point(aes(x = size, y = kos, color = host)) +
  scale_color_manual(values = host_order_color_dark) +
  labs(x = "# Bees", y = "# KOs", color = "Host") +
  make_theme(setFill = F, setCol = F, guide_nrow = 1)
  ggsave("results/figures/08-gene_content_plots/KO_cumulative_curve.pdf")

df_curves_genus_combined <- data.frame()
for (genus in genera) {
# for (genus in unique(df_ko_counts_genus$genus)) {
  df_curves_genus <- read.csv(paste0("results/figures/08-summarize_functions/cumulative_curves/ko_curves_", genus, ".tsv"), sep = "\t")
  df_curves_genus$genus <- genus
  df_curves_genus_combined <- rbind(df_curves_genus_combined, df_curves_genus)
}
ggplot(df_curves_genus_combined) +
  geom_smooth(aes(x = size, y = kos, color = host), method = "loess", se = T) +
  # geom_point(aes(x = size, y = kos, color = host)) +
  facet_wrap(~genus, scales = "free") +
  scale_color_manual(values = host_order_color_dark) +
  labs(x = "# Bees", y = "# KOs", color = "Host") +
  make_theme(setFill = F, setCol = F, guide_nrow = 1)
  ggsave("results/figures/08-gene_content_plots/KO_cumulative_curve_by_genus.pdf", width = 15, height = 20)
```

## KO PCoA

```{r}
# get presence absence matrix
ko_matrix_pa <- (ko_matrix > 0) * 1
# get distance matrix
dist_mat <- beta.pair(ko_matrix_pa, "jaccard")
ado_res <- adonis2(dist_mat$beta.jac ~ Species, 
          rownames(dist_mat$beta.jac %>% as.matrix) %>% as.data.frame() %>%
            mutate(Species = Vectorize(get_host_from_sample_name)(.)))
pcoa_plot(dist_mat$beta.jac, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark) +
  annotate("text", x = -0.15,
                      y = 0.2,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]])
          )
ggsave("results/figures/08-gene_content_plots/KO_pcoa_host_species_pa_jac_full.pdf")

# get distance matrix
dist_mat <- beta.pair(ko_matrix_pa, "jaccard")
ado_res <- adonis2(dist_mat$beta.jtu ~ Species, 
          rownames(dist_mat$beta.jtu %>% as.matrix) %>% as.data.frame() %>%
            mutate(Species = Vectorize(get_host_from_sample_name)(.)))
pcoa_plot(dist_mat$beta.jtu, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark) +
  annotate("text", x = -0.15,
                      y = 0.2,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]])
          )
ggsave("results/figures/08-gene_content_plots/KO_pcoa_host_species_pa_jac_turnover.pdf")

ko_matrix_norm <- decostand(ko_matrix, method = "normalize")
dist_mat <- vegdist(ko_matrix, method = "robust.aitchison")
ado_res <- adonis2(dist_mat ~ Species, 
          rownames(as.matrix(dist_mat)) %>% as.data.frame() %>%
            mutate(Species = Vectorize(get_host_from_sample_name)(.)))
pcoa_plot(dist_mat, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark) +
  annotate("text", x = -0.15,
                      y = 0.2,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]])
          )
ggsave("results/figures/08-gene_content_plots/KO_pcoa_host_species_norm_aitchison.pdf")
```


# Collect CAZyme counts

```{r}
df_cazyme_counts_sample <- data.frame()
df_cazyme_counts_species <- data.frame()
df_cazyme_counts_genus <- data.frame()
for (sample in samples_IN_MY) {
  df_cazyme <- read.csv(paste0("results/figures/08-summarize_functions/cazyme_counts/", sample, "_cazyme_counts.csv")) %>%
    # remove cazymes starting with AA and GT as they are not infomrmative
    filter(!grepl("^AA", cazyme), !grepl("^GT", cazyme))
  if (nrow(df_cazyme) == 0) {
    next
  }
  df_cazyme$sample <- sample
  df_cazyme_sample <- df_cazyme %>%
    group_by(cazyme, sample) %>%
    summarize(count = sum(rpkm), .groups = "keep") %>%
    ungroup()
  df_cazyme_species <- df_cazyme %>%
    group_by(cazyme, species, sample) %>%
    summarize(count = sum(rpkm), .groups = "keep") %>%
    ungroup()
  df_cazyme_genus <- df_cazyme %>%
    group_by(cazyme, genus, sample) %>%
    summarize(count = sum(rpkm), .groups = "keep") %>%
    ungroup()
  df_cazyme_counts_sample <- rbind(df_cazyme_counts_sample, df_cazyme_sample)
  df_cazyme_counts_species <- rbind(df_cazyme_counts_species, df_cazyme_species)
  df_cazyme_counts_genus <- rbind(df_cazyme_counts_genus, df_cazyme_genus)
}
# save the data
write.csv(df_cazyme_counts_sample, "results/figures/08-summarize_functions/cazyme_counts_sample.csv", row.names = FALSE)
write.csv(df_cazyme_counts_species, "results/figures/08-summarize_functions/cazyme_counts_species.csv", row.names = FALSE)
write.csv(df_cazyme_counts_genus, "results/figures/08-summarize_functions/cazyme_counts_genus.csv", row.names = FALSE)
df_cazyme_counts_sample <- read.csv("results/figures/08-summarize_functions/cazyme_counts_sample.csv")
cazyme_matrix <- df_cazyme_counts_sample %>% 
  pivot_wider(names_from = cazyme, values_from = count, values_fill = list(count = 0)) %>%
  column_to_rownames(var = "sample")
```

# Collect cazyme information

```{r}
completed_substrate_annotations <- read.csv("data/cayman_gene_db/20230607_glycan_annotations_cleaned_manually.tsv", header = T, sep = "\t")
glycan_annotations_final_cleaned <- completed_substrate_annotations %>% select(c(Family:Subfamily,ORIGIN:FUNCTION_AT_DESTINATION_3, Glycan_annotation))
glycan_annotations_final_cleaned_long <- glycan_annotations_final_cleaned %>% separate_rows(ORIGIN,sep=",") %>% separate_rows(FUNCTION_IN_ORIGIN,sep=",") %>% separate_rows(FUNCTION_AT_DESTINATION_1,sep=",") %>% separate_rows(FUNCTION_AT_DESTINATION_2,sep=",") %>% separate_rows(FUNCTION_AT_DESTINATION_3,sep=",")
glycan_annotations_cleaned <- glycan_annotations_final_cleaned

glycan_annotations_cleaned_filt <- glycan_annotations_cleaned %>% filter(!is.na(FUNCTION_AT_DESTINATION_3))
```

# Plots - CAZyme

## CAZyme per sample boxplot

Number of CAZymes per sample for each host

```{r}
cazyme_per_sample <- df_cazyme_counts_sample %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample) %>%
                    summarize(n_cazymes = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(cazyme_per_sample) +
  geom_boxplot(aes(y = n_cazymes, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_cazymes, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of CAZymes", fill = "Host") +
  # add stats
  geom_signif(data = cazyme_per_sample,
              aes(x = host, y = n_cazymes),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_cazymes_per_sample.pdf")

cazyme_per_sample <- df_cazyme_counts_sample %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample) %>%
                    left_join(glycan_annotations_cleaned_filt %>%
                                select(Subfamily, FUNCTION_AT_DESTINATION_3)
                        , by = c("cazyme" = "Subfamily")) %>%
                    filter(grepl("ectin|ellulose|emicellulose", FUNCTION_AT_DESTINATION_3)) %>%
                    filter(FUNCTION_AT_DESTINATION_3 != "Amylose_and_Amylopectin") %>%
                    summarize(n_cazymes = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(cazyme_per_sample) +
  geom_boxplot(aes(y = n_cazymes, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_cazymes, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of CAZymes", fill = "Host") +
  # add stats
  geom_signif(data = cazyme_per_sample,
              aes(x = host, y = n_cazymes),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_cazymes_per_sample_subset.pdf")

cazyme_per_sample_genus <- df_cazyme_counts_genus %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample, genus) %>%
                    summarize(n_cazymes = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(cazyme_per_sample_genus) +
  geom_boxplot(aes(y = n_cazymes, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_cazymes, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  facet_wrap(~genus, scales = "free") +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of CAZymes", fill = "Host") +
  # add stats
  geom_signif(data = cazyme_per_sample_genus %>%
                      group_by(genus) %>%
                      mutate(n_host = n_distinct(host)) %>%
                      filter(n_host > 1) %>%
                      ungroup(),
              aes(x = host, y = n_cazymes),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_cazymes_per_sample_by_genus.pdf", width = 15, height = 20)

cazyme_per_sample_genus <- df_cazyme_counts_genus %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample, genus) %>%
                    left_join(glycan_annotations_cleaned_filt %>%
                                select(Subfamily, FUNCTION_AT_DESTINATION_3)
                        , by = c("cazyme" = "Subfamily")) %>%
                    filter(grepl("ectin|ellulose|emicellulose", FUNCTION_AT_DESTINATION_3)) %>%
                    filter(FUNCTION_AT_DESTINATION_3 != "Amylose_and_Amylopectin") %>%
                    summarize(n_cazymes = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(cazyme_per_sample_genus) +
  geom_boxplot(aes(y = n_cazymes, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_cazymes, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  facet_wrap(~genus, scales = "free_x") +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of CAZymes", fill = "Host") +
  # add stats
  geom_signif(data = cazyme_per_sample_genus %>%
                      group_by(genus) %>%
                      mutate(n_host = n_distinct(host)) %>%
                      filter(n_host > 1) %>%
                      ungroup(),
              aes(x = host, y = n_cazymes),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_cazymes_per_sample_by_genus_subset.pdf", width = 15, height = 20)

cazyme_per_sample_species <- df_cazyme_counts_species %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample, species) %>%
                    summarize(n_cazymes = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(cazyme_per_sample_species) +
  geom_boxplot(aes(y = n_cazymes, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_cazymes, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  facet_wrap(~species, scales = "free") +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of CAZymes", fill = "Host") +
  # add stats
  geom_signif(data = cazyme_per_sample_species %>%
                      group_by(species) %>%
                      mutate(n_hosts = n_distinct(host)) %>%
                      filter(n_hosts > 1) %>%
                      ungroup(),
              aes(x = host, y = n_cazymes),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_cazymes_per_sample_by_species.pdf", width = 30, height = 20)

cazyme_per_sample_species <- df_cazyme_counts_species %>%
                    mutate(present = ifelse(count > 0, 1, 0)) %>%
                    group_by(sample, species) %>%
                    left_join(glycan_annotations_cleaned_filt %>%
                                select(Subfamily, FUNCTION_AT_DESTINATION_3)
                        , by = c("cazyme" = "Subfamily")) %>%
                    filter(grepl("ectin|ellulose|emicellulose", FUNCTION_AT_DESTINATION_3)) %>%
                    filter(FUNCTION_AT_DESTINATION_3 != "Amylose_and_Amylopectin") %>%
                    summarize(n_cazymes = sum(present)) %>%
                    ungroup() %>%
                    mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(cazyme_per_sample_species) +
  geom_boxplot(aes(y = n_cazymes, x = factor(host, host_order), 
                fill = factor(host, host_order),
                ),
                outlier.shape = NA
              ) +
  geom_point(aes(y = n_cazymes, x = factor(host, host_order), group = host,
                 fill = factor(host)),
              size = 2,
              position = position_jitterdodge(0.75)) +
  facet_wrap(~species, scales = "free_x") +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "Number of CAZymes", fill = "Host") +
  # add stats
  geom_signif(data = cazyme_per_sample_species %>%
                      group_by(species) %>%
                      mutate(n_hosts = n_distinct(host)) %>%
                      filter(n_hosts > 1) %>%
                      ungroup(),
              aes(x = host, y = n_cazymes),
              test = "wilcox.test",
              step_increase = 0.05,
              tip_length = 0,
              comparisons = list(c("Apis florea", "Apis andreniformis"), 
                                #  c("Apis mellifera", "Apis florea"),
                                #  c("Apis mellifera", "Apis andreniformis"),
                                #  c("Apis cerana", "Apis dorsata"),
                                #  c("Apis cerana", "Apis florea"),
                                #  c("Apis dorsata", "Apis florea"),
                                 c("Apis mellifera", "Apis cerana"),
                                 c("Apis dorsata", "Apis andreniformis"),
                                 c("Apis dorsata", "Apis cerana"),
                                 c("Apis mellifera", "Apis dorsata")),
              map_signif_level = TRUE, textsize = 6) +
  make_theme(setFill = FALSE, leg_pos = "bottom",
              x_angle = 90, x_size = 0,
              x_hj = 10, x_vj = 1,
              y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/Number_of_cazymes_per_sample_by_species_subset.pdf", width = 30, height = 20)
```

## CAZyme PCoA

```{r}
# get presence absence matrix
cazyme_matrix_pa <- (cazyme_matrix > 0) * 1
# get distance matrix
dist_mat <- beta.pair(cazyme_matrix_pa, "jaccard")
ado_res <- adonis2(dist_mat$beta.jac ~ Species, 
          rownames(dist_mat$beta.jac %>% as.matrix) %>% as.data.frame() %>%
            mutate(Species = Vectorize(get_host_from_sample_name)(.)))
pcoa_plot(dist_mat$beta.jac, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark) +
  annotate("text", x = -0.15,
                      y = 0.2,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]])
          )
ggsave("results/figures/08-gene_content_plots/cazyme_pcoa_host_species_pa_jac_full.pdf")

# get distance matrix
dist_mat <- beta.pair(cazyme_matrix_pa, "jaccard")
ado_res <- adonis2(dist_mat$beta.jtu ~ Species, 
          rownames(dist_mat$beta.jtu %>% as.matrix) %>% as.data.frame() %>%
            mutate(Species = Vectorize(get_host_from_sample_name)(.)))
pcoa_plot(dist_mat$beta.jtu, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark) +
  annotate("text", x = -0.15,
                      y = 0.2,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]])
          )
ggsave("results/figures/08-gene_content_plots/cazyme_pcoa_host_species_pa_jac_turnover.pdf")

cazyme_matrix_norm <- decostand(cazyme_matrix, method = "normalize")
dist_mat <- vegdist(cazyme_matrix_norm, method = "robust.aitchison")
ado_res <- adonis2(dist_mat ~ Species, 
          rownames(as.matrix(dist_mat)) %>% as.data.frame() %>%
            mutate(Species = Vectorize(get_host_from_sample_name)(.)))
pcoa_plot(dist_mat, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark) +
  annotate("text", x = -0.15,
                      y = 0.2,
                      size = 8,
          label = paste0("omega2 = ", round(adonis_OmegaSq(ado_res)$parOmegaSq[[1]], 3),
                         ", \np = ", adonis_OmegaSq(ado_res)$`Pr(>F`[[1]])
          )
ggsave("results/figures/08-gene_content_plots/cazyme_pcoa_host_species_norm_aitchinson.pdf")
```

# GH and PLs for figure

```{r}
# limit to genera with high prevalence in host
df_cazyme_counts_plot <- df_cazyme_counts_genus %>%
                          select(genus, sample) %>%
                          unique() %>%
                          mutate(present = 1) %>%
                          mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
                          group_by(host) %>%
                          mutate(n_in_host = n_distinct(sample)) %>%
                          ungroup() %>%
                          group_by(genus, host) %>%
                          mutate(n_samples = sum(present)) %>%
                          ungroup() %>%
                          filter(n_samples/n_in_host > 0.05) %>%
                          select(-n_samples, -n_in_host) %>%
                          filter(genus %in% genera_hp) %>%
                          left_join(df_cazyme_counts_genus, by = c("sample", "genus")) %>%
                          left_join(glycan_annotations_cleaned_filt %>%
                                      select(Subfamily, FUNCTION_AT_DESTINATION_3)
                                  , by = c("cazyme" = "Subfamily")) %>%
                          filter(grepl("ectin|ellulose|emicellulose", FUNCTION_AT_DESTINATION_3)) %>%
                          filter(FUNCTION_AT_DESTINATION_3 != "Amylose_and_Amylopectin") %>%
                          mutate(host = Vectorize(get_host_from_sample_name)(sample))
df_cazyme_counts_dot_plot <- df_cazyme_counts_plot %>%
                              mutate(present = ifelse(count > 0, 1, 0)) %>%
                              group_by(cazyme, sample) %>%
                              summarize(count_cazyme = sum(present)) %>%
                              ungroup() %>%
                              mutate(host = Vectorize(get_host_from_sample_name)(sample))
df_cazyme_counts_bar_plot <- df_cazyme_counts_plot %>%
                              group_by(genus, sample) %>% 
                              summarize(rpkm_cazy_in_genus = sum(count)) %>%
                              ungroup() %>%
                              mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
                              group_by(genus, host) %>%
                              summarize(mean_rpkm_cazy_in_genus = mean(rpkm_cazy_in_genus),
                                        sd_rpkm_cazy_in_genus = sd(rpkm_cazy_in_genus),
                                        n = n()) %>%
                              ungroup() %>%
                              # make relative proportion
                              group_by(host) %>%
                              mutate(mean_prop_rpkm_cazy_in_genus = mean_rpkm_cazy_in_genus/sum(mean_rpkm_cazy_in_genus)) %>%
                              ungroup()
df_cazyme_tiles_plot <- df_cazyme_counts_plot %>%
                          select(cazyme, FUNCTION_AT_DESTINATION_3) %>%
                          mutate(Pectin = ifelse(grepl("Pectin", FUNCTION_AT_DESTINATION_3), "Y", "N")) %>%
                          mutate(Hemicellulose = ifelse(grepl("Hemicellulose", FUNCTION_AT_DESTINATION_3), "Y", "N")) %>%
                          mutate(Cellulose = ifelse(grepl("Cellulose", FUNCTION_AT_DESTINATION_3), "Y", "N")) %>%
                          mutate(Others = ifelse(FUNCTION_AT_DESTINATION_3 %in% c("Pectin", "Hemicellulose", "Cellulose"), "N", "Y")) %>%
                          select(!FUNCTION_AT_DESTINATION_3) %>%
                          pivot_longer(!c(cazyme), names_to = "Substrate", values_to = "Present")
cazyme_order <- df_cazyme_counts_plot %>%
                          select(cazyme, FUNCTION_AT_DESTINATION_3) %>%
                          mutate(Pectin = ifelse(grepl("Pectin", FUNCTION_AT_DESTINATION_3), "Y", "N")) %>%
                          mutate(Hemicellulose = ifelse(grepl("Hemicellulose", FUNCTION_AT_DESTINATION_3), "Y", "N")) %>%
                          mutate(Cellulose = ifelse(grepl("Cellulose", FUNCTION_AT_DESTINATION_3), "Y", "N")) %>%
                          mutate(Others = ifelse(FUNCTION_AT_DESTINATION_3 %in% c("Pectin", "Hemicellulose", "Cellulose"), "N", "Y")) %>%
                          select(!FUNCTION_AT_DESTINATION_3) %>%
                          unique() %>%
                          arrange(desc(Pectin), desc(Hemicellulose), desc(Cellulose), desc(Others)) %>%
                          pull(cazyme)
                          
genusColors_custom <- c("g__Bombilactobacillus" = "#9E0142",
                        "g__Lactobacillus" = "#BE5581",
                        "g__CALYQJ01" = "#fa9fb5",
                        "g__Dysgonomonas" = "#D53E4F",
                        "g__Bifidobacterium" = "#F46D43",
                        "g__Apibacter" = "#E6AB02",
                        "g__Commensalibacter" = "#FFFFBF",
                        "g__Frischella" = "#ABDDA4",
                        "g__Snodgrassella" = "#3288BD",
                        "g__Gilliamella" = "#4a1486",
                        "g__CALYQQ01" = "#5E4FA2",
                        "g__Pectinatus" = "#1B9E77",
                        "g__Saezia" = "#FDAE61",
                        "g__Entomomonas" = "#B28B5B"
)
p_tiles <- ggplot(df_cazyme_tiles_plot) +
            geom_tile(aes(x = factor(cazyme, cazyme_order), 
                          y = factor(Substrate, rev(c("Pectin", "Hemicellulose", "Cellulose", "Others"))), fill = Present),
                      color = "white") +
            scale_fill_manual(values = c("N" = "#ffffff", "Y" = "#045a8d")) +
            labs(x = "CAZyme", y = "Substrate", fill = "Present") +
            make_theme(setFill = FALSE, leg_pos = "bottom",
                        x_angle = 90, x_vj = 0.5, x_hj = 1,
                        y_size = 10, y_hj = 1)
ggsave("results/figures/08-gene_content_plots/cazyme_figure/CAZyme_tiles.pdf", width = 15, height = 10)
p_bars <- ggplot(df_cazyme_counts_bar_plot) +
            geom_bar(aes(x = mean_prop_rpkm_cazy_in_genus, y = host, fill = genus),
                      stat = "identity") +
            scale_fill_manual(values = genusColors_custom) +
            labs(x = "Count", y = "Number of CAZymes", fill = "Genus") +
            make_theme(setFill = FALSE, leg_pos = "bottom",
                        x_angle = 90, x_vj = 0.5, x_hj = 1,
                        y_size = 10, y_hj = 1)
ggsave("results/figures/08-gene_content_plots/cazyme_figure/CAZyme_bars.pdf", width = 15, height = 20)
p_dots <- ggplot(df_cazyme_counts_dot_plot) +
            geom_point(aes(x = factor(cazyme, cazyme_order), y = host, size = count_cazyme),
                      color = "#045a8d") +
            # scale_color_manual(values = host_order_color_dark) +
            scale_size_continuous(range = c(1, 5)) +
            labs(x = "Count", y = "Number of CAZymes") +
            make_theme(setFill = FALSE, leg_pos = "bottom",
                        x_angle = 90, x_vj = 0.5, x_hj = 1,
                        y_size = 10, y_hj = 1)
ggsave("results/figures/08-gene_content_plots/cazyme_figure/CAZyme_dots.pdf", width = 15, height = 20)
# plot the three together
pdf("results/figures/08-gene_content_plots/cazyme_figure/CAZyme_figure.pdf", width = 15, height = 20)
grid.arrange(p_dots, p_bars, p_tiles,
             widths = c(0.75, 0.25, 0.75),
             layout_matrix = rbind(c(1, 1, 2),
                                   c(3, 3, NA)
             )
)
dev.off()
```

# Parse minpath

```{r}
pathways_include_df <- read.csv("scripts/visualization/pathways_to_include.tsv", sep = "\t")
pathways_include <- pathways_include_df %>% filter(include == "y") %>% pull(pathway_name)
pathways_include_clean <- pathways_include_df %>%
                          filter(include == "y") %>%
                          mutate(pathway_name = make.names(pathway_name)) %>%
                          pull(pathway_name)
kegg_path_info <- read.csv("scripts/visualization/kegg_pathway_categories.tsv", sep = '\t') %>%
                    distinct(path_name, .keep_all = TRUE)
kegg_path_info_clean <- kegg_path_info
kegg_path_info_clean$path_name <- make.names(kegg_path_info_clean$path_name)

df_counts_all <- data.frame()
for (sample in samples) {
  if (!file.exists(paste0("results/figures/08-summarize_functions/minpath_outputs/by_sample/", sample, "_kos.path.rpkm"))) {
    next
  }
  df_counts_iter <- read.csv(paste0("results/figures/08-summarize_functions/minpath_outputs/by_sample/", sample, "_kos.path.rpkm"), sep = "\t") %>%
                      mutate(Sample = sample)
  df_counts_all <- rbind(df_counts_all, df_counts_iter)
}

length(df_reads$Sample)
min_reads = min(df_reads$MAGs_DB_mf)
# for normalizing coverage across samples

min_copies <- min(qpcr_plot_df$copy_num)
max_copies <- max(qpcr_plot_df$copy_num)
df_copy_norm <- qpcr_plot_df %>% 
                  group_by(Host) %>% 
                    summarise(Median_cp = median(copy_num)) %>% 
                    # summarise(Median_cp = median(copy_num)/min_copies) %>% 
                      unique() %>%
                      mutate(norm_factor = Median_cp/min(Median_cp))
                      # mutate(Median_cp = Median_cp/min(Median_cp))
sample_norm_factor_df <- samples_IN_MY %>% as.data.frame() %>% 
                          mutate(Host = Vectorize(get_host_from_sample_name)(samples_IN_MY)) %>%
                          left_join(df_copy_norm, by = c("Host" = "Host")) %>%
                          mutate(sample = samples_IN_MY) %>%
                          select(Sample = sample, norm_factor)
# # get the median copy number for that host species
# qpcr_normalizing_factor <- function(x) {
#   num = df_copy_norm[df_copy_norm$Host== x, "Median_cp"][[1]]
#   return(num)
# }
# # get the copy number of that sample if available and if not, get the median copy number for that species 
# not doing this as the copy number is not available for India samples.
# qpcr_normalizing_factor_indiv <- function(sample) {
#   num = 1
#   if (sample %in% qpcr_plot_df$Sample.Name) {
#     num = qpcr_plot_df[qpcr_plot_df$Sample.Name == sample, "copy_num"] %>% pull %>% unlist
#   } else {
#     h = get_host_from_sample_name(sample)
#     num = df_copy_norm[df_copy_norm$Host== h, "Median_cp"][[1]] %>% unlist
#     print(paste0("gettin median for ", sample, " as it is not found in the qpcr data"))
#   }
#   return(num)
# }

df_counts <- df_counts_all %>%
              filter(pathway_name %in% pathways_include) %>%
              left_join(sample_norm_factor_df, by = c("Sample" = "Sample")) %>%
              mutate(rpkm = as.numeric(rpkm)*norm_factor)
# df_counts <- df_counts_all %>%
#               filter(pathway_name %in% pathways_include)
df_counts_mat <- df_counts %>%
                  mutate(rpkm = as.numeric(rpkm)) %>%
                  pivot_wider(id_cols = c(Sample), names_from = pathway_name, values_from = rpkm) %>%
                  # replace NAs with 0
                  mutate(across(where(is.numeric), ~replace_na(., 0)))
```

## plot rpkm of pathways

```{r}
pathway_order <- df_counts %>%
                  left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  # filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        # filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = section), outlier.shape = NA) +
  geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order)), size = 0.05) +
  facet_grid(~factor(Host, host_order), scales = "free") +
  labs(y = "Pathway name", x = "RPKM", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 6,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_all.pdf", width = 20, height = 25)

pathway_order <- df_counts %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = section), outlier.shape = NA) +
  geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order)), size = 0.05) +
  facet_grid(~factor(Host, host_order), scales = "free") +
  labs(y = "Pathway name (>10 samples)", x = "RPKM", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 6,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_gtr10.pdf", width = 20, height = 25)

pathway_order <- df_counts %>%
                  left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
                  filter(section != "6. Human Diseases") %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        filter(section != "6. Human Diseases") %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = section), outlier.shape = NA) +
  geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order)), size = 0.05) +
  facet_grid(~factor(Host, host_order), scales = "free") +
  labs(y = "Pathway name (> 10 samples excluding section 6)", x = "RPKM", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 6,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_notNum6.pdf", width = 20, height = 25)

pathway_order <- df_counts %>%
                  left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
                  filter(section == "1. Metabolism") %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        filter(section == "1. Metabolism") %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = subsection), outlier.shape = NA) +
  geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order)), size = 0.05) +
  facet_grid(~factor(Host, host_order), scales = "free") +
  labs(y = "Pathway name (> 10 samples, only Metabolism)", x = "RPKM", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 6,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_Num1.pdf", width = 20, height = 25)

pathway_order <- df_counts %>%
                  left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
                  filter(section == "1. Metabolism") %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        filter(section == "1. Metabolism") %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = Host), outlier.shape = NA) +
  # geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order), color = Host), size = 0.05) +
  # facet_wrap(subsection~., scales = "free") +
  labs(y = "Pathway name (> 10 samples, only Metabolism)", x = "RPKM", fill = "Host") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 2,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_Num1_compared.pdf", width = 20, height = 25)

pathway_order <- df_counts %>%
                  left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
                  filter(section == "1. Metabolism") %>%
                  filter(pathway_name != "Zeatin biosynthesis") %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        filter(section == "1. Metabolism") %>%
        filter(pathway_name != "Zeatin biosynthesis") %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = subsection), outlier.shape = NA) +
  geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order)), size = 0.05) +
  facet_grid(~factor(Host, host_order), scales = "free") +
  labs(y = "Pathway name (> 10 samples, only Metabolism)", x = "RPKM", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 6,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_Num1_sub.pdf", width = 20, height = 25)

pathway_order <- df_counts %>%
                  left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
                  filter(section == "1. Metabolism") %>%
                  filter(pathway_name != "Zeatin biosynthesis") %>%
                  group_by(pathway_name) %>%
                  mutate(count = n_distinct(Sample)) %>%
                  filter(count > 10) %>%
                  ungroup() %>%
                  mutate(max_count = max(rpkm)) %>%
                  arrange(max_count, count) %>%
                  pull(pathway_name) %>% unique
ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        filter(section == "1. Metabolism") %>%
        filter(pathway_name != "Zeatin biosynthesis") %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_boxplot(aes(x = rpkm, y = factor(pathway_name, pathway_order), fill = Host), outlier.shape = NA) +
  # geom_point(aes(x = rpkm, y = factor(pathway_name, pathway_order), color = Host), size = 0.05) +
  # facet_wrap(subsection~., scales = "free") +
  labs(y = "Pathway name (> 10 samples, only Metabolism)", x = "RPKM", fill = "Host") +
  scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "bottom", 
             x_angle = 45, x_size = 10,
             palettefill = "Spectral",
             guide_nrow = 2,
             y_size = 10, y_hj = 1, y_vj = 0.5)
  ggsave("results/figures/08-gene_content_plots/pathway_rpkm_plots/pathway_rpkm_boxplot_Num1_compared_sub.pdf", width = 20, height = 25)
```

## run maaslin2

```{r}
input_data <- df_counts_mat %>% column_to_rownames("Sample") %>% as.matrix
meta_table <- data.frame(Host = Vectorize(get_host_from_sample_name)(rownames(input_data)))
meta_table$Host <- factor(meta_table$Host, levels = c("Apis mellifera", "Apis cerana",
                                                      "Apis dorsata", "Apis florea", "Apis andreniformis"))
input_metadata <- meta_table
saveRDS(input_data, "results/figures/08-gene_content_plots/maaslin2_input_data.rds")
saveRDS(input_metadata, "results/figures/08-gene_content_plots/maaslin2_input_metadata.rds")
# input_data <- readRDS("results/figures/08-gene_content_plots/maaslin2_input_data.rds")
# input_metadata <- readRDS("results/figures/08-gene_content_plots/maaslin2_input_metadata.rds")
# fit_data = Maaslin2(
#     input_data = input_data, 
#     input_metadata = input_metadata, 
#     output = "results/figures/08-summarize_functions/maaslin2_results_minpath", 
#     analysis_method = "LM", # consider
#     transform = "LOG", # consider
#     normalization = "TSS", # consider
#     fixed_effects = c("Host"),
#     min_abundance = 0,
#     save_models = T,
#     plot_scatter = F,
#     plot_heatmap = F)
df_pairwise <- data.frame()
models_data <- readRDS("results/figures/08-summarize_functions/maaslin2_results_minpath/fits/models.rds")
for (i in 1:length(models_data)) {
  model <- models_data[[i]]
  conts <- summary(lsmeans(model, pairwise ~ Host, adjust = "FDR"))
  df_to_add <- data.frame(
                    feature = names(models_data)[[i]],
                    contrast = conts$contrasts$contrast,
                    p.value = conts$contrasts$p.value,
                    estimate = conts$contrasts$estimate
                  )
  df_pairwise <- rbind(df_pairwise, df_to_add)
}
write.csv(df_pairwise, "results/figures/08-summarize_functions/maaslin2_results_minpath_parsed/maaslin2_results_minpath.csv", row.names = F, quote = F)
```

## maaslin2 Plots

```{r}
df_minpath_out <- read.csv("results/figures/08-summarize_functions/maaslin2_results_minpath/all_results.tsv", sep = "\t")
df_minpath_out_amRef <- df_minpath_out %>%
                      filter(qval < 0.05) %>%
                      filter(feature %in% pathways_include_clean)
df_minpath_out_amRef_names <- df_minpath_out_amRef %>%
                      group_by(feature) %>%
                      mutate(count = n_distinct(value)) %>%
                      ungroup() %>%
                      arrange(count) %>%
                      pull(feature) %>% unique
# make a heatmap 
ggplot(df_minpath_out_amRef) +
  geom_tile(aes(x = value, y = factor(feature, df_minpath_out_amRef_names), fill = coef)) +
  labs(y = "Pathway", x = "Host", fill = "Estimate") +
  scale_fill_gradientn(
    colours = c("red", "white", "blue"),
  ) + 
  make_theme(setFill = FALSE, leg_pos = "right", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/maaslin2_results_minpath_heatmap.pdf", width = 10, height = 15)

ggplot(df_minpath_out_amRef %>% left_join(kegg_path_info_clean %>% select(path_name, section), by = c("feature" = "path_name"))) +
  geom_tile(aes(x = value, y = factor(feature, df_minpath_out_amRef_names), fill = coef)) +
  labs(y = "Pathway", x = "Host", fill = "Estimate") +
  scale_fill_gradientn(
    colours = c("red", "white", "blue"),
  ) + 
  facet_grid(section~., scales = "free", space = "free_y") +
  make_theme(setFill = FALSE, leg_pos = "right", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 10, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/maaslin2_results_minpath_heatmap_sections.pdf", width = 10, height = 15)

ggplot(df_counts %>%
        # remove those in 5 samples or less
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Sample)) %>%
        filter(count > 10) %>%
        ungroup() %>%
        arrange(count) %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name")) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))) +
  geom_tile(aes(x = Sample, y = pathway_name, fill = rpkm)) +
  facet_grid(section~Host, scales = "free", space = "free_y") +
  labs(y = "Pathway (>5 samples)", x = "Sample", fill = "RPKM") +
  # log scale for color fill
  scale_fill_continuous(trans = "log10") +
  make_theme(setFill = FALSE, leg_pos = "right", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 6, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/pathway_rpkm_heatmap.pdf", width = 20, height = 25)
ggplot(df_counts %>%
        # remove those in 5 samples or less
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Host)) %>%
        mutate(sample_count = n_distinct(Sample)) %>%
        filter(count > 2) %>%
        filter(sample_count > 10) %>%
        ungroup() %>%
        arrange(count) %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name"))) +
  geom_tile(aes(x = Sample, y = pathway_name, fill = rpkm)) +
  facet_grid(section~Host, scales = "free", space = "free_y") +
  labs(y = "Pathway (>5 samples)", x = "Sample", fill = "RPKM") +
  # log scale for color fill
  scale_fill_continuous(trans = "log10") +
  make_theme(setFill = FALSE, leg_pos = "right", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 6, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/pathway_rpkm_heatmap_shared.pdf", width = 20, height = 25)
ggplot(df_counts %>%
        # remove those in 5 samples or less
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
        group_by(pathway_name) %>%
        mutate(count = n_distinct(Host)) %>%
        mutate(sample_count = n_distinct(Sample)) %>%
        filter(count <= 2) %>%
        # filter(sample_count > 10) %>%
        ungroup() %>%
        arrange(count) %>%
        left_join(kegg_path_info, by = c("pathway_name" = "path_name"))) +
  geom_tile(aes(x = Sample, y = pathway_name, fill = rpkm)) +
  facet_grid(section~Host, scales = "free", space = "free_y") +
  labs(y = "Pathway (>5 samples)", x = "Sample", fill = "RPKM") +
  # log scale for color fill
  scale_fill_continuous(trans = "log10") +
  make_theme(setFill = FALSE, leg_pos = "right", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 6, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/pathway_rpkm_heatmap_unshared.pdf", width = 20, height = 25)

all_pairs <- read.csv("results/figures/08-summarize_functions/maaslin2_results_minpath_parsed/maaslin2_results_minpath.csv", header = T)$contrast %>% unique
for (pair in all_pairs) {
  # get the pathways that are differentially abundant
  df_pairwise <- read.csv("results/figures/08-summarize_functions/maaslin2_results_minpath_parsed/maaslin2_results_minpath.csv", header = T) %>%
                  filter(feature %in% pathways_include_clean) %>%
                  filter(contrast == pair) %>%
                  # filter(p.value < 0.05) %>%
                  mutate(pair1 = strsplit(pair, " - ") %>% sapply(function(x) x[1]),
                         pair2 = strsplit(pair, " - ") %>% sapply(function(x) x[2])) %>%
                  mutate(host = ifelse(estimate > 0, pair1, pair2)) %>%
                  mutate(host = ifelse(p.value > 0.05, NA, host))
  # volcano plot
  ggplot(df_pairwise) +
    geom_point(aes(x = estimate, y = -log10(p.value), color = host), size = 3) +
    labs(x = "Estimate", y = "-log10(p.value)", color = "Host") +
    ggtitle(pair) +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    scale_color_manual(values = host_order_color_dark) +
    geom_text_repel(aes(x = estimate, y = -log10(p.value), label = feature), box.padding = 0.5) +
    make_theme(setFill = F, setCol = F, leg_pos = "right", 
               x_angle = 45, x_size = 10, modify_guide = F,
               y_size = 6, y_hj = 1)
    ggsave(paste0("results/figures/08-gene_content_plots/masslin2_volcano/maaslin2_results_minpath_volcano_", pair, ".pdf"))
}
for (pair in all_pairs) {
  # get the pathways that are differentially abundant
  df_pairwise <- read.csv("results/figures/08-summarize_functions/maaslin2_results_minpath_parsed/maaslin2_results_minpath.csv", header = T) %>%
                  filter(feature %in% pathways_include_clean) %>%
                  filter(contrast == pair) %>%
                  # filter(p.value < 0.05) %>%
                  mutate(pair1 = strsplit(pair, " - ") %>% sapply(function(x) x[1]),
                         pair2 = strsplit(pair, " - ") %>% sapply(function(x) x[2])) %>%
                  mutate(host = ifelse(estimate > 0, pair1, pair2)) %>%
                  mutate(host = ifelse(p.value > 0.05, NA, host)) %>%
                  left_join(kegg_path_info_clean %>% select(path_name, section), by = c("feature" = "path_name")) %>%
                  filter(section != "6. Human Diseases")
  # volcano plot
  ggplot(df_pairwise) +
    geom_point(aes(x = estimate, y = -log10(p.value), color = host), size = 3) +
    labs(x = "Estimate", y = "-log10(p.value)", color = "Host") +
    ggtitle(pair) +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    scale_color_manual(values = host_order_color_dark) +
    geom_text_repel(aes(x = estimate, y = -log10(p.value), label = feature), box.padding = 0.5) +
    make_theme(setFill = F, setCol = F, leg_pos = "right", 
               x_angle = 45, x_size = 10, modify_guide = F,
               y_size = 6, y_hj = 1)
    ggsave(paste0("results/figures/08-gene_content_plots/masslin2_volcano/maaslin2_results_minpath_volcano_noHumanDis_", pair, ".pdf"))
}
for (pair in all_pairs) {
  # get the pathways that are differentially abundant
  df_pairwise <- read.csv("results/figures/08-summarize_functions/maaslin2_results_minpath_parsed/maaslin2_results_minpath.csv", header = T) %>%
                  filter(feature %in% pathways_include_clean) %>%
                  filter(contrast == pair) %>%
                  # filter(p.value < 0.05) %>%
                  mutate(pair1 = strsplit(pair, " - ") %>% sapply(function(x) x[1]),
                         pair2 = strsplit(pair, " - ") %>% sapply(function(x) x[2])) %>%
                  mutate(host = ifelse(estimate > 0, pair1, pair2)) %>%
                  mutate(host = ifelse(p.value > 0.05, NA, host)) %>%
                  left_join(kegg_path_info_clean %>% select(path_name, section), by = c("feature" = "path_name")) %>%
                  filter(section == "1. Metabolism")
  # volcano plot
  ggplot(df_pairwise) +
    geom_point(aes(x = estimate, y = -log10(p.value), color = host), size = 3) +
    labs(x = "Estimate", y = "-log10(p.value)", color = "Host") +
    ggtitle(pair) +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    scale_color_manual(values = host_order_color_dark) +
    geom_text_repel(aes(x = estimate, y = -log10(p.value), label = feature), box.padding = 0.5) +
    make_theme(setFill = F, setCol = F, leg_pos = "right", 
               x_angle = 45, x_size = 10, modify_guide = F,
               y_size = 6, y_hj = 1)
    ggsave(paste0("results/figures/08-gene_content_plots/masslin2_volcano/maaslin2_results_minpath_volcano_Metab_", pair, ".pdf"))
}

# plot the boxplot for the top signif results from maaslin2
df_counts_sig <- df_counts %>%
                  mutate(pathway_name = make.names(pathway_name)) %>%
                  filter(pathway_name %in% (df_minpath_out %>% filter(qval < 0.05) %>% pull(feature) %>% unique)) %>%
                  left_join(kegg_path_info_clean %>% select(path_name, section), by = c("pathway_name" = "path_name")) %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample))
ggplot(df_counts_sig %>% filter(section == "1. Metabolism")) +
  geom_boxplot(aes(x = rpkm, y = pathway_name, fill = Host), outlier.shape = NA) +
  # geom_point(aes(x = rpkm, y = pathway_name, color = section), size = 0.5) +
  # facet_grid(~Host) +
  labs(y = "RPKM", x = "Host", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "bottom", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 6, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/pathway_rpkm_boxplot_sig_metab.pdf", width = 20, height = 25)
df_counts_sig <- df_counts %>%
                  mutate(pathway_name = make.names(pathway_name)) %>%
                  filter(pathway_name %in% (df_minpath_out %>% filter(qval < 0.05) %>% pull(feature) %>% unique)) %>%
                  left_join(kegg_path_info_clean %>% select(path_name, section), by = c("pathway_name" = "path_name")) %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample))
ggplot(df_counts_sig %>% filter(section != "1. Metabolism")) +
  geom_boxplot(aes(x = rpkm, y = pathway_name, fill = Host), outlier.shape = NA) +
  # geom_point(aes(x = rpkm, y = pathway_name, color = section), size = 0.5) +
  # facet_grid(~Host) +
  labs(y = "RPKM", x = "Host", fill = "Host") +
  # scale_color_manual(values = host_order_color_dark) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = F, setCol = F, leg_pos = "bottom", 
             x_angle = 45, x_size = 10, modify_guide = F,
             y_size = 6, y_hj = 1)
  ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/pathway_rpkm_boxplot_sig_nonMetab.pdf", width = 20, height = 25)
```

## run kruskal + random forest

```{r}
pvalue.threshold <- 0.05
# excluded pathways of non-interest here:
abund_table <- df_counts_mat %>% column_to_rownames("Sample") %>% as.data.frame()
# filter and normalize (log-relative) data
abund_table <- abund_table[rowSums(abund_table) > 0,]
abund_table <- abund_table / rowSums(abund_table)
abund_table <- log(abund_table + 1)
meta_table <- data.frame(Groups = Vectorize(get_host_from_sample_name)(rownames(abund_table)))
meta_table$Groups <- as.factor(meta_table$Groups)

kruskal.wallis.table <- data.frame()
data <- as.data.frame(abund_table)
for (i in 1:dim(data)[2]){
    ks.test <- kruskal.test(data[,i], g=meta_table$Groups)
    # Store the result in the data frame
    kruskal.wallis.table <- rbind(kruskal.wallis.table,data.frame(id=names(data)[i],p.value=ks.test$p.value))
}
kruskal.wallis.table$E.value <- kruskal.wallis.table$p.value * dim(kruskal.wallis.table)[1]
kruskal.wallis.table$FWER <- pbinom(q=0, p=kruskal.wallis.table$p.value, size=dim(kruskal.wallis.table)[1], lower.tail=FALSE)
kruskal.wallis.table <- kruskal.wallis.table[order(kruskal.wallis.table$p.value, decreasing=FALSE), ]
kruskal.wallis.table$q.value.factor <- dim(kruskal.wallis.table)[1] / 1:dim(kruskal.wallis.table)[1]
kruskal.wallis.table$q.value <- kruskal.wallis.table$p.value * kruskal.wallis.table$q.value.factor
rownames(kruskal.wallis.table) <- kruskal.wallis.table$id

#==================significant feature selection =================================#
last.significant.element <- max(which(kruskal.wallis.table$q.value <= pvalue.threshold))
selected <- 1:last.significant.element
sig_res <-kruskal.wallis.table$id[selected]

#==random forest classifier ==#
subset.data<-data.frame(data[,as.character(kruskal.wallis.table[rownames(kruskal.wallis.table),"id"])], check.names=FALSE)
kruskal.wallis.table$id <- colnames(subset.data) #enforce that ids and colnames of subset data remain the same for easy indexing later on
names(subset.data) <- make.names(names(subset.data))
saveRDS(subset.data, "results/figures/08-summarize_functions/selected_features.rds")
saveRDS(meta_table, "results/figures/08-summarize_functions/meta_table.rds")
# Random Forest
# subset.data <- readRDS("results/figures/08-summarize_functions/selected_features.rds")
# meta_table <- readRDS("results/figures/08-summarize_functions/meta_table.rds")
# rf_res <- randomforest_res(subset.data, meta_table$Groups)
# saveRDS(rf_res, "results/figures/08-summarize_functions/randomforest_res.rds")
rf_res <- readRDS("results/figures/08-summarize_functions/randomforest_res.rds")
df_accuracy <- rf_res$importance

df <- NULL
for(i in df_accuracy$Sample){
    rank <- (subset(df_accuracy, df_accuracy$Sample==i))$rank
    tmp<-data.frame(subset.data[,i],meta_table$Groups, rep(rank), rep(paste(i," p.adj = ",sprintf("%.10g",kruskal.wallis.table[kruskal.wallis.table$id==i,"q.value"]),sep=""),dim(data)[1]))
    colnames(tmp)<-c("Value","Groups","Rank","Taxa")
    if(is.null(df)){df<-tmp} else { df<-rbind(df,tmp)}
    df <- na.omit(df)
}

out <- list("SignfeaturesTable"=kruskal.wallis.table, "plotdata"=df, "importance"=df_accuracy)
write.csv(out$SignfeaturesTable, "results/figures/08-summarize_functions/minpath_compared_kruskal_wallis_table.csv")
write.csv(out$importance, "results/figures/08-summarize_functions/minpath_compared_kruskal_wallis_importance.csv")

max_rank <- max(out$importance$rank)
write.csv(out$importance %>% left_join(kegg_path_info_clean,
                    by = c("Sample" = "path_name")), "results/figures/08-summarize_functions/minpath_compared_kruskal_wallis_importance_added.csv")
out_importance <- read.csv("results/figures/08-summarize_functions/minpath_compared_kruskal_wallis_importance.csv") %>%
                    filter(Sample %in% pathways_include_clean)
path_order <- out_importance %>%
        arrange(Value) %>% pull(Sample) %>% as.character %>% unique
```

## random forest plots

```{r}
ggplot(out_importance %>%
        left_join(kegg_path_info_clean,
                    by = c("Sample" = "path_name")) %>%
        arrange(-rank) %>% filter(rank < max_rank)) +
  geom_bar(aes(x = Value, y = factor(Sample, path_order), fill = section), stat = "identity") +
  # facet_wrap(~section, scales = "free") +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", palettefill = "Spectral", 
             x_angle = 45, x_size = 10, guide_nrow = 6,
             y_size = 6, y_hj = 1)
    ggsave("results/figures/08-summarize_functions/minpath_maaslin_rf_plots/minpath_compared_kruskal_wallis.pdf", width = 10, height = 15)
ggplot(out_importance %>%
        left_join(kegg_path_info_clean,
                    by = c("Sample" = "path_name")) %>%
        filter(section != "6. Human Diseases") %>%
        arrange(-rank) %>% filter(rank < max_rank)) +
  geom_bar(aes(x = Value, y = factor(Sample, path_order), fill = section), stat = "identity") +
  # facet_wrap(~section, scales = "free") +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", palettefill = "Spectral", 
             x_angle = 45, x_size = 10, guide_nrow = 6,
             y_size = 6, y_hj = 1)
    ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/minpath_compared_kruskal_wallis_sub.pdf", width = 10, height = 15)
ggplot(out_importance %>%
        left_join(kegg_path_info_clean,
                    by = c("Sample" = "path_name")) %>%
        filter(section == "1. Metabolism") %>%
        arrange(-rank) %>% filter(rank < max_rank)) +
  geom_bar(aes(x = Value, y = factor(Sample, path_order), fill = subsection), stat = "identity") +
  # facet_wrap(~section, scales = "free") +
  # scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = T, setCol = T, leg_pos = "bottom", palettefill = "Spectral", 
             x_angle = 45, x_size = 10, guide_nrow = 6,
             y_size = 6, y_hj = 1)
    ggsave("results/figures/08-gene_content_plots/minpath_maaslin_rf_plots/minpath_compared_kruskal_wallis_sub_metabolism.pdf", width = 10, height = 15)
```

# Plot gene maps

```{r}
# read minpath info
minpath_map <- read.csv("scripts/MinPath/data/KEGG-mapping.txt", sep = "\t", header = F)
minpath_pathnames <- read.csv("scripts/MinPath/data/KEGG-pathway.txt", sep = "\t", header = F)
# minpath_ko_info <- KEGG-family.txt
colnames(minpath_map) <- c("pathway_id", "ko_id")
colnames(minpath_pathnames) <- c("pathway_id", "pathway_name")
pathway_info <- left_join(minpath_map, minpath_pathnames, by = "pathway_id") %>%
                  mutate(pathway_name = ifelse(pathway_name %in% pathways_include, pathway_name, "Other"))
# get_gene_name_from_desc <- function(description){
#   # description = "Hemolysin coregulated protein Hcp (TssD)"
#   description = strsplit(description, "\\(")[[1]][[2]]
#   gene_name = strsplit(description, "\\)")[[1]][[1]]
#   return(gene_name)
# }
clean_dram_gene_name <- function(gene_name) {
  #  gene_name = "D3-4_D3-4_NODE_10001_length_6406_cov_22.032908_2"
   gene_name = strsplit(gene_name, "_") %>% unlist
   gene_name = gene_name[2:(length(gene_name) - 0)]
   gene_name = paste(gene_name, collapse = "_")
   return(gene_name)
}
get_molecule_from_gene <- function(gene_id) {
  # remove only last _ and the number after it
  # gene_id = "D3-4_NODE_34_length_225669_cov_18.609235_187"
  molecule <- strsplit(gene_id, "_") %>% unlist
  molecule <- molecule[1:(length(molecule) - 1)]
  molecule <- paste(molecule, collapse = "_")
  return(molecule)
}
cazymes_of_interest <- glycan_annotations_final_cleaned %>%
                        filter(grepl("ectin|ellulose|emicellulose", FUNCTION_AT_DESTINATION_3)) %>%
                        filter(FUNCTION_AT_DESTINATION_3 != "Amylose_and_Amylopectin") %>%
                        mutate(cazyme = Family) %>%
                        select(cazyme, FUNCTION_AT_DESTINATION_3) 
# search for pectin genes!
cazymes_of_interest <- glycan_annotations_final_cleaned %>%
                        filter(grepl("Pectin", FUNCTION_AT_DESTINATION_3)) %>%
                        mutate(cazyme = Family) %>%
                        select(cazyme, FUNCTION_AT_DESTINATION_3) 
# sample = "D3-2"
for (sample in samples_AD) {
# for (sample in c("D3-1", "D3-2", "D3-3", "D3-4", "D4-5", "D2-3", "D8-1", "D8-2", "D8-3")) {
  df_covs_pos <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_cov.csv"))
  df_info <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_info.csv"))
  df_annotations <- read.csv(paste0("results/08_gene_content/02_DRAM_annotations/", sample, "/annotations.tsv"), skip = 1, header = F, sep = "\t")
  colnames(df_annotations) <- c(
      "gene_dram_name",
      "sample",
      "rank",
      "ko_id",
      "kegg_hit",
      "peptidase_id",
      "peptidase_family",
      "peptidase_hit",
      "peptidase_RBH",
      "peptidase_identity",
      "peptidase_bitScore",
      "peptidase_eVal",
      "pfam_hits",
      "cazy_ids",
      "cazy_hits",
      "cazy_subfam_ec",
      "cazy_best_hit",
      "heme_regulatory_motif_count"
  )
  df_annotations_parsed <- df_annotations %>%
                    left_join(pathway_info, by = "ko_id") %>%
                    filter(ko_id != "") %>% 
                    mutate(gene = Vectorize(clean_dram_gene_name)(gene_dram_name)) %>%
                    mutate(gene_desc_kegg = kegg_hit) %>%
                    mutate(gene_desc_pfam = pfam_hits) %>%
                    select(gene, pathway_name, pathway_id, ko_id, gene_desc_pfam, gene_desc_kegg)
  df_all_gene_info <- left_join(df_info, df_covs_pos, by = "gene") %>%
                        mutate(molecule = Vectorize(get_molecule_from_gene)(gene)) %>%
                        left_join(df_annotations_parsed, by = "gene") %>%
                          left_join(glycan_annotations_final_cleaned, by = c("cazyme" = "Subfamily"))
  genes_chosen_ranges <- df_all_gene_info %>%
                    # filter(mag == "D3-4_25") %>%
                    filter(genus == "g__Dysgonomonas") %>%
                    filter(cazyme != "") %>%
                    filter(cazyme %in% cazymes_of_interest$cazyme) %>%
                    group_by(molecule) %>%
                    mutate(molecule_start = min(start),
                          molecule_end = max(end)) %>%
                    mutate(molecule_range = molecule_end - molecule_start) %>%
                    # filter(molecule_range > 10000) %>% # remove small molecules
                    filter(molecule_range < 1000000) %>% # remove those with genes too far away
                    mutate(num_genes = n()) %>%
                    filter(num_genes > 5) %>%
                    ungroup() %>%
                    group_by(molecule) %>%
                    mutate(molecule_name = molecule) %>%
                    summarise(molecule_start = min(molecule_start),
                              molecule_end = max(molecule_end))
  # just select the molecules that are listed in genes_chosen_ranges
  # and check that only genes in that range are selected
  genes_chosen <- df_all_gene_info %>%
                    left_join(genes_chosen_ranges, by = "molecule") %>%
                    filter(!is.na(molecule_start) & !is.na(molecule_end)) %>%
                    mutate(include = ifelse(start >= molecule_start & end <= molecule_end, T, F)) %>%
                    filter(include) %>%
                    mutate(middle = (start + end) / 2) %>%
                    group_by(molecule) %>%
                    mutate(molecule_order = min(start)) %>%
                    arrange(molecule_order) %>%
                    ungroup() %>%
                    group_by(gene) %>%
                    mutate(pathway_name = paste(pathway_name, collapse = " | ")) %>%
                    mutate(pathway_id = paste(pathway_id, collapse = " | ")) %>%
                    select(!X) %>%
                    ungroup() %>%
                    unique() %>%
                    mutate(cazy_substrate = ifelse(grepl("Pectin", FUNCTION_AT_DESTINATION_3), "Pectin",
                                                  ifelse(grepl("Hemicellulose", FUNCTION_AT_DESTINATION_3), "Hemicellulose",
                                                  ifelse(grepl("Cellulose", FUNCTION_AT_DESTINATION_3), "Cellulose", "Other"))))
  if (nrow(genes_chosen) == 0) {
    print("No genes found in")
    print(sample)
    next
  }
  ggplot(genes_chosen, aes(xmin = start,
                            xmax = end,
                            x = middle,
                            y = molecule,
                            label = cazyme,
                            fill = cazy_substrate)) +
    geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
    geom_gene_label(align = "left") +
    geom_text(aes(label = gene_desc_kegg), angle = 50, hjust = 0, size = 2) +
    # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 2) +
    scale_fill_manual(values = c("Pectin" = "#b3cde3",
                                "Hemicellulose" = "#fed9a6",
                                  "Cellulose" = "#fbb4ae",
                                  "Other" = "#d2d2d2")) +
    # scale_fill_manual(values = c("Pectin" = "#0570b0",
    #                              "Hemicellulose" = "#fdae61",
    #                               "Cellulose" = "#d73027",
    #                               "Other" = "#999999")) +
    facet_wrap(~ molecule, scales = "free", ncol = 1) +
    labs(x = "Position", y = "Molecule", fill = "Cazyme") +
    make_theme(theme_name = theme_genes(),
              setFill = F, setCol = F,
              guide_nrow = 6, y_size = 5,
              leg_pos = "bottom")
    ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/Dysgonomonas_cazyme_genes_", sample, ".pdf"), width = 20, height = 25)
  ggplot(genes_chosen, aes(xmin = start,
                            xmax = end,
                            x = middle,
                            y = molecule,
                            label = cazyme,
                            fill = FUNCTION_AT_DESTINATION_3)) +
    geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
    geom_gene_label(align = "left") +
    geom_text(aes(label = pathway_name), angle = 50, hjust = 0, size = 2) +
    # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 1) +
    scale_fill_manual(values = c("Pectin" = "#b3cde3",
                                "Hemicellulose" = "#fed9a6",
                                  "Cellulose" = "#fbb4ae",
                                  "Other" = "#d2d2d2")) +
    # scale_fill_manual(values = c("Pectin" = "#0570b0",
    #                              "Hemicellulose" = "#fdae61",
    #                               "Cellulose" = "#d73027",
    #                               "Other" = "#999999")) +
    facet_wrap(~ molecule, scales = "free", ncol = 1) +
    labs(x = "Position", y = "Molecule", fill = "Cazyme") +
    make_theme(theme_name = theme_genes(),
              setFill = F, setCol = F,
              guide_nrow = 6, y_size = 5,
              leg_pos = "bottom")
    ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/Dysgonomonas_cazyme_genes_pathway_", sample, ".pdf"), width = 20, height = 25)
    print("completed")
    print(sample)
    write.csv(genes_chosen, paste0("results/figures/08-gene_content_plots/gene_maps/Dysgonomonas_cazyme_genes_", sample, ".csv"), row.names = F, quote = F)
    saveRDS(genes_chosen, paste0("results/figures/08-gene_content_plots/gene_maps/Dysgonomonas_cazyme_genes_", sample, ".rds"))
}
# for ease of work, print those contigs and respective samples that have the following genes of interest
for (sample in samples_AD) {
  if (!(file.exists(paste0("results/figures/08-gene_content_plots/gene_maps/Dysgonomonas_cazyme_genes_", sample, ".csv")))) {
    next
  }
  genes_chosen <- readRDS(paste0("results/figures/08-gene_content_plots/gene_maps/Dysgonomonas_cazyme_genes_", sample, ".rds"))
  if (nrow(genes_chosen) == 0) {
    next
  }
  genes_chosen_grouped <- genes_chosen %>%
                            group_by(molecule) %>%
                            summarise(cazymes = paste(cazyme, collapse = " | "),
                                      susCD = ifelse(any(grepl("SusC|SusD", gene_desc_kegg)), "SusCD", "No SusCD"))
  print(sample)
  print(genes_chosen_grouped)
}
```

```{r}
# read minpath info
minpath_map <- read.csv("scripts/MinPath/data/KEGG-mapping.txt", sep = "\t", header = F)
minpath_pathnames <- read.csv("scripts/MinPath/data/KEGG-pathway.txt", sep = "\t", header = F)
# minpath_ko_info <- KEGG-family.txt
colnames(minpath_map) <- c("pathway_id", "ko_id")
colnames(minpath_pathnames) <- c("pathway_id", "pathway_name")
pathway_info <- left_join(minpath_map, minpath_pathnames, by = "pathway_id") %>%
                  mutate(pathway_name = ifelse(pathway_name %in% pathways_include, pathway_name, "Other"))
# get_gene_name_from_desc <- function(description){
#   # description = "Hemolysin coregulated protein Hcp (TssD)"
#   description = strsplit(description, "\\(")[[1]][[2]]
#   gene_name = strsplit(description, "\\)")[[1]][[1]]
#   return(gene_name)
# }
clean_dram_gene_name <- function(gene_name) {
  #  gene_name = "D3-4_D3-4_NODE_10001_length_6406_cov_22.032908_2"
   gene_name = strsplit(gene_name, "_") %>% unlist
   gene_name = gene_name[2:(length(gene_name) - 0)]
   gene_name = paste(gene_name, collapse = "_")
   return(gene_name)
}
get_molecule_from_gene <- function(gene_id) {
  # remove only last _ and the number after it
  # gene_id = "D3-4_NODE_34_length_225669_cov_18.609235_187"
  molecule <- strsplit(gene_id, "_") %>% unlist
  molecule <- molecule[1:(length(molecule) - 1)]
  molecule <- paste(molecule, collapse = "_")
  return(molecule)
}
for (sample in samples_IN_MY) {
# for (sample in c("D3-1", "D3-2", "D3-3", "D3-4", "D4-5", "D2-3", "D8-1", "D8-2", "D8-3")) {
  if (file.exists(paste0("results/figures/08-gene_content_plots/gene_maps/T6SS_genes/T6SS_genes_", sample, ".pdf"))) {
    next
  }
  df_covs_pos <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_cov.csv"))
  df_info <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_info.csv"))
  df_annotations <- read.csv(paste0("results/08_gene_content/02_DRAM_annotations/", sample, "/annotations.tsv"), skip = 1, header = F, sep = "\t")
  colnames(df_annotations) <- c(
      "gene_dram_name",
      "sample",
      "rank",
      "ko_id",
      "kegg_hit",
      "peptidase_id",
      "peptidase_family",
      "peptidase_hit",
      "peptidase_RBH",
      "peptidase_identity",
      "peptidase_bitScore",
      "peptidase_eVal",
      "pfam_hits",
      "cazy_ids",
      "cazy_hits",
      "cazy_subfam_ec",
      "cazy_best_hit",
      "heme_regulatory_motif_count"
  )
  df_annotations_parsed <- df_annotations %>%
                    left_join(pathway_info, by = "ko_id") %>%
                    filter(ko_id != "") %>% 
                    mutate(gene = Vectorize(clean_dram_gene_name)(gene_dram_name)) %>%
                    mutate(gene_desc_kegg = kegg_hit) %>%
                    mutate(gene_desc_pfam = pfam_hits) %>%
                    select(gene, pathway_name, pathway_id, ko_id, gene_desc_pfam, gene_desc_kegg)
  df_all_gene_info <- left_join(df_info, df_covs_pos, by = "gene") %>%
                        mutate(molecule = Vectorize(get_molecule_from_gene)(gene)) %>%
                        left_join(df_annotations_parsed, by = "gene") #%>%
                          # left_join(glycan_annotations_final_cleaned, by = c("cazyme" = "Subfamily"))
  genes_chosen_ranges <- df_all_gene_info %>%
                    filter(grepl("ecretion ", gene_desc_kegg)) %>%
                    # remove the phrase "type VI secretion system protein" and 
                    # "type VI secretion system secreted protein" from name description
                    mutate(gene_desc_kegg = gsub("type VI secretion system protein", "", gene_desc_kegg)) %>%
                    mutate(gene_desc_kegg = gsub("type VI secretion system secreted protein", "secreted ", gene_desc_kegg)) %>%
                    group_by(molecule) %>%
                    mutate(molecule_start = min(start),
                          molecule_end = max(end)) %>%
                    mutate(molecule_range = molecule_end - molecule_start) %>%
                    # filter(molecule_range > 10000) %>% # remove small molecules
                    filter(molecule_range < 1000000) %>% # remove those with genes too far away
                    mutate(num_genes = n()) %>%
                    filter(num_genes > 5) %>%
                    ungroup() %>%
                    group_by(molecule) %>%
                    mutate(molecule_name = molecule) %>%
                    summarise(molecule_start = min(molecule_start),
                              molecule_end = max(molecule_end)) %>%
                    mutate(length = molecule_end - molecule_start) %>%
                    filter(length > 10000) %>%
                    filter(length < 100000)
  # just select the molecules that are listed in genes_chosen_ranges
  # and check that only genes in that range are selected
  genes_chosen <- df_all_gene_info %>%
                    filter(genus %in% names(genusColors)) %>%
                    left_join(genes_chosen_ranges, by = "molecule") %>%
                    filter(!is.na(molecule_start) & !is.na(molecule_end)) %>%
                    mutate(include = ifelse(start >= molecule_start & end <= molecule_end, T, F)) %>%
                    filter(include) %>%
                    mutate(middle = (start + end) / 2) %>%
                    group_by(molecule) %>%
                    mutate(molecule_order = min(start)) %>%
                    arrange(molecule_order) # %>%
                    # ungroup() %>%
                    # group_by(gene) %>%
                    # mutate(pathway_name = paste(pathway_name, collapse = " | ")) %>%
                    # mutate(pathway_id = paste(pathway_id, collapse = " | ")) %>%
                    # select(!X) %>%
                    # ungroup() %>%
                    # unique() %>%
                    # mutate(cazy_substrate = ifelse(grepl("Pectin", FUNCTION_AT_DESTINATION_3), "Pectin",
                    #                               ifelse(grepl("Hemicellulose", FUNCTION_AT_DESTINATION_3), "Hemicellulose",
                    #                               ifelse(grepl("Cellulose", FUNCTION_AT_DESTINATION_3), "Cellulose", "Other"))))
  if (nrow(genes_chosen) == 0) {
    print("No genes found in")
    print(sample)
    next
  }
  ggplot(genes_chosen, aes(xmin = start,
                            xmax = end,
                            x = middle,
                            y = molecule,
                            label = ko,
                            fill = genus)) +
    geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
    geom_gene_label(align = "left") +
    geom_text(aes(label = gene_desc_kegg), angle = 50, hjust = 0, size = 2) +
    # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 2) +
    scale_fill_manual(values = genusColors) +
    facet_wrap(~ molecule, scales = "free", ncol = 1) +
    labs(x = "Position", y = "Molecule", fill = "Cazyme") +
    make_theme(theme_name = theme_genes(),
              setFill = F, setCol = F,
              guide_nrow = 6, y_size = 5,
              leg_pos = "bottom")
    ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/T6SS_genes/T6SS_genes_", sample, ".pdf"), width = 20, height = 25)
  ggplot(genes_chosen, aes(xmin = start,
                            xmax = end,
                            x = middle,
                            y = molecule,
                            label = ko,
                            fill = genus)) +
    geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
    geom_gene_label(align = "left") +
    geom_text(aes(label = gene_desc_pfam), angle = 50, hjust = 0, size = 2) +
    # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 1) +
    scale_fill_manual(values = genusColors) +
    facet_wrap(~ molecule, scales = "free", ncol = 1) +
    labs(x = "Position", y = "Molecule", fill = "Cazyme") +
    make_theme(theme_name = theme_genes(),
              setFill = F, setCol = F,
              guide_nrow = 6, y_size = 5,
              leg_pos = "bottom")
    ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/T6SS_genes/T6SS_genes_genes_pfamName_", sample, ".pdf"), width = 20, height = 25)
    print("completed")
    print(sample)
    write.csv(genes_chosen, paste0("results/figures/08-gene_content_plots/gene_maps/T6SS_genes/T6SS_genes__genes_", sample, ".csv"), row.names = F, quote = F)
    saveRDS(genes_chosen, paste0("results/figures/08-gene_content_plots/gene_maps/T6SS_genes/T6SS_genes__genes_", sample, ".rds"))
}
```
```{r}
for (sample in samples_IN_MY) {
  # if (file.exists(paste0("results/figures/08-gene_content_plots/gene_maps/cazy_genes_all/cazy_genes_all_", sample, ".pdf"))) {
  #   next
  # }
  df_covs_pos <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_cov.csv"))
  df_info <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_info.csv"))
  df_annotations <- read.csv(paste0("results/08_gene_content/02_DRAM_annotations/", sample, "/annotations.tsv"), skip = 1, header = F, sep = "\t")
  cazymes_of_interest <- glycan_annotations_final_cleaned %>%
                        filter(grepl("Pectin|Cellulose|Hemicellulose", FUNCTION_AT_DESTINATION_3)) %>%
                        filter(FUNCTION_AT_DESTINATION_3 != "Amylose_and_Amylopectin") %>%
                        mutate(cazyme = Family) %>%
                        select(cazyme, FUNCTION_AT_DESTINATION_3)
  # # search for pectin genes!
  # cazymes_of_interest <- glycan_annotations_final_cleaned %>%
  #                         filter(grepl("Pectin", FUNCTION_AT_DESTINATION_3)) %>%
  #                         mutate(cazyme = Family) %>%
  #                         select(cazyme, FUNCTION_AT_DESTINATION_3) 
  colnames(df_annotations) <- c(
      "gene_dram_name",
      "sample",
      "rank",
      "ko_id",
      "kegg_hit",
      "peptidase_id",
      "peptidase_family",
      "peptidase_hit",
      "peptidase_RBH",
      "peptidase_identity",
      "peptidase_bitScore",
      "peptidase_eVal",
      "pfam_hits",
      "cazy_ids",
      "cazy_hits",
      "cazy_subfam_ec",
      "cazy_best_hit",
      "heme_regulatory_motif_count"
  )
  df_annotations_parsed <- df_annotations %>%
                    left_join(pathway_info, by = "ko_id") %>%
                    filter(ko_id != "") %>% 
                    mutate(gene = Vectorize(clean_dram_gene_name)(gene_dram_name)) %>%
                    mutate(gene_desc_kegg = kegg_hit) %>%
                    mutate(gene_desc_pfam = pfam_hits) %>%
                    select(gene, pathway_name, pathway_id, ko_id, gene_desc_pfam, gene_desc_kegg)
  df_all_gene_info <- left_join(df_info, df_covs_pos, by = "gene") %>%
                        mutate(molecule = Vectorize(get_molecule_from_gene)(gene)) %>%
                        left_join(df_annotations_parsed, by = "gene") %>%
                          left_join(glycan_annotations_final_cleaned, by = c("cazyme" = "Subfamily"))
  genes_chosen_ranges <- df_all_gene_info %>%
                    filter(cazyme != "") %>%
                    filter(cazyme %in% cazymes_of_interest$cazyme) %>%
                    group_by(molecule) %>%
                    mutate(molecule_start = min(start),
                          molecule_end = max(end)) %>%
                    mutate(molecule_range = molecule_end - molecule_start) %>%
                    # filter(molecule_range > 10000) %>% # remove small molecules
                    filter(molecule_range < 100000) %>% # remove those with genes too far away
                    mutate(num_genes = n()) %>%
                    filter(num_genes > 5) %>%
                    ungroup() %>%
                    group_by(molecule) %>%
                    mutate(molecule_name = molecule) %>%
                    summarise(molecule_start = min(molecule_start),
                              molecule_end = max(molecule_end)) %>%
                    mutate(length = molecule_end - molecule_start) %>%
                    filter(length > 10000) %>%
                    filter(length < 100000)
  # just select the molecules that are listed in genes_chosen_ranges
  # and check that only genes in that range are selected
  genes_chosen <- df_all_gene_info %>%
                    filter(genus %in% names(genusColors)) %>%
                    left_join(genes_chosen_ranges, by = "molecule") %>%
                    filter(!is.na(molecule_start) & !is.na(molecule_end)) %>%
                    mutate(include = ifelse(start >= molecule_start & end <= molecule_end, T, F)) %>%
                    filter(include) %>%
                    mutate(middle = (start + end) / 2) %>%
                    group_by(molecule) %>%
                    mutate(molecule_order = min(start)) %>%
                    arrange(molecule_order) %>%
                    ungroup() %>%
                    group_by(gene) %>%
                    mutate(pathway_name = paste(pathway_name, collapse = " | ")) %>%
                    mutate(pathway_id = paste(pathway_id, collapse = " | ")) %>%
                    select(!X) %>%
                    ungroup() %>%
                    unique() %>%
                    mutate(cazy_substrate = ifelse(grepl("Pectin", FUNCTION_AT_DESTINATION_3), "Pectin",
                                                  ifelse(grepl("Hemicellulose", FUNCTION_AT_DESTINATION_3), "Hemicellulose",
                                                  ifelse(grepl("Cellulose", FUNCTION_AT_DESTINATION_3), "Cellulose", "Other"))))
  if (nrow(genes_chosen) == 0) {
    print("No genes found in")
    print(sample)
    next
  }
  ggplot(genes_chosen, aes(xmin = start,
                            xmax = end,
                            x = middle,
                            y = molecule,
                            label = cazyme,
                            fill = genus)) +
    geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
    geom_gene_label(align = "left") +
    geom_text(aes(label = gene_desc_kegg), angle = 50, hjust = 0, size = 2) +
    # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 2) +
    scale_fill_manual(values = genusColors) +
    facet_wrap(~ molecule, scales = "free", ncol = 1) +
    labs(x = "Position", y = "Molecule", fill = "Cazyme") +
    make_theme(theme_name = theme_genes(),
              setFill = F, setCol = F,
              guide_nrow = 6, y_size = 5,
              leg_pos = "bottom")
    ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/cazy_genes_all/cazy_genes_all_", sample, ".pdf"), width = 20, height = 25)
  ggplot(genes_chosen, aes(xmin = start,
                            xmax = end,
                            x = middle,
                            y = molecule,
                            label = cazyme,
                            fill = genus)) +
    geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
    geom_gene_label(align = "left") +
    geom_text(aes(label = gene_desc_pfam), angle = 50, hjust = 0, size = 2) +
    # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 1) +
    scale_fill_manual(values = genusColors) +
    facet_wrap(~ molecule, scales = "free", ncol = 1) +
    labs(x = "Position", y = "Molecule", fill = "Cazyme") +
    make_theme(theme_name = theme_genes(),
              setFill = F, setCol = F,
              guide_nrow = 6, y_size = 5,
              leg_pos = "bottom")
    ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/cazy_genes_all/cazy_genes_all_pathwayname_", sample, ".pdf"), width = 20, height = 25)
    print("completed")
    print(sample)
    write.csv(genes_chosen, paste0("results/figures/08-gene_content_plots/gene_maps/cazy_genes_all/cazy_genes_all_", sample, ".csv"), row.names = F, quote = F)
    saveRDS(genes_chosen, paste0("results/figures/08-gene_content_plots/gene_maps/cazy_genes_all/cazy_genes_all_", sample, ".rds"))
}
```

# Plot gene maps for regions marked by antiSMASH

```{r}
all_region_genes <- data.frame()
# read antismash regions for all samples
for (sample in samples_IN_MY) {
  # ("results/figures/08-summarize_functions/09-antismash_parsed/D1-1_regions_summary.txt")
  # sample <- "D1-1"
  if (file.exists(paste0("results/figures/08-summarize_functions/09-antismash_parsed/", sample, "_regions_summary.txt"))) {
    print(sample)
  } else {
    print(paste0("No antismash regions available for ", sample))
    next
  }
  bgc_regions <- read.csv(paste0("results/figures/08-summarize_functions/09-antismash_parsed/", sample, "_regions_summary.txt"), sep = "\t") %>%
                  filter(edge != "True") %>%
                  rename(mag_antismash = mag) %>%
                  rename(species_antismash = species) %>%
                  rename(region_start = start) %>%
                  rename(region_end = end)
  df_covs_pos <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_cov.csv"))
  df_info <- read.csv(paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_info.csv"))
  df_annotations <- read.csv(paste0("results/08_gene_content/02_DRAM_annotations/", sample, "/annotations.tsv"), skip = 1, header = F, sep = "\t")
  colnames(df_annotations) <- c(
      "gene_dram_name",
      "sample",
      "rank",
      "ko_id",
      "kegg_hit",
      "peptidase_id",
      "peptidase_family",
      "peptidase_hit",
      "peptidase_RBH",
      "peptidase_identity",
      "peptidase_bitScore",
      "peptidase_eVal",
      "pfam_hits",
      "cazy_ids",
      "cazy_hits",
      "cazy_subfam_ec",
      "cazy_best_hit",
      "heme_regulatory_motif_count"
  )
  df_annotations_parsed <- df_annotations %>%
                    left_join(pathway_info, by = "ko_id") %>%
                    filter(ko_id != "") %>% 
                    mutate(gene = Vectorize(clean_dram_gene_name)(gene_dram_name)) %>%
                    mutate(gene_desc_kegg = kegg_hit) %>%
                    mutate(gene_desc_pfam = pfam_hits) %>%
                    select(gene, pathway_name, pathway_id, ko_id, gene_desc_pfam, gene_desc_kegg)
  df_all_gene_info <- left_join(df_info, df_covs_pos, by = "gene") %>%
                        mutate(molecule = Vectorize(get_molecule_from_gene)(gene)) %>%
                        left_join(df_annotations_parsed, by = "gene") %>%
                          left_join(glycan_annotations_final_cleaned, by = c("cazyme" = "Subfamily"))
  genes_chosen_ranges <- df_all_gene_info %>%
                    left_join(bgc_regions, by = c("molecule" = "scaffold")) %>%
                    filter(start >= region_start & end <= region_end) %>%
                    group_by(molecule) %>%
                    mutate(molecule_start = min(start),
                          molecule_end = max(end)) %>%
                    mutate(molecule_range = molecule_end - molecule_start) %>%
                    # filter(molecule_range > 10000) %>% # remove small molecules
                    filter(molecule_range < 1000000) %>% # remove those with genes too far away
                    mutate(num_genes = n()) %>%
                    filter(num_genes > 5) %>%
                    ungroup() %>%
                    group_by(molecule) %>%
                    mutate(molecule_name = molecule) %>%
                    summarise(molecule_start = min(molecule_start),
                              molecule_end = max(molecule_end))
  genes_chosen_sample <- df_all_gene_info %>%
                    left_join(bgc_regions, by = c("molecule" = "scaffold")) %>%
                    filter(start >= region_start & end <= region_end) %>%
                    left_join(genes_chosen_ranges, by = "molecule") %>%
                    filter(!is.na(molecule_start) & !is.na(molecule_end)) %>%
                    mutate(include = ifelse(start >= molecule_start & end <= molecule_end, T, F)) %>%
                    filter(include) %>%
                    mutate(middle = (start + end) / 2) %>%
                    group_by(molecule) %>%
                    mutate(molecule_order = min(start)) %>%
                    arrange(molecule_order) %>%
                    ungroup() %>%
                    group_by(gene) %>%
                    mutate(pathway_name = paste(pathway_name, collapse = " | ")) %>%
                    mutate(pathway_id = paste(pathway_id, collapse = " | ")) %>%
                    select(!X) %>%
                    ungroup() %>%
                    unique() %>%
                    mutate(sample = sample)
  if (nrow(genes_chosen) == 0) {
    print("No genes found in")
    print(sample)
    next
  }
  all_region_genes <- rbind(all_region_genes, genes_chosen_sample)
}
view(all_region_genes)
  # ggplot(genes_chosen, aes(xmin = start,
  #                           xmax = end,
  #                           x = middle,
  #                           y = molecule,
  #                           label = cazyme,
  #                           fill = genus)) +
  #   geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) +
  #   geom_gene_label(align = "left") +
  #   geom_text(aes(label = gene_desc_kegg), angle = 50, hjust = 0, size = 2) +
  #   # geom_text_repel(aes(label = gene_desc_kegg), angle = 50, hjust = -0.2, size = 2) +
  #   scale_fill_manual(values = genusColors) +
  #   facet_wrap(~ molecule, scales = "free", ncol = 1) +
  #   labs(x = "Position", y = "Molecule", fill = "Cazyme") +
  #   make_theme(theme_name = theme_genes(),
  #             setFill = F, setCol = F,
  #             guide_nrow = 6, y_size = 5,
  #             leg_pos = "bottom")
  #   ggsave(paste0("results/figures/08-gene_content_plots/gene_maps/antismash_genes/antismash_genes_", sample, ".pdf"), width = 20, height = 25)
```

# Taxonomy resolved functional comparison

```{r}
# MAG ANIs
ani_matrix <- drep_N %>%
             select(reference, querry, ani) %>%
             pivot_wider(names_from=querry, values_from=ani) %>%
             replace(is.na(.), 0.7) %>%
             column_to_rownames("reference") %>%
             as.matrix
# per mag pcoa
genera_clean_sub_all <- c(
    "Lactobacillus",
    "Bifidobacterium",
    "Bombilactobacillus",
    "CALYQJ01",
    "Snodgrassella",
    "Gilliamella",
    "CALYQQ01",
    "Apibacter",
    "Bartonella_A",
    "Frischella",
    "Commensalibacter",
    "Apilactobacillus",
    # "Bombella",
    "Dysgonomonas",
    # "Enterobacter",
    "Pectinatus",
    # "Spiroplasma",
    # "Zymobacter",
    "Entomomonas",
    "Saezia",
    # "Parolsenella",
    "WRHT01",
    # "Cronobacter",
    # "Melissococcus",
    # "Rosenbergiella",
    "Pluralibacter"
    # "KACC-21507"
  )
all_mag_metadata <- read.csv("results/figures/magOTU_metadata_full_exported.csv") %>%
                      select(ID, Completeness, Contamination, Quality, Genus, MAG_species_name_final, Representative, Sample) %>%
                      mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
                      mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
                      mutate(Genus = Vectorize(clean_g_name)(Genus))
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    filter(Genus %in% genera_clean_sub_all) %>%
                    pull(ID)
# ran everything with rpkms as its redoing with "abs" where rpkm is adjusted with qPCR data
# df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_kos.csv")
df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")
df_rpkms_by_func <- df_rpkms_by_func_all %>%
                      filter(mag %in% mags_to_select)
ani_matrix <- ani_matrix[mags_to_select, mags_to_select]
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = Genus),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   size = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all))
                ) +
    geom_text_repel(aes(x = ev1, y = ev2, label = ifelse(Representative == 1, MAG_species_name_final, "")),
                      box.padding = 0.1, color = "#000000", size = 3) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    ggtitle("PCoA of MAGs based on adjusted RPKM (high and medium MAGs)") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
    ggsave("results/figures/11-figures/tax_func_plots/all_kos_abs_genus_circle_spec_marked.pdf", width = 15, height = 15)
    # ggsave("results/figures/11-figures/tax_func_plots/all_kos_genus_circle_spec_marked.pdf", width = 15, height = 15)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = Genus, label = Genus),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "elbow",
                        label.fontsize = 12, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.90, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    # add genus name next to species representative mag
    # add ellipse around each species
    ggtitle("PCoA of MAGs based on adjusted RPKM (high and medium MAGs)") +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
    ggsave("results/figures/11-figures/tax_func_plots/all_kos_abs_genus_circle_marked.pdf", width = 15, height = 15)
    # ggsave("results/figures/11-figures/tax_func_plots/all_kos_genus_circle_marked.pdf", width = 15, height = 15)
# dev.off()
# disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
# anova(disp_res)
ado_res = adonis2(dist_matrix ~ Genus + Host, 
        rownames(dist_matrix %>% as.matrix) %>% as.data.frame() %>%
        left_join(all_mag_metadata %>% select(ID, Sample, Genus), by = c("." = "ID")) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))
)
adonis_OmegaSq(ado_res)
# mags_to_select <- all_mag_metadata %>%
#                     filter(Quality != "low") %>%
#                     # filter(Completeness >= 95) %>%
#                     # filter(Contamination < 4) %>%
#                     # filter(Representative == "1") %>%
#                     filter(Genus %in% genera_clean_sub_all) %>%
#                     pull(ID)
# ani_matrix_df <- ani_matrix[mags_to_select, mags_to_select] %>%
#                   as.data.frame() %>%
#                   rownames_to_column("ID") %>%
#                   gather(key = "ID2", value = "ANI", -ID)
# func_dist_df <-  dist_matrix %>%
#                   as.matrix() %>%
#                   as.data.frame() %>%
#                   rownames_to_column("ID") %>%
#                   gather(key = "ID2", value = "func_dist", -ID)
# all_dist_df <- left_join(ani_matrix_df, func_dist_df, by = c("ID" = "ID", "ID2" = "ID2")) %>%
#                 filter(!is.na(ANI)) %>%
#                 filter(ANI != 0.7) %>%
#                 filter(ID2 != ID) %>%
#                 left_join(all_mag_metadata %>% select(ID, Sample, Genus, MAG_species_name_final, Host), by = c("ID" = "ID")) %>%
#                 left_join(all_mag_metadata %>% select(ID, Sample, Genus, MAG_species_name_final, Host), by = c("ID2" = "ID")) %>%
#                 mutate(Genus = Genus.x) %>%
#                 mutate(Genus2 = Genus.y) %>%
#                 mutate(MAG_species_name_final = MAG_species_name_final.x) %>%
#                 mutate(MAG_species_name_final2 = MAG_species_name_final.y) %>%
#                 mutate(Host = Host.x) %>%
#                 mutate(Host2 = Host.y)
# ggplot(all_dist_df) +
#   geom_point(aes(x = ANI, y = func_dist, color = Host, fill = Host2),
#                 width=0.001, size = 1, shape = 21, stroke = 2) +
#   geom_smooth(aes(x = ANI, y = func_dist), method = "lm", se = T) +
#   labs(x = "ANI", y = "Functional distance", color = "Genus", shape = "Host") +
#   scale_color_manual(values=host_order_color_dark) +
#   scale_fill_manual(values=host_order_color_dark) +
#   # facet_wrap(~Genus, scales = "free", ncol = 3) +
#   scale_shape_manual(values=c(
#     "Apis mellifera" = 16,
#     "Apis cerana" = 15,
#     "Apis dorsata" = 8,
#     "Apis florea" = 18,
#     "Apis andreniformis" = 17
#   )) +
#   make_theme(setFill = F, setCol = F, 
#     leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
#     x_size = 10, y_size = 10, modify_guide = F)



genera_clean_sub <- c(
    "Lactobacillus",
    "Bifidobacterium",
    "Bombilactobacillus",
    "CALYQJ01",
    "Apilactobacillus"
  )
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    # filter(MAG_species_name_final != "Bombilactobacillus SGB-37_0") %>%
                    # filter(MAG_species_name_final != "Lactobacillus SGB-28_1") %>%
                    filter(Genus %in% genera_clean_sub) %>%
                    pull(ID)
df_rpkms_by_func <- df_rpkms_by_func_all %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub),
                   color = factor(Genus, genera_clean_sub)),
                   size=4) +
    facet_wrap(~factor(Host, host_order)) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    # scale_shape_manual(values=c(
    #   "Apis mellifera" = 16,
    #   "Apis cerana" = 15,
    #   "Apis dorsata" = 8,
    #   "Apis florea" = 18,
    #   "Apis andreniformis" = 17
    # )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_kos_abs_lac_bif_spec_circle_marked.pdf", width = 15, height = 15)
  # ggsave("results/figures/11-figures/tax_func_plots/all_kos_lac_bif_spec_circle_marked.pdf", width = 15, height = 15)



genera_clean_sub <- c(
    "Gilliamella",
    "Frischella",
    "CALYQQ01"
  )
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    # filter(MAG_species_name_final != "Bombilactobacillus SGB-37_0") %>%
                    # filter(MAG_species_name_final != "Lactobacillus SGB-28_1") %>%
                    filter(Genus %in% genera_clean_sub) %>%
                    pull(ID)
df_rpkms_by_func <- df_rpkms_by_func_all %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub),
                   color = factor(Genus, genera_clean_sub)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    facet_wrap(~factor(Host, host_order)) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_kos_abs_gilli_frisch_spec_circle_marked.pdf", width = 15, height = 15)
  # ggsave("results/figures/11-figures/tax_func_plots/all_kos_gilli_frisch_spec_circle_marked.pdf", width = 15, height = 15)




# all_mag_metadata %>%
#                     # filter(Quality != "low") %>%
#                     filter(Quality == "high") %>% dim
# all_mag_metadata %>%
#                     # filter(Quality != "low") %>%
#                     filter(Quality != "low") %>% dim
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Quality == "high") %>%
                    # filter(Completeness > 70) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    group_by(MAG_species_name_final) %>%
                    mutate(num_mags_spec = n()) %>%
                    filter(num_mags_spec > 1) %>%
                    ungroup() %>%
                    filter(Genus %in% genera_clean_sub_all) %>%
                    pull(ID)
df_rpkms_by_func <- df_rpkms_by_func_all %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  # ggsave("results/figures/11-figures/tax_func_plots/all_kos_by_host_spec_circle_marked_highq.pdf", width = 20, height = 20)
  # ggsave("results/figures/11-figures/tax_func_plots/all_kos_by_host_spec_circle_marked.pdf", width = 20, height = 20)
  ggsave("results/figures/11-figures/tax_func_plots/all_kos_abs_by_host_spec_circle_marked.pdf", width = 20, height = 20)

mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    filter(Genus %in% genera_clean_sub_all) %>%
                    pull(ID)
df_rpkms_by_func <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
ani_matrix <- ani_matrix[mags_to_select, mags_to_select]
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = Genus),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   size = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all))
                ) +
    geom_text_repel(aes(x = ev1, y = ev2, label = ifelse(Representative == 1, MAG_species_name_final, "")),
                      box.padding = 0.1, color = "#000000", size = 3) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
    ggsave("results/figures/11-figures/tax_func_plots/all_cazy_genus_circle_spec_marked.pdf", width = 15, height = 15)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = Genus, label = Genus),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "elbow",
                        label.fontsize = 12, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.90, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    # add genus name next to species representative mag
    # add ellipse around each species
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
    ggsave("results/figures/11-figures/tax_func_plots/all_cazy_genus_circle_marked.pdf", width = 15, height = 15)
# dev.off()
# disp_res = betadisper(dist_obj, rownames(dist_obj %>% as.matrix) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host))
# anova(disp_res)
ado_res = adonis2(dist_matrix ~ Genus + Host, 
        rownames(dist_matrix %>% as.matrix) %>% as.data.frame() %>%
        left_join(all_mag_metadata %>% select(ID, Sample, Genus), by = c("." = "ID")) %>%
        mutate(Host = Vectorize(get_host_from_sample_name)(Sample))
)
adonis_OmegaSq(ado_res)
mags_to_select <- all_mag_metadata %>%
                    filter(Quality == "high") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    filter(Genus %in% genera_clean_sub_all) %>%
                    pull(ID)
ani_matrix_df <- ani_matrix[mags_to_select, mags_to_select] %>%
                  as.data.frame() %>%
                  rownames_to_column("ID") %>%
                  gather(key = "ID2", value = "ANI", -ID)
func_dist_df <-  dist_matrix %>%
                  as.matrix() %>%
                  as.data.frame() %>%
                  rownames_to_column("ID") %>%
                  gather(key = "ID2", value = "func_dist", -ID)
all_dist_df <- left_join(ani_matrix_df, func_dist_df, by = c("ID" = "ID", "ID2" = "ID2")) %>%
                filter(!is.na(ANI)) %>%
                filter(ANI != 0.7) %>%
                filter(ID2 != ID) %>%
                left_join(all_mag_metadata %>% select(ID, Sample, Genus, MAG_species_name_final, Host), by = c("ID" = "ID")) %>%
                left_join(all_mag_metadata %>% select(ID, Sample, Genus, MAG_species_name_final, Host), by = c("ID2" = "ID")) %>%
                mutate(Genus = Genus.x) %>%
                mutate(Genus2 = Genus.y) %>%
                mutate(MAG_species_name_final = MAG_species_name_final.x) %>%
                mutate(MAG_species_name_final2 = MAG_species_name_final.y) %>%
                mutate(Host = Host.x) %>%
                mutate(Host2 = Host.y)
ggplot(all_dist_df) +
  geom_point(aes(x = ANI, y = func_dist, color = Host, fill = Host2),
                width=0.001, size = 1, shape = 21, stroke = 2) +
  geom_smooth(aes(x = ANI, y = func_dist), method = "lm", se = T) +
  labs(x = "ANI", y = "Functional distance", color = "Genus", shape = "Host") +
  scale_color_manual(values=host_order_color_dark) +
  scale_fill_manual(values=host_order_color_dark) +
  # facet_wrap(~Genus, scales = "free", ncol = 3) +
  scale_shape_manual(values=c(
    "Apis mellifera" = 16,
    "Apis cerana" = 15,
    "Apis dorsata" = 8,
    "Apis florea" = 18,
    "Apis andreniformis" = 17
  )) +
  make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)



genera_clean_sub <- c(
    "Lactobacillus",
    "Bifidobacterium",
    "Bombilactobacillus",
    "CALYQJ01",
    "Apilactobacillus"
  )
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Quality == "high") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    filter(MAG_species_name_final != "Bombilactobacillus SGB-37_0") %>%
                    filter(MAG_species_name_final != "Lactobacillus SGB-28_1") %>%
                    filter(Genus %in% genera_clean_sub) %>%
                    pull(ID)
df_rpkms_by_func <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub),
                   color = factor(Genus, genera_clean_sub)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_cazy_lac_bif_spec_circle_marked.pdf", width = 15, height = 15)



genera_clean_sub <- c(
    "Gilliamella",
    "Frischella",
    "CALYQQ01"
  )
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Quality == "high") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    filter(MAG_species_name_final != "Bombilactobacillus SGB-37_0") %>%
                    filter(MAG_species_name_final != "Lactobacillus SGB-28_1") %>%
                    filter(Genus %in% genera_clean_sub) %>%
                    pull(ID)
df_rpkms_by_func <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub),
                   color = factor(Genus, genera_clean_sub)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 15,
      "Apis dorsata" = 8,
      "Apis florea" = 18,
      "Apis andreniformis" = 17
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3.5,
      "Apis florea" = 3,
      "Apis andreniformis" = 2
    )) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_cazy_gilli_frisch_spec_circle_marked.pdf", width = 15, height = 15)




mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Quality != "high") %>%
                    # filter(Representative == "1") %>%
                    group_by(MAG_species_name_final) %>%
                    mutate(num_mags_spec = n()) %>%
                    filter(num_mags_spec > 1) %>%
                    ungroup() %>%
                    filter(ID != "M5-3_19") %>% # had a very high ev1 = 4.8073 possible contamination...
                    filter(Genus %in% genera_clean_sub_all) %>%
                    pull(ID)
df_rpkms_by_func <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_cazy_by_host_spec_circle_marked.pdf", width = 20, height = 20)
```

```{r}
# Make genus-level resolved plots all KOs but tie them into the same multi-page pdf using
# using marrange grobs rather than ggsave so there is more space for each
# genus = "Lactobacillus"
p_list <- list()
for (genus in genera_clean_sub_all) {
  mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Completeness >= 95) %>%
                      # filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
  df_rpkms_by_func <- df_rpkms_by_func_all %>%
  # df_rpkms_by_func <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_kos.csv") %>%
                        filter(mag %in% mags_to_select)
  rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
  dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
  matrix <- as.matrix(dist_matrix)
  dist <- as.dist(matrix)
  res_pcoa <- pcoa(dist)
  ev1 <- res_pcoa$vectors[,1]
  ev2 <- res_pcoa$vectors[,2]
  df_pcoa_new <- data.frame(cbind(ev1,ev2))
  df_pcoa_new$Sample <- rownames(df_pcoa_new)
  rownames(df_pcoa_new) <- NULL
  df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
  perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
  axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
  axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
  p <- ggplot(df_pcoa_new) +
      geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                          color = NA, fill = "#000000", geom = "polygon",
                          label.margin = margin(0.5, 0.5, 0.5, 0.5, "mm"),
                          con.type = "straight",
                          label.fontsize = 10, con.cap = unit(1, "mm"),
                          alpha = 0.05, level = 0.95, size = 0.5) +
      geom_point(aes(x = ev1,
                     y = ev2,
                     shape = factor(Host, host_order),
                     fill = factor(Genus, genera_clean_sub_all),
                     color = factor(Genus, genera_clean_sub_all)),
                     size=4) +
      labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
      scale_color_manual(values=genusColorsClean) +
      scale_fill_manual(values=genusColorsClean) +
      # facet_wrap(~factor(Host, host_order)) +
      scale_shape_manual(values=c(
        "Apis mellifera" = 16,
        "Apis cerana" = 15,
        "Apis dorsata" = 8,
        "Apis florea" = 18,
        "Apis andreniformis" = 17
      )) +
      scale_size_manual(values=c(
        "Apis mellifera" = 3,
        "Apis cerana" = 3,
        "Apis dorsata" = 3.5,
        "Apis florea" = 3,
        "Apis andreniformis" = 2
      )) +
      guides(fill = "none", size = "none") +
      make_theme(setFill = F, setCol = F, 
      leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
      x_size = 10, y_size = 10, modify_guide = F)
    # ggsave(paste0("results/figures/11-figures/tax_func_plots/all_kos_by_genus_",genus,"_circle_marked.pdf"), width = 15, height = 15)
    p_list[[genus]] <- p
}
pdf("results/figures/11-figures/tax_func_plots/all_kos_abs_by_genus_spec_circle_marked_combined.pdf", width = 15, height = 20)
# pdf("results/figures/11-figures/tax_func_plots/all_kos_by_genus_spec_circle_marked_faceted_combined.pdf", width = 15, height = 20)
marrangeGrob(p_list, nrow = 2, ncol = 1)
dev.off()

```

# Make KO heatmaps to look at patterns per genus

```{r}
# here rpkm is scalled by a norm factor instead redo this by using a more appropriate measure that can actually be based on copy number! (?)
df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")
ko_category_info <- read.csv("results/figures/08-summarize_functions/kegg_info_dict.txt", sep = "\t")
kos_to_exclude <- ko_category_info %>%
                    filter(A %in% c("Human Diseases", "Organismal Systems")) %>%
                    # filter(A %in% c("Metabolism", "Genetic Information Processing", "Environmental Information Processing")) %>%
                    pull(ko)

# plot for all species but only kos in "Metabolism of cofactors and vitamins" as category B
rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
kos_b_to_include <- ko_category_info %>%
                    filter(B %in% c("Metabolism of cofactors and vitamins")) %>%
                    pull(ko)
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_b_to_include]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
col_split = category_C %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_metabolism_cofactors_vitamins.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

# plot for all species but only kos in "Metabolism of cofactors and vitamins" as category B
rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
kos_b_to_include <- ko_category_info %>%
                    filter(B %in% c("Amino acid metabolism", "Metabolism of other amino acids")) %>%
                    pull(ko)
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_b_to_include]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
col_split = category_C %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_metabolism_amino_acid.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


# plot for all species but only kos in "Metabolism of cofactors and vitamins" as category B
rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
kos_b_to_include <- ko_category_info %>%
                    filter(!A %in% c("Human Diseases", "Organismal Systems")) %>%
                    filter(B %in% c("Xenobiotics biodegradation and metabolism", "Metabolism of terpenoids and polyketides", "Biosynthesis of other secondary metabolites")) %>%
                    pull(ko)
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_b_to_include]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
col_split = category_C %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_metabolism_secondary.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

# do secretion system better...
# plot for all species but only kos in "Metabolism of cofactors and vitamins" as category B
rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
df_ko_info_anno <- data.frame()
for (sample in samples_IN_MY) {
  df_annotations <- read.csv(paste0("results/08_gene_content/02_DRAM_annotations/", sample, "/annotations.tsv"), skip = 1, header = F, sep = "\t")
  colnames(df_annotations) <- c(
      "gene_dram_name",
      "sample",
      "rank",
      "ko_id",
      "kegg_hit",
      "peptidase_id",
      "peptidase_family",
      "peptidase_hit",
      "peptidase_RBH",
      "peptidase_identity",
      "peptidase_bitScore",
      "peptidase_eVal",
      "pfam_hits",
      "cazy_ids",
      "cazy_hits",
      "cazy_subfam_ec",
      "cazy_best_hit",
      "heme_regulatory_motif_count"
  )
  df_ko_info_anno_iter <- df_annotations %>% filter(grepl("ecretion", pfam_hits) | grepl("ecretion", kegg_hit)) %>%
    select(ko_id, kegg_hit, pfam_hits) #%>%
  df_ko_info_anno <- rbind(df_ko_info_anno, df_ko_info_anno_iter)
    # mutate(desc = ifelse(grepl("Type I", pfam_hits) | grepl("type I", kegg_hit) ))
}
df_ko_info_anno_sub <- df_ko_info_anno %>% select(ko_id, kegg_hit) %>% unique() %>% filter(ko_id != "")
# kos_b_to_include <- ko_category_info %>%
#                     filter(!A %in% c("Human Diseases", "Organismal Systems")) %>%
#                     # filter(B %in% c("Membrane transport")) %>%
#                     filter(grepl("ecretion", C)) %>%
#                     pull(ko)
kos_to_include <- df_ko_info_anno$ko_id
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_to_include]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                pull(kegg_hit) %>%
                abbreviate(50)
# category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(A)
# category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(B)
# category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(C)
col_split = desc_names %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_secretion.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
kos_b_to_include <- ko_category_info %>%
                    filter(!A %in% c("Human Diseases", "Organismal Systems")) %>%
                    filter(B %in% c("Lipid metabolism")) %>%
                    # filter(grepl("ecretion", C)) %>%
                    pull(ko)
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_b_to_include]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
col_split = category_C %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_lipid_metabolism.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
kos_b_to_include <- ko_category_info %>%
                    filter(!A %in% c("Human Diseases", "Organismal Systems")) %>%
                    filter(B %in% c("Signaling molecules and interaction")) %>%
                    pull(ko)
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_b_to_include]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
col_split = category_C %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_signalling.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

# look at the heatmap of respiratory genes from DRAM
rpkm_matrix <- df_rpkms_by_func_all %>% column_to_rownames("mag") %>% as.matrix()
# remove genes found in < 5 samples
rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
df_dram_kos_info <- read.csv("scripts/visualization/DRAM_KOs_list_from_excel.txt", sep = "\t")
colnames(df_dram_kos_info) <- c("ko", "description", "sub_category", "type", "category")
kos_DRAM_to_include <- df_dram_kos_info %>%
                        filter(category %in% c("Energy")) %>% 
                        # select(sub_category) %>% unique
                        pull(ko)
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_DRAM_to_include]
sub_category_names <-  data.frame(ko = colnames(rpkm_matrix)) %>%
                        left_join(df_dram_kos_info, by = "ko") %>%
                        pull(sub_category) %>% as.character %>%
                        abbreviate(50)
type_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_dram_kos_info, by = "ko") %>%
                pull(type)
updated_rpkm_matrix_df <- rpkm_matrix %>% t() %>% as.data.frame() %>% rownames_to_column("ko") %>%
                      left_join(df_dram_kos_info, by = "ko") %>%
                      group_by(ko) %>%
                      # add a serial number for kos numbered from 1 for each ko
                      mutate(serial_num = row_number()) %>%               
                      mutate(ko_uniq = paste0(ko, "_", serial_num)) %>%
                      unique
updated_rpkm_matrix <- updated_rpkm_matrix_df %>% column_to_rownames("ko_uniq") %>% select(-ko, -description, -sub_category, -type, -category, -serial_num) %>% t
ko_info_df_sub <- updated_rpkm_matrix_df %>% select(ko, description, sub_category, type, category)
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
sub_category_names <- data.frame(ko = colnames(updated_rpkm_matrix)) %>%
                        left_join(updated_rpkm_matrix_df %>% ungroup() %>% select(ko_uniq, sub_category), by = c("ko" = "ko_uniq")) %>%
                        pull(sub_category) %>% as.character %>%
                        abbreviate(50)
type_names <- data.frame(ko = colnames(updated_rpkm_matrix)) %>%
                        left_join(updated_rpkm_matrix_df %>% ungroup() %>% select(ko_uniq, type), by = c("ko" = "ko_uniq")) %>%
                        pull(type) %>% as.character %>%
                        abbreviate(50)
ko_names <- colnames(updated)
col_split = type_names %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj = Heatmap(t(updated_rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # right_annotation = anno_sub_cat_text,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_all_species_energy_DRAM.pdf", width = 30, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

# first heatmap split by dram subcategotu labelled color bar for type made for each category
# do Transporters, carbon utilization (Woodcroft), carbon utilization, 
# and for the heatmap, do pcoas for interesting subsets
# plot faceted across hosts but only choose kos involved in "carbon metabolism"
# as defined in the DRAM output and do one including also cazymes with it and one without
mags_to_select <- all_mag_metadata %>%
                    filter(Quality != "low") %>%
                    # filter(Completeness >= 95) %>%
                    # filter(Contamination < 4) %>%
                    # filter(Representative == "1") %>%
                    filter(Genus %in% genera_clean_sub_all) %>%
                    pull(ID)
# df_rpkms_by_func # all kos
df_dram_kos_info <- read.csv("scripts/visualization/DRAM_KOs_list_from_excel.txt", sep = "\t")
colnames(df_dram_kos_info) <- c("ko", "description", "sub_category", "type", "category")
kos_DRAM_to_include <- df_dram_kos_info %>%
                        # filter(category %in% c("carbon utilization", "carbon utilization (woodcroft)")) %>% 
                        filter(category %in% c("Transporters")) %>% 
                        pull(ko)
rpkm_matrix <- df_rpkms_by_func %>% column_to_rownames("mag") %>% as.matrix()
rpkm_matrix <- rpkm_matrix[, colnames(rpkm_matrix) %in% kos_DRAM_to_include]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
dram_sub_category <- data.frame(ko = ko_names) %>%
                left_join(df_dram_kos_info %>% unique, by = "ko") %>%
                group_by(ko) %>%
                summarise(sub_category = first(sub_category)) %>%
                pull(sub_category) %>% as.character
col_split = dram_sub_category %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj <- Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T        
)
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_DRAM_Transporters.pdf", width = 30, height = 30) 
draw(heatmap_obj, merge_legend = TRUE)
dev.off()
# remove the mag A5-3_6
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_dram_ko_Transporters_cazy_by_host.pdf", width = 20, height = 20)

kos_DRAM_to_include <- df_dram_kos_info %>%
                        filter(category %in% c("carbon utilization (Woodcroft)", "carbon utilization")) %>% 
                        pull(ko)
df_rpkms_by_func_cazy <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func %>% select(any_of(c("mag", kos_DRAM_to_include))) %>% column_to_rownames("mag") %>% as.matrix()
# rpkm_matrix <- cbind(df_rpkms_by_func %>% select(any_of(kos_DRAM_to_include)), df_rpkms_by_func_cazy) %>% column_to_rownames("mag") %>% as.matrix()
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
dram_sub_category <- data.frame(ko = ko_names) %>%
                left_join(df_dram_kos_info %>% unique, by = "ko") %>%
                group_by(ko) %>%
                summarise(sub_category = first(sub_category)) %>%
                pull(sub_category) %>% as.character
dram_type <- data.frame(ko = ko_names) %>%
                left_join(df_dram_kos_info %>% unique, by = "ko") %>%
                group_by(ko) %>%
                summarise(type = first(type)) %>%
                pull(type) %>% as.character
col_split = dram_sub_category %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj <- Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T        
)
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_carbon_utilization_all.pdf", width = 30, height = 30) 
draw(heatmap_obj, merge_legend = TRUE)
dev.off()
# remove empty rows and cols
rpkm_matrix <- rpkm_matrix[-which(rownames(rpkm_matrix) %in% c("A2-3_13", "A5-3_6")),] # extreme values
rpkm_matrix <- rpkm_matrix[rowSums(rpkm_matrix) > 0, colSums(rpkm_matrix) > 0]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_dram_ko_carbon_util_all_by_host.pdf", width = 20, height = 20)

df_rpkms_by_func_cazy <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func_cazy %>% column_to_rownames("mag") %>% as.matrix()
# remove empty rows and cols
rpkm_matrix <- rpkm_matrix[-which(rownames(rpkm_matrix) %in% c("M5-3_19")),] # extreme values
rpkm_matrix <- rpkm_matrix[rowSums(rpkm_matrix) > 0, colSums(rpkm_matrix) > 0]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_cazy_by_host.pdf", width = 20, height = 20)


kos_DRAM_to_include <- df_dram_kos_info %>%
                        filter(category %in% c("carbon utilization (Woodcroft)", "carbon utilization")) %>% 
                        filter(type %in% c("central carbon")) %>%
                        pull(ko)
rpkm_matrix <- df_rpkms_by_func %>% select(any_of(c("mag", kos_DRAM_to_include))) %>% column_to_rownames("mag") %>% as.matrix()
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
genus_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            ungroup() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus) %>% as.character
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                   Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_genus_col = HeatmapAnnotation(Host = host_names,
                                       Genus = genus_names,
                                        col = list(Host = host_order_color,
                                                   Genus = unlist(extend_colors(genus_names, genusColorsClean))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
dram_sub_category <- data.frame(ko = ko_names) %>%
                left_join(df_dram_kos_info %>% unique, by = "ko") %>%
                group_by(ko) %>%
                summarise(sub_category = first(sub_category)) %>%
                pull(sub_category) %>% as.character
col_split = dram_sub_category %>% as.factor
row_split = species_names %>% as.factor
heatmap_obj <- Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 1000, 10000), c("#ffffff", "#c6dbef", "#6baed6", "#4292c6", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_genus_col,
            # top_annotation = anno_host_spec_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes abundance (RPKM)"),
            # cell_fun = my_cell_fun,
            cluster_columns = F,
            column_split = row_split,
            row_split = col_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            column_title_gp = grid::gpar(fontsize = 10),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T        
)
pdf("results/figures/11-figures/tax_func_heatmaps/ko_heatmap_central_carbon_utilization_all.pdf", width = 30, height = 30) 
draw(heatmap_obj, merge_legend = TRUE)
dev.off()
# remove empty rows and cols
rpkm_matrix <- rpkm_matrix[-which(rownames(rpkm_matrix) %in% c("F7-1_9")),] # extreme values
rpkm_matrix <- rpkm_matrix[rowSums(rpkm_matrix) > 0, colSums(rpkm_matrix) > 0]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_dram_ko_central_carbon_util_all_by_host.pdf", width = 20, height = 20)

df_rpkms_by_func_cazy <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
rpkm_matrix <- df_rpkms_by_func_cazy %>% column_to_rownames("mag") %>% as.matrix()
# remove empty rows and cols
rpkm_matrix <- rpkm_matrix[-which(rownames(rpkm_matrix) %in% c("M5-3_19")),] # extreme values
rpkm_matrix <- rpkm_matrix[rowSums(rpkm_matrix) > 0, colSums(rpkm_matrix) > 0]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_cazy_by_host.pdf", width = 20, height = 20)


df_rpkms_by_func_cazy <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_all_cazy.csv") %>%
                      filter(mag %in% mags_to_select)
cazy_to_exclude <- df_rpkms_by_func_cazy %>% select(starts_with("GT"), starts_with("AA")) %>% colnames
rpkm_matrix <- df_rpkms_by_func_cazy %>% column_to_rownames("mag") %>% as.matrix()
# only keep columns not starting with GT and AA
rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% cazy_to_exclude]
# rpkm_matrix <- rpkm_matrix[-which(rownames(rpkm_matrix) %in% c("M5-3_19")),] # extreme values
# remove empty rows and cols
rpkm_matrix <- rpkm_matrix[rowSums(rpkm_matrix) > 0, colSums(rpkm_matrix) > 0]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
                        con.type = "straight",
                        label.fontsize = 5, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all),
                   color = factor(Genus, genera_clean_sub_all)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 16,
      "Apis cerana" = 16,
      "Apis dorsata" = 16,
      "Apis florea" = 16,
      "Apis andreniformis" = 16
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave("results/figures/11-figures/tax_func_plots/all_cazy_exclude_AA_GT_by_host.pdf", width = 20, height = 20)


# extra code
# # remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
# # make a presence absence matrix
# rpkm_matrix_pa <- rpkm_matrix
# rpkm_matrix_pa[rpkm_matrix > 0] <- 1
```

# Identify differentially abundant functions between bacterial species / hosts

## Using the randomforest_res approach

# Identify differentially abundant functions between bacterial species / hosts

## Between related but host-specific species

Apibacter
Commensalibacter
eg. Apibacter adventoris and Apibacter sp002964915

```{r}
pathway_categories <- read.csv("scripts/visualization/kegg_pathway_categories.tsv", sep = "\t")
```

```{r}
df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")
dir.create("results/figures/11-figures/maaslin2_by_taxa/", showWarnings = F)

df_ko_info_anno <- data.frame()
for (sample in samples_IN_MY) {
  df_annotations <- read.csv(paste0("results/08_gene_content/02_DRAM_annotations/", sample, "/annotations.tsv"), skip = 1, header = F, sep = "\t")
  colnames(df_annotations) <- c(
      "gene_dram_name",
      "sample",
      "rank",
      "ko_id",
      "kegg_hit",
      "peptidase_id",
      "peptidase_family",
      "peptidase_hit",
      "peptidase_RBH",
      "peptidase_identity",
      "peptidase_bitScore",
      "peptidase_eVal",
      "pfam_hits",
      "cazy_ids",
      "cazy_hits",
      "cazy_subfam_ec",
      "cazy_best_hit",
      "heme_regulatory_motif_count"
  )
  df_ko_info_anno_iter <- df_annotations %>% filter(grepl("ecretion", pfam_hits) | grepl("ecretion", kegg_hit)) %>%
    select(ko_id, kegg_hit, pfam_hits) #%>%
  df_ko_info_anno <- rbind(df_ko_info_anno, df_ko_info_anno_iter)
    # mutate(desc = ifelse(grepl("Type I", pfam_hits) | grepl("type I", kegg_hit) ))
}
df_ko_info_anno_sub <- df_ko_info_anno %>% select(ko_id, kegg_hit) %>% unique() %>% filter(ko_id != "")
df_ko_info_anno_pfam <- df_ko_info_anno %>% select(ko_id, pfam_hits) %>% unique() %>% filter(ko_id != "")

genus = "Apibacter"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)

dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
# for heatmap
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      filter(Completeness >= 95) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()

# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Apibacter.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Commensalibacter"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)

dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      filter(Completeness >= 95) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n > 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Commensalibacter.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Frischella"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order)) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Completeness >= 95) %>%
                      filter(Completeness >= 70) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n > 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      reference = "Species,Frischella perrara",
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color_dark,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Frischella.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Snodgrassella"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)

dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      filter(Completeness >= 95) %>%
                      # filter(Completeness >= 70) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n > 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      reference = "Species,Snodgrassella alvi",
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color_dark,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Snodgrassella.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Bifidobacterium"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(Sample != "F7-1_9") # too far off removed for visualization
        # filter(Sample != "A6-5_3") # too far off removed for visualization
    ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(0:20)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(Sample != "F7-1_9") # too far off removed for visualization
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(Sample != "F7-1_9") # too far off removed for visualization
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new %>%
        filter(Sample != "A6-5_3") # too far off removed for visualization
    ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(0:20)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Completeness >= 70) %>%
                      # filter(Completeness >= 70) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n > 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      reference = "Species,Bifidobacterium indicum",
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color_dark,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Bifidobacterium.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Bombilactobacillus"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Completeness >= 70) %>%
                      # filter(Completeness >= 70) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n > 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      reference = "Species,Bombilactobacillus mellis",
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color_dark,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Bombilactobacillus.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Lactobacillus"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("F7-1_7", "F7-1_2")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(0:20)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("F7-1_7", "F7-1_2")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("F7-1_7", "F7-1_2")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values=c(11:19)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Completeness >= 70) %>%
                      # filter(Completeness >= 70) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n > 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      reference = "Species,Lactobacillus apis",
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color_dark,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Lactobacillus.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


genus = "Gilliamella"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("D3-5_1")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(0:20)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("D3-5_1")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("D3-5_1")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)
mags_to_select <- all_mag_metadata %>%
                      # filter(Quality != "low") %>%
                      filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values=c(0:19)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_hq.pdf"), width = 20, height = 20)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus), showWarnings = F)
dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed"), showWarnings = F)
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Completeness >= 70) %>%
                      # filter(Completeness >= 70) %>%
                      filter(Contamination < 4) %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      group_by(MAG_species_name_final) %>%
                      mutate(n = n()) %>%
                      filter(n >= 5) %>%
                      ungroup() %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# rpkm_matrix[rpkm_matrix > 0] <- 1
# prepare data as needed for maasiln2
input_data <- rpkm_matrix
meta_table <- left_join(data.frame("ID" = rownames(input_data)), all_mag_metadata %>% mutate(Species = MAG_species_name_final) %>% select(ID, Host, Species), by = c("ID" = "ID"))
rownames(meta_table) <- meta_table$ID
input_metadata <- meta_table
saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  
  "
  input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
  input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
  fit_data = Maaslin2(
      input_data = input_data, 
      input_metadata = input_metadata,
      output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
      analysis_method = "LM", # consider
      transform = "LOG", # consider
      normalization = "TSS", # consider
      fixed_effects = c("Species"),
      max_significance = 0.05,
      min_prevalence = 0.1,
      min_abundance = 0,
      save_models = T,
      reference = "Species,Gilliamella apicola",
      plot_scatter = F,
      plot_heatmap = F)
  "
sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
            pull(feature)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(sig_kos)) %>%
                        as.matrix()
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color_dark,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                        col = list(Host = host_order_color_dark
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = species_names %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_host_col,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_Gilliamella.pdf", width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

genus = "Dysgonomonas"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(15:25)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)

genus = "CALYQJ01"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(15:25)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)

genus = "Bartonella_A"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus == genus) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("M5-3_19", "M4-1_4")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(10:25)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("M5-3_19", "M4-1_4")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("M5-3_19", "M4-1_4")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(10:25)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        filter(!Sample %in% c("M5-3_19", "M4-1_4")) # too far removed for visualization
  ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)

genus = "SnodSaez"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus %in% c("Snodgrassella", "Saezia")) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(10:25)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    scale_shape_manual(values = c(10:25)) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                  #  shape = MAG_species_name_final,
                   color = factor(Host, host_order),
                   fill = factor(Host, host_order)),
                   size=4) +
    facet_wrap(~MAG_species_name_final) +
    labs(x=axis_x_title, y = axis_y_title, color = "Host", fill = "Host") +
    scale_fill_manual(values=host_order_color_dark) +
    scale_color_manual(values=host_order_color_dark) +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_colors_only_spec.pdf"), width = 20, height = 20)
  
genus = "FrischGilli"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      filter(Genus %in% c("Frischella", "Gilliamella")) %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)


genus = "All"
mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      pull(ID)
# now using abs transformed RPKMs
rpkm_matrix <- df_rpkms_by_func_all %>% 
                filter(mag %in% mags_to_select) %>%
                column_to_rownames("mag") %>%
                as.matrix()
# first plot pcoa of all KOs faceted by host
# remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
dist_matrix <- vegdist(rpkm_matrix, method = "robust.aitchison")
matrix <- as.matrix(dist_matrix)
dist <- as.dist(matrix)
res_pcoa <- pcoa(dist)
ev1 <- res_pcoa$vectors[,1]
ev2 <- res_pcoa$vectors[,2]
df_pcoa_new <- data.frame(cbind(ev1,ev2))
df_pcoa_new$Sample <- rownames(df_pcoa_new)
rownames(df_pcoa_new) <- NULL
df_pcoa_new <- left_join(df_pcoa_new, all_mag_metadata, by = c("Sample" = "ID"))
perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
ggplot(df_pcoa_new) +
    geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final),
                        color = NA, fill = "#000000", geom = "polygon",
                        label.margin = margin(0.01, 0.01, 0.01, 0.01, "mm"),
                        con.type = "straight",
                        label.fontsize = 10, con.cap = unit(1, "mm"),
                        alpha = 0.05, level = 0.95, size = 0.5) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = factor(Genus, genera_clean_sub_all)),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_color_manual(values=genusColorsClean) +
    scale_fill_manual(values=genusColorsClean) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(fill = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host.pdf"), width = 20, height = 20)
ggplot(df_pcoa_new %>%
        left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index))
        ) +
    geom_point(aes(x = ev1,
                   y = ev2,
                   shape = factor(Host, host_order),
                   fill = rohdes_index),
                   color = "#000000",
                   size=4) +
    labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
    scale_fill_gradientn(limits=c(0,1),
                         colors = brewer.pal(9, "Blues")) +
    scale_shape_manual(values=c(
      "Apis mellifera" = 21,
      "Apis cerana" = 21,
      "Apis dorsata" = 21,
      "Apis florea" = 21,
      "Apis andreniformis" = 21
    )) +
    scale_size_manual(values=c(
      "Apis mellifera" = 3,
      "Apis cerana" = 3,
      "Apis dorsata" = 3,
      "Apis florea" = 3,
      "Apis andreniformis" = 3
    )) +
    facet_wrap(~factor(Host, host_order), drop = F) +
    guides(shape = "none", size = "none") +
    make_theme(setFill = F, setCol = F, 
    leg_pos = "right", leg_size = 8, guide_nrow = 30,
    x_size = 10, y_size = 10, modify_guide = F)
  ggsave(paste0("results/figures/11-figures/tax_func_plots/all_ko_", genus, "_by_host_with_spec.pdf"), width = 20, height = 20)

# df_pairwise <- data.frame()
# models_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/fits/models.rds"))
# for (i in 1:length(models_data)) {
#     model <- models_data[[i]]
#     conts <- summary(lsmeans(model, pairwise ~ Species, adjust = "FDR"))
#     df_to_add <- data.frame(
#                       feature = names(models_data)[[i]],
#                       contrast = conts$contrasts$contrast,
#                       p.value = conts$contrasts$p.value,
#                       estimate = conts$contrasts$estimate
#                     )
#     df_pairwise <- rbind(df_pairwise, df_to_add)
# }
# write.csv(df_pairwise, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed/pairwise_output.csv"), row.names = F, quote = F)
# df_sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
#                   left_join(ko_category_info, by = c("feature" = "ko")) %>%
#                   filter(A != "Human Diseases")
#   # make a bar of cat C
# category_order <- df_sig_kos %>% group_by(C) %>% summarize(n = n()) %>% arrange(n) %>% pull(C)
#   ggplot(df_sig_kos) +
#     geom_bar(aes(y = factor(C, category_order), fill = A), stat = "count") +
#     labs(title = "Significant KOs by category C") +
#     make_theme(setFill = T, setCol = F, palettefill = "Spectral",
#       leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
#       y_hj = 1,
#       x_size = 10, y_size = 10, modify_guide = F)
#     ggsave(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed/sig_kos_by_cat_C.pdf"), width = 10, height = 10)
#   # make a bar of cat B
# category_order <- df_sig_kos %>% group_by(B) %>% summarize(n = n()) %>% arrange(n) %>% pull(B)
#   ggplot(df_sig_kos) +
#     geom_bar(aes(y = factor(B, category_order), fill = A), stat = "count") +
#     labs(title = "Significant KOs by category B") +
#     make_theme(setFill = T, setCol = F, palettefill = "Spectral",
#       leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
#       y_hj = 1,
#       x_size = 10, y_size = 10, modify_guide = F)
#     ggsave(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results_parsed/sig_kos_by_cat_B.pdf"), width = 10, height = 10)


# # make a volcano
# ggplot(df_sig_kos) +
#   geom_point(aes(x = coef, y = -log10(qval), color = A)) +
#   geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
#   labs(title = paste0(df_sig_kos)) +
#   make_theme(setFill = F, setCol = F, 
#     leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
#     x_size = 10, y_size = 10, modify_guide = F)

# genus = "Snodgrassella"


```

```{r}
# Maaslin2 to see what are the most "different functional groups" between host-specific and generalist species
# do this using specificity index as a continuous variable and either
  # across all species
  # across all species but include species as a variable
  # by genus and focus on genera with both specificists and generalists

# Get out KOs that are found in at least 2 MAGs of all species of the genus
# Then those that are present in >80 % of generalists and <20 % of specialists and vice versa

# First consider a certain genus.

df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")  
kos_to_plot <- read.csv("results/figures/08-summarize_functions/enriched_kos_generalist.txt", sep = "\t", header = F)$V1

mags_to_select <- all_mag_metadata %>%
                      filter(Quality != "low") %>%
                      # filter(Completeness > 90) %>%
                      # filter(Quality == "high") %>%
                      # filter(Representative == "1") %>%
                      pull(ID)
rpkm_matrix <- df_rpkms_by_func_all %>%
                        filter(mag %in% mags_to_select) %>%
                        column_to_rownames("mag") %>%
                        select(any_of(kos_to_plot)) %>%
                        as.matrix()
# rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(MAG_species_name_final)
host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
                            left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Host)
rohdes_index = species_names %>% as.data.frame() %>%
                            left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index), by = c("." = "MAG_species_name_final")) %>%
                              pull(rohdes_index)
specific_or_not = species_names %>% as.data.frame() %>%
                            left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index), by = c("." = "MAG_species_name_final")) %>%
                              pull(specific)
anno_host_spec_col = HeatmapAnnotation(Host = host_names,
                                       Species = species_names,
                                        col = list(Host = host_order_color,
                                                     Species = make_col_list(species_names)
                                                  ),
                                          border = T
                                  )
anno_host_col = HeatmapAnnotation(Host = host_names,
                                  Specificity = rohdes_index,
                                        col = list(Host = host_order_color_dark,
                                                   Specificity = colorRamp2(c(0, 1), c("#ffffff", "#000000"))
                                                  ),
                                          border = T
                                  )
anno_spec_species = HeatmapAnnotation(Species = species_names,
                                      Specificity = rohdes_index,
                                        col = list(Species = make_col_list(species_names),
                                                   Specificity = colorRamp2(c(0, 1), c("#ffffff", "#000000"))
                                                  ),
                                          border = T
                                  )
ko_names <- colnames(rpkm_matrix)
category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(A)
category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(B)
category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
                left_join(ko_category_info, by = c(. = "ko")) %>%
                pull(C)
desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
                left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
                group_by(ko) %>%
                summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
                # filter(!is.na(pfam_hits)) %>%
                left_join(ko_category_info, by = c(ko = "ko")) %>%
                mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
                pull(desc) %>%
                abbreviate(50)
col_split = specific_or_not %>% as.factor
row_split = category_C %>% as.factor
heatmap_obj = Heatmap(t(rpkm_matrix),
            # col = col_input,
            # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
            col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
            # top_annotation = top_annotation_obj,
            # bottom_annotation = bottom_annotation_obj,
            top_annotation = anno_spec_species,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 5),
            heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
            # cell_fun = my_cell_fun,
            cluster_columns = T,
            column_split = col_split,
            row_split = row_split,
            column_title_rot = 90,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 8),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = T
            )
pdf(paste0("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_enriched_ko_list.pdf"), width = 20, height = 30)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

##### not so clear pattern using Maaslin2

# genus = "gilliamella_spec_vs_gen"
# df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")  
# mags_to_select <- all_mag_metadata %>%
#                       filter(Quality != "low") %>%
#                       filter(Completeness > 90) %>%
#                       # filter(Quality == "high") %>%
#                       # filter(Representative == "1") %>%
#                       filter(Genus == "Gilliamella") %>%
#                       pull(ID)
# # now using abs transformed RPKMs
# rpkm_matrix <- df_rpkms_by_func_all %>% 
#                 filter(mag %in% mags_to_select) %>%
#                 column_to_rownames("mag") %>%
#                 as.matrix()
# # remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# # rpkm_matrix[rpkm_matrix > 0] <- 1
# # prepare data as needed for maasiln2
# input_data <- rpkm_matrix
# meta_table <- left_join(data.frame("ID" = rownames(input_data)),
#                         all_mag_metadata %>%
#                         left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index)) %>%
#                           mutate(Species = MAG_species_name_final) %>%
#                           select(ID, Host, Species, rohdes_index), by = c("ID" = "ID")
#                   ) %>%
#                 mutate(specificity = ifelse(rohdes_index >= 0.95, "specific", "general"))
# rownames(meta_table) <- meta_table$ID
# input_metadata <- meta_table
# dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus))
# dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"))
# saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
# saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
#   "
#   input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
#   input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
#   fit_data = Maaslin2(
#       input_data = input_data, 
#       input_metadata = input_metadata,
#       output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
#       analysis_method = "LM", # consider
#       transform = "LOG", # consider
#       normalization = "TSS", # consider
#       fixed_effects = c("rohdes_index"),
#       max_significance = 0.05,
#       min_prevalence = 0.1,
#       min_abundance = 0,
#       save_models = T,
#       plot_scatter = F,
#       plot_heatmap = F)
#   "
# sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
#             filter(metadata == "rohdes_index") %>%
#             pull(feature)
# mags_to_select <- all_mag_metadata %>%
#                       filter(Quality != "low") %>%
#                       # filter(Completeness > 90) %>%
#                       # filter(Quality == "high") %>%
#                       # filter(Representative == "1") %>%
#                       filter(Genus == "Gilliamella") %>%
#                       pull(ID)
# rpkm_matrix <- df_rpkms_by_func_all %>%
#                         filter(mag %in% mags_to_select) %>%
#                         column_to_rownames("mag") %>%
#                         select(any_of(sig_kos)) %>%
#                         as.matrix()
# # rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
# species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
#                             left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
#                               pull(MAG_species_name_final)
# host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
#                             left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
#                               pull(Host)
# rohdes_index = species_names %>% as.data.frame() %>%
#                             left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index), by = c("." = "MAG_species_name_final")) %>%
#                               pull(rohdes_index)
# specific_or_not = species_names %>% as.data.frame() %>%
#                             left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index), by = c("." = "MAG_species_name_final")) %>%
#                               pull(specific)
# anno_host_spec_col = HeatmapAnnotation(Host = host_names,
#                                        Species = species_names,
#                                         col = list(Host = host_order_color,
#                                                      Species = make_col_list(species_names)
#                                                   ),
#                                           border = T
#                                   )
# anno_host_col = HeatmapAnnotation(Host = host_names,
#                                   Specificity = rohdes_index,
#                                         col = list(Host = host_order_color_dark,
#                                                    Specificity = colorRamp2(c(0, 1), c("#ffffff", "#000000"))
#                                                   ),
#                                           border = T
#                                   )
# anno_spec_species = HeatmapAnnotation(Species = species_names,
#                                       Specificity = rohdes_index,
#                                         col = list(Species = make_col_list(species_names),
#                                                    Specificity = colorRamp2(c(0, 1), c("#ffffff", "#000000"))
#                                                   ),
#                                           border = T
#                                   )
# ko_names <- colnames(rpkm_matrix)
# category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(A)
# category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(B)
# category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(C)
# desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
#                 left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
#                 group_by(ko) %>%
#                 summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
#                 # filter(!is.na(pfam_hits)) %>%
#                 left_join(ko_category_info, by = c(ko = "ko")) %>%
#                 mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
#                 pull(desc) %>%
#                 abbreviate(50)
# col_split = specific_or_not %>% as.factor
# row_split = category_C %>% as.factor
# heatmap_obj = Heatmap(t(rpkm_matrix),
#             # col = col_input,
#             # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
#             col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
#             # top_annotation = top_annotation_obj,
#             # bottom_annotation = bottom_annotation_obj,
#             top_annotation = anno_spec_species,
#             column_names_gp = grid::gpar(fontsize = 0),
#             row_names_gp = grid::gpar(fontsize = 5),
#             heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
#             # cell_fun = my_cell_fun,
#             cluster_columns = T,
#             column_split = col_split,
#             row_split = row_split,
#             column_title_rot = 90,
#             row_title_rot = 0,
#             row_title_gp = grid::gpar(fontsize = 8),
#             border = T,
#             # row_dend_reorder = T,
#             cluster_rows = T
#             )
# pdf(paste0("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_", genus,".pdf"), width = 20, height = 30)
# draw(heatmap_obj, merge_legend = TRUE)
# dev.off()

# genus = "bifidobacterium_spec_vs_gen"
# df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")  
# mags_to_select <- all_mag_metadata %>%
#                       filter(Quality != "low") %>%
#                       filter(Completeness > 90) %>%
#                       # filter(Quality == "high") %>%
#                       # filter(Representative == "1") %>%
#                       filter(Genus == "Bifidobacterium") %>%
#                       pull(ID)
# # now using abs transformed RPKMs
# rpkm_matrix <- df_rpkms_by_func_all %>% 
#                 filter(mag %in% mags_to_select) %>%
#                 column_to_rownames("mag") %>%
#                 as.matrix()
# # remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# # rpkm_matrix[rpkm_matrix > 0] <- 1
# # prepare data as needed for maasiln2
# input_data <- rpkm_matrix
# meta_table <- left_join(data.frame("ID" = rownames(input_data)),
#                         all_mag_metadata %>%
#                         left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index)) %>%
#                           mutate(Species = MAG_species_name_final) %>%
#                           select(ID, Host, Species, rohdes_index), by = c("ID" = "ID")
#                     )
# rownames(meta_table) <- meta_table$ID
# input_metadata <- meta_table
# dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus))
# dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"))
# saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
# saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
#   "
#   input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
#   input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
#   fit_data = Maaslin2(
#       input_data = input_data, 
#       input_metadata = input_metadata,
#       output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
#       analysis_method = "LM", # consider
#       transform = "LOG", # consider
#       normalization = "TSS", # consider
#       fixed_effects = c("rohdes_index", "Species"),
#       reference = "Species,Bifidobacterium apousia",
#       max_significance = 0.05,
#       min_prevalence = 0.1,
#       min_abundance = 0,
#       save_models = T,
#       plot_scatter = F,
#       plot_heatmap = F)
#   "
# sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
#             filter(metadata == "rohdes_index") %>%
#             pull(feature)
# mags_to_select <- all_mag_metadata %>%
#                       filter(Quality != "low") %>%
#                       # filter(Completeness > 90) %>%
#                       # filter(Quality == "high") %>%
#                       # filter(Representative == "1") %>%
#                       filter(Genus == "Gilliamella") %>%
#                       pull(ID)
# rpkm_matrix <- df_rpkms_by_func_all %>%
#                         filter(mag %in% mags_to_select) %>%
#                         column_to_rownames("mag") %>%
#                         select(any_of(sig_kos)) %>%
#                         as.matrix()
# # rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
# species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
#                             left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
#                               pull(MAG_species_name_final)
# host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
#                             left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
#                               pull(Host)
# rohdes_index = species_names %>% as.data.frame() %>%
#                             left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index), by = c("." = "MAG_species_name_final")) %>%
#                               pull(rohdes_index)
# specific_or_not = species_names %>% as.data.frame() %>%
#                             left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index), by = c("." = "MAG_species_name_final")) %>%
#                               pull(specific)
# anno_host_spec_col = HeatmapAnnotation(Host = host_names,
#                                        Species = species_names,
#                                         col = list(Host = host_order_color,
#                                                      Species = make_col_list(species_names)
#                                                   ),
#                                           border = T
#                                   )
# anno_host_col = HeatmapAnnotation(Host = host_names,
#                                   Specificity = rohdes_index,
#                                         col = list(Host = host_order_color_dark,
#                                                    Specificity = colorRamp2(c(0, 1), c("#ffffff", "#000000"))
#                                                   ),
#                                           border = T
#                                   )
# anno_spec_species = HeatmapAnnotation(Species = species_names,
#                                       Specificity = rohdes_index,
#                                         col = list(Species = make_col_list(species_names),
#                                                    Specificity = colorRamp2(c(0, 1), c("#ffffff", "#000000"))
#                                                   ),
#                                           border = T
#                                   )
# ko_names <- colnames(rpkm_matrix)
# category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(A)
# category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(B)
# category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(C)
# desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
#                 left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
#                 group_by(ko) %>%
#                 summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
#                 # filter(!is.na(pfam_hits)) %>%
#                 left_join(ko_category_info, by = c(ko = "ko")) %>%
#                 mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
#                 pull(desc) %>%
#                 abbreviate(50)
# col_split = specific_or_not %>% as.factor
# row_split = category_C %>% as.factor
# heatmap_obj = Heatmap(t(rpkm_matrix),
#             # col = col_input,
#             # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
#             col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
#             # top_annotation = top_annotation_obj,
#             # bottom_annotation = bottom_annotation_obj,
#             top_annotation = anno_host_col,
#             column_names_gp = grid::gpar(fontsize = 0),
#             row_names_gp = grid::gpar(fontsize = 5),
#             heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
#             # cell_fun = my_cell_fun,
#             cluster_columns = T,
#             column_split = col_split,
#             row_split = row_split,
#             column_title_rot = 90,
#             row_title_rot = 0,
#             row_title_gp = grid::gpar(fontsize = 8),
#             border = T,
#             # row_dend_reorder = T,
#             cluster_rows = T
#             )
# pdf(paste0("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_", genus,".pdf"), width = 20, height = 30)
# draw(heatmap_obj, merge_legend = TRUE)
# dev.off()

# genus = "all_genera_spec_vs_gen"
# df_rpkms_by_func_all <- read.csv("results/figures/08-summarize_functions/rpkm_matrix_by_function/rpkm_matrix_abs_all_kos.csv")  
# mags_to_select <- all_mag_metadata %>%
#                       filter(Quality != "low") %>%
#                       # filter(Quality == "high") %>%
#                       # filter(Representative == "1") %>%
#                       pull(ID)
# # now using abs transformed RPKMs
# rpkm_matrix <- df_rpkms_by_func_all %>% 
#                 filter(mag %in% mags_to_select) %>%
#                 column_to_rownames("mag") %>%
#                 as.matrix()
# # remove genes found in < 5 samples
# rpkm_matrix <- rpkm_matrix[, colSums(rpkm_matrix > 0) > 5]
# # rpkm_matrix[rpkm_matrix > 0] <- 1
# # prepare data as needed for maasiln2
# input_data <- rpkm_matrix
# meta_table <- left_join(data.frame("ID" = rownames(input_data)),
#                         all_mag_metadata %>%
#                         left_join(df_rohdes %>% filter(type == "prev") %>% select(specific, MAG_species_name_final, rohdes_index)) %>%
#                           mutate(Species = MAG_species_name_final) %>%
#                           select(ID, Host, Species, rohdes_index), by = c("ID" = "ID")
#                     )
# rownames(meta_table) <- meta_table$ID
# input_metadata <- meta_table
# dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus))
# dir.create(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"))
# saveRDS(input_data, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
# saveRDS(input_metadata, paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
#   "
#   input_data <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_data.rds"))
#   input_metadata <- readRDS(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_input_metadata.rds"))
#   fit_data = Maaslin2(
#       input_data = input_data, 
#       input_metadata = input_metadata,
#       output = paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results"), 
#       analysis_method = "LM", # consider
#       transform = "LOG", # consider
#       normalization = "TSS", # consider
#       fixed_effects = c("rohdes_index", "Species"),
#       reference = "Species,Bombilactobacillus mellis",
#       max_significance = 0.05,
#       min_prevalence = 0.1,
#       min_abundance = 0,
#       save_models = T,
#       plot_scatter = F,
#       plot_heatmap = F)
#   "
# sig_kos <- read.csv(paste0("results/figures/11-figures/maaslin2_by_taxa/",genus,"/maaslin2_results/significant_results.tsv"), sep = "\t") %>%
#             filter(metadata == "rohdes_index") %>%
#             pull(feature)
# rpkm_matrix <- df_rpkms_by_func_all %>%
#                         filter(mag %in% mags_to_select) %>%
#                         column_to_rownames("mag") %>%
#                         select(any_of(sig_kos)) %>%
#                         as.matrix()
# # rpkm_matrix <- rpkm_matrix[, !colnames(rpkm_matrix) %in% kos_to_exclude]
# species_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
#                             left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
#                               pull(MAG_species_name_final)
# host_names = rownames(rpkm_matrix) %>% as.data.frame() %>%
#                             left_join(all_mag_metadata %>% ungroup(), by = c(. = "ID")) %>%
#                               pull(Host)
# anno_host_spec_col = HeatmapAnnotation(Host = host_names,
#                                        Species = species_names,
#                                         col = list(Host = host_order_color,
#                                                      Species = make_col_list(species_names)
#                                                   ),
#                                           border = T
#                                   )
# anno_host_col = HeatmapAnnotation(Host = host_names,
#                                         col = list(Host = host_order_color_dark
#                                                   ),
#                                           border = T
#                                   )
# ko_names <- colnames(rpkm_matrix)
# category_A <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(A)
# category_B <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(B)
# category_C <- colnames(rpkm_matrix) %>% as.data.frame() %>%
#                 left_join(ko_category_info, by = c(. = "ko")) %>%
#                 pull(C)
# desc_names <- data.frame(ko = colnames(rpkm_matrix)) %>%
#                 left_join(df_ko_info_anno_sub, by = c(ko = "ko_id")) %>%
#                 group_by(ko) %>%
#                 summarize(kegg_hit = paste(kegg_hit, collapse = "|")) %>%
#                 # filter(!is.na(pfam_hits)) %>%
#                 left_join(ko_category_info, by = c(ko = "ko")) %>%
#                 mutate(desc = ifelse(is.na(kegg_hit), C, kegg_hit)) %>%
#                 pull(desc) %>%
#                 abbreviate(50)
# col_split = species_names %>% as.factor
# row_split = category_C %>% as.factor
# heatmap_obj = Heatmap(t(rpkm_matrix),
#             # col = col_input,
#             # col = colorRamp2(c(0, 1), c("#ffffff", "#08306b")),
#             col = colorRamp2(c(0, 10, 100, 10000, 100000), c("#ffffff", "#c6dbef", "#6baed6", "#2171b5", "#08306b")),
#             # top_annotation = top_annotation_obj,
#             # bottom_annotation = bottom_annotation_obj,
#             top_annotation = anno_host_col,
#             column_names_gp = grid::gpar(fontsize = 0),
#             row_names_gp = grid::gpar(fontsize = 5),
#             heatmap_legend_param = list(title = "Genes presence (adjusted RPKM > 0)"),
#             # cell_fun = my_cell_fun,
#             cluster_columns = T,
#             column_split = col_split,
#             row_split = row_split,
#             column_title_rot = 90,
#             row_title_rot = 0,
#             row_title_gp = grid::gpar(fontsize = 8),
#             border = T,
#             # row_dend_reorder = T,
#             cluster_rows = T
#             )
# pdf(paste0("results/figures/11-figures/tax_func_heatmaps/ko_sig_heatmap_", genus,".pdf"), width = 20, height = 30)
# draw(heatmap_obj, merge_legend = TRUE)
# dev.off()
```

### Showing strain-level diversity


### No strain-level diversity


```{r}
# ggplot(df_pcoa_new) +
#     geom_mark_ellipse(aes(x = ev1, y = ev2, group = MAG_species_name_final, label = MAG_species_name_final),
#                         color = NA, fill = "#000000", geom = "polygon",
#                         label.margin = margin(0.1, 0.1, 0.1, 0.1, "mm"),
#                         con.type = "straight",
#                         label.fontsize = 5, con.cap = unit(1, "mm"),
#                         alpha = 0.05, level = 0.95, size = 0.5) +
#     geom_point(aes(x = ev1,
#                    y = ev2,
#                    shape = factor(Host, host_order),
#                    fill = factor(Genus, genera_clean_sub_all),
#                    color = factor(Genus, genera_clean_sub_all)),
#                    size=4) +
#     labs(x=axis_x_title, y = axis_y_title, color = "Genus", shape = "Host") +
#     scale_color_manual(values=genusColorsClean) +
#     scale_fill_manual(values=genusColorsClean) +
#     scale_shape_manual(values=c(
#       "Apis mellifera" = 16,
#       "Apis cerana" = 15,
#       "Apis dorsata" = 8,
#       "Apis florea" = 18,
#       "Apis andreniformis" = 17
#     )) +
#     scale_size_manual(values=c(
#       "Apis mellifera" = 3,
#       "Apis cerana" = 3,
#       "Apis dorsata" = 3.5,
#       "Apis florea" = 3,
#       "Apis andreniformis" = 2
#     )) +
#     facet_wrap(~Genus, scales = "free", ncol = 3) +
#     guides(fill = "none", size = "none") +
#     make_theme(setFill = F, setCol = F, 
#     leg_pos = "bottom", leg_size = 8, guide_nrow = 30,
#     x_size = 10, y_size = 10, modify_guide = F)
#   ggsave("results/figures/11-figures/tax_func_plots/all_kos_by_genus_spec_circle_marked.pdf", width = 15, height = 20)
```