---
title: "Honeybee cross-species analysis - 11"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Initialization

To intialize complete all of or relevant sections of `scripts/visualization/Load_data.R`.



```{r}
df_genes_per_sample <- read.csv("results/figures/visualize_temp/genes_per_sample.tsv", sep = "\t") %>%
                          mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
                          mutate(location = Vectorize(get_location_from_sample_name)(sample)) %>%
                            filter(sample %in% samples)

ggplot(df_genes_per_sample) +
  geom_bar(aes(y = factor(sample, samples), x = genes, fill = factor(host)), stat = "identity") +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE, leg_pos = "bottom", 
             x_angle = 90, x_size = 0,
             x_hj = 10, x_vj = 1,
             y_size = 10, y_hj = 1)

ggplot(df_genes_per_sample) +
  geom_boxplot(aes(y = genes, x = host, fill = factor(host))) +
  geom_point(aes(y = genes, x = host, group = host,
                 shape = location, fill = factor(host)), 
              size = 4,
              position = position_jitterdodge(0.75)) +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE, leg_pos = "bottom", 
             x_angle = 90, x_size = 0,
             x_hj = 10, x_vj = 1,
             y_size = 10, y_hj = 1)
ggsave("results/figures/Number_of_genes_per_sample.pdf")
```



```{r}
df_cdhit_clusters_info <- read.csv("results/08_gene_content/00_cdhit_clustering/cluster_info.tsv", sep = "\t")
hist(df_cdhit_clusters_info %>% pull(num_genes))

ggplot(df_cdhit_clusters_info) +
  geom_histogram(aes(x = num_genes, fill = factor(num_hosts), group = num_hosts), binwidth = 1) +
  make_theme()

ggplot(df_cdhit_clusters_info %>% filter(num_genes > 50)) +
  geom_histogram(aes(x = num_genes, fill = factor(num_hosts), group = num_hosts), binwidth = 1) +
  make_theme()

ggplot(df_cdhit_clusters_info %>% filter(num_genes < 50)) +
  geom_histogram(aes(x = num_genes, fill = factor(num_hosts), group = num_hosts), binwidth = 1) +
  make_theme()

df_clusters_host_info <- df_cdhit_clusters_info %>%
  mutate(host_summarised = if_else(hosts == "M", "Apis mellifera", 
             if_else(hosts == "C", "Apis cerana", 
             if_else(hosts == "D", "Apis dorsata", 
             if_else(hosts == "F", "Apis florea", 
             if_else(hosts == "A", "Apis andreniformis", "Mixed"))))))
                          
ggplot(df_clusters_host_info) +
  geom_histogram(aes(x = num_genes, fill = factor(host_summarised), group = host_summarised), binwidth = 1) +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE)

ggplot(df_clusters_host_info %>% filter(representative_GH != "NA")) +
  geom_bar(aes(x = representative_GH, 
               fill = factor(host_summarised))) +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE)

# df_clusters_host_info order by count of GH
gh_order <- df_clusters_host_info %>%
  filter(representative_GH != "NA") %>%
  group_by(representative_GH) %>%
  mutate(count = n()) %>%
  unique() %>%
  arrange(count) %>%
  pull(representative_GH) %>% unique()

ko_order <- df_clusters_host_info %>%
  filter(representative_KO != "NA") %>%
  group_by(representative_KO) %>%
  mutate(count = n()) %>%
  unique() %>%
  arrange(count) %>%
  pull(representative_KO) %>% unique()

ggplot(df_clusters_host_info %>% 
        group_by(representative_GH) %>% 
        mutate(freq = n()) %>%
        filter(freq > 10) %>% 
        ungroup() %>%
        filter(representative_GH != "NA")) +
  labs(y = "CAZy GH family", x = "Count", fill = "Host") +
  geom_bar(aes(y = factor(representative_GH, gh_order), 
               fill = factor(host_summarised))) +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE)
  ggsave("results/figures/11-GH_family_clusters_per_host_filt_10.pdf")

# heatmap of host vs GH and group by hierarchy

# read in gene profiling results
df_profiling <- data.frame()
for (sample in samples) {
  flagstat_path = paste0("/scratch/aprasad/20230313_apis_species_comparison/results/08_gene_content/01_profiling/", sample, "_mapped.flagstat")
  if (file.exists(flagstat_path)) {
    reads_mapped = get_read_count_from_flagstat(flagstat_path)
  }
  total = df_reads %>% filter(Sample == sample) %>% pull(Total_clean)
  unmapped = total - reads_mapped
  df_profiling <- rbind(df_profiling, c(sample, unmapped, reads_mapped))
}
colnames(df_profiling) <- c("sample", "b_Mapped_catalog", "a_Unmapped_catalog")
df_profiling$b_Mapped_catalog <- as.numeric(df_profiling$b_Mapped_catalog)
df_profiling$a_Unmapped_catalog <- as.numeric(df_profiling$a_Unmapped_catalog)
ggplot(df_profiling %>% pivot_longer(!sample, names_to = "type", values_to = "reads")) +
  geom_bar(aes(y = sample, 
               x = reads, 
              fill = type), stat = "identity", position = "stack"
              ) +
    make_theme(y_size = 5)
# Pcoa of KO families

df_ko_counts <- read.csv("/scratch/aprasad/20230313_apis_species_comparison/results/08_gene_content/03_gene_counts/ko_family_genes_matrix.txt", sep = "\t", stringsAsFactors = FALSE, na.strings=0)
df_gh_counts <- read.csv("/scratch/aprasad/20230313_apis_species_comparison/results/08_gene_content/03_gene_counts/GH_family_genes_matrix.txt", sep = "\t", stringsAsFactors = FALSE, na.strings=0)
colnames(df_ko_counts) = c("KO_group", unlist(lapply(colnames(df_ko_counts)[-1], function(x) convert_sample_name(x))))
colnames(df_gh_counts) = c("GH_group", unlist(lapply(colnames(df_gh_counts)[-1], function(x) convert_sample_name(x))))
ko_mat <- data.matrix(df_ko_counts)[,-1]
rownames(ko_mat) <- data.matrix(df_ko_counts)[,1]
gh_mat <- data.matrix(df_gh_counts)[,-1]
rownames(gh_mat) <- data.matrix(df_gh_counts)[,1]

heatmap(ko_mat)
heatmap(gh_mat)
distance_mat <- dist(t(ko_mat[, ]))
pcoa_plot(distance_mat, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark)

distance_mat_gh <- dist(t(gh_mat[, which(colSums(gh_mat) > 1)]))
pcoa_plot(distance_mat_gh, df_meta_complete, "Species", color_add = T, color_list = host_order_color_dark)

view(distance_mat)
view(df_meta_complete)

all_gene_counts <- read.csv("/scratch/aprasad/20230313_apis_species_comparison/results/08_gene_content/03_gene_counts/all_genes_matrix.txt", sep = "\t", stringsAsFactors = FALSE, na.strings=0)
colnames(all_gene_counts) = c("gene_ID", unlist(lapply(colnames(df_ko_counts)[-1], function(x) convert_sample_name(x))))
```

```{r gggenes}
library(gggenes)

```


```{r}
# visualize cazy

cazy_df <- read.csv("results/figures/cazy_df.tsv", sep = "\t") %>%
            filter(sample %in% samples) %>%
            mutate(Host = Vectorize(get_host_from_sample_name)(sample))
host_count <- cazy_df %>% group_by(Host) %>% summarize(count = n_distinct(sample))
get_prev <- function(present, host) {
  total <- pull(filter(host_count, Host == host), count)
  return(present / total)
}

cazy_df <- read.csv("results/figures/cazy_df.tsv", sep = "\t") %>%
            mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
              mutate(present = ifelse(uniq_rpkm > 0, 1, 0)) %>%
              group_by(feature, Host) %>%
              mutate(count = sum(present)) %>%
              ungroup() %>%
              mutate(Prevalence_host = Vectorize(get_prev)(count, Host))

# count is number of samples of host in which it is present
# present if uniq_rpkm > 0
order_feature_count <- cazy_df %>%
            group_by(feature) %>%
            mutate(sum = sum(count)) %>%
            arrange(sum) %>%
            pull(feature) %>% unique

order_feature_prev <- cazy_df %>% 
            reframe(feature, Host, present, Prevalence_host) %>%
            group_by(feature) %>%
            mutate(sum = sum(Prevalence_host)) %>%
            arrange(sum) %>%
            pull(feature) %>% unique

ggplot(cazy_df %>% reframe(feature, Host, count, Prevalence_host) %>% unique) +
  geom_tile(aes(x = Host, y = feature, fill = Prevalence_host)) +
  labs(y = "CAZy GH family", x = "Host", fill = "Prevalence") +
  scale_y_discrete(limits = order_feature_prev) +
  coord_fixed(ratio = 0.05) +
  make_theme(setFill = FALSE, leg_pos = "right", 
             x_angle = 45, x_size = 10,
             y_size = 6, y_hj = 1)


# completed_substrate_annotations <- read_xlsx("data/cayman_gene_db/20230607_glycan_annotations_cleaned.xlsx")
completed_substrate_annotations <- read.csv("data/cayman_gene_db/20230607_glycan_annotations_cleaned_manually.tsv", header = T, sep = "\t")
glycan_annotations_final_cleaned <- completed_substrate_annotations %>% select(c(Family:Subfamily,ORIGIN:FUNCTION_AT_DESTINATION_3, Glycan_annotation))
glycan_annotations_final_cleaned_long <- glycan_annotations_final_cleaned %>% separate_rows(ORIGIN,sep=",") %>% separate_rows(FUNCTION_IN_ORIGIN,sep=",") %>% separate_rows(FUNCTION_AT_DESTINATION_1,sep=",") %>% separate_rows(FUNCTION_AT_DESTINATION_2,sep=",") %>% separate_rows(FUNCTION_AT_DESTINATION_3,sep=",")
glycan_annotations_cleaned <- glycan_annotations_final_cleaned

glycan_annotations_cleaned_filt <- glycan_annotations_cleaned %>% filter(!is.na(FUNCTION_AT_DESTINATION_3))

# and abbreviate long glycan names
cazy_df_annotated <- cazy_df %>%
                      left_join(glycan_annotations_cleaned, by = c("feature" = "Subfamily")) %>%
                      mutate(mod_names_pec = ifelse(grepl("pectin", FUNCTION_AT_DESTINATION_3), 1, 0)) %>%
                      mutate(short_names = abbreviate(FUNCTION_AT_DESTINATION_3, minlength = 50))

write.csv(cazy_df_annotated, "results/figures/cazy_df_annotated.txt", row.names = FALSE, quote = FALSE)

# heatmap of Glycan_annotatoin vs host showing prevalence as size of dot
ggplot(cazy_df_annotated %>% filter(sample %in% samples_IN_MY) %>% filter(uniq_rpkm > 0) %>% filter(grepl("ectin", FUNCTION_AT_DESTINATION_3))) +
  geom_tile(aes(x = sample, y = feature, fill = uniq_rpkm)) +
  labs(y = "Cazyme family", x = "sample", fill = "uniq_rpkm") +
  scale_fill_viridis_c(trans = "log10") +
  new_scale_fill() +
  geom_tile(aes(x = sample, y = 0, fill = Host)) +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE, leg_pos = "right", guide_nrow = 10, modify_guide = F,
             x_angle = 90, x_size = 0,
             x_hj = 10, x_vj = 1,
             y_size = 10, y_hj = 1)
             ggsave("results/figures/cazyme_uniq_rpkm_pectin_assoc_familes.pdf")

ggplot(cazy_df_annotated %>% filter(sample %in% samples_IN_MY) %>% filter(uniq_rpkm > 0)) +
# ggplot(cazy_df_annotated %>% filter(sample %in% samples_IN_MY) %>% filter(FUNCTION_AT_DESTINATION_3 != "Unknown") %>% filter(!is.na(FUNCTION_AT_DESTINATION_3))) +
  geom_tile(aes(x = sample, y = short_names, fill = uniq_rpkm)) +
  labs(y = "Cazyme family", x = "sample", fill = "uniq_rpkm") +
  scale_fill_viridis_c(trans = "log10") +
  new_scale_fill() +
  geom_tile(aes(x = sample, y = 0, fill = Host)) +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = FALSE, leg_pos = "right", guide_nrow = 10, modify_guide = F,
             x_angle = 90, x_size = 0,
             x_hj = 10, x_vj = 1,
             y_size = 10, y_hj = 1)
             ggsave("results/figures/cazyme_uniq_rpkm_glyc_annotation.pdf")

# make this with complex heatmap and add glycan annotation

cazyme_uniq_mat_pec <- cazy_df_annotated %>%
                        select(sample, feature, uniq_rpkm) %>%
                        pivot_wider(names_from = sample, values_from = uniq_rpkm) %>%                        
                        column_to_rownames("feature") %>% as.matrix %>% 


column_ba <- HeatmapAnnotation(Host = colnames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
glycan_anno_names <- rownames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown")
row_la <- rowAnnotation(Glycan = glycan_anno_names,
                        col = list(Glycan = make_col_list(glycan_anno_names)
                        ), border = T, show_legend = c(F)
                    )
column_split <- colnames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.character
row_split <- rownames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown") %>% as.factor %>% as.character
col_fun = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9]))
heatmap_obj = Heatmap(cazyme_uniq_mat_pec,
        # col = c("#e1ebf4", "#081d58"),
        col = col_fun,
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # border = T,
        heatmap_legend_param = list(title = "RPKM (Unique)"),
        bottom_annotation = column_ba, 
        # right_annotation = row_ra,
        left_annotation = row_la,
        # top_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 7),
          cluster_columns = F,
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)
pdf("results/figures/cazyme_uniq_rpkm.pdf", width = 20, height = 25)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


# make_without unknown
cazy_df_annotated %>% select(feature, FUNCTION_AT_DESTINATION_3) %>% unique %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(FUNCTION_AT_DESTINATION_3 != "Unknown") %>% pull(FUNCTION_AT_DESTINATION_3) %>% unique
known_families <- cazy_df_annotated %>% select(feature, FUNCTION_AT_DESTINATION_3) %>% unique %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(FUNCTION_AT_DESTINATION_3 != "Unknown" & FUNCTION_AT_DESTINATION_3 != "") %>% pull(feature)
cazyme_uniq_mat_pec_no_na <- cazy_df_annotated %>%
                        filter(feature %in% known_families) %>%
                        select(sample, feature, uniq_rpkm) %>%
                        pivot_wider(names_from = sample, values_from = uniq_rpkm) %>%
                        column_to_rownames("feature") %>% as.matrix

column_ba <- HeatmapAnnotation(Host = colnames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
glycan_anno_names <- rownames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown")
row_la <- rowAnnotation(Glycan = glycan_anno_names,
                        col = list(Glycan = make_col_list(glycan_anno_names)
                        ), border = T, show_legend = c(F)
                    )
column_split <- colnames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.character
row_split <- rownames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown") %>% as.factor %>% as.character
col_fun = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9]))
heatmap_obj = Heatmap(cazyme_uniq_mat_pec_no_na,
        # col = c("#e1ebf4", "#081d58"),
        col = col_fun,
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # border = T,
        heatmap_legend_param = list(title = "RPKM (Unique)"),
        bottom_annotation = column_ba, 
        # right_annotation = row_ra,
        left_annotation = row_la,
        # top_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 7),
        # show_row_dend = F,
        # show_column_dend = F,
          cluster_columns = F,
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)
pdf("results/figures/cazyme_uniq_rpkm_no_unknown.pdf", width = 15, height = 25)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

cazy_df_annotated <- cazy_df %>%
                      left_join(glycan_annotations_cleaned, by = c("feature" = "Subfamily")) %>%
                      mutate(mod_names_pec = ifelse(grepl("pectin", FUNCTION_AT_DESTINATION_3), 1, 0)) %>%
                      mutate(short_names = abbreviate(FUNCTION_AT_DESTINATION_3, minlength = 50)) %>%
                      filter(sample %in% samples_IN_MY)

cazyme_uniq_mat_pec <- cazy_df_annotated %>%
                        select(sample, feature, uniq_rpkm) %>%
                        pivot_wider(names_from = sample, values_from = uniq_rpkm) %>%                        
                        column_to_rownames("feature") %>% as.matrix


column_ba <- HeatmapAnnotation(Host = colnames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
glycan_anno_names <- rownames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown")
row_la <- rowAnnotation(Glycan = glycan_anno_names,
                        col = list(Glycan = make_col_list(glycan_anno_names)
                        ), border = T, show_legend = c(F)
                    )
column_split <- colnames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.character
row_split <- rownames(cazyme_uniq_mat_pec) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown") %>% as.factor %>% as.character
col_fun = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9]))
heatmap_obj = Heatmap(cazyme_uniq_mat_pec,
        # col = c("#e1ebf4", "#081d58"),
        col = col_fun,
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # border = T,
        heatmap_legend_param = list(title = "RPKM (Unique)"),
        bottom_annotation = column_ba, 
        # right_annotation = row_ra,
        left_annotation = row_la,
        # top_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 7),
          cluster_columns = F,
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)
pdf("results/figures/cazyme_uniq_rpkm_IN_MY.pdf", width = 20, height = 25)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()


# make_without unknown
cazy_df_annotated %>% select(feature, FUNCTION_AT_DESTINATION_3) %>% unique %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(FUNCTION_AT_DESTINATION_3 != "Unknown") %>% pull(FUNCTION_AT_DESTINATION_3) %>% unique
known_families <- cazy_df_annotated %>% select(feature, FUNCTION_AT_DESTINATION_3) %>% unique %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(!is.na(FUNCTION_AT_DESTINATION_3)) %>% filter(FUNCTION_AT_DESTINATION_3 != "Unknown" & FUNCTION_AT_DESTINATION_3 != "") %>% pull(feature)
cazyme_uniq_mat_pec_no_na <- cazy_df_annotated %>%
                        filter(feature %in% known_families) %>%
                        select(sample, feature, uniq_rpkm) %>%
                        pivot_wider(names_from = sample, values_from = uniq_rpkm) %>%
                        column_to_rownames("feature") %>% as.matrix

column_ba <- HeatmapAnnotation(Host = colnames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
glycan_anno_names <- rownames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown")
row_la <- rowAnnotation(Glycan = glycan_anno_names,
                        col = list(Glycan = make_col_list(glycan_anno_names)
                        ), border = T, show_legend = c(F)
                    )
column_split <- colnames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor %>% as.character
row_split <- rownames(cazyme_uniq_mat_pec_no_na) %>% as.data.frame() %>% left_join(cazy_df_annotated %>% select(feature, short_names) %>% filter(!is.na(short_names)) %>% unique, by = c("." = "feature")) %>% pull(short_names) %>% replace_na(., "Unknown") %>% as.factor %>% as.character
col_fun = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9]))
heatmap_obj = Heatmap(cazyme_uniq_mat_pec_no_na,
        # col = c("#e1ebf4", "#081d58"),
        col = col_fun,
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # border = T,
        heatmap_legend_param = list(title = "RPKM (Unique)"),
        bottom_annotation = column_ba, 
        # right_annotation = row_ra,
        left_annotation = row_la,
        # top_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 7),
        # show_row_dend = F,
        # show_column_dend = F,
          cluster_columns = F,
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)
pdf("results/figures/cazyme_uniq_rpkm_no_unknown.pdf", width = 15, height = 25)
draw(heatmap_obj, merge_legend = TRUE)
dev.off()

```

```{r}
# do this later using python!!!!!!!!!!!!!!!!!
# read gz sample file
df_gene_counts <- data.frame()
for (sample in samples_IN_MY) {
  print(paste0("reading ", sample, " gene counts"))
  # df_gene_counts_iter <- read.csv(gzfile(paste0("/scratch/aprasad/20230313_apis_species_comparison/results/08_gene_content/06_cayman/", sample, ".gene_counts.txt.gz")
  #                             ),
  #                             sep = "\t") %>%
  #                               select(gene, uniq_rpkm) %>%
  #                               mutate(uniq_rpkm = as.numeric(uniq_rpkm))
  df_gene_counts_iter <- read.csv(paste0("results/08_gene_content/01_profiling/", sample, "_mapped.coverage"), sep = "\t") %>%
                                select(`#rname`, meandepth) %>%
                                mutate(meandepth = as.numeric(meandepth))
  names(df_gene_counts_iter) <- c("gene", "depth")
  if (dim(df_gene_counts)[[1]] == 0) {
    df_gene_counts <- df_gene_counts_iter 
  } else {
    df_gene_counts <- left_join(df_gene_counts, df_gene_counts_iter, by = "gene")
  }
}
all_scaffolds_info <- read.csv("results/09_MAGs_collection/all_scaffold_to_bin.tsv", sep = "\t")

get_scaffold_from_gene <- function(gene_name){
  # get everything but the stuff after the last underscore
  scaff <- paste(strsplit(gene_name, "_")[[1]][1:(length(strsplit(gene_name, "_")[[1]]) - 1)], collapse = "_")
  return(scaff)
}

get_index_from_gene <- function(gene_name){
  # get everything the num after the last underscore
  ind <- strsplit(gene_name, "_")[[1]][length(strsplit(gene_name, "_")[[1]])]
  return(ind)
}
parse_out_og_id <- function(og) {
  return(strsplit(og, "--")[[1]][2])
}
gene_cazyme_info <- read.csv("data/cayman_gene_db/20230313_gene_catalog_db.tsv", header = F, sep = "\t")
names(gene_cazyme_info) <- c("gene", "a", "b", "cazyme_family")
gene_cazyme_info <- gene_cazyme_info %>%
                      select(gene, cazyme_family) %>%
                        mutate(duplicated = duplicated(gene)) %>%
                          filter(duplicated == F & !grepl("_", cazyme_family))
og_gene_df <- read.csv("results/08_gene_content/07_OG_coreness/gene_og_coreness.tsv", sep = "\t") %>%
                mutate(og_id = Vectorize(parse_out_og_id)(og))
mag_species_info <- vis_magOTUs_df %>%
  left_join(handmade_species_names %>% select(cluster, MAG_species_name_final), by = c("magOTU" = "cluster"))
all_genes <- read.csv("results/figures/20230313_gene_catalog_list.txt", header = F, sep = "\t") %>%
                pull(V1)
# genes_info <- data.frame("gene" = df_gene_counts$gene) %>%
genes_info <- data.frame("gene" = all_genes) %>%
  mutate(scaffold = Vectorize(get_scaffold_from_gene)(gene)) %>%
  mutate(gene_ind = Vectorize(get_index_from_gene)(gene)) %>%
    left_join(all_scaffolds_info %>% select(scaffold, mag), by = "scaffold") %>%
      left_join(mag_species_info, by = c("mag" = "ID")) %>%
        left_join(gene_cazyme_info, by = "gene") %>%
          mutate(id_in_OG_analysis = paste0(mag,"_", gene_ind)) %>%
          left_join(og_gene_df, by = c("id_in_OG_analysis" = "gene"))
write.csv(genes_info, "results/figures/gene_info.csv", row.names = FALSE, quote = FALSE)
```

```{r}
# [1] "gene"                      "scaffold"                 
# [3] "gene_ind"                  "mag"                      
# [5] "Completeness"              "Contamination"            
# [7] "Size"                      "scaffolds"                
# [9] "contigs"                   "N50_scaffolds"            
# [11] "N50_contigs"               "Mean_scaffold_length"     
# [13] "Mean_contig_length"        "Longest_scaffold"         
# [15] "Longest_contig"            "genes"                    
# [17] "Size_readable"             "Sample"                   
# [19] "Type"                      "Quality"                  
# [21] "Domain"                    "Phylum"                   
# [23] "Class"                     "Order"                    
# [25] "Family"                    "Genus"                    
# [27] "Species"                   "Score"                    
# [29] "magOTU"                    "closest_cluster_member"   
# [31] "furthest_cluster_member.x" "Representative"           
# [33] "completeness"              "contamination"            
# [35] "length"                    "N50"                      
# [37] "Completeness_quality"      "Contamination_quality"    
# [39] "N50_quality"               "Host"                     
# [41] "Num_mags"                  "score_representative"     
# [43] "cluster_members"           "furthest_cluster_member.y"
# [45] "representative_genome"     "score"                    
# [47] "MAG_species_name_final"    "cazyme_family"            
# [49] "id_in_OG_analysis"         "og"                       
# [51] "coreness"                  "og_id"     
genes_of_interest <- genes_info %>%
                      filter(!is.na(mag)) %>%
                      filter(!is.na(cazyme_family)) %>%
                      filter(Genus == "g__Dysgonomonas") %>%
                      # filter(Genus == "g__Bombilactobacillus") %>%
                      pull(gene)
# replace NAs with 0s
matrix_to_plot <- df_gene_counts %>%
                    filter(gene %in% genes_of_interest) %>%
                      column_to_rownames("gene") %>%
                          as.matrix %>%
                            replace(is.na(.), 0)
dim(matrix_to_plot)
matrix_to_plot <- matrix_to_plot[apply(matrix_to_plot, 1, function(x) sum(x > 0, na.rm = T)) > 10,]
dim(matrix_to_plot)
genes_of_interest <- rownames(matrix_to_plot)
genus_names <- genes_of_interest %>%
                as.data.frame() %>%
                left_join(genes_info %>% select(gene, Genus), by = c("." = "gene")) %>%
                  pull(Genus)
species_names <- genes_of_interest %>%
                as.data.frame() %>%
                left_join(genes_info %>% select(gene, MAG_species_name_final), by = c("." = "gene")) %>%
                  pull(MAG_species_name_final)
cazy_names <- genes_of_interest %>%
                as.data.frame() %>%
                left_join(genes_info %>% select(gene, cazyme_family), by = c("." = "gene")) %>%
                  pull(cazyme_family)
coreness_names <- genes_of_interest %>%
                as.data.frame() %>%
                left_join(genes_info %>% select(gene, coreness), by = c("." = "gene")) %>%
                  pull(coreness)
species_names_to_mark <- unique(species_names)
species_indices <- sapply(species_names_to_mark, function(x) which(species_names == x)[1]) %>% as.numeric
species_names_to_mark <- species_names_to_mark[order(species_indices)]
anno_species_row <- rowAnnotation(Species = species_names,
                        col = list(Species = make_col_list(species_names)
                        ), border = T, show_legend = c(F)
                    )
anno_species_text_row <- rowAnnotation(Spec_labales = anno_mark(at = species_indices, side = "left",
                                                          labels = species_names_to_mark),
                                       Species = species_names,
                        col = list(Species = make_col_list(species_names)
                        ), border = T, show_legend = c(F)
                    )
anno_cazyme_row <- rowAnnotation(Cazyme = cazy_names,
                                 coreness = coreness_names,
                        col = list(Cazyme = make_col_list(cazy_names),
                                    coreness = make_col_list(coreness_names)
                        ), border = T, show_legend = c(F, T)
                    )
anno_coreness_row <- rowAnnotation(coreness = coreness_names,
                        col = list(coreness = make_col_list(coreness_names)
                        ), border = T
                    )
anno_genus_row <- rowAnnotation(Genus = genus_names,
                        col = list(Genus = make_col_list(genus_names)
                        ), border = T, show_legend = c(F)
                    )
anno_host_col <- HeatmapAnnotation(Host = colnames(matrix_to_plot) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color
                              ), border = T
                          )
row_split <- as.character(as.factor(cazy_names))
row_split <- as.character(as.factor(coreness_names))
# remove rows with all 0s and / or NAs
heatmap_obj = Heatmap(matrix_to_plot,
        # col = c("#e1ebf4", "#081d58"),
        col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        row_split = row_split,
        # border = T,
        heatmap_legend_param = list(title = "RPKM (Unique)"),
        left_annotation = anno_cazyme_row,
        right_annotation = anno_species_row,
        bottom_annotation = anno_host_col, 
        # top_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 0),
          cluster_columns = F,
          cluster_rows = F)
draw(heatmap_obj, merge_legend = TRUE)

gene_compiled <- df_gene_counts %>%
        pivot_longer(!gene, names_to = "sample", values_to = "uniq_rpkm") %>%
        left_join(genes_info, by = "gene")
head(gene_compiled)

```

```{r}
clusters_host <- read.csv("results/figures/visualize_temp/cluster_host_barplot.tsv", sep = "\t") %>%
                  mutate(variable = ifelse(variable == "M", "Apis mellifera", variable)) %>%
                  mutate(variable = ifelse(variable == "C", "Apis cerana", variable)) %>%
                  mutate(variable = ifelse(variable == "D", "Apis dorsata", variable)) %>%
                  mutate(variable = ifelse(variable == "F", "Apis florea", variable)) %>%
                  mutate(variable = ifelse(variable == "A", "Apis andreniformis", variable)) %>%
                  mutate(index = as.factor(index))
index_order <- clusters_host %>%
                  group_by(index) %>%
                  summarise(count = sum(value)) %>%
                  arrange(desc(count)) %>%
                  pull(index)
ggplot(clusters_host) +
  geom_bar(aes(y = factor(index, index_order), x = value, fill = variable), stat = "identity") +
  scale_fill_manual(values = host_order_color) +
  make_theme(setFill = F)

```

```{r}
df_num_genes <- read.csv("results/figures/visualize_temp/genes_per_sample.tsv", sep = "\t") %>%
                  mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(df_num_genes %>% filter(sample %in% samples)) +
  geom_bar(aes(y = factor(sample, samples),
               x = genes,
               fill = host
             ),
             stat = "identity"
  ) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "# Genes", y = "Sample", color = "Host") +
  make_theme(setFill = F)

ggplot(df_num_genes %>% filter(sample %in% samples)) +
  geom_boxplot(aes(x = factor(host, host_order),
               y = genes,
               fill = host
             ),
             outlier.shape = NA
  ) +
  geom_jitter(aes(x = factor(host, host_order),
               y = genes
             ),
             width = 0.1,
             color = "#000000"
  ) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "# Genes", color = "Host") +
  make_theme(setFill = F, x_hj = 1, x_vj = 1, x_angle = 30)
```

```{r}
df_num_clusters <- read.csv("results/figures/visualize_temp/clusters_per_sample.tsv", sep = "\t") %>%
                  mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(df_num_clusters %>% filter(sample %in% samples)) +
  geom_bar(aes(y = factor(sample, samples),
               x = clusters,
               fill = host
             ),
             stat = "identity"
  ) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "# Genes", y = "Sample", color = "Host") +
  make_theme(setFill = F)

ggplot(df_num_clusters %>% filter(sample %in% samples)) +
  geom_boxplot(aes(x = factor(host, host_order),
               y = clusters,
               fill = host
             ),
             outlier.shape = NA
  ) +
  geom_jitter(aes(x = factor(host, host_order),
               y = clusters
             ),
             width = 0.1,
             color = "#000000"
  ) +
  scale_fill_manual(values = host_order_color_dark) +
  labs(x = "Host", y = "# Genes", color = "Host") +
  make_theme(setFill = F, x_hj = 1, x_vj = 1, x_angle = 30)
```

```{r}
df_gene_clust <- df_num_genes %>%
                    left_join(df_num_clusters %>% select(sample, clusters)) %>%
                      pivot_longer(!c(sample, host), names_to = "type", values_to = "num")
ggplot(df_gene_clust %>% filter(sample %in% samples)) +
  geom_bar(aes(y = factor(sample, samples),
               x = num,
               fill = type
             ),
             stat = "identity", position = "dodge"
  ) +
  facet_wrap(~host, scales = "free_y") +
  labs(x = "# Genes", y = "Sample", color = "Type") +
  make_theme(setFill = F)
```

```{r}
# plot cumulative curves

df_cum_data <- read.csv("results/figures/visualize_temp/cluster_cum_curve.tsv", sep = "\t")
ggplot() +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis mellifera"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis mellifera"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis cerana"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis cerana"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis dorsata"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis dorsata"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis florea"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis florea"), aes(x =ize, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis andreniformis"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis andreniformis"), aes(x = size, y = genes, color = host)) +
  scale_color_manual(values = host_order_color_dark) +
  labs(x = "# Bees", y = "# Clusters", color = "Host") +
  scale_y_continuous(labels = label_number(suffix = " K", scale = 1e-3)) +
  make_theme(setFill = F, setCol = F, guide_nrow = 1)
```

```{r}
df_num_clusters <- read.csv("results/figures/visualize_temp/clusters_per_sample.tsv", sep = "\t") %>%
                  mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(df_num_clusters %>% filter(sample %in% samples)) +
  geom_bar(aes(y = factor(sample, samples),
               x = clusters,
               fill = host
             ),
             stat = "identity"
  ) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = F)
ggplot(df_num_clusters %>% filter(sample %in% samples)) +
  geom_boxplot(aes(x = factor(host, host_order),
               y = clusters,
               fill = host
             ),
             outlier.shape = NA
  ) +
  geom_jitter(aes(x = factor(host, host_order),
               y = clusters
             ),
             width = 0.1,
             color = "#000000"
  ) +
  scale_fill_manual(values = host_order_color_dark)
  make_theme(setFill = F)
```

```{r}
# plot cumulative curves
df_cum_data <- read.csv("results/figures/visualize_temp/clusters_cum_curve.tsv", sep = "\t")

ggplot() +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis mellifera"), aes(x = genes, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis mellifera"), aes(x = genes, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis cerana"), aes(x = genes, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis cerana"), aes(x = genes, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis dorsata"), aes(x = genes, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis dorsata"), aes(x = genes, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis florea"), aes(x = genes, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis florea"), aes(x = genes, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis andreniformis"), aes(x = genes, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis andreniformis"), aes(x = genes, y = genes, color = host)) +
  scale_color_manual(values = host_order_color) +
  scale_y_continuous(trans = "log10") +
  facet_wrap(~host) +
  make_theme(setFill = F, setCol = F)

df_cum_data <- read.csv("results/figures/visualize_temp/genes_cum_curve_extended.tsv", sep = "\t")

ggplot() +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis mellifera"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis mellifera"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis cerana"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis cerana"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis dorsata"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis dorsata"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis florea"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis florea"), aes(x = size, y = genes, color = host)) +
  geom_smooth(data = df_cum_data %>% filter(host == "Apis andreniformis"), aes(x = size, y = genes, color = host), method = "loess", se = T) +
  geom_point(data = df_cum_data %>% filter(host == "Apis andreniformis"), aes(x = size, y = genes, color = host)) +
  scale_color_manual(values = host_order_color_dark) +
  # scale_y_continuous(trans = "log10") +
  # facet_wrap(~host) +
  make_theme(setFill = F, setCol = F)

```

```{r}
df_ko_counts <- read.csv("results/figures/visualize_temp/kos_per_sample.tsv", sep = "\t") %>%
                  pivot_longer(!sample, names_to = "type", values_to = "num") %>%
                  mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(df_ko_counts %>% filter(sample %in% samples)) +
  geom_bar(aes(y = factor(sample, samples),
               x = num,
               fill = type
             ),
             stat = "identity", position = "stack"
  ) +
  facet_wrap(~host, scales = "free_y") +
  labs(x = "# Genes", y = "Sample", color = "Type") +
  scale_fill_manual(labels = c("kos" = "annotated"), 
                    values = c("kos" = brewer.pal(11, 'Spectral')[4],
                               "unannotated" = brewer.pal(11, 'Spectral')[1]
                    )
  ) +
  make_theme(setFill = F)

ggplot(df_ko_counts %>% filter(sample %in% samples)) +
  geom_boxplot(aes(y = num,
               x = host,
               fill = type
             )
  ) +
  # geom_jitter(aes(y = num,
  #              x = host,
  #              color = type
  #            )
  # ) +
  labs(x = "Host", y = "# Genes", color = "Type") +
  scale_fill_manual(labels = c("kos" = "annotated"), 
                    values = c("kos" = brewer.pal(11, 'Spectral')[4],
                               "unannotated" = brewer.pal(11, 'Spectral')[1]
                    )
  ) +
  scale_color_manual(labels = c("kos" = "annotated"), 
                    values = c("kos" = brewer.pal(11, 'Spectral')[4],
                               "unannotated" = brewer.pal(11, 'Spectral')[1]
                    )
  ) +
  make_theme(setFill = F, setCol = F)

df_kos_per_sample <- read.csv("results/figures/visualize_temp/kos_per_sample.tsv", sep = "\t") %>%
                      mutate(host = Vectorize(get_host_from_sample_name)(sample))
ggplot(df_kos_per_sample %>% filter(sample %in% samples)) +
  geom_bar(aes(y = factor(sample, samples),
               x = kos,
               fill = host
             ),
             stat = "identity"
  ) +
  scale_fill_manual(values = host_order_color_dark) +
  make_theme(setFill = F)
ggplot(df_kos_per_sample %>% filter(sample %in% samples)) +
  geom_boxplot(aes(x = factor(host, host_order),
               y = kos,
               fill = host
             ),
             outlier.shape = NA
  ) +
  geom_jitter(aes(x = factor(host, host_order),
               y = kos
             ),
             width = 0.1,
             color = "#000000"
  ) +
  scale_fill_manual(values = host_order_color_dark)
  make_theme(setFill = F)
```

```{r}
df_ko_cum_data <- read.csv("results/figures/visualize_temp/KO_cum_curve.tsv", sep = "\t")
ggplot() +
  geom_smooth(data = df_ko_cum_data %>% filter(host == "Apis mellifera"), aes(x = size, y = kos, color = host), method = "loess", se = T) +
  geom_point(data = df_ko_cum_data %>% filter(host == "Apis mellifera"), aes(x = size, y = kos, color = host)) +
  geom_smooth(data = df_ko_cum_data %>% filter(host == "Apis cerana"), aes(x = size, y = kos, color = host), method = "loess", se = T) +
  geom_point(data = df_ko_cum_data %>% filter(host == "Apis cerana"), aes(x = size, y = kos, color = host)) +
  geom_smooth(data = df_ko_cum_data %>% filter(host == "Apis dorsata"), aes(x = size, y = kos, color = host), method = "loess", se = T) +
  geom_point(data = df_ko_cum_data %>% filter(host == "Apis dorsata"), aes(x = size, y = kos, color = host)) +
  geom_smooth(data = df_ko_cum_data %>% filter(host == "Apis florea"), aes(x = size, y = kos, color = host), method = "loess", se = T) +
  geom_point(data = df_ko_cum_data %>% filter(host == "Apis florea"), aes(x = size, y = kos, color = host)) +
  geom_smooth(data = df_ko_cum_data %>% filter(host == "Apis andreniformis"), aes(x = size, y = kos, color = host), method = "loess", se = T) +
  geom_point(data = df_ko_cum_data %>% filter(host == "Apis andreniformis"), aes(x = size, y = kos, color = host)) +
  scale_color_manual(values = host_order_color_dark) +
  labs(x = "# Bees", y = "# KOs", color = "Host") +
  scale_y_continuous(labels = label_number(suffix = " K", scale = 1e-3)) +
  make_theme(setFill = F, setCol = F, guide_nrow = 1)
```

```{r}
df_ko_cat_B_data <- read.csv("results/figures/visualize_temp/kos_B_per_sample.tsv", sep = "\t") %>%
                      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
                      filter(sample %in% samples)
ggplot(df_ko_cat_B_data) +
  geom_bar(aes(x = sample,
               y = count,
               fill = kos
  ), color = "#000000", linewidth = 0.1,
  stat = "identity", position = "stack"
  ) +
  make_theme(max_colors = length(unique(df_ko_cat_B_data$kos)))

df_ko_cat_B_data_sub <- df_ko_cat_B_data %>%
        filter(!is.na(kos)) %>% 
        filter(kos != "Poorly characterized") %>%
        mutate(kos = ifelse(grepl("genetic information processing", kos), "genetic information processing", kos)) %>%
        mutate(kos = ifelse(grepl("signaling and cellular processes", kos), "genetic information processing", kos)) %>%
        mutate(kos = ifelse(grepl("etabolism", kos), "Metabolism", kos))
ggplot(df_ko_cat_B_data_sub) +
  geom_bar(aes(x = sample,
               y = count,
               fill = kos
  ), color = "#000000", linewidth = 0.1,
  stat = "identity", position = "stack"
  ) +
  facet_wrap(~host, scales = "free_x") +
  make_theme(leg_pos = 'right', guide_nrow = 40,
      max_colors = length(unique(df_ko_cat_B_data_sub$kos))
  )
df_ko_cat_B_data_sub <- df_ko_cat_B_data_sub %>%
                          group_by(kos) %>%
                          mutate(tot_count = sum(count)) %>%
                          filter(tot_count > 5) %>%
                          select(!tot_count) %>%
                          ungroup()
ggplot(df_ko_cat_B_data_sub) +
  geom_tile(aes(y = sample,
               fill = count,
               x = kos
  )) +
  scale_fill_gradient(trans = "log") +
  make_theme(setCol = F, setFill = F,
      leg_pos = 'right', guide_nrow = 40,
      modify_guide = F,
      x_angle = 30, x_hj = 1, x_vj = 1,
      max_colors = length(unique(df_ko_cat_B_data_sub$kos))
  )

df_mat_ko_B <- read.csv("results/figures/visualize_temp/kos_B_per_sample.tsv", sep = "\t", stringsAsFactors = F) %>%
              filter(sample %in% samples) %>%
              mutate(count = as.numeric(count)) %>%
              pivot_wider(id_cols = sample, names_from = kos, values_from = count)
```

```{r}

ko_info_table <- read.csv("results/figures/08-summarize_functions/kegg_info_dict.txt", sep = "\t")

df_ko_cat_C_data <- read.csv("results/figures/visualize_temp/kos_C_per_sample.tsv", sep = "\t") %>%
                      mutate(host = Vectorize(get_host_from_sample_name)(sample)) %>%
                    filter(sample %in% samples)
df_ko_cat_C_data_sub <- df_ko_cat_C_data %>%
        filter(!is.na(kos)) %>% 
        filter(kos != "Poorly characterized")
df_ko_cat_C_data_sub <- df_ko_cat_C_data_sub %>%
                          group_by(kos) %>%
                          mutate(tot_count = sum(count)) %>%
                          filter(tot_count > 10) %>%
                          select(!tot_count) %>%
                          ungroup()
ggplot(df_ko_cat_C_data_sub) +
  geom_tile(aes(y = sample,
               fill = count,
               x = kos
  )) +
  scale_fill_gradient(trans = "log") +
  make_theme(setCol = F, setFill = F,
      leg_pos = 'right', guide_nrow = 40,
      modify_guide = F,
      x_angle = 30, x_hj = 1, x_vj = 1,
      max_colors = length(unique(df_ko_cat_C_data_sub$kos))
  )

df_mat_ko_C <- read.csv("results/figures/visualize_temp/kos_C_per_sample.tsv", sep = "\t", stringsAsFactors = F) %>%
              filter(sample %in% samples) %>%
              mutate(count = as.numeric(count)) %>%
              pivot_wider(id_cols = sample, names_from = kos, values_from = count)
```

```{r}
ko_df <- read.csv("results/figures/08-summarize_functions/feature_matrices/ko_feature_matrix.tsv")
# make a heatmap of the KOs
```

```{r}
all_sample_genes_df <- data.frame()
for (sample in samples) {
  print(paste0("reading ", sample, " gene info"))
  sample_info_path <- paste0("results/figures/08-summarize_functions/gene_info_tables/", sample, "_df_detected_genes_info.csv")
  if (file.exists(sample_info_path)) {
    sample_genes_df_iter <- read.csv(sample_info_path) %>%
                                mutate(sample = sample)
    if (dim(all_sample_genes_df)[[1]] == 0) {
      all_sample_genes_df <- sample_genes_df_iter 
    } else {
      all_sample_genes_df <- rbind(all_sample_genes_df, sample_genes_df_iter)
    }
  }
}
split_og_name <- function(og_name) {
  return(strsplit(og_name, "--")[[1]][2])
}
all_sample_genes_df <- all_sample_genes_df %>%
                        mutate(og_id = Vectorize(split_og_name)(og))
write.csv(all_sample_genes_df, "results/figures/all_sample_genes_df.csv", row.names = FALSE, quote = FALSE)


ko_mat <- read.csv("results/figures/08-summarize_functions/ko_matrix.csv") %>%
            filter(sample %in% samples) %>%
            column_to_rownames("sample")
# plot pcoa of ko matrix
pcoa_plot(vegdist(ko_mat, "robust.aitchison"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(ko_mat, "robust.aitchison"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)



ko_mat <- all_sample_genes_df %>%
            filter(sample %in% samples) %>%
            # filter(species == "Bombilactobacillus mellis") %>%
            filter(grepl("Bombilactobacillus", species)) %>%
            filter(!is.na(ko)) %>%
            filter(ko != "") %>%
            select(sample, ko) %>%
            group_by(sample, ko) %>%
            summarise(count = n()) %>%
            pivot_wider(id_cols = sample, names_from = ko, values_from = count) %>%
            # replace NAs with 0s
            column_to_rownames("sample") %>%
            as.matrix %>%
            replace(is.na(.), 0)
# convert to a presence-absence numeric matrix of 1 and 0
ko_mat <- ko_mat > 0
ko_mat <- ko_mat * 1
anno_host_row <- rowAnnotation(Host = rownames(ko_mat) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
Heatmap(ko_mat,
        col = c("#e1ebf4", "#081d58"),
        # col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # row_split = row_split,
        # border = T,
        left_annotation = anno_host_row,    
        column_names_gp = grid::gpar(fontsize = 8),
        row_names_gp = grid::gpar(fontsize = 8),
          cluster_columns = T,
          cluster_rows = T)
# plot pcoa of ko matrix
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)


og_mat <- all_sample_genes_df %>%
            filter(sample %in% samples) %>%
            filter(species == "Bombilactobacillus mellifer") %>%
            # filter(grepl("Bombilactobacillus", species)) %>%
            filter(!is.na(og)) %>%
            filter(og != "") %>%
            select(sample, og) %>%
            group_by(sample, og) %>%
            summarise(count = n()) %>%
            pivot_wider(id_cols = sample, names_from = og, values_from = count) %>%
            # replace NAs with 0s
            column_to_rownames("sample") %>%
            as.matrix %>%
            replace(is.na(.), 0)
og_mat <- og_mat > 0
og_mat <- og_mat * 1
# remove columns with only 0s
og_mat <- og_mat[,apply(og_mat, 2, function(x) sum(x > 0, na.rm = T)) > 5]
anno_host_row <- rowAnnotation(Host = rownames(og_mat) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
Heatmap(og_mat,
        col = c("#ffffff", "#081d58"),
        # col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # row_split = row_split,
        # border = T,
        left_annotation = anno_host_row,    
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 8),
          cluster_columns = T,
          cluster_rows = T)
# plot pcoa of og matrix
pcoa_plot(vegdist(og_mat, "jaccard"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(og_mat, "jaccard"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)
pcoa_plot(beta.pair(og_mat, "sorensen")$beta.sim, df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)

ko_mat <- all_sample_genes_df %>%
            filter(sample %in% samples) %>%
            # filter(species == "Bombilactobacillus mellifer") %>%
            filter(grepl("Bombilactobacillus", species)) %>%
            filter(!is.na(ko)) %>%
            filter(ko != "") %>%
            select(sample, ko) %>%
            group_by(sample, ko) %>%
            summarise(count = n()) %>%
            pivot_wider(id_cols = sample, names_from = ko, values_from = count) %>%
            # replace NAs with 0s
            column_to_rownames("sample") %>%
            as.matrix %>%
            replace(is.na(.), 0)
ko_mat <- ko_mat > 0
ko_mat <- ko_mat * 1
# remove columns with only 0s
ko_mat <- ko_mat[,apply(ko_mat, 2, function(x) sum(x > 0, na.rm = T)) > 5]
anno_host_row <- rowAnnotation(Host = rownames(ko_mat) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
Heatmap(ko_mat,
        col = c("#ffffff", "#081d58"),
        # col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # row_split = row_split,
        # border = T,
        left_annotation = anno_host_row,    
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 8),
          cluster_columns = T,
          cluster_rows = T)
# plot pcoa of ko matrix
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)
pcoa_plot(beta.pair(ko_mat, "sorensen")$beta.sim, df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)


ko_mat <- all_sample_genes_df %>%
            filter(sample %in% samples) %>%
            filter(grepl("Gilliamella", species)) %>%
            filter(!is.na(ko)) %>%
            filter(ko != "") %>%
            select(sample, ko) %>%
            group_by(sample, ko) %>%
            summarise(count = n()) %>%
            pivot_wider(id_cols = sample, names_from = ko, values_from = count) %>%
            # replace NAs with 0s
            column_to_rownames("sample") %>%
            as.matrix %>%
            replace(is.na(.), 0)
ko_mat <- ko_mat > 0
ko_mat <- ko_mat * 1

# remove columns with only 0s
# ko_mat <- ko_mat[,apply(ko_mat, 2, function(x) sum(x > 0, na.rm = T)) > 5]
anno_host_row <- rowAnnotation(Host = rownames(ko_mat) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
categories <- colnames(ko_mat) %>% as.data.frame() %>% left_join(ko_info_table %>% select(ko, B), by = c("." = "ko")) %>% pull(B)
anno_ko_column <- HeatmapAnnotation(Category = categories,
                              col = list(Category = make_col_list(categories)
                              ), border = T
                          
                          )
Heatmap(ko_mat,
        col = c("#ffffff", "#081d58"),
        # col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # row_split = row_split,
        column_split = categories %>% as.factor,
        # border = T,
        left_annotation = anno_host_row,    
        top_annotation = anno_ko_column,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 8),
          cluster_columns = T,
          cluster_rows = T)
# plot pcoa of ko matrix
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)
pcoa_plot(beta.pair(ko_mat, "sorensen")$beta.sim, df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)


ko_mat <- all_sample_genes_df %>%
            filter(sample %in% samples) %>%
            filter(grepl("Frischella", species)) %>%
            filter(!is.na(ko)) %>%
            filter(ko != "") %>%
            select(sample, ko) %>%
            group_by(sample, ko) %>%
            summarise(count = n()) %>%
            pivot_wider(id_cols = sample, names_from = ko, values_from = count) %>%
            # replace NAs with 0s
            column_to_rownames("sample") %>%
            as.matrix %>%
            replace(is.na(.), 0)
ko_mat <- ko_mat > 0
ko_mat <- ko_mat * 1
# remove columns with only 0s
# ko_mat <- ko_mat[,apply(ko_mat, 2, function(x) sum(x > 0, na.rm = T)) > 5]
anno_host_row <- rowAnnotation(Host = rownames(ko_mat) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
categories <- colnames(ko_mat) %>% as.data.frame() %>% left_join(ko_info_table %>% select(ko, B), by = c("." = "ko")) %>% pull(B)
anno_ko_column <- HeatmapAnnotation(Category = categories,
                              col = list(Category = make_col_list(categories)
                              ), border = T
                          
                          )
Heatmap(ko_mat,
        col = c("#ffffff", "#081d58"),
        # col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # row_split = row_split,
        column_split = categories %>% as.factor,
        # border = T,
        left_annotation = anno_host_row,    
        top_annotation = anno_ko_column,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 8),
          cluster_columns = T,
          cluster_rows = T)
# plot pcoa of ko matrix
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)
pcoa_plot(beta.pair(ko_mat, "sorensen")$beta.sim, df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)


ko_mat <- all_sample_genes_df %>%
            filter(sample %in% samples) %>%
            filter(grepl("Snodgrassella", species)) %>%
            filter(!is.na(ko)) %>%
            filter(ko != "") %>%
            select(sample, ko) %>%
            group_by(sample, ko) %>%
            summarise(count = n()) %>%
            pivot_wider(id_cols = sample, names_from = ko, values_from = count) %>%
            # replace NAs with 0s
            column_to_rownames("sample") %>%
            as.matrix %>%
            replace(is.na(.), 0)
ko_mat <- ko_mat > 0
ko_mat <- ko_mat * 1
# remove columns with only 0s
# ko_mat <- ko_mat[,apply(ko_mat, 2, function(x) sum(x > 0, na.rm = T)) > 5]
anno_host_row <- rowAnnotation(Host = rownames(ko_mat) %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
categories <- colnames(ko_mat) %>% as.data.frame() %>% left_join(ko_info_table %>% select(ko, B), by = c("." = "ko")) %>% pull(B)
anno_ko_column <- HeatmapAnnotation(Category = categories,
                              col = list(Category = make_col_list(categories)
                              ), border = T
                          
                          )
Heatmap(ko_mat,
        col = c("#ffffff", "#081d58"),
        # col = colorRamp2(c(0, 0.1, 1, 10, 1000), c(brewer.pal(9, "PuBu")[1], brewer.pal(9, "PuBu")[3], brewer.pal(9, "PuBu")[4], brewer.pal(9, "PuBu")[7], brewer.pal(9, "PuBu")[9])),
        clustering_method_columns = 'ward.D2',
        column_dend_side = "bottom",
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # row_split = row_split,
        column_split = categories %>% as.factor,
        # border = T,
        left_annotation = anno_host_row,    
        top_annotation = anno_ko_column,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 8),
          cluster_columns = T,
          cluster_rows = T)
# plot pcoa of ko matrix
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, color_list=host_order_color_dark)
pcoa_plot(vegdist(ko_mat, "jaccard"), df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)
pcoa_plot(beta.pair(ko_mat, "sorensen")$beta.sim, df_meta_complete, "Species", color_add=T, shape_add=T, shape_var= "Country", color_list=host_order_color_dark)



```

```{r}

# # devtools::install_github("noriakis/ggkegg", dependencies=F)
# library(ggkegg)
# library(ggfx)
# library(igraph)
# library(tidygraph)
# library(dplyr)


# pathway("ko00071") |>
# pathway("ko00020") |>
# pathway("ko01100") |>
#     process_line() |>
#     # highlight_module(module("M00021")) |>
#     # highlight_module(module("M00338")) |>
#     ggraph(x=x, y=y) +
#         geom_node_point(size=1, aes(color=I(fgcolor),
#             filter=fgcolor!="none" & type!="line")) +
#         geom_edge_link0(width=0.1, aes(color=I(fgcolor),
#             filter=type=="line"& fgcolor!="none")) +
#         # with_outer_glow(
#         #     geom_edge_link0(width=1,
#         #         aes(color=I(fgcolor),
#         #             filter=(M00021 | M00338))),
#         #     colour="red", expand=5
#         # ) +
#         # with_outer_glow(
#         #     geom_node_point(size=1.5,
#         #         aes(color=I(fgcolor),
#         #             filter=(M00021 | M00338))),
#         #     colour="red", expand=5
#         # ) +
#         geom_node_text(size=2,
#             aes(x=x, y=y,
#                 label=graphics_name,
#                 filter=name=="path:ko00270"),
#             repel=TRUE, family="sans", bg.colour="white") +
#         theme_void()

# export_kos <- all_sample_genes_df %>%
#             filter(sample %in% c("C3-2")) %>%
#             filter(grepl("Snodgrassella", species)) %>%
#             filter(!is.na(ko)) %>%
#             filter(ko != "") %>%
#             pull(ko)
#             # select(gene, ko)


# pathway("ko01100") |>
#     process_line() |>
#     ggraph(x=x, y=y) +
#         geom_node_point(size=1, aes(color=I(fgcolor),
#             filter=fgcolor!="none" & type!="line")) +
#         geom_edge_link0(width=0.1, aes(color=I(fgcolor),
#             filter=type=="line"& fgcolor!="none")) +
#         geom_node_text(size=12,
#             aes(x=x, y=y,
#                 label=graphics_name,
#                 filter=name=="path:00020"),
#             repel=TRUE, family="sans", bg.colour="white") +
#         theme_void()

# # K04343 - AMR in melli snod

# write.table(export_kos, "results/figures/visualize_temp/KOs.csv", row.names = FALSE, quote = FALSE)
```

```{r}
df_by_sample <- read.csv("results/figures/visualize_temp/microbeannotator_out/by_sample_module_completeness.tab", check.names = F, sep = "\t") %>%
                  mutate(name = abbreviate(name, 55))
format_colnames <- function(list_cols) {
  sapply(list_cols, function(x) if (x %in% c("name", "module", "pathway group")) {return(x)} else {return(strsplit(x, "_kos.csv")[[1]][1])})
}
colnames(df_by_sample) <- df_by_sample %>% colnames %>% format_colnames
by_sample_mat <- df_by_sample %>% 
  select(!c(module, `pathway group`), ) %>%
  column_to_rownames("name") %>% as.matrix
# remove sparse rows
by_sample_mat <- by_sample_mat[apply(by_sample_mat, 1, function(x) sum(x > 0, na.rm = T)) > 10,]
anno_top_host <- HeatmapAnnotation(Host = by_sample_mat %>% colnames %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                              col = list(Host = host_order_color_dark
                              ), border = T
                          )
anno_left_module <- rowAnnotation(Group = by_sample_mat %>% rownames %>% as.data.frame() %>% left_join(df_by_sample %>% select(name, `pathway group`), by = c("." = "name")) %>% pull(`pathway group`),
                              col = list(Group = make_col_list(by_sample_mat %>% rownames %>% as.data.frame() %>% left_join(df_by_sample %>% select(name, `pathway group`), by = c("." = "name")) %>% pull(`pathway group`), "Set3")
                              ), border = T,
                              show_legend = c(F)
                          )
row_split <- by_sample_mat %>% rownames %>% as.data.frame() %>% left_join(df_by_sample %>% select(name, `pathway group`), by = c("." = "name")) %>% pull(`pathway group`) %>% as.factor
column_split <- by_sample_mat %>% colnames %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor
col_fun = colorRamp2(c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100), rev(colorRampPalette(c(brewer.pal(9, "Blues")[9], "#FFFFFF"))(11)))
hp <- Heatmap(by_sample_mat,
        col = col_fun,
        clustering_method_columns = 'ward.D2',
        column_split = column_split,
        row_split = row_split,
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # border = T,
        heatmap_legend_param = list(title = "Completeness (%)"),
        top_annotation = anno_top_host,
        left_annotation = anno_left_module,
        # bottom_annotation = column_ba, 
        # right_annotation = row_ra,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 5),
        cluster_columns = T,
        cluster_rows = T
        # column_dend_side = "bottom",
)
pdf("results/figures/11-gene_function/heatmap_by_sample_module_completeness.pdf", width = 15, height = 20)
draw(hp, heatmap_legend_side = "top", merge_legend = TRUE)
dev.off()
hp <- Heatmap(by_sample_mat,
        col = col_fun,
        clustering_method_columns = 'ward.D2',
        # column_split = column_split,
        row_split = row_split,
        column_title_gp = grid::gpar(fontsize = 0),
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_rot = 0,
        # border = T,
        heatmap_legend_param = list(title = "Completeness (%)"),
        top_annotation = anno_top_host,
        left_annotation = anno_left_module,
        # bottom_annotation = column_ba, 
        # right_annotation = row_ra,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 5),
        cluster_columns = T,
        cluster_rows = T
        # column_dend_side = "bottom",
)
pdf("results/figures/11-gene_function/heatmap_by_sample_module_completeness_clust.pdf", width = 15, height = 20)
draw(hp, heatmap_legend_side = "top", merge_legend = TRUE)
dev.off()
```

```{r}
# genus <- "g__Gilliamella"
for (genus in c(genera, "g__Saezia_snod")) {
  if (file.exists(paste0("results/figures/visualize_temp/microbeannotator_out/by_genus/", genus, "/", genus,"__module_completeness.tab"))) {
    if (1==1) {
    # if (!dir.exists(paste0("results/figures/11-gene_function/by_genus/", genus))) {
      dir.create(paste0("results/figures/11-gene_function/by_genus/", genus))
      df_by_genus <- read.csv(paste0("results/figures/visualize_temp/microbeannotator_out/by_genus/", genus, "/", genus,"__module_completeness.tab"), check.names = F, sep = "\t") %>%
                        mutate(name = abbreviate(name, 55))
      format_colnames <- function(list_cols) {
        sapply(list_cols, function(x) if (x %in% c("name", "module", "pathway group")) {return(x)} else {return(strsplit(x, "_kos.csv")[[1]][1])})
      }
      colnames(df_by_genus) <- df_by_genus %>% colnames %>% format_colnames
      by_genus_mat <- df_by_genus %>% 
        select(!c(module, `pathway group`), ) %>%
        column_to_rownames("name") %>% as.matrix
      # for each species / genus get a set of samples that have enough coverage of the species / genus to proceed
      # subset those columns here - for now only removing "samples_to_exclude"
      by_genus_mat <- by_genus_mat[, colnames(by_genus_mat) %in% samples]
      # remove sparse rows
      min <- 10
      if (dim(by_genus_mat)[[2]] < 10) {
        min <- 1
      }
      by_genus_mat <- by_genus_mat[apply(by_genus_mat, 1, function(x) sum(x > 0, na.rm = T)) > min,]
      anno_top_host <- HeatmapAnnotation(Host = by_genus_mat %>% colnames %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host),
                                    col = list(Host = host_order_color_dark
                                    ), border = T
                                )
      anno_left_module <- rowAnnotation(Group = by_genus_mat %>% rownames %>% as.data.frame() %>% left_join(df_by_genus %>% select(name, `pathway group`), by = c("." = "name")) %>% pull(`pathway group`),
                                    col = list(Group = make_col_list(by_genus_mat %>% rownames %>% as.data.frame() %>% left_join(df_by_genus %>% select(name, `pathway group`), by = c("." = "name")) %>% pull(`pathway group`), "Set3")
                                    ), border = T,
                                    show_legend = c(F)
                                )
      row_split <- by_genus_mat %>% rownames %>% as.data.frame() %>% left_join(df_by_genus %>% select(name, `pathway group`), by = c("." = "name")) %>% pull(`pathway group`) %>% as.factor
      column_split <- by_genus_mat %>% colnames %>% as.data.frame() %>% mutate(Host = Vectorize(get_host_from_sample_name)(.)) %>% pull(Host) %>% as.factor
      col_fun = colorRamp2(c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100), rev(colorRampPalette(c(brewer.pal(9, "Blues")[9], "#FFFFFF"))(11)))
      hp <- Heatmap(by_genus_mat,
              col = col_fun,
              clustering_method_columns = 'ward.D2',
              column_split = column_split,
              row_split = row_split,
              column_title_gp = grid::gpar(fontsize = 0),
              row_title_gp = grid::gpar(fontsize = 12),
              row_title_rot = 0,
              # border = T,
              heatmap_legend_param = list(title = "Completeness (%)"),
              top_annotation = anno_top_host,
              left_annotation = anno_left_module,
              # bottom_annotation = column_ba, 
              # right_annotation = row_ra,
              column_names_gp = grid::gpar(fontsize = 0),
              row_names_gp = grid::gpar(fontsize = 5),
              cluster_columns = T,
              cluster_rows = T
              # column_dend_side = "bottom",
      )
      draw(hp, heatmap_legend_side = "bottom", merge_legend = TRUE, column_title = paste0(genus, " module completeness"))
      pdf(paste0("results/figures/11-gene_function/by_genus/", genus, "/", genus, "_module_completeness.pdf"), width = 15, height = 20)
      draw(hp, heatmap_legend_side = "bottom", merge_legend = TRUE, column_title = paste0(genus, " module completeness"))
      dev.off()
      hp <- Heatmap(by_genus_mat,
              col = col_fun,
              clustering_method_columns = 'ward.D2',
              # column_split = column_split,
              row_split = row_split,
              column_title_gp = grid::gpar(fontsize = 0),
              row_title_gp = grid::gpar(fontsize = 12),
              row_title_rot = 0,
              # border = T,
              heatmap_legend_param = list(title = "Completeness (%)"),
              top_annotation = anno_top_host,
              left_annotation = anno_left_module,
              # bottom_annotation = column_ba, 
              # right_annotation = row_ra,
              column_names_gp = grid::gpar(fontsize = 0),j
              row_names_gp = grid::gpar(fontsize = 5),
              cluster_columns = T,
              cluster_rows = T
              # column_dend_side = "bottom",
      )
      draw(hp, heatmap_legend_side = "bottom", merge_legend = TRUE, column_title = paste0(genus, " module completeness"))
      pdf(paste0("results/figures/11-gene_function/by_genus/", genus, "/", genus, "_module_completeness_clust.pdf"), width = 15, height = 20)
      draw(hp, heatmap_legend_side = "bottom", merge_legend = TRUE, column_title = paste0(genus, " module completeness"))
      dev.off()
    } else {
      print(paste0("directory already exists for ", genus))
    }
  } else {
    print(paste0("no file for ", genus, " at ", paste0("results/figures/visualize_temp/microbeannotator_out/by_genus/", genus, "/", genus,"__module_completeness.tab")))
  }
}
```