---
title: "Honeybee cross-species analysis"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---
<!-- comment out for pdf compiling -->
<!-- some css for formatting html -->
<!-- <style>
    h1.title {
        font-size: 40px;
        font-family: Serif;
        text-align: center;
        font-weight: normal;
        /* color: DarkRed; */
    }
    h1 {
        font-size: 24px;
        font-family: Serif;
        font-weight: bold;
        /* font-weight: bold; */
        /* text-align: center; */
        /* color: DarkRed; */
    }
    h2 {
        font-size: 22px;
        font-family: Serif;
        font-weight: bold;
        /* text-align: center; */
        /* color: DarkRed; */
    }
    h3 {
        font-size: 20px;
        font-family: Serif;
        font-weight: bold;
        /* text-align: center; */
        /* color: DarkRed; */
    }
    body .main-container {
        /* max-width: 1000px; */
        font-size: 18px;
        font-family: Serif;
    }
</style> -->

<!-- ```{css echo=FALSE, warning=FALSE}
/* To make hoverable links. (does not have to be called hint) Usage: */
/* [Message to show on hover]{.hint} */
.hint {
  visibility: hidden;
}

.hint::before {
  visibility: visible;
  content: "Hint";
  color: blue;
}

.hint:hover {
  visibility: visible;
  font-weight: bold;
}

.hint:hover::before {
  display: none;
}
``` -->

# Introduction

Collection of files, definitions and functions for visualization scripts. The chunks on this can be run first before any of the other scripts for visualization.

## General definitions and functions

### Load libraries

Check that the working directory is specified correctly.

If there are any specific changes to functions be sure to re-write the RData object before closing!

```{r load_libraries}

# library(picante)
# library(ggpubr) - version confilct
# Previously saved env data not clean - redo next time
# save.image("/scratch/aprasad/20230313_apis_species_comparison/results/figures/workspace_generaldata_chunks_20230611.RData")
# load("/scratch/aprasad/20230313_apis_species_comparison/results/figures/workspace_generaldata_chunks_20230611.RData")
source('/scratch/aprasad/20230313_apis_species_comparison/scripts/visualization/utilitites.R')
```

### Read metadata and parse into dataframe

Read file with metadata. Working directory set here. Check full path.

```{r parse_metadata}
path_df_meta_complete <- "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/DataCollection/221220_metadata_compiled.xlsx"
df_meta_complete <- read_xlsx(path_df_meta_complete, sheet = "Compiled", range = "A368:AH661", col_names = F)
colnames(df_meta_complete) <- read_xlsx(path_df_meta_complete, sheet = "Compiled", range = "A1:AH1", col_names = F)
df_meta_complete <- df_meta_complete %>%
                      rename(ID_long = ID) %>%
                        rename(ID = Short_name) %>%
                          mutate(Species_long = Species) %>%
                            mutate(ID = Vectorize(convert_sample_name)(ID)) %>%
                            mutate(Species = Vectorize(get_host_from_sample_name)(ID)) %>%
                            mutate(Country = Vectorize(get_location_from_sample_name)(ID))
df_meta_complete %>% select(Species, ID)

```

## Initialization for 00a plot DNA concentration
### Load and parse data
```{r init_00a}
df_concentrations <- df_meta_complete %>%
  select(ID, Concentration) %>%
      mutate(Host = Vectorize(get_host_from_sample_name)(ID)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(ID))
```

## Initialization for 00b plot qPCR values

### Definitions and functions
```{r 00b-functions_and_definitions}
qpcr_results <- c(
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate1/plate1-2022-11-17_115607_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate2/plate2-2022-11-17_142003_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate3/plate3-2022-11-17_161529_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate4/plate4-2022-11-17_181034_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate5/plate5-2023-02-Malaysia_samples_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate6/plate6-2023-02-Malaysia_samples_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate7/plate7-2023-02-Malaysia_samples_Results.xls"
)
```
### Load and parse data

sample C7.1 was repeated in plate 6 and 7. Consider the value from plate 7 (both values were close)

```{r 00b-load_and_parse}
qpcr_df <- data.frame()
plate_num <- 1
for(x in qpcr_results){
  # read just the rows and columns with the values and leave out metadata cells
  # Limits the cells from rows=(36 to 132) and columns=(2 to 11)
  t <- read_excel(x, sheet = 1, range = cell_limits(c(36,2), c(132,11)))
  t <- t %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H1", "H2", "H3"), paste0("Control_p_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H4", "H5", "H6"), paste0("Control_n_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H7", "H8", "H9"), paste0("Control_act_p_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H10", "H11", "H12"), paste0("Control_act_n_", plate_num), `Sample Name`))
  # get only columns "Well Position", "Sample Name", "Target Name" and "CT" from the xls report
  t <- t[,c(1,3,4,8,9,10)]
  names(t)[1:4] <- c("Well", "Sample.Name","Target.Name","Ct")
  t$Ct <- as.numeric(t$Ct) # NAs will be introduced at the place of 'undetermined' values
  t$Target.Name <- factor(t$Target.Name)
  t$Sample.Name <- factor(t$Sample.Name)
  # t$sd <- factor(t$sd)
  t <- as.data.frame(t)
  # add in the plate number
  t <- cbind(t, "Plate" = as.factor(rep(plate_num, dim(t)[1])))
  # bind the previous plate values with the next before moving to the next file
  qpcr_df <- rbind(qpcr_df, t)
  plate_num = plate_num + 1
}
```

```{r 00b-intialize}
# qpcr_df %>% reframe(Sample.Name, Ct) %>% filter(grepl("Control", Sample.Name))
# qpcr_df_info %>% reframe()
qpcr_df_info <- qpcr_df %>%
          mutate(Target.Name = ifelse(Target.Name == "UV_0356/0772", "Bacteria", "Host")) %>%
                filter(!is.na(Sample.Name)) %>%
                filter(Sample.Name!="EMPTY") %>%
                filter(!(Sample.Name=="C7.1" & Plate == 6)) %>%
                filter(!is.na(`Ct Mean`)) %>%
                mutate(Sample.Name = Vectorize(convert_sample_name)(Sample.Name)) %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample.Name)) %>%
                  mutate(Host = ifelse(grepl("Control_n", Sample.Name), "-veControl", Host)) %>%
                  mutate(Host = ifelse(grepl("Control_p", Sample.Name), "+veControl", Host)) %>%
                  mutate(Host = as.factor(Host)) %>%
                  filter(Target.Name == "Bacteria") %>%
                  group_by(Sample.Name) %>%
                    reframe(`Ct Mean`, Host, `Ct SD`, Plate) %>%
                      group_by(Host) %>%
                      unique() %>%
                        reframe(Sample.Name, `Ct Mean`, `Ct SD`, Plate) %>%
                          rename(Ct_mean = `Ct Mean`) %>%
                          rename(Ct_SD = `Ct SD`)
UV_eff <- 1.971
UV_int <- 36.94109669
# Actin_eff <- 2.001
# Actin_int <- 36.71918987
elution_vol <- 30
dilution_factor <- 10
LOD <- 100
qpcr_plot_df <- qpcr_df_info %>%
                  mutate(copy_num = UV_eff^(UV_int - Ct_mean)*elution_vol*dilution_factor) %>%
                  mutate(Location = Vectorize(get_location_from_sample_name)(Sample.Name)) %>%
                  filter(!is.na(Host)) %>%
                  filter(Host %in% host_order) %>%
                    unique()
write.csv(qpcr_plot_df, "results/figures/qpcr_plot_df.csv", row.names = F, quote = F)
```

To do a sanity check use `qpcr_df_info %>% reframe(n())`. 
`r qpcr_df_info %>% reframe(n())`

## Initialization for 01 plot mapping results

### Definitions and functions
### Load and parse data

```{r load_and_parse_data}
df_raw_info <- read.csv("config/raw_sequence_library_info.txt", header = T, sep = "\t", stringsAsFactors = F)
df_raw_info <- df_raw_info %>%
                group_by(libraryName) %>%
                mutate(reads = Vectorize(read_comma_numbers)(clustersPF)) %>%
                  mutate(raw_reads = 2*sum(reads)) %>%
                    ungroup() %>%
                      select(libraryName, raw_reads) %>%
                      rename(Sample = libraryName)
df_raw_ind <- read.csv("config/raw_sequence_library_info_india_samples.txt", header = T, sep = "\t", stringsAsFactors = F) %>%
                select(Sample, Total_reads) %>%
                  mutate(raw_reads = Vectorize(read_comma_numbers)(Total_reads)) %>%
                    select(Sample, raw_reads)
df_raw_info <- rbind(df_raw_info, df_raw_ind)
```

```{r 01-init}
# use bash script to get number of reads from raw and  files (before concatenating summed per sample)
# extremely slow when done in R (login node)
df_reads <- data.frame()
for (sample in samples) {
  raw_reads <- NA
  if (sample %in% df_raw_info$Sample) {
    raw_reads <- df_raw_info[which(df_raw_info$Sample == sample), "raw_reads"] %>% pull %>% head(1)
  }
  total_clean <- NA
  assembly_mapped <- NA
  host_mapped <- NA
  rep_db_mapped <- NA
  gene_catalog_mapped <- NA
  host_unmpd_mag_mapd <- NA
  mag_unmaped_host_mapped <- NA
  # get raw reads
    # number of reads from raw and  files (before concatenating summed per sample)
    # for now only getting data from repair cleaning not fastqc
  if (file.exists(paste0("results/01_cleanreads/", sample, "_repaired.log.cluster.e"))) {
    lines <- readLines(paste0("results/01_cleanreads/", sample, "_repaired.log.cluster.e"))
    for (line in lines) {
      if (startsWith(line, "Input:")) {
      }
      else if (startsWith(line, "Pairs:")) {
        total_clean <- as.numeric(strsplit(strsplit(line, "\t")[[1]][2], " reads")[[1]][1])
      } else {
        if (file.exists(paste0("results/01_cleanreads/", sample, "_repaired.log"))) {
          lines <- readLines(paste0("results/01_cleanreads/", sample, "_repaired.log"))
          for (line in lines) {
            if (startsWith(line, "Input:")) {
            }
            else if (startsWith(line, "Pairs:")) {
              total_clean <- as.numeric(strsplit(strsplit(line, "\t")[[1]][2], " reads")[[1]][1])
            }
          }
        }
      }
    }
  }
  # reads mapped to assembly
  assembly_summary <- "results/05_assembly/assembly_summary.txt"
  assembly_summary_df <- read.csv(assembly_summary, header = T, stringsAsFactors = F)
  if (sample %in% assembly_summary_df$Sample) {
    total_clean <- assembly_summary_df[which(assembly_summary_df$Sample == sample), "Number.of.reads"] %>% as.numeric()
  }
  if (sample %in% assembly_summary_df$Sample) {
    assembly_mapped <- assembly_summary_df[which(assembly_summary_df$Sample == sample), "Number.mapped"] %>% as.numeric()
  }
  # reads mapped to host reference
  host_flagstat <- paste0("results/03_host_mapping/", sample, "_flagstat.txt")
  if (file.exists(host_flagstat)) {
    host_mapped <- get_read_count_from_flagstat(host_flagstat, "and mate mapped")
  }
  # reads mapped to rep mag database
  rep_flagstat <- paste0("results/10_instrain/01_mapping/", sample, "/", sample, "_flagstat.tsv")
  if (file.exists(rep_flagstat)) {
    rep_db_mapped <- get_read_count_from_flagstat(rep_flagstat, "and mate mapped")
  }
  # reads mapped to gene catalog
  gene_catalog_flagstat <- paste0("results/08_gene_content/01_profiling/", sample, "_mapped.flagstat")
  if (file.exists(gene_catalog_flagstat)) {
    gene_catalog_mapped <- get_read_count_from_flagstat(gene_catalog_flagstat)
  }
  # reads mapped to gene catalog
  host_unmpd_mag_mapd_flagstat <- paste0("results/04_MapToDBs/", sample, "/", sample, "_host_unmapd_map_MAGs_rep.flagstat")
  if (file.exists(host_unmpd_mag_mapd_flagstat)) {
    host_unmpd_mag_mapd <- get_read_count_from_flagstat(host_unmpd_mag_mapd_flagstat)
  }
  mag_unmaped_host_mapped_flagstat <- paste0("results/04_MapToDBs/", sample, "/", sample, "_unmapd_rep_MAGs_host_mapping.flagstat")
  if (file.exists(mag_unmaped_host_mapped_flagstat)) {
    mag_unmaped_host_mapped <- get_read_count_from_flagstat(mag_unmaped_host_mapped_flagstat)
  }
  values_to_bind <- c(sample, raw_reads, total_clean, assembly_mapped, host_mapped, rep_db_mapped, gene_catalog_mapped, host_unmpd_mag_mapd, mag_unmaped_host_mapped)
  df_reads <- rbind(df_reads, values_to_bind)
}
# parse this later
# read.csv(paste0("02_HostMapping/", sample, "_coverage.tsv"), sep = "\t")
```

```{r load_reads}
# at the moment number of raw reads is not eastimated
# hf means host mapping was done first then the other
# mf means mag mapping was done first then the other
df_colnames <- c("Sample", "Raw", "Total_clean", "Assembly", "Host_mapped_hf", "MAGs_DB_mf", "Gene_catalog", "MAGs_DB_hf", "Host_mapped_mf")
colnames(df_reads) <- df_colnames
df_reads <- df_reads %>%
              mutate(across(!c("Sample"), as.integer))
df_reads_plot <- df_reads %>%
                          mutate(Unmapped_hf = Total_clean - MAGs_DB_hf - Host_mapped_hf) %>%
                          mutate(Unmapped_mf = Total_clean - MAGs_DB_mf - Host_mapped_mf) %>%
                          mutate(microbe_to_host_ratio = MAGs_DB_mf/Host_mapped_mf) %>%
                            pivot_longer(!Sample, names_to = "Type", values_to = "Number") %>%
                              mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
                              mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
df_reads_updated <-  df_reads %>%
  mutate(sum = Host_mapped_hf + MAGs_DB_hf) %>%
  mutate(Unmapped_hf = Total_clean - MAGs_DB_hf - Host_mapped_hf) %>%
  mutate(Unmapped_mf = Total_clean - MAGs_DB_mf - Host_mapped_mf) %>%
  select(Sample, MAGs_DB_hf, Host_mapped_hf, Unmapped_hf, Total_clean) %>%
    mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
    mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
        mutate(MAGs_DB_range = ifelse(MAGs_DB_hf >= 30e+06, "High", NA)) %>%
        mutate(MAGs_DB_range = ifelse(MAGs_DB_hf >= 15e+06 & MAGs_DB_hf < 30e+06, "Sufficient", MAGs_DB_range)) %>%
        mutate(MAGs_DB_range = ifelse(MAGs_DB_hf >= 0 & MAGs_DB_hf < 15e+06, "Low", MAGs_DB_range)) %>%
          mutate(Total_depth = ifelse(Total_clean >= 30e+06, "Target_reached", "Insufficient"))
```

## Initialization for 02 plot motus result

### Definitions and functions

```{r 02-functions_and definitions}
condense_motu <- function(motu){
  motu_condensed = strsplit(motu, " ")[[1]][1]
  return(motu_condensed)
}

extend_colors_motus <- function(names_vec){
  final_list <- list()
  for (a_name in names_vec) {
    if (a_name %in% names(motuColors)) {
      final_list[a_name] = motuColors[a_name]
    } else {
      final_list[a_name] = "grey"
    }
  }
  return(final_list)
}
```

### Load and parse data

```{r load_and_parse_data}
df_motus_raw <- read.csv("results/02_motus_profile/samples_merged.motus", sep = "\t", skip = 2, stringsAsFactors=FALSE)
```

```{r 02-init}
colnames(df_motus_raw)[1] <- "taxonomy"
df_motus_raw <- df_motus_raw %>%
              mutate(across(!c("taxonomy"), as.numeric)) %>%
              mutate(sum_ab = rowSums(across(where(is.numeric)))) %>%
                filter(sum_ab > 0) %>%
                  select(!sum_ab) %>%
                    column_to_rownames("taxonomy")

df_motus <- as.data.frame(t(df_motus_raw)) %>%
              rownames_to_column("Sample") %>%
              mutate(Sample = Vectorize(convert_sample_name)(Sample)) %>%
              pivot_longer(!Sample, names_to = "motu", values_to = "rel_ab") %>%
                group_by(Sample, motu) %>%
                  mutate(Present = ifelse(rel_ab > 0, 1, 0)) %>%
                    group_by(motu) %>%
                     mutate(Prevalence_num = sum(Present)) %>%
                      mutate(Prevalence = mean(Present))

df_motus_combined <- left_join(df_motus, df_meta_complete, by = c("Sample" = "ID"))  %>%
                      group_by(Species, motu) %>%
                        mutate(Present_host = ifelse(rel_ab > 0, 1, 0)) %>%
                          group_by(Species, motu) %>%
                            mutate(Prevalence_num_host = sum(Present_host)) %>%
                            mutate(Prevalence_host = mean(Present_host)) %>%
                              mutate(mean_rel_ab_host = mean(rel_ab)) %>%
                              group_by(Sample) %>%
                              mutate(mean_rel_ab = mean(rel_ab)) %>%
                                mutate(motu_condensed = Vectorize(condense_motu)(motu))
```

## Initialization for 03 Assembly summary
### Definitions and functions
### Load and parse data

```{r 03-load_and_parse}
df_assembly <- read.csv("results/05_assembly/assembly_summary.txt", sep = ',') %>%
                left_join(df_meta_complete, by=c("Sample" = "ID"))
```

```{r 03-init}
df_assembly_plot <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    rename(Unmapped = Number.unmapped, Mapped = Number.mapped) %>%
     pivot_longer(cols = 3:4, names_to ="Type", values_to = "Number")

df_assembly_plot <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    mutate(percent_mapped = Number.mapped/Number.of.reads*100) %>%
    mutate(percent_unmapped = Number.unmapped/Number.of.reads*100) %>%
    rename(Unmapped = percent_unmapped, Mapped = percent_mapped, Unmapped_number = Number.unmapped, Mapped_number = Number.mapped) %>%
    select(Sample, Mapped, Unmapped) %>%
     pivot_longer(cols = 2:3, names_to ="Type", values_to = "Percent")

df_assembly_plot_number <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    mutate(percent_mapped = Number.mapped/Number.of.reads*100) %>%
    mutate(percent_unmapped = Number.unmapped/Number.of.reads*100) %>%
    rename(Unmapped = Number.unmapped, Mapped = Number.mapped) %>%
    select(Sample, Mapped, Unmapped) %>%
     pivot_longer(cols = 2:3, names_to ="Type", values_to = "Number")
```


## Initialization for 04 xxx
### Definitions and functions
### Load and parse data

## Initialization for 05 xxx
### Definitions and functions
### Load and parse data

## Initialization for 06 plot and summarize mags

### Definitions and functions

```{r 06-functions_and_definitions}
samples_am  <- samples_AM %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_ac <- samples_AC %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_ad <- samples_AD %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_af <- samples_AF %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_aa <- samples_AA %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
line_list <- c()
for (num in 1:94){
  add_line <- geom_vline(xintercept=num+0.5, size=0.1, color="black")
  # add_line <- geom_vline(xintercept=num+0.5, size=0.1, alpha=0.5, color="grey")
  line_list <- c(line_list, add_line)
}
make_col_list <- function(my_vec, my_palette = "Spectral") {
  uniq_names <- unique(ifelse(is.na(my_vec), 0, my_vec))
  num_ele <- length(uniq_names)
  if (num_ele <= 3) {
    cols_used <- brewer.pal(5, my_palette)[1:num_ele]
  }
  if (num_ele <= 9 & num_ele > 3) {
    cols_used <- brewer.pal(num_ele, my_palette)
  } else {
    if (my_palette == "Pastel2") {
      cols_used <- colorRampPalette(brewer.pal(8, my_palette))(num_ele)
    } else {
      cols_used <- colorRampPalette(brewer.pal(9, my_palette))(num_ele)
    }
  }
  col_list <- c(cols_used)
  names(col_list) = uniq_names
  return(col_list)
}
my_col_fun <- colorRamp2(c(0.7, 0.8, 0.9, 0.93, 0.94, 0.95, 1), c("#ffffe5", "#f7fcb9", "#d9f0a3", "#4d9221", "#fddbc7", "#b2182b", "#67000d"))
pretty_ani_heatmaps <- function(matrix, mags_info_df=vis_magOTUs_df, row_split = "magotus", add_values = F, value_size = 10, limit_l = 0.95, limit_h = 1, col_fun = colorRamp2(c(limit_l, limit_h), c("#9ecae1", "#08306b"))) {
  host_names <- rownames(matrix) %>% as.data.frame() %>%
                  as.data.frame() %>%
                        left_join(mags_info_df %>%
                          ungroup() %>%
                            select(ID, Host), by = c(. = "ID")) %>% pull(Host)
  magOTU_names <- rownames(matrix) %>% 
                      as.data.frame() %>%
                        left_join(mags_info_df %>%
                          ungroup() %>%
                            select(ID, magOTU), by = c(. = "ID")) %>% pull(magOTU)
  Genus_names <- rownames(matrix) %>% 
                    as.data.frame() %>%
                      left_join(mags_info_df %>%
                        ungroup() %>%
                          select(ID, Genus), by = c(. = "ID")) %>% pull(Genus)
  species_names <- rownames(matrix) %>% 
                    as.data.frame() %>%
                      left_join(mags_info_df %>%
                        ungroup() %>%
                          select(ID, Species), by = c(. = "ID")) %>% pull(Species)
  anno_host_col = HeatmapAnnotation(Host = host_names,
                                      col = list(Host = host_order_color),
                                      border = TRUE
                               )
  anno_host_row = rowAnnotation(Host = host_names,
                                      col = list(Host = host_order_color),
                                      border = TRUE
                                                  )
  anno_gtdb_col = HeatmapAnnotation(`GTDB Species` = species_names,
                                    col = list(`GTDB Species` = make_col_list(species_names, "Spectral")),
                                    border = TRUE
                          )
  anno_gtdb_row = rowAnnotation(`GTDB Species` = species_names,
                                    col = list(`GTDB Species` = make_col_list(species_names, "Spectral")),
                                    border = TRUE
                          )
  anno_magotu_col = HeatmapAnnotation(`magOTU` = magOTU_names,
                                col = list(`magOTU` = make_col_list(magOTU_names, "Spectral")
                                          ),
                                          border = TRUE
                          )
  anno_magotu_row = rowAnnotation(`magOTU` = magOTU_names,
                                col = list(`magOTU` = make_col_list(magOTU_names, "Spectral")
                                          ),
                                          border = TRUE
                          )
  anno_tax_col = HeatmapAnnotation(`GTDB Species` = species_names,
                                   `magOTU` = magOTU_names,
                                   col = list(`magOTU` = make_col_list(magOTU_names, "Spectral"),
                                              `GTDB Species` = make_col_list(species_names, "Spectral")
                                             ),
                                             border = TRUE
                                   )
  anno_tax_row = rowAnnotation(`GTDB Species` = species_names,
                                   `magOTU` = magOTU_names,
                                   col = list(`magOTU` = make_col_list(magOTU_names, "Spectral"),
                                              `GTDB Species` = make_col_list(species_names, "Spectral")
                                             ),
                                             border = TRUE
                                   )
  if (row_split == "magotus") {
    row_split = as.character(as.factor(magOTU_names))
    # column_split = as.numeric(as.factor(host_names))
  }
  if (add_values) {
    cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(ifelse(matrix[i,j] > 0.7, sprintf("%.2f", matrix[i, j]), "-"), x, y, gp = gpar(fontsize = value_size))
          }
    heatmap_obj = Heatmap(matrix, 
            col = col_fun,
            # col = brewer.pal(2, "Pastel1"),
            # clustering_method_columns = 'ward.D2',
            # top_annotation = anno_host_col,
            right_annotation = anno_gtdb_row,
            left_annotation = anno_host_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 0),
            heatmap_legend_param = list(title = "ANI", color_bar = "Continuous"),
            row_split = row_split,
            row_title_rot = 0,
            cell_fun = cell_fun,
            cluster_columns = F,
            border = TRUE,
            # row_dend_reorder = T,
            cluster_rows = F
            )
  } else {
    heatmap_obj = Heatmap(matrix, 
            col = col_fun,
            # col = brewer.pal(2, "Pastel1"),
            # clustering_method_columns = 'ward.D2',
            # top_annotation = anno_host_col,
            right_annotation = anno_gtdb_row,
            left_annotation = anno_host_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 0),
            heatmap_legend_param = list(title = "ANI", color_bar = "Continuous"),
            row_split = row_split,
            row_title_rot = 0,
            # cell_fun = cell_fun,
            cluster_columns = F,
            border = TRUE,
            # row_dend_reorder = T,
            cluster_rows = F
            )
  }
  return(heatmap_obj)
  # draw(heatmap_obj, merge_legend = TRUE)
}
get_ani_matrix <- function(genera_list = c(), mag_names = c(), magOTUs = c(), mags_info_df=vis_magOTUs_df) {
  if (length(genera_list) > 0) {
    matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% mags_info_df$ID) %>%
                    right_join(mags_info_df %>% ungroup() %>% select(ID, magOTU, Host, Species, Genus), by = c("reference" = "ID")) %>%
                      filter(Genus %in% genera_list) %>%
                      # mutate(reference = paste0(magOTU, "_", Genus, "_", reference)) %>%
                      # mutate(querry = paste0(magOTU, "_", Genus, "_", querry)) %>%
                      arrange(Genus, magOTU, Host, Species) %>%
                      select(!c(Genus, magOTU)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix
  }
  if (length(mag_names) > 0) {
    matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% mags_info_df$ID) %>%
                    right_join(mags_info_df %>% ungroup() %>% select(ID, Cluster, Host, Species, Genus), by = c("reference" = "ID")) %>%
                      filter(reference %in% mag_names) %>%
                      # filter(reference %in% mag_names | querry %in% mag_names) %>%
                      # mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      # mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      arrange(Genus, Cluster, Host, Species) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix  
  }
  if (length(magOTUs) > 0) {
    matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% mags_info_df$ID) %>%
                    right_join(mags_info_df %>% ungroup() %>% select(ID, Cluster, Host, Genus), by = c("reference" = "ID")) %>%
                      filter(Cluster %in% genera_list) %>%
                      # mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      # mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      arrange(Genus, Cluster, Host, Species) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix  
  }
  return(matrix)
}
```


### Load and parse data

```{r 06-load}
df_gtdbk_bac <- read.csv("results/09_MAGs_collection/gtdb_output/classify/20230313_MAGs.bac120.summary.tsv", sep = "\t")
df_gtdbk_ar <- read.csv("results/09_MAGs_collection/gtdb_output/classify/20230313_MAGs.ar53.summary.tsv", sep = "\t")
drep_Bdb <- read.csv("results/09_MAGs_collection/drep_output/data_tables/Bdb.csv", sep = ",")
drep_gInformation <- read.csv("results/09_MAGs_collection/drep_output/data_tables/genomeInfo.csv", sep = ",")
drep_Sdb <- read.csv("results/09_MAGs_collection/drep_output/data_tables/Sdb.csv", sep = ",")
drep_Widb <- read.csv("results/09_MAGs_collection/drep_output/data_tables/Widb.csv", sep = ",")
MAGs_collated <- read.csv("results/09_MAGs_collection/MAGs_metadata_summary.tsv", sep = "\t")
colnames(MAGs_collated)
# system("bash scripts/calc_genome_sizes.sh /scratch/aprasad/20230313_apis_species_comparison/results/09_MAGs_collection/MAGs_high_medium /scratch/aprasad/20230313_apis_species_comparison/results/figures/Genome_sizes_good_quality.csv fa")
# system("bash scripts/calc_genome_sizes.sh /scratch/aprasad/20230313_apis_species_comparison/results/09_MAGs_collection/MAGs /scratch/aprasad/20230313_apis_species_comparison/results/figures/Genome_sizes.csv fa")
# system("bash scripts/calc_genome_sizes.sh /scratch/aprasad/20230313_apis_species_comparison/results/11_phylogenies/downloads /scratch/aprasad/20230313_apis_species_comparison/results/figures/Genome_sizes_refs.csv fna")
genome_lengths_df <- read.csv("results/figures/Genome_sizes.csv")
genome_lengths_df_hq <- read.csv("results/figures/Genome_sizes_good_quality.csv")
genome_lengths_df_ref <- read.csv("results/figures/Genome_sizes_refs.csv")
reference_genomes_info <- read.csv("config/Outgroup_isolate_genomes.tsv", "\t", header = T)
drep_N <- read.csv("results/09_MAGs_collection/drep_output/data_tables/Ndb.csv") %>%
            mutate(reference = Vectorize(format_genome_name)(reference)) %>%
              mutate(querry = Vectorize(format_genome_name)(querry))
```
```{r 06-for_contig_depths}
# samples_am <- c(vis_magOTUs_df_all %>% filter(Host == "Apis mellifera") %>% pull(Sample) %>% unique %>% as.vector)
# samples_ac <- c(vis_magOTUs_df_all %>% filter(Host == "Apis cerana") %>% pull(Sample) %>% unique %>% as.vector)
# samples_ad <-c(vis_magOTUs_df_all %>% filter(Host == "Apis dorsata") %>% pull(Sample) %>% unique %>% as.vector)
# samples_af <-c(vis_magOTUs_df_all %>% filter(Host == "Apis florea") %>% pull(Sample) %>% unique %>% as.vector)
# contig_fates_df_am <- data.frame()
# for (sample in samples_am) {
#   contig_fates_df_am <- rbind(contig_fates_df_am, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
# }
# contig_fates_df_am <- cbind(contig_fates_df_am, host = rep("Apis mellifera", dim(contig_fates_df_am)[[1]]))
# contig_fates_df_am_pf <- contig_fates_df_am %>%
#                               group_by(sample, passed_filter) %>%
#                                 reframe(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
# contig_fates_df_am_bin <- contig_fates_df_am %>%
#                               group_by(sample, binned) %>%
#                                 reframe(binned_length = sum(length), binned_num = n(), .groups = "keep")


# contig_fates_df_ac <- data.frame()
# for (sample in samples_ac) {
#   contig_fates_df_ac <- rbind(contig_fates_df_ac, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
# }
# contig_fates_df_ac <- cbind(contig_fates_df_ac, host = rep("Apis cerana", dim(contig_fates_df_ac)[[1]]))
# contig_fates_df_ac_pf <- contig_fates_df_ac %>%
#                               group_by(sample, passed_filter) %>%
#                                 reframe(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
# contig_fates_df_ac_bin <- contig_fates_df_ac %>%
#                               group_by(sample, binned) %>%
#                                 reframe(binned_length = sum(length), binned_num = n(), .groups = "keep")

# contig_fates_df_ad <- data.frame()
# for (sample in samples_ad) {
#   contig_fates_df_ad <- rbind(contig_fates_df_ad, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
# }
# contig_fates_df_ad <- cbind(contig_fates_df_ad, host = rep("Apis dorsata", dim(contig_fates_df_ad)[[1]]))
# contig_fates_df_ad_pf <- contig_fates_df_ad %>%
#                               group_by(sample, passed_filter) %>%
#                                 reframe(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
# contig_fates_df_ad_bin <- contig_fates_df_ad %>%
#                               group_by(sample, binned) %>%
#                                 reframe(binned_length = sum(length), binned_num = n(), .groups = "keep")

# contig_fates_df_af <- data.frame()
# for (sample in samples_af) {
#   contig_fates_df_af <- rbind(contig_fates_df_af, read.csv(paste0("06_MAG_binning/contig_fates/",sample,"_contig_fates.csv"), sep = ","))
# }
# contig_fates_df_af <- cbind(contig_fates_df_af, host = rep("Apis florea", dim(contig_fates_df_af)[[1]]))
# contig_fates_df_af_pf <- contig_fates_df_af %>%
#                               group_by(sample, passed_filter) %>%
#                                 reframe(pass_fail_length = sum(length), pf_num = n(), .groups = "keep")
# contig_fates_df_af_bin <- contig_fates_df_af %>%
#                               group_by(sample, binned) %>%
#                                 reframe(binned_length = sum(length), binned_num = n(), .groups = "keep")

# contigs_length_df <- rbind(
#                         reframe(group_by(contig_fates_df_am, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop"),
#                         reframe(group_by(contig_fates_df_ac, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop"),
#                         reframe(group_by(contig_fates_df_ad, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop"),
#                         reframe(group_by(contig_fates_df_af, sample), contig = contig_name, length = length, binned = binned, bin = bin_name, passed = passed_filter, .groups = "drop")
#                   )
# length_bin_sum_df <- contigs_length_df %>%
#                     mutate(length_bin = cut(length ,breaks=c(0, 500, 1000, 10000, 50000, 100000, 200000, 500000, 1000000, Inf), labels = c("500", "0.5-1kb", "1-10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", ">1Mb"))) %>%
#                       group_by(sample, binned, length_bin) %>%
#                         reframe(length_bin_sum = sum(length), num_contigs = n())
# length_bin_sum_df_passed <- contigs_length_df %>%
#                     filter(passed == "P") %>%
#                     mutate(length_bin = cut(length ,breaks=c(0, 500, 1000, 10000, 50000, 100000, 200000, 500000, 1000000, Inf), labels = c("500", "0.5-1kb", "1-10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", ">1Mb"))) %>%
#                       group_by(sample, binned, length_bin) %>%
#                         reframe(length_bin_sum = sum(length), num_contigs = n())
# all_depths <- data.frame()
# for (sample in samples) {
#   sample_contig_depths <- read.csv(paste0("06_MAG_binning/backmapping/",sample,"/",sample,"_mapped_to_",sample,".depth"), sep = "\t")
#   depth_column <- paste0(sample,"_mapped_to_",sample,".bam")
#   all_depths <- rbind(all_depths, rename(select(sample_contig_depths, contigName, all_of(depth_column)), depth = depth_column))
# }
# contigs_depths_df <- left_join(filter(contigs_length_df, passed == "P"), all_depths, by = c("contig" = "contigName"))
# rm(all_depths)
# contig_fates_df_pf <- rbind(contig_fates_df_am_pf,
#                             contig_fates_df_ac_pf,
#                             contig_fates_df_ad_pf,
#                             contig_fates_df_af_pf
#                       )
# contig_fates_df_bin <- rbind(contig_fates_df_am_bin,
#                             contig_fates_df_ac_bin,
#                             contig_fates_df_ad_bin,
#                             contig_fates_df_af_bin
#                       )
# contig_fates_df_am_mag <- contig_fates_df_am %>%
#                             # filter(binned == "Y") %>%
#                             group_by(sample, bin_name) %>%
#                               reframe(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
#                                 left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
# contig_fates_df_ac_mag <- contig_fates_df_ac %>%
#                             # filter(binned == "Y") %>%
#                             group_by(sample, bin_name) %>%
#                               reframe(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
#                                 left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
# contig_fates_df_ad_mag <- contig_fates_df_ad %>%
#                             # filter(binned == "Y") %>%
#                             group_by(sample, bin_name) %>%
#                               reframe(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
#                                 left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
# contig_fates_df_af_mag <- contig_fates_df_af %>%
#                             # filter(binned == "Y") %>%
#                             group_by(sample, bin_name) %>%
#                               reframe(num_contigs = n(), len_contigs = sum(length), .groups = "keep") %>%
#                                 left_join(select(vis_magOTUs_df_all, ID, Host, Sample, Cluster, Family, Genus, N50, Prevalence), by = c("bin_name" = "ID"))
# contigs_depths_df_genus <- left_join(contigs_depths_df, select(vis_magOTUs_df_all, ID, Genus, Cluster), by = c("bin" = "ID")) %>%
#   ungroup() %>%
#   select(!Host) %>%
#   left_join(rename(select(df_meta_complete, sample, Species), Host = Species), by = c("sample" = "Sample"))
```


```{r 06-init}
cluster_info <- drep_Widb %>%
                  select(genome, score, cluster, cluster_members, furthest_cluster_member) %>%
                    rename(score_representative = score) %>%
                    mutate(representative_genome = Vectorize(format_genome_name)(genome)) %>%
                      select(!genome)
MAGs_collated_info <- left_join(MAGs_collated, mutate(drep_gInformation, genome = Vectorize(format_genome_name)(genome)), by = c("ID"="genome"))
# head(MAGs_collated_info)
# colnames(MAGs_collated_info)
vis_magOTUs_df_all <- MAGs_collated_info %>%
                            mutate(Completeness_quality = cut(Completeness ,breaks=c(-1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 100, 110),
                                                            labels = c("<10", "10-20", "20-30", "30-40", "40-50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-100", "100"))
                                                          ) %>%
                              mutate(Contamination_quality = cut(Contamination ,breaks=c(-1, 0, 5, 10, 100),
                                                              labels = c("0", "0-5", "5-10", ">10"))
                                                          ) %>%
                              mutate(N50_quality = cut(N50 ,breaks=c(0, 10000, 50000, 100000, 200000, 500000, 1000000, 2000000, Inf), labels = c("<10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", "1-2Mb", ">2Mb"))) %>%
                                  mutate(Host = Vectorize(get_host_name)(ID)) %>%
                                  filter(!grepl("unbinned", ID))
vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(magOTU) %>%
                            mutate(Num_mags = n())
# vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
#                           group_by(magOTU, Host) %>%
#                             mutate(Present = n()) %>%
#                             mutate(Prevalence = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
#                             mutate(Prevalence = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence)) %>%
#                             mutate(Prevalence = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence)) %>%
#                             mutate(Prevalence = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence)) %>%
#                             mutate(Prevalence = ifelse(Host=="Apis andreniformis", Present/length(samples_aa), Prevalence)) %>%
#                               # left_join(reframe(group_by(contigs_depths_df, bin), mean_coverage = mean(depth), .groups = "drop"), by = c("ID" = "bin")) %>%
#                                 arrange(Genus)
vis_magOTUs_df_all <- left_join(vis_magOTUs_df_all, cluster_info, by = c("magOTU" = "cluster")) %>%
                        left_join(drep_Sdb %>%
                                    mutate(ID = Vectorize(format_genome_name)(genome)) %>%
                                    select(!genome)
                        )
vis_magOTUs_df_all <- vis_magOTUs_df_all %>% 
                        mutate(Genus = ifelse(Genus == "", NA, Genus)) %>%
                        mutate(Family = ifelse(Family == "", NA, Family)) %>%
                        mutate(Host = as.factor(Host)) %>%
                        mutate(magOTU = as.factor(magOTU)) %>%
                        mutate(Genus = as.factor(Genus)) %>%
                        mutate(completeness = Completeness) %>%
                        mutate(contamination = Contamination) %>%
                        mutate(length = Size)
vis_magOTUs_df <- vis_magOTUs_df_all %>%
                    filter(Quality != "low")
# vis_magOTUs_df$Host <- as.factor(recode(vis_magOTUs_df$Host, Am="Apis mellifera", Ac="Apis cerana", Ad="Apis dorsata", Af="Apis florea"))
vis_magOTUs_df$magOTU <- as.factor(vis_magOTUs_df$magOTU)
vis_magOTUs_df$Sample <- as.factor(vis_magOTUs_df$Sample)
vis_magOTUs_df$Family <- as.factor(vis_magOTUs_df$Family)
vis_magOTUs_df$Genus <- as.factor(vis_magOTUs_df$Genus)
vis_magOTUs_df <- vis_magOTUs_df[order(vis_magOTUs_df$Genus), ]
vis_magOTUs_df_all_means <- vis_magOTUs_df_all %>%
                        group_by(Sample) %>%
                          reframe(Sample, completeness, contamination, N50, Host) %>%
                            mutate(Completeness_mean = mean(completeness)) %>%
                              mutate(Contamination_mean = mean(contamination)) %>%
                              mutate(N50_mean = mean(N50))
cluster_tax_info <- vis_magOTUs_df %>%
                      ungroup() %>%
                      select(Genus, Species, magOTU, Num_mags) %>%
                      unique() %>%
                        mutate(cluster_name = paste0(Genus, "-", magOTU))
# vis_magOTUs_df_all_shared_cluster <- vis_magOTUs_df_all %>%
#                               group_by(Cluster) %>%
#                                 mutate(Num_hosts = n_distinct(Host)) %>%
#                                   filter(Num_hosts > 1)
reference_genomes <- reference_genomes_info %>%
                            rename("ID" = Name) %>%
                              mutate(Genus = Group) %>%
                                left_join(genome_lengths_df_ref) %>%
                                  group_by(Genus) %>%
                                  mutate(length_med = median(length))
matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% vis_magOTUs_df$ID) %>%
                    right_join(vis_magOTUs_df %>% ungroup() %>% select(ID, magOTU, Genus), by = c("reference" = "ID")) %>%
                      # filter(Genus %in% c("g__Bifidobacterium")) %>%
                      mutate(reference = paste0(magOTU, "_", Genus, "_", reference)) %>%
                      mutate(querry = paste0(magOTU, "_", Genus, "_", querry)) %>%
                      select(!c(Genus, magOTU)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix
drep_dist_ordered <- matrix %>%
                      as.data.frame() %>%
                        rownames_to_column(var = "reference") %>%
                          pivot_longer(!reference, values_to = "ani", names_to = "querry") %>%
                            mutate(ani = as.numeric(ani))


drep_dist <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% vis_magOTUs_df$ID) %>%
                  left_join(vis_magOTUs_df %>% ungroup() %>% select(ID, magOTU, Genus), by = c("reference" = "ID"))
```

## Initialization for 07 xxx

### Definitions and functions

```{r 07-func_def}
get_genecount_matrix <- function(my_genus, scps_only = F, upper_count = 10, combined_info = combined_info_df, mags_info = vis_magOTUs_df, only_mags = T, with_isolates = F) {
  # change motupsn paths as needed!
  if (with_isolates & !only_mags) {
    OG_numbers_df <- read.csv(paste0("results/11_phylogenies/02_orthofinder_results/", my_genus, "/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
    # motupan_out <- paste0("/scratch/aprasad/211018_Medgenome_india_samples/15_FurtherProcessing/motupan_test/",my_genus,"/mOTUpan.tsv")
  }
  # if (!with_isolates && only_mags) {
  #   OG_numbers_df <- read.csv(paste0("database/MAGs_database_Orthofinder/", my_genus,"/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  #   # motupan_out <- paste0("/scratch/aprasad/211018_Medgenome_india_samples/15_FurtherProcessing/motupan_test/",my_genus,"/mOTUpan.tsv")
  # }
  # if (file.exists(motupan_out)) {
  #   # motupan_df <- read.csv(motupan_out, skip = 16, sep = "\t") %>%
  #                   column_to_rownames("trait_name") %>%
  #                   select(type) %>%
  #                   rownames_to_column("OG")
  # }
  single_copy_OGs <- OG_numbers_df %>%
                    select(!Total) %>%
                    filter(if_all(is.numeric, ~ .x <= 1)) %>%
                    pull(Orthogroup)
  OG_status_summary <- OG_numbers_df %>%
                        column_to_rownames("Orthogroup") %>%
                        select(!Total) %>%
                        mutate(across(is.numeric, ~+as.logical(.x))) %>%
                        select(starts_with(c("MAG"))) %>%
                          rownames_to_column("OG") %>%
                          mutate(scp = ifelse(OG %in% single_copy_OGs, 1, 0)) %>%
                          mutate(core = scp) %>%
                          # left_join(motupan_df) %>%
                          # mutate(type_motupan = type) %>%
                          # select(!type) %>%
                          select(!matches("MAG_"))
  if (scps_only) {
    OG_numbers_df_core_isolate <- OG_numbers_df %>%
                      column_to_rownames("Orthogroup") %>%
                        select(!Total) %>%
                        filter(if_all(is.numeric, ~ .x <= 1)) %>%
                          t() %>%
                          as.data.frame() %>%
                          rownames_to_column("ID") %>%
                          mutate(Type = ifelse(startsWith(ID, "MAG_"), "MAG", "Isolate")) %>%
                          left_join(combined_info %>%
                                              ungroup() %>%
                                                select(ID, Host, Cluster) %>%
                                                unique(), by = "ID") %>%
                          left_join(vis_magOTUs_df %>%
                                              ungroup() %>%
                                                select(ID, completeness) %>%
                                                unique(), by = "ID") %>%
                            mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>%
                            mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>%
                            arrange(Type, Host, Cluster, completeness) %>%
                            select(!c(Type, Host, Cluster, completeness)) %>%
                            column_to_rownames("ID") %>% 
                            as.matrix
  } else {
    OG_numbers_df_core_isolate <- OG_numbers_df %>%
                      select(!Total) %>%
                      column_to_rownames("Orthogroup") %>%
                          t() %>%
                          as.data.frame() %>%
                          rownames_to_column("ID") %>%
                          mutate(Type = ifelse(startsWith(ID, "MAG_"), "MAG", "Isolate")) %>%
                          left_join(combined_info %>%
                                              ungroup() %>%
                                                select(ID, Host, Cluster) %>%
                                                unique(), by = "ID") %>%
                          left_join(vis_magOTUs_df %>%
                                              ungroup() %>%
                                                select(ID, completeness) %>%
                                                unique(), by = "ID") %>%
                            mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>%
                            mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>%
                            arrange(Type, Host, Cluster, completeness) %>%
                            select(!c(Type, Host, Cluster, completeness)) %>%
                            column_to_rownames("ID") %>% 
                            as.matrix
    OG_status_summary <- OG_status_summary %>%
                          filter(OG %in% single_copy_OGs)
  }
    types_df <- OG_status_summary %>%
                  mutate(type = ifelse(core == 0.5, "half_core", "accessory")) %>%
                  mutate(type = ifelse(core == 1, "core", type)) %>%
                  select(!core)
    return(list(matrix = OG_numbers_df_core_isolate, info = types_df))
}

pretty_orthogene_heatmaps <- function(matrix, do_motupan_anno = F, motupan_info = NA, combined_info=combined_info_df, ref_info=reference_genomes_info, mags_info = vis_magOTUs_df_all, row_split = "magotus", add_values = F, value_size = 10, limit_l = 0.95, limit_h = 1, col_fun ) {
  mag_completeness <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(mags_info %>%
                          ungroup() %>%
                            select(ID, completeness) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(completeness = ifelse(is.na(completeness), 100, completeness)) %>% 
                            pull(completeness)
  Representative <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(mags_info %>%
                          ungroup() %>%
                            select(ID, Representative) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(Representative = ifelse(is.na(Representative), 1, Representative)) %>% 
                            pull(Representative)
  magOTU_names <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(combined_info_df %>%
                          ungroup() %>%
                            select(ID, magOTU) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(magOTU = ifelse(is.na(magOTU), "Unknown", magOTU)) %>% 
                            pull(magOTU)
  host_names <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(combined_info_df %>%
                          ungroup() %>%
                            select(ID, Host) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>% 
                            pull(Host)
  genome_types <- rownames(matrix) %>% as.data.frame() %>%
                        mutate(Type = ifelse(startsWith(., "MAG_"), "MAG", "Isolate")) %>% pull(Type)
  anno_completeness_ref_row = rowAnnotation(Completeness = mag_completeness,
                                          `Reference status` = Representative,
                                            col = list(Completeness = colorRamp2(c(0, 50, 100), c("#d73027", "#ffffbf", "#1a9850")),
                                                      `Reference status` = make_col_list(Representative, "Greys")
                                            ),
                                            border = TRUE
                               )
  anno_host_row = rowAnnotation(Host = host_names,
                                col = list(Host = make_col_list(host_names, "Pastel1")),
                                border = TRUE
                               )
  anno_type_host_row = rowAnnotation(Host = host_names,
                                    `Genome type` = genome_types,
                                      col = list(`Genome type` = make_col_list(genome_types, "Spectral"),
                                                Host = make_col_list(host_names, "Pastel1")
                                                ),
                                      border = TRUE
                          )
  row_split  = as.character(as.factor(magOTU_names))
    heatmap_obj = Heatmap(matrix, 
            col = col_fun,
            # col = brewer.pal(3, "Set1")[1:2],
            # clustering_method_columns = 'ward.D2',
            # top_annotation = anno_type_col,
            right_annotation = anno_completeness_ref_row,
            left_annotation = anno_type_host_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 0),
            heatmap_legend_param = list(title = "Gene count"),
            row_split = row_split,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            cluster_columns = T,
            border = TRUE,
            cluster_rows = F
            )
  return(heatmap_obj)
  draw(heatmap_obj, merge_legend = TRUE)
}
```

### Load and parse data

```{r 07-Load}
files <- list.files("results/11_phylogenies/02_orthofinder_results/")
genera_list <- files[which(sapply(files, function(x) !grepl("[.-]", x)))]

all_Isolates <- c()
all_MAGs <- c()
Number_genomes_df <- data.frame()
Groups = genera_list
for (group in Groups) {
  OG_numbers_df <- read.csv(paste0("results/11_phylogenies/02_orthofinder_results/", group, "/Results_", group, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  genomes <- head(colnames(OG_numbers_df)[-1], -1)
  MAGs <- genomes[which(grepl("MAG", genomes, fixed=TRUE))]
  all_MAGs <- c(all_MAGs, MAGs)
  Isolates <- genomes[which(!grepl("MAG", genomes, fixed=TRUE))]
  all_Isolates <- c(all_Isolates, Isolates)
  Number_genomes_df <- rbind(Number_genomes_df,
                             cbind(
                                MAGs = length(MAGs),
                                Isolates = length(Isolates),
                                genomes = length(genomes),
                                Group = group
                              )
                            )
}
Number_genomes_df <- pivot_longer(Number_genomes_df, !Group, names_to = "Type", values_to = "Number") %>%
                      mutate(Number = as.numeric(Number))
# ortho_mags_df <- data.frame()
# ortho_mags_filt_df <- data.frame()
# for (genus_name in genera_list) {
#   orthofinder_filtered_summary <- read.csv(paste0("database/MAGs_database_Orthofinder/", genus_name,"_Orthogroups_filtered_summary.csv"))
#   orthofinder_summary <- read.csv(paste0("database/MAGs_database_Orthofinder/", genus_name,"_Orthogroups_summary.csv"))
#   ortho_mags_df <- rbind(ortho_mags_df, orthofinder_summary)
#   ortho_mags_filt_df <- rbind(ortho_mags_filt_df, orthofinder_filtered_summary)
# }
# ortho_w_isolates_df <- data.frame()
# for (group in Groups) {
#   orthofinder_summary_w_isolates <- read.csv(paste0("11_phylogenies/02_orthofinder_results", group, "_Orthogroups_summary.csv"))
#   ortho_w_isolates_df <- rbind(ortho_w_isolates_df, orthofinder_summary_w_isolates)
# }

# Orthogroups_summary_df <- data.frame()
# for (group in Groups) {
#   OG_summary <- read.csv(paste0("11_phylogenies/02_orthofinder_results", group, "_Orthogroups_summary.csv"), sep = ",")
#   Orthogroups_summary_df <- rbind(Orthogroups_summary_df,
#                                   OG_summary
#                             ) %>%
#                               filter(present_in != "None")
# }
# ortho_w_isolates_df <- ortho_w_isolates_df %>%
#                         left_join(cluster_tax_info, by = c("group" = "Genus")) %>%
# ortho_mags_df <- ortho_mags_df %>%
#                         left_join(cluster_tax_info, by = c("group" = "Genus")) %>%
# ortho_mags_filt_df <- ortho_mags_filt_df %>%
#                         left_join(cluster_tax_info, by = c("group" = "Genus")) %>%
# plot_number_OGs <- ortho_w_isolates_df %>%
#   mutate(Type = ifelse(status %in% c("Isolates:ScpCore ; MAGs:ScpCore0.5"), "Present in at least 50% MAGs", NA)) %>%
#   mutate(Type = ifelse(status %in% c("Isolates:ScpCore ; MAGs:ScpCore"), "Present in all MAGs", Type)) %>%
#   filter(!is.na(Type)) %>%
#     group_by(group, Type) %>%
#      reframe(Number = n())
# plot_number_OGs_mags <- ortho_mags_df %>%
#   mutate(Type = ifelse(status %in% c("; MAGs:ScpCore0.5"), "Present in at least 50% MAGs", NA)) %>%
#   mutate(Type = ifelse(status %in% c("; MAGs:ScpCore"), "Present in all MAGs", Type)) %>%
#   filter(!is.na(Type)) %>%
#     group_by(group, Type) %>%
#     filter(group %in% Groups) %>%
#     group_by(group, Type) %>%
#      reframe(Type, Number = n())
# plot_number_OGs_mags %>%
#                     select(group) %>% unique %>%
#                     left_join(Number_genomes_df %>%
#                                 pivot_wider(names_from = Type, values_from = Number), by = c("group" = "Group")) %>%
#                         ungroup() %>%
#                         filter(group %in% Groups) %>%
#                            mutate(SampleID = row_number()) %>% pull(SampleID)
```

## Initialization for 08 plot DRAM results

```{r 08-fun_def}
pcoa_plot <- function(df_pcoa, metadata, variable, color_add=F, color_list, colname_in_metadata = "ID", shape_add = F, shape_var) {
          matrix <- as.matrix(df_pcoa)
          dist <- as.dist(matrix)
          res_pcoa <- pcoa(dist)
          ev1 <- res_pcoa$vectors[,1]
          ev2 <- res_pcoa$vectors[,2]
          df_pcoa_new <- data.frame(cbind(ev1,ev2))
          df_pcoa_new$Sample <- rownames(df_pcoa_new)
          rownames(df_pcoa_new) <- NULL
          df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = colname_in_metadata))
          perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
          axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
          axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
          if(color_add & shape_add) {
            p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       shape = get(shape_var),
                                       colour = get(variable)))+
                geom_point(stat="identity", size=4) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable, shape = shape_var) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 7, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
          } else {
            if (color_add) {
              p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       colour = get(variable)))+
                geom_point(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 7, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
            } else {
              p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       color = get(variable)))+
                geom_point(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme( max_colors = length(unique(df_pcoa_new[, variable])), guide_nrow = 7, leg_size = 10 ) 
            }
          }
          return(p)
}
pcoa_plot_jitter <- function(df_pcoa, metadata, variable, color_add=F, color_list, colname_in_metadata = "ID", shape_add = F, shape_var) {
          matrix <- as.matrix(df_pcoa)
          dist <- as.dist(matrix)
          res_pcoa <- pcoa(dist)
          ev1 <- res_pcoa$vectors[,1]
          ev2 <- res_pcoa$vectors[,2]
          df_pcoa_new <- data.frame(cbind(ev1,ev2))
          df_pcoa_new$Sample <- rownames(df_pcoa_new)
          rownames(df_pcoa_new) <- NULL
          df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = colname_in_metadata))
          perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
          axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
          axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
          if(color_add & shape_add) {
            p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       shape = get(shape_var),
                                       colour = get(variable)))+
                geom_jitter(stat="identity", size=4) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable, shape = shape_var) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 7, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
          } else {
            if (color_add) {
              p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       colour = get(variable)))+
                geom_jitter(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 7, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
            } else {
              p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       color = get(variable)))+
                geom_jitter(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme( max_colors = length(unique(df_pcoa_new[, variable])), guide_nrow = 7, leg_size = 10 ) 
            }
          }
          return(p)
}
make_col_list_species <- function(my_vec, my_palette = "Spectral") {
  uniq_names <- unique(my_vec)
  num_ele <- length(uniq_names)
  if (num_ele <= 3) {
    diff_ele = 3 - num_ele
    num_ele = 3
    uniq_names <- c(uniq_names, rep("s__", diff_ele))
    cols_used <- brewer.pal(5, my_palette)[1:3]
  }
  if (num_ele <= 9 & num_ele > 3) {
    cols_used <- brewer.pal(num_ele, my_palette)
  } else {
    if (my_palette == "Pastel2") {
      cols_used <- colorRampPalette(brewer.pal(8, my_palette))(num_ele)
    } else {
      cols_used <- colorRampPalette(brewer.pal(9, my_palette))(num_ele)
    }
  }
  col_list <- c(cols_used)
  names(col_list) = uniq_names
  return(col_list)
}
pretty_dram_heatmap <- function(my_matrix,
                                bottom_annotation_name = "Genus",
                                col_input, clust_r = F, clust_c = T,
                                row_names_size = 10,
                                col_split_by = "Genus",
                                # or magOTU or Species
                                col_label_size = 10,
                                grid_line_col = "white") {
  my_layer_fun = function (j, i, x, y, w, h, fill) {
                grid.rect(x = x, y = y, width = w, heigh = h,
                          gp = gpar(col = grid_line_col, fill = NA)
                )
                # ind_mat = restore_matrix(j, i, x, y)
                # ind = unique(c(ind_mat[2, ], ind_mat[, 3]))
                # grid.points(x[ind], y[ind], pch = 16, size = unit(4, "mm"))

             }
  # my_cell_fun = function (j, i, x, y, w, h, fill) {
  #               grid.rect(x = x, y = y, width = w, heigh = h,
  #                         gp = gpar(col = "white", fill = NA)
  #               )
  #               grid.circle(x = x, y = y, r = abs(my_matrix[i, j]/2 * min(unit.c(w, h))),
  #                         gp = gpar(col = NA, fill = "black")
  #               )

  #            }
  genus_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus)
  species_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Species)
  magOTU_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Cluster)
  host_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                            pull(Host) %>% as.character()
  row_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus)
  anno_host_col = HeatmapAnnotation(Host = host_names,
                                          col = list(Host = host_order_color),
                                          border = T
                                    )
  anno_genus_col = HeatmapAnnotation(Genus = genus_names,
                                            col = list(Genus = genusColors_char
                                                      ),
                                            border = T
                                     )
  anno_magOTU_col = HeatmapAnnotation(magOTU = magOTU_names,
                                            col = list(magOTU = make_col_list(magOTU_names)
                                                      ),
                                            border = T
                                     )
  anno_species_col = HeatmapAnnotation(Species = species_names,
                                            col = list(Species = make_col_list_species(species_names)
                                                      ),
                                            border = T
                                     )
  anno_all_col = HeatmapAnnotation(Genus = genus_names,
                                   Species = species_names,
                                   magOTU = magOTU_names,
                                            col = list(Genus = genusColors_char,
                                                       Species = make_col_list_species(species_names, "Spectral"),
                                                       magOTU = make_col_list(magOTU_names, "Paired")
                                                        ),
                                            border = T
                                     )
  anno_g_s_col = HeatmapAnnotation(Genus = genus_names,
                                   Species = species_names,
                                            col = list(Genus = genusColors_char,
                                                       Species = make_col_list_species(species_names, "Paired")
                                                        ),
                                            border = T
                                     )
  anno_g_h_col = HeatmapAnnotation(Genus = genus_names,
                                   Host = host_names,
                                            col = list(Genus = genusColors_char,
                                                       Host =  host_order_color
                                                        ),
                                            border = T
                                     )
  anno_m_s_col = HeatmapAnnotation(Species = species_names,
                                   magOTU = magOTU_names,
                                            col = list(Species = make_col_list_species(species_names, "Spectral"),
                                                       magOTU = make_col_list(magOTU_names, "Paired")
                                                        ),
                                            border = T
                                     )
  anno_m_g_col = HeatmapAnnotation(Genus = genus_names,
                                   magOTU = magOTU_names,
                                            col = list(Genus = genusColors_char,
                                                       magOTU = make_col_list(magOTU_names, "Paired")
                                                        ),
                                            border = T
                                     )
  bottom_annotation_obj = anno_genus_col
  top_annotation_obj = anno_host_col
  if (bottom_annotation_name == "Genus") {
    bottom_annotation_obj = anno_genus_col
  }
  if (bottom_annotation_name == "Species") {
    bottom_annotation_obj = anno_species_col
  }
  if (bottom_annotation_name == "magOTU") {
    bottom_annotation_obj = anno_magOTU_col
  }
  if (bottom_annotation_name == "Genus_Species") {
    bottom_annotation_obj = anno_g_s_col
  }
  if (bottom_annotation_name == "Genus_Host") {
    bottom_annotation_obj = anno_g_h_col
    # top_annotation_obj = anno_empty()
  }
  if (bottom_annotation_name == "magOTU_Species") {
    bottom_annotation_obj = anno_g_s_col
  }
  if (bottom_annotation_name == "magOTU_Genus") {
    bottom_annotation_obj = anno_m_g_col
  }
  if (bottom_annotation_name == "All") {
    bottom_annotation_obj = anno_all_col
  }
  if (col_split_by == "magOTU") {
    column_split = as.character(as.factor(magOTU_names))
  }
  if (col_split_by == "Genus") {
    column_split = as.character(as.factor(genus_names))
  }
  if (col_split_by == "Species") {
    column_split = as.character(as.factor(species_names))
  }
  if (col_split_by == "Host") {
    column_split = as.character(as.factor(host_names))
  }
  dram_heatmap_obj = Heatmap(t(my_matrix),
            col = col_input,
            top_annotation = top_annotation_obj,
            bottom_annotation = bottom_annotation_obj,
            # left_annotation = anno_host_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = row_names_size),
            heatmap_legend_param = list(title = "Number genes copies"),
            layer_fun = my_layer_fun,
            # cell_fun = my_cell_fun,
            cluster_columns = clust_c,
            column_split = column_split,
            column_title_rot = 90,
            column_title_gp = grid::gpar(fontsize = col_label_size),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = clust_r
            )
  draw(dram_heatmap_obj, merge_legend = TRUE)
}
logic_to_num <- function(x) {
  if (x) {
    return(1)
  }
  return(0)
}
get_matrix_from_dram <- function(dram_df, goi = genes_of_interest, convert_logic_to_num = F) {
  # use convert_logic _to_num if working with product tsv output
  if (convert_logic_to_num) {
    dram_df_mat <- dram_df %>%
                  select(!c(genome)) %>%
                              t %>%
                              as.data.frame %>%
                              rowwise() %>% 
                              mutate(across(everything(), 
                                     ~ifelse(. == "True", "1", .))) %>%
                              mutate(across(everything(), 
                                     ~ifelse(. == "False", "0", .)))
    dram_df_mat <- dram_df_mat %>%
                    mutate(across(where(is.character), as.numeric))
    colnames(dram_df_mat) <- dram_df$genome
  } else {
    dram_df_mat <- dram_df %>%
                  select(!c(gene_id, gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
    dram_df_mat <- dram_df_mat[-1,] %>%
                      mutate_if(is.character,as.numeric)
    colnames(dram_df_mat) <- dram_df$gene_id
  }
  return(dram_df_mat)
}
prepare_matrix <- function(my_matrix, subset_genes = T, genes_list = c(),
                           subset_mags = F, mags_info_df = vis_magOTUs_df_all,
                           mags_list = c(), by_genus = F,
                           genus_list = c(), row_sum_cutoff = 0
                          ) {
                    # for now, get_rid of duplicates - there are ~ 120
                    # later find a better way to deal with them
                    duplicated_names <- names(my_matrix)[names(my_matrix) %>% duplicated]
                    my_matrix <- my_matrix %>%
                                    select(-one_of(duplicated_names))
                    if (by_genus) {
                      subset_mags = T
                      mags_list <- mags_info_df %>%
                                    ungroup() %>%
                                    filter(Genus %in% genus_list) %>%
                                    pull(ID)
                    }
                    if (subset_mags) {
                      my_matrix <- my_matrix %>%
                                    filter(rownames(.) %in% mags_list)
                    }
                    if (subset_genes) {
                      my_matrix <- my_matrix %>%
                                    select_if(names(.) %in% genes_list)
                    }
                    my_matrix <- my_matrix %>%
                                  rownames_to_column("rownames") %>%
                                  left_join(vis_magOTUs_df_all %>%
                                                ungroup() %>%
                                                select(ID, Genus, Host, Cluster),
                                              by = c("rownames" = "ID")
                                  ) %>%
                                  arrange(Genus, Cluster, Host) %>%
                                  # arrange(Host, Genus, Cluster) %>%
                                  select(!c(Genus, Host, Cluster)) %>%
                                  column_to_rownames("rownames") %>%
                                  filter(rowSums(.) > row_sum_cutoff)
}
```
### Load and parse data

```{r 08-load}
# made using scripts/count_genes_in_contigs.py
num_genes_binned <- read.csv("/scratch/aprasad/211018_Medgenome_india_samples/Figures/Number_genes_binned.csv")
metabolism_carbon <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "carbon utilization")
metabolism_transporters <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "Transporters")
metabolism_energy <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "Energy")
metabolism_nitrogen <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "Organic Nitrogen")
product_tsv <- read.csv("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/product.tsv", sep = "\t")

metabolism_nitrogen_mat <- metabolism_nitrogen %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
# colnames(metabolism_nitrogen_mat) <- metabolism_nitrogen_mat[1,]
metabolism_nitrogen_mat <- metabolism_nitrogen_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
dist_matrix_nitrogen <- vegdist(metabolism_nitrogen_mat, method = "bray")
metabolism_energy_mat <- metabolism_energy %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
# colnames(metabolism_energy_mat) <- metabolism_energy_mat[1,]
metabolism_energy_mat <- metabolism_energy_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
dist_matrix_energy <- vegdist(metabolism_energy_mat, method = "bray")
metabolism_transporters_mat <- metabolism_transporters %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
metabolism_transporters_mat <- metabolism_transporters_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
dist_matrix_transporters <- vegdist(metabolism_transporters_mat, method = "bray")
colnames(metabolism_transporters_mat) <- metabolism_transporters_mat[1,]
metabolism_carbon_mat <- metabolism_carbon %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
metabolism_carbon_mat <- metabolism_carbon_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
colnames(metabolism_carbon_mat) <- metabolism_carbon_mat[1,]
dist_matrix_carbon <- vegdist(metabolism_carbon_mat, method = "bray")
metabolism_carbon_df <- metabolism_carbon %>%
                              mutate(Type = ifelse(gene_id %in% c(genes_of_interest), "pectin", "other")) %>%
                              # select(!c(gene_description, module, header, subheader)) %>%
                              group_by(gene_id) %>%
                              reframe(subheader, across(where(is.numeric), sum)) %>% t %>% as.data.frame

colnames(metabolism_carbon_df) <- metabolism_carbon_df[1,]
metabolism_carbon_df <- metabolism_carbon_df[-1,] %>%
                                    mutate_if(is.character,as.numeric) %>%
                                    rownames_to_column("ID") %>%
                                    left_join(vis_magOTUs_df_all)

dist_matrix_product_pathway <- vegdist(product_tsv %>%
                                select(c(1:14)) %>%
                                  column_to_rownames("genome"))
dist_matrix_product_others <- vegdist(product_tsv %>%
                                select(c(1,34:52)) %>%
                                  column_to_rownames("genome") %>% mutate(across(!where(is.numeric), Vectorize(logic_to_num))) %>%
                                  filter(rowSums(.) > 0), 
                                  method = "jaccard"
                                  )
df_other <- product_tsv %>%
                  select(c(1,34:52)) %>%
                   column_to_rownames("genome") %>%
                    mutate(across(!where(is.numeric), Vectorize(logic_to_num))) %>%
                    filter(rowSums(.) > 0) %>%
                      as.matrix
```

## Initialization for 09 plot community composition

### Definitions and functions

```{r 09-def_func}
shorten_list <- function(covs_list_full, start = 1, end = 2) {
  covs_list_short = strsplit(covs_list_full, ",")[[1]][start:end]
  return(paste0(covs_list_short, collapse = ","))
}

trim_side_mean <- function(x, trim, type="both") {
    if (type == "both") {
        mean(x,trim)}
    else if (type == "right") {
        x <- sort(x)
        mean(x[1:floor(length(x)*(1-trim))])}
    else if (type == "left"){
        x <- sort(x)
        mean(x[max(1,floor(length(x)*trim)):length(x)])}
}
make_cum_curve <- function(samples_vector, pa_df, iterations, name = NA) {
  num_clusters_matrix <- matrix(nrow = length(samples_vector), ncol = iterations)
  for (iter in 1:iterations) {
    clusters_found_cumulative <- c()
    for (num_samples in 1:length(samples_vector)) {
      num_new_clusters = 0
      selected_samples <- sample(samples_vector, num_samples)
      clusters_found <- colnames(pa_df[selected_samples, which(colSums(pa_df[selected_samples, ]) > 1)])
      for (cluster in clusters_found) {
        if (cluster %in% clusters_found_cumulative) {
          invisible()
        } else {
          num_new_clusters <- num_new_clusters + 1
          clusters_found_cumulative <- c(clusters_found_cumulative, cluster)
        }
      }
      num_clusters_matrix[num_samples, iter] = length(clusters_found_cumulative)
    }
  }
  num_clusters_df <- as.data.frame(num_clusters_matrix)
  colnames(num_clusters_df) <- do.call(function(x) paste0("curve_", x), list(c(1:iterations)))
  num_clusters_df <- cbind(sample_size = c(1:length(samples_vector)), num_clusters_df)
  plot_cum_curve_df <- pivot_longer(num_clusters_df, !sample_size, values_to = "number_of_clusters", names_to = "curve")
  plot_cum_curve_df <- cbind("name" = name, plot_cum_curve_df)
  return(plot_cum_curve_df)
}
pretty_pcoa_plot <- function(df_pcoa, metadata, variable, color_add=F, color_list, colname_in_metadata = "ID") {
          matrix <- as.matrix(df_pcoa)
          dist <- as.dist(matrix)
          res_pcoa <- pcoa(dist)
          ev1 <- res_pcoa$vectors[,1]
          ev2 <- res_pcoa$vectors[,2]
          df_pcoa_new <- data.frame(cbind(ev1,ev2))
          df_pcoa_new$Sample <- rownames(df_pcoa_new)
          rownames(df_pcoa_new) <- NULL
          df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = colname_in_metadata))
          perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
          axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
          axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
          if(color_add) {
            p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       colour = get(variable)))+
                geom_point(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 4, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
          } else {
            p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       color = get(variable)))+
                geom_point(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme( max_colors = length(unique(df_pcoa_new[, variable])), guide_nrow = 4, leg_size = 10 ) 
          }
          return(p)
}
get_ID_from_scaffold <- function(my_name) {
  split_name <- strsplit(my_name, "|", fixed = T)[[1]]
  if (length(split_name) == 3) {
    return(paste0(strsplit(split_name[[3]], "_")[[1]][1:3], collapse="_"))
  } else {
    return(my_name)
  }
}
remove_extension <- function(name, extension) {
  return(strsplit(name, extension)[[1]][[1]])
}
```

### Load and parse data

```{r 09-load_parse}
read_length <- 150
instrain_genomes_info_df <- data.frame()
for (sample in samples_IN_MY) {
  genome_info_path <- paste0("results/10_instrain/02_instrain_profile/", sample, "/output/", sample, "_genome_info.tsv")
  if (file.exists(genome_info_path)) {
    print(paste0("Reading ", sample))
    df_read <- read.csv(genome_info_path, sep = "\t") %>%
      mutate(genome = Vectorize(remove_extension)(genome, ".fna")) %>%
        left_join(vis_magOTUs_df_all %>% ungroup() %>% select(ID, magOTU, Genus, Representative, Quality), by = c("genome" = "ID")) %>%
          mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
            mutate(magOTU = paste0(Genus, "_", magOTU, "-", genome))
    instrain_genomes_info_df <- bind_rows(instrain_genomes_info_df, cbind(df_read, "sample"=sample))
  }
}

instrain_genomes_info_df_all <- data.frame()
for (sample in samples) {
  genome_info_path <- paste0("results/10_instrain/02_instrain_profile/", sample, "/output/", sample, "_genome_info.tsv")
  if (file.exists(genome_info_path)) {
    print(paste0("Reading ", sample))
    df_read <- read.csv(genome_info_path, sep = "\t") %>%
      mutate(genome = Vectorize(remove_extension)(genome, ".fna")) %>%
        left_join(vis_magOTUs_df_all %>% ungroup() %>% select(ID, magOTU, Genus, Representative, Quality), by = c("genome" = "ID")) %>%
          mutate(Host = Vectorize(get_host_from_sample_name)(sample)) %>%
            mutate(magOTU = paste0(Genus, "_", magOTU, "-", genome))
    instrain_genomes_info_df_all <- bind_rows(instrain_genomes_info_df_all, cbind(df_read, "sample"=sample))
  }
}
breadth_cutoff = 0.5

instrain_breadth_depth <- instrain_genomes_info_df %>%
                            group_by(sample, magOTU) %>%
                            reframe(coverage_median, Host, breadth, length, sample, Genus, filtered_read_pair_count, 
                                      reads_unfiltered_pairs, reads_unfiltered_reads) %>%
                            filter(breadth > breadth_cutoff) %>%
                              mutate(coverage_factor = cut(coverage_median ,breaks=c(-1, 1e-04, 1e-03, 1e-02, 1e-01, 1, 10, 100, 1000, Inf),labels = c("<0.0001", "0.0001 - 0.001", "0.001 - 0.01", "0.01 - 0.1", "0.1 - 1", "1 - 10", "10 - 100", "100 - 1000", ">1000"))) %>%
                              mutate(expected_number_of_genomes = filtered_read_pair_count*read_length*2/(length)) %>%
                              ungroup() %>%
                              group_by(sample) %>%
                              mutate(percentage_reads = filtered_read_pair_count/sum(filtered_read_pair_count)*100,
                                     percentage_expected_number_of_genomes = expected_number_of_genomes/sum(expected_number_of_genomes)*100
                              ) %>%
                              ungroup()
instrain_breadth_depth_unfiltered <- instrain_genomes_info_df %>%
                            group_by(sample, magOTU) %>%
                            reframe(coverage_median, Host, breadth, length, sample, Genus, filtered_read_pair_count, 
                                      reads_unfiltered_pairs, reads_unfiltered_reads) %>%
                              mutate(coverage_factor = cut(coverage_median ,breaks=c(-1, 1e-04, 1e-03, 1e-02, 1e-01, 1, 10, 100, 1000, Inf),labels = c("<0.0001", "0.0001 - 0.001", "0.001 - 0.01", "0.01 - 0.1", "0.1 - 1", "1 - 10", "10 - 100", "100 - 1000", ">1000"))) %>%
                              mutate(expected_number_of_genomes = filtered_read_pair_count*read_length*2/(length)) %>%
                              ungroup() %>%
                              group_by(sample) %>%
                              mutate(percentage_reads = filtered_read_pair_count/sum(filtered_read_pair_count)*100,
                                     percentage_expected_number_of_genomes = expected_number_of_genomes/sum(expected_number_of_genomes)*100
                              ) %>%
                              ungroup()

instrain_breadth_depth_all <- instrain_genomes_info_df_all %>%
                            group_by(sample, magOTU) %>%
                            reframe(coverage_median, Host, breadth, length, sample, Genus, filtered_read_pair_count, 
                                      reads_unfiltered_pairs, reads_unfiltered_reads) %>%
                            filter(breadth > breadth_cutoff) %>%
                              mutate(coverage_factor = cut(coverage_median ,breaks=c(-1, 1e-04, 1e-03, 1e-02, 1e-01, 1, 10, 100, 1000, Inf),labels = c("<0.0001", "0.0001 - 0.001", "0.001 - 0.01", "0.01 - 0.1", "0.1 - 1", "1 - 10", "10 - 100", "100 - 1000", ">1000"))) %>%
                              mutate(expected_number_of_genomes = filtered_read_pair_count*read_length*2/(length)) %>%
                              ungroup() %>%
                              group_by(sample) %>%
                              mutate(percentage_reads = filtered_read_pair_count/sum(filtered_read_pair_count)*100,
                                     percentage_expected_number_of_genomes = expected_number_of_genomes/sum(expected_number_of_genomes)*100
                              ) %>%
                              ungroup()
instrain_breadth_depth_unfiltered_all <- instrain_genomes_info_df_all %>%
                            group_by(sample, magOTU) %>%
                            reframe(coverage_median, Host, breadth, length, sample, Genus, filtered_read_pair_count, 
                                      reads_unfiltered_pairs, reads_unfiltered_reads) %>%
                              mutate(coverage_factor = cut(coverage_median ,breaks=c(-1, 1e-04, 1e-03, 1e-02, 1e-01, 1, 10, 100, 1000, Inf),labels = c("<0.0001", "0.0001 - 0.001", "0.001 - 0.01", "0.01 - 0.1", "0.1 - 1", "1 - 10", "10 - 100", "100 - 1000", ">1000"))) %>%
                              mutate(expected_number_of_genomes = filtered_read_pair_count*read_length*2/(length)) %>%
                              ungroup() %>%
                              group_by(sample) %>%
                              mutate(percentage_reads = filtered_read_pair_count/sum(filtered_read_pair_count)*100,
                                     percentage_expected_number_of_genomes = expected_number_of_genomes/sum(expected_number_of_genomes)*100
                              ) %>%
                              ungroup()

magOTU_tax_info_all <- instrain_breadth_depth_unfiltered_all %>%
                    select(Genus, magOTU) %>% unique()
magOTU_tax_info <- instrain_breadth_depth_unfiltered %>%
                    select(Genus, magOTU) %>% unique()


# adjust coverage for sequencing depth
length(df_reads$Sample)
min_reads = min(df_reads$MAGs_DB_mf)
# for normalizing coverage across samples

min_copies <- min(qpcr_plot_df$copy_num)
max_copies <- max(qpcr_plot_df$copy_num)
df_copy_norm <- qpcr_plot_df %>% 
                  group_by(Host) %>% 
                    summarise(Median_cp = median(copy_num)) %>% 
                    # summarise(Median_cp = median(copy_num)/min_copies) %>% 
                      unique() #%>%
                      # mutate(Median_cp = Median_cp/min(Median_cp))
# # get the median copy number for that host species
# qpcr_normalizing_factor <- function(x) {
#   num = df_copy_norm[df_copy_norm$Host== x, "Median_cp"][[1]]
#   return(num)
# }
# get the copy number of that sample if available and if not, get the median copy number for that species 
qpcr_normalizing_factor_indiv <- function(sample) {
  if (sample %in% qpcr_plot_df$Sample.Name) {
    num = qpcr_plot_df[qpcr_plot_df$Sample.Name == sample, "copy_num"] %>% pull
  } else {
    h = get_host_from_sample_name(sample)
    num = df_copy_norm[df_copy_norm$Host== h, "Median_cp"][[1]]
    print(paste0("gettin median for ", sample, " as it is not found in the qpcr data"))
  }
  return(num)
}

abundance_df <-  instrain_breadth_depth_unfiltered %>%
                                filter(!is.na(magOTU)) %>%
                                left_join(df_reads %>% select(Sample, MAGs_DB_mf), by = c("sample" = "Sample")) %>%
                                mutate(coverage = coverage_median) %>%
                                group_by(sample) %>%
                                mutate(rel_cov = coverage/sum(coverage)) %>%
                                ungroup() %>%
                                mutate(copies_cov = rel_cov*Vectorize(qpcr_normalizing_factor_indiv)(sample))
                                  # for now not doing any adjusting - using median coverage as is
                                  # # normalize by sequencing depth (of bacterial reads)
                                  # mutate(coverage_adj = coverage_median/MAGs_DB_mf*min_reads) %>%
                                  # # normalize by qpcr - community size
                                  # mutate(coverage_norm = coverage_adj/max_copies*Vectorize(qpcr_normalizing_factor_indiv)(sample)) %>%
                                  # mutate(coverage = coverage_norm) %>%
                                  #   select(!MAGs_DB_mf) %>%
                                  #     unique()
abundance_df_matrix <-  abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            select(sample, magOTU, coverage) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=coverage, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))

abundance_df_matrix_copies <-  abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            select(sample, magOTU, copies_cov) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=copies_cov, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))

abundance_df_matrix_rel <-  abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            select(sample, magOTU, rel_cov) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=rel_cov, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))

abundance_df_matrix_pa <-  abundance_df %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage >= 1 & breadth > breadth_cutoff, 1, 0))  %>%
                            group_by(magOTU) %>%
                            # count number of samples where magOTU is detected
                            mutate(numSamples = sum(detected)) %>%
                            # only keep magOTUs that are detected in more than 3 samples
                            filter(numSamples > 3) %>%
                            ungroup() %>%
                            select(sample, magOTU, detected) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=detected, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))
write.csv(abundance_df_matrix_copies, "results/figures/magOTU_abundance_matrix_copies.csv", row.names = T, quote = F)
write.csv(abundance_df_matrix_rel, "results/figures/magOTU_abundance_matrix_rel.csv", row.names = T, quote = F)
write.csv(abundance_df_matrix_pa, "results/figures/magOTU_abundance_matrix_presence_absence.csv", row.names = T, quote = F)


abundance_df_all <-  instrain_breadth_depth_unfiltered_all %>%
                                filter(!is.na(magOTU)) %>%
                                left_join(df_reads %>% select(Sample, MAGs_DB_mf), by = c("sample" = "Sample")) %>%
                                mutate(coverage = coverage_median) %>%
                                group_by(sample) %>%
                                mutate(rel_cov = coverage/sum(coverage)) %>%
                                ungroup() %>%
                                mutate(copies_cov = rel_cov*Vectorize(qpcr_normalizing_factor_indiv)(sample))
                                  # for now not doing any adjusting - using median coverage as is
                                  # # normalize by sequencing depth (of bacterial reads)
                                  # mutate(coverage_adj = coverage_median/MAGs_DB_mf*min_reads) %>%
                                  # # normalize by qpcr - community size
                                  # mutate(coverage_norm = coverage_adj/max_copies*Vectorize(qpcr_normalizing_factor_indiv)(sample)) %>%
                                  # mutate(coverage = coverage_norm) %>%
                                  #   select(!MAGs_DB_mf) %>%
                                  #     unique()
abundance_df_matrix_all <-  abundance_df_all %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            select(sample, magOTU, coverage) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=coverage, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))

abundance_df_matrix_copies_all <-  abundance_df_all %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            select(sample, magOTU, copies_cov) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=copies_cov, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))

abundance_df_matrix_rel_all <-  abundance_df_all %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            select(sample, magOTU, rel_cov) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=rel_cov, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))

abundance_df_matrix_pa_all <-  abundance_df_all %>% filter(!is.na(magOTU)) %>%
                            filter(breadth > breadth_cutoff) %>%
                            mutate(detected = ifelse(coverage > 1 & breadth > breadth_cutoff, 1, 0))  %>%
                            group_by(magOTU) %>%
                            # count number of samples where magOTU is detected
                            mutate(numSamples = sum(detected)) %>%
                            # only keep magOTUs that are detected in more than 3 samples
                            filter(numSamples > 3) %>%
                            ungroup() %>%
                            select(sample, magOTU, detected) %>%
                            unique() %>%
                              pivot_wider(names_from=magOTU, values_from=detected, values_fn = function(x) as.numeric(x)) %>% remove_rownames %>%
                              column_to_rownames(var="sample") %>% 
                              mutate_all(~replace_na(.,0))
write.csv(abundance_df_matrix_all, "results/figures/magOTU_abundance_matrix_coverage_all.csv", row.names = T, quote = F)
write.csv(abundance_df_matrix_pa_all, "results/figures/magOTU_abundance_matrix_presence_absence_all.csv", row.names = T, quote = F)

sample_data_mat <- df_meta_complete %>%
                   filter(ID %in% samples_MY) %>%
                    mutate(Host = Vectorize(get_host_from_sample_name)(ID)) %>%
                    column_to_rownames("ID")
# colnames(abundance_df$magOTU)
get_rep_genome_magotu <- function(magOTU_name, genus_name) {
  temp = strsplit(magOTU_name, paste0(genus_name, "_"))[[1]][2]
  temp_split = strsplit(temp, paste0("-"))[[1]]
  genome = paste0(temp_split[2:length(temp_split)], collapse = "-")
  return(genome)
} 
otu_data_mat <- abundance_df %>%
                  filter(breadth > breadth_cutoff & coverage > 1) %>%
                  mutate(representative_genome = Vectorize(get_rep_genome_magotu)(magOTU, Genus)) %>%
                    select(sample, coverage, representative_genome) %>%
                        unique %>%
                        rownames_to_column() %>%
                        pivot_wider(!c(rowname), names_from=representative_genome, values_from=coverage, values_fn = as.numeric) %>%
                        column_to_rownames(var="sample") %>%
                        mutate_all(~replace_na(.,0)) %>%
                            as.matrix                  
tax_table_mat <- df_gtdbk_bac %>% select(user_genome, classification) %>%
                    separate(into = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), col = classification, sep = ";") %>%
                    filter(user_genome %in% colnames(otu_data_mat)) %>%
                    mutate("Name" = ifelse(Species == "s__", Genus, Species)) %>%
                    mutate("Name" = ifelse(Name == "g__", Family, Name)) %>%
                    column_to_rownames("user_genome") %>% as.matrix
dim(otu_data_mat)
dim(tax_table_mat)
ps <- phyloseq(sample_data(sample_data_mat),
               otu_table(otu_data_mat, taxa_are_rows = F),
               tax_table(tax_table_mat)
)
# mags_tree_path <- "/scratch/aprasad/211018_Medgenome_india_samples/15_FurtherProcessing/gtdb_inferred_tree/gtdbtk.unrooted.tree"

# host tree made from NCBI mito sequences
# from this paper:
# https://link.springer.com/article/10.1007/s11756-022-01123-6
# Apis anreniformis	Thailand	KF736157
# Apis florea	Thailand	AP018491
# Apis dorsata	Thailand	AP018369
# Apis mellifera	carnica N/A	MN250878
# Apis cerana	Borneo Island	AP018149
host_tree_path <- "/scratch/aprasad/20230313_apis_species_comparison/data/host_tree_published.tree"
host_tree <- read.tree(host_tree_path)

# print samples not in rownames(abundance_df_matrix)
dim(abundance_df_matrix)
print("samples dropped for community analysis:")
samples[!(samples %in% rownames(abundance_df_matrix))]
samples_kept <- samples[(samples %in% rownames(abundance_df_matrix))]
```

## Initialization for 10 plot_instrain_results

```{r 10-load_and_parse}
# 
shared_genera <- vis_magOTUs_df %>%
  ungroup() %>%
  group_by(Genus) %>%
    mutate(num_genus = n_distinct(Host)) %>%
      reframe(num_genus) %>%
      filter(num_genus > 1) %>%
        pull(Genus) %>%
          unique
```

### Definitions and functions
### Load and parse data

### Definitions and functions
### Load and parse data
### Initialization for 11 xxx

### Definitions and functions
### Load and parse data
### Initialization for 12 xxx

```{r unused_code}
```

```{r 11-load_data}
df_cdhit_clusters_info <- read.csv("results/08_gene_content/00_cdhit_clustering/cluster_info.tsv", sep = "\t")
```


```{r query_dataset}
# vis_magOTUs_df %>%
#   ungroup() %>%
#   group_by(Genus) %>%
#   reframe(n=n()) %>%
#   arrange(n) %>% pull(Genus)

# # g__Lactobacillus
# # g__Bifidobacterium
# # g__Gilliamella
# vis_magOTUs_df %>%
#   filter(Genus == "g__Dysgonomonas") %>%
#   mutate(custom_name = paste0(ID, "-", Cluster)) %>%
#     pull(custom_name)
# vis_magOTUs_df %>%
#   filter(Genus == "g__Dysgonomonas") %>%
#   select(ID,Cluster,Cluster,Host,completeness) %>% print(n=50)
# # g__Bombilactobacillus
# vis_magOTUs_df %>%
#   filter(Genus == "g__Snodgrassella") %>%
#     pull(ID)
# vis_magOTUs_df %>%
#   filter(Genus == "g__Frischella") %>%
#     pull(ID)
# vis_magOTUs_df %>%
#   filter(Genus == "g__Apibacter") %>%
#     pull(ID)
# vis_magOTUs_df %>%
#   filter(Genus == "g__Pectinatus") %>%
#     pull(ID)
# vis_magOTUs_df %>%
#   filter(Genus == "g__Commensalibacter") %>%
#     pull(ID)

# vis_magOTUs_df_all %>%
#   filter(ID == "MAG_D2.4_17") %>%
#     select(length)
# instrain_breadth_depth_by_genus <- instrain_breadth_depth %>%
#                                     ungroup() %>%
#                                     group_by(sample, Genus) %>%
#                                     reframe(sample, Genus, Host, 
#                                               rel_abundance = sum(percentage_expected_number_of_genomes),
#                                               abundance_raw = sum(filtered_read_pair_count),
#                                               abundance = sum(filtered_read_pair_count)
#                                     ) %>% unique                        
# instrain_breadth_depth_by_genus_unfiltered <- instrain_breadth_depth_unfiltered %>%
#                                     ungroup() %>%
#                                     group_by(sample, Genus) %>%
#                                     reframe(sample, Genus, Host, 
#                                               rel_abundance = sum(percentage_expected_number_of_genomes),
#                                               abundance_raw = sum(filtered_read_pair_count),
#                                               abundance = sum(filtered_read_pair_count)
#                                     ) %>% unique                        
# instrain_breadth_depth_by_genus_host <- instrain_breadth_depth %>%
#                                     ungroup() %>%
#                                     group_by(Host, Genus) %>%
#                                     reframe(Genus, Host, 
#                                               rel_abundance = mean(percentage_expected_number_of_genomes),
#                                               abundance_raw = mean(filtered_read_pair_count),
#                                               abundance = mean(filtered_read_pair_count)
#                                     ) %>% unique
# instrain_breadth_depth_by_magOTU_host <- instrain_breadth_depth %>%
#                                     ungroup() %>%
#                                     group_by(Host, magOTU) %>%
#                                     reframe(magOTU, Host, 
#                                               rel_abundance = mean(percentage_expected_number_of_genomes),
#                                               abundance_raw = mean(filtered_read_pair_count),
#                                               abundance = mean(filtered_read_pair_count)
#                                     ) %>% unique
# View(instrain_breadth_depth)
# instrain_pa_by_magOTU <- instrain_breadth_depth %>%
#                                     ungroup() %>%
#                                     mutate(Present = ifelse(percentage_reads > 0, 1, 0))
# samples_am <- instrain_breadth_depth %>% filter(Host == "Apis mellifera") %>% select(sample) %>% unique() %>% unlist()
# samples_ac <- instrain_breadth_depth %>% filter(Host == "Apis cerana") %>% select(sample) %>% unique() %>% unlist()
# samples_ad <- instrain_breadth_depth %>% filter(Host == "Apis dorsata") %>% select(sample) %>% unique() %>% unlist()
# samples_af <- instrain_breadth_depth %>% filter(Host == "Apis florea") %>% select(sample) %>% unique() %>% unlist()
# samples_aa <- instrain_breadth_depth %>% filter(Host == "Apis andreniformis") %>% select(sample) %>% unique() %>% unlist()

# instrain_pa_by_magOTU_by_host <- instrain_pa_by_magOTU %>%
#                                     select(Host, Present, magOTU) %>%
#                                     unique() %>%
#                                     group_by(Host, magOTU) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis mellifera", 100*sum(Present)/length(samples_am), NA)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis cerana", 100*sum(Present)/length(samples_ac), Prevalence)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis dorsata", 100*sum(Present)/length(samples_ad), Prevalence)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis florea", 100*sum(Present)/length(samples_af), Prevalence)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis andreniformis", 100*sum(Present)/length(samples_aa), Prevalence))
# instrain_pa_by_genus_by_host <- instrain_pa_by_magOTU %>%
#                                     select(Host, Present, Genus) %>%
#                                     unique() %>%
#                                     group_by(Host, Genus) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis mellifera", 100*sum(Present)/length(samples_am), NA)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis cerana", 100*sum(Present)/length(samples_ac), Prevalence)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis dorsata", 100*sum(Present)/length(samples_ad), Prevalence)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis florea", 100*sum(Present)/length(samples_af), Prevalence)) %>%
#                                       mutate(Prevalence = ifelse(Host=="Apis andreniformis", 100*sum(Present)/length(samples_aa), Prevalence))
                                    
# mutate(Detected = ifelse(abundance_raw > 10, 1, 0)) %>%
# mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
# ungroup() %>%
# group_by(Host, Genus) %>%
# mutate(Present = sum(Detected)) %>%
# mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
# mutate(Prevalence_host = ifelse(Host=="Apis cerana",Present/length(samples_ac), Prevalence_host)) %>%
# mutate(Prevalence_host = ifelse(Host=="Apis dorsata",Present/length(samples_ad), Prevalence_host)) %>%
# mutate(Prevalence_host = ifelse(Host=="Apis florea",Present/length(samples_af), Prevalence_host))

# corecov_data_dir <- "09_MagDatabaseProfiling/CoverageEstimation/Merged"
# coords_df <- data.frame()
# for (file in list.files(corecov_data_dir)) {
#   if (endsWith(file, "_coord.txt")) {
#     df_temp <- read.csv(paste0(corecov_data_dir, "/", file), sep = "\t", header = TRUE)
#     coords_df <- rbind(coords_df, df_temp)
#   }
# }
# cluster_tax_info <- vis_magOTUs_df %>%
#                       ungroup() %>%
#                       select(Genus, Species, magOTU, Num_mags) %>%
#                       unique() %>%
#                         mutate(magOTU_name = paste0(Genus, "-", magOTU))
# samples_am <- c(coords_df %>%
#                 mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                 filter(Host == "Apis mellifera") %>%
#                 pull(Sample) %>% unique %>% as.vector)
# samples_ac <- c(coords_df %>%
#                 mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                 filter(Host == "Apis cerana") %>%
#                 pull(Sample) %>% unique %>% as.vector)
# samples_ad <- c(coords_df %>%
#                 mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                 filter(Host == "Apis dorsata") %>%
#                 pull(Sample) %>% unique %>% as.vector)
# samples_af <- c(coords_df %>%
#                 mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                 filter(Host == "Apis florea") %>%
#                 pull(Sample) %>% unique %>% as.vector)
# samples_aa <- c(coords_df %>%
#                 mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                 filter(Host == "Apis andreniformis") %>%
#                 pull(Sample) %>% unique %>% as.vector)
# coverage_df <- coords_df %>%
#   rename(Cluster = magOTU) %>%
#     left_join(cluster_tax_info) %>%
#       mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#       mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
#         mutate(Detected = ifelse(Cov > 0, 1, 0)) %>%
#           mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
#           mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
#           group_by(Host, Cluster) %>%
#             mutate(Present = sum(Detected)) %>%
#             mutate(Mean = mean(Cov)) %>%
#             mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
#             mutate(Prevalence_host = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence_host)) %>%
#             mutate(Prevalence_host = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence_host)) %>%
#             mutate(Prevalence_host = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence_host)) %>%
#             mutate(Prevalence_host = ifelse(Host=="Apis andreniformis", Present/length(samples_aa), Prevalence_host))
# coverage_df_genus <- coords_df %>%
#                 rename(Cluster = magOTU) %>%
#                   left_join(cluster_tax_info) %>%
#                       mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                       mutate(Detected = ifelse(Host != "Apis dorsata" & Cov > 1, 1, 0)) %>%
#                       mutate(Detected = ifelse(Host == "Apis dorsata" & Cov > 1, 1, Detected)) %>%
#                       # mutate(Detected = ifelse(Cov > 100, 1, 0)) %>%
#                         mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
#                         mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
#                         ungroup() %>%
                        # group_by(Sample) %>%
                        # mutate(total_cov = sum(Cov)) %>%
                        #   ungroup() %>%
                        #   group_by(Genus, Sample) %>%
                        #     mutate(rel_cov = sum(Cov)/total_cov*100) %>%
                        #     mutate(rel_cov = ifelse(total_cov == 0, 0, rel_cov)) %>%
#                               reframe(rel_cov, Genus_Present = ifelse(any(Detected == 1), 1, 0), Num_mags = sum(Num_mags), Genus_cov = trim_side_mean(Cov, trim = 0.01)) %>%
#                               mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                                 group_by(Host, Genus) %>%
#                                   mutate(Mean = round(mean(Genus_cov), 2)) %>%
                                  # mutate(Mean_rel_cov = round(mean(rel_cov), 2)) %>%
#                                   mutate(Median = round(mean(Genus_cov)), 2) %>%
#                                   mutate(covs_list = paste0(round(Genus_cov, 0), collapse = ",")) %>%
#                                   mutate(covs_list = Vectorize(shorten_list)(covs_list, 1, 3)) %>%
#                                   mutate(Present = sum(Genus_Present)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis cerana",Present/length(samples_ac), Prevalence_host)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis dorsata",Present/length(samples_ad), Prevalence_host)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis florea",Present/length(samples_af), Prevalence_host)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis andreniformis",Present/length(samples_aa), Prevalence_host))
# coverage_df_genus <- coords_df %>%
#                 rename(Cluster = magOTU) %>%
#                   left_join(cluster_tax_info) %>%
#                       mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                       mutate(Detected = ifelse(Host != "Apis dorsata" & Cov > 10, 1, 0)) %>%
#                       mutate(Detected = ifelse(Host == "Apis dorsata"& Cov > 1, 1, Detected)) %>%
#                       # mutate(Detected = ifelse(Cov > 100, 1, 0)) %>%
#                         mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
#                         mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
#                         ungroup() %>%
#                            group_by(Sample) %>%
#                         mutate(total_cov = sum(Cov)) %>%
#                           ungroup() %>%
#                           group_by(Genus, Sample) %>%
#                           # fix this part!!
#                             # mutate(rel_cov = sum(Cov)/total_cov*100) %>%
#                             # mutate(rel_cov = ifelse(total_cov == 0, 0, rel_cov)) %>%
#                             # mutate(Mean_rel_cov = round(mean(rel_cov), 2)) %>%
#                               # reframe(Genus_Present = ifelse(any(Detected == 1), 1, 0), Num_mags = sum(Num_mags), Genus_cov = Mean_rel_cov) %>%
#                               reframe(Genus_Present = ifelse(any(Detected == 1), 1, 0), Num_mags = sum(Num_mags), Genus_cov = mean(Cov)) %>%
#                                mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
#                                 group_by(Host, Genus) %>%
#                                   mutate(Mean = round(trim_side_mean(Genus_cov, trim = 0.01)), 2) %>%
#                                   mutate(Median = round(mean(Genus_cov)), 2) %>%
#                                   mutate(covs_list = paste0(round(Genus_cov, 0), collapse = ",")) %>%
#                                   # mutate(Genus_cov = Genus_cov/total_cov*100) %>%
#                                   mutate(covs_list = Vectorize(shorten_list)(covs_list, 1, 3)) %>%
#                                   mutate(Present = sum(Genus_Present)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis cerana",Present/length(samples_ac), Prevalence_host)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis dorsata",Present/length(samples_ad), Prevalence_host)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis florea",Present/length(samples_af), Prevalence_host)) %>%
#                                   mutate(Prevalence_host = ifelse(Host=="Apis andreniformis",Present/length(samples_aa), Prevalence_host))
```