---
title: "Honeybee cross-species analysis"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---
<!-- comment out for pdf compiling -->
<!-- some css for formatting html -->
<!-- <style>
    h1.title {
        font-size: 40px;
        font-family: Serif;
        text-align: center;
        font-weight: normal;
        /* color: DarkRed; */
    }
    h1 {
        font-size: 24px;
        font-family: Serif;
        font-weight: bold;
        /* font-weight: bold; */
        /* text-align: center; */
        /* color: DarkRed; */
    }
    h2 {
        font-size: 22px;
        font-family: Serif;
        font-weight: bold;
        /* text-align: center; */
        /* color: DarkRed; */
    }
    h3 {
        font-size: 20px;
        font-family: Serif;
        font-weight: bold;
        /* text-align: center; */
        /* color: DarkRed; */
    }
    body .main-container {
        /* max-width: 1000px; */
        font-size: 18px;
        font-family: Serif;
    }
</style> -->

<!-- ```{css echo=FALSE, warning=FALSE}
/* To make hoverable links. (does not have to be called hint) Usage: */
/* [Message to show on hover]{.hint} */
.hint {
  visibility: hidden;
}

.hint::before {
  visibility: visible;
  content: "Hint";
  color: blue;
}

.hint:hover {
  visibility: visible;
  font-weight: bold;
}

.hint:hover::before {
  display: none;
}
``` -->

# Introduction

Collection of files, definitions and functions for visualization scripts. The chunks on this can be run first before any of the other scripts for visualization.

## General definitions and functions

### Load libraries

Check that the working directory is specified correctly.

```{r load_libraries}
working_dir <- "/scratch/aprasad/211018_Medgenome_india_samples"
setwd(working_dir)
library(ggplot2)
library(kableExtra)
library(readxl)
library(knitr)
library(tidyverse)
library(viridis)
library(hrbrthemes)
library(ggthemes)
library(RColorBrewer)
library(scales)
library(dplyr)
library(gridExtra)
library(ggVennDiagram)
library(vegan)
library(ape)
library(ggforce)
library(VennDiagram)
library(circlize)
library(phyloseq)
library(ggdendro)
library(dendextend)
library(ComplexHeatmap)
library(ggnewscale)
library(ggsignif)
# library(ggpubr) - version confilct
```

### General definitions

Manually define vectors and lists for sample names, colors etc.

```{r general_definitions}
samples_IN <- c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5",
              "C1.1", "C1.2", "C1.3", "C1.4", "C1.5",
              "C2.1", "C2.2", "C2.3", "C2.4", "C2.5",
              "C3.1", "C3.2", "C3.3", "C3.4", "C3.5",
              "D1.1","D1.2","D1.3","D1.4","D1.5",
              "D2.1","D2.2","D2.3","D2.4","D2.5",
              "D3.1","D3.2","D3.3","D3.4","D3.5",
              "F1.1","F1.2","F1.3","F1.4","F1.5",
              "F2.1","F2.2","F2.3","F2.4","F2.5",
              "F3.1","F3.2","F3.3","F3.4","F3.5"
            )
samples_KE <- c("AmAi01", "AmAi02", "AmAi03", "AmAi04", "AmAi05",
              "AmAi06", "AmAi07", "AmAi08", "AmAi09", "AmAi10",
              "AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05",
              "AmIu06", "AmIu07", "AmIu08", "AmIu09", "AmIu10",
              "AcKn01", "AcKn02", "AcKn03", "AcKn04", "AcKn05",
              "AcKn06", "AcKn07", "AcKn08", "AcKn09", "AcKn10",
              "AcCh01", "AcCh02", "AcCh03", "AcCh04", "AcCh05",
              "AcCh06", "AcCh07", "AcCh08", "AcCh09", "AcCh10",
              "DrY1_F1", "DrY1_F2", "DrY1_F3", "DrY1_F4", "DrY1_F5", "DrY1_F6",
              "DrY1_N1", "DrY1_N2", "DrY1_N3", "DrY1_N4", "DrY1_N5", "DrY1_N6",
              "DrY1_W1", "DrY1_W2", "DrY1_W3", "DrY1_W4", "DrY1_W5", "DrY1_W6",
              "DrY2_F1", "DrY2_F2", "DrY2_F3", "DrY2_F4", "DrY2_F5", "DrY2_F6",
              "DrY2_N1", "DrY2_N2", "DrY2_N3", "DrY2_N4", "DrY2_N5", "DrY2_N6",
              "DrY2_W1", "DrY2_W2", "DrY2_W3", "DrY2_W4", "DrY2_W5", "DrY2_W6",
              "GrY2_F1", "GrY2_F2", "GrY2_F3", "GrY2_F4", "GrY2_F5", "GrY2_F6",
              "GrY2_N1", "GrY2_N2", "GrY2_N3", "GrY2_N4", "GrY2_N5", "GrY2_N6",
              "GrY2_W1", "GrY2_W2", "GrY2_W3", "GrY2_W4", "GrY2_W5", "GrY2_W6"
            )
samples_MY <- c("M2.1", "M2.2", "M2.3", "M2.4", "M2.5",
              "M3.1", "M3.2", "M3.3", "M3.4", "M3.5",
              "M4.1", "M4.2", "M4.3", "M4.4", "M4.5",
              "M5.1", "M5.2", "M5.3", "M5.4", "M5.5",
              "M6.1", "M6.2", "M6.3", "M6.4", "M6.5",
              "M7.1", "M7.2", "M7.3", "M7.4", "M7.5",
              "C4.1", "C4.2", "C4.3", "C4.4", "C4.5",
              "C5.1", "C5.2", "C5.3", "C5.4", "C5.5",
              "C6.1", "C6.2", "C6.3", "C6.4", "C6.5",
              "C7.1", "C7.2", "C7.3", "C7.4", "C7.5",
              "C8.1", "C8.2", "C8.3", "C8.4", "C8.5",
              "C9.1", "C9.2", "C9.3", "C9.4", "C9.5",
              "D4.1","D4.2","D4.3","D4.4","D4.5",
              "D5.1","D5.2","D5.3","D5.4","D5.5",
              "D6.1","D6.2","D6.3","D6.4","D6.5",
              "D7.1","D7.2","D7.3","D7.4","D7.5",
              "D8.1","D8.2","D8.3","D8.4","D8.5",
              "D9.1","D9.2","D9.3","D9.4","D9.5",
              "F4.1","F4.2","F4.3","F4.4","F4.5",
              "F5.1","F5.2","F5.3","F5.4","F5.5",
              "F6.1","F6.2","F6.3","F6.4","F6.5",
              "F7.1","F7.2","F7.3","F7.4","F7.5",
              "F8.1","F8.2","F8.3","F8.4","F8.5",
              "F9.1","F9.2","F9.3","F9.4","F9.5",
              "A1.1","A1.2","A1.3","A1.4","A1.5",
              "A2.1","A2.2","A2.3","A2.4","A2.5",
              "A3.1","A3.2","A3.3","A3.4","A3.5",
              "A4.1","A4.2","A4.3","A4.4","A4.5",
              "A5.1","A5.2","A5.3","A5.4","A5.5",
              "A6.1","A6.2","A6.3","A6.4","A6.5"
            )
samples <- c(samples_IN, samples_MY, samples_KE)
samples_AM <- c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5",
              "M2.1", "M2.2", "M2.3", "M2.4", "M2.5",
              "M3.1", "M3.2", "M3.3", "M3.4", "M3.5",
              "M4.1", "M4.2", "M4.3", "M4.4", "M4.5",
              "M5.1", "M5.2", "M5.3", "M5.4", "M5.5",
              "M6.1", "M6.2", "M6.3", "M6.4", "M6.5",
              "M7.1", "M7.2", "M7.3", "M7.4", "M7.5",
              "AmAi01", "AmAi02", "AmAi03", "AmAi04", "AmAi05",
              "AmAi06", "AmAi07", "AmAi08", "AmAi09", "AmAi10",
              "AmIu01", "AmIu02", "AmIu03", "AmIu04", "AmIu05",
              "DrY2_W1", "DrY2_W2", "DrY2_W3", "DrY2_W4", "DrY2_W5", "DrY2_W6",
              "GrY2_F1", "GrY2_F2", "GrY2_F3", "GrY2_F4", "GrY2_F5", "GrY2_F6",
              "GrY2_N1", "GrY2_N2", "GrY2_N3", "GrY2_N4", "GrY2_N5", "GrY2_N6",
              "GrY2_W1", "GrY2_W2", "GrY2_W3", "GrY2_W4", "GrY2_W5", "GrY2_W6"
            )
samples_AC <- c("C1.1", "C1.2", "C1.3", "C1.4", "C1.5",
              "C2.1", "C2.2", "C2.3", "C2.4", "C2.5",
              "C3.1", "C3.2", "C3.3", "C3.4", "C3.5",
              "C4.1", "C4.2", "C4.3", "C4.4", "C4.5",
              "C5.1", "C5.2", "C5.3", "C5.4", "C5.5",
              "C6.1", "C6.2", "C6.3", "C6.4", "C6.5",
              "C7.1", "C7.2", "C7.3", "C7.4", "C7.5",
              "C8.1", "C8.2", "C8.3", "C8.4", "C8.5",
              "C9.1", "C9.2", "C9.3", "C9.4", "C9.5",
              "AcKn01", "AcKn02", "AcKn03", "AcKn04", "AcKn05",
              "AcKn06", "AcKn07", "AcKn08", "AcKn09", "AcKn10",
              "AcCh01", "AcCh02", "AcCh03", "AcCh04", "AcCh05",
              "AcCh06", "AcCh07", "AcCh08", "AcCh09", "AcCh10"
            )
samples_AD <- c("D1.1","D1.2","D1.3","D1.4","D1.5",
              "D2.1","D2.2","D2.3","D2.4","D2.5",
              "D3.1","D3.2","D3.3","D3.4","D3.5",
              "D4.1","D4.2","D4.3","D4.4","D4.5",
              "D5.1","D5.2","D5.3","D5.4","D5.5",
              "D6.1","D6.2","D6.3","D6.4","D6.5",
              "D7.1","D7.2","D7.3","D7.4","D7.5",
              "D8.1","D8.2","D8.3","D8.4","D8.5",
              "D9.1","D9.2","D9.3","D9.4","D9.5"
            )
samples_AF <- c("F1.1","F1.2","F1.3","F1.4","F1.5",
              "F2.1","F2.2","F2.3","F2.4","F2.5",
              "F3.1","F3.2","F3.3","F3.4","F3.5",
              "F4.1","F4.2","F4.3","F4.4","F4.5",
              "F5.1","F5.2","F5.3","F5.4","F5.5",
              "F6.1","F6.2","F6.3","F6.4","F6.5",
              "F7.1","F7.2","F7.3","F7.4","F7.5",
              "F8.1","F8.2","F8.3","F8.4","F8.5",
              "F9.1","F9.2","F9.3","F9.4","F9.5"
            )
samples_AA <- c("A1.1","A1.2","A1.3","A1.4","A1.5",
              "A2.1","A2.2","A2.3","A2.4","A2.5",
              "A3.1","A3.2","A3.3","A3.4","A3.5",
              "A4.1","A4.2","A4.3","A4.4","A4.5",
              "A5.1","A5.2","A5.3","A5.4","A5.5",
              "A6.1","A6.2","A6.3","A6.4","A6.5"
            )
samples_IN_MY <- c("M1.1", "M1.2", "M1.3", "M1.4", "M1.5",
              "M2.1", "M2.2", "M2.3", "M2.4", "M2.5",
              "M3.1", "M3.2", "M3.3", "M3.4", "M3.5",
              "M4.1", "M4.2", "M4.3", "M4.4", "M4.5",
              "M5.1", "M5.2", "M5.3", "M5.4", "M5.5",
              "M6.1", "M6.2", "M6.3", "M6.4", "M6.5",
              "M7.1", "M7.2", "M7.3", "M7.4", "M7.5",
              "C1.1", "C1.2", "C1.3", "C1.4", "C1.5",
              "C2.1", "C2.2", "C2.3", "C2.4", "C2.5",
              "C3.1", "C3.2", "C3.3", "C3.4", "C3.5",
              "C4.1", "C4.2", "C4.3", "C4.4", "C4.5",
              "C5.1", "C5.2", "C5.3", "C5.4", "C5.5",
              "C6.1", "C6.2", "C6.3", "C6.4", "C6.5",
              "C7.1", "C7.2", "C7.3", "C7.4", "C7.5",
              "C8.1", "C8.2", "C8.3", "C8.4", "C8.5",
              "C9.1", "C9.2", "C9.3", "C9.4", "C9.5",
              "D1.1","D1.2","D1.3","D1.4","D1.5",
              "D2.1","D2.2","D2.3","D2.4","D2.5",
              "D3.1","D3.2","D3.3","D3.4","D3.5",
              "D4.1","D4.2","D4.3","D4.4","D4.5",
              "D5.1","D5.2","D5.3","D5.4","D5.5",
              "D6.1","D6.2","D6.3","D6.4","D6.5",
              "D7.1","D7.2","D7.3","D7.4","D7.5",
              "D8.1","D8.2","D8.3","D8.4","D8.5",
              "D9.1","D9.2","D9.3","D9.4","D9.5",
              "F1.1","F1.2","F1.3","F1.4","F1.5",
              "F2.1","F2.2","F2.3","F2.4","F2.5",
              "F3.1","F3.2","F3.3","F3.4","F3.5",
              "F4.1","F4.2","F4.3","F4.4","F4.5",
              "F5.1","F5.2","F5.3","F5.4","F5.5",
              "F6.1","F6.2","F6.3","F6.4","F6.5",
              "F7.1","F7.2","F7.3","F7.4","F7.5",
              "F8.1","F8.2","F8.3","F8.4","F8.5",
              "F9.1","F9.2","F9.3","F9.4","F9.5",
              "A1.1","A1.2","A1.3","A1.4","A1.5",
              "A2.1","A2.2","A2.3","A2.4","A2.5",
              "A3.1","A3.2","A3.3","A3.4","A3.5",
              "A4.1","A4.2","A4.3","A4.4","A4.5",
              "A5.1","A5.2","A5.3","A5.4","A5.5",
              "A6.1","A6.2","A6.3","A6.4","A6.5"
            )
location_order <- c("Malaysia", "India", "Switzerland", "Japan")
location_order_color <- c("Malaysia" = "#1f78b4", "India" = "#33a02c", "Switzerland" = "#e31a1c", "Japan" = "#fb9a99")
host_order <- c("Apis mellifera", "Apis cerana", "Apis dorsata", "Apis florea", "Apis andreniformis")
host_order_color <- c("Apis mellifera" = brewer.pal(9, "Pastel1")[2], "Apis cerana" = brewer.pal(9, "Pastel1")[1], "Apis dorsata" = brewer.pal(9, "Pastel1")[4], "Apis florea" = brewer.pal(9, "Pastel1")[3], "Apis andreniformis" = brewer.pal(9, "Pastel1")[5])
host_order_color_dark <- c("Apis mellifera" = brewer.pal(9, "Set1")[2], "Apis cerana" = brewer.pal(9, "Set1")[1], "Apis dorsata" = brewer.pal(9, "Set1")[4], "Apis florea" = brewer.pal(9, "Set1")[3], "Apis andreniformis" = brewer.pal(9, "Set1")[5])
phylotypes <- c("firm4", "firm5", "api", "bifido", "bom", "com", "bapis", "fper", "lkun", "snod", "gilli")
phylotypes_heatmap_order <- c("snod", "gilli", "firm4", "firm5", "bifido", "bapis", "fper", "api", "lkun", "bom", "com")
sdps <- c('firm4_1', 'firm4_2',
          'firm5_1', 'firm5_2', 'firm5_3', 'firm5_4', 'firm5_7', 'firm5_bombus',
          # 'bifido_1', 'bifido_2', 'bifido_bombus',
          'bifido_1.1', 'bifido_1.2', 'bifido_1.3', 'bifido_1.4', 'bifido_1.5', 'bifido_2', 'bifido_1_cerana', 'bifido_bombus',
          'api_1', 'api_apis_dorsa', 'api_bombus',
          'bom_1', 'bom_apis_melli', 'bom_bombus',
          'com_1', 'com_drosophila', 'com_monarch',
          'bapis',
          'fper_1',
          'lkun',
          'snod_1', 'snod_2', 'snod_bombus',
          'gilli_1', 'gilli_2', 'gilli_3', 'gilli_4', 'gilli_5', 'gilli_6',
          'gilli_apis_andre', 'gilli_apis_dorsa', 'gilli_bombus')
# Data_dir <- "04_CoreCov_211018_Medgenome_india_samples"
# species <- c('s__Bombilactobacillus mellis', 's__Lactobacillus panisapium', 's__', 's__Gilliamella apicola_E', 's__Bombilactobacillus mellifer', 's__Lactobacillus apis', 's__Snodgrassella alvi', 's__Lactobacillus melliventris', 's__Lactobacillus helsingborgensis', 's__Frischella perrara', 's__Enterobacter hormaechei_A', 's__Apibacter sp002964915', 's__Snodgrassella alvi_E', 's__Frischella japonica', 's__Gilliamella apicola_F', 's__Spiroplasma melliferum', 's__Bartonella apis', 's__Apibacter adventoris', 's__Apilactobacillus kunkeei_A', 's__Hafnia paralvei', 's__Pantoea vagans', 's__Gilliamella apicola_K', 's__Gilliamella apicola', 's__Snodgrassella alvi_G', 's__Bifidobacterium indicum', 's__Gilliamella apicola_N', 's__Klebsiella variicola', 's__Commensalibacter sp003202795', 's__Gilliamella apicola_Q')
genera <- c(
    "g__Lactobacillus",
    "g__Bifidobacterium",
    "g__Bombilactobacillus",
    "g__Snodgrassella",
    "g__Gilliamella",
    "g__Apibacter",
    "g__Bartonella",
    "g__Frischella",
    "g__Commensalibacter",
    "g__Apilactobacillus",
    "g__Bombella",
    "g__Dysgonomonas",
    "g__Enterobacter",
    "g__Pectinatus",
    "g__Spiroplasma",
    "g__Zymobacter",
    "g__Entomomonas",
    "g__Saezia",
    "g__Parolsenella",
    "g__WRHT01",
    "g__"
  )
phy_group_dict = c("firm4" = "g__Bombilactobacillus",
            "g__Bombilactobacillus_outgroup" = "g__Bombilactobacillus",
            "firm5" = "g__Lactobacillus",
            "lacto" = "g__Lactobacillus",
            "g__Lactobacillus_outgroup" = "g__Lactobacillus",
            "bifido" = "g__Bifidobacterium",
            "g__Bifidobacterium_outgroup" = "g__Bifidobacterium",
            "gilli" = "g__Gilliamella",
            "entero" = "g__Gilliamella",
            "g__Gilliamella_outgroup" = "g__Gilliamella",
            "fper" = "g__Frischella",
            "g__Frischella_outgroup" = "g__Frischella",
            "snod" = "g__Snodgrassella",
            "g__Snodgrassella_outgroup" = "g__Snodgrassella",
            "bapis" = "g__Bartonella",
            "g__Bartonella_outgroup" = "g__Bartonella",
            # "" = "g__Enterobacter",
            "g__Enterobacter_outgroup" = "g__Enterobacter",
            # "" = "g__",
            # "" = "g__Pectinatus",
            "g__Pectinatus_outgroup" = "g__Pectinatus",
            "api" = "g__Apibacter",
            "g__Apibacter_outgroup" = "g__Apibacter",
            # "" = "g__Dysgonomonas",
            "g__Dysgonomonas_outgroup" = "g__Dysgonomonas",
            # "" = "g__Spiroplasma",
            "g__Spiroplasma_outgroup" = "g__Spiroplasma",
            # "" = "g__Zymobacter",
            "g__Zymobacter_outgroup" = "g__Zymobacter",
            # "" = "g__Entomomonas",
            "g__Entomomonas_outgroup" = "g__Entomomonas",
            # "" = "g__Saezia",
            "g__Saezia_outgroup" = "g__Saezia",
            # "" = "g__Parolsenella",
            "g__Parolsenella_outgroup" = "g__Parolsenella",
            # "" = "g__WRHT01",
            "g__WRHT01_outgroup" = "g__WRHT01",
            "com" = "g__Commensalibacter",
            "g__Commensalibacter_outgroup" = "g__Commensalibacter",
            "lkun" = "g__Apilactobacillus",
            "g__Apilactobacillus_outgroup" = "g__Apilactobacillus",
            "bom" = "g__Bombella",
            "g__Bombella_outgroup" = "g__Bombella"
          )
genusColors <- list("g__Bombilactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[1],
                    "g__Lactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[4],
                    "g__Bifidobacterium" = brewer.pal(11, "Spectral")[3],
                    "g__Gilliamella" = brewer.pal(11, "Spectral")[11],
                    "g__Frischella" = brewer.pal(11, "Spectral")[8],
                    "g__Bartonella" = brewer.pal(11, "Spectral")[7],
                    "g__Snodgrassella" = brewer.pal(11, "Spectral")[10],
                    "g__Apibacter" = brewer.pal(11, "Spectral")[4],
                    "g__Commensalibacter" = brewer.pal(11, "Spectral")[6],
                    "g__Bombella" = brewer.pal(11, "Spectral")[5],
                    "g__Apilactobacillus" = brewer.pal(11, "Spectral")[9],
                    "g__Dysgonomonas" = brewer.pal(11, "Spectral")[2],
                    "g__Spiroplasma" = brewer.pal(8, "Set1")[8],
                    "g__WRHT01" = brewer.pal(8, "Dark2")[3],
                    "g__Pectinatus" = brewer.pal(8, "Dark2")[1],
                    "g__Enterobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[1],
                    "g__Zymobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[2],
                    "g__Entomomonas"= head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
                    "g__Saezia" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
                    "g__Parolsenella" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[8],
                    "g__" = "#000000"
)
genusColors_char <- c("g__Bombilactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[1],
                    "g__Lactobacillus" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(10), -1)[4],
                    "g__Bifidobacterium" = brewer.pal(11, "Spectral")[3],
                    "g__Gilliamella" = brewer.pal(11, "Spectral")[11],
                    "g__Frischella" = brewer.pal(11, "Spectral")[8],
                    "g__Bartonella" = brewer.pal(11, "Spectral")[7],
                    "g__Snodgrassella" = brewer.pal(11, "Spectral")[10],
                    "g__Apibacter" = brewer.pal(11, "Spectral")[4],
                    "g__Commensalibacter" = brewer.pal(11, "Spectral")[6],
                    "g__Bombella" = brewer.pal(11, "Spectral")[5],
                    "g__Apilactobacillus" = brewer.pal(11, "Spectral")[9],
                    "g__Dysgonomonas" = brewer.pal(11, "Spectral")[2],
                    "g__Spiroplasma" = brewer.pal(8, "Set1")[8],
                    "g__WRHT01" = brewer.pal(8, "Dark2")[3],
                    "g__Pectinatus" = brewer.pal(8, "Dark2")[1],
                    "g__Enterobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[1],
                    "g__Zymobacter" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[2],
                    "g__Entomomonas"= head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
                    "g__Saezia" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
                    "g__Parolsenella" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[8],
                    "g__Klebsiella" = "#000000",
                    "g__Hafnia" = "#000000",
                    "g__JAATFO01" = "#000000",
                    "g__Floricoccus" = "#000000",
                    "g__Pantoea" = "#000000",
                    "g__" = "#000000"
)

families <- c("f__Lactobacillaceae", "f__Bifidobacteriaceae", "f__Enterobacteriaceae", "f__Neisseriaceae", "f__Rhizobiaceae_A", "f__Selenomonadaceae", "f__Weeksellaceae", "f__Dysgonomonadaceae", "f__Mycoplasmataceae", "f__Halomonadaceae", "f__Pseudomonadaceae", "f__Burkholderiaceae", "f__Atopobiaceae", "f__Desulfovibrionaceae", "f__Acetobacteraceae", "f__", "f__Streptococcaceae")
familyColors <- list(
  "f__Lactobacillaceae" = brewer.pal(11, "Spectral")[1],
  "f__Bifidobacteriaceae" = brewer.pal(11, "Spectral")[3],
  "f__Enterobacteriaceae" = brewer.pal(11, "Spectral")[11],
  "f__Neisseriaceae" = brewer.pal(11, "Spectral")[10],
  "f__Rhizobiaceae_A" = brewer.pal(11, "Spectral")[7],
  "f__Weeksellaceae" = brewer.pal(11, "Spectral")[4],
  "f__Acetobacteraceae" = brewer.pal(11, "Spectral")[6],
  "f__Dysgonomonadaceae" = brewer.pal(11, "Spectral")[2],
  "f__Mycoplasmataceae" = brewer.pal(8, "Set1")[8],
  "f__Desulfovibrionaceae" = brewer.pal(8, "Dark2")[3],
  "f__Selenomonadaceae" = brewer.pal(8, "Dark2")[1],
  "f__Halomonadaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[2],
  "f__Pseudomonadaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[4],
  "f__Burkholderiaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[6],
  "f__Atopobiaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[8],
  "f__Streptococcaceae" = head(colorRampPalette(c(brewer.pal(11, "BrBG")[2], "#FFFFFF"))(10), -1)[9],
  "f__" = "#000000"
)
PhylotypeColors <- brewer.pal(11,"Spectral")
names(PhylotypeColors) <- phylotypes
# each color from the Spectral palatte corresponds to a phylotype
# each SDP of the phylotype gets a color made by colorRampPalette
# it falls in the sange from the color of the phylotype to #FFFFFF (white)
SDPColors <- c()
# 'firm4'
# 'firm4_1''firm4_2':2
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(3), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[1], "#FFFFFF"))(3), -1))
# 'firm5'
# 'firm5_1''firm5_2''firm5_3''firm5_4''firm5_7''firm5_bombus':6
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[2], "#FFFFFF"))(7), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[2], "#FFFFFF"))(7), -1))
# 'bifido'
# # 'bifido_1''bifido_2''bifido_bombus':3 - no
# 'bifido_1.1', 'bifido_1.2', 'bifido_1.3', 'bifido_1.4', 'bifido_1.5', 'bifido_2', 'bifido_1_cerana' 'bifido_bombus': 8
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[3], "#FFFFFF"))(9), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[3], "#FFFFFF"))(4), -1))
# 'api'
# 'api_1''api_apis_dorsa''api_bombus':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[4], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[4], "#FFFFFF"))(4), -1))
# 'bom''bom_apis_melli''bom_bombus':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[5], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[5], "#FFFFFF"))(4), -1))
# 'com'
# 'com_1''com_drosophila''com_monarch':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[6], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[6], "#FFFFFF"))(4), -1))
# 'bapis'
# 'bapis':1
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[7], "#FFFFFF"))(2), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[7], "#FFFFFF"))(2), -1))
# 'fper'
# 'fper_1':1
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[8], "#FFFFFF"))(2), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[8], "#FFFFFF"))(2), -1))
# 'lkun'
# 'lkun':1
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[9], "#FFFFFF"))(2), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[9], "#FFFFFF"))(2), -1))
# 'snod'
# 'snod_1''snod_2''snod_bombus':3
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[10], "#FFFFFF"))(4), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[10], "#FFFFFF"))(4), -1))
# 'gilli'
# 'gilli_1''gilli_2''gilli_3''gilli_4''gilli_5''gilli_6''gilli_apis_andre''gilli_apis_dorsa''gilli_bombus':9
SDPColors <- c(SDPColors, head(colorRampPalette(c(brewer.pal(11, "Spectral")[11], "#FFFFFF"))(10), -1))
# show_col(head(colorRampPalette(c(brewer.pal(11, "Spectral")[11], "#FFFFFF"))(10), -1))
names(SDPColors) <- sdps
Groups <- c("g__Bombilactobacillus",
            "g__Lactobacillus",
            "g__Bifidobacterium",
            "g__Gilliamella",
            "g__Frischella",
            "g__Snodgrassella",
            "g__Bartonella",
            "g__Enterobacter",
            # "g__",
            "g__Pectinatus",
            "g__Apibacter",
            "g__Dysgonomonas",
            "g__Spiroplasma",
            # "g__Zymobacter",
            "g__Entomomonas",
            "g__Saezia",
            "g__Parolsenella",
            "g__WRHT01",
            "g__Commensalibacter",
            "g__Apilactobacillus",
            "g__Bombella")
motus <- c("Lactobacillus", "Bifidobacterium", "Gilliamella", "Frischella", "Snodgrassella", "Bartonella", "Bombella", "Acetobacteraceae", "Dysgonomonadaceae", "Spiroplasma", "Flavobacteriaceae", "Fructobacillus", "unassigned")
motuColors <- list(
  "Lactobacillus" = brewer.pal(11, "Spectral")[1],
  "Bifidobacterium" = brewer.pal(11, "Spectral")[3],
  "Flavobacteriaceae" = brewer.pal(11, "Spectral")[4],
  "Bombella" = brewer.pal(11, "Spectral")[5],
  "Acetobacteraceae" = brewer.pal(11, "Spectral")[6],
  "Bartonella" = brewer.pal(11, "Spectral")[7],
  "Frischella" = brewer.pal(11, "Spectral")[8],
  "Gilliamella" = brewer.pal(11, "Spectral")[11],
  "Snodgrassella" = brewer.pal(11, "Spectral")[10],
  "Dysgonomonadaceae" = brewer.pal(11, "Spectral")[2],
  "Fructobacillus" = brewer.pal(11, "Spectral")[9],
  "Spiroplasma" = head(colorRampPalette(c(brewer.pal(11, "Spectral")[2], "#FFFFFF"))(10), -1)[8],
  "unassigned" = "black"
)
```

### General functions

```{r general_functions}
# function to edit the look of ggplot objects to make plots uniform
# add to any ggplot object with desired parameters modified
make_theme <- function(theme_name=theme_classic() ,max_colors=0, palettefill="Pastel1", palettecolor="Dark2", modify_guide = T,
                        setFill=TRUE, setCol=TRUE,
                        guide_nrow=2, guide_nrow_byrow=TRUE, leg_pos="top", leg_size=12,
                        x_angle=0 ,x_vj=0, x_hj=0, x_size=12,
                        y_angle=0 ,y_vj=0, y_hj=0, y_size=12){
  n_11 = c("BrBG", "PiYG", "PRGn", "PuOr", "RdBu", "RdGy", "RdYlBu", "RdYlGn", "Spectral")
  n_12 = c("Paired", "Set3")
  n_8 = c("Accent", "Dark2", "Pastel2", "Set2")
  if (palettefill %in% n_12) {
    n_f = 12
  } else {
    if (palettefill %in% n_11) {
      n_f = 11
    } else {
      if (palettefill %in% n_8) {
        n_f  = 8
      } else {
        n_f = 9
      }
    }
  }
  if (palettecolor %in% n_12) {
    n_c = 12
  } else {
    if (palettecolor %in% n_11) {
      n_c = 11
    } else {
      if (palettecolor %in% n_8) {
        n_c  = 8
      } else {
        n_c = 9
      }
    }
  }
  getFill = colorRampPalette(brewer.pal(n_f, palettefill))
  getColor = colorRampPalette(brewer.pal(n_c, palettecolor))
  theme_params <- theme(axis.text.x = element_text(angle = x_angle,
    vjust = x_vj, hjust=x_hj,
    size = x_size),
    axis.text.y = element_text(angle = y_angle,
      vjust = y_vj, hjust=y_hj,
      size = y_size),
      # axis.title.x = element_text(margin=margin(t=5)),
      # axis.title.y = element_text(margin=margin(r=10)),
      legend.position=leg_pos,
      legend.text = element_text(size=leg_size)
    )
  if (modify_guide == T) {
    guide_params <- guides(fill = guide_legend(
                                    nrow=guide_nrow,
                                    byrow=guide_nrow_byrow
                                  ),
                          col = guide_legend(
                                    nrow=guide_nrow,
                                    byrow=guide_nrow_byrow
                                  )
                    )
  my_theme <- list(
                theme_name,
                theme_params,
                guide_params
              )
  } else {
    my_theme <- list(
                  theme_name,
                  theme_params
                )
  }
  if(setFill) {
    if (n_f < max_colors) {
      my_theme <- list(
                    my_theme,
                    scale_fill_manual(values = getFill(max_colors), na.value="grey")
                  )

    } else {
      my_theme <- list(
                    my_theme,
                    scale_fill_brewer(palette=palettefill, na.value="grey")
                  )
    }
  }
  if(setCol) {
    if (n_c < max_colors) {
      my_theme <- list(
                    my_theme,
                    scale_color_manual(values = getColor(max_colors), na.value="grey")
                  )

    } else {
      my_theme <- list(
                    my_theme,
                    scale_color_brewer(palette=palettecolor, na.value="grey")
                  )
    }
  }
  return(my_theme)
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_phylotype <- function(SDP){
  # only works if sdps are written in the right format
  # phylotype_x (eg. firm5_1)
  phy = strsplit(SDP, "_")[[1]][1]
  return(phy)
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_host_from_colony <- function(colony_name){
  # only works if sdps are written in the right format
  # phylotype_x (eg. Am_xx)
  host_name = strsplit(colony_name, "_")[[1]][1]
  if (host_name == "Am"){
    return("Apis mellifera")
  }
  if (host_name == "Ac"){
    return("Apis cerana")
  }
  if (host_name == "Ad"){
    return("Apis dorsata")
  }
  if (host_name == "Af"){
    return("Apis florea")
  }
  if (host_name == "Aa"){
    return("Apis andreniformis")
  }
  return(NA)
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_sample_name <- function(magname){
  if (startsWith(magname, "MAG_")){
    paste0(head(strsplit(strsplit(magname, "MAG_")[[1]][2], "_")[[1]], -1), collapse="_")
  } else {
    strsplit(magname, "_MAG")[[1]][1]
  }
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_host_name <- function(magname){
  if (startsWith(magname, "MAG_")){
    sample_name = strsplit(magname, "MAG_")[[1]][2]
    if (grepl("Dr|Gr", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Am", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Ac", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("Apis dorsata")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("Apis florea")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
  else {
    sample_name = strsplit(magname, "_MAG")[[1]][1]
    if (grepl("Dr|Gr", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Am", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("Ac", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("Apis mellifera")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("Apis cerana")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("Apis dorsata")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("Apis florea")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_host_from_sample_name <- function(sample_name){
  if (grepl("Am", sample_name)) {
    return("Apis mellifera")
  }
  if (grepl("Ac", sample_name)) {
    return("Apis cerana")
  }
  if (grepl("M1.|M2.|M3.|M4.|M5.|M6.|M7.|M8.|M9.|DrY|GrY|AmAi|AmIu", sample_name)) {
    return("Apis mellifera")
  }
  if (grepl("C1.|C2.|C3.|C4.|C5.|C6.|C7.|C8.|C9.|AcCh|AcKn", sample_name)) {
    return("Apis cerana")
  }
  if (grepl("D1.|D2.|D3.|D4.|D5.|D6.|D7.|D8.|D9.", sample_name)) {
    return("Apis dorsata")
  }
  if (grepl("F1.|F2.|F3.|F4.|F5.|F6.|F7.|F8.|F9.", sample_name)) {
    return("Apis florea")
  }
  if (grepl("A1.|A2.|A3.|A4.|A5.|A6.|A7.|A8.|A9.", sample_name)) {
    return("Apis andreniformis")
  }
  return(NA)
}
samples
# function to be used (with Vectorize)
# to mutate in dataframes
get_colony_from_sample_name <- function(sample_name){
  if (grepl("AmAi", sample_name)) {
    return("AmAi")
  }
  if (grepl("AmIu", sample_name)) {
    return("AmIu")
  }
  if (grepl("AcKn", sample_name)) {
    return("AcKn")
  }
  if (grepl("AcCh", sample_name)) {
    return("AcCh")
  }
  if (grepl("DrY1", sample_name)) {
    return("DrY1")
  }
  if (grepl("DrY2", sample_name)) {
    return("DrY2")
  }
  if (grepl("GrY1", sample_name)) {
    return("GrY1")
  }
  if (grepl("GrY2", sample_name)) {
    return("GrY2")
  }
  return(strsplit(sample_name, "\\.")[[1]][[1]])
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_colonyid_from_sample_name <- function(sample_name){
  if (grepl("AmAi", sample_name)) {
    return("1")
  }
  if (grepl("AmIu", sample_name)) {
    return("2")
  }
  if (grepl("AcKn", sample_name)) {
    return("1")
  }
  if (grepl("AcCh", sample_name)) {
    return("2")
  }
  if (grepl("DrY1", sample_name)) {
    return("1")
  }
  if (grepl("DrY2", sample_name)) {
    return("3")
  }
  if (grepl("GrY1", sample_name)) {
    return("2")
  }
  if (grepl("GrY2", sample_name)) {
    return("4")
  }
  split_sample = strsplit(sample_name, "\\.")[[1]][[1]]
  return(strsplit(split_sample, "")[[1]][[2]])
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_location_from_sample_name <- function(sample_name){
  if (grepl("Am|Ac", sample_name)) {
    return("Japan")
  }
  if (grepl("DrY|GrY", sample_name)) {
    return("Switzerland")
  }
  if (grepl("M1.", sample_name)) {
    return("India")
  }
  if (grepl("M2.|M3.|M4.|M5.|M6.|M7.|M8.|M9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("C1.|C2.|C3.", sample_name)) {
    return("India")
  }
  if (grepl("C4.|C5.|C6.|C7.|C8.|C9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("D1.|D2.|D3.", sample_name)) {
    return("India")
  }
  if (grepl("D4.|D5.|D6.|D7.|D8.|D9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("F1.|F2.|F3.", sample_name)) {
    return("India")
  }
  if (grepl("F4.|F5.|F6.|F7.|F8.|F9.", sample_name)) {
    return("Malaysia")
  }
  if (grepl("A1.|A2.|A3.|A4.|A5.|A6.|A7.|A8.|A9.", sample_name)) {
    return("Malaysia")
  }
  return(NA)
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_origin_name <- function(magname){
  if (startsWith(magname, "MAG_")){
    sample_name = strsplit(magname, "MAG_")[[1]][2]
    if (grepl("Dr|Gr", sample_name)) {
      return("Switzerland, Engel apiary")
    }
    if (grepl("Am", sample_name)) {
      return("Japan")
    }
    if (grepl("Ac", sample_name)) {
      return("Japan")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("India")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("India")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("India")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("India")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
  else {
    sample_name = strsplit(magname, "_MAG")[[1]][1]
    if (grepl("Dr|Gr", sample_name)) {
      return("Switzerland, Engel apiary")
    }
    if (grepl("Am", sample_name)) {
      return("Japan")
    }
    if (grepl("Ac", sample_name)) {
      return("Japan")
    }
    if (grepl("M1.|M2.|M3.", sample_name)) {
      return("India")
    }
    if (grepl("C1.|C2.|C3.", sample_name)) {
      return("India")
    }
    if (grepl("D1.|D2.|D3.", sample_name)) {
      return("India")
    }
    if (grepl("F1.|F2.|F3.", sample_name)) {
      return("India")
    }
    if (grepl("A1.|A2.|A3.", sample_name)) {
      return("Apis andreniformis")
    }
    return(NA)
  }
}
# function to be used (with Vectorize)
# to mutate in dataframes
format_genome_name <- function(genome){
  MAG = strsplit(genome, ".fa")[[1]]
  return(MAG)
}
# function to be used (with Vectorize)
# to mutate in dataframes
get_sdp <- function(cluster){
  if (is.na(clusters_sdp[cluster])) {
    return(cluster)
  } else{
    return(clusters_sdp[cluster])
  }
}
# extend color definitions for a vector given a named vector
# and add to it names not already defined make it grey
extend_colors <- function(names_vec, colors_vec){
  final_list <- list()
  for (a_name in names_vec) {
    if (a_name %in% names(colors_vec)) {
      final_list[a_name] = colors_vec[a_name]
    } else {
      final_list[a_name] = "grey"
    }
  }
  return(final_list)
}
extend_colors_genera <- function(names_vec) {
  extend_colors(names_vec, genusColors)
}
extend_colors_family <- function(names_vec) {
  extend_colors(names_vec, familyColors)
}
get_only_legend <- function(plot) {
  # get tabular interpretation of plot
  plot_table <- ggplot_gtable(ggplot_build(plot))
  #  Mark only legend in plot
  legend_plot <- which(sapply(plot_table$grobs, function(x) x$name) == "guide-box")
  # extract legend
  legend <- plot_table$grobs[[legend_plot]]
  # return legend
  return(legend)
}
```

### Read metadata and parse into dataframe

Read file with metadata. Working directory set here. Check full path.

```{r parse_metadata}
path_df_meta_complete <- "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/DataCollection/221220_metadata_compiled.xlsx"
df_meta_complete <- read_xlsx(path_df_meta_complete, sheet = "Compiled", range = "A368:AH661", col_names = F)
colnames(df_meta_complete) <- read_xlsx(path_df_meta_complete, sheet = "Compiled", range = "A1:AH1", col_names = F)
df_meta_complete <- df_meta_complete %>%
                      rename(ID_long = ID) %>%
                        rename(ID = Short_name) %>%
                          mutate(Species_long = Species) %>%
                            mutate(Species = Vectorize(get_host_from_sample_name)(ID)) %>%
                            mutate(Country = Vectorize(get_location_from_sample_name)(ID))
```

## Initialization for 00a plot DNA concentration
### Load and parse data
```{r init_00a}
df_concentrations <- df_meta_complete %>%
  select(ID, Concentration) %>%
      mutate(Host = Vectorize(get_host_from_sample_name)(ID)) %>%
        mutate(Location = Vectorize(get_location_from_sample_name)(ID))
```

## Initialization for 00b plot qPCR values

### Definitions and functions
```{r 00b-functions_and_definitions}
qpcr_results <- c(
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate1/plate1-2022-11-17_115607_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate2/plate2-2022-11-17_142003_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate3/plate3-2022-11-17_161529_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate4/plate4-2022-11-17_181034_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate5/plate5-2023-02-Malaysia_samples_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate6/plate6-2023-02-Malaysia_samples_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/qPCRs/plate7/plate7-2023-02-Malaysia_samples_Results.xls"
)
```
### Load and parse data

sample C7.1 was repeated in plate 6 and 7. Consider the value from plate 7 (both values were close)

```{r 00b-load_and_parse}
qpcr_df <- data.frame()
plate_num <- 1
for(x in qpcr_results){
  # read just the rows and columns with the values and leave out metadata cells
  # Limits the cells from rows=(36 to 132) and columns=(2 to 11)
  t <- read_excel(x, sheet = 1, range = cell_limits(c(36,2), c(132,11)))
  t <- t %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H1", "H2", "H3"), paste0("Control_p_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H4", "H5", "H6"), paste0("Control_n_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H7", "H8", "H9"), paste0("Control_act_p_", plate_num), `Sample Name`)) %>%
        mutate(`Sample Name` = ifelse(`Well Position` %in% c("H10", "H11", "H12"), paste0("Control_act_n_", plate_num), `Sample Name`))
  # get only columns "Well Position", "Sample Name", "Target Name" and "CT" from the xls report
  t <- t[,c(1,3,4,8,9,10)]
  names(t)[1:4] <- c("Well", "Sample.Name","Target.Name","Ct")
  t$Ct <- as.numeric(t$Ct) # NAs will be introduced at the place of 'undetermined' values
  t$Target.Name <- factor(t$Target.Name)
  t$Sample.Name <- factor(t$Sample.Name)
  # t$sd <- factor(t$sd)
  t <- as.data.frame(t)
  # add in the plate number
  t <- cbind(t, "Plate" = as.factor(rep(plate_num, dim(t)[1])))
  # bind the previous plate values with the next before moving to the next file
  qpcr_df <- rbind(qpcr_df, t)
  plate_num = plate_num + 1
}
```

```{r 00b-intialize}
# qpcr_df %>% summarise(Sample.Name, Ct) %>% filter(grepl("Control", Sample.Name))
# qpcr_df_info %>% summarise()
qpcr_df_info <- qpcr_df %>%
          mutate(Target.Name = ifelse(Target.Name == "UV_0356/0772", "Bacteria", "Host")) %>%
                filter(!is.na(Sample.Name)) %>%
                filter(Sample.Name!="EMPTY") %>%
                filter(!(Sample.Name=="C7.1" & Plate == 6)) %>%
                filter(!is.na(`Ct Mean`)) %>%
                  mutate(Host = Vectorize(get_host_from_sample_name)(Sample.Name)) %>%
                  mutate(Host = ifelse(grepl("Control_n", Sample.Name), "-veControl", Host)) %>%
                  mutate(Host = ifelse(grepl("Control_p", Sample.Name), "+veControl", Host)) %>%
                  mutate(Host = as.factor(Host)) %>%
                  filter(Target.Name == "Bacteria") %>%
                  group_by(Sample.Name) %>%
                    summarise(`Ct Mean`, Host, `Ct SD`, Plate) %>%
                      group_by(Host) %>%
                      unique() %>%
                        summarise(Sample.Name, `Ct Mean`, `Ct SD`, Plate) %>%
                          rename(Ct_mean = `Ct Mean`) %>%
                          rename(Ct_SD = `Ct SD`)
```

To do a sanity check use `qpcr_df_info %>% summarise(n())`. 
`r qpcr_df_info %>% summarise(n())`

## Initialization for 01 plot mapping results

### Definitions and functions
### Load and parse data

```{r load_and_parse_data}
df_reads <- data.frame()
for (sample in samples) {
  if (file.exists(paste0("fastqc/raw/", sample, "_R1_fastqc.zip"))) {
    number_raw_R1 <- read.csv(unz(paste0("fastqc/raw/", sample, "_R1_fastqc.zip"), paste0(sample, "_R1_fastqc/fastqc_data.txt")), sep = "\t") %>%
                filter(.[[1]] == "Total Sequences") %>%
                  pull() %>%
                    as.integer()
  } else {
    number_raw_R1 <- NA
  }
  if (file.exists(paste0("fastqc/raw/", sample, "_R2_fastqc.zip"))) {
    number_raw_R2 <- read.csv(unz(paste0("fastqc/raw/", sample, "_R2_fastqc.zip"), paste0(sample, "_R2_fastqc/fastqc_data.txt")), sep = "\t") %>%
                filter(.[[1]] == "Total Sequences") %>%
                  pull() %>%
                    as.integer()
  } else {
    number_raw_R2 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"))) {
    number_trimmed_R1 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"), paste0(sample, "_R1_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R1 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"))) {
    number_trimmed_R2 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"), paste0(sample, "_R2_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R2 <- NA
  }
  raw_reads <- number_raw_R1 + number_raw_R2
  trimmed <- number_trimmed_R1 + number_trimmed_R2
  if (file.exists(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_flagstat.tsv"))) {
    mapped_full_db <- read.csv(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
    mapped_full_db <- NA
  }
  if (file.exists(paste0("09_MagDatabaseProfiling/SNVProfiling/Mapping/", sample, "_flagstat.tsv"))) {
    mapped_rep_db <- read.csv(paste0("09_MagDatabaseProfiling/SNVProfiling/Mapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
    mapped_rep_db <- NA
  }
  if (file.exists(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host_mapping_flagstat.tsv"))) {
    mapped_host_db <- read.csv(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host_mapping_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
   mapped_host_db <- NA
  }
  host_filtered <- trimmed - mapped_host_db
  unmapped_all <- trimmed - mapped_host_db - mapped_full_db
  if (file.exists(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host-filtered_flagstat.tsv"))) {
    mapped_full_db_host_filtered <- read.csv(paste0("09_MagDatabaseProfiling/MAGsDatabaseMapping/", sample, "_host-filtered_flagstat.tsv"), sep = "\t") %>%
                        filter(.[[3]] == "with itself and mate mapped") %>%
                          pull(1) %>%
                            as.integer()
  } else {
   mapped_full_db_host_filtered <- NA
  }
  values_to_bind <- c(sample, as.integer(c(raw_reads, trimmed, mapped_full_db, mapped_rep_db, host_filtered, mapped_full_db_host_filtered, mapped_host_db, unmapped_all)))
  df_reads <- rbind(df_reads, values_to_bind)
}
# parse this later
# read.csv(paste0("02_HostMapping/", sample, "_coverage.tsv"), sep = "\t")
```

```{r load_reads}
df_colnames <- c("Sample", "Raw", "Trimmed", "MAGs_DB", "Mapped_to_rep_MAGs", "Host_filtered_reads", "MAGs_DB_after_host_filtering", "Host_mapped", "Unmapped")
colnames(df_reads) <- df_colnames
df_reads <- df_reads %>%
              mutate(across(!c("Sample"), as.integer))
df_reads_plot <- df_reads %>%
                          mutate(sum = Host_mapped + MAGs_DB) %>%
                          mutate(perc_mags = MAGs_DB/Trimmed*100) %>%
                          mutate(perc_mags_no_host = MAGs_DB/(Trimmed-Host_mapped)*100) %>%
                          mutate(perc_unmapped_no_host = 100 - perc_mags_no_host) %>%
                          mutate(perc_host = Host_mapped/Trimmed*100) %>%
                          mutate(perc_unmapped_all = 100-perc_mags-perc_host) %>%
                          mutate(perc_mapped = MAGs_DB/Trimmed*100) %>%
                          mutate(perc_unmapped = 100-perc_mapped) %>%
                          mutate(perc_mapped_hf = MAGs_DB_after_host_filtering/host_filtered*100) %>%
                          mutate(perc_unmapped_hf = 100-perc_mapped_hf) %>%
                          mutate(microbe_to_host_ratio = MAGs_DB/Host_mapped) %>%
                            pivot_longer(!Sample, names_to = "Type", values_to = "Number") %>%
                              mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
                              mutate(Location = Vectorize(get_location_from_sample_name)(Sample))
df_reads_updated <-  df_reads %>%
  select(Sample, MAGs_DB, Host_mapped, Unmapped, Trimmed) %>%
    mutate(Host = Vectorize(get_host_from_sample_name)(Sample)) %>%
    mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
        mutate(MAGs_DB_range = ifelse(MAGs_DB >= 30e+06, "High", NA)) %>%
        mutate(MAGs_DB_range = ifelse(MAGs_DB >= 15e+06 & MAGs_DB < 30e+06, "Sufficient", MAGs_DB_range)) %>%
        mutate(MAGs_DB_range = ifelse(MAGs_DB >= 0 & MAGs_DB < 15e+06, "Low", MAGs_DB_range)) %>%
          mutate(Total_depth = ifelse(Trimmed >= 50e+06, "Target_reached", "Insufficient"))
```

## Initialization for 02 plot motus result

### Definitions and functions

```{r 02-functions_and definitions}
condense_motu <- function(motu){
  motu_condensed = strsplit(motu, " ")[[1]][1]
  return(motu_condensed)
}

extend_colors_motus <- function(names_vec){
  final_list <- list()
  for (a_name in names_vec) {
    if (a_name %in% names(motuColors)) {
      final_list[a_name] = motuColors[a_name]
    } else {
      final_list[a_name] = "grey"
    }
  }
  return(final_list)
}
```

### Load and parse data

```{r load_and_parse_data}
df_motus_raw <- read.csv("02_motus_profile/samples_merged.motus", sep = "\t", skip = 2, stringsAsFactors=FALSE)
```

```{r 02-init}
colnames(df_motus_raw)[1] <- "taxonomy"
df_motus_raw <- df_motus_raw %>%
              mutate(across(!c("taxonomy"), as.numeric)) %>%
              mutate(sum_ab = rowSums(across(where(is.numeric)))) %>%
                filter(sum_ab > 0) %>%
                  select(!sum_ab) %>%
                    column_to_rownames("taxonomy")

df_motus <- as.data.frame(t(df_motus_raw)) %>%
              rownames_to_column("Sample") %>%
              pivot_longer(!Sample, names_to = "motu", values_to = "rel_ab") %>%
                group_by(Sample, motu) %>%
                  mutate(Present = ifelse(rel_ab > 0, 1, 0)) %>%
                    group_by(motu) %>%
                     mutate(Prevalence_num = sum(Present)) %>%
                      mutate(Prevalence = mean(Present))

df_motus_combined <- left_join(df_motus, df_meta_complete, by = c("Sample" = "ID"))  %>%
                      group_by(Species, motu) %>%
                        mutate(Present_host = ifelse(rel_ab > 0, 1, 0)) %>%
                          group_by(Species, motu) %>%
                            mutate(Prevalence_num_host = sum(Present_host)) %>%
                            mutate(Prevalence_host = mean(Present_host)) %>%
                              mutate(mean_rel_ab_host = mean(rel_ab)) %>%
                              group_by(Sample) %>%
                              mutate(mean_rel_ab = mean(rel_ab)) %>%
                                mutate(motu_condensed = Vectorize(condense_motu)(motu))
```

## Initialization for 03 Assembly summary
### Definitions and functions
### Load and parse data

```{r 03-load_and_parse}
df_assembly <- read.csv("05_Assembly/MapToAssembly/Assembly_mapping_summary.csv", sep = ',') %>%
                left_join(df_meta_complete, by=c("Sample" = "ID"))
```

```{r 03-init}
df_assembly_plot <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    rename(Unmapped = Number.unmapped, Mapped = Number.mapped) %>%
     pivot_longer(cols = 3:4, names_to ="Type", values_to = "Number")

df_assembly_plot <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    mutate(percent_mapped = Number.mapped/Number.of.reads*100) %>%
    mutate(percent_unmapped = Number.unmapped/Number.of.reads*100) %>%
    rename(Unmapped = percent_unmapped, Mapped = percent_mapped, Unmapped_number = Number.unmapped, Mapped_number = Number.mapped) %>%
    select(Sample, Mapped, Unmapped) %>%
     pivot_longer(cols = 2:3, names_to ="Type", values_to = "Percent")

df_assembly_plot_number <- df_assembly %>%
  select(Sample, Number.of.reads, Number.mapped) %>%
    mutate(Number.unmapped = Number.of.reads - Number.mapped) %>%
    mutate(percent_mapped = Number.mapped/Number.of.reads*100) %>%
    mutate(percent_unmapped = Number.unmapped/Number.of.reads*100) %>%
    rename(Unmapped = Number.unmapped, Mapped = Number.mapped) %>%
    select(Sample, Mapped, Unmapped) %>%
     pivot_longer(cols = 2:3, names_to ="Type", values_to = "Number")
```


## Initialization for 04 xxx
### Definitions and functions
### Load and parse data

## Initialization for 05 xxx
### Definitions and functions
### Load and parse data

## Initialization for 06 plot and summarize mags

### Definitions and functions

```{r 06-functions_and_definitions}
samples_am  <- samples_AM %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_ac <- samples_AC %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_ad <- samples_AD %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_af <- samples_AF %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
samples_aa <- samples_AA %>% as.data.frame() %>% filter(. %in% samples_IN) %>% pull(.)
line_list <- c()
for (num in 1:94){
  add_line <- geom_vline(xintercept=num+0.5, size=0.1, color="black")
  # add_line <- geom_vline(xintercept=num+0.5, size=0.1, alpha=0.5, color="grey")
  line_list <- c(line_list, add_line)
}
make_col_list <- function(my_vec, my_palette = "Spectral") {
  uniq_names <- unique(my_vec)
  num_ele <- length(uniq_names)
  if (num_ele <= 3) {
    cols_used <- brewer.pal(5, my_palette)[1:num_ele]
  }
  if (num_ele <= 9 & num_ele > 3) {
    cols_used <- brewer.pal(num_ele, my_palette)
  } else {
    if (my_palette == "Pastel2") {
      cols_used <- colorRampPalette(brewer.pal(8, my_palette))(num_ele)
    } else {
      cols_used <- colorRampPalette(brewer.pal(9, my_palette))(num_ele)
    }
  }
  col_list <- c(cols_used)
  names(col_list) = uniq_names
  return(col_list)
}
my_col_fun <- colorRamp2(c(0.7, 0.8, 0.9, 0.93, 0.94, 0.95, 1), c("#c6dbef", "#9ecae1", "#2171b5", "#4d9221", "#fddbc7", "#b2182b", "#67000d"))
pretty_ani_heatmaps <- function(matrix, mags_info_df=vis_magOTUs_df, row_split = "magotus", add_values = F, value_size = 10, limit_l = 0.95, limit_h = 1, col_fun = colorRamp2(c(limit_l, limit_h), c("#9ecae1", "#08306b"))) {
  host_names <- rownames(matrix) %>% as.data.frame() %>%
                  as.data.frame() %>%
                        left_join(mags_info_df %>%
                          ungroup() %>%
                            select(ID, Host), by = c(. = "ID")) %>% pull(Host)
  magOTU_names <- rownames(matrix) %>% 
                      as.data.frame() %>%
                        left_join(mags_info_df %>%
                          ungroup() %>%
                            select(ID, Cluster), by = c(. = "ID")) %>% pull(Cluster)
  Genus_names <- rownames(matrix) %>% 
                    as.data.frame() %>%
                      left_join(mags_info_df %>%
                        ungroup() %>%
                          select(ID, Genus), by = c(. = "ID")) %>% pull(Genus)
  species_names <- rownames(matrix) %>% 
                    as.data.frame() %>%
                      left_join(mags_info_df %>%
                        ungroup() %>%
                          select(ID, Species), by = c(. = "ID")) %>% pull(Species)
  anno_host_col = HeatmapAnnotation(Host = host_names,
                                      col = list(Host = host_order_color_dark),
                                      border = TRUE
                               )
  anno_host_row = rowAnnotation(Host = host_names,
                                      col = list(Host = host_order_color_dark),
                                      border = TRUE
                                                  )
  anno_gtdb_col = HeatmapAnnotation(`GTDB Species` = species_names,
                                    col = list(`GTDB Species` = make_col_list(species_names, "Paired")),
                                    border = TRUE
                          )
  anno_gtdb_row = rowAnnotation(`GTDB Species` = species_names,
                                    col = list(`GTDB Species` = make_col_list(species_names, "Paired")),
                                    border = TRUE
                          )
  anno_magotu_col = HeatmapAnnotation(`magOTU` = magOTU_names,
                                col = list(`magOTU` = make_col_list(magOTU_names, "Spectral")
                                          ),
                                          border = TRUE
                          )
  anno_magotu_row = rowAnnotation(`magOTU` = magOTU_names,
                                col = list(`magOTU` = make_col_list(magOTU_names, "Spectral")
                                          ),
                                          border = TRUE
                          )
  anno_tax_col = HeatmapAnnotation(`GTDB Species` = species_names,
                                   `magOTU` = magOTU_names,
                                   col = list(`magOTU` = make_col_list(magOTU_names, "Spectral"),
                                              `GTDB Species` = make_col_list(species_names, "Paired")
                                             ),
                                             border = TRUE
                                   )
  anno_tax_row = rowAnnotation(`GTDB Species` = species_names,
                                   `magOTU` = magOTU_names,
                                   col = list(`magOTU` = make_col_list(magOTU_names, "Spectral"),
                                              `GTDB Species` = make_col_list(species_names, "Paired")
                                             ),
                                             border = TRUE
                                   )
  if (row_split == "magotus") {
    row_split = as.character(as.factor(magOTU_names))
    # column_split = as.numeric(as.factor(host_names))
  }
  if (add_values) {
    cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(ifelse(matrix[i,j] > 0.7, sprintf("%.2f", matrix[i, j]), "-"), x, y, gp = gpar(fontsize = value_size))
          }
    heatmap_obj = Heatmap(matrix, 
            col = col_fun,
            # col = brewer.pal(2, "Pastel1"),
            # clustering_method_columns = 'ward.D2',
            top_annotation = anno_host_col,
            right_annotation = anno_gtdb_row,
            left_annotation = anno_magotu_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 0),
            heatmap_legend_param = list(title = "ANI", color_bar = "Continuous"),
            row_split = row_split,
            row_title_rot = 0,
            cell_fun = cell_fun,
            cluster_columns = F,
            border = TRUE,
            # row_dend_reorder = T,
            cluster_rows = F
            )
  } else {
    heatmap_obj = Heatmap(matrix, 
            col = col_fun,
            # col = brewer.pal(2, "Pastel1"),
            # clustering_method_columns = 'ward.D2',
            top_annotation = anno_host_col,
            right_annotation = anno_gtdb_row,
            left_annotation = anno_magotu_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 0),
            heatmap_legend_param = list(title = "ANI", color_bar = "Continuous"),
            row_split = row_split,
            row_title_rot = 0,
            # cell_fun = cell_fun,
            cluster_columns = F,
            border = TRUE,
            # row_dend_reorder = T,
            cluster_rows = F
            )
  }
  return(heatmap_obj)
  # draw(heatmap_obj, merge_legend = TRUE)
}
get_ani_matrix <- function(genera_list = c(), mag_names = c(), magOTUs = c(), mags_info_df=vis_magOTUs_df) {
  if (length(genera_list) > 0) {
    matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% mags_info_df$ID) %>%
                    right_join(mags_info_df %>% ungroup() %>% select(ID, Cluster, Host, Species, Genus), by = c("reference" = "ID")) %>%
                      filter(Genus %in% genera_list) %>%
                      # mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      # mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      arrange(Genus, Cluster, Host, Species) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix
  }
  if (length(mag_names) > 0) {
    matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% mags_info_df$ID) %>%
                    right_join(mags_info_df %>% ungroup() %>% select(ID, Cluster, Host, Species, Genus), by = c("reference" = "ID")) %>%
                      filter(reference %in% mag_names) %>%
                      # filter(reference %in% mag_names | querry %in% mag_names) %>%
                      # mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      # mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      arrange(Genus, Cluster, Host, Species) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix  
  }
  if (length(magOTUs) > 0) {
    matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% mags_info_df$ID) %>%
                    right_join(mags_info_df %>% ungroup() %>% select(ID, Cluster, Host, Genus), by = c("reference" = "ID")) %>%
                      filter(Cluster %in% genera_list) %>%
                      # mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      # mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      arrange(Genus, Cluster, Host, Species) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix  
  }
  return(matrix)
}
```


### Load and parse data

```{r 06-load}
df_gtdbk_bac <- read.csv("06_MAG_binning/gtdbtk_out_dir/classify/gtdbtk.bac120.summary.tsv", sep = "\t")
drep_Bdb <- read.csv("06_MAG_binning/drep_results/data_tables/Bdb.csv", sep = ",")
drep_gInformation <- read.csv("06_MAG_binning/drep_results/data_tables/genomeInfo.csv", sep = ",")
drep_Sdb <- read.csv("06_MAG_binning/drep_results/data_tables/Sdb.csv", sep = ",")
drep_Widb <- read.csv("06_MAG_binning/drep_results/data_tables/Widb.csv", sep = ",")
MAGs_collated <- read.csv("06_MAG_binning/all_GenomeInfo_auto.tsv", sep = "\t")
# system("bash scripts/calc_genome_sizes.sh")
genome_lengths_df <- read.csv("/scratch/aprasad/211018_Medgenome_india_samples/Figures/Genome_sizes.csv")
reference_genomes_info <- read.csv("/scratch/aprasad/211018_Medgenome_india_samples/06_MAG_binning/all_GenomeInfo_auto.csv")
drep_N <- read.csv("06_MAG_binning/drep_results/data_tables/Ndb.csv") %>%
            mutate(reference = Vectorize(format_genome_name)(reference)) %>%
              mutate(querry = Vectorize(format_genome_name)(querry))
```


```{r 06-init}
cluster_info <- drep_Widb %>%
                  select(genome, score, cluster, cluster_members, furthest_cluster_member) %>%
                    rename(score_representative = score) %>%
                    mutate(representative_genome = Vectorize(format_genome_name)(genome)) %>%
                      select(!genome)
cluster_tax_info <- vis_magOTUs_df %>%
                      ungroup() %>%
                      select(Genus, Species, Cluster, Num_mags) %>%
                      unique() %>%
                        mutate(cluster_name = paste0(Genus, "-", Cluster))
MAGs_collated_info <- left_join(MAGs_collated, mutate(drep_gInformation, genome = Vectorize(format_genome_name)(genome)), by = c("ID"="genome"))
vis_magOTUs_df_all <- MAGs_collated_info %>%
                        group_by(Sample) %>%
                            mutate(all_quality = ifelse(completeness > 50 & contamination < 5 & N50 > 10000, "Pass", "Fail")) %>%
                            mutate(Completeness_quality = cut(completeness ,breaks=c(-1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 100, 110),
                                                            labels = c("<10", "10-20", "20-30", "30-40", "40-50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-100", "100"))
                                                          ) %>%
                              mutate(Contamination_quality = cut(contamination ,breaks=c(-1, 0, 5, 10, 100),
                                                              labels = c("0", "0-5", "5-10", ">10"))
                                                          ) %>%
                              mutate(N50_quality = cut(N50 ,breaks=c(0, 10000, 50000, 100000, 200000, 500000, 1000000, 2000000, Inf), labels = c("<10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", "1-2Mb", ">2Mb"))) %>%
                                mutate(sample = Vectorize(get_sample_name)(ID)) %>%
                                filter(grepl("MAG_", ID)) %>%
                                  mutate(Host = Vectorize(get_host_name)(ID)) # %>%
                                    # mutate(num_contigs = Vectorize(get_num_contigs_per_bin)(ID))
vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(Cluster) %>%
                            mutate(Num_mags = n()) %>%
                              mutate(Prevalence_overall = Num_mags/length(samples))

vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(Cluster, Host) %>%
                            mutate(Present = n()) %>%
                            mutate(Prevalence = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis andreniformis", Present/length(samples_aa), Prevalence)) %>%
                              # left_join(summarise(group_by(contigs_depths_df, bin), mean_coverage = mean(depth), .groups = "drop"), by = c("ID" = "bin")) %>%
                                arrange(Genus)
vis_magOTUs_df_all <- left_join(vis_magOTUs_df_all, cluster_info, by = c("Cluster" = "cluster")) %>%
                        left_join(drep_Sdb %>%
                                    mutate(ID = Vectorize(format_genome_name)(genome)) %>%
                                    select(!genome)
                        )
vis_magOTUs_df <- vis_magOTUs_df_all %>%
                    filter(all_quality == "Pass")
vis_magOTUs_df$Host <- as.factor(recode(vis_magOTUs_df$Host, Am="Apis mellifera", Ac="Apis cerana", Ad="Apis dorsata", Af="Apis florea"))
vis_magOTUs_df$Cluster <- as.factor(vis_magOTUs_df$Cluster)
vis_magOTUs_df$sample <- as.factor(vis_magOTUs_df$sample)
vis_magOTUs_df$Family <- as.factor(vis_magOTUs_df$Family)
vis_magOTUs_df$Genus <- as.factor(vis_magOTUs_df$Genus)
vis_magOTUs_df <- vis_magOTUs_df[order(vis_magOTUs_df$Genus), ]
vis_magOTUs_df_all_means <- vis_magOTUs_df_all %>%
                        group_by(Sample) %>%
                          summarise(Sample, completeness, contamination, N50, Host) %>%
                            mutate(Completeness_mean = mean(completeness)) %>%
                              mutate(Contamination_mean = mean(contamination)) %>%
                              mutate(N50_mean = mean(N50))
# vis_magOTUs_df_all_shared_cluster <- vis_magOTUs_df_all %>%
#                               group_by(Cluster) %>%
#                                 mutate(Num_hosts = n_distinct(Host)) %>%
#                                   filter(Num_hosts > 1)
reference_genomes <- reference_genomes_info %>%
                            filter(Source_database != "MAGs" & !grepl("MAG_", ID)) %>%
                            select(ID,Group_auto,Host,Species,Phylotype) %>%
                              rename(Genus = Group_auto) %>%
                                left_join(genome_lengths_df) %>%
                                  group_by(Genus) %>%
                                  mutate(length_med = median(length))
matrix <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% vis_magOTUs_df$ID) %>%
                    right_join(vis_magOTUs_df %>% ungroup() %>% select(ID, Cluster, Genus), by = c("reference" = "ID")) %>%
                      # filter(Genus %in% c("g__Bifidobacterium")) %>%
                      mutate(reference = paste0(Cluster, "_", Genus, "_", reference)) %>%
                      mutate(querry = paste0(Cluster, "_", Genus, "_", querry)) %>%
                      select(!c(Genus, Cluster)) %>%
                        pivot_wider(names_from=querry, values_from=ani) %>%
                          replace(is.na(.), 0.7) %>%
                            column_to_rownames("reference") %>%
                            select(rownames(.)) %>%
                                as.matrix
drep_dist_ordered <- matrix %>%
                      as.data.frame() %>%
                        rownames_to_column(var = "reference") %>%
                          pivot_longer(!reference, values_to = "ani", names_to = "querry") %>%
                            mutate(ani = as.numeric(ani))


drep_dist <- drep_N %>%
                  select(reference, querry, ani) %>%
                  filter(reference %in% vis_magOTUs_df$ID) %>%
                  left_join(vis_magOTUs_df %>% ungroup() %>% select(ID, Cluster, Genus), by = c("reference" = "ID"))
```

## Initialization for 07 xxx

### Definitions and functions

```{r 07-func_def}
get_genecount_matrix <- function(my_genus, upper_count = 10, combined_info = combined_info_df, mags_info = vis_magOTUs_df, only_mags = F, with_isolates = T) {
  if (with_isolates & !only_mags) {
    OG_numbers_df <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", my_genus, "/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  if (!with_isolates && only_mags) {
    OG_numbers_df <- read.csv(paste0("database/MAGs_database_Orthofinder/", my_genus,"/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  OG_numbers_df_core_isolate <- OG_numbers_df %>%
  column_to_rownames("Orthogroup") %>%
    select(!Total) %>%
    filter(if_all(is.numeric, ~ .x <= upper_count)) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column("ID") %>%
      mutate(Type = ifelse(startsWith(ID, "MAG_"), "MAG", "Isolate")) %>%
      left_join(combined_info %>%
                          ungroup() %>%
                            select(ID, Host, Cluster) %>%
                            unique(), by = "ID") %>%
      left_join(vis_magOTUs_df %>%
                          ungroup() %>%
                            select(ID, completeness) %>%
                            unique(), by = "ID") %>%
        mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>%
        mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>%
        arrange(Type, Host, Cluster, completeness) %>%
        select(!c(Type, Host, Cluster, completeness)) %>%
        column_to_rownames("ID") %>% 
        as.matrix
}

get_scp_matrix <- function(my_genus, combined_info = combined_info_df, mags_info = vis_magOTUs_df only_mags = F, with_isolates = T) {
  if (with_isolates & !only_mags) {
    OG_numbers_df <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", my_genus, "/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  if (!with_isolates && only_mags) {
    OG_numbers_df <- read.csv(paste0("database/MAGs_database_Orthofinder/", my_genus,"/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  OG_numbers_df_core_isolate <- OG_numbers_df %>%
  column_to_rownames("Orthogroup") %>%
    select(!Total) %>%
    filter(if_all(is.numeric, ~ .x <= 1)) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column("ID") %>%
      mutate(Type = ifelse(startsWith(ID, "MAG_"), "MAG", "Isolate")) %>%
      left_join(combined_info %>%
                          ungroup() %>%
                            select(ID, Host, Cluster) %>%
                            unique(), by = "ID") %>%
      left_join(vis_magOTUs_df %>%
                          ungroup() %>%
                            select(ID, completeness) %>%
                            unique(), by = "ID") %>%
        mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>%
        mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>%
        arrange(Type, Host, Cluster, completeness) %>%
        select(!c(Type, Host, Cluster, completeness)) %>%
        column_to_rownames("ID") %>% 
        as.matrix
}

get_isolate_core_matrix <- function(my_genus, combined_info = combined_info_df, mags_info = vis_magOTUs_df, only_mags = F, with_isolates = T) {
  if (with_isolates & !only_mags) {
    OG_numbers_df <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", my_genus, "/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  if (!with_isolates && only_mags) {
    OG_numbers_df <- read.csv(paste0("database/MAGs_database_Orthofinder/", my_genus,"/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  OG_numbers_df_core_isolate <- OG_numbers_df %>%
  column_to_rownames("Orthogroup") %>%
    select(!Total) %>%
    filter(if_all(is.numeric, ~ .x <= 1)) %>%
    filter(if_all(!starts_with(c("MAG")), ~ .x == 1)) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column("ID") %>%
      mutate(Type = ifelse(startsWith(ID, "MAG_"), "MAG", "Isolate")) %>%
      left_join(combined_info %>%
                          ungroup() %>%
                            select(ID, Host, Cluster) %>%
                            unique(), by = "ID") %>%
      left_join(vis_magOTUs_df %>%
                          ungroup() %>%
                            select(ID, completeness) %>%
                            unique(), by = "ID") %>%
        mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>%
        mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>%
        arrange(Type, Host, Cluster, completeness) %>%
        select(!c(Type, Host, Cluster, completeness)) %>%
        column_to_rownames("ID") %>% 
        as.matrix
}

get_half_core_mag_matrix <- function(my_genus, combined_info = combined_info_df, mags_info = vis_magOTUs_df, only_mags = F, with_isolates = T) {
  if (with_isolates & !only_mags) {
    OG_numbers_df <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", my_genus, "/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  if (!with_isolates && only_mags) {
    OG_numbers_df <- read.csv(paste0("database/MAGs_database_Orthofinder/", my_genus,"/OrthoFinder/Results_", my_genus, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  }
  dim_mags <- OG_numbers_df %>%
                select(matches("MAG_")) %>%
                dim()
  num_mags <- dim_mags[[2]]
  OG_numbers_df_half_core_mags <- OG_numbers_df %>%
              column_to_rownames("Orthogroup") %>%
                select(!Total) %>%
                  filter(if_all(is.numeric, ~ .x <= 1)) %>%
                    select(starts_with(c("MAG"))) %>%
                      mutate(half_core = ifelse(rowSums(.) > num_mags/2, "Yes", "No")) %>%
                      select(half_core) %>%
                      rownames_to_column("OG")
  OG_numbers_df_core_isolate <- OG_numbers_df %>%
  column_to_rownames("Orthogroup") %>%
    select(!Total) %>%
    filter(if_all(is.numeric, ~ .x <= 1)) %>%
    filter(if_all(!starts_with(c("MAG")), ~ .x == 1)) %>%
    rownames_to_column("OG") %>%
    left_join(OG_numbers_df_half_core_mags) %>%
      filter(half_core == "Yes") %>%
      select(!c(half_core)) %>%
      column_to_rownames("OG") %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column("ID") %>%
      mutate(Type = ifelse(startsWith(ID, "MAG_"), "MAG", "Isolate")) %>%
      left_join(combined_info %>%
                          ungroup() %>%
                            select(ID, Host, Cluster) %>%
                            unique(), by = "ID") %>%
      left_join(vis_magOTUs_df %>%
                          ungroup() %>%
                            select(ID, completeness) %>%
                            unique(), by = "ID") %>%
        mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>%
        mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>%
        arrange(Type, Host, Cluster, completeness) %>%
        select(!c(Type, Host, Cluster, completeness)) %>%
        column_to_rownames("ID") %>% 
        as.matrix
}
pretty_orthogene_heatmaps <- function(matrix, combined_info=combined_info_df, ref_info=reference_genomes_info, mags_info = vis_magOTUs_df_all, row_split = "magotus", add_values = F, value_size = 10, limit_l = 0.95, limit_h = 1, col_fun ) {
  mag_completeness <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(mags_info %>%
                          ungroup() %>%
                            select(ID, completeness) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(completeness = ifelse(is.na(completeness), 100, completeness)) %>% 
                            pull(completeness)
  ref_status <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(mags_info %>%
                          ungroup() %>%
                            select(ID, Ref_status) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(Ref_status = ifelse(is.na(Ref_status), 1, Ref_status)) %>% 
                            pull(Ref_status)
  magOTU_names <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(combined_info_df %>%
                          ungroup() %>%
                            select(ID, Cluster) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(Cluster = ifelse(is.na(Cluster), "Unknown", Cluster)) %>% 
                            pull(Cluster)
  host_names <- rownames(matrix) %>% as.data.frame() %>%
                        left_join(combined_info_df %>%
                          ungroup() %>%
                            select(ID, Host) %>%
                            unique(), by = c(. = "ID")) %>%
                            mutate(Host = ifelse(is.na(Host), "Unknown", Host)) %>% 
                            pull(Host)
  genome_types <- rownames(matrix) %>% as.data.frame() %>%
                        mutate(Type = ifelse(startsWith(., "MAG_"), "MAG", "Isolate")) %>% pull(Type)
  anno_completeness_ref_row = rowAnnotation(Completeness = mag_completeness,
                                          `Reference status` = ref_status,
                                            col = list(Completeness = colorRamp2(c(0, 50, 100), c("#d73027", "#ffffbf", "#1a9850")),
                                                      `Reference status` = make_col_list(ref_status, "Greys")
                                            ),
                                            border = TRUE
                               )
  anno_host_row = rowAnnotation(Host = host_names,
                                col = list(Host = make_col_list(host_names, "Pastel1")),
                                border = TRUE
                               )
  anno_type_host_row = rowAnnotation(Host = host_names,
                                    `Genome type` = genome_types,
                                      col = list(`Genome type` = make_col_list(genome_types, "Spectral"),
                                                Host = make_col_list(host_names, "Pastel1")
                                                ),
                                      border = TRUE
                          )
  row_split  = as.character(as.factor(magOTU_names))
  heatmap_obj = Heatmap(matrix, 
            col = col_fun,
            # col = brewer.pal(3, "Set1")[1:2],
            # clustering_method_columns = 'ward.D2',
            # top_annotation = anno_type_col,
            right_annotation = anno_completeness_ref_row,
            left_annotation = anno_type_host_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = 0),
            heatmap_legend_param = list(title = "Gene count"),
            row_split = row_split,
            row_title_rot = 0,
            row_title_gp = grid::gpar(fontsize = 10),
            cluster_columns = T,
            border = TRUE,
            cluster_rows = F
            )
  return(heatmap_obj)
  # draw(heatmap_obj, merge_legend = TRUE)
}
```

### Load and parse data

```{r 07-Load}
genera_list <- c()
for (file_name in list.files("database/MAGs_database_Orthofinder/")) {
  if (!grepl(".csv", file_name)) {
    genus_name <- file_name
  }
  if (!(genus_name %in% genera_list)) {
    genera_list <- c(genera_list, genus_name)
  }
}
all_Isolates <- c()
all_MAGs <- c()
Number_genomes_df <- data.frame()
for (group in Groups) {
  OG_numbers_df <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", group, "/OrthoFinder/Results_", group, "/Orthogroups/Orthogroups.GeneCount.tsv"), sep = "\t")
  genomes <- head(colnames(OG_numbers_df)[-1], -1)
  MAGs <- genomes[which(grepl("MAG", genomes, fixed=TRUE))]
  all_MAGs <- c(all_MAGs, MAGs)
  Isolates <- genomes[which(!grepl("MAG", genomes, fixed=TRUE))]
  all_Isolates <- c(all_Isolates, Isolates)
  Number_genomes_df <- rbind(Number_genomes_df,
                             cbind(
                                MAGs = length(MAGs),
                                Isolates = length(Isolates),
                                genomes = length(genomes),
                                Group = group
                              )
                            )
}
Number_genomes_df <- pivot_longer(Number_genomes_df, !Group, names_to = "Type", values_to = "Number") %>%
                      mutate(Number = as.numeric(Number))
ortho_mags_df <- data.frame()
ortho_mags_filt_df <- data.frame()
for (genus_name in genera_list) {
  orthofinder_filtered_summary <- read.csv(paste0("database/MAGs_database_Orthofinder/", genus_name,"_Orthogroups_filtered_summary.csv"))
  orthofinder_summary <- read.csv(paste0("database/MAGs_database_Orthofinder/", genus_name,"_Orthogroups_summary.csv"))
  ortho_mags_df <- rbind(ortho_mags_df, orthofinder_summary)
  ortho_mags_filt_df <- rbind(ortho_mags_filt_df, orthofinder_filtered_summary)
}
ortho_w_isolates_df <- data.frame()
for (group in Groups) {
  orthofinder_summary_w_isolates <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", group, "_Orthogroups_summary.csv"))
  ortho_w_isolates_df <- rbind(ortho_w_isolates_df, orthofinder_summary_w_isolates)
}

Orthogroups_summary_df <- data.frame()
for (group in Groups) {
  OG_summary <- read.csv(paste0("07_AnnotationAndPhylogenies/02_orthofinder/", group, "_Orthogroups_summary.csv"), sep = ",")
  Orthogroups_summary_df <- rbind(Orthogroups_summary_df,
                                  OG_summary
                            ) %>%
                              filter(present_in != "None")
}
ortho_w_isolates_df <- ortho_w_isolates_df %>%
                        left_join(cluster_tax_info, by = c("group" = "Genus")) %>%
ortho_mags_df <- ortho_mags_df %>%
                        left_join(cluster_tax_info, by = c("group" = "Genus")) %>%
ortho_mags_filt_df <- ortho_mags_filt_df %>%
                        left_join(cluster_tax_info, by = c("group" = "Genus")) %>%
plot_number_OGs <- ortho_w_isolates_df %>%
  mutate(Type = ifelse(status %in% c("Isolates:ScpCore ; MAGs:ScpCore0.5"), "Present in at least 50% MAGs", NA)) %>%
  mutate(Type = ifelse(status %in% c("Isolates:ScpCore ; MAGs:ScpCore"), "Present in all MAGs", Type)) %>%
  filter(!is.na(Type)) %>%
    group_by(group, Type) %>%
     summarise(Number = n())
plot_number_OGs_mags <- ortho_mags_df %>%
  mutate(Type = ifelse(status %in% c("; MAGs:ScpCore0.5"), "Present in at least 50% MAGs", NA)) %>%
  mutate(Type = ifelse(status %in% c("; MAGs:ScpCore"), "Present in all MAGs", Type)) %>%
  filter(!is.na(Type)) %>%
    group_by(group, Type) %>%
    filter(group %in% Groups) %>%
    group_by(group, Type) %>%
     summarise(Type, Number = n())
plot_number_OGs_mags %>%
                    select(group) %>% unique %>%
                    left_join(Number_genomes_df %>%
                                pivot_wider(names_from = Type, values_from = Number), by = c("group" = "Group")) %>%
                        ungroup() %>%
                        filter(group %in% Groups) %>%
                           mutate(SampleID = row_number()) %>% pull(SampleID)     
```

### Initialization for 08 plot DRAM results

```{r 08-fun_def}
pcoa_plot_dram <- function(df_pcoa, metadata, variable, color_add=F, color_list, colname_in_metadata = "ID", shape_add = F, shape_var) {
          matrix <- as.matrix(df_pcoa)
          dist <- as.dist(matrix)
          res_pcoa <- pcoa(dist)
          ev1 <- res_pcoa$vectors[,1]
          ev2 <- res_pcoa$vectors[,2]
          df_pcoa_new <- data.frame(cbind(ev1,ev2))
          df_pcoa_new$Sample <- rownames(df_pcoa_new)
          rownames(df_pcoa_new) <- NULL
          df_pcoa_new <- left_join(df_pcoa_new, metadata, by = c("Sample" = colname_in_metadata))
          perc_axis <- round(((res_pcoa$values$Relative_eig[c(1,2)])*100), digits=1)
          axis_x_title <- paste0("PCo1 (",perc_axis[1],"%)")
          axis_y_title <- paste0("PCo2 (",perc_axis[2],"%)")
          if(color_add & shape_add) {
            p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       shape = get(shape_var),
                                       colour = get(variable)))+
                geom_point(stat="identity", size=4) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable, shape = shape_var) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 7, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
          } else {
            if (color_add) {
              p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       colour = get(variable)))+
                geom_point(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme(setFill = F, setCol = F, guide_nrow = 7, leg_size = 10 ) +
                      scale_color_manual(values=color_list)
            } else {
              p <- ggplot(df_pcoa_new, aes(x = ev1,
                                       y = ev2,
                                       color = get(variable)))+
                geom_point(stat="identity", size=4, shape=19) +
                  labs(x=axis_x_title, y = axis_y_title, color = variable) +
                    make_theme( max_colors = length(unique(df_pcoa_new[, variable])), guide_nrow = 7, leg_size = 10 ) 
            }
          }
          return(p)
}
make_col_list <- function(my_vec, my_palette = "Spectral") {
  uniq_names <- unique(my_vec)
  num_ele <- length(uniq_names)
  if (num_ele <= 3) {
    diff_ele = 3 - num_ele
    num_ele = 3
    uniq_names <- c(uniq_names, rep("s__", diff_ele))
    cols_used <- brewer.pal(5, my_palette)[1:3]
  }
  if (num_ele <= 9 & num_ele > 3) {
    cols_used <- brewer.pal(num_ele, my_palette)
  } else {
    if (my_palette == "Pastel2") {
      cols_used <- colorRampPalette(brewer.pal(8, my_palette))(num_ele)
    } else {
      cols_used <- colorRampPalette(brewer.pal(9, my_palette))(num_ele)
    }
  }
  col_list <- c(cols_used)
  names(col_list) = uniq_names
  return(col_list)
}
pretty_dram_heatmap <- function(my_matrix,
                                bottom_annotation_name = "Genus",
                                col_input, clust_r = F, clust_c = T,
                                row_names_size = 10,
                                col_split_by = "Genus",
                                # or magOTU or Species
                                col_label_size = 10,
                                grid_line_col = "white") {
  my_layer_fun = function (j, i, x, y, w, h, fill) {
                grid.rect(x = x, y = y, width = w, heigh = h,
                          gp = gpar(col = grid_line_col, fill = NA)
                )
                # ind_mat = restore_matrix(j, i, x, y)
                # ind = unique(c(ind_mat[2, ], ind_mat[, 3]))
                # grid.points(x[ind], y[ind], pch = 16, size = unit(4, "mm"))

             }
  # my_cell_fun = function (j, i, x, y, w, h, fill) {
  #               grid.rect(x = x, y = y, width = w, heigh = h,
  #                         gp = gpar(col = "white", fill = NA)
  #               )
  #               grid.circle(x = x, y = y, r = abs(my_matrix[i, j]/2 * min(unit.c(w, h))),
  #                         gp = gpar(col = NA, fill = "black")
  #               )

  #            }
  genus_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus)
  species_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Species)
  magOTU_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Cluster)
  host_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                            pull(Host) %>% as.character()
  row_names = rownames(my_matrix) %>% as.data.frame() %>%
                            left_join(vis_magOTUs_df_all %>% ungroup(), by = c(. = "ID")) %>%
                              pull(Genus)
  anno_host_col = HeatmapAnnotation(Host = host_names,
                                          col = list(Host = host_order_color),
                                          border = T
                                    )
  anno_genus_col = HeatmapAnnotation(Genus = genus_names,
                                            col = list(Genus = genusColors_char
                                                      ),
                                            border = T
                                     )
  anno_magOTU_col = HeatmapAnnotation(magOTU = magOTU_names,
                                            col = list(magOTU = make_col_list(magOTU_names)
                                                      ),
                                            border = T
                                     )
  anno_species_col = HeatmapAnnotation(Species = species_names,
                                            col = list(Species = make_col_list(species_names)
                                                      ),
                                            border = T
                                     )
  anno_all_col = HeatmapAnnotation(Genus = genus_names,
                                   Species = species_names,
                                   magOTU = magOTU_names,
                                            col = list(Genus = genusColors_char,
                                                       Species = make_col_list(species_names, "Spectral"),
                                                       magOTU = make_col_list(magOTU_names, "Paired")
                                                        ),
                                            border = T
                                     )
  anno_g_s_col = HeatmapAnnotation(Genus = genus_names,
                                   Species = species_names,
                                            col = list(Genus = genusColors_char,
                                                       Species = make_col_list(species_names, "Paired")
                                                        ),
                                            border = T
                                     )
  anno_m_s_col = HeatmapAnnotation(Species = species_names,
                                   magOTU = magOTU_names,
                                            col = list(Species = make_col_list(species_names, "Spectral"),
                                                       magOTU = make_col_list(magOTU_names, "Paired")
                                                        ),
                                            border = T
                                     )
  anno_m_g_col = HeatmapAnnotation(Genus = genus_names,
                                   magOTU = magOTU_names,
                                            col = list(Genus = genusColors_char,
                                                       magOTU = make_col_list(magOTU_names, "Paired")
                                                        ),
                                            border = T
                                     )
  bottom_annotation_obj = anno_genus_col
  if (bottom_annotation_name == "Genus") {
    bottom_annotation_obj = anno_genus_col
  }
  if (bottom_annotation_name == "Species") {
    bottom_annotation_obj = anno_species_col
  }
  if (bottom_annotation_name == "magOTU") {
    bottom_annotation_obj = anno_magOTU_col
  }
  if (bottom_annotation_name == "Genus_Species") {
    bottom_annotation_obj = anno_g_s_col
  }
  if (bottom_annotation_name == "magOTU_Species") {
    bottom_annotation_obj = anno_g_s_col
  }
  if (bottom_annotation_name == "magOTU_Genus") {
    bottom_annotation_obj = anno_m_g_col
  }
  if (bottom_annotation_name == "All") {
    bottom_annotation_obj = anno_all_col
  }
  if (col_split_by == "magOTU") {
    column_split = as.character(as.factor(magOTU_names))
  }
  if (col_split_by == "Genus") {
    column_split = as.character(as.factor(genus_names))
  }
  if (col_split_by == "Species") {
    column_split = as.character(as.factor(species_names))
  }
  heatmap_obj = Heatmap(t(my_matrix), 
            # col = colorRamp2(c(limit_l, limit_h), c("#9ecae1", "#08306b")),
            col = col_input,
            # clustering_method_columns = 'ward.D2',
            top_annotation = anno_host_col,
            bottom_annotation = bottom_annotation_obj,
            # left_annotation = anno_host_row,
            column_names_gp = grid::gpar(fontsize = 0),
            row_names_gp = grid::gpar(fontsize = row_names_size),
            # heatmap_legend_param = list(title = "Number of genes", color_bar = "Continuous"),
            layer_fun = my_layer_fun,
            # cell_fun = my_cell_fun,
            cluster_columns = clust_c,
            column_split = column_split,
            column_title_rot = 90,
            column_title_gp = grid::gpar(fontsize = col_label_size),
            border = T,
            # row_dend_reorder = T,
            cluster_rows = clust_r
            )
  draw(heatmap_obj, merge_legend = TRUE)
}
logic_to_num <- function(x) {
  if (x) {
    return(1)
  }
  return(0)
}
get_matrix_from_dram <- function(dram_df, goi = genes_of_interest, convert_logic_to_num = F) {
  # use convert_logic _to_num if working with product tsv output
  if (convert_logic_to_num) {
    dram_df_mat <- dram_df %>%
                  select(!c(genome)) %>%
                              t %>%
                              as.data.frame %>%
                              rowwise() %>% 
                              mutate(across(everything(), 
                                     ~ifelse(. == "True", "1", .))) %>%
                              mutate(across(everything(), 
                                     ~ifelse(. == "False", "0", .)))
    dram_df_mat <- dram_df_mat %>%
                    mutate(across(where(is.character), as.numeric))
    colnames(dram_df_mat) <- dram_df$genome
  } else {
    dram_df_mat <- dram_df %>%
                  select(!c(gene_id, gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
    dram_df_mat <- dram_df_mat[-1,] %>%
                      mutate_if(is.character,as.numeric)
    colnames(dram_df_mat) <- dram_df$gene_id
  }
  return(dram_df_mat)
}
prepare_matrix <- function(my_matrix, subset_genes = T, genes_list = c(),
                           subset_mags = F, mags_info_df = vis_magOTUs_df_all,
                           mags_list = c(), by_genus = F,
                           genus_list = c()
                          ) {
                    # for now, get_rid of duplicates - there are ~ 120
                    # later find a better way to deal with them
                    duplicated_names <- names(my_matrix)[names(my_matrix) %>% duplicated]
                    my_matrix <- my_matrix %>%
                                    select(-one_of(duplicated_names))
                    if (by_genus) {
                      subset_mags = T
                      mags_list <- mags_info_df %>%
                                    ungroup() %>%
                                    filter(Genus %in% genus_list) %>%
                                    pull(ID)
                    }
                    if (subset_mags) {
                      my_matrix <- my_matrix %>%
                                    filter(rownames(.) %in% mags_list)
                    }
                    if (subset_genes) {
                      my_matrix <- my_matrix %>%
                                    select(any_of(a_genes_list))
                    }
                    my_matrix <- my_matrix %>%
                                  rownames_to_column("rownames") %>%
                                  left_join(vis_magOTUs_df_all %>%
                                                ungroup() %>%
                                                select(ID, Genus, Host, Cluster),
                                              by = c("rownames" = "ID")
                                  ) %>%
                                  arrange(Genus, Cluster, Host) %>%
                                  select(!c(Genus, Host, Cluster)) %>%
                                  column_to_rownames("rownames") %>%
                                  filter(rowSums(.) > 0)
}
```
### Load and parse data

```{r 08-load}
# made using scripts/count_genes_in_contigs.py
num_genes_binned <- read.csv("/scratch/aprasad/211018_Medgenome_india_samples/Figures/Number_genes_binned.csv")
metabolism_carbon <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "carbon utilization")
metabolism_transporters <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "Transporters")
metabolism_energy <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "Energy")
metabolism_nitrogen <- read_excel("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/metabolism_summary.xlsx", sheet = "Organic Nitrogen")
product_tsv <- read.csv("/scratch/aprasad/211018_Medgenome_india_samples/08_DRAM_annotations/MAGs_distill_manual/distill_all/distilled/product.tsv", sep = "\t")

metabolism_nitrogen_mat <- metabolism_nitrogen %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
# colnames(metabolism_nitrogen_mat) <- metabolism_nitrogen_mat[1,]
metabolism_nitrogen_mat <- metabolism_nitrogen_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
dist_matrix_nitrogen <- vegdist(metabolism_nitrogen_mat, method = "bray")
metabolism_energy_mat <- metabolism_energy %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
# colnames(metabolism_energy_mat) <- metabolism_energy_mat[1,]
metabolism_energy_mat <- metabolism_energy_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
dist_matrix_energy <- vegdist(metabolism_energy_mat, method = "bray")
metabolism_transporters_mat <- metabolism_transporters %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
metabolism_transporters_mat <- metabolism_transporters_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
dist_matrix_transporters <- vegdist(metabolism_transporters_mat, method = "bray")
colnames(metabolism_transporters_mat) <- metabolism_transporters_mat[1,]
metabolism_carbon_mat <- metabolism_carbon %>%
                              select(!c(gene_description, module, header, subheader)) %>%
                              t %>%
                              as.data.frame
metabolism_carbon_mat <- metabolism_carbon_mat[-1,] %>%
                                    mutate_if(is.character,as.numeric)
colnames(metabolism_carbon_mat) <- metabolism_carbon_mat[1,]
dist_matrix_carbon <- vegdist(metabolism_carbon_mat, method = "bray")
metabolism_carbon_df <- metabolism_carbon %>%
                              mutate(Type = ifelse(gene_id %in% c(genes_of_interest), "pectin", "other")) %>%
                              # select(!c(gene_description, module, header, subheader)) %>%
                              group_by(gene_id) %>%
                              summarise(subheader, across(where(is.numeric), sum)) %>% t %>% as.data.frame

colnames(metabolism_carbon_df) <- metabolism_carbon_df[1,]
metabolism_carbon_df <- metabolism_carbon_df[-1,] %>%
                                    mutate_if(is.character,as.numeric) %>%
                                    rownames_to_column("ID") %>%
                                    left_join(vis_magOTUs_df_all)

dist_matrix_product_pathway <- vegdist(product_tsv %>%
                                select(c(1:14)) %>%
                                  column_to_rownames("genome"))
dist_matrix_product_others <- vegdist(product_tsv %>%
                                select(c(1,34:52)) %>%
                                  column_to_rownames("genome") %>% mutate(across(!where(is.numeric), Vectorize(logic_to_num))) %>%
                                  filter(rowSums(.) > 0), 
                                  method = "jaccard"
                                  )
df_other <- product_tsv %>%
                  select(c(1,34:52)) %>%
                   column_to_rownames("genome") %>%
                    mutate(across(!where(is.numeric), Vectorize(logic_to_num))) %>%
                    filter(rowSums(.) > 0) %>%
                      as.matrix
```

### Definitions and functions
### Load and parse data
### Initialization for 09 xxx

### Definitions and functions
### Load and parse data
### Initialization for 10 xxx

### Definitions and functions
### Load and parse data
### Initialization for 11 xxx

### Definitions and functions
### Load and parse data
### Initialization for 12 xxx

```{r unused_code}
shorten_list <- function(covs_list_full, start = 1, end = 2) {
  covs_list_short = strsplit(covs_list_full, ",")[[1]][start:end]
  return(paste0(covs_list_short, collapse = ","))
}

trim_side_mean <- function(x, trim, type="both") {
    if (type == "both") {
        mean(x,trim)}
    else if (type == "right") {
        x <- sort(x)
        mean(x[1:floor(length(x)*(1-trim))])}
    else if (type == "left"){
        x <- sort(x)
        mean(x[max(1,floor(length(x)*trim)):length(x)])}
}




paginate_save <- function(plot, variable, plot_name, pass_nrow = 5, pass_ncol = 5, position = "top", pass_scales = "fixed"){
  for(i in 1:n_pages(plot)){
  plot +
    facet_wrap_paginate(as.formula(paste("~", variable)), ncol = pass_ncol, nrow = pass_nrow, strip.position = position, scales = pass_scales, page = i)
    ggsave(filename = paste0(strsplit(plot_name, ".pdf")[[1]][1], "_", i, '.pdf'))
  }
}

##############
# define important vectors (manually done)
##############

  
colonies <- c("M_1", "M_1", "M_1", "M_1", "M_1",
             "M_DrY2_F","M_DrY2_F","M_Ai","M_Iu",
              "C_1", "C_1", "C_1", "C_1", "C_1",
              "C_2", "C_2", "C_2", "C_2", "C_2",
              "C_3", "C_3", "C_3", "C_3", "C_3",
              "C_Ch","C_Kn",
              "D_1","D_1","D_1","D_1","D_1",
              "D_2","D_2","D_2","D_2","D_2",
              "D_3","D_3","D_3","D_3","D_3",
              "F_1","F_1","F_1","F_1","F_1",
              "F_2","F_2","F_2","F_2","F_2",
              "F_3","F_3","F_3","F_3","F_3")

colony_order <- c("M_1", "M_Iu", "M_Ai", "M_DrY2_F", "C_1", "C_2", "C_3", "C_Kn", "C_Ch", "D_1", "D_2", "D_3", "F_1", "F_2", "F_3")
location_order <- c("AIST_Am", "UT_Am", "Bee park, GKVK_Am","Les Droites_Am",
                    "NCBS campus_Ac", "Bee park, GKVK_Ac", "Chiba_Ac", "Kanagawa_Ac",
                    "Biological sciences building, IISc_Ad","House near NCBS_Ad","Naideli hostel_Ad",
                    "Bangalore outskirts_Af")

#

#
##############
# files read (manual)
##############

qpcr_results <- c(
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate1/plate1-2022-11-17_115607_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate2/plate2-2022-11-17_142003_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate3/plate3-2022-11-17_161529_Results.xls",
  "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate4/plate4-2022-11-17_181034_Results.xls"
  # "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate5/plate5-2023-01-xx_xxxxxx_Results.xls"
  # "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate6/plate6-2023-01-xx_xxxxxx_Results.xls"
  # "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20211018_aprasad_ApisCrossSpeciesAnalysis/Analysis/SampleProcessing/qPCRs/plate7/plate7-2023-01-xx_xxxxxx_Results.xls"
)



df_reads <- data.frame()
for (sample in samples) {
  if (file.exists(paste0("fastqc/raw/", sample, "_R1_fastqc.zip"))) {
    number_raw_R1 <- read.csv(unz(paste0("fastqc/raw/", sample, "_R1_fastqc.zip"), paste0(sample, "_R1_fastqc/fastqc_data.txt")), sep = "\t") %>%
                filter(.[[1]] == "Total Sequences") %>%
                  pull() %>%
                    as.integer()  
  } else {
    number_raw_R1 <- NA
  }
  if (file.exists(paste0("fastqc/raw/", sample, "_R2_fastqc.zip"))) {
    number_raw_R2 <- read.csv(unz(paste0("fastqc/raw/", sample, "_R2_fastqc.zip"), paste0(sample, "_R2_fastqc/fastqc_data.txt")), sep = "\t") %>%
                filter(.[[1]] == "Total Sequences") %>%
                  pull() %>%
                    as.integer()
  } else {
    number_raw_R2 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"))) {
    number_trimmed_R1 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R1_trim_fastqc.zip"), paste0(sample, "_R1_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R1 <- NA
  }
  if (file.exists(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"))) {
    number_trimmed_R2 <- read.csv(unz(paste0("fastqc/trim/", sample, "_R2_trim_fastqc.zip"), paste0(sample, "_R2_trim_fastqc/fastqc_data.txt")), sep = "\t") %>%
                  filter(.[[1]] == "Total Sequences") %>%
                    pull() %>%
                      as.integer()
  } else {
    number_trimmed_R2 <- NA
  }
  raw_reads <- number_raw_R1 + number_raw_R2
  trimmed <- number_trimmed_R1 + number_trimmed_R2
  if (file.exists(paste0("02_HostMapping/", sample, "_flagstat.tsv"))) {
    mapped_host_db <- read.csv(paste0("02_HostMapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                      filter(.[[3]] == "with itself and mate mapped") %>%
                        pull(1) %>%
                          as.integer()
  } else {
   mapped_host_db <- NA 
  }
  unmapped_host_db <- trimmed - mapped_host_db
  if (file.exists(paste0("04_MicrobiomeMappingDirect/", sample, "_flagstat.tsv"))) {
    mapped_mic_db <- read.csv(paste0("04_MicrobiomeMappingDirect/", sample, "_flagstat.tsv"), sep = "\t") %>%
                      filter(.[[3]] == "with itself and mate mapped") %>%
                        pull(1) %>%
                          as.integer()
  } else {
   mapped_mic_db <- NA 
  }
  unmapped_mic_db <- trimmed - mapped_mic_db
  if (file.exists(paste0("03_MicrobiomeMapping/", sample, "_flagstat.tsv"))) {
    mapped_mic_db_host_filtered <- read.csv(paste0("03_MicrobiomeMapping/", sample, "_flagstat.tsv"), sep = "\t") %>%
                      filter(.[[3]] == "with itself and mate mapped") %>%
                        pull(1) %>%
                          as.integer()
  } else {
   mapped_mic_db_host_filtered <- NA 
  }
  unmapped_sequential <- unmapped_host_db - mapped_mic_db_host_filtered
  values_to_bind <- c(sample, as.integer(c(raw_reads, trimmed, mapped_host_db, unmapped_host_db, mapped_mic_db, unmapped_mic_db, mapped_mic_db_host_filtered, unmapped_sequential)))
  df_reads <- rbind(df_reads, values_to_bind)
}
# parse this later
# read.csv(paste0("02_HostMapping/", sample, "_coverage.tsv"), sep = "\t")

##############
# analyse data and plot
##############

df_colnames <- c("Sample", "Raw", "Trimmed", "Mapped_host", "Unmapped_host", "Mapped_microbiome", "Unmapped_microbiome", "Mapped_microbiome_filtered", "Unmapped_filtered")
colnames(df_reads) <- df_colnames
df_reads <- df_reads %>%
              mutate(across(!c("Sample"), as.integer))
df_meta <- read.csv("config/Metadata_211018_Medgenome_india_samples.csv", sep = ',')
colnames(df_meta)[which(colnames(df_meta) == "ID")] <- "Sample"
df_meta$SpeciesID <- recode(df_meta$SpeciesID, "Am" = "Apis mellifera", "Ac" = "Apis cerana", "Af" = "Apis florea", "Ad" = "Apis dorsata")
df_meta <- df_meta %>%
            filter(Sample %in% samples) %>%
              arrange(match(Sample, samples))
df_meta %>% group_by(SpeciesID) %>% tally()
df_meta %>% group_by(SpeciesID, Country) %>% tally()
df_meta %>% group_by(SpeciesID, Country, Colony) %>% tally()
df_plot_reads <- pivot_longer(df_reads, !Sample, values_to = "Number", names_to = "Type") %>%
                  merge(select(df_meta, Sample, SpeciesID), by="Sample")
df_plot_reads$Number <- as.integer(df_plot_reads$Number)

df_gtdbk_bac <- read.csv("06_MAG_binning/gtdbtk_out_dir/classify/gtdbtk.bac120.summary.tsv", sep = "\t")
drep_Bdb <- read.csv("06_MAG_binning/drep_results/data_tables/Bdb.csv", sep = ",")
drep_gInformation <- read.csv("06_MAG_binning/drep_results/data_tables/genomeInfo.csv", sep = ",")
drep_Sdb <- read.csv("06_MAG_binning/drep_results/data_tables/Sdb.csv", sep = ",")
drep_Widb <- read.csv("06_MAG_binning/drep_results/data_tables/Widb.csv", sep = ",")
MAGs_collated <- read.csv("06_MAG_binning/all_GenomeInfo_auto.tsv", sep = "\t")

##############
# analyse data and plot
##############

cluster_info <- drep_Widb %>%
                  select(genome, score, cluster, cluster_members, furthest_cluster_member) %>%
                    rename(score_representative = score) %>%
                    mutate(representative_genome = Vectorize(format_genome_name)(genome)) %>%
                      select(!genome)
MAGs_collated_info <- left_join(MAGs_collated, mutate(drep_gInformation, genome = Vectorize(format_genome_name)(genome)), by = c("ID"="genome"))
# Number of MAGs coming from each host species
vis_magOTUs_df_all <- MAGs_collated_info %>%
                        group_by(Sample) %>%
                            mutate(all_quality = ifelse(completeness > 50 & contamination < 5 & N50 > 10000, "Pass", "Fail")) %>%
                            mutate(Completeness_quality = cut(completeness ,breaks=c(-1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 100, 110),
                                                            labels = c("<10", "10-20", "20-30", "30-40", "40-50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-100", "100"))
                                                          ) %>%
                              mutate(Contamination_quality = cut(contamination ,breaks=c(-1, 0, 5, 10, 100),
                                                              labels = c("0", "0-5", "5-10", ">10"))
                                                          ) %>%
                              mutate(N50_quality = cut(N50 ,breaks=c(0, 10000, 50000, 100000, 200000, 500000, 1000000, 2000000, Inf), labels = c("<10Kb", "10-50Kb", "50-100Kb", "100-200Kb", "200-500Kb", "0.5-1Mb", "1-2Mb", ">2Mb"))) %>%
                                mutate(sample = Vectorize(get_sample_name)(ID)) %>%
                                filter(grepl("MAG_", ID)) %>%
                                  mutate(Host = Vectorize(get_host_name)(ID)) # %>%
                                    # mutate(num_contigs = Vectorize(get_num_contigs_per_bin)(ID))
vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(Cluster) %>%
                            mutate(Num_mags = n()) %>%
                              mutate(Prevalence_overall = Num_mags/length(samples))
samples_am <- c(vis_magOTUs_df_all %>% filter(Host == "Apis mellifera") %>% pull(Sample) %>% unique %>% as.vector)
samples_ac <- c(vis_magOTUs_df_all %>% filter(Host == "Apis cerana") %>% pull(Sample) %>% unique %>% as.vector)
samples_ad <-c(vis_magOTUs_df_all %>% filter(Host == "Apis dorsata") %>% pull(Sample) %>% unique %>% as.vector)
samples_af <-c(vis_magOTUs_df_all %>% filter(Host == "Apis florea") %>% pull(Sample) %>% unique %>% as.vector)
vis_magOTUs_df_all <- vis_magOTUs_df_all %>%
                          group_by(Cluster, Host) %>%
                            mutate(Present = n()) %>%
                            mutate(Prevalence = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence)) %>%
                            mutate(Prevalence = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence)) %>%
                              # left_join(summarise(group_by(contigs_depths_df, bin), mean_coverage = mean(depth), .groups = "drop"), by = c("ID" = "bin")) %>%
                                arrange(Genus)
vis_magOTUs_df_all <- left_join(vis_magOTUs_df_all, cluster_info, by = c("Cluster" = "cluster")) %>%
                        left_join(drep_Sdb %>%
                                    mutate(ID = Vectorize(format_genome_name)(genome)) %>%
                                    select(!genome)
                        )
vis_magOTUs_df <- vis_magOTUs_df_all %>%
                    filter(all_quality == "Pass")
vis_magOTUs_df$Host <- as.factor(recode(vis_magOTUs_df$Host, Am="Apis mellifera", Ac="Apis cerana", Ad="Apis dorsata", Af="Apis florea"))
vis_magOTUs_df$Cluster <- as.factor(vis_magOTUs_df$Cluster)
vis_magOTUs_df$sample <- as.factor(vis_magOTUs_df$sample)
vis_magOTUs_df$Family <- as.factor(vis_magOTUs_df$Family)
vis_magOTUs_df$Genus <- as.factor(vis_magOTUs_df$Genus)
vis_magOTUs_df <- vis_magOTUs_df[order(vis_magOTUs_df$Genus), ]

corecov_data_dir <- "09_MagDatabaseProfiling/CoverageEstimation/Merged"
coords_df <- data.frame()
for (file in list.files(corecov_data_dir)) {
  if (endsWith(file, "_coord.txt")) {
    df_temp <- read.csv(paste0(corecov_data_dir, "/", file), sep = "\t", header = TRUE)
    coords_df <- rbind(coords_df, df_temp)
  }
}

coverage_df <- coords_df %>%
  rename(Cluster = magOTU) %>%
    left_join(cluster_tax_info) %>%
      mutate(Host = Vectorize(get_host_name)(Sample)) %>%
      mutate(Location = Vectorize(get_location_from_sample_name)(Sample)) %>%
        mutate(Detected = ifelse(Cov > 0, 1, 0)) %>%
          mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
          mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
          group_by(Host, Cluster) %>%
            mutate(Present = sum(Detected)) %>%
            mutate(Mean = mean(Cov)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis cerana", Present/length(samples_ac), Prevalence_host)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis dorsata", Present/length(samples_ad), Prevalence_host)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis florea", Present/length(samples_af), Prevalence_host)) %>%
            mutate(Prevalence_host = ifelse(Host=="Apis andreniformis", Present/length(samples_aa), Prevalence_host))

coverage_df_genus <- coords_df %>%
                rename(Cluster = magOTU) %>%
                  left_join(cluster_tax_info) %>%
                      mutate(Detected = ifelse(Cov > 1, 1, 0)) %>%
                        mutate(Detected = ifelse(is.na(Detected), 0, Detected)) %>%
                        mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
                        ungroup() %>%
                          group_by(Genus, Sample) %>%
                              summarise(Genus_Present = ifelse(any(Detected == 1), 1, 0), Num_mags = sum(Num_mags), Genus_cov = mean(Cov)) %>%
                               mutate(Host = Vectorize(get_host_name)(Sample)) %>%
                                group_by(Host, Genus) %>%
                                  mutate(Mean = round(trim_side_mean(Genus_cov, trim = 0.01)), 2) %>%
                                  mutate(Median = round(mean(Genus_cov)), 2) %>%
                                  mutate(covs_list = paste0(round(Genus_cov, 0), collapse = ",")) %>%
                                  mutate(covs_list = Vectorize(shorten_list)(covs_list, 1, 3)) %>%
                                  mutate(Present = sum(Genus_Present)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis mellifera", Present/length(samples_am), NA)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis cerana",Present/length(samples_ac), Prevalence_host)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis dorsata",Present/length(samples_ad), Prevalence_host)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis florea",Present/length(samples_af), Prevalence_host)) %>%
                                  mutate(Prevalence_host = ifelse(Host=="Apis andreniformis",Present/length(samples_aa), Prevalence_host))

abundance_df <- coords_df %>%
  rename(Cluster = magOTU) %>%
    left_join(cluster_tax_info %>% filter(!is.na(cluster_name))) %>%
      mutate(Host = Vectorize(get_host_name)(Sample)) %>%
        mutate(Cov = ifelse(is.na(Cov), 0, Cov)) %>%
          mutate(MedianCov = ifelse(is.na(MedianCov), 0, MedianCov)) %>%
            select(Sample, Cov, cluster_name, Host) %>%
              left_join(select(df_reads, Trimmed, Sample), by = "Sample") %>%
                  rename(NumReads = Trimmed) %>%
                    mutate(CovNorm = Cov/NumReads)

abundance_df_matrix <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          mutate(PresAb = ifelse(CovNorm > 0, 1, 0)) %>%
                          select(Sample, cluster_name, PresAb) %>%
                            pivot_wider(names_from=cluster_name, values_from=PresAb, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

abundance_df_matrix_pa <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          mutate(PresAb = ifelse(Cov > 1, 1, 0)) %>%
                          select(Sample, cluster_name, PresAb) %>%
                            pivot_wider(names_from=cluster_name, values_from=PresAb, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")

abundance_df_matrix_Cov_rel <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample") %>%
                              mutate(across(-1)*100/rowSums(across(-1)))

abundance_df_matrix_Cov <-  abundance_df %>% filter(!is.na(cluster_name)) %>%
                          select(Sample, cluster_name, Cov) %>%
                          filter(Sample %in% samples_IN) %>%
                            pivot_wider(names_from=cluster_name, values_from=Cov, values_fn = as.vector) %>% remove_rownames %>%
                            column_to_rownames(var="Sample")
```


```{r query_dataset}
vis_magOTUs_df %>%
  ungroup() %>%
  group_by(Genus) %>%
  summarise(n=n()) %>%
  arrange(n) %>% pull(Genus)

# g__Lactobacillus
# g__Bifidobacterium
# g__Gilliamella
vis_magOTUs_df %>%
  filter(Genus == "g__Dysgonomonas") %>%
  mutate(custom_name = paste0(ID, "-", Cluster)) %>%
    pull(custom_name)
# g__Bombilactobacillus
vis_magOTUs_df %>%
  filter(Genus == "g__Snodgrassella") %>%
    pull(ID)
vis_magOTUs_df %>%
  filter(Genus == "g__Frischella") %>%
    pull(ID)
vis_magOTUs_df %>%
  filter(Genus == "g__Apibacter") %>%
    pull(ID)
vis_magOTUs_df %>%
  filter(Genus == "g__Pectinatus") %>%
    pull(ID)
vis_magOTUs_df %>%
  filter(Genus == "g__Commensalibacter") %>%
    pull(ID)
```