#!/bin/bash

remote_server_nas=aiswarya@130.223.110.124:/home/aiswarya/mnt
# change this to copy from backup of backup
specifier=/

mkdir -p envs  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p scripts  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p Figures  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p logs  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p database  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p 00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log

# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}01_Trimmed /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# or
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}00_RawData /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# List raw file paths here or later, in an external file or table
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C1.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C1.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C1.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C1.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C1.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C1.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C1.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C1.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C1.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C1.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C2.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C2.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C2.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C3.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C3.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C3.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C3.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C3.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C3.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C3.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/C3.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C3.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/C3.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D1.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D2.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D2.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D2.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D2.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D2.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D2.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D2.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D2.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D2.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D2.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D3.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D3.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D3.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D3.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D3.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D3.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D3.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/D3.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D3.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/D3.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F1.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F1.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F1.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F2.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F2.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F2.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F3.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F3.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F3.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/F3.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F3.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F3.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F3.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F3.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F3.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211102_Medgenome_india_samples_resequenced/00_rawdata/F3.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.1_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.1_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.2_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.2_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.3_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.3_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.4_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.4_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.5_R1.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress /nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/211018_Medgenome_india_samples/00_rawdata/M1.5_R2.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/00_RawData  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log

mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/01_Trimmed/
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}01_Trimmed/*.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/01_Trimmed/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/fastqc/
rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}fastqc/* /scratch/aprasad/211018_Medgenome_india_samples/fastqc/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/02_HostMapping/
rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}02_HostMapping/* /scratch/aprasad/211018_Medgenome_india_samples/02_HostMapping/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/03_MicrobiomeMapping/
rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}03_MicrobiomeMapping/* /scratch/aprasad/211018_Medgenome_india_samples/03_MicrobiomeMapping/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/04_MicrobiomeMappingDirect/
rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}04_MicrobiomeMappingDirect/* /scratch/aprasad/211018_Medgenome_india_samples/04_MicrobiomeMappingDirect/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/05_Assembly/
rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}05_Assembly/* /scratch/aprasad/211018_Medgenome_india_samples/05_Assembly/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/09_MapToAssembly/
rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}09_MapToAssembly/* /scratch/aprasad/211018_Medgenome_india_samples/09_MapToAssembly/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log

mkdir -p 06_MAG_binning/gtdbtk_out_dir/
mkdir -p 06_MAG_binning/gtdbtk_out_dir/classify/
rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}/06_MAG_binning/gtdbtk_out_dir/classify/gtdbtk.bac120.summary.tsv 06_MAG_binning/gtdbtk_out_dir/classify/
mkdir -p 06_MAG_binning/drep_results
mkdir -p 06_MAG_binning/drep_results/data_tables/
rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}/06_MAG_binning/drep_results/data_tables/*.csv 06_MAG_binning/drep_results/data_tables/
mkdir -p 05_Assembly/host_unmapped
mkdir -p 05_Assembly/host_unmapped/check_assembly/
rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}/05_Assembly/host_unmapped/check_assembly/Assembly_mapping_summary.csv 05_Assembly/host_unmapped/check_assembly
mkdir -p 06_MAG_binning/contig_fates
rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}/06_MAG_binning/contig_fates/*_contig_fates.csv 06_MAG_binning/contig_fates
mkdir -p 06_MAG_binning/backmapping
rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}/06_MAG_binning/backmapping/* 06_MAG_binning/backmapping/
# for group in "g__Bombilactobacillus" "g__Lactobacillus" "g__Lactobacillus" "g__Bifidobacterium" "g__Gilliamella" "g__Gilliamella" "g__Frischella" "g__Snodgrassella" "g__Bartonella" "g__Enterobacter" "g__Pectinatus" "g__Apibacter" "g__Dysgonomonas" "g__Spiroplasma" "g__Entomomonas" "g__Saezia" "g__Parolsenella" "g__WRHT01" "g__Commensalibacter" "g__Apilactobacillus" "g__Bombella"
# do
#   mkdir -p 07_Phylogenies
#   mkdir -p 07_Phylogenies/02_orthofinder
#   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/
#   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}
#   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Comparative_Genomics_Statistics
#   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups
#   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.txt 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.txt
#   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.GeneCount.tsv 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.GeneCount.tsv
#   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Comparative_Genomics_Statistics/Statistics_Overall.tsv 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Comparative_Genomics_Statistics/Statistics_Overall.tsv
#   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}_Orthogroups_summary.csv 07_Phylogenies/02_orthofinder/${group}_Orthogroups_summary.csv
# done


rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}envs/* /scratch/aprasad/211018_Medgenome_india_samples/envs/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}scripts/* /scratch/aprasad/211018_Medgenome_india_samples/scripts/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log

# for database in ...
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/genome_db_220315_AP /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/genome_db_220315_AP*metafile* /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/4_host_db /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log


# # For KE's core coverage estimation
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/bed_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress --exclude "*_ortho_sequences*" --exclude "*OrthoFinder*" ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/faa_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/ffn_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/gff_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log

rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}/IsolateGenomeInfo.csv .
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}211018_Medgenome_india_samples_Report.Rmd /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}make_phylo_table.R /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}IsolateGenomeInfo.csv /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}Adapters-PE.fa /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}Snakefile /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}config.yaml /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}Metadata_211018_Medgenome_india_samples.csv /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}MakePhylogenies.smk /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}index_table.csv /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log

# specifier=BACKUP/
#
# mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/01_Trimmed/
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}01_Trimmed/*.fastq.gz /scratch/aprasad/211018_Medgenome_india_samples/01_Trimmed/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/fastqc/
# rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}fastqc/* /scratch/aprasad/211018_Medgenome_india_samples/fastqc/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/02_HostMapping/
# rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}02_HostMapping/* /scratch/aprasad/211018_Medgenome_india_samples/02_HostMapping/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/03_MicrobiomeMapping/
# rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}03_MicrobiomeMapping/* /scratch/aprasad/211018_Medgenome_india_samples/03_MicrobiomeMapping/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/04_MicrobiomeMappingDirect/
# rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}04_MicrobiomeMappingDirect/* /scratch/aprasad/211018_Medgenome_india_samples/04_MicrobiomeMappingDirect/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# mkdir -p /scratch/aprasad/211018_Medgenome_india_samples/05_Assembly/
# rsync -r -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}05_Assembly/* /scratch/aprasad/211018_Medgenome_india_samples/05_Assembly/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
#
# mkdir -p 06_MAG_binning/gtdbtk_out_dir/
# mkdir -p 06_MAG_binning/gtdbtk_out_dir/classify/
# rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}06_MAG_binning/gtdbtk_out_dir/classify/gtdbtk.bac120.summary.tsv 06_MAG_binning/gtdbtk_out_dir/classify/
# mkdir -p 06_MAG_binning/drep_results
# mkdir -p 06_MAG_binning/drep_results/data_tables/
# rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}06_MAG_binning/drep_results/data_tables/*.csv 06_MAG_binning/drep_results/data_tables/
# mkdir -p 05_Assembly/host_unmapped
# mkdir -p 05_Assembly/host_unmapped/check_assembly/
# rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}05_Assembly/host_unmapped/check_assembly/Assembly_mapping_summary.csv 05_Assembly/host_unmapped/check_assembly
# rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}IsolateGenomeInfo.csv .
# mkdir -p 06_MAG_binning/contig_fates
# rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}06_MAG_binning/contig_fates/*_contig_fates.csv 06_MAG_binning/contig_fates
# mkdir -p 06_MAG_binning/backmapping
# rsync -r --progress -v ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}06_MAG_binning/backmapping/* 06_MAG_binning/backmapping/
# # for group in "g__Bombilactobacillus" "g__Lactobacillus" "g__Lactobacillus" "g__Bifidobacterium" "g__Gilliamella" "g__Gilliamella" "g__Frischella" "g__Snodgrassella" "g__Bartonella" "g__Enterobacter" "g__Pectinatus" "g__Apibacter" "g__Dysgonomonas" "g__Spiroplasma" "g__Entomomonas" "g__Saezia" "g__Parolsenella" "g__WRHT01" "g__Commensalibacter" "g__Apilactobacillus" "g__Bombella"
# # do
# #   mkdir -p 07_Phylogenies
# #   mkdir -p 07_Phylogenies/02_orthofinder
# #   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/
# #   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}
# #   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Comparative_Genomics_Statistics
# #   mkdir -p 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups
# #   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.txt 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.txt
# #   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.GeneCount.tsv 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Orthogroups/Orthogroups.GeneCount.tsv
# #   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Comparative_Genomics_Statistics/Statistics_Overall.tsv 07_Phylogenies/02_orthofinder/${group}/OrthoFinder/Results_${group}/Comparative_Genomics_Statistics/Statistics_Overall.tsv
# #   rsync -r --progress -v aprasad@curnagl.dcsr.unil.ch:/scratch/aprasad/211018_Medgenome_india_samples/07_Phylogenies/02_orthofinder/${group}_Orthogroups_summary.csv 07_Phylogenies/02_orthofinder/${group}_Orthogroups_summary.csv
# # done
#
#
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}envs/* /scratch/aprasad/211018_Medgenome_india_samples/envs/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}scripts/* /scratch/aprasad/211018_Medgenome_india_samples/scripts/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
#
# # for database in ...
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/genome_db_220315_AP /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/genome_db_220315_AP*metafile* /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/4_host_db /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
#
#
# # # For KE's core coverage estimation
# # rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/bed_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# # rsync -a -v --progress --exclude "*_ortho_sequences*" --exclude "*OrthoFinder*" ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/faa_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# # rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/ffn_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# # rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}database/gff_files /scratch/aprasad/211018_Medgenome_india_samples/database/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
#
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}211018_Medgenome_india_samples_Report.Rmd /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}make_phylo_table.R /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}IsolateGenomeInfo.csv /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}Adapters-PE.fa /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}Snakefile /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}config.yaml /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}Metadata_211018_Medgenome_india_samples.csv /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}MakePhylogenies.smk /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
# rsync -a -v --progress ${remote_server_nas}/aprasad/Backups/working_dir_backup/Cluster/211018_Medgenome_india_samples/${specifier}index_table.csv /scratch/aprasad/211018_Medgenome_india_samples/  | tee -a /scratch/aprasad/211018_Medgenome_india_samples/project_reinitialization.log
