{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from itertools import combinations\n",
    "import glob\n",
    "\n",
    "orf_table = 'Ruminococcus.orfs.csv'\n",
    "pa_table = 'master_presence_absence.csv'\n",
    "genome_location = '../../../Data/Genomes/ruminococcus/*.fasta'\n",
    "\n",
    "# Dictionary of average core genome divergence for each population\n",
    "# calculated from MSA, available in phybreak directory\n",
    "pi_dict = {0.1: 0.0138272139869,\n",
    "           0.0: 0.00521122600799,\n",
    "           0.2: 0.02343396299783631}\n",
    "\n",
    "# Map of populations\n",
    "pops = pd.read_table('../../../Data/Clusters/rumino_0.000355362.txt.cluster.tab.txt', index_col=0)\n",
    "\n",
    "# Master table of all orfs\n",
    "df_seq = pd.read_csv(orf_table, index_col = 'protein_accession')\n",
    "\n",
    "# Master table of presence/absence\n",
    "df_presence_absenence = pd.read_csv(pa_table, index_col='member')['cluster']\n",
    "df_concat = df_seq.join(df_presence_absenence)\n",
    "\n",
    "# Map contigs to populations\n",
    "pop_map = pd.DataFrame(columns=['contig', 'population', 'strain'])\n",
    "for f in glob.glob(genome_location):\n",
    "    strain = '.'.join(os.path.basename(f).split('.')[0:-1])\n",
    "    population = pops.loc[strain, 'Cluster_ID']\n",
    "    for s in SeqIO.parse(f, 'fasta'):\n",
    "        temp = pd.DataFrame([[s.id,\n",
    "                              population,\n",
    "                              strain]], columns=pop_map.columns)\n",
    "        pop_map = pop_map.append(temp)\n",
    "pop_map.index = pop_map.contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Methods for calculating pi and alignments\n",
    "def align_seqs(unaligned_seqs):\n",
    "    \n",
    "    # Align sequences with muscle\n",
    "    \n",
    "    with open('temp_unalign.fasta', 'w') as outfile:\n",
    "        for i, s in enumerate(unaligned_seqs):\n",
    "            outfile.write('>{0}\\n'.format(str(i)))\n",
    "            outfile.write(s + '\\n')\n",
    "    os.system('muscle.exe -in temp_unalign.fasta -out temp_align.fasta')\n",
    "    seqs = [str(s.seq) for s in SeqIO.parse('temp_align.fasta', 'fasta')]\n",
    "    return seqs\n",
    "\n",
    "def is_recent(aligned_seqs, alpha, pi):\n",
    "    # Given aligned seqs, a significance level, and an average nucleotide divergence\n",
    "    # determines whether gene is \"recent\"\n",
    "\n",
    "    pi_obs = calculate_pi(aligned_seqs)\n",
    "    seq_len = len(aligned_seqs[0])\n",
    "    low, high = stats.binom.interval(alpha, seq_len, pi)\n",
    "    if pi_obs == 0 or pi_obs < low / seq_len:\n",
    "        recent = True\n",
    "    else:\n",
    "        recent = False\n",
    "    return recent\n",
    "    \n",
    "def count_divs(s1, s2):\n",
    "    # just counts divergences\n",
    "    d = 0\n",
    "    for b1, b2 in zip(s1, s2):\n",
    "        if b1 != b2:\n",
    "            d += 1\n",
    "    return d * 1.0 / len(s1)\n",
    "\n",
    "def calculate_pi(aligned_seqs):\n",
    "    # Calculates average nucleotide divergence\n",
    "    aves = []\n",
    "    for seq1, seq2 in combinations(aligned_seqs, 2):\n",
    "        aves.append(count_divs(seq1, seq2))\n",
    "    return(np.average(aves))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix df_concat fasta ids\n",
    "df_concat['fasta_id'] = ['_'.join(i.split('_')[0:-1]) for i in df_concat.index]\n",
    "# assign population to contigs\n",
    "for index, row in df_concat.iterrows():\n",
    "    contig = row.fasta_id\n",
    "    population = pop_map.loc[contig, 'population']\n",
    "    df_concat.loc[index, 'population'] = population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process annotations\n",
    "dfs = []\n",
    "colnames = (\"protein_accession\",\n",
    "            \"md5\",\n",
    "            \"sequence_length\",\n",
    "            \"analysis\",\n",
    "            \"signature_accession\",\n",
    "            \"signature_description\",\n",
    "            \"start\",\n",
    "            \"stop\",\n",
    "            \"score\",\n",
    "            \"status\",\n",
    "            \"date\",\n",
    "            \"interpro_accession\",\n",
    "            \"interpro_description\",\n",
    "            \"go_annotation\",\n",
    "            \"pathway_annotation\")\n",
    "\n",
    "for f in glob.glob('annotations/*.tsv'):\n",
    "    newlines = [line for line in open(f) if len(line.split('\\t')) == 15]\n",
    "    with open(f + '.filtered.tsv', 'w') as outfile:\n",
    "        outfile.writelines(newlines)\n",
    "for f in glob.glob('annotations/*filtered.tsv'):\n",
    "    temp = pd.read_table(f, header=None, names=colnames)\n",
    "    dfs.append(temp)\n",
    "    \n",
    "annotations = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "finaldf = pd.DataFrame(index=set(df_concat.cluster), columns=['Population', 'Annotation'])\n",
    "for prot_cluster, prot_df in df_concat.groupby('cluster'):\n",
    "    num_pops = len(set(prot_df.population))\n",
    "    num_strains = len(set(prot_df.fasta_id))\n",
    "    # Checks if the orf cluster only contains one population\n",
    "    if num_pops == 1:\n",
    "        pop = list(set(prot_df.population))[0]\n",
    "        pop_length = len(set(pops[pops.Cluster_ID == pop].Clonal_complex))\n",
    "        # Checks if all members of the population have this protein\n",
    "        if num_strains == pop_length:\n",
    "            aligned_seqs = align_seqs(prot_df.orf_seq)\n",
    "            \n",
    "            # Check nucleotide diversity cutoff\n",
    "            if is_recent(aligned_seqs, 0.05, pi_dict[pop]):\n",
    "                annots = annotations[annotations.protein_accession == prot_cluster]\n",
    "                finaldf.loc[prot_cluster, 'Population'] = pop\n",
    "                if set(annots.interpro_description) != set():\n",
    "                    finaldf.loc[prot_cluster, 'Annotation'] = '; '.join(set(annots.interpro_description))\n",
    "                else:\n",
    "                    finaldf.loc[prot_cluster, 'Annotation'] = 'Unannotated'\n",
    "\n",
    "finaldf = finaldf.dropna()\n",
    "finaldf = finaldf.sort_values('Population')\n",
    "finaldf.to_csv('flex_genome_results.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
