#! /usr/bin/perl -w
use strict;
use Bio::Seq;
use Bio::SeqIO;
use Getopt::Long;
use Pod::Usage;
pod2usage(-verbose => 2) if ((@ARGV==0)&&(-t STDIN));

=head1 SYNOPSIS

perl parse_speades_metagenome.pl contigs.fasta > parsed_contigs.fasta

=head1 DESCRIPTION

This script reads the contigs.fasta file, generated by the spades assembler, and produces a new file with contigs passing the filtering threshold. The minimum kmer coverage is set to 1, and parsed from the fasta header. The minimum contig length is set to 500bp, and is calculated with bioperl.

=cut

GetOptions(
    'help' => sub { pod2usage( -exitstatus => 0, -verbose => 2 ) },
);

my $input_file = shift;
my $seq_in = Bio::SeqIO->new(
                      -file => $input_file, 
                       -format => 'fasta',
                       );

while (my $seq_obj = $seq_in->next_seq()) {
    my $header = $seq_obj->id;
    my @split_head = split('_',$header);
    my $length = $split_head[3];
    my $cov = $split_head[5];
    my $seq = $seq_obj->seq();
    if ($length>500 && $cov>1) {
	print ">",$header,"\n";
	print $seq,"\n";
    }
}
